// @ve-design/web/web/dist/16491.js
import { css, html, nothing } from "lit";
import { componentReset, VeElement } from "./78928.js";
import { token } from "./97605.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
const ve_dropdown_group = [
    componentReset,
    token,
    css`
    :host {
      display: block;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-dropdown-group-title {
      margin-top: var(--ve-dropdown-group-title-margin-top);
      box-sizing: border-box;
      block-size: var(--ve-dropdown-group-title-height);
      padding-block: var(--ve-dropdown-group-title-padding-block);
      padding-inline: var(--ve-dropdown-group-title-padding-inline);
      color: var(--ve-dropdown-group-title-color);
      font-family: var(--font-sans);
      font-size: var(--ve-dropdown-group-title-font-size);
      font-weight: var(--ve-dropdown-group-title-font-weight);
      line-height: var(--ve-dropdown-group-title-line-height);
      text-align: left;
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
      pointer-events: none;
    }

    :host(:first-child) .ve-dropdown-group-title {
      margin-top: calc(
        var(--ve-dropdown-group-title-margin-top) - var(
            --ve-dropdown-padding-block
          )
      );
    }

    .ve-dropdown-group-content {
      display: flex;
      flex-direction: column;
      gap: var(--ve-dropdown-group-gap);
    }
  `
];
const VE_DROPDOWN_GROUP_TAG = 've-dropdown-group';
let dropdownGroupTitleSeed = 0;
class VeDropdownGroupElement extends VeElement {
    static styles = ve_dropdown_group;
    titleId = `ve-dropdown-group-title-${++dropdownGroupTitleSeed}`;
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    render() {
        const labelledBy = this.title ? this.titleId : nothing;
        return html`
      ${this.title ? html`<div id=${this.titleId} class="ve-dropdown-group-title" part="title">${this.title}</div>` : null}
      <div
        class="ve-dropdown-group-content"
        part="content"
        role="group"
        aria-labelledby=${labelledBy}
      >
        <slot></slot>
      </div>
    `;
    }
}
export { VE_DROPDOWN_GROUP_TAG, VeDropdownGroupElement };
