// @ve-design/web/web/dist/30129.js
import { css, html, nothing } from "lit";
import { property, query, state } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createFloatingAutoUpdate, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
import { closeIcon } from "./icons/close.js";
import { loadingProgressIcon } from "./icons/loading-progress.js";
import { chevronUpSmIcon } from "./icons/chevron-up-sm.js";
import { chevronDownSmIcon } from "./67641.js";
import { checkIcon } from "./76270.js";
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _init_value, _init_extra__init_value, _init_disabled, _init_extra__init_disabled, _init_closable, _init_extra__init_closable;
const VE_SELECT_ITEM_TAG = 've-select-item';
const VE_SELECT_OPTIONS_CHANGE_EVENT = 've-select-options-change';
const CLOSABLE_CONVERTER = {
    fromAttribute (value) {
        return null === value || 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
_dec = property(), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    converter: CLOSABLE_CONVERTER
});
class VeSelectItemElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, _init_disabled, _init_extra__init_disabled, _init_closable, _init_extra__init_closable] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "value"
            ],
            [
                _dec1,
                1,
                "disabled"
            ],
            [
                _dec2,
                1,
                "closable"
            ]
        ]));
    }
    static styles = [
        componentReset,
        css`
      :host {
        display: contents;
      }
    `
    ];
    #___private_value_1 = (()=>{
        const _value = _init_value(this, void 0);
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_disabled_2 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_2;
    }
    set disabled(_v) {
        this.#___private_disabled_2 = _v;
    }
    #___private_closable_3 = (()=>{
        const _value = _init_closable(this, true);
        _init_extra__init_closable(this);
        return _value;
    })();
    get closable() {
        return this.#___private_closable_3;
    }
    set closable(_v) {
        this.#___private_closable_3 = _v;
    }
    get label() {
        const text = this.textContent?.trim() ?? '';
        return text.length > 0 ? text : void 0 === this.value ? '' : String(this.value);
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('value') || changedProperties.has('disabled') || changedProperties.has('closable')) this.dispatchOptionsChange();
    }
    dispatchOptionsChange() {
        this.dispatchEvent(new CustomEvent(VE_SELECT_OPTIONS_CHANGE_EVENT, {
            bubbles: true,
            composed: true
        }));
    }
    render() {
        return html`<slot @slotchange=${this.dispatchOptionsChange}></slot>`;
    }
}
function ve_select_group_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_select_group_element_toPropertyKey(arg) {
    var key = ve_select_group_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_select_group_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_select_group_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_select_group_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_select_group_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_select_group_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_select_group_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_select_group_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_select_group_element_dec, ve_select_group_element_dec1, _init_label, _init_extra__init_label, ve_select_group_element_init_disabled, ve_select_group_element_init_extra_init_disabled;
const VE_SELECT_GROUP_TAG = 've-select-group';
ve_select_group_element_dec = property({
    type: String
}), ve_select_group_element_dec1 = property({
    type: Boolean,
    reflect: true
});
class VeSelectGroupElement extends VeElement {
    static{
        ({ e: [_init_label, _init_extra__init_label, ve_select_group_element_init_disabled, ve_select_group_element_init_extra_init_disabled] } = ve_select_group_element_apply_decs_2311(this, [], [
            [
                ve_select_group_element_dec,
                1,
                "label"
            ],
            [
                ve_select_group_element_dec1,
                1,
                "disabled"
            ]
        ]));
    }
    static styles = [
        componentReset,
        css`
      :host {
        display: contents;
      }
    `
    ];
    #___private_label_1 = (()=>{
        const _value = _init_label(this, '');
        _init_extra__init_label(this);
        return _value;
    })();
    get label() {
        return this.#___private_label_1;
    }
    set label(_v) {
        this.#___private_label_1 = _v;
    }
    #___private_disabled_2 = (()=>{
        const _value = ve_select_group_element_init_disabled(this, false);
        ve_select_group_element_init_extra_init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_2;
    }
    set disabled(_v) {
        this.#___private_disabled_2 = _v;
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('label') || changedProperties.has('disabled')) this.dispatchOptionsChange();
    }
    dispatchOptionsChange() {
        this.dispatchEvent(new CustomEvent(VE_SELECT_OPTIONS_CHANGE_EVENT, {
            bubbles: true,
            composed: true
        }));
    }
    render() {
        return html`<slot @slotchange=${this.dispatchOptionsChange}></slot>`;
    }
}
const veSelectTokenDefaults = css`
  :host {
    --ve-select-font-family: var(--font-sans);
    --ve-select-font-weight: var(--font-weight-normal);
    --ve-select-line-height-small: var(--line-height-body);
    --ve-select-line-height-default: var(--line-height-body);
    --ve-select-line-height-large: var(--line-height-body);

    --ve-select-font-size-small: var(--text-body-sm);
    --ve-select-font-size-default: var(--text-body);
    --ve-select-font-size-large: var(--text-body);

    --ve-select-height-small: var(--height-small);
    --ve-select-height-default: var(--height-default);
    --ve-select-height-large: var(--height-large);

    --ve-select-control-content-padding-inline: var(--space-xxs);
    --ve-select-control-multiple-tag-padding-inline-start: var(--space-xxxs);
    --ve-select-padding-inline-small: var(
      --ve-select-control-content-padding-inline
    );
    --ve-select-padding-inline-default: var(
      --ve-select-control-content-padding-inline
    );
    --ve-select-padding-inline-large: var(
      --ve-select-control-content-padding-inline
    );
    --ve-select-padding-block-small: var(--space-xxxs);
    --ve-select-padding-block-default: var(--space-xxxs);
    --ve-select-padding-block-large: var(--space-xxs);
    --ve-select-radius: var(--radius-md);
    --ve-select-border-width: var(--stroke-weight-base);
    --ve-select-control-border-width: var(--ve-select-border-width);
    --ve-select-gap: var(--space-xxxs);
    --ve-select-label-padding-inline: var(--space-xxs);
    --ve-select-label-color: var(--color-text-secondary);
    --ve-select-tag-gap: var(--space-xxxs);
    --ve-select-tag-padding-inline: calc(var(--space-xxs) - 2px);
    --ve-select-tag-height: calc(
      var(--ve-select-current-height) -
        var(--ve-select-current-control-border-width) -
        var(--ve-select-current-control-border-width) -
        var(--ve-select-current-padding-block) -
        var(--ve-select-current-padding-block)
    );
    --ve-select-tag-radius: var(--radius-sm);
    --ve-select-action-icon-size: var(--text-caption);
    --ve-select-arrow-icon-size: var(--space-s);
    --ve-select-tag-remove-size: var(--text-caption);
    --ve-select-clear-size: var(--space-s);

    --ve-select-control-background: var(--color-bg-control);
    --ve-select-control-background-hover: var(--color-bg-base);
    --ve-select-control-background-focus: var(--color-bg-base);
    --ve-select-control-background-disabled: var(--color-bg-control);
    --ve-select-control-text: var(--color-text-primary);
    --ve-select-control-text-disabled: var(--color-text-disable);
    --ve-select-control-placeholder: var(--color-text-disable);
    --ve-select-control-border-default: var(--color-border-default);
    --ve-select-control-border-hover: var(--color-border-interactive);
    --ve-select-control-border-focus: var(--color-border-interactive-active);
    --ve-select-control-border-disabled: var(--color-border-default);
    --ve-select-control-borderless-background: transparent;
    --ve-select-control-borderless-background-hover: var(--color-bg-surface);
    --ve-select-control-borderless-background-focus: transparent;
    --ve-select-control-borderless-border: transparent;
    --ve-select-focus-ring-default: var(--ve-select-control-border-focus);
    --ve-select-focus-shadow-strong-color: rgb(
      from var(--color-border-interactive) r g b / 25%
    );
    --ve-select-focus-shadow-soft-color: rgb(
      from var(--color-border-interactive) r g b / 10%
    );
    --ve-select-shadow-focus-default:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        var(--ve-select-focus-shadow-strong-color),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        var(--ve-select-focus-shadow-soft-color);
    --ve-select-status-warning-background: var(--color-bg-warning-secondary);
    --ve-select-status-warning-background-hover: color-mix(
      var(--ve-select-status-warning-background) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-select-status-warning-border: var(--color-border-warning);
    --ve-select-status-warning-border-hover: var(--color-text-warning);
    --ve-select-status-warning-border-focus: var(--color-text-warning);
    --ve-select-status-warning-shadow-focus: var(
      --ve-select-shadow-focus-default
    );
    --ve-select-status-error-background: var(--color-bg-danger-secondary);
    --ve-select-status-error-background-hover: color-mix(
      var(--ve-select-status-error-background) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-select-status-error-border: var(--color-border-danger);
    --ve-select-status-error-border-hover: var(--color-text-danger);
    --ve-select-status-error-border-focus: var(--color-text-danger);
    --ve-select-status-error-shadow-focus: var(
      --ve-select-shadow-focus-default
    );
    --ve-select-description-color: var(--color-text-tertiary);
    --ve-select-status-warning-description-color: var(--color-text-warning);
    --ve-select-status-error-description-color: var(--color-text-danger);
    --ve-select-feedback-font-size: var(--text-caption);
    --ve-select-feedback-line-height: var(--line-height-body);
    --ve-select-feedback-min-height: 18px;

    --ve-select-prefix-color: var(--ve-select-control-text);
    --ve-select-arrow-color: var(--color-icon-disable);
    --ve-select-clear-color: var(--color-text-tertiary);
    --ve-select-clear-hover-color: var(--color-text-secondary);
    --ve-select-clear-hover-background: var(--color-bg-muted);

    --ve-select-panel-background: var(--color-bg-overlay);
    --ve-select-panel-border-color: var(--color-border-default);
    --ve-select-panel-shadow: var(--shadow-md);
    --ve-select-panel-radius: var(--radius-lg);
    --ve-select-panel-z-index: 1000;
    --ve-select-panel-gap: var(--space-xxxs);
    --ve-select-panel-padding: var(--space-xxs);
    --ve-select-panel-max-height: 220px;

    --ve-select-option-min-height: 38px;
    --ve-select-option-padding-inline: var(--space-xxs);
    --ve-select-option-padding-block: var(--space-xxs);
    --ve-select-option-gap: var(--space-xxs);
    --ve-select-option-radius: var(--radius-md);
    --ve-select-option-text: var(--color-text-primary);
    --ve-select-option-text-disabled: var(--color-text-disable);
    --ve-select-option-hover-background: var(--color-bg-surface);
    --ve-select-option-active-background: color-mix(
      var(--color-bg-overlay) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );
    --ve-select-option-selected-background: transparent;
    --ve-select-option-selected-text: var(--color-text-primary);
    --ve-select-option-icon-color: var(--color-text-secondary);
    --ve-select-option-selected-icon: var(--color-text-accent);
    --ve-select-option-prefix-icon-size: var(--space-s);
    --ve-select-option-check-size: var(--space-s);
    --ve-select-option-check-border-width: var(--stroke-weight-strong);
    --ve-select-option-check-radius: var(--radius-xs);
    --ve-select-option-check-border: var(--color-border-strong);
    --ve-select-option-check-border-hover: var(--color-border-interactive);
    --ve-select-option-check-selected-border: transparent;
    --ve-select-option-check-background: transparent;
    --ve-select-option-check-selected-background: var(--color-bg-primary);
    --ve-select-option-check-selected-color: var(--color-text-foreground);
    --ve-select-option-check-selected-hover-background: color-mix(
      var(--ve-select-option-check-selected-background)
        var(--state-hover-solid-ratio),
      var(--state-hover-solid-mix)
    );

    --ve-select-group-color: var(--color-text-tertiary);
    --ve-select-group-padding-inline: var(--space-xxs);
    --ve-select-group-padding-block: var(--space-xxs);

    --ve-select-tag-background: var(--color-bg-surface);
    --ve-select-tag-color: var(--color-text-secondary);
    --ve-select-tag-remove-color: var(--color-text-tertiary);
    --ve-select-tag-remove-hover-background: var(--color-bg-surface);

    --ve-select-empty-color: var(--color-text-tertiary);
    --ve-select-empty-padding: var(--space-s);
    --ve-select-loading-color: var(--color-text-tertiary);
    --ve-select-loading-spin-duration: 1200ms;
    --ve-select-disabled-opacity: var(--state-disabled-ratio);
    --ve-select-transition-duration: var(--motion-fast);
    --ve-select-transition-timing: var(--ease-standard);

    --ve-select-current-height: var(--ve-select-height-default);
    --ve-select-current-padding-inline: var(--ve-select-padding-inline-default);
    --ve-select-current-padding-block: var(--ve-select-padding-block-default);
    --ve-select-current-font-size: var(--ve-select-font-size-default);
    --ve-select-current-line-height: var(--ve-select-line-height-default);
    --ve-select-current-control-border-width: var(
      --ve-select-control-border-width
    );
    --ve-select-current-background: var(--ve-select-control-background);
    --ve-select-current-background-hover: var(
      --ve-select-control-background-hover
    );
    --ve-select-current-border-color: var(--ve-select-control-border-default);
    --ve-select-current-border-color-hover: var(
      --ve-select-control-border-hover
    );
    --ve-select-current-focus-ring-color: var(--ve-select-focus-ring-default);
    --ve-select-current-shadow-focus: var(--ve-select-shadow-focus-default);
    --ve-select-current-description-color: var(--ve-select-description-color);
  }

  :host([size='small']) {
    --ve-select-current-height: var(--ve-select-height-small);
    --ve-select-current-padding-inline: var(--ve-select-padding-inline-small);
    --ve-select-current-padding-block: var(--ve-select-padding-block-small);
    --ve-select-current-font-size: var(--ve-select-font-size-small);
    --ve-select-current-line-height: var(--ve-select-line-height-small);
  }

  :host([size='large']) {
    --ve-select-current-height: var(--ve-select-height-large);
    --ve-select-current-padding-inline: var(--ve-select-padding-inline-large);
    --ve-select-current-padding-block: var(--ve-select-padding-block-large);
    --ve-select-current-font-size: var(--ve-select-font-size-large);
    --ve-select-current-line-height: var(--ve-select-line-height-large);
  }

  :host([status='warning']) {
    --ve-select-current-background: var(--ve-select-status-warning-background);
    --ve-select-current-background-hover: var(
      --ve-select-status-warning-background-hover
    );
    --ve-select-current-border-color: var(--ve-select-status-warning-border);
    --ve-select-current-border-color-hover: var(
      --ve-select-status-warning-border-hover
    );
    --ve-select-current-focus-ring-color: var(
      --ve-select-status-warning-border-focus
    );
    --ve-select-current-shadow-focus: var(
      --ve-select-status-warning-shadow-focus
    );
    --ve-select-current-description-color: var(
      --ve-select-status-warning-description-color
    );
  }

  :host([status='error']) {
    --ve-select-current-background: var(--ve-select-status-error-background);
    --ve-select-current-background-hover: var(
      --ve-select-status-error-background-hover
    );
    --ve-select-current-border-color: var(--ve-select-status-error-border);
    --ve-select-current-border-color-hover: var(
      --ve-select-status-error-border-hover
    );
    --ve-select-current-focus-ring-color: var(
      --ve-select-status-error-border-focus
    );
    --ve-select-current-shadow-focus: var(
      --ve-select-status-error-shadow-focus
    );
    --ve-select-current-description-color: var(
      --ve-select-status-error-description-color
    );
  }

  :host([borderless]) {
    --ve-select-current-control-border-width: 0px;
    --ve-select-current-background: var(
      --ve-select-control-borderless-background
    );
    --ve-select-current-background-hover: var(
      --ve-select-control-borderless-background-hover
    );
    --ve-select-control-background-focus: var(
      --ve-select-control-borderless-background-focus
    );
    --ve-select-control-background-disabled: var(
      --ve-select-control-borderless-background
    );
    --ve-select-current-border-color: var(
      --ve-select-control-borderless-border
    );
    --ve-select-current-border-color-hover: var(
      --ve-select-control-borderless-border
    );
    --ve-select-current-focus-ring-color: var(
      --ve-select-control-borderless-border
    );
    --ve-select-current-shadow-focus: none;
    --ve-select-control-border-disabled: var(
      --ve-select-control-borderless-border
    );
  }
`;
const token = veSelectTokenDefaults;
const ve_select = [
    componentReset,
    token,
    css`
    :host {
      position: relative;
      display: inline-flex;
      inline-size: 100%;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-select {
      position: relative;
      inline-size: 100%;
    }

    .ve-select-field {
      display: flex;
      align-items: stretch;
      inline-size: 100%;
    }

    .ve-select-control {
      appearance: none;
      display: flex;
      align-items: center;
      gap: var(--ve-select-gap);
      inline-size: 100%;
      block-size: var(--ve-select-current-height);
      min-inline-size: 0;
      box-sizing: border-box;
      padding-inline: var(--ve-select-current-padding-inline);
      padding-block: var(--ve-select-current-padding-block);
      border: var(--ve-select-current-control-border-width) solid
        var(--ve-select-current-border-color);
      border-radius: var(--ve-select-radius);
      background: var(--ve-select-current-background);
      color: var(--ve-select-control-text);
      font-family: var(--ve-select-font-family);
      font-size: var(--ve-select-current-font-size);
      font-weight: var(--ve-select-font-weight);
      line-height: var(--ve-select-current-line-height);
      cursor: pointer;
      transition:
        border-color var(--ve-select-transition-duration)
          var(--ve-select-transition-timing),
        box-shadow var(--ve-select-transition-duration)
          var(--ve-select-transition-timing),
        background-color var(--ve-select-transition-duration)
          var(--ve-select-transition-timing),
        opacity var(--ve-select-transition-duration)
          var(--ve-select-transition-timing);
      overflow: hidden;
    }

    .ve-select-control-filled.ve-select-control-multiple {
      padding-inline-start: var(
        --ve-select-control-multiple-tag-padding-inline-start
      );
    }

    :host([borderless]) .ve-select-control {
      inline-size: fit-content;
      max-inline-size: 100%;
    }

    .ve-select-label {
      display: inline-flex;
      align-items: center;
      flex: none;
      box-sizing: border-box;
      block-size: var(--ve-select-current-height);
      min-inline-size: 0;
      padding-inline: var(--ve-select-label-padding-inline);
      padding-block: var(--ve-select-current-padding-block);
      border: var(--ve-select-current-control-border-width) solid
        var(--ve-select-current-border-color);
      border-start-start-radius: var(--ve-select-radius);
      border-end-start-radius: var(--ve-select-radius);
      border-start-end-radius: 0;
      border-end-end-radius: 0;
      background: var(--ve-select-current-background);
      color: var(--ve-select-label-color);
      font-family: var(--ve-select-font-family);
      font-size: var(--ve-select-current-font-size);
      font-weight: var(--ve-select-font-weight);
      line-height: var(--ve-select-current-line-height);
      white-space: nowrap;
      cursor: pointer;
      transition:
        border-color var(--ve-select-transition-duration)
          var(--ve-select-transition-timing),
        background-color var(--ve-select-transition-duration)
          var(--ve-select-transition-timing),
        opacity var(--ve-select-transition-duration)
          var(--ve-select-transition-timing);
    }

    .ve-select-prefix.ve-select-prefix-empty {
      display: none;
    }

    .ve-select-label-text {
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ve-select-with-label .ve-select-control {
      border-start-start-radius: 0;
      border-end-start-radius: 0;
      margin-inline-start: calc(
        var(--ve-select-current-control-border-width) * -1
      );
    }

    :host(:not([disabled])) .ve-select-control:hover {
      border-color: var(--ve-select-current-border-color-hover);
      background: var(--ve-select-current-background-hover);
    }

    :host(:not([disabled])) .ve-select-label:hover {
      border-color: var(--ve-select-current-border-color-hover);
      background: var(--ve-select-current-background-hover);
    }

    :host(:not([disabled])) .ve-select-control:focus,
    :host([data-open='true']) .ve-select-control {
      outline: none;
      background: var(--ve-select-control-background-focus);
      border-color: var(--ve-select-current-focus-ring-color);
      box-shadow: var(--ve-select-current-shadow-focus);
    }

    :host([data-open='true']) .ve-select-label {
      background: var(--ve-select-control-background-focus);
      border-color: var(--ve-select-current-focus-ring-color);
    }

    :host([disabled]) .ve-select-control,
    :host([disabled]) .ve-select-label {
      color: var(--ve-select-control-text-disabled);
      background: var(--ve-select-control-background-disabled);
      border-color: var(--ve-select-control-border-disabled);
      opacity: var(--ve-select-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-select-prefix,
    .ve-select-arrow,
    .ve-select-loading {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
    }

    .ve-select-prefix {
      color: var(--ve-select-prefix-color);
    }

    :host([disabled]) .ve-select-prefix {
      color: var(--ve-select-control-text-disabled);
    }

    .ve-select-arrow {
      color: var(--ve-select-arrow-color);
    }

    .ve-select-loading {
      color: var(--ve-select-loading-color);
    }

    .ve-select-arrow ve-icon,
    .ve-select-loading ve-icon,
    .ve-select-option-icon ve-icon {
      --ve-icon-size: var(--ve-select-arrow-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-select-loading ve-icon {
      animation: ve-select-loading-spin
        var(--ve-select-loading-spin-duration) linear infinite;
    }

    .ve-select-value {
      display: flex;
      align-items: center;
      gap: var(--ve-select-tag-gap);
      min-inline-size: 0;
      flex: 1;
      flex-wrap: wrap;
    }

    :host([borderless]) .ve-select-value {
      flex: 0 1 auto;
    }

    .ve-select-value-text {
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-select-value-text-placeholder {
      color: var(--ve-select-control-placeholder);
    }

    .ve-select-tag {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      flex: none;
      max-inline-size: 100%;
      block-size: var(--ve-select-tag-height);
      padding-inline: var(--ve-select-tag-padding-inline);
      border-radius: var(--ve-select-tag-radius);
      background: var(--ve-select-tag-background);
      color: var(--ve-select-tag-color);
      font-size: var(--text-caption);
      line-height: var(--line-height-body);
      box-sizing: border-box;
      overflow: hidden;
    }

    .ve-select-tag-label {
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-select-tag-remove {
      appearance: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: var(--ve-select-tag-remove-size);
      block-size: var(--ve-select-tag-remove-size);
      border: none;
      border-radius: var(--radius-full);
      background: transparent;
      color: var(--ve-select-tag-remove-color);
      cursor: pointer;
      padding: 0;
    }

    .ve-select-tag-remove:hover {
      background: var(--ve-select-tag-remove-hover-background);
    }

    .ve-select-actions {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      margin-inline-start: auto;
      flex: none;
      pointer-events: none;
    }

    :host([borderless]) .ve-select-actions {
      margin-inline-start: 0;
    }

    .ve-select-actions > * {
      pointer-events: auto;
    }

    .ve-select-clear {
      appearance: none;
      display: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-select-clear-size);
      block-size: var(--ve-select-clear-size);
      border: none;
      border-radius: var(--radius-full);
      background: transparent;
      color: var(--ve-select-clear-color);
      cursor: pointer;
      padding: 0;
    }

    :host(:not([disabled])) .ve-select-control:hover .ve-select-clear,
    :host([data-open='true']) .ve-select-clear {
      display: inline-flex;
    }

    .ve-select-clear:hover {
      color: var(--ve-select-clear-hover-color);
      background: var(--ve-select-clear-hover-background);
    }

    .ve-select-tag-remove ve-icon,
    .ve-select-clear ve-icon {
      --ve-icon-size: var(--ve-select-action-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-select-panel {
      position: fixed;
      inset: auto;
      top: 0;
      left: 0;
      z-index: var(--ve-select-panel-z-index);
      isolation: isolate;
      box-sizing: border-box;
      margin: 0;
      padding: 0;
      border: var(--ve-select-border-width) solid
        var(--ve-select-panel-border-color);
      border-radius: var(--ve-select-panel-radius);
      background: var(--ve-select-panel-background);
      box-shadow: var(--ve-select-panel-shadow);
      overflow: hidden;
      will-change: transform;
    }

    .ve-select-panel-inner {
      padding: var(--ve-select-panel-padding);
      max-block-size: var(--ve-select-panel-max-height);
      overflow: auto;
      overscroll-behavior: contain;
    }

    .ve-select-group {
      display: block;
    }

    .ve-select-group-label {
      display: block;
      padding-inline: var(--ve-select-group-padding-inline);
      padding-block: var(--ve-select-group-padding-block);
      color: var(--ve-select-group-color);
      font-size: var(--text-caption);
    }

    .ve-select-options,
    .ve-select-group-options {
      margin: 0;
      padding: 0;
      list-style: none;
    }

    .ve-select-group-options {
      display: block;
    }

    .ve-select-option {
      appearance: none;
      display: flex;
      align-items: center;
      justify-content: space-between;
      inline-size: 100%;
      min-block-size: var(--ve-select-option-min-height);
      box-sizing: border-box;
      gap: var(--ve-select-option-gap);
      padding-inline: var(--ve-select-option-padding-inline);
      padding-block: var(--ve-select-option-padding-block);
      border: none;
      border-radius: var(--ve-select-option-radius);
      background: transparent;
      color: var(--ve-select-option-text);
      cursor: pointer;
      font: inherit;
      text-align: left;
      transition:
        background-color var(--ve-select-transition-duration)
          var(--ve-select-transition-timing),
        color var(--ve-select-transition-duration)
          var(--ve-select-transition-timing);
    }

    .ve-select-option:hover,
    .ve-select-option:focus-visible {
      outline: none;
      background: var(--ve-select-option-hover-background);
    }

    .ve-select-option:active,
    .ve-select-option-active {
      outline: none;
      background: var(--ve-select-option-active-background);
    }

    .ve-select-option-selected {
      background: var(--ve-select-option-selected-background);
      color: var(--ve-select-option-selected-text);
      font-weight: var(--font-weight-normal);
    }

    .ve-select-option-active.ve-select-option-selected {
      background: var(--ve-select-option-active-background);
    }

    .ve-select-option-disabled,
    .ve-select-option-disabled:hover,
    .ve-select-option-disabled:active,
    .ve-select-option-disabled:focus-visible {
      color: var(--ve-select-option-text-disabled);
      cursor: not-allowed;
      background: transparent;
      opacity: var(--ve-select-disabled-opacity);
      pointer-events: none;
    }

    .ve-select-option-label,
    .ve-select-option-content {
      min-inline-size: 0;
      flex: 1;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-select-option-content {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-select-option-gap);
    }

    .ve-select-option-content ::slotted(*) {
      flex: none;
    }

    .ve-select-option-check,
    .ve-select-option-prefix,
    .ve-select-option-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      color: var(--ve-select-option-icon-color);
    }

    .ve-select-option-check {
      inline-size: var(--ve-select-option-check-size);
      block-size: var(--ve-select-option-check-size);
      box-sizing: border-box;
      border: var(--ve-select-option-check-border-width) solid
        var(--ve-select-option-check-border);
      border-radius: var(--ve-select-option-check-radius);
      background: var(--ve-select-option-check-background);
    }

    .ve-select-option-check-selected {
      border-color: var(--ve-select-option-check-selected-border);
      background: var(--ve-select-option-check-selected-background);
      color: var(--ve-select-option-check-selected-color);
    }

    .ve-select-option:hover
      .ve-select-option-check:not(.ve-select-option-check-selected),
    .ve-select-option:focus-visible
      .ve-select-option-check:not(.ve-select-option-check-selected) {
      border-color: var(--ve-select-option-check-border-hover);
    }

    .ve-select-option:hover .ve-select-option-check-selected,
    .ve-select-option:focus-visible .ve-select-option-check-selected {
      background: var(--ve-select-option-check-selected-hover-background);
    }

    .ve-select-option-check-icon {
      display: block;
      flex: none;
      inline-size: 12px;
      block-size: 12px;
    }

    .ve-select-option-prefix ve-icon {
      --ve-icon-size: var(--ve-select-option-prefix-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-select-option-icon-selected {
      color: var(--ve-select-option-selected-icon);
    }

    .ve-select-empty {
      padding: var(--ve-select-empty-padding);
      color: var(--ve-select-empty-color);
      text-align: center;
      font-size: var(--text-caption);
    }

    .ve-select-feedback {
      display: flex;
      flex-direction: column;
      gap: var(--ve-select-gap);
      margin-block-start: var(--ve-select-gap);
      inline-size: 100%;
      font-family: var(--ve-select-font-family);
      font-size: var(--ve-select-feedback-font-size);
      font-weight: var(--ve-select-font-weight);
      line-height: var(--ve-select-feedback-line-height);
    }

    .ve-select-description {
      min-block-size: var(--ve-select-feedback-min-height);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-select-description {
      color: var(--ve-select-current-description-color);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-select-control,
      .ve-select-arrow,
      .ve-select-loading,
      .ve-select-option {
        transition: none;
        animation: none;
      }
    }

    @keyframes ve-select-loading-spin {
      to {
        transform: rotate(360deg);
      }
    }
  `
];
function ve_select_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_select_element_toPropertyKey(arg) {
    var key = ve_select_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_select_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_select_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_select_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_select_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_select_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_select_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_select_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_select_element_dec, ve_select_element_dec1, ve_select_element_dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _dec24, _dec25, _dec26, ve_select_element_init_value, ve_select_element_init_extra_init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_placeholder, _init_extra__init_placeholder, ve_select_element_init_label, ve_select_element_init_extra_init_label, _init_description, _init_extra__init_description, _init_mode, _init_extra__init_mode, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_borderless, _init_extra__init_borderless, ve_select_element_init_disabled, ve_select_element_init_extra_init_disabled, _init_loading, _init_extra__init_loading, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_maxTagCount, _init_extra__init_maxTagCount, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_emptyText, _init_extra__init_emptyText, _init_maxHeight, _init_extra__init_maxHeight, _init_controlElement, _init_extra__init_controlElement, _init_fieldElement, _init_extra__init_fieldElement, _init_emptySlot, _init_extra__init_emptySlot, _init_panelElement, _init_extra__init_panelElement, _init_normalizedOptions, _init_extra__init_normalizedOptions, _init_activeOptionKey, _init_extra__init_activeOptionKey, _init_panelPlacement, _init_extra__init_panelPlacement, _init_innerOpen, _init_extra__init_innerOpen, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot;
const VE_SELECT_TAG = 've-select';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby',
    'aria-required',
    'aria-invalid'
];
const VALUE_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return '';
        const trimmed = value.trim();
        if (!trimmed) return '';
        try {
            const parsed = JSON.parse(trimmed);
            if (Array.isArray(parsed)) return parsed.filter(isOptionValue);
            if (isOptionValue(parsed)) return parsed;
        } catch  {}
        return trimmed;
    },
    toAttribute (value) {
        if (Array.isArray(value)) return value.length > 0 ? JSON.stringify(value) : null;
        return '' === value ? null : String(value);
    }
};
const MAX_TAG_COUNT_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return;
        if ('responsive' === value) return 'responsive';
        const numeric = Number(value);
        return Number.isFinite(numeric) && numeric >= 0 ? Math.floor(numeric) : void 0;
    },
    toAttribute (value) {
        return void 0 === value ? null : String(value);
    }
};
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        return null === value || 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
function isOptionValue(value) {
    return 'string' == typeof value || 'number' == typeof value;
}
function isGroup(item) {
    return Array.isArray(item.options);
}
function toOptionKey(value) {
    return `${typeof value}:${String(value)}`;
}
function isSameValue(left, right) {
    return toOptionKey(left) === toOptionKey(right);
}
function areSelectValuesEqual(left, right) {
    if (Array.isArray(left) || Array.isArray(right)) {
        if (!Array.isArray(left) || !Array.isArray(right) || left.length !== right.length) return false;
        return left.every((value, index)=>isSameValue(value, right[index]));
    }
    return left === right;
}
function normalizeAriaBooleanString(value) {
    if (null === value) return;
    return 'false' === value ? 'false' : 'true';
}
ve_select_element_dec = property({
    converter: VALUE_CONVERTER
}), ve_select_element_dec1 = property({
    attribute: 'default-value',
    converter: VALUE_CONVERTER
}), ve_select_element_dec2 = property({
    type: String
}), _dec3 = property({
    type: String
}), _dec4 = property({
    type: String
}), _dec5 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec6 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec8 = property({
    type: Boolean,
    reflect: true
}), _dec9 = property({
    type: Boolean,
    reflect: true
}), _dec10 = property({
    type: Boolean,
    reflect: true
}), _dec11 = property({
    type: Boolean,
    attribute: 'allow-clear',
    reflect: true
}), _dec12 = property({
    type: Boolean,
    reflect: true
}), _dec13 = property({
    type: Boolean,
    attribute: 'default-open'
}), _dec14 = property({
    attribute: 'max-tag-count',
    converter: MAX_TAG_COUNT_CONVERTER
}), _dec15 = property({
    attribute: 'popup-match-select-width',
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER
}), _dec16 = property({
    type: String,
    attribute: 'empty-text'
}), _dec17 = property({
    type: Number,
    attribute: 'max-height'
}), _dec18 = query('.ve-select-control'), _dec19 = query('.ve-select-field'), _dec20 = query('slot[name="empty"]'), _dec21 = query('.ve-select-panel'), _dec22 = state(), _dec23 = state(), _dec24 = state(), _dec25 = state(), _dec26 = state();
class VeSelectElement extends VeElement {
    static{
        ({ e: [ve_select_element_init_value, ve_select_element_init_extra_init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_placeholder, _init_extra__init_placeholder, ve_select_element_init_label, ve_select_element_init_extra_init_label, _init_description, _init_extra__init_description, _init_mode, _init_extra__init_mode, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_borderless, _init_extra__init_borderless, ve_select_element_init_disabled, ve_select_element_init_extra_init_disabled, _init_loading, _init_extra__init_loading, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_maxTagCount, _init_extra__init_maxTagCount, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_emptyText, _init_extra__init_emptyText, _init_maxHeight, _init_extra__init_maxHeight, _init_controlElement, _init_extra__init_controlElement, _init_fieldElement, _init_extra__init_fieldElement, _init_emptySlot, _init_extra__init_emptySlot, _init_panelElement, _init_extra__init_panelElement, _init_normalizedOptions, _init_extra__init_normalizedOptions, _init_activeOptionKey, _init_extra__init_activeOptionKey, _init_panelPlacement, _init_extra__init_panelPlacement, _init_innerOpen, _init_extra__init_innerOpen, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot] } = ve_select_element_apply_decs_2311(this, [], [
            [
                ve_select_element_dec,
                1,
                "value"
            ],
            [
                ve_select_element_dec1,
                1,
                "defaultValue"
            ],
            [
                ve_select_element_dec2,
                1,
                "placeholder"
            ],
            [
                _dec3,
                1,
                "label"
            ],
            [
                _dec4,
                1,
                "description"
            ],
            [
                _dec5,
                1,
                "mode"
            ],
            [
                _dec6,
                1,
                "size"
            ],
            [
                _dec7,
                1,
                "status"
            ],
            [
                _dec8,
                1,
                "borderless"
            ],
            [
                _dec9,
                1,
                "disabled"
            ],
            [
                _dec10,
                1,
                "loading"
            ],
            [
                _dec11,
                1,
                "allowClear"
            ],
            [
                _dec12,
                1,
                "open"
            ],
            [
                _dec13,
                1,
                "defaultOpen"
            ],
            [
                _dec14,
                1,
                "maxTagCount"
            ],
            [
                _dec15,
                1,
                "popupMatchSelectWidth"
            ],
            [
                _dec16,
                1,
                "emptyText"
            ],
            [
                _dec17,
                1,
                "maxHeight"
            ],
            [
                _dec18,
                1,
                "controlElement"
            ],
            [
                _dec19,
                1,
                "fieldElement"
            ],
            [
                _dec20,
                1,
                "emptySlot"
            ],
            [
                _dec21,
                1,
                "panelElement"
            ],
            [
                _dec22,
                1,
                "normalizedOptions"
            ],
            [
                _dec23,
                1,
                "activeOptionKey"
            ],
            [
                _dec24,
                1,
                "panelPlacement"
            ],
            [
                _dec25,
                1,
                "innerOpen"
            ],
            [
                _dec26,
                1,
                "hasPrefixSlot"
            ]
        ]));
    }
    static styles = ve_select;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement,
        [VE_SELECT_ITEM_TAG]: VeSelectItemElement,
        [VE_SELECT_GROUP_TAG]: VeSelectGroupElement
    };
    #___private_value_1 = (()=>{
        const _value = ve_select_element_init_value(this, '');
        ve_select_element_init_extra_init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_defaultValue_2 = (()=>{
        const _value = _init_defaultValue(this, '');
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_2;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_2 = _v;
    }
    #___private_placeholder_3 = (()=>{
        const _value = _init_placeholder(this, 'Please select');
        _init_extra__init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_3;
    }
    set placeholder(_v) {
        this.#___private_placeholder_3 = _v;
    }
    #___private_label_4 = (()=>{
        const _value = ve_select_element_init_label(this, '');
        ve_select_element_init_extra_init_label(this);
        return _value;
    })();
    get label() {
        return this.#___private_label_4;
    }
    set label(_v) {
        this.#___private_label_4 = _v;
    }
    #___private_description_5 = (()=>{
        const _value = _init_description(this, '');
        _init_extra__init_description(this);
        return _value;
    })();
    get description() {
        return this.#___private_description_5;
    }
    set description(_v) {
        this.#___private_description_5 = _v;
    }
    #___private_mode_6 = (()=>{
        const _value = _init_mode(this, 'single');
        _init_extra__init_mode(this);
        return _value;
    })();
    get mode() {
        return this.#___private_mode_6;
    }
    set mode(_v) {
        this.#___private_mode_6 = _v;
    }
    #___private_size_7 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_7;
    }
    set size(_v) {
        this.#___private_size_7 = _v;
    }
    #___private_status_8 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_8;
    }
    set status(_v) {
        this.#___private_status_8 = _v;
    }
    #___private_borderless_9 = (()=>{
        const _value = _init_borderless(this, false);
        _init_extra__init_borderless(this);
        return _value;
    })();
    get borderless() {
        return this.#___private_borderless_9;
    }
    set borderless(_v) {
        this.#___private_borderless_9 = _v;
    }
    #___private_disabled_10 = (()=>{
        const _value = ve_select_element_init_disabled(this, false);
        ve_select_element_init_extra_init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_10;
    }
    set disabled(_v) {
        this.#___private_disabled_10 = _v;
    }
    #___private_loading_11 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_11;
    }
    set loading(_v) {
        this.#___private_loading_11 = _v;
    }
    #___private_allowClear_12 = (()=>{
        const _value = _init_allowClear(this, false);
        _init_extra__init_allowClear(this);
        return _value;
    })();
    get allowClear() {
        return this.#___private_allowClear_12;
    }
    set allowClear(_v) {
        this.#___private_allowClear_12 = _v;
    }
    #___private_open_13 = (()=>{
        const _value = _init_open(this, void 0);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_13;
    }
    set open(_v) {
        this.#___private_open_13 = _v;
    }
    #___private_defaultOpen_14 = (()=>{
        const _value = _init_defaultOpen(this, false);
        _init_extra__init_defaultOpen(this);
        return _value;
    })();
    get defaultOpen() {
        return this.#___private_defaultOpen_14;
    }
    set defaultOpen(_v) {
        this.#___private_defaultOpen_14 = _v;
    }
    #___private_maxTagCount_15 = (()=>{
        const _value = _init_maxTagCount(this, void 0);
        _init_extra__init_maxTagCount(this);
        return _value;
    })();
    get maxTagCount() {
        return this.#___private_maxTagCount_15;
    }
    set maxTagCount(_v) {
        this.#___private_maxTagCount_15 = _v;
    }
    #___private_popupMatchSelectWidth_16 = (()=>{
        const _value = _init_popupMatchSelectWidth(this, true);
        _init_extra__init_popupMatchSelectWidth(this);
        return _value;
    })();
    get popupMatchSelectWidth() {
        return this.#___private_popupMatchSelectWidth_16;
    }
    set popupMatchSelectWidth(_v) {
        this.#___private_popupMatchSelectWidth_16 = _v;
    }
    #___private_emptyText_17 = (()=>{
        const _value = _init_emptyText(this, 'No data');
        _init_extra__init_emptyText(this);
        return _value;
    })();
    get emptyText() {
        return this.#___private_emptyText_17;
    }
    set emptyText(_v) {
        this.#___private_emptyText_17 = _v;
    }
    #___private_maxHeight_18 = (()=>{
        const _value = _init_maxHeight(this, 220);
        _init_extra__init_maxHeight(this);
        return _value;
    })();
    get maxHeight() {
        return this.#___private_maxHeight_18;
    }
    set maxHeight(_v) {
        this.#___private_maxHeight_18 = _v;
    }
    #___private_controlElement_19 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_19;
    }
    set controlElement(_v) {
        this.#___private_controlElement_19 = _v;
    }
    #___private_fieldElement_20 = (()=>{
        const _value = _init_fieldElement(this, null);
        _init_extra__init_fieldElement(this);
        return _value;
    })();
    get fieldElement() {
        return this.#___private_fieldElement_20;
    }
    set fieldElement(_v) {
        this.#___private_fieldElement_20 = _v;
    }
    #___private_emptySlot_21 = (()=>{
        const _value = _init_emptySlot(this, null);
        _init_extra__init_emptySlot(this);
        return _value;
    })();
    get emptySlot() {
        return this.#___private_emptySlot_21;
    }
    set emptySlot(_v) {
        this.#___private_emptySlot_21 = _v;
    }
    #___private_panelElement_22 = (()=>{
        const _value = _init_panelElement(this, null);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_22;
    }
    set panelElement(_v) {
        this.#___private_panelElement_22 = _v;
    }
    #___private_normalizedOptions_23 = (()=>{
        const _value = _init_normalizedOptions(this, []);
        _init_extra__init_normalizedOptions(this);
        return _value;
    })();
    get normalizedOptions() {
        return this.#___private_normalizedOptions_23;
    }
    set normalizedOptions(_v) {
        this.#___private_normalizedOptions_23 = _v;
    }
    #___private_activeOptionKey_24 = (()=>{
        const _value = _init_activeOptionKey(this, '');
        _init_extra__init_activeOptionKey(this);
        return _value;
    })();
    get activeOptionKey() {
        return this.#___private_activeOptionKey_24;
    }
    set activeOptionKey(_v) {
        this.#___private_activeOptionKey_24 = _v;
    }
    #___private_panelPlacement_25 = (()=>{
        const _value = _init_panelPlacement(this, 'bottom-start');
        _init_extra__init_panelPlacement(this);
        return _value;
    })();
    get panelPlacement() {
        return this.#___private_panelPlacement_25;
    }
    set panelPlacement(_v) {
        this.#___private_panelPlacement_25 = _v;
    }
    #___private_innerOpen_26 = (()=>{
        const _value = _init_innerOpen(this, false);
        _init_extra__init_innerOpen(this);
        return _value;
    })();
    get innerOpen() {
        return this.#___private_innerOpen_26;
    }
    set innerOpen(_v) {
        this.#___private_innerOpen_26 = _v;
    }
    #___private_hasPrefixSlot_27 = (()=>{
        const _value = _init_hasPrefixSlot(this, false);
        _init_extra__init_hasPrefixSlot(this);
        return _value;
    })();
    get hasPrefixSlot() {
        return this.#___private_hasPrefixSlot_27;
    }
    set hasPrefixSlot(_v) {
        this.#___private_hasPrefixSlot_27 = _v;
    }
    hasInitializedDefaultValue = false;
    openChangeReason = 'api';
    repositionScheduler = createRafScheduler();
    positionRequestId = 0;
    cleanupAutoUpdate = null;
    listboxId = '';
    descriptionId = '';
    flatOptions = [];
    enabledOptions = [];
    optionByKey = new Map();
    filteredOptions = [];
    filteredEnabledOptions = [];
    selectedOptionsCache = [];
    selectedOptionKeySet = new Set();
    lastScrolledActiveOptionKey = '';
    ignoreNextControlClick = false;
    optionContentSlotId = 0;
    optionContentSlotByElement = new WeakMap();
    optionContentSlotByKey = new Map();
    constructor(){
        super();
        this.listboxId = `ve-select-listbox-${Math.random().toString(36).slice(2)}`;
        this.descriptionId = `ve-select-description-${Math.random().toString(36).slice(2)}`;
    }
    addEventListener(type, listener, options) {
        super.addEventListener(type, listener, options);
    }
    removeEventListener(type, listener, options) {
        super.removeEventListener(type, listener, options);
    }
    handleDocumentPointerDown = (event)=>{
        if (!this.isOpen || event.composedPath().includes(this)) return;
        this.setOpen(false, 'outside');
    };
    handleDocumentKeydown = (event)=>{
        if (!this.isOpen || 'Escape' !== event.key) return;
        event.stopPropagation();
        this.setOpen(false, 'escape');
        this.focus();
    };
    handleOptionsSourceChange = ()=>{
        this.refreshOptions();
        this.refreshFilteredOptions();
        if (this.isOpen) {
            this.ensureActiveOption();
            this.scheduleUpdatePosition();
        }
    };
    get isMultiple() {
        return 'multiple' === this.mode;
    }
    get isOpen() {
        return 'boolean' == typeof this.open ? this.open : this.innerOpen;
    }
    get isOpenControlled() {
        return 'boolean' == typeof this.open;
    }
    get normalizedMode() {
        return 'multiple' === this.mode ? 'multiple' : 'single';
    }
    get shouldShowClear() {
        if (!this.allowClear || this.disabled || this.loading) return false;
        return this.isMultipleValue(this.value) ? this.value.length > 0 : '' !== this.value;
    }
    get hasLeadingLabel() {
        return (this.label ?? '').trim().length > 0;
    }
    get hasDescriptionFeedback() {
        return (this.description ?? '').trim().length > 0;
    }
    get isFilled() {
        return this.hasValue(this.value);
    }
    get activeOptionId() {
        return this.activeOptionKey ? this.getOptionId(this.activeOptionKey) : void 0;
    }
    get controlAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get controlAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get controlAriaDescribedby() {
        const describedBy = this.getAttribute('aria-describedby');
        const ids = [
            describedBy,
            this.hasDescriptionFeedback ? this.descriptionId : void 0
        ].filter((id)=>'string' == typeof id && id.length > 0);
        return ids.length > 0 ? ids.join(' ') : void 0;
    }
    get controlAriaRequired() {
        return normalizeAriaBooleanString(this.getAttribute('aria-required'));
    }
    get controlAriaInvalid() {
        return normalizeAriaBooleanString(this.getAttribute('aria-invalid')) ?? ('error' === this.status ? 'true' : void 0);
    }
    getOptionFromKey(key) {
        return key ? this.optionByKey.get(key) : void 0;
    }
    focus(options) {
        this.controlElement?.focus(options);
    }
    blur() {
        this.controlElement?.blur();
    }
    show() {
        this.setOpen(true, 'api');
    }
    hide() {
        this.setOpen(false, 'api');
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.handleOptionsSourceChange();
        this.applyDefaultValueIfNeeded();
        if (this.isOpen) {
            this.installGlobalListeners();
            this.activateInitialOption();
        }
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('mode')) {
            const normalizedMode = this.normalizedMode;
            if (this.mode !== normalizedMode) this.mode = normalizedMode;
        }
        if (changedProperties.has('mode') || changedProperties.has('value')) {
            const normalizedValue = this.normalizeValueForMode(this.value);
            if (!areSelectValuesEqual(this.value, normalizedValue)) this.value = normalizedValue;
        }
        if (changedProperties.has('mode') || changedProperties.has('value')) this.syncSelectedOptions();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('disabled') && this.disabled && this.isOpen) return void this.setOpen(false, 'api');
        if (changedProperties.has('open') && 'boolean' == typeof this.open) this.innerOpen = this.open;
        if (changedProperties.has('defaultOpen') && 1 === changedProperties.size && !this.isOpenControlled) this.innerOpen = this.defaultOpen;
        const didInnerOpenChange = changedProperties.has('innerOpen');
        if (changedProperties.has('open') || didInnerOpenChange) {
            this.toggleAttribute('data-open', this.isOpen);
            this.handleOpenChanged();
        }
        if (this.isOpen && changedProperties.has('value')) {
            this.ensureActiveOption();
            this.scheduleUpdatePosition();
        }
        if (this.isOpen && this.activeOptionKey && this.activeOptionKey !== this.lastScrolledActiveOptionKey) {
            this.scrollActiveOptionIntoView();
            this.lastScrolledActiveOptionKey = this.activeOptionKey;
        }
        if (!this.isOpen || '' === this.activeOptionKey) this.lastScrolledActiveOptionKey = '';
    }
    connectedCallback() {
        super.connectedCallback();
        this.innerOpen = this.open ?? this.defaultOpen;
        super.addEventListener(VE_SELECT_OPTIONS_CHANGE_EVENT, this.handleOptionsSourceChange);
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    disconnectedCallback() {
        super.removeEventListener(VE_SELECT_OPTIONS_CHANGE_EVENT, this.handleOptionsSourceChange);
        this.removeGlobalListeners();
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        this.teardownAutoUpdate();
        super.disconnectedCallback();
    }
    isMultipleValue(value) {
        return Array.isArray(value);
    }
    normalizeValueForMode(value) {
        if (this.isMultiple) {
            if (Array.isArray(value)) return value;
            return '' === value ? [] : [
                value
            ];
        }
        return Array.isArray(value) ? value[0] ?? '' : value;
    }
    applyDefaultValueIfNeeded() {
        if (this.hasInitializedDefaultValue) return;
        this.hasInitializedDefaultValue = true;
        if (this.hasValue(this.value) || !this.hasValue(this.defaultValue)) return;
        this.value = this.normalizeValueForMode(this.defaultValue);
    }
    hasValue(value) {
        return Array.isArray(value) ? value.length > 0 : '' !== value;
    }
    installGlobalListeners() {
        document.addEventListener('pointerdown', this.handleDocumentPointerDown);
        document.addEventListener('keydown', this.handleDocumentKeydown);
    }
    removeGlobalListeners() {
        document.removeEventListener('pointerdown', this.handleDocumentPointerDown);
        document.removeEventListener('keydown', this.handleDocumentKeydown);
    }
    normalizeItemElement(element, parentDisabled = false, allowSlottedContent = false) {
        const value = element.value;
        if (!isOptionValue(value)) return null;
        const option = {
            label: element.label || String(value),
            value,
            disabled: parentDisabled || element.disabled,
            closable: element.closable
        };
        if (allowSlottedContent) {
            const contentSlotName = this.getOptionContentSlotName(element);
            this.optionContentSlotByKey.set(toOptionKey(value), contentSlotName);
            if (element.slot !== contentSlotName) element.slot = contentSlotName;
        }
        return option;
    }
    getOptionContentSlotName(element) {
        const current = this.optionContentSlotByElement.get(element);
        if (current) return current;
        const next = `${this.listboxId}-option-content-${this.optionContentSlotId++}`;
        this.optionContentSlotByElement.set(element, next);
        return next;
    }
    normalizeOptions() {
        const normalized = [];
        for (const item of Array.from(this.children)){
            if (item instanceof VeSelectItemElement) {
                const option = this.normalizeItemElement(item, false, true);
                if (option) normalized.push(option);
                continue;
            }
            if (item instanceof VeSelectGroupElement) {
                const groupOptions = Array.from(item.children).filter((child)=>child instanceof VeSelectItemElement).map((option)=>this.normalizeItemElement(option, item.disabled)).filter((option)=>null !== option);
                if (groupOptions.length > 0) normalized.push({
                    label: (item.label ?? '').trim().length > 0 ? item.label : 'Group',
                    disabled: item.disabled,
                    options: groupOptions
                });
            }
        }
        return normalized;
    }
    refreshOptions() {
        this.optionContentSlotByKey = new Map();
        this.normalizedOptions = this.normalizeOptions();
        this.flatOptions = this.flattenOptions(this.normalizedOptions);
        this.enabledOptions = this.flatOptions.filter(({ option })=>!option.disabled);
        this.optionByKey = new Map(this.flatOptions.map(({ key, option })=>[
                key,
                option
            ]));
        this.syncSelectedOptions();
    }
    flattenOptions(items) {
        const flattened = [];
        for (const item of items){
            const options = isGroup(item) ? item.options : [
                item
            ];
            for (const option of options)flattened.push({
                option,
                key: toOptionKey(option.value)
            });
        }
        return flattened;
    }
    refreshFilteredOptions() {
        this.filteredOptions = this.computeFilteredOptions();
        this.filteredEnabledOptions = this.flattenOptions(this.filteredOptions).filter(({ option })=>!option.disabled);
    }
    computeFilteredOptions() {
        return this.normalizedOptions;
    }
    findOptionByValue(value) {
        return this.optionByKey.get(toOptionKey(value));
    }
    getSelectedOptions() {
        return this.selectedOptionsCache;
    }
    syncSelectedOptions() {
        const selectedOptions = Array.isArray(this.value) ? this.value.map((item)=>this.findOptionByValue(item)).filter((item)=>void 0 !== item) : '' === this.value ? [] : (()=>{
            const option = this.findOptionByValue(this.value);
            return option ? [
                option
            ] : [];
        })();
        this.selectedOptionsCache = selectedOptions;
        this.selectedOptionKeySet = new Set(selectedOptions.map((option)=>toOptionKey(option.value)));
    }
    getSelectedOptionsFromValue(value) {
        if (Array.isArray(value)) return value.map((item)=>this.findOptionByValue(item)).filter((item)=>void 0 !== item);
        if ('' === value) return [];
        const option = this.findOptionByValue(value);
        return option ? [
            option
        ] : [];
    }
    setOpen(nextOpen, reason) {
        if (nextOpen && this.disabled) return;
        if (this.isOpen === nextOpen) return;
        this.openChangeReason = reason;
        if (!this.isOpenControlled) this.innerOpen = nextOpen;
        this.emitOpenChange(nextOpen, reason);
    }
    handleOpenChanged() {
        if (this.isOpen && !this.disabled) {
            this.installGlobalListeners();
            this.activateInitialOption();
            this.updateComplete.then(()=>{
                if (!this.isConnected || !this.isOpen || this.disabled) return;
                showPanelPopover(this.panelElement);
                this.setupAutoUpdate();
                this.scheduleUpdatePosition();
            });
        } else {
            hidePanelPopover(this.panelElement);
            this.removeGlobalListeners();
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            this.teardownAutoUpdate();
            this.activeOptionKey = '';
        }
        this.openChangeReason = 'api';
    }
    emit(type, detail) {
        this.dispatchEvent(new CustomEvent(type, {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    emitChange(option) {
        this.emit('ve-change', {
            value: Array.isArray(this.value) ? [
                ...this.value
            ] : this.value,
            option
        });
    }
    emitClear(value) {
        this.emit('ve-clear', {
            value
        });
    }
    emitOpenChange(open, reason) {
        this.emit('ve-open-change', {
            open,
            reason
        });
    }
    emitRemove(option) {
        this.emit('ve-remove', {
            value: option.value,
            option
        });
    }
    selectOption(option) {
        if (option.disabled || this.disabled || this.loading) return;
        if (this.isMultiple) {
            const values = Array.isArray(this.value) ? [
                ...this.value
            ] : [];
            const exists = values.some((item)=>isSameValue(item, option.value));
            if (exists) {
                const nextValue = values.filter((item)=>!isSameValue(item, option.value));
                this.value = nextValue;
                this.emitRemove(option);
                this.emitChange(this.getSelectedOptionsFromValue(nextValue));
                return;
            }
            const nextValue = [
                ...values,
                option.value
            ];
            this.value = nextValue;
            this.emitChange(this.getSelectedOptionsFromValue(nextValue));
            return;
        }
        if (!Array.isArray(this.value) && '' !== this.value && isSameValue(this.value, option.value)) return void this.setOpen(false, 'select');
        this.value = option.value;
        this.emitChange(option);
        this.setOpen(false, 'select');
    }
    clearValue(event) {
        event?.preventDefault();
        event?.stopPropagation();
        if (this.disabled || this.loading) return;
        const nextValue = this.isMultiple ? [] : '';
        this.value = nextValue;
        this.emitClear(nextValue);
        this.emitChange(void 0);
    }
    removeValue(option, event) {
        event.preventDefault();
        event.stopPropagation();
        if (!this.isMultiple || !Array.isArray(this.value) || this.disabled || this.loading) return;
        const nextValue = this.value.filter((item)=>!isSameValue(item, option.value));
        this.value = nextValue;
        this.emitRemove(option);
        this.emitChange(this.getSelectedOptionsFromValue(nextValue));
    }
    isEmbeddedActionTarget(target) {
        return target instanceof Element && null !== target.closest('.ve-select-clear, .ve-select-tag-remove');
    }
    handleControlPointerDown = (event)=>{
        if (0 !== event.button || this.disabled || this.isEmbeddedActionTarget(event.target)) return;
        event.preventDefault();
        this.ignoreNextControlClick = true;
        this.controlElement?.focus();
        this.setOpen(!this.isOpen, 'trigger');
    };
    handleControlClick() {
        if (this.ignoreNextControlClick) {
            this.ignoreNextControlClick = false;
            return;
        }
        if (this.disabled) return;
        this.setOpen(!this.isOpen, 'trigger');
    }
    handleControlKeydown(event) {
        if (this.disabled || event.target !== this.controlElement) return;
        this.handleNavigationKeydown(event);
    }
    handleNavigationKeydown(event) {
        switch(event.key){
            case 'Enter':
                event.preventDefault();
                this.handleEnterKey();
                break;
            case ' ':
                if (!this.isOpen || event.target === this.controlElement) {
                    event.preventDefault();
                    this.setOpen(true, 'trigger');
                }
                break;
            case 'ArrowDown':
                event.preventDefault();
                this.setOpen(true, 'trigger');
                this.moveActiveOption(1);
                break;
            case 'ArrowUp':
                event.preventDefault();
                this.setOpen(true, 'trigger');
                this.moveActiveOption(-1);
                break;
            case 'Home':
                if (this.isOpen) {
                    event.preventDefault();
                    this.activateBoundaryOption('first');
                }
                break;
            case 'End':
                if (this.isOpen) {
                    event.preventDefault();
                    this.activateBoundaryOption('last');
                }
                break;
            case 'Escape':
                if (this.isOpen) {
                    event.preventDefault();
                    this.setOpen(false, 'escape');
                }
                break;
            default:
                break;
        }
    }
    handleEnterKey() {
        if (!this.isOpen) return void this.setOpen(true, 'trigger');
        const activeOption = this.activeOptionKey ? this.optionByKey.get(this.activeOptionKey) : void 0;
        if (activeOption) return void this.selectOption(activeOption);
    }
    activateInitialOption() {
        const selected = this.getSelectedOptions()[0];
        const selectedKey = selected ? toOptionKey(selected.value) : '';
        const candidate = this.filteredEnabledOptions.find(({ key })=>key === selectedKey);
        this.activeOptionKey = candidate?.key ?? this.filteredEnabledOptions[0]?.key ?? '';
    }
    ensureActiveOption() {
        if (this.filteredEnabledOptions.some(({ key })=>key === this.activeOptionKey)) return;
        this.activeOptionKey = this.filteredEnabledOptions[0]?.key ?? '';
    }
    moveActiveOption(offset) {
        if (0 === this.filteredEnabledOptions.length) {
            this.activeOptionKey = '';
            return;
        }
        const currentIndex = this.filteredEnabledOptions.findIndex(({ key })=>key === this.activeOptionKey);
        const startIndex = currentIndex >= 0 ? currentIndex : offset > 0 ? -1 : 0;
        const nextIndex = (startIndex + offset + this.filteredEnabledOptions.length) % this.filteredEnabledOptions.length;
        this.activeOptionKey = this.filteredEnabledOptions[nextIndex]?.key ?? '';
    }
    activateBoundaryOption(position) {
        this.activeOptionKey = 'first' === position ? this.filteredEnabledOptions[0]?.key ?? '' : this.filteredEnabledOptions.at(-1)?.key ?? '';
    }
    setupAutoUpdate() {
        if (!this.isConnected || !this.isOpen || !this.controlElement || !this.panelElement) return;
        this.teardownAutoUpdate();
        this.cleanupAutoUpdate = createFloatingAutoUpdate(this.fieldElement ?? this.controlElement, this.panelElement, ()=>this.scheduleUpdatePosition());
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    async updatePosition() {
        if (!this.isOpen || !this.isConnected || !this.controlElement || !this.panelElement) return;
        const requestId = ++this.positionRequestId;
        const referenceElement = this.fieldElement ?? this.controlElement;
        if (this.popupMatchSelectWidth) {
            const rect = referenceElement.getBoundingClientRect();
            this.panelElement.style.width = `${rect.width}px`;
        } else this.panelElement.style.width = '';
        const styles = window.getComputedStyle(this);
        const gapValue = Number.parseFloat(styles.getPropertyValue('--ve-select-panel-gap')) || 4;
        const result = await computeFloatingPanelPosition({
            reference: referenceElement,
            floating: this.panelElement,
            placement: 'bottom-start',
            distance: gapValue
        });
        if (requestId !== this.positionRequestId || !this.isOpen || !this.isConnected || !this.panelElement) return;
        this.panelPlacement = result.placement;
        Object.assign(this.panelElement.style, {
            left: '0',
            top: '0',
            transform: `translate3d(${Math.round(result.x)}px, ${Math.round(result.y)}px, 0)`
        });
    }
    getOptionId(key) {
        return `${this.listboxId}-option-${CSS.escape(key)}`;
    }
    isSelected(option) {
        return this.selectedOptionKeySet.has(toOptionKey(option.value));
    }
    scrollActiveOptionIntoView() {
        if (!this.activeOptionKey || !this.panelElement) return;
        const optionElement = this.panelElement.querySelector(`#${CSS.escape(this.getOptionId(this.activeOptionKey))}`);
        optionElement?.scrollIntoView({
            block: 'nearest'
        });
    }
    handleTagRemoveClick = (event)=>{
        const option = this.getOptionFromKey(event.currentTarget?.dataset.optionKey);
        if (option) this.removeValue(option, event);
    };
    handleClearClick = (event)=>{
        this.clearValue(event);
    };
    handleOptionMouseDown = (event)=>{
        event.preventDefault();
    };
    handleOptionPointerMove = (event)=>{
        const key = event.currentTarget?.dataset.optionKey;
        const option = this.getOptionFromKey(key);
        if (key && option && !option.disabled) this.activeOptionKey = key;
    };
    handleOptionClick = (event)=>{
        const option = this.getOptionFromKey(event.currentTarget?.dataset.optionKey);
        if (option) this.selectOption(option);
    };
    handlePrefixSlotChange = (event)=>{
        const slot = event.currentTarget;
        const hasAssignedNodes = (slot?.assignedNodes({
            flatten: true
        }).filter((node)=>node.nodeType !== Node.TEXT_NODE || (node.textContent?.trim().length ?? 0) > 0).length ?? 0) > 0;
        this.hasPrefixSlot = hasAssignedNodes;
    };
    handleLabelClick = ()=>{
        if (this.disabled) return;
        this.controlElement?.focus();
        this.setOpen(!this.isOpen, 'trigger');
    };
    renderLabel() {
        const label = (this.label ?? '').trim();
        if (!label) return nothing;
        return html`
      <span
        class="ve-select-label"
        part="label"
        @click=${this.handleLabelClick}
      >
        <span class="ve-select-label-text">${label}</span>
      </span>
    `;
    }
    renderPrefix() {
        return html`<span
      class=${classMap({
            've-select-prefix': true,
            've-select-prefix-empty': !this.hasPrefixSlot
        })}
      ><slot name="prefix" @slotchange=${this.handlePrefixSlotChange}></slot
    ></span>`;
    }
    renderTag(option) {
        const closable = false !== option.closable && !this.disabled && !this.loading;
        return html`
      <span class="ve-select-tag">
        <span class="ve-select-tag-label">${option.label}</span>
        ${closable ? html`<button
              class="ve-select-tag-remove"
              type="button"
              aria-label=${`Remove ${option.label}`}
              data-option-key=${toOptionKey(option.value)}
              @click=${this.handleTagRemoveClick}
            >
              <ve-icon .definition=${closeIcon}></ve-icon>
            </button>` : nothing}
      </span>
    `;
    }
    renderMultipleValue() {
        const selected = this.getSelectedOptions();
        if (0 === selected.length) return html`<span
        class="ve-select-value-text ve-select-value-text-placeholder"
        >${this.placeholder}</span
      >`;
        let visible = selected;
        let overflowCount = 0;
        if ('number' == typeof this.maxTagCount) {
            visible = selected.slice(0, this.maxTagCount);
            overflowCount = Math.max(selected.length - visible.length, 0);
        }
        return html`
      ${visible.map((option)=>this.renderTag(option))}
      ${overflowCount > 0 ? html`<span class="ve-select-tag">+${overflowCount}</span>` : nothing}
    `;
    }
    renderSingleValue() {
        const selected = this.getSelectedOptions()[0];
        if (!selected) return html`<span
        class="ve-select-value-text ve-select-value-text-placeholder"
        >${this.placeholder}</span
      >`;
        return html`<span class="ve-select-value-text">${selected.label}</span>`;
    }
    renderValue() {
        return html`<span class="ve-select-value"
      >${this.isMultiple ? this.renderMultipleValue() : this.renderSingleValue()}</span
    >`;
    }
    renderActions() {
        return html`
      <span class="ve-select-actions">
        ${this.shouldShowClear ? html`<button
              class="ve-select-clear"
              part="clear"
              type="button"
              aria-label="Clear selection"
              @click=${this.handleClearClick}
            >
              <ve-icon .definition=${closeIcon}></ve-icon>
            </button>` : nothing}
        ${this.loading ? html`<span class="ve-select-loading" aria-hidden="true"
              ><ve-icon .definition=${loadingProgressIcon}></ve-icon
            ></span>` : html`<span class="ve-select-arrow" aria-hidden="true"
              ><ve-icon
                .definition=${this.isOpen ? chevronUpSmIcon : chevronDownSmIcon}
              ></ve-icon
            ></span>`}
      </span>
    `;
    }
    renderOptionCheckIcon() {
        return html`
      <svg
        class="ve-select-option-check-icon"
        aria-hidden="true"
        focusable="false"
        width="12"
        height="12"
        viewBox="0 0 12 12"
        fill="none"
      >
        <path
          d="M2 6l3 3 5-5"
          stroke="currentColor"
          stroke-width="1.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        ></path>
      </svg>
    `;
    }
    renderOptionContent(option) {
        const contentSlotName = this.optionContentSlotByKey.get(toOptionKey(option.value));
        if (contentSlotName) return html`
        <span class="ve-select-option-content">
          <slot name=${contentSlotName}></slot>
        </span>
      `;
        return html`<span class="ve-select-option-label">${option.label}</span>`;
    }
    renderOption(option) {
        const selected = this.isSelected(option);
        const key = toOptionKey(option.value);
        const active = key === this.activeOptionKey;
        return html`
      <li role="presentation">
        <button
          id=${this.getOptionId(key)}
          class=${classMap({
            've-select-option': true,
            've-select-option-active': active,
            've-select-option-selected': selected,
            've-select-option-disabled': true === option.disabled
        })}
          data-option-key=${key}
          part="option"
          type="button"
          role="option"
          tabindex="-1"
          aria-selected=${selected ? 'true' : 'false'}
          ?disabled=${true === option.disabled}
          @mousedown=${this.handleOptionMouseDown}
          @pointermove=${this.handleOptionPointerMove}
          @click=${this.handleOptionClick}
        >
          ${this.isMultiple ? html`<span
                class=${classMap({
            've-select-option-check': true,
            've-select-option-check-selected': selected
        })}
                aria-hidden="true"
              >
                ${selected ? this.renderOptionCheckIcon() : nothing}
              </span>` : nothing}
          ${this.renderOptionContent(option)}
          ${selected && !this.isMultiple ? html`<span
                class="ve-select-option-icon ve-select-option-icon-selected"
                aria-hidden="true"
                ><ve-icon .definition=${checkIcon}></ve-icon
              ></span>` : nothing}
        </button>
      </li>
    `;
    }
    renderFeedback() {
        const description = (this.description ?? '').trim();
        if (!description) return nothing;
        return html`
      <div class="ve-select-feedback">
        <div
          class="ve-select-description"
          id=${this.descriptionId}
          part="description"
        >
          ${description}
        </div>
      </div>
    `;
    }
    renderOptionsList(items) {
        if (this.loading) return html`<div class="ve-select-empty ve-select-empty-loading">
        Loading...
      </div>`;
        if (0 === items.length) return this.renderEmpty();
        return html`
      <div class="ve-select-panel-inner">
        <ul
          class="ve-select-options"
          id=${this.listboxId}
          role="listbox"
          aria-multiselectable=${ifDefined(this.isMultiple ? 'true' : void 0)}
        >
          ${items.map((item)=>{
            if (!isGroup(item)) return this.renderOption(item);
            return html`
              <li class="ve-select-group" role="presentation">
                <span class="ve-select-group-label">${item.label}</span>
                <ul
                  class="ve-select-group-options"
                  role="group"
                  aria-label=${item.label}
                >
                  ${item.options.map((option)=>this.renderOption(option))}
                </ul>
              </li>
            `;
        })}
        </ul>
      </div>
    `;
    }
    renderEmpty() {
        const hasEmptySlot = (this.emptySlot?.assignedNodes({
            flatten: true
        }).length ?? 0) > 0;
        return html`
      <div class="ve-select-empty">
        ${hasEmptySlot ? html`<slot name="empty"></slot>` : this.emptyText}
      </div>
    `;
    }
    renderPanel() {
        if (!this.isOpen) return nothing;
        return html`
      <div
        class="ve-select-panel"
        part="panel"
        popover=${"manual"}
        data-placement=${this.panelPlacement}
        style=${`--ve-select-panel-max-height: ${this.maxHeight}px;`}
      >
        ${this.renderOptionsList(this.filteredOptions)}
      </div>
    `;
    }
    render() {
        return html`
      <div
        class=${classMap({
            've-select': true,
            've-select-with-label': this.hasLeadingLabel,
            've-select-with-feedback': this.hasDescriptionFeedback
        })}
      >
        <div hidden aria-hidden="true">
          <slot @slotchange=${this.handleOptionsSourceChange}></slot>
        </div>
        <div class="ve-select-field">
          ${this.renderLabel()}
          <div
            class=${classMap({
            've-select-control': true,
            've-select-control-filled': this.isFilled,
            've-select-control-multiple': this.isMultiple
        })}
            part="control"
            tabindex=${this.disabled ? '-1' : '0'}
            role="combobox"
            aria-haspopup="listbox"
            aria-controls=${ifDefined(this.isOpen ? this.listboxId : void 0)}
            aria-activedescendant=${ifDefined(this.isOpen ? this.activeOptionId : void 0)}
            aria-expanded=${this.isOpen ? 'true' : 'false'}
            aria-disabled=${this.disabled ? 'true' : 'false'}
            aria-busy=${this.loading ? 'true' : 'false'}
            aria-label=${ifDefined(this.controlAriaLabel)}
            aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
            aria-describedby=${ifDefined(this.controlAriaDescribedby)}
            aria-required=${ifDefined(this.controlAriaRequired)}
            aria-invalid=${ifDefined(this.controlAriaInvalid)}
            @pointerdown=${this.handleControlPointerDown}
            @click=${this.handleControlClick}
            @keydown=${this.handleControlKeydown}
          >
            ${this.renderPrefix()} ${this.renderValue()} ${this.renderActions()}
          </div>
        </div>
        ${this.renderFeedback()} ${this.renderPanel()}
      </div>
    `;
    }
}
export { VE_SELECT_GROUP_TAG, VE_SELECT_ITEM_TAG, VE_SELECT_TAG, VeSelectElement, VeSelectGroupElement, VeSelectItemElement };
