// @ve-design/web/web/dist/77133.js
import { css, html, nothing, svg } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { query, state } from "lit/decorators.js";
import { componentReset, VeElement } from "./78928.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
const veEmptyTokenDefaults = css`
  :host {
    --ve-empty-gap: var(--space-s);
    --ve-empty-body-gap: var(--space-xxs);
    --ve-empty-actions-gap: var(--space-xs);

    --ve-empty-image-color: var(--color-text-disable);

    --ve-empty-title-font-size: var(--text-ui-strong);
    --ve-empty-title-font-weight: var(--font-weight-medium);
    --ve-empty-title-line-height: var(--line-height-body);
    --ve-empty-title-color: var(--color-text-primary);

    --ve-empty-description-font-size: var(--text-body);
    --ve-empty-description-font-weight: var(--font-weight-normal);
    --ve-empty-description-line-height: var(--line-height-body);
    --ve-empty-description-color: var(--color-text-tertiary);

    --ve-empty-font-family: var(--font-sans);
  }
`;
const token = veEmptyTokenDefaults;
const ve_empty = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      font-family: var(--ve-empty-font-family);
    }

    :host([hidden]) {
      display: none;
    }

    [hidden] {
      display: none !important;
    }

    .ve-empty {
      box-sizing: border-box;
      display: flex;
      flex-direction: column;
      align-items: center;
      gap: var(--ve-empty-gap);
      text-align: center;
    }

    .ve-empty-image {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: center;
      flex: none;
      max-inline-size: 100%;
      color: var(--ve-empty-image-color);
    }

    .ve-empty-default-illustration {
      display: block;
    }

    .ve-empty-image ::slotted(*) {
      display: block;
      flex: none;
      max-inline-size: 100%;
      max-block-size: 100%;
    }

    .ve-empty-body {
      display: flex;
      flex-direction: column;
      align-items: center;
      gap: var(--ve-empty-body-gap);
      inline-size: 100%;
    }

    .ve-empty-title {
      margin: 0;
      inline-size: 100%;
      color: var(--ve-empty-title-color);
      font-size: var(--ve-empty-title-font-size);
      font-weight: var(--ve-empty-title-font-weight);
      line-height: var(--ve-empty-title-line-height);
      word-break: break-word;
    }

    .ve-empty-description {
      margin: 0;
      inline-size: 100%;
      color: var(--ve-empty-description-color);
      font-size: var(--ve-empty-description-font-size);
      font-weight: var(--ve-empty-description-font-weight);
      line-height: var(--ve-empty-description-line-height);
      word-break: break-word;
    }

    .ve-empty-actions {
      display: flex;
      align-items: center;
      justify-content: center;
      gap: var(--ve-empty-actions-gap);
      flex-wrap: wrap;
    }

    .ve-empty-body[hidden],
    .ve-empty-title[hidden],
    .ve-empty-description[hidden],
    .ve-empty-actions[hidden] {
      display: none;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _init_descriptionSlot, _init_extra__init_descriptionSlot, _init_imageSlot, _init_extra__init_imageSlot, _init_titleSlot, _init_extra__init_titleSlot, _init_defaultSlot, _init_extra__init_defaultSlot, _init_actionsSlot, _init_extra__init_actionsSlot, _init_hasDescriptionSlot, _init_extra__init_hasDescriptionSlot, _init_hasImageSlot, _init_extra__init_hasImageSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasDefaultSlot, _init_extra__init_hasDefaultSlot, _init_hasActionsSlot, _init_extra__init_hasActionsSlot;
const VE_EMPTY_TAG = 've-empty';
_dec = query('slot[name="description"]'), _dec1 = query('slot[name="image"]'), _dec2 = query('slot[name="title"]'), _dec3 = query('slot:not([name])'), _dec4 = query('slot[name="actions"]'), _dec5 = state(), _dec6 = state(), _dec7 = state(), _dec8 = state(), _dec9 = state();
class VeEmptyElement extends VeElement {
    static{
        ({ e: [_init_descriptionSlot, _init_extra__init_descriptionSlot, _init_imageSlot, _init_extra__init_imageSlot, _init_titleSlot, _init_extra__init_titleSlot, _init_defaultSlot, _init_extra__init_defaultSlot, _init_actionsSlot, _init_extra__init_actionsSlot, _init_hasDescriptionSlot, _init_extra__init_hasDescriptionSlot, _init_hasImageSlot, _init_extra__init_hasImageSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasDefaultSlot, _init_extra__init_hasDefaultSlot, _init_hasActionsSlot, _init_extra__init_hasActionsSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "descriptionSlot"
            ],
            [
                _dec1,
                1,
                "imageSlot"
            ],
            [
                _dec2,
                1,
                "titleSlot"
            ],
            [
                _dec3,
                1,
                "defaultSlot"
            ],
            [
                _dec4,
                1,
                "actionsSlot"
            ],
            [
                _dec5,
                1,
                "hasDescriptionSlot"
            ],
            [
                _dec6,
                1,
                "hasImageSlot"
            ],
            [
                _dec7,
                1,
                "hasTitleSlot"
            ],
            [
                _dec8,
                1,
                "hasDefaultSlot"
            ],
            [
                _dec9,
                1,
                "hasActionsSlot"
            ]
        ]));
    }
    static styles = ve_empty;
    #___private_descriptionSlot_1 = (()=>{
        const _value = _init_descriptionSlot(this, null);
        _init_extra__init_descriptionSlot(this);
        return _value;
    })();
    get descriptionSlot() {
        return this.#___private_descriptionSlot_1;
    }
    set descriptionSlot(_v) {
        this.#___private_descriptionSlot_1 = _v;
    }
    #___private_imageSlot_2 = (()=>{
        const _value = _init_imageSlot(this, null);
        _init_extra__init_imageSlot(this);
        return _value;
    })();
    get imageSlot() {
        return this.#___private_imageSlot_2;
    }
    set imageSlot(_v) {
        this.#___private_imageSlot_2 = _v;
    }
    #___private_titleSlot_3 = (()=>{
        const _value = _init_titleSlot(this, null);
        _init_extra__init_titleSlot(this);
        return _value;
    })();
    get titleSlot() {
        return this.#___private_titleSlot_3;
    }
    set titleSlot(_v) {
        this.#___private_titleSlot_3 = _v;
    }
    #___private_defaultSlot_4 = (()=>{
        const _value = _init_defaultSlot(this, null);
        _init_extra__init_defaultSlot(this);
        return _value;
    })();
    get defaultSlot() {
        return this.#___private_defaultSlot_4;
    }
    set defaultSlot(_v) {
        this.#___private_defaultSlot_4 = _v;
    }
    #___private_actionsSlot_5 = (()=>{
        const _value = _init_actionsSlot(this, null);
        _init_extra__init_actionsSlot(this);
        return _value;
    })();
    get actionsSlot() {
        return this.#___private_actionsSlot_5;
    }
    set actionsSlot(_v) {
        this.#___private_actionsSlot_5 = _v;
    }
    #___private_hasDescriptionSlot_6 = (()=>{
        const _value = _init_hasDescriptionSlot(this, false);
        _init_extra__init_hasDescriptionSlot(this);
        return _value;
    })();
    get hasDescriptionSlot() {
        return this.#___private_hasDescriptionSlot_6;
    }
    set hasDescriptionSlot(_v) {
        this.#___private_hasDescriptionSlot_6 = _v;
    }
    #___private_hasImageSlot_7 = (()=>{
        const _value = _init_hasImageSlot(this, false);
        _init_extra__init_hasImageSlot(this);
        return _value;
    })();
    get hasImageSlot() {
        return this.#___private_hasImageSlot_7;
    }
    set hasImageSlot(_v) {
        this.#___private_hasImageSlot_7 = _v;
    }
    #___private_hasTitleSlot_8 = (()=>{
        const _value = _init_hasTitleSlot(this, false);
        _init_extra__init_hasTitleSlot(this);
        return _value;
    })();
    get hasTitleSlot() {
        return this.#___private_hasTitleSlot_8;
    }
    set hasTitleSlot(_v) {
        this.#___private_hasTitleSlot_8 = _v;
    }
    #___private_hasDefaultSlot_9 = (()=>{
        const _value = _init_hasDefaultSlot(this, false);
        _init_extra__init_hasDefaultSlot(this);
        return _value;
    })();
    get hasDefaultSlot() {
        return this.#___private_hasDefaultSlot_9;
    }
    set hasDefaultSlot(_v) {
        this.#___private_hasDefaultSlot_9 = _v;
    }
    #___private_hasActionsSlot_10 = (()=>{
        const _value = _init_hasActionsSlot(this, false);
        _init_extra__init_hasActionsSlot(this);
        return _value;
    })();
    get hasActionsSlot() {
        return this.#___private_hasActionsSlot_10;
    }
    set hasActionsSlot(_v) {
        this.#___private_hasActionsSlot_10 = _v;
    }
    get showDescription() {
        return this.hasDescriptionSlot || this.hasDefaultSlot;
    }
    get showBody() {
        return true;
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.syncSlotState();
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
        this.syncSlotState();
    }
    syncSlotState() {
        this.hasDescriptionSlot = hasAssignedSlotContent(this.descriptionSlot) || hasNamedSlottedContent(this, "description");
        this.hasImageSlot = hasAssignedSlotContent(this.imageSlot) || hasNamedSlottedContent(this, 'image');
        this.hasTitleSlot = hasAssignedSlotContent(this.titleSlot) || hasNamedSlottedContent(this, 'title');
        this.hasDefaultSlot = hasAssignedSlotContent(this.defaultSlot);
        this.hasActionsSlot = hasAssignedSlotContent(this.actionsSlot) || hasNamedSlottedContent(this, 'actions');
    }
    handleSlotChange() {
        this.syncSlotState();
    }
    renderDefaultIllustration() {
        return svg`
      <svg
        class="ve-empty-default-illustration"
        width="184"
        height="152"
        viewBox="0 0 184 152"
        xmlns="http://www.w3.org/2000/svg"
        role="img"
        aria-label="No Data"
      >
        <title>No Data</title>
        <g fill="none" fill-rule="evenodd">
          <g transform="translate(24 31.7)">
            <ellipse fill-opacity=".8" fill="#F5F5F7" cx="67.8" cy="106.9" rx="67.8" ry="12.7"></ellipse>
            <path fill="#aeb8c2" d="M122 69.7 98.1 40.2a6 6 0 0 0-4.6-2.2H42.1a6 6 0 0 0-4.6 2.2l-24 29.5V85H122z"></path>
            <path fill="#f5f5f7" d="M33.8 0h68a4 4 0 0 1 4 4v93.3a4 4 0 0 1-4 4h-68a4 4 0 0 1-4-4V4a4 4 0 0 1 4-4"></path>
            <path fill="#dce0e6" d="M42.7 10h50.2a2 2 0 0 1 2 2v25a2 2 0 0 1-2 2H42.7a2 2 0 0 1-2-2V12a2 2 0 0 1 2-2m.2 39.8h49.8a2.3 2.3 0 1 1 0 4.5H42.9a2.3 2.3 0 0 1 0-4.5m0 11.7h49.8a2.3 2.3 0 1 1 0 4.6H42.9a2.3 2.3 0 0 1 0-4.6m79 43.5a7 7 0 0 1-6.8 5.4H20.5a7 7 0 0 1-6.7-5.4l-.2-1.8V69.7h26.3c2.9 0 5.2 2.4 5.2 5.4s2.4 5.4 5.3 5.4h34.8c2.9 0 5.3-2.4 5.3-5.4s2.3-5.4 5.2-5.4H122v33.5q0 1-.2 1.8"></path>
          </g>
          <path fill="#dce0e6" d="m149.1 33.3-6.8 2.6a1 1 0 0 1-1.3-1.2l2-6.2q-4.1-4.5-4.2-10.4c0-10 10.1-18.1 22.6-18.1S184 8.1 184 18.1s-10.1 18-22.6 18q-6.8 0-12.3-2.8"></path>
          <g fill="#fff" transform="translate(149.7 15.4)">
            <circle cx="20.7" cy="3.2" r="2.8"></circle>
            <path d="M5.7 5.6H0L2.9.7zM9.3.7h5v5h-5z"></path>
          </g>
        </g>
      </svg>
    `;
    }
    renderDescriptionFallback() {
        return html` <slot @slotchange=${this.handleSlotChange}></slot> `;
    }
    render() {
        const emptyClass = classMap({
            've-empty': true,
            've-empty-with-title': true,
            "ve-empty-with-description": this.showDescription,
            've-empty-with-actions': this.hasActionsSlot
        });
        return html`
      <div class=${emptyClass} part="base">
        <div class="ve-empty-image" part="image">
          <slot name="image" @slotchange=${this.handleSlotChange}></slot>
          ${this.hasImageSlot ? nothing : this.renderDefaultIllustration()}
        </div>

        <div class="ve-empty-body" part="body" ?hidden=${!this.showBody}>
          <div class="ve-empty-title" part="title">
            <slot name="title" @slotchange=${this.handleSlotChange}></slot>
            ${this.hasTitleSlot ? nothing : 'No Data'}
          </div>

          <div
            class="ve-empty-description"
            part="description"
            ?hidden=${!this.showDescription}
          >
            <slot
              name="description"
              @slotchange=${this.handleSlotChange}
            ></slot>
            ${this.hasDescriptionSlot ? nothing : this.renderDescriptionFallback()}
          </div>
        </div>

        <div
          class="ve-empty-actions"
          part="actions"
          ?hidden=${!this.hasActionsSlot}
        >
          <slot name="actions" @slotchange=${this.handleSlotChange}></slot>
        </div>
      </div>
    `;
    }
}
export { VE_EMPTY_TAG, VeEmptyElement };
