// @ve-design/web/web/dist/78928.js
import { LitElement, css } from "lit";
class VeElement extends LitElement {
    static dependencies;
    static shadowRootOptions = {
        ...LitElement.shadowRootOptions,
        delegatesFocus: true
    };
}
function defineCustomElementInternal(tagName, elementClass, visited) {
    if ("u" < typeof customElements) return;
    if (customElements.get(tagName) || visited.has(tagName)) return;
    visited.add(tagName);
    for (const [dependencyTagName, dependencyClass] of Object.entries(elementClass.dependencies ?? {}))defineCustomElementInternal(dependencyTagName, dependencyClass, visited);
    if (!customElements.get(tagName)) customElements.define(tagName, elementClass);
}
function defineCustomElement(tagName, elementClass) {
    defineCustomElementInternal(tagName, elementClass, new Set());
}
const componentReset = css`
  :host {
    box-sizing: border-box;
    line-height: 1.15;
    -webkit-text-size-adjust: 100%;
  }

  :host *,
  :host *::before,
  :host *::after {
    box-sizing: inherit;
  }

  :host([hidden]) {
    display: none !important;
  }

  main {
    display: block;
  }

  h1 {
    font-size: 2em;
    margin: 0.67em 0;
  }

  hr {
    box-sizing: content-box;
    height: 0;
    overflow: visible;
  }

  pre {
    font-family: monospace, monospace;
    font-size: 1em;
  }

  a {
    background-color: transparent;
  }

  abbr[title] {
    border-bottom: none;
    text-decoration: underline;
    text-decoration: underline dotted;
  }

  b,
  strong {
    font-weight: bolder;
  }

  code,
  kbd,
  samp {
    font-family: monospace, monospace;
    font-size: 1em;
  }

  small {
    font-size: 80%;
  }

  sub,
  sup {
    position: relative;
    font-size: 75%;
    line-height: 0;
    vertical-align: baseline;
  }

  sub {
    bottom: -0.25em;
  }

  sup {
    top: -0.5em;
  }

  img {
    border-style: none;
  }

  button,
  input,
  optgroup,
  select,
  textarea {
    margin: 0;
    font-family: inherit;
    font-size: 100%;
    line-height: 1.15;
  }

  button,
  input {
    overflow: visible;
  }

  button,
  select {
    text-transform: none;
  }

  button,
  [type='button'],
  [type='reset'],
  [type='submit'] {
    -webkit-appearance: button;
  }

  button::-moz-focus-inner,
  [type='button']::-moz-focus-inner,
  [type='reset']::-moz-focus-inner,
  [type='submit']::-moz-focus-inner {
    padding: 0;
    border-style: none;
  }

  button:-moz-focusring,
  [type='button']:-moz-focusring,
  [type='reset']:-moz-focusring,
  [type='submit']:-moz-focusring {
    outline: 1px dotted ButtonText;
  }

  fieldset {
    padding: 0.35em 0.75em 0.625em;
  }

  legend {
    box-sizing: border-box;
    display: table;
    max-width: 100%;
    padding: 0;
    color: inherit;
    white-space: normal;
  }

  progress {
    vertical-align: baseline;
  }

  textarea {
    overflow: auto;
  }

  [type='checkbox'],
  [type='radio'] {
    box-sizing: border-box;
    padding: 0;
  }

  [type='number']::-webkit-inner-spin-button,
  [type='number']::-webkit-outer-spin-button {
    height: auto;
  }

  [type='search'] {
    -webkit-appearance: textfield;
    outline-offset: -2px;
  }

  [type='search']::-webkit-search-decoration {
    -webkit-appearance: none;
  }

  ::-webkit-file-upload-button {
    font: inherit;
    -webkit-appearance: button;
  }

  details {
    display: block;
  }

  summary {
    display: list-item;
  }

  template {
    display: none;
  }

  [hidden] {
    display: none;
  }

  :host {
    --ve-scrollbar-size: 6px;
    --ve-scrollbar-radius: 999px;
    --ve-scrollbar-thumb-color: var(--color-border-strong, #b8b7c3);
    --ve-scrollbar-thumb-background: color-mix(
      in srgb,
      var(--ve-scrollbar-thumb-color) 50%,
      transparent
    );
    --ve-scrollbar-track-color: transparent;
  }

  :host *,
  :host *::before,
  :host *::after {
    scrollbar-width: thin;
    scrollbar-color: var(--ve-scrollbar-thumb-background)
      var(--ve-scrollbar-track-color);
  }

  :host *::-webkit-scrollbar {
    inline-size: var(--ve-scrollbar-size);
    block-size: var(--ve-scrollbar-size);
  }

  :host *::-webkit-scrollbar-thumb {
    border-radius: var(--ve-scrollbar-radius);
    background: var(--ve-scrollbar-thumb-background);
  }

  :host *::-webkit-scrollbar-track,
  :host *::-webkit-scrollbar-corner {
    background: var(--ve-scrollbar-track-color);
  }
`;
export { VeElement, componentReset, defineCustomElement };
