// @ve-design/web/web/dist/97605.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { checkIcon } from "./76270.js";
import { createFloatingAutoUpdate, getFloatingDistance, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
const veDropdownTokenDefaults = css`
  :host {
    --ve-dropdown-z-index: 1000;
    --ve-dropdown-offset: var(--space-xxxs);

    --ve-dropdown-min-width: 200px;
    --ve-dropdown-max-width: min(400px, calc(100vw - var(--space-xxs)));
    --ve-dropdown-max-height: none;
    --ve-dropdown-padding-block: var(--space-xxs);
    --ve-dropdown-padding-inline: var(--space-xxs);
    --ve-dropdown-radius: var(--radius-lg);
    --ve-dropdown-border-width: var(--stroke-0_5);
    --ve-dropdown-border-color: var(--color-border-default);
    --ve-dropdown-background: var(--color-bg-overlay);
    --ve-dropdown-shadow: var(--shadow-md);
    --ve-dropdown-menu-gap: 0px;

    --ve-dropdown-trigger-focus-ring: var(--color-text-accent);
    --ve-dropdown-disabled-opacity: var(--state-disabled-ratio);
    --ve-dropdown-enter-offset: var(--space-xxxs);

    --ve-dropdown-item-min-height: 38px;
    --ve-dropdown-item-padding-block: var(--space-xxs);
    --ve-dropdown-item-padding-inline: var(--space-xxs);
    --ve-dropdown-item-font-size: var(--text-body);
    --ve-dropdown-item-font-weight: var(--font-weight-normal);
    --ve-dropdown-item-label-line-height: 22px;
    --ve-dropdown-item-radius: var(--radius-md);
    --ve-dropdown-item-gap: var(--space-xxs);
    --ve-dropdown-item-color: var(--color-text-primary);
    --ve-dropdown-item-danger-color: var(--color-text-danger);
    --ve-dropdown-item-disabled-color: var(--color-text-disable);
    --ve-dropdown-item-extra-color: var(--color-text-tertiary);
    --ve-dropdown-item-extra-margin-inline-start: var(--space-xs);
    --ve-dropdown-item-selected-icon-color: currentColor;
    --ve-dropdown-item-selected-icon-size: 16px;
    --ve-dropdown-item-indicator-size: 16px;
    --ve-dropdown-item-selected-bg: transparent;
    --ve-dropdown-item-hover-bg: var(--color-bg-surface);
    --ve-dropdown-item-active-bg: color-mix(
      var(--color-bg-overlay) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-dropdown-item-danger-surface: color-mix(
      var(--color-bg-overlay) 92%,
      var(--color-bg-danger)
    );
    --ve-dropdown-item-danger-hover-bg: color-mix(
      var(--ve-dropdown-item-danger-surface) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-dropdown-item-danger-active-bg: color-mix(
      var(--ve-dropdown-item-danger-surface) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );
    --ve-dropdown-item-focus-shadow: 0 0 0 var(--stroke-2)
      var(--color-bg-primary) inset;
    --ve-dropdown-item-checkbox-size: 16px;
    --ve-dropdown-item-checkbox-border-width: var(
      --ve-checkbox-mask-border-width,
      var(--stroke-weight-strong)
    );
    --ve-dropdown-item-checkbox-radius: var(--radius-xs);
    --ve-dropdown-item-checkbox-border-color: var(--color-border-strong);
    --ve-dropdown-item-checkbox-background: var(--color-bg-base);
    --ve-dropdown-item-checkbox-check-color: var(--color-text-foreground);
    --ve-dropdown-item-checkbox-check-width: 8px;
    --ve-dropdown-item-checkbox-check-height: 5px;
    --ve-dropdown-item-checkbox-check-stroke: 2px;
    --ve-dropdown-item-checkbox-selected-border-color: var(
      --ve-checkbox-mask-color-border-checked,
      transparent
    );
    --ve-dropdown-item-checkbox-selected-background: var(
      --ve-checkbox-mask-color-bg-checked,
      var(--color-bg-primary)
    );
    --ve-dropdown-item-switch-width: 24px;
    --ve-dropdown-item-switch-height: 14px;
    --ve-dropdown-item-switch-border-width: var(--stroke-0_5);
    --ve-dropdown-item-switch-radius: 99px;
    --ve-dropdown-item-switch-border-color: var(--color-border-default);
    --ve-dropdown-item-switch-background: var(--color-gray-200);
    --ve-dropdown-item-switch-selected-border-color: var(--color-bg-primary);
    --ve-dropdown-item-switch-selected-background: var(--color-bg-primary);
    --ve-dropdown-item-switch-thumb-size: 10px;
    --ve-dropdown-item-switch-thumb-offset: 1.5px;
    --ve-dropdown-item-switch-thumb-background: var(--color-white);
    --ve-dropdown-item-switch-thumb-shadow: 0 1px 2px rgba(16, 16, 19, 0.18);

    --ve-dropdown-group-title-padding-block: 0px;
    --ve-dropdown-group-title-padding-inline: var(--space-xs);
    --ve-dropdown-group-title-margin-top: var(--space-xxxs);
    --ve-dropdown-group-title-height: 20px;
    --ve-dropdown-group-title-color: var(--color-text-tertiary);
    --ve-dropdown-group-title-font-size: var(--text-caption);
    --ve-dropdown-group-title-line-height: var(
      --ve-dropdown-group-title-height
    );
    --ve-dropdown-group-title-font-weight: var(--font-weight-normal);
    --ve-dropdown-group-gap: 0px;

    --ve-dropdown-divider-margin-block: var(--space-xxxs);
    --ve-dropdown-divider-margin-inline: var(--space-xxxs);
    --ve-dropdown-divider-width: var(--stroke-0_5);
    --ve-dropdown-divider-color: var(--color-border-default);
    --ve-dropdown-sub-item-gap: var(--space-xxxs);
    --ve-dropdown-sub-item-arrow-color: var(--ve-dropdown-item-extra-color);
    --ve-dropdown-sub-item-arrow-size: 16px;
    --ve-dropdown-sub-item-arrow-icon-size: 6px;
    --ve-dropdown-sub-item-arrow-stroke: var(--stroke-weight-strong);
    --ve-dropdown-sub-item-arrow-offset-y: 0px;

    --ve-dropdown-transition-duration: var(--motion-fast);
    --ve-dropdown-transition-easing: cubic-bezier(0.34, 0.69, 0.1, 1);
    --ve-dropdown-current-transform-origin: 0% 0%;

    --ve-dropdown-current-enter-transform: translate3d(
      0,
      var(--ve-dropdown-enter-offset),
      0
    );
  }
`;
const token = veDropdownTokenDefaults;
const ve_dropdown_item = [
    componentReset,
    token,
    css`
    :host {
      display: block;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-dropdown-item {
      appearance: none;
      display: flex;
      align-items: center;
      justify-content: space-between;
      inline-size: 100%;
      min-block-size: var(--ve-dropdown-item-min-height);
      box-sizing: border-box;
      padding-block: var(--ve-dropdown-item-padding-block);
      padding-inline: var(--ve-dropdown-item-padding-inline);
      gap: var(--ve-dropdown-item-gap);
      border: none;
      border-radius: var(--ve-dropdown-item-radius);
      background: transparent;
      color: var(--ve-dropdown-item-color);
      font-family: var(--font-sans);
      font-size: var(--ve-dropdown-item-font-size);
      font-weight: var(--ve-dropdown-item-font-weight);
      line-height: var(--ve-dropdown-item-label-line-height);
      text-align: left;
      cursor: pointer;
      z-index: 1;
      transition:
        background-color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing);
    }

    .ve-dropdown-item-content {
      min-inline-size: 0;
      display: block;
      flex: 1;
      overflow: hidden;
      text-align: left;
    }

    .ve-dropdown-item-extra {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxs);
      flex: none;
      color: var(--ve-dropdown-item-extra-color);
      margin-inline-start: var(--ve-dropdown-item-extra-margin-inline-start);
      min-block-size: var(--ve-dropdown-item-label-line-height);
    }

    .ve-dropdown-item-check {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-dropdown-item-indicator-size);
      block-size: var(--ve-dropdown-item-indicator-size);
      color: var(--ve-dropdown-item-selected-icon-color);
      font-size: var(--ve-dropdown-item-selected-icon-size);
      line-height: 1;
    }

    .ve-dropdown-item-checkbox {
      position: relative;
      box-sizing: border-box;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-dropdown-item-checkbox-size);
      block-size: var(--ve-dropdown-item-checkbox-size);
      flex: none;
      border: var(--ve-dropdown-item-checkbox-border-width) solid
        var(--ve-dropdown-item-checkbox-border-color);
      border-radius: var(--ve-dropdown-item-checkbox-radius);
      background: var(--ve-dropdown-item-checkbox-background);
      color: var(--ve-dropdown-item-checkbox-check-color);
      transition:
        background-color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        border-color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing);
    }

    .ve-dropdown-item-checkbox::after {
      content: '';
      display: none;
      inline-size: var(--ve-dropdown-item-checkbox-check-width);
      block-size: var(--ve-dropdown-item-checkbox-check-height);
      border-color: currentColor;
      border-style: solid;
      border-width: 0 0 var(--ve-dropdown-item-checkbox-check-stroke)
        var(--ve-dropdown-item-checkbox-check-stroke);
      transform: rotate(-45deg) translateY(-1px);
      transform-origin: center;
    }

    .ve-dropdown-item-switch {
      box-sizing: border-box;
      position: relative;
      display: inline-flex;
      align-items: center;
      inline-size: var(--ve-dropdown-item-switch-width);
      block-size: var(--ve-dropdown-item-switch-height);
      flex: none;
      border: var(--ve-dropdown-item-switch-border-width) solid
        var(--ve-dropdown-item-switch-border-color);
      border-radius: var(--ve-dropdown-item-switch-radius);
      background: var(--ve-dropdown-item-switch-background);
      transition:
        background-color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        border-color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing);
    }

    .ve-dropdown-item-switch-thumb {
      position: absolute;
      inset-block-start: var(--ve-dropdown-item-switch-thumb-offset);
      inset-inline-start: var(--ve-dropdown-item-switch-thumb-offset);
      inline-size: var(--ve-dropdown-item-switch-thumb-size);
      block-size: var(--ve-dropdown-item-switch-thumb-size);
      border-radius: 50%;
      background: var(--ve-dropdown-item-switch-thumb-background);
      box-shadow: var(--ve-dropdown-item-switch-thumb-shadow);
      transition: inset-inline-start var(--ve-dropdown-transition-duration)
        var(--ve-dropdown-transition-easing);
    }

    :host([selected]) .ve-dropdown-item-checkbox {
      border-color: var(--ve-dropdown-item-checkbox-selected-border-color);
      background: var(--ve-dropdown-item-checkbox-selected-background);
    }

    :host([selected]) .ve-dropdown-item-checkbox::after {
      display: block;
    }

    :host([selected]) .ve-dropdown-item-switch {
      border-color: var(--ve-dropdown-item-switch-selected-border-color);
      background: var(--ve-dropdown-item-switch-selected-background);
    }

    :host([selected]) .ve-dropdown-item-switch-thumb {
      inset-inline-start: calc(
        var(--ve-dropdown-item-switch-width) - var(
            --ve-dropdown-item-switch-thumb-size
          ) - var(--ve-dropdown-item-switch-thumb-offset) - var(
            --ve-dropdown-item-switch-border-width
          ) *
          2
      );
    }

    .ve-dropdown-item-check[hidden] {
      display: none;
    }

    .ve-dropdown-item.is-selected {
      background: var(--ve-dropdown-item-selected-bg);
    }

    .ve-dropdown-item:hover {
      background: var(--ve-dropdown-item-hover-bg);
    }

    .ve-dropdown-item:active {
      background: var(--ve-dropdown-item-active-bg);
    }

    .ve-dropdown-item:focus-visible {
      outline: none;
      box-shadow: var(--ve-dropdown-item-focus-shadow);
    }

    :host([danger]) .ve-dropdown-item {
      color: var(--ve-dropdown-item-danger-color);
    }

    :host([danger]) .ve-dropdown-item:hover,
    :host([danger]) .ve-dropdown-item:focus-visible {
      background: var(--ve-dropdown-item-danger-hover-bg);
    }

    :host([danger]) .ve-dropdown-item:active {
      background: var(--ve-dropdown-item-danger-active-bg);
    }

    :host([disabled]) .ve-dropdown-item,
    :host([disabled]) .ve-dropdown-item:hover,
    :host([disabled]) .ve-dropdown-item:active,
    :host([disabled]) .ve-dropdown-item:focus-visible {
      color: var(--ve-dropdown-item-disabled-color);
      background: transparent;
      cursor: not-allowed;
      opacity: var(--ve-dropdown-disabled-opacity);
      pointer-events: none;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _init_value, _init_extra__init_value, _init_disabled, _init_extra__init_disabled, _init_danger, _init_extra__init_danger, _init_selected, _init_extra__init_selected, _init_selection, _init_extra__init_selection, _init_buttonElement, _init_extra__init_buttonElement;
const VE_DROPDOWN_ITEM_TAG = 've-dropdown-item';
const VE_DROPDOWN_ITEM_SELECT_EVENT = 've-dropdown-item-select';
_dec = property({
    type: String
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec5 = query('.ve-dropdown-item');
class VeDropdownItemElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, _init_disabled, _init_extra__init_disabled, _init_danger, _init_extra__init_danger, _init_selected, _init_extra__init_selected, _init_selection, _init_extra__init_selection, _init_buttonElement, _init_extra__init_buttonElement] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "value"
            ],
            [
                _dec1,
                1,
                "disabled"
            ],
            [
                _dec2,
                1,
                "danger"
            ],
            [
                _dec3,
                1,
                "selected"
            ],
            [
                _dec4,
                1,
                "selection"
            ],
            [
                _dec5,
                1,
                "buttonElement"
            ]
        ]));
    }
    static styles = ve_dropdown_item;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_value_1 = (()=>{
        const _value = _init_value(this, '');
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_disabled_2 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_2;
    }
    set disabled(_v) {
        this.#___private_disabled_2 = _v;
    }
    #___private_danger_3 = (()=>{
        const _value = _init_danger(this, false);
        _init_extra__init_danger(this);
        return _value;
    })();
    get danger() {
        return this.#___private_danger_3;
    }
    set danger(_v) {
        this.#___private_danger_3 = _v;
    }
    #___private_selected_4 = (()=>{
        const _value = _init_selected(this, false);
        _init_extra__init_selected(this);
        return _value;
    })();
    get selected() {
        return this.#___private_selected_4;
    }
    set selected(_v) {
        this.#___private_selected_4 = _v;
    }
    #___private_selection_5 = (()=>{
        const _value = _init_selection(this, 'none');
        _init_extra__init_selection(this);
        return _value;
    })();
    get selection() {
        return this.#___private_selection_5;
    }
    set selection(_v) {
        this.#___private_selection_5 = _v;
    }
    #___private_buttonElement_6 = (()=>{
        const _value = _init_buttonElement(this);
        _init_extra__init_buttonElement(this);
        return _value;
    })();
    get buttonElement() {
        return this.#___private_buttonElement_6;
    }
    set buttonElement(_v) {
        this.#___private_buttonElement_6 = _v;
    }
    get textLabel() {
        return (this.textContent ?? '').trim();
    }
    get roleType() {
        if ('checkbox' === this.selection || 'switch' === this.selection) return 'menuitemcheckbox';
        if ('check' === this.selection) return 'menuitemradio';
        return 'menuitem';
    }
    get isSelectable() {
        return 'none' !== this.selection;
    }
    focus(options) {
        this.buttonElement?.focus(options);
    }
    commitSelectionState() {
        if ('checkbox' === this.selection || 'switch' === this.selection) {
            this.selected = !this.selected;
            return;
        }
        if ('check' === this.selection) this.selected = true;
    }
    select() {
        if (this.disabled) return;
        this.commitSelectionState();
        const value = (this.value ?? '').trim().length > 0 ? this.value : this.textLabel;
        this.dispatchEvent(new CustomEvent(VE_DROPDOWN_ITEM_SELECT_EVENT, {
            detail: {
                value,
                label: this.textLabel,
                item: this,
                selection: this.selection,
                selected: this.selected
            },
            bubbles: true,
            composed: true
        }));
    }
    handleClick = (event)=>{
        event.stopPropagation();
        this.select();
    };
    handleKeydown = (event)=>{
        if ('Enter' !== event.key && ' ' !== event.key) return;
        event.preventDefault();
        this.select();
    };
    render() {
        const showCheckIndicator = this.selected && 'check' === this.selection;
        const showCheckboxIndicator = 'checkbox' === this.selection;
        const showSwitchIndicator = 'switch' === this.selection;
        const hasExtra = showCheckIndicator || showSwitchIndicator;
        const itemClass = classMap({
            've-dropdown-item': true,
            'is-selected': this.selected,
            [`is-selection-${this.selection}`]: true
        });
        return html`
      <button
        class=${itemClass}
        part="button"
        type="button"
        tabindex="-1"
        role=${this.roleType}
        aria-checked=${this.isSelectable ? this.selected ? 'true' : 'false' : nothing}
        ?disabled=${this.disabled}
        aria-disabled=${this.disabled ? 'true' : 'false'}
        @click=${this.handleClick}
        @keydown=${this.handleKeydown}
      >
        ${showCheckboxIndicator ? html`<span
              class="ve-dropdown-item-checkbox"
              part="checkbox"
              aria-hidden="true"
            ></span>` : nothing}

        <span class="ve-dropdown-item-content">
          <slot></slot>
        </span>

        ${hasExtra ? html`<span class="ve-dropdown-item-extra">
              <span
                class="ve-dropdown-item-check"
                aria-hidden="true"
                ?hidden=${!showCheckIndicator}
              >
                <ve-icon .definition=${checkIcon}></ve-icon>
              </span>
              ${showSwitchIndicator ? html`<span
                    class="ve-dropdown-item-switch"
                    part="switch"
                    aria-hidden="true"
                  >
                    <span class="ve-dropdown-item-switch-thumb"></span>
                  </span>` : nothing}
            </span>` : nothing}
      </button>
    `;
    }
}
const MENU_NODE_TAGS = [
    've-dropdown-item',
    've-dropdown-sub-item'
];
function isMenuNodeCandidate(element) {
    return element instanceof HTMLElement && MENU_NODE_TAGS.some((tagName)=>element.matches(tagName));
}
function appendMenuNode(candidate, nodes, visited) {
    if (!isMenuNodeCandidate(candidate) || candidate.hasAttribute('disabled')) return;
    if (visited.has(candidate)) return;
    visited.add(candidate);
    nodes.push(candidate);
}
function hasAssignedSlotContent(slot) {
    const assignedNodes = slot?.assignedNodes({
        flatten: true
    }) ?? [];
    return assignedNodes.some((node)=>{
        if (node.nodeType === Node.ELEMENT_NODE) return true;
        if (node.nodeType === Node.TEXT_NODE) return (node.textContent ?? '').trim().length > 0;
        return false;
    });
}
function focusElement(element) {
    if (!element) return;
    element.focus({
        preventScroll: true
    });
    element.scrollIntoView({
        block: 'nearest',
        inline: 'nearest'
    });
}
function getAssignedMenuNodes(slot) {
    const assignedElements = slot?.assignedElements({
        flatten: true
    }) ?? [];
    const nodes = [];
    const visited = new Set();
    assignedElements.forEach((element)=>{
        const isDirectMenuNode = isMenuNodeCandidate(element);
        appendMenuNode(element, nodes, visited);
        if (isDirectMenuNode) return;
        if (!(element instanceof HTMLElement)) return;
        Array.from(element.children).forEach((child)=>{
            appendMenuNode(child, nodes, visited);
        });
    });
    return nodes;
}
const ve_dropdown = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
      color-scheme: light;
    }

    :host([hidden]) {
      display: none;
    }

    :host([block]) {
      display: flex;
      inline-size: 100%;
    }

    :host([block]) .ve-dropdown {
      display: flex;
      inline-size: 100%;
    }

    :host([block]) .ve-dropdown-trigger {
      inline-size: 100%;
    }

    .ve-dropdown {
      position: relative;
      display: inline-flex;
      max-inline-size: 100%;
    }

    .ve-dropdown-trigger {
      display: inline-flex;
      align-items: inherit;
      justify-content: inherit;
      inline-size: fit-content;
      max-inline-size: 100%;
      outline: none;
    }

    :host(:not([disabled])) .ve-dropdown-trigger:focus-visible {
      border-radius: var(--ve-dropdown-radius);
      box-shadow: 0 0 0 var(--stroke-2)
        color-mix(var(--ve-dropdown-trigger-focus-ring) 24%, transparent);
    }

    :host([disabled]) .ve-dropdown-trigger {
      cursor: not-allowed;
    }

    .ve-dropdown-panel {
      position: fixed;
      inset: auto;
      margin: 0;
      padding: 0;
      z-index: var(--ve-dropdown-z-index);
      box-sizing: border-box;
      min-inline-size: var(--ve-dropdown-min-width);
      max-inline-size: var(--ve-dropdown-max-width);
      border: var(--ve-dropdown-border-width) solid
        var(--ve-dropdown-border-color);
      border-radius: var(--ve-dropdown-radius);
      background: var(--ve-dropdown-background);
      box-shadow: var(--ve-dropdown-shadow);
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
      transform: var(--ve-dropdown-current-enter-transform);
      transform-origin: var(--ve-dropdown-current-transform-origin);
      transition:
        opacity var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        transform var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        visibility 0s linear var(--ve-dropdown-transition-duration);
      overflow: visible;
    }

    .ve-dropdown-panel.is-open {
      opacity: 1;
      visibility: visible;
      pointer-events: auto;
      transform: none;
      transition-delay: 0s;
    }

    .ve-dropdown-panel.is-closed {
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
    }

    .ve-dropdown-panel.ve-dropdown-placement-top,
    .ve-dropdown-panel.ve-dropdown-placement-top-start,
    .ve-dropdown-panel.ve-dropdown-placement-top-end {
      --ve-dropdown-current-enter-transform: scaleY(0.9)
        translate3d(0, var(--ve-dropdown-enter-offset), 0);
      --ve-dropdown-current-transform-origin: 100% 100%;
    }

    .ve-dropdown-panel.ve-dropdown-placement-bottom,
    .ve-dropdown-panel.ve-dropdown-placement-bottom-start,
    .ve-dropdown-panel.ve-dropdown-placement-bottom-end {
      --ve-dropdown-current-enter-transform: scaleY(0.9)
        translate3d(0, calc(var(--ve-dropdown-enter-offset) * -1), 0);
      --ve-dropdown-current-transform-origin: 0% 0%;
    }

    .ve-dropdown-panel.ve-dropdown-placement-left,
    .ve-dropdown-panel.ve-dropdown-placement-left-start,
    .ve-dropdown-panel.ve-dropdown-placement-left-end,
    .ve-dropdown-panel.ve-dropdown-placement-right,
    .ve-dropdown-panel.ve-dropdown-placement-right-start,
    .ve-dropdown-panel.ve-dropdown-placement-right-end {
      --ve-dropdown-current-enter-transform: scaleY(0.9) translateZ(0);
      --ve-dropdown-current-transform-origin: 0% 0%;
    }

    .ve-dropdown-menu {
      position: relative;
      display: flex;
      flex-direction: column;
      gap: var(--ve-dropdown-menu-gap);
      padding-block: var(--ve-dropdown-padding-block);
      padding-inline: var(--ve-dropdown-padding-inline);
      max-block-size: var(--ve-dropdown-max-height);
      overflow: auto;
      overscroll-behavior: contain;
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-dropdown-panel {
        transition: none;
      }
    }
  `
];
function ve_dropdown_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_dropdown_element_toPropertyKey(arg) {
    var key = ve_dropdown_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_dropdown_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_dropdown_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_dropdown_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_dropdown_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_dropdown_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_dropdown_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_dropdown_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_dropdown_element_dec, ve_dropdown_element_dec1, ve_dropdown_element_dec2, ve_dropdown_element_dec3, ve_dropdown_element_dec4, ve_dropdown_element_dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _init_position, _init_extra__init_position, _init_trigger, _init_extra__init_trigger, ve_dropdown_element_init_selection, ve_dropdown_element_init_extra_init_selection, ve_dropdown_element_init_value, ve_dropdown_element_init_extra_init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, ve_dropdown_element_init_disabled, ve_dropdown_element_init_extra_init_disabled, _init_block, _init_extra__init_block, _init_delay, _init_extra__init_delay, _init_innerOpen, _init_extra__init_innerOpen, _init_hasMenuContent, _init_extra__init_hasMenuContent, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_triggerSlot, _init_extra__init_triggerSlot, _init_menuSlot, _init_extra__init_menuSlot;
const VE_DROPDOWN_TAG = 've-dropdown';
const DEFAULT_HOVER_DELAY = 400;
const HOVER_SAFE_CLOSE_DELAY = 400;
let dropdownPanelSeed = 0;
const VALUE_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return '';
        const trimmed = value.trim();
        if (!trimmed) return '';
        try {
            const parsed = JSON.parse(trimmed);
            if (Array.isArray(parsed)) return parsed.filter((item)=>'string' == typeof item);
            if ('string' == typeof parsed) return parsed;
        } catch  {}
        return trimmed;
    },
    toAttribute (value) {
        if (Array.isArray(value)) return value.length > 0 ? JSON.stringify(value) : null;
        return '' === value ? null : value;
    }
};
function cloneDropdownValue(value) {
    return Array.isArray(value) ? [
        ...value
    ] : value;
}
function hasDropdownValue(value) {
    return Array.isArray(value) ? value.length > 0 : '' !== value;
}
function normalizeDropdownValue(value, selection) {
    if ('multiple' === selection) {
        if (Array.isArray(value)) return [
            ...new Set(value.filter((item)=>item.trim().length > 0))
        ];
        return value.trim().length > 0 ? [
            value
        ] : [];
    }
    if (Array.isArray(value)) return value[0] ?? '';
    return value;
}
function areDropdownValuesEqual(left, right) {
    if (Array.isArray(left) || Array.isArray(right)) {
        if (!Array.isArray(left) || !Array.isArray(right)) return false;
        return left.length === right.length && left.every((item, index)=>item === right[index]);
    }
    return left === right;
}
ve_dropdown_element_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), ve_dropdown_element_dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), ve_dropdown_element_dec2 = property({
    type: String,
    reflect: true,
    useDefault: true
}), ve_dropdown_element_dec3 = property({
    converter: VALUE_CONVERTER
}), ve_dropdown_element_dec4 = property({
    attribute: 'default-value',
    converter: VALUE_CONVERTER
}), ve_dropdown_element_dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = property({
    type: Boolean,
    attribute: 'default-open'
}), _dec7 = property({
    type: Boolean,
    reflect: true
}), _dec8 = property({
    type: Boolean,
    reflect: true
}), _dec9 = property({
    type: Number
}), _dec10 = state(), _dec11 = state(), _dec12 = state(), _dec13 = query('.ve-dropdown-trigger'), _dec14 = query('.ve-dropdown-panel'), _dec15 = query('slot[name="trigger"]'), _dec16 = query('slot:not([name])');
class VeDropdownElement extends VeElement {
    static{
        ({ e: [_init_position, _init_extra__init_position, _init_trigger, _init_extra__init_trigger, ve_dropdown_element_init_selection, ve_dropdown_element_init_extra_init_selection, ve_dropdown_element_init_value, ve_dropdown_element_init_extra_init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, ve_dropdown_element_init_disabled, ve_dropdown_element_init_extra_init_disabled, _init_block, _init_extra__init_block, _init_delay, _init_extra__init_delay, _init_innerOpen, _init_extra__init_innerOpen, _init_hasMenuContent, _init_extra__init_hasMenuContent, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_triggerSlot, _init_extra__init_triggerSlot, _init_menuSlot, _init_extra__init_menuSlot] } = ve_dropdown_element_apply_decs_2311(this, [], [
            [
                ve_dropdown_element_dec,
                1,
                "position"
            ],
            [
                ve_dropdown_element_dec1,
                1,
                "trigger"
            ],
            [
                ve_dropdown_element_dec2,
                1,
                "selection"
            ],
            [
                ve_dropdown_element_dec3,
                1,
                "value"
            ],
            [
                ve_dropdown_element_dec4,
                1,
                "defaultValue"
            ],
            [
                ve_dropdown_element_dec5,
                1,
                "open"
            ],
            [
                _dec6,
                1,
                "defaultOpen"
            ],
            [
                _dec7,
                1,
                "disabled"
            ],
            [
                _dec8,
                1,
                "block"
            ],
            [
                _dec9,
                1,
                "delay"
            ],
            [
                _dec10,
                1,
                "innerOpen"
            ],
            [
                _dec11,
                1,
                "hasMenuContent"
            ],
            [
                _dec12,
                1,
                "resolvedPlacement"
            ],
            [
                _dec13,
                1,
                "triggerElement"
            ],
            [
                _dec14,
                1,
                "panelElement"
            ],
            [
                _dec15,
                1,
                "triggerSlot"
            ],
            [
                _dec16,
                1,
                "menuSlot"
            ]
        ]));
    }
    static styles = ve_dropdown;
    #___private_position_1 = (()=>{
        const _value = _init_position(this, 'bottom-start');
        _init_extra__init_position(this);
        return _value;
    })();
    get position() {
        return this.#___private_position_1;
    }
    set position(_v) {
        this.#___private_position_1 = _v;
    }
    #___private_trigger_2 = (()=>{
        const _value = _init_trigger(this, 'hover');
        _init_extra__init_trigger(this);
        return _value;
    })();
    get trigger() {
        return this.#___private_trigger_2;
    }
    set trigger(_v) {
        this.#___private_trigger_2 = _v;
    }
    #___private_selection_3 = (()=>{
        const _value = ve_dropdown_element_init_selection(this, 'none');
        ve_dropdown_element_init_extra_init_selection(this);
        return _value;
    })();
    get selection() {
        return this.#___private_selection_3;
    }
    set selection(_v) {
        this.#___private_selection_3 = _v;
    }
    #___private_value_4 = (()=>{
        const _value = ve_dropdown_element_init_value(this, '');
        ve_dropdown_element_init_extra_init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_4;
    }
    set value(_v) {
        this.#___private_value_4 = _v;
    }
    #___private_defaultValue_5 = (()=>{
        const _value = _init_defaultValue(this, '');
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_5;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_5 = _v;
    }
    #___private_open_6 = (()=>{
        const _value = _init_open(this, void 0);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_6;
    }
    set open(_v) {
        this.#___private_open_6 = _v;
    }
    #___private_defaultOpen_7 = (()=>{
        const _value = _init_defaultOpen(this, false);
        _init_extra__init_defaultOpen(this);
        return _value;
    })();
    get defaultOpen() {
        return this.#___private_defaultOpen_7;
    }
    set defaultOpen(_v) {
        this.#___private_defaultOpen_7 = _v;
    }
    #___private_disabled_8 = (()=>{
        const _value = ve_dropdown_element_init_disabled(this, false);
        ve_dropdown_element_init_extra_init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_8;
    }
    set disabled(_v) {
        this.#___private_disabled_8 = _v;
    }
    #___private_block_9 = (()=>{
        const _value = _init_block(this, false);
        _init_extra__init_block(this);
        return _value;
    })();
    get block() {
        return this.#___private_block_9;
    }
    set block(_v) {
        this.#___private_block_9 = _v;
    }
    #___private_delay_10 = (()=>{
        const _value = _init_delay(this, DEFAULT_HOVER_DELAY);
        _init_extra__init_delay(this);
        return _value;
    })();
    get delay() {
        return this.#___private_delay_10;
    }
    set delay(_v) {
        this.#___private_delay_10 = _v;
    }
    #___private_innerOpen_11 = (()=>{
        const _value = _init_innerOpen(this, false);
        _init_extra__init_innerOpen(this);
        return _value;
    })();
    get innerOpen() {
        return this.#___private_innerOpen_11;
    }
    set innerOpen(_v) {
        this.#___private_innerOpen_11 = _v;
    }
    #___private_hasMenuContent_12 = (()=>{
        const _value = _init_hasMenuContent(this, false);
        _init_extra__init_hasMenuContent(this);
        return _value;
    })();
    get hasMenuContent() {
        return this.#___private_hasMenuContent_12;
    }
    set hasMenuContent(_v) {
        this.#___private_hasMenuContent_12 = _v;
    }
    #___private_resolvedPlacement_13 = (()=>{
        const _value = _init_resolvedPlacement(this, this.normalizedPosition);
        _init_extra__init_resolvedPlacement(this);
        return _value;
    })();
    get resolvedPlacement() {
        return this.#___private_resolvedPlacement_13;
    }
    set resolvedPlacement(_v) {
        this.#___private_resolvedPlacement_13 = _v;
    }
    #___private_triggerElement_14 = (()=>{
        const _value = _init_triggerElement(this);
        _init_extra__init_triggerElement(this);
        return _value;
    })();
    get triggerElement() {
        return this.#___private_triggerElement_14;
    }
    set triggerElement(_v) {
        this.#___private_triggerElement_14 = _v;
    }
    #___private_panelElement_15 = (()=>{
        const _value = _init_panelElement(this);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_15;
    }
    set panelElement(_v) {
        this.#___private_panelElement_15 = _v;
    }
    #___private_triggerSlot_16 = (()=>{
        const _value = _init_triggerSlot(this);
        _init_extra__init_triggerSlot(this);
        return _value;
    })();
    get triggerSlot() {
        return this.#___private_triggerSlot_16;
    }
    set triggerSlot(_v) {
        this.#___private_triggerSlot_16 = _v;
    }
    #___private_menuSlot_17 = (()=>{
        const _value = _init_menuSlot(this);
        _init_extra__init_menuSlot(this);
        return _value;
    })();
    get menuSlot() {
        return this.#___private_menuSlot_17;
    }
    set menuSlot(_v) {
        this.#___private_menuSlot_17 = _v;
    }
    panelId = `ve-dropdown-panel-${++dropdownPanelSeed}`;
    openTimer = null;
    closeTimer = null;
    documentBound = false;
    positionRequestId = 0;
    contextMenuPoint = null;
    cleanupAutoUpdate = null;
    repositionScheduler = createRafScheduler();
    slottedTriggerElement = null;
    slottedTriggerHadDisabledAttribute = false;
    suppressTriggerFocusOpen = false;
    get isOpen() {
        return 'boolean' == typeof this.open ? this.open : this.innerOpen;
    }
    get isControlled() {
        return 'boolean' == typeof this.open;
    }
    get canShowPanel() {
        return this.isOpen && this.hasMenuContent && !this.disabled;
    }
    get normalizedTrigger() {
        return this.trigger;
    }
    get normalizedPosition() {
        return this.position;
    }
    get canInteract() {
        return !this.disabled;
    }
    get referenceElement() {
        return this.getSlottedTriggerElement() ?? this.triggerElement;
    }
    focus(options) {
        const trigger = this.referenceElement ?? this.triggerElement;
        trigger?.focus(options);
    }
    restoreTriggerFocus() {
        if (!this.referenceElement) {
            this.suppressTriggerFocusOpen = false;
            return;
        }
        this.suppressTriggerFocusOpen = 'hover' === this.normalizedTrigger;
        focusElement(this.referenceElement);
    }
    connectedCallback() {
        super.connectedCallback();
        this.innerOpen = this.open ?? this.defaultOpen;
        this.applyDefaultValue();
        this.normalizeAndSyncSelection();
    }
    firstUpdated() {
        this.syncMenuContentState();
        this.syncSlottedTriggerDisabledState();
        if (this.canShowPanel) this.updateComplete.then(()=>{
            if (!this.isConnected || !this.canShowPanel) return;
            showPanelPopover(this.panelElement);
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
            this.syncDocumentListener();
        });
    }
    disconnectedCallback() {
        this.clearDelayTimers();
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        this.contextMenuPoint = null;
        hidePanelPopover(this.panelElement);
        this.teardownAutoUpdate();
        this.removeDocumentListener();
        this.restoreSlottedTriggerDisabledState();
        super.disconnectedCallback();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('open') && 'boolean' == typeof this.open) this.innerOpen = this.open;
        if (changedProperties.has('defaultOpen') && 1 === changedProperties.size && 'boolean' != typeof this.open) this.innerOpen = this.defaultOpen;
        if (changedProperties.has('disabled') && this.disabled && !this.isControlled) {
            this.innerOpen = false;
            this.contextMenuPoint = null;
        }
        if (changedProperties.has('selection') || changedProperties.has('value') || changedProperties.has('defaultValue')) this.normalizeAndSyncSelection();
        if ((changedProperties.has('position') || changedProperties.has('trigger')) && this.canShowPanel) this.scheduleUpdatePosition();
        if (changedProperties.has('disabled')) this.syncSlottedTriggerDisabledState();
        if ((changedProperties.has('open') || changedProperties.has('innerOpen') || changedProperties.has('disabled')) && !this.isOpen) {
            this.closeDescendantSubItems();
            this.repositionScheduler.cancel();
            hidePanelPopover(this.panelElement);
            this.teardownAutoUpdate();
            this.contextMenuPoint = null;
        }
        if (changedProperties.has('open') || changedProperties.has('innerOpen') || changedProperties.has('trigger')) this.syncDocumentListener();
        if ((changedProperties.has('open') || changedProperties.has('innerOpen')) && this.canShowPanel) this.updateComplete.then(()=>{
            if (!this.isConnected || !this.canShowPanel) return;
            showPanelPopover(this.panelElement);
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
        });
    }
    dispatchOpenChange(nextOpen, source) {
        this.dispatchEvent(new CustomEvent('ve-open-change', {
            detail: {
                open: nextOpen,
                source
            },
            bubbles: true,
            composed: true
        }));
    }
    setOpen(nextOpen, source) {
        if (nextOpen && this.disabled) return;
        if (this.isOpen === nextOpen) return;
        if (nextOpen && !this.hasMenuContent) return;
        if (!this.isControlled) this.innerOpen = nextOpen;
        this.dispatchOpenChange(nextOpen, source);
        if (!nextOpen) {
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            this.contextMenuPoint = null;
            hidePanelPopover(this.panelElement);
            this.teardownAutoUpdate();
            this.syncDocumentListener();
            return;
        }
        this.updateComplete.then(()=>{
            if (!this.isConnected || !this.canShowPanel) return;
            showPanelPopover(this.panelElement);
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
            this.syncDocumentListener();
        });
    }
    applyDefaultValue() {
        if (hasDropdownValue(this.value) || !hasDropdownValue(this.defaultValue)) return;
        this.value = normalizeDropdownValue(this.defaultValue, this.selection);
    }
    normalizeAndSyncSelection() {
        const normalizedValue = normalizeDropdownValue(this.value, this.selection);
        if (!areDropdownValuesEqual(this.value, normalizedValue)) this.value = normalizedValue;
        this.syncSelectionState(normalizedValue);
    }
    syncSelectionState(value = this.value) {
        if ('none' === this.selection) return;
        const selectedValues = Array.isArray(value) ? value : '' === value ? [] : [
            value
        ];
        const selectedSet = new Set(selectedValues);
        const itemSelection = 'multiple' === this.selection ? 'checkbox' : 'check';
        this.getAllDropdownItems().forEach((item)=>{
            if ('switch' !== item.selection) item.selection = itemSelection;
            item.selected = selectedSet.has(this.getItemValue(item));
        });
    }
    getAllDropdownItems() {
        return Array.from(this.querySelectorAll(VE_DROPDOWN_ITEM_TAG)).filter((element)=>element instanceof VeDropdownItemElement);
    }
    getItemValue(item) {
        const explicitValue = (item.value ?? '').trim();
        if (explicitValue.length > 0) return explicitValue;
        return (item.textContent ?? '').trim();
    }
    syncMenuContentState() {
        this.hasMenuContent = hasAssignedSlotContent(this.menuSlot);
        this.normalizeAndSyncSelection();
        if (!this.hasMenuContent && this.isOpen && !this.isControlled) this.setOpen(false, 'api');
        if (this.canShowPanel) this.scheduleUpdatePosition();
        this.syncDocumentListener();
    }
    getSlottedTriggerElement() {
        const assignedElements = this.triggerSlot?.assignedElements({
            flatten: true
        }) ?? [];
        return assignedElements.find((element)=>element instanceof HTMLElement) ?? null;
    }
    syncSlottedTriggerDisabledState() {
        const triggerElement = this.getSlottedTriggerElement();
        if (this.slottedTriggerElement && this.slottedTriggerElement !== triggerElement) {
            if (!this.slottedTriggerHadDisabledAttribute) {
                this.slottedTriggerElement.removeAttribute('disabled');
                Reflect.set(this.slottedTriggerElement, 'disabled', false);
            }
            this.slottedTriggerElement = null;
        }
        if (!triggerElement) return;
        if (this.slottedTriggerElement !== triggerElement) {
            this.slottedTriggerElement = triggerElement;
            this.slottedTriggerHadDisabledAttribute = triggerElement.hasAttribute('disabled');
        }
        if (this.disabled) {
            triggerElement.setAttribute('disabled', '');
            Reflect.set(triggerElement, 'disabled', true);
            return;
        }
        if (!this.slottedTriggerHadDisabledAttribute) {
            triggerElement.removeAttribute('disabled');
            Reflect.set(triggerElement, 'disabled', false);
        }
    }
    restoreSlottedTriggerDisabledState() {
        if (!this.slottedTriggerElement) return;
        if (!this.slottedTriggerHadDisabledAttribute) {
            this.slottedTriggerElement.removeAttribute('disabled');
            Reflect.set(this.slottedTriggerElement, 'disabled', false);
        }
        this.slottedTriggerElement = null;
    }
    clearDelayTimers() {
        if (null !== this.openTimer) {
            window.clearTimeout(this.openTimer);
            this.openTimer = null;
        }
        if (null !== this.closeTimer) {
            window.clearTimeout(this.closeTimer);
            this.closeTimer = null;
        }
    }
    scheduleOpen() {
        this.clearDelayTimers();
        if (this.delay <= 0) return void this.setOpen(true, 'trigger');
        this.openTimer = window.setTimeout(()=>{
            this.setOpen(true, 'trigger');
            this.openTimer = null;
        }, Math.max(0, this.delay));
    }
    scheduleClose(delay = this.delay) {
        this.clearDelayTimers();
        if (delay <= 0) return void this.setOpen(false, 'trigger');
        this.closeTimer = window.setTimeout(()=>{
            this.setOpen(false, 'trigger');
            this.closeTimer = null;
        }, Math.max(0, delay));
    }
    handleTriggerMouseEnter = ()=>{
        if (!this.canInteract || 'hover' !== this.normalizedTrigger) return;
        this.scheduleOpen();
    };
    handleTriggerMouseLeave = (event)=>{
        if (!this.canInteract || 'hover' !== this.normalizedTrigger) return;
        const nextTarget = event.relatedTarget;
        if (nextTarget && this.panelElement?.contains(nextTarget)) return;
        this.scheduleClose(Math.max(this.delay, HOVER_SAFE_CLOSE_DELAY));
    };
    handleTriggerClick = (event)=>{
        if (!this.canInteract || 'click' !== this.normalizedTrigger) return;
        event.stopPropagation();
        this.setOpen(!this.isOpen, 'trigger');
    };
    handleTriggerContextMenu = (event)=>{
        if (!this.canInteract || 'contextmenu' !== this.normalizedTrigger) return;
        event.preventDefault();
        event.stopPropagation();
        this.contextMenuPoint = {
            x: event.clientX,
            y: event.clientY
        };
        this.setOpen(true, 'contextmenu');
    };
    handleTriggerFocusIn = ()=>{
        if (!this.canInteract || 'hover' !== this.normalizedTrigger) return;
        if (this.suppressTriggerFocusOpen) {
            this.suppressTriggerFocusOpen = false;
            return;
        }
        this.setOpen(true, 'trigger');
    };
    handleTriggerFocusOut = (event)=>{
        if (!this.canInteract || 'hover' !== this.normalizedTrigger) return;
        const nextTarget = event.relatedTarget;
        if (nextTarget && (this.contains(nextTarget) || this.panelElement?.contains(nextTarget))) return;
        this.scheduleClose();
    };
    handlePanelMouseEnter = ()=>{
        if (!this.canInteract || 'hover' !== this.normalizedTrigger) return;
        this.clearDelayTimers();
    };
    handlePanelMouseLeave = (event)=>{
        if (!this.canInteract || 'hover' !== this.normalizedTrigger) return;
        const nextTarget = event.relatedTarget;
        if (nextTarget && this.referenceElement?.contains(nextTarget)) return;
        this.scheduleClose(Math.max(this.delay, HOVER_SAFE_CLOSE_DELAY));
    };
    handleDocumentPointerDown = (event)=>{
        if (!this.canShowPanel || !this.canInteract) return;
        if ('click' !== this.normalizedTrigger && 'contextmenu' !== this.normalizedTrigger) return;
        const path = event.composedPath();
        const clickedInside = path.includes(this) || path.includes(this.referenceElement) || path.includes(this.panelElement);
        if (!clickedInside) this.setOpen(false, 'outside');
    };
    getMenuItems() {
        return getAssignedMenuNodes(this.menuSlot);
    }
    closeDescendantSubItems() {
        this.querySelectorAll('ve-dropdown-sub-item').forEach((subItem)=>{
            subItem.open = false;
        });
    }
    focusFirstMenuItem() {
        focusElement(this.getMenuItems()[0]);
    }
    focusLastMenuItem() {
        const items = this.getMenuItems();
        focusElement(items.at(-1));
    }
    focusAdjacentMenuItem(step) {
        const items = this.getMenuItems();
        if (0 === items.length) return;
        const activeElement = document.activeElement;
        const activeHost = items.find((item)=>item === activeElement || item.shadowRoot?.activeElement === activeElement);
        const activeIndex = activeHost ? items.indexOf(activeHost) : -1;
        const nextIndex = -1 === activeIndex ? 1 === step ? 0 : items.length - 1 : (activeIndex + step + items.length) % items.length;
        focusElement(items[nextIndex]);
    }
    handleKeydown = (event)=>{
        if (!this.canInteract) return;
        if ('ArrowDown' === event.key) {
            event.preventDefault();
            if (!this.canShowPanel) {
                this.setOpen(true, 'trigger');
                this.updateComplete.then(()=>{
                    this.focusFirstMenuItem();
                });
                return;
            }
            this.focusAdjacentMenuItem(1);
            return;
        }
        if ('ArrowUp' === event.key) {
            event.preventDefault();
            if (!this.canShowPanel) {
                this.setOpen(true, 'trigger');
                this.updateComplete.then(()=>{
                    this.focusLastMenuItem();
                });
                return;
            }
            this.focusAdjacentMenuItem(-1);
            return;
        }
        if ('Home' === event.key) {
            if (!this.canShowPanel) return;
            event.preventDefault();
            this.focusFirstMenuItem();
            return;
        }
        if ('End' === event.key) {
            if (!this.canShowPanel) return;
            event.preventDefault();
            this.focusLastMenuItem();
            return;
        }
        if ('Escape' === event.key) {
            if (!this.canShowPanel) return;
            event.stopPropagation();
            this.setOpen(false, 'escape');
            this.restoreTriggerFocus();
        }
    };
    emitSelectionChange(value, previousValue, item, selected) {
        if (areDropdownValuesEqual(value, previousValue)) return;
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail: {
                value: cloneDropdownValue(value),
                previousValue: cloneDropdownValue(previousValue),
                item,
                selected
            },
            bubbles: true,
            composed: true
        }));
    }
    updateSelectionValue(detail) {
        if ('none' === this.selection || 'switch' === detail.selection) return detail.selected;
        const previousValue = cloneDropdownValue(this.value);
        if ('single' === this.selection) {
            this.value = detail.value;
            this.syncSelectionState(this.value);
            this.emitSelectionChange(this.value, previousValue, detail.item, true);
            return true;
        }
        const currentValues = Array.isArray(this.value) ? [
            ...this.value
        ] : '' === this.value ? [] : [
            this.value
        ];
        const valueSet = new Set(currentValues);
        if (detail.selected) valueSet.add(detail.value);
        else valueSet.delete(detail.value);
        this.value = Array.from(valueSet);
        this.syncSelectionState(this.value);
        this.emitSelectionChange(this.value, previousValue, detail.item, detail.selected);
        return detail.selected;
    }
    handleMenuItemSelect = (event)=>{
        event.stopPropagation();
        const selected = this.updateSelectionValue(event.detail);
        this.dispatchEvent(new CustomEvent('ve-select', {
            detail: {
                value: event.detail.value,
                label: event.detail.label,
                item: event.detail.item,
                selection: event.detail.selection,
                selected,
                dropdownValue: cloneDropdownValue(this.value)
            },
            bubbles: true,
            composed: true
        }));
        const shouldClose = this.canInteract && ('none' === event.detail.selection || 'check' === event.detail.selection || 'single' === this.selection);
        if (shouldClose) {
            this.setOpen(false, 'menu');
            this.restoreTriggerFocus();
        }
    };
    syncDocumentListener() {
        if (!this.isConnected) return void this.removeDocumentListener();
        if (('click' === this.normalizedTrigger || 'contextmenu' === this.normalizedTrigger) && this.canShowPanel && !this.documentBound) {
            document.addEventListener('pointerdown', this.handleDocumentPointerDown, true);
            this.documentBound = true;
            return;
        }
        if ('click' !== this.normalizedTrigger && 'contextmenu' !== this.normalizedTrigger || !this.canShowPanel) this.removeDocumentListener();
    }
    removeDocumentListener() {
        if (!this.documentBound) return;
        document.removeEventListener('pointerdown', this.handleDocumentPointerDown, true);
        this.documentBound = false;
    }
    setupAutoUpdate() {
        if (!this.isConnected) return;
        if (this.contextMenuPoint) return void this.teardownAutoUpdate();
        const referenceElement = this.referenceElement;
        if (!this.canShowPanel || !referenceElement || !this.panelElement) return;
        this.teardownAutoUpdate();
        this.cleanupAutoUpdate = createFloatingAutoUpdate(referenceElement, this.panelElement, ()=>{
            this.scheduleUpdatePosition();
        });
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    updateContextMenuPosition() {
        if (!this.panelElement || !this.contextMenuPoint) return;
        const viewportPadding = 8;
        const panelRect = this.panelElement.getBoundingClientRect();
        const maxX = Math.max(viewportPadding, window.innerWidth - panelRect.width - viewportPadding);
        const maxY = Math.max(viewportPadding, window.innerHeight - panelRect.height - viewportPadding);
        const x = Math.min(Math.max(this.contextMenuPoint.x, viewportPadding), maxX);
        const y = Math.min(Math.max(this.contextMenuPoint.y, viewportPadding), maxY);
        this.panelElement.style.position = 'fixed';
        this.panelElement.style.left = `${x}px`;
        this.panelElement.style.top = `${y}px`;
        this.panelElement.style.visibility = 'visible';
    }
    async updatePosition() {
        const referenceElement = this.referenceElement;
        if (!this.canShowPanel || !this.isConnected || !this.panelElement) return;
        if (this.contextMenuPoint) return void this.updateContextMenuPosition();
        if (!referenceElement) return;
        const requestId = ++this.positionRequestId;
        const distance = getFloatingDistance(this, {
            offsetVarName: '--ve-dropdown-offset',
            arrowVarName: '--ve-dropdown-arrow-size',
            fallbackOffset: 4,
            fallbackArrowSize: 0,
            arrowGap: 0
        });
        this.panelElement.style.position = 'fixed';
        this.panelElement.style.top = '0px';
        this.panelElement.style.left = '0px';
        const position = await computeFloatingPanelPosition({
            reference: referenceElement,
            floating: this.panelElement,
            placement: this.normalizedPosition,
            distance,
            viewportPadding: 8
        });
        if (requestId !== this.positionRequestId || !this.canShowPanel || !this.isConnected) return;
        this.resolvedPlacement = position.placement;
        this.panelElement.style.left = `${position.x}px`;
        this.panelElement.style.top = `${position.y}px`;
        this.panelElement.style.visibility = position.hidden ? 'hidden' : 'visible';
    }
    handleMenuSlotChange = ()=>{
        this.syncMenuContentState();
    };
    handleTriggerSlotChange = ()=>{
        this.syncSlottedTriggerDisabledState();
        if (this.canShowPanel) {
            this.teardownAutoUpdate();
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
        }
    };
    render() {
        const panelClass = classMap({
            've-dropdown-panel': true,
            [`ve-dropdown-placement-${this.resolvedPlacement}`]: true,
            'is-open': this.canShowPanel,
            'is-closed': !this.canShowPanel
        });
        return html`
      <span class="ve-dropdown">
        <span
          class="ve-dropdown-trigger"
          part="trigger"
          tabindex=${this.disabled ? -1 : 0}
          aria-haspopup="menu"
          aria-expanded=${this.canShowPanel ? 'true' : 'false'}
          aria-disabled=${this.disabled ? 'true' : 'false'}
          aria-controls=${this.canShowPanel ? this.panelId : nothing}
          @mouseenter=${this.handleTriggerMouseEnter}
          @mouseleave=${this.handleTriggerMouseLeave}
          @click=${this.handleTriggerClick}
          @contextmenu=${this.handleTriggerContextMenu}
          @focusin=${this.handleTriggerFocusIn}
          @focusout=${this.handleTriggerFocusOut}
          @keydown=${this.handleKeydown}
        >
          <slot
            name="trigger"
            @slotchange=${this.handleTriggerSlotChange}
          ></slot>
        </span>

        <div
          id=${this.panelId}
          class=${panelClass}
          part="panel"
          popover=${"manual"}
          role="menu"
          aria-hidden=${this.canShowPanel ? 'false' : 'true'}
          @mouseenter=${this.handlePanelMouseEnter}
          @mouseleave=${this.handlePanelMouseLeave}
          @keydown=${this.handleKeydown}
          @ve-dropdown-item-select=${this.handleMenuItemSelect}
        >
          <div class="ve-dropdown-menu" part="menu">
            <slot @slotchange=${this.handleMenuSlotChange}></slot>
          </div>
        </div>
      </span>
    `;
    }
}
export { VE_DROPDOWN_ITEM_SELECT_EVENT, VE_DROPDOWN_ITEM_TAG, VE_DROPDOWN_TAG, VeDropdownElement, VeDropdownItemElement, focusElement, getAssignedMenuNodes, hasAssignedSlotContent, token };
