// @ve-design/web/web/dist/ve-avatar.js
import { css, html, nothing } from "lit";
import { property, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
const veAvatarTokenDefaults = css`
  :host {
    --ve-avatar-default-size: 32px;
    --ve-avatar-current-size: var(
      --ve-avatar-size,
      var(--ve-avatar-context-size, var(--ve-avatar-default-size))
    );
    --ve-avatar-radius: var(--radius-full);

    --ve-avatar-background: var(--color-bg-muted);
    --ve-avatar-text-color: var(--color-text-secondary);
    --ve-avatar-border-color: transparent;
    --ve-avatar-border-width: 0px;
    --ve-avatar-image-background: var(--color-bg-surface);

    --ve-avatar-font-family: var(--font-sans);
    --ve-avatar-font-weight: var(--font-weight-medium);
    --ve-avatar-font-size: var(--text-caption);
    --ve-avatar-line-height: var(--line-height-body);
    --ve-avatar-letter-spacing: var(--letter-spacing-normal);
    --ve-avatar-icon-size: calc(var(--ve-avatar-current-size) * 0.583333);
    --ve-avatar-icon-color: var(--color-icon-primary);
    --ve-avatar-disabled-opacity: var(--state-disabled-ratio);

    --ve-avatar-group-size: var(
      --ve-avatar-size,
      var(--ve-avatar-default-size)
    );
    --ve-avatar-group-overlap: calc(var(--ve-avatar-group-size) / 6);
    --ve-avatar-group-child-size: var(--ve-avatar-group-size);
    --ve-avatar-group-avatar-border-color: var(--color-bg-base);
    --ve-avatar-group-avatar-border-width: var(--stroke-weight-base);
    --ve-avatar-group-summary-background: var(--color-bg-muted);
    --ve-avatar-group-summary-text-color: var(--color-text-tertiary);
    --ve-avatar-group-summary-font-family: var(--font-en);
    --ve-avatar-group-summary-font-weight: var(--font-weight-normal);
    --ve-avatar-group-summary-line-height: calc(
      var(--text-caption) + var(--space-xxxs)
    );
    --ve-avatar-group-summary-border-color: var(
      --ve-avatar-group-avatar-border-color
    );
    --ve-avatar-group-summary-border-width: var(
      --ve-avatar-group-avatar-border-width
    );
  }
`;
const token = veAvatarTokenDefaults;
const ve_avatar = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      inline-size: var(--ve-avatar-current-size);
      block-size: var(--ve-avatar-current-size);
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    :host([size]) {
      inline-size: var(--ve-avatar-current-size);
      block-size: var(--ve-avatar-current-size);
    }

    .ve-avatar {
      position: relative;
      display: inline-flex;
      inline-size: 100%;
      block-size: 100%;
    }

    .ve-avatar-control {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: 100%;
      block-size: 100%;
      box-sizing: border-box;
      overflow: hidden;
      border-style: solid;
      border-width: var(--ve-avatar-border-width);
      border-color: var(--ve-avatar-border-color);
      border-radius: var(--ve-avatar-radius);
      background: var(--ve-avatar-background);
      color: var(--ve-avatar-text-color);
      font-family: var(--ve-avatar-font-family);
      font-size: var(--ve-avatar-font-size);
      font-weight: var(--ve-avatar-font-weight);
      line-height: var(--ve-avatar-line-height);
      letter-spacing: var(--ve-avatar-letter-spacing);
      user-select: none;
    }

    .ve-avatar-slot {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: 100%;
      block-size: 100%;
      box-sizing: border-box;
      text-align: center;
      overflow: hidden;
      white-space: nowrap;
      text-rendering: optimizeLegibility;
    }

    .ve-avatar-slot ::slotted(img),
    .ve-avatar-slot ::slotted(picture) {
      inline-size: 100%;
      block-size: 100%;
    }

    .ve-avatar-slot ::slotted(img) {
      object-fit: cover;
      background: var(--ve-avatar-image-background);
    }

    .ve-avatar-slot ::slotted(*:not(img):not(picture)) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-inline-size: 100%;
      max-block-size: 100%;
      box-sizing: border-box;
    }

    .ve-avatar-slot ::slotted(ve-icon),
    .ve-avatar-slot ::slotted(svg) {
      flex: none;
      inline-size: var(--ve-avatar-icon-size);
      block-size: var(--ve-avatar-icon-size);
      color: var(--ve-avatar-icon-color);
      font-size: var(--ve-avatar-icon-size);
      --ve-icon-size: var(--ve-avatar-icon-size);
    }

    .ve-avatar-control:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--stroke-2) color-mix(currentColor 24%, transparent);
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _init_size, _init_extra__init_size;
const VE_AVATAR_TAG = 've-avatar';
const DEFAULT_AVATAR_SIZE = '32';
const DEFAULT_AVATAR_SIZE_VALUE = `${DEFAULT_AVATAR_SIZE}px`;
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
});
class VeAvatarElement extends VeElement {
    static{
        ({ e: [_init_size, _init_extra__init_size] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "size"
            ]
        ]));
    }
    static styles = ve_avatar;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                'style'
            ])
        ];
    }
    #___private_size_1 = (()=>{
        const _value = _init_size(this, DEFAULT_AVATAR_SIZE);
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_1;
    }
    set size(_v) {
        this.#___private_size_1 = _v;
    }
    get hasExplicitSize() {
        return this.hasAttribute('size') || String(this.size).trim() !== DEFAULT_AVATAR_SIZE;
    }
    get normalizedSize() {
        const rawSize = String(this.size).trim();
        const numericSize = Number(rawSize);
        if (!rawSize) return DEFAULT_AVATAR_SIZE_VALUE;
        return Number.isFinite(numericSize) ? `${numericSize}px` : rawSize;
    }
    connectedCallback() {
        super.connectedCallback();
        this.syncHostSizeToken();
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('size')) this.syncHostSizeToken();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if ('size' === name && oldValue !== newValue) {
            this.syncHostSizeToken();
            this.requestUpdate();
        }
    }
    syncHostSizeToken() {
        if (!this.hasExplicitSize) return void this.style.removeProperty('--ve-avatar-size');
        this.style.setProperty('--ve-avatar-size', this.normalizedSize);
    }
    render() {
        return html`
      <span class="ve-avatar">
        <span class="ve-avatar-control" part="control">
          <span class="ve-avatar-slot" part="content">
            <slot></slot>
          </span>
        </span>
      </span>
    `;
    }
}
const ve_avatar_group = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-avatar-group {
      display: inline-flex;
      align-items: center;
      min-inline-size: 0;
      isolation: isolate;
    }

    .ve-avatar-group-item,
    .ve-avatar-group-summary {
      position: relative;
      display: inline-flex;
      flex: none;
    }

    ::slotted(ve-avatar:not(:first-child)) {
      margin-inline-start: calc(var(--ve-avatar-group-overlap) * -1);
    }

    ::slotted(ve-avatar) {
      position: relative;
      --ve-avatar-context-size: var(--ve-avatar-group-child-size);
      --ve-avatar-border-color: var(--ve-avatar-group-avatar-border-color);
      --ve-avatar-border-width: var(--ve-avatar-group-avatar-border-width);
    }

    .ve-avatar-group-item:not(:first-child),
    .ve-avatar-group-summary {
      margin-inline-start: calc(var(--ve-avatar-group-overlap) * -1);
    }

    .ve-avatar-group-summary {
      --ve-avatar-background: var(--ve-avatar-group-summary-background);
      --ve-avatar-text-color: var(--ve-avatar-group-summary-text-color);
      --ve-avatar-border-color: var(--ve-avatar-group-summary-border-color);
      --ve-avatar-border-width: var(--ve-avatar-group-summary-border-width);
      --ve-avatar-font-family: var(--ve-avatar-group-summary-font-family);
      --ve-avatar-font-weight: var(--ve-avatar-group-summary-font-weight);
      --ve-avatar-font-size: var(--text-caption);
      --ve-avatar-line-height: var(--ve-avatar-group-summary-line-height);
      z-index: 1;
    }
  `
];
function ve_avatar_group_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_avatar_group_element_toPropertyKey(arg) {
    var key = ve_avatar_group_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_avatar_group_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_avatar_group_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_avatar_group_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_avatar_group_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_avatar_group_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_avatar_group_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_avatar_group_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_avatar_group_element_dec, _dec1, _dec2, ve_avatar_group_element_init_size, ve_avatar_group_element_init_extra_init_size, _init_maxCount, _init_extra__init_maxCount, _init_avatars, _init_extra__init_avatars;
const VE_AVATAR_GROUP_TAG = 've-avatar-group';
const DEFAULT_AVATAR_GROUP_SIZE = '32';
const DEFAULT_AVATAR_GROUP_SIZE_VALUE = `${DEFAULT_AVATAR_GROUP_SIZE}px`;
const parsePixelSize = (size)=>{
    const rawSize = String(size ?? '').trim();
    const numericSize = Number(rawSize);
    if (!rawSize) return;
    if (Number.isFinite(numericSize) && numericSize >= 0) return numericSize;
    const pxSize = /^(\d+(?:\.\d+)?)px$/i.exec(rawSize)?.[1];
    if (!pxSize) return;
    return Number(pxSize);
};
const formatGroupOverlap = (avatarSize)=>`${Math.floor(avatarSize / 6)}px`;
ve_avatar_group_element_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: Number,
    attribute: 'max-count',
    reflect: true
}), _dec2 = state();
class VeAvatarGroupElement extends VeElement {
    static{
        ({ e: [ve_avatar_group_element_init_size, ve_avatar_group_element_init_extra_init_size, _init_maxCount, _init_extra__init_maxCount, _init_avatars, _init_extra__init_avatars] } = ve_avatar_group_element_apply_decs_2311(this, [], [
            [
                ve_avatar_group_element_dec,
                1,
                "size"
            ],
            [
                _dec1,
                1,
                "maxCount"
            ],
            [
                _dec2,
                1,
                "avatars"
            ]
        ]));
    }
    static styles = ve_avatar_group;
    static dependencies = {
        [VE_AVATAR_TAG]: VeAvatarElement
    };
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                'style'
            ])
        ];
    }
    #___private_size_1 = (()=>{
        const _value = ve_avatar_group_element_init_size(this, DEFAULT_AVATAR_GROUP_SIZE);
        ve_avatar_group_element_init_extra_init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_1;
    }
    set size(_v) {
        this.#___private_size_1 = _v;
    }
    #___private_maxCount_2 = (()=>{
        const _value = _init_maxCount(this, 0);
        _init_extra__init_maxCount(this);
        return _value;
    })();
    get maxCount() {
        return this.#___private_maxCount_2;
    }
    set maxCount(_v) {
        this.#___private_maxCount_2 = _v;
    }
    #___private_avatars_3 = (()=>{
        const _value = _init_avatars(this, []);
        _init_extra__init_avatars(this);
        return _value;
    })();
    get avatars() {
        return this.#___private_avatars_3;
    }
    set avatars(_v) {
        this.#___private_avatars_3 = _v;
    }
    get hasExplicitSize() {
        return this.hasAttribute('size') || String(this.size).trim() !== DEFAULT_AVATAR_GROUP_SIZE;
    }
    get normalizedSize() {
        const rawSize = String(this.size).trim();
        const numericSize = Number(rawSize);
        if (!rawSize) return DEFAULT_AVATAR_GROUP_SIZE_VALUE;
        return Number.isFinite(numericSize) ? `${numericSize}px` : rawSize;
    }
    get normalizedOverlap() {
        const avatarSize = this.resolveOverlapAvatarSize();
        return void 0 === avatarSize ? void 0 : formatGroupOverlap(avatarSize);
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('size')) this.syncGroupSizeToken();
    }
    connectedCallback() {
        super.connectedCallback();
        this.syncGroupSizeToken();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if ('size' === name && oldValue !== newValue) {
            this.syncGroupSizeToken();
            this.requestUpdate();
        }
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
        this.syncGroupSizeToken();
        this.syncChildAvatarSizeTokens();
        this.syncAvatarZIndex();
    }
    handleSlotChange(event) {
        const slot = event.target instanceof HTMLSlotElement ? event.target : null;
        const assignedAvatars = slot?.assignedElements({
            flatten: true
        }).filter((element)=>element.localName === VE_AVATAR_TAG) ?? [];
        this.avatars = assignedAvatars;
        this.syncGroupSizeToken();
        this.syncChildAvatarSizeTokens();
        this.syncAvatarZIndex();
    }
    syncGroupSizeToken() {
        if (!this.hasExplicitSize) {
            this.style.removeProperty('--ve-avatar-group-size');
            this.style.removeProperty('--ve-avatar-group-child-size');
            this.syncGroupOverlapToken();
            return;
        }
        this.style.setProperty('--ve-avatar-group-size', this.normalizedSize);
        this.style.setProperty('--ve-avatar-group-child-size', this.normalizedSize);
        this.syncGroupOverlapToken();
    }
    syncGroupOverlapToken() {
        const normalizedOverlap = this.normalizedOverlap;
        if (void 0 === normalizedOverlap) return void this.style.removeProperty('--ve-avatar-group-overlap');
        this.style.setProperty('--ve-avatar-group-overlap', normalizedOverlap);
    }
    resolveOverlapAvatarSize() {
        if (this.hasExplicitSize) return parsePixelSize(this.size);
        const firstExplicitAvatarSize = this.avatars.find((avatar)=>avatar.hasAttribute('size'))?.getAttribute('size');
        return parsePixelSize(firstExplicitAvatarSize);
    }
    syncChildAvatarSizeTokens() {
        const normalizedOverlap = this.normalizedOverlap;
        this.avatars.forEach((avatar)=>{
            if (void 0 === normalizedOverlap) avatar.style.removeProperty('--ve-avatar-group-overlap');
            else avatar.style.setProperty('--ve-avatar-group-overlap', normalizedOverlap);
            if (!this.hasExplicitSize) return void avatar.style.removeProperty('--ve-avatar-context-size');
            avatar.style.setProperty('--ve-avatar-context-size', this.normalizedSize);
        });
    }
    syncAvatarZIndex() {
        const visibleAvatars = this.getVisibleAvatars();
        this.avatars.forEach((avatar)=>{
            avatar.style.removeProperty('z-index');
            avatar.toggleAttribute('hidden', !visibleAvatars.includes(avatar));
        });
        visibleAvatars.forEach((avatar, index)=>{
            avatar.style.setProperty('z-index', String(visibleAvatars.length - index + 1));
        });
    }
    getVisibleAvatars() {
        const avatars = this.avatars ?? [];
        const visibleCount = this.maxCount > 0 ? Math.min(this.maxCount, avatars.length) : avatars.length;
        return avatars.slice(0, visibleCount);
    }
    firstUpdated() {
        this.syncChildAvatarSizeTokens();
        this.syncAvatarZIndex();
    }
    render() {
        const avatars = this.avatars ?? [];
        const visibleAvatars = this.getVisibleAvatars();
        const hiddenCount = Math.max(avatars.length - visibleAvatars.length, 0);
        return html`
      <div class="ve-avatar-group" part="group">
        <slot @slotchange=${this.handleSlotChange}></slot>
        ${hiddenCount > 0 ? html`
              <ve-avatar
                class="ve-avatar-group-summary"
                size=${this.hasExplicitSize ? String(this.size) : nothing}
                part="summary"
                aria-label=${`还有 ${hiddenCount} 个头像`}
              >
                +${hiddenCount}
              </ve-avatar>
            ` : nothing}
      </div>
    `;
    }
}
defineCustomElement(VE_AVATAR_TAG, VeAvatarElement);
defineCustomElement(VE_AVATAR_GROUP_TAG, VeAvatarGroupElement);
export { VE_AVATAR_GROUP_TAG, VE_AVATAR_TAG, VeAvatarElement, VeAvatarGroupElement };
