// @ve-design/web/web/dist/ve-authorization.js
import { css, html, nothing } from "lit";
import { property } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
const veAuthorizationTokenDefaults = css`
  :host {
    --ve-authorization-background: var(--color-bg-base);
    --ve-authorization-border-color: var(--color-border-default);
    --ve-authorization-border-strong: var(--color-border-strong);
    --ve-authorization-text-primary: var(--color-text-primary);
    --ve-authorization-text-secondary: var(--color-text-secondary);
    --ve-authorization-text-foreground: var(--color-text-foreground);
    --ve-authorization-text-accent: var(--color-text-accent);
    --ve-authorization-content-background: var(--color-bg-surface);
    --ve-authorization-primary-action-background: var(--color-bg-primary);
    --ve-authorization-primary-action-text: var(--color-text-foreground);
    --ve-authorization-action-background: transparent;
    --ve-authorization-action-text: var(--color-text-primary);
    --ve-authorization-action-border: var(--color-border-strong);
    --ve-authorization-shadow: var(--shadow-md);
    --ve-authorization-radius: var(--radius-2xl);
    --ve-authorization-inner-radius: var(--radius-lg);
    --ve-authorization-action-radius: var(--radius-md);
    --ve-authorization-gap: var(--space-xs);
    --ve-authorization-gap-small: var(--space-xxs);
    --ve-authorization-padding-block: var(--space-xs);
    --ve-authorization-padding-inline: var(--space-xs);
    --ve-authorization-action-height: 28px;
    --ve-authorization-action-padding-inline: var(--space-xs);
    --ve-authorization-min-width: 370px;
    --ve-authorization-max-width-vertical: 600px;
    --ve-authorization-code-font: var(--font-mono);
    --ve-authorization-body-font: var(--font-sans);
    --ve-authorization-body-size: var(--text-body);
    --ve-authorization-content-body-size: var(--text-body-lg);
    --ve-authorization-code-size: var(--text-body-sm);
    --ve-authorization-disabled-opacity: var(--state-disabled-ratio);
  }
`;
const token = veAuthorizationTokenDefaults;
const ve_authorization = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      min-inline-size: var(--ve-authorization-min-width);
      color: var(--ve-authorization-text-primary);
      font-family: var(--ve-authorization-body-font);
    }

    :host([hidden]) {
      display: none;
    }

    :host([disabled]) {
      cursor: not-allowed;
    }

    :host([orientation='vertical']) {
      max-inline-size: var(--ve-authorization-max-width-vertical);
    }

    .ve-authorization {
      box-sizing: border-box;
      display: flex;
      flex-direction: column;
      align-items: stretch;
      gap: var(--ve-authorization-gap);
      width: 100%;
      padding: var(--ve-authorization-padding-block)
        var(--ve-authorization-padding-inline) var(--space-s);
      border: var(--stroke-weight-base) solid
        var(--ve-authorization-border-color);
      border-radius: var(--ve-authorization-radius);
      background: var(--ve-authorization-background);
      box-shadow: var(--ve-authorization-shadow);
    }

    .ve-authorization-main {
      display: flex;
      flex-direction: column;
      gap: var(--ve-authorization-gap-small);
      min-inline-size: 0;
    }

    .ve-authorization-body {
      display: flex;
      flex: 1 1 auto;
      flex-direction: column;
      gap: var(--ve-authorization-gap-small);
      min-inline-size: 0;
    }

    .ve-authorization-title {
      margin: 0;
      color: var(--ve-authorization-text-accent);
      font-size: var(--ve-authorization-body-size);
      font-weight: var(--font-weight-medium);
      line-height: 1.5;
      word-break: break-word;
    }

    .ve-authorization-content {
      display: block;
      overflow: hidden;
      box-sizing: border-box;
      width: 100%;
      margin: 0;
      padding: var(--space-xxs) var(--space-xs);
      border-radius: var(--ve-authorization-inner-radius);
      background: var(--ve-authorization-content-background);
      color: var(--ve-authorization-text-secondary);
      font-family: var(--ve-authorization-code-font);
      font-size: var(--ve-authorization-code-size);
      font-weight: var(--font-weight-regular);
      line-height: 20px;
      white-space: pre-wrap;
      word-break: break-word;
    }

    .ve-authorization-content:empty {
      display: none;
    }

    .ve-authorization-actions {
      display: flex;
      flex-wrap: wrap;
      align-items: center;
      justify-content: flex-end;
      gap: var(--ve-authorization-gap);
      width: 100%;
    }

    .ve-authorization-actions slot[name='actions'] {
      display: contents;
    }

    :host([orientation='vertical']) .ve-authorization-actions {
      flex-direction: column;
      align-items: stretch;
    }

    .ve-authorization-action {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      min-inline-size: fit-content;
      block-size: var(--ve-authorization-action-height);
      padding: 5px var(--ve-authorization-action-padding-inline);
      border: var(--stroke-weight-base) solid
        var(--ve-authorization-action-border);
      border-radius: var(--ve-authorization-action-radius);
      background: var(--ve-authorization-action-background);
      color: var(--ve-authorization-action-text);
      font: inherit;
      font-size: var(--ve-authorization-body-size);
      line-height: 22px;
      white-space: nowrap;
      cursor: pointer;
      transition:
        background-color 160ms ease,
        border-color 160ms ease,
        color 160ms ease,
        opacity 160ms ease,
        transform 90ms ease-out;
    }

    .ve-authorization-action:hover {
      background: color-mix(
        var(--ve-authorization-action-background)
          var(--state-hover-subtle-ratio),
        var(--state-hover-subtle-mix)
      );
    }

    .ve-authorization-action:active {
      background: color-mix(
        var(--ve-authorization-action-background)
          var(--state-pressed-subtle-ratio),
        var(--state-pressed-subtle-mix)
      );
      transform: translateY(1px);
    }

    .ve-authorization-action:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--stroke-2)
        color-mix(currentColor 24%, transparent);
    }

    .ve-authorization-action-primary {
      border-color: transparent;
      background: var(--ve-authorization-primary-action-background);
      color: var(--ve-authorization-primary-action-text);
    }

    .ve-authorization-action-primary:hover {
      background: color-mix(
        var(--ve-authorization-primary-action-background)
          var(--state-hover-solid-ratio),
        var(--state-hover-solid-mix)
      );
    }

    .ve-authorization-action-primary:active {
      background: color-mix(
        var(--ve-authorization-primary-action-background)
          var(--state-pressed-solid-ratio),
        var(--state-pressed-solid-mix)
      );
    }

    :host([orientation='vertical']) .ve-authorization-action {
      width: 100%;
    }

    :host([disabled]) .ve-authorization {
      opacity: var(--ve-authorization-disabled-opacity);
    }

    :host([disabled]) .ve-authorization-action {
      cursor: not-allowed;
      pointer-events: none;
    }

    @media (max-width: 520px) {
      :host {
        min-inline-size: 0;
      }

      .ve-authorization-actions {
        flex-direction: column;
        align-items: stretch;
      }

      .ve-authorization-action {
        width: 100%;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _init_content, _init_extra__init_content, _init_orientation, _init_extra__init_orientation, _init_disabled, _init_extra__init_disabled, _init_denyText, _init_extra__init_denyText, _init_allowOnceText, _init_extra__init_allowOnceText, _init_allowAlwaysText, _init_extra__init_allowAlwaysText;
const VE_AUTHORIZATION_TAG = 've-authorization';
const ACTIONS = [
    'deny',
    'allow-once',
    'allow-always'
];
_dec = property({
    type: String
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: String,
    attribute: 'deny-text'
}), _dec4 = property({
    type: String,
    attribute: 'allow-once-text'
}), _dec5 = property({
    type: String,
    attribute: 'allow-always-text'
});
class VeAuthorizationElement extends VeElement {
    static{
        ({ e: [_init_content, _init_extra__init_content, _init_orientation, _init_extra__init_orientation, _init_disabled, _init_extra__init_disabled, _init_denyText, _init_extra__init_denyText, _init_allowOnceText, _init_extra__init_allowOnceText, _init_allowAlwaysText, _init_extra__init_allowAlwaysText] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "content"
            ],
            [
                _dec1,
                1,
                "orientation"
            ],
            [
                _dec2,
                1,
                "disabled"
            ],
            [
                _dec3,
                1,
                "denyText"
            ],
            [
                _dec4,
                1,
                "allowOnceText"
            ],
            [
                _dec5,
                1,
                "allowAlwaysText"
            ]
        ]));
    }
    static styles = ve_authorization;
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    #___private_content_1 = (()=>{
        const _value = _init_content(this, '');
        _init_extra__init_content(this);
        return _value;
    })();
    get content() {
        return this.#___private_content_1;
    }
    set content(_v) {
        this.#___private_content_1 = _v;
    }
    #___private_orientation_2 = (()=>{
        const _value = _init_orientation(this, 'horizontal');
        _init_extra__init_orientation(this);
        return _value;
    })();
    get orientation() {
        return this.#___private_orientation_2;
    }
    set orientation(_v) {
        this.#___private_orientation_2 = _v;
    }
    #___private_disabled_3 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_3;
    }
    set disabled(_v) {
        this.#___private_disabled_3 = _v;
    }
    get hasCustomContent() {
        return null !== this.querySelector('[slot="content"]');
    }
    #___private_denyText_4 = (()=>{
        const _value = _init_denyText(this, '拒绝');
        _init_extra__init_denyText(this);
        return _value;
    })();
    get denyText() {
        return this.#___private_denyText_4;
    }
    set denyText(_v) {
        this.#___private_denyText_4 = _v;
    }
    #___private_allowOnceText_5 = (()=>{
        const _value = _init_allowOnceText(this, '允许一次');
        _init_extra__init_allowOnceText(this);
        return _value;
    })();
    get allowOnceText() {
        return this.#___private_allowOnceText_5;
    }
    set allowOnceText(_v) {
        this.#___private_allowOnceText_5 = _v;
    }
    #___private_allowAlwaysText_6 = (()=>{
        const _value = _init_allowAlwaysText(this, '始终允许');
        _init_extra__init_allowAlwaysText(this);
        return _value;
    })();
    get allowAlwaysText() {
        return this.#___private_allowAlwaysText_6;
    }
    set allowAlwaysText(_v) {
        this.#___private_allowAlwaysText_6 = _v;
    }
    getActionText(action) {
        if ('deny' === action) return this.denyText;
        if ('allow-once' === action) return this.allowOnceText;
        return this.allowAlwaysText;
    }
    emitAction(action) {
        if (this.disabled) return;
        this.dispatchEvent(new CustomEvent('ve-authorization-action', {
            detail: {
                action
            },
            bubbles: true,
            composed: true
        }));
    }
    handleActionClick = (event)=>{
        const button = event.currentTarget;
        const action = button.value;
        if (ACTIONS.includes(action)) this.emitAction(action);
    };
    renderContent() {
        if (!this.content && !this.hasCustomContent) return nothing;
        return html`
      <pre class="ve-authorization-content" part="content">
<slot name="content">${this.content}</slot></pre
      >
    `;
    }
    renderAction(action) {
        const isPrimary = 'allow-always' === action;
        const classes = {
            've-authorization-action': true,
            've-authorization-action-primary': isPrimary
        };
        return html`
      <button
        class=${classMap(classes)}
        part="action"
        type="button"
        value=${action}
        ?disabled=${this.disabled}
        @click=${this.handleActionClick}
      >
        <span>${this.getActionText(action)}</span>
      </button>
    `;
    }
    renderActions() {
        return html`
      <slot name="actions">
        ${ACTIONS.map((action)=>this.renderAction(action))}
      </slot>
    `;
    }
    render() {
        return html`
      <div class="ve-authorization" part="container">
        <div class="ve-authorization-main">
          <div class="ve-authorization-body">
            <p class="ve-authorization-title" part="title">
              <slot name="title">${this.title}</slot>
            </p>
            ${this.renderContent()}
          </div>
        </div>
        <div class="ve-authorization-actions" part="actions">
          ${this.renderActions()}
        </div>
      </div>
    `;
    }
}
defineCustomElement(VE_AUTHORIZATION_TAG, VeAuthorizationElement);
export { VE_AUTHORIZATION_TAG, VeAuthorizationElement };
