// @ve-design/web/web/dist/ve-alert.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { infoIcon } from "./icons/info.js";
import { checkCircleIcon } from "./icons/check-circle.js";
import { warningIcon } from "./icons/warning.js";
import { errorIcon } from "./icons/error.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
import { closeIcon } from "./icons/close.js";
const veAlertTokenDefaults = css`
  :host {
    --ve-alert-border-width: var(--stroke-weight-base);
    --ve-alert-border-radius: var(--radius-lg);
    --ve-alert-gap: var(--space-s);
    --ve-alert-content-gap: var(--space-xxs);
    --ve-alert-line-height: var(--line-height-body);

    --ve-alert-padding-horizontal: var(--space-s);
    --ve-alert-padding-vertical: var(--space-xs);

    --ve-alert-font-family: var(--font-sans);
    --ve-alert-font-size-text-content: var(--text-body);
    --ve-alert-font-size-icon: var(--text-subsection-title);
    --ve-alert-icon-align-self: flex-start;
    --ve-alert-close-button-padding: var(--space-xxxs);
    --ve-alert-close-button-radius: var(--radius-xs);
    --ve-alert-close-icon-size: var(--text-body);
    --ve-alert-close-color: var(--color-icon-secondary);
    --ve-alert-close-color-hover: var(--color-icon-primary);
    --ve-alert-close-color-bg-hover: color-mix(
      var(--color-bg-secondary) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );

    --ve-alert-neutral-color-bg: var(--color-bg-secondary);
    --ve-alert-neutral-color-border: var(--color-border-default);
    --ve-alert-neutral-color-icon: var(--color-icon-primary);
    --ve-alert-neutral-color-text-content: var(--color-text-primary);

    --ve-alert-info-color-bg: var(--color-bg-info-secondary);
    --ve-alert-info-color-border: var(--color-border-info);
    --ve-alert-info-color-icon: var(--color-text-info);
    --ve-alert-info-color-text-content: var(--color-text-info);

    --ve-alert-warning-color-bg: var(--color-bg-warning-secondary);
    --ve-alert-warning-color-border: var(--color-border-warning);
    --ve-alert-warning-color-icon: var(--color-text-warning);
    --ve-alert-warning-color-text-content: var(--color-text-warning);

    --ve-alert-error-color-bg: var(--color-bg-danger-secondary);
    --ve-alert-error-color-border: var(--color-border-danger);
    --ve-alert-error-color-icon: var(--color-text-danger);
    --ve-alert-error-color-text-content: var(--color-text-danger);

    --ve-alert-success-color-bg: var(--color-bg-success-secondary);
    --ve-alert-success-color-border: var(--color-border-success);
    --ve-alert-success-color-icon: var(--color-text-success);
    --ve-alert-success-color-text-content: var(--color-text-success);
  }
`;
const token = veAlertTokenDefaults;
const ve_alert = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      width: 100%;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-alert {
      box-sizing: border-box;
      border: var(--ve-alert-border-width) solid transparent;
      border-radius: var(--ve-alert-border-radius);
      padding: var(--ve-alert-padding-vertical)
        var(--ve-alert-padding-horizontal);
      font-family: var(--ve-alert-font-family);
      font-size: var(--ve-alert-font-size-text-content);
      overflow: hidden;
      display: flex;
      gap: var(--ve-alert-gap);
      width: 100%;
      text-align: left;
      align-items: flex-start;
      line-height: var(--ve-alert-line-height);
    }

    .ve-alert-neutral {
      border-color: var(--ve-alert-neutral-color-border);
      background-color: var(--ve-alert-neutral-color-bg);
    }

    .ve-alert-info {
      border-color: var(--ve-alert-info-color-border);
      background-color: var(--ve-alert-info-color-bg);
    }

    .ve-alert-success {
      border-color: var(--ve-alert-success-color-border);
      background-color: var(--ve-alert-success-color-bg);
    }

    .ve-alert-warning {
      border-color: var(--ve-alert-warning-color-border);
      background-color: var(--ve-alert-warning-color-bg);
    }

    .ve-alert-error {
      border-color: var(--ve-alert-error-color-border);
      background-color: var(--ve-alert-error-color-bg);
    }

    .ve-alert-main {
      display: flex;
      flex: 1 1 auto;
      gap: var(--ve-alert-content-gap);
      align-items: flex-start;
      min-width: 0;
    }

    .ve-alert-content {
      flex: 1 1 auto;
      min-width: 0;
      font-size: var(--ve-alert-font-size-text-content);
      line-height: var(--ve-alert-line-height);
      word-break: break-word;
    }

    .ve-alert-neutral .ve-alert-content {
      color: var(--ve-alert-neutral-color-text-content);
    }

    .ve-alert-info .ve-alert-content {
      color: var(--ve-alert-info-color-text-content);
    }

    .ve-alert-success .ve-alert-content {
      color: var(--ve-alert-success-color-text-content);
    }

    .ve-alert-warning .ve-alert-content {
      color: var(--ve-alert-warning-color-text-content);
    }

    .ve-alert-error .ve-alert-content {
      color: var(--ve-alert-error-color-text-content);
    }

    .ve-alert-icon-wrapper {
      align-self: var(--ve-alert-icon-align-self);
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      width: var(--ve-alert-font-size-icon);
      height: calc(
        var(--ve-alert-font-size-text-content) * var(--ve-alert-line-height)
      );
      line-height: 1;
    }

    .ve-alert-icon-wrapper ve-icon,
    .ve-alert-icon-wrapper svg,
    .ve-alert-icon-wrapper ::slotted(svg) {
      font-size: var(--ve-alert-font-size-icon);
      width: 1em;
      height: 1em;
      color: inherit;
      flex: none;
    }

    .ve-alert-icon-wrapper ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-width: 100%;
      max-height: 100%;
      line-height: 1;
      flex: none;
    }

    .ve-alert-neutral .ve-alert-icon-wrapper {
      color: var(--ve-alert-neutral-color-icon);
    }

    .ve-alert-info .ve-alert-icon-wrapper {
      color: var(--ve-alert-info-color-icon);
    }

    .ve-alert-success .ve-alert-icon-wrapper {
      color: var(--ve-alert-success-color-icon);
    }

    .ve-alert-warning .ve-alert-icon-wrapper {
      color: var(--ve-alert-warning-color-icon);
    }

    .ve-alert-error .ve-alert-icon-wrapper {
      color: var(--ve-alert-error-color-icon);
    }

    .ve-alert-close-button {
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      padding: var(--ve-alert-close-button-padding);
      border: 0;
      border-radius: var(--ve-alert-close-button-radius);
      background: transparent;
      color: var(--ve-alert-close-color);
      line-height: 1;
      cursor: pointer;
    }

    .ve-alert-close-button:hover {
      background: var(--ve-alert-close-color-bg-hover);
      color: var(--ve-alert-close-color-hover);
    }

    .ve-alert-close-button ve-icon,
    .ve-alert-close-button svg {
      width: var(--ve-alert-close-icon-size);
      height: var(--ve-alert-close-icon-size);
      color: inherit;
      flex: none;
    }

    :host([dir='rtl']) .ve-alert,
    :host-context([dir='rtl']) .ve-alert {
      direction: rtl;
      text-align: right;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _init_type, _init_extra__init_type, _init_showIcon, _init_extra__init_showIcon, _init_closable, _init_extra__init_closable, _init_closed, _init_extra__init_closed, _init_iconSlot, _init_extra__init_iconSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot;
const VE_ALERT_TAG = 've-alert';
const DEFAULT_ICON_BY_TYPE = {
    neutral: infoIcon,
    info: infoIcon,
    success: checkCircleIcon,
    warning: warningIcon,
    error: errorIcon
};
const ICON_SLOT_SIZE = 'var(--ve-alert-font-size-icon)';
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
const FALSE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        return null !== value && 'false' !== value;
    },
    toAttribute (value) {
        return value ? '' : null;
    }
};
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    attribute: 'show-icon',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec2 = property({
    attribute: 'closable',
    reflect: true,
    converter: FALSE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec3 = state(), _dec4 = query('slot[name="icon"]'), _dec5 = state();
class VeAlertElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_showIcon, _init_extra__init_showIcon, _init_closable, _init_extra__init_closable, _init_closed, _init_extra__init_closed, _init_iconSlot, _init_extra__init_iconSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "type"
            ],
            [
                _dec1,
                1,
                "showIcon"
            ],
            [
                _dec2,
                1,
                "closable"
            ],
            [
                _dec3,
                1,
                "closed"
            ],
            [
                _dec4,
                1,
                "iconSlot"
            ],
            [
                _dec5,
                1,
                "hasIconSlot"
            ]
        ]));
    }
    static styles = ve_alert;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'neutral');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_showIcon_2 = (()=>{
        const _value = _init_showIcon(this, true);
        _init_extra__init_showIcon(this);
        return _value;
    })();
    get showIcon() {
        return this.#___private_showIcon_2;
    }
    set showIcon(_v) {
        this.#___private_showIcon_2 = _v;
    }
    #___private_closable_3 = (()=>{
        const _value = _init_closable(this, false);
        _init_extra__init_closable(this);
        return _value;
    })();
    get closable() {
        return this.#___private_closable_3;
    }
    set closable(_v) {
        this.#___private_closable_3 = _v;
    }
    #___private_closed_4 = (()=>{
        const _value = _init_closed(this, false);
        _init_extra__init_closed(this);
        return _value;
    })();
    get closed() {
        return this.#___private_closed_4;
    }
    set closed(_v) {
        this.#___private_closed_4 = _v;
    }
    #___private_iconSlot_5 = (()=>{
        const _value = _init_iconSlot(this, null);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_5;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_5 = _v;
    }
    #___private_hasIconSlot_6 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_6;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_6 = _v;
    }
    firstUpdated() {
        this.syncSlotState();
    }
    syncSlotState() {
        this.hasIconSlot = hasAssignedSlotContent(this.iconSlot) || hasNamedSlottedContent(this, 'icon');
        this.syncSlottedIconSize();
    }
    syncSlottedIconSize() {
        const icons = this.iconSlot?.assignedElements({
            flatten: true
        }).filter((element)=>element.localName === VE_ICON_TAG);
        icons?.forEach((icon)=>{
            if ('' === `${icon.getAttribute('size') ?? ''}`.trim() && '' === `${icon.size ?? ''}`.trim()) icon.size = ICON_SLOT_SIZE;
        });
    }
    handleSlotChange() {
        this.syncSlotState();
    }
    handleCloseClick(sourceEvent) {
        const shouldClose = this.dispatchEvent(new CustomEvent('ve-close', {
            detail: {
                sourceEvent
            },
            bubbles: true,
            composed: true,
            cancelable: true
        }));
        if (shouldClose) this.closed = true;
    }
    renderDefaultIcon() {
        const icon = DEFAULT_ICON_BY_TYPE[this.type] ?? DEFAULT_ICON_BY_TYPE.info;
        return html`<ve-icon
      .definition=${icon}
      size="var(--ve-alert-font-size-icon)"
    ></ve-icon>`;
    }
    renderIcon() {
        if (!this.showIcon) return nothing;
        return html`
      <div class="ve-alert-icon-wrapper" part="icon" aria-hidden="true">
        <slot name="icon" @slotchange=${this.handleSlotChange}></slot>
        ${this.hasIconSlot ? nothing : this.renderDefaultIcon()}
      </div>
    `;
    }
    renderCloseButton() {
        if (!this.closable) return nothing;
        return html`
      <button
        class="ve-alert-close-button"
        part="close-button"
        type="button"
        aria-label="Close"
        @click=${this.handleCloseClick}
      >
        <ve-icon .definition=${closeIcon} size="var(--ve-alert-close-icon-size)"></ve-icon>
      </button>
    `;
    }
    render() {
        if (this.closed) return nothing;
        const alertClass = classMap({
            've-alert': true,
            [`ve-alert-${this.type}`]: true,
            've-alert-closable': this.closable
        });
        return html`
      <div class=${alertClass} part="alert" role="alert">
        <div class="ve-alert-main">
          ${this.renderIcon()}
          <div class="ve-alert-content" part="content">
            <slot></slot>
          </div>
        </div>
        ${this.renderCloseButton()}
      </div>
    `;
    }
}
defineCustomElement(VE_ALERT_TAG, VeAlertElement);
export { VE_ALERT_TAG, VeAlertElement };
