// @ve-design/web/web/dist/ve-bubble-divider.js
import { css, html, nothing, svg } from "lit";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
const veBubbleDividerToken = css`
  :host {
    /* ── 通用 ── */
    --ve-bubble-divider-font-family: var(--font-sans);
    --ve-bubble-divider-font-weight: var(--font-weight-normal);
    --ve-bubble-divider-box-sizing: border-box;
    --ve-bubble-divider-width: 100%;
    --ve-bubble-divider-max-width: 100%;

    /* ── 分割线 ── */
    --ve-bubble-divider-line-height: 32px;
    --ve-bubble-divider-line-padding-block: var(--space-xxxs, 4px);
    --ve-bubble-divider-line-size: var(--stroke-0_5, 0.5px);
    --ve-bubble-divider-line-color: var(--color-border-default, #e5e6eb);
    --ve-bubble-divider-label-height: 20px;
    --ve-bubble-divider-label-padding-inline: var(--space-xs, 8px);
    --ve-bubble-divider-label-color: var(--color-text-tertiary, #86909c);
    --ve-bubble-divider-label-font-size: var(--text-caption, 12px);
    --ve-bubble-divider-label-font-weight: var(--font-weight-normal, 400);
    --ve-bubble-divider-label-line-height: 1.5;

    /* ── 系统提示 ── */
    --ve-bubble-divider-notice-padding-block: var(--space-xs, 8px);
    --ve-bubble-divider-notice-padding-inline: var(--space-s, 12px);
    --ve-bubble-divider-notice-radius: var(--radius-lg, 8px);
    --ve-bubble-divider-notice-border-width: var(--stroke-0_5, 0.5px);
    --ve-bubble-divider-notice-border-color: var(
      --color-border-default,
      #e5e6eb
    );
    --ve-bubble-divider-notice-background: var(--color-bg-secondary, #f7f8fa);
    --ve-bubble-divider-notice-color: var(--color-text-secondary, #4e5969);

    /* ── 系统提示 error 状态 ── */
    --ve-bubble-divider-notice-error-background: color-mix(
      in srgb,
      var(--color-bg-danger, #f53f3f) 8%,
      var(--color-bg-base, #ffffff)
    );
    --ve-bubble-divider-notice-error-border-color: color-mix(
      in srgb,
      var(--color-bg-danger, #f53f3f) 24%,
      var(--color-bg-base, #ffffff)
    );
    --ve-bubble-divider-notice-error-color: var(
      --color-text-danger,
      #f53f3f
    );
    --ve-bubble-divider-notice-gap: var(--space-xxs, 6px);
    --ve-bubble-divider-notice-font-size: var(--text-body, 14px);
    --ve-bubble-divider-notice-line-height: 22px;
    --ve-bubble-divider-notice-icon-size: 16px;
    --ve-bubble-divider-notice-icon-box-height: 22px;
    --ve-bubble-divider-notice-icon-stroke-width: 1.35px;

    /* ── 系统提示操作按钮 ── */
    --ve-bubble-divider-action-height: 28px;
    --ve-bubble-divider-action-padding-block: 3px;
    --ve-bubble-divider-action-padding-inline: var(--space-xs, 8px);
    --ve-bubble-divider-action-radius: var(--radius-md, 6px);
    --ve-bubble-divider-action-border-width: var(--stroke-0_5, 0.5px);
    --ve-bubble-divider-action-border-color: var(
      --color-border-strong,
      #c9cdd4
    );
    --ve-bubble-divider-action-background: transparent;
    --ve-bubble-divider-action-color: var(--color-text-primary, #1d2129);
    --ve-bubble-divider-action-font-size: var(--text-body, 14px);
    --ve-bubble-divider-action-hover-ratio: var(
      --state-hover-subtle-ratio,
      92%
    );
    --ve-bubble-divider-action-hover-mix: var(
      --state-hover-subtle-mix,
      #000000
    );
    --ve-bubble-divider-action-pressed-ratio: var(
      --state-pressed-subtle-ratio,
      88%
    );
    --ve-bubble-divider-action-pressed-mix: var(
      --state-pressed-subtle-mix,
      #000000
    );
    --ve-bubble-divider-action-transition: background-color 160ms ease,
      border-color 160ms ease, transform 90ms ease-out;
    --ve-bubble-divider-action-pressed-transform: translateY(1px);
    --ve-bubble-divider-focus-ring-width: var(--stroke-2, 2px);
    --ve-bubble-divider-focus-ring-color: color-mix(
      in srgb,
      currentColor 24%,
      transparent
    );
  }
`;
const token = veBubbleDividerToken;
const ve_bubble_divider = [
    componentReset,
    token,
    css`
    :host {
      box-sizing: var(--ve-bubble-divider-box-sizing);
      display: flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-bubble-divider-width);
      max-inline-size: var(--ve-bubble-divider-max-width);
      font-family: var(--ve-bubble-divider-font-family);
      font-weight: var(--ve-bubble-divider-font-weight);
    }

    /* ── 分割线形态 ── */
    :host([type='divider']) {
      block-size: var(--ve-bubble-divider-line-height);
      padding-block: var(--ve-bubble-divider-line-padding-block);
    }

    /* ── 系统提示形态 ── */
    :host([type='notice']) {
      justify-content: flex-start;
      padding-block: var(--ve-bubble-divider-notice-padding-block);
      padding-inline: var(--ve-bubble-divider-notice-padding-inline);
      border: var(--ve-bubble-divider-notice-border-width) solid
        var(--ve-bubble-divider-notice-border-color);
      border-radius: var(--ve-bubble-divider-notice-radius);
      background: var(--ve-bubble-divider-notice-background);
      color: var(--ve-bubble-divider-notice-color);
      gap: var(--ve-bubble-divider-notice-gap);
    }

    /* ── error 状态 ── */
    :host([type='notice'][status='error']) {
      background: var(--ve-bubble-divider-notice-error-background);
      border-color: var(--ve-bubble-divider-notice-error-border-color);
      color: var(--ve-bubble-divider-notice-error-color);
    }

    :host([hidden]) {
      display: none;
    }

    *,
    *::before,
    *::after {
      box-sizing: border-box;
    }

    /* ── 分割线 ── */
    .ve-bubble-divider__line {
      flex: 1 1 0;
      min-inline-size: 0;
      block-size: var(--ve-bubble-divider-line-size);
      background: var(--ve-bubble-divider-line-color);
    }

    .ve-bubble-divider__label {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      block-size: var(--ve-bubble-divider-label-height);
      padding-inline: var(--ve-bubble-divider-label-padding-inline);
      color: var(--ve-bubble-divider-label-color);
      font-size: var(--ve-bubble-divider-label-font-size);
      font-weight: var(--ve-bubble-divider-label-font-weight);
      line-height: var(--ve-bubble-divider-label-line-height);
      white-space: nowrap;
      word-break: break-word;
    }

    /* ── 系统提示内容区 ── */
    .ve-bubble-divider__body {
      display: flex;
      flex: 1 1 0;
      align-items: center;
      gap: var(--ve-bubble-divider-notice-gap);
      min-inline-size: 0;
    }

    .ve-bubble-divider__icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      block-size: var(--ve-bubble-divider-notice-icon-box-height);
      flex: none;
      inline-size: var(--ve-bubble-divider-notice-icon-size);
      color: currentColor;
    }

    .ve-bubble-divider__icon-svg {
      display: block;
      inline-size: 100%;
      block-size: 100%;
      fill: none;
      stroke: currentColor;
      stroke-width: var(--ve-bubble-divider-notice-icon-stroke-width);
      stroke-linecap: round;
      stroke-linejoin: round;
    }

    .ve-bubble-divider__icon ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-inline-size: 100%;
      max-block-size: 100%;
      line-height: 1;
      flex: none;
    }

    .ve-bubble-divider__message {
      flex: 1 1 0;
      min-inline-size: 0;
      color: currentColor;
      font-size: var(--ve-bubble-divider-notice-font-size);
      font-weight: var(--ve-bubble-divider-font-weight);
      line-height: var(--ve-bubble-divider-notice-line-height);
      word-break: break-word;
    }

    /* ── 操作按钮 ── */
    .ve-bubble-divider__action {
      appearance: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      min-block-size: var(--ve-bubble-divider-action-height);
      block-size: var(--ve-bubble-divider-action-height);
      padding-block: var(--ve-bubble-divider-action-padding-block);
      padding-inline: var(--ve-bubble-divider-action-padding-inline);
      border: var(--ve-bubble-divider-action-border-width) solid
        var(--ve-bubble-divider-action-border-color);
      border-radius: var(--ve-bubble-divider-action-radius);
      background: var(--ve-bubble-divider-action-background);
      color: var(--ve-bubble-divider-action-color);
      font: inherit;
      font-size: var(--ve-bubble-divider-action-font-size);
      line-height: var(--ve-bubble-divider-notice-line-height);
      white-space: nowrap;
      cursor: pointer;
      transition: var(--ve-bubble-divider-action-transition);
    }

    .ve-bubble-divider__action ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-inline-size: 100%;
      line-height: inherit;
    }

    .ve-bubble-divider__action:hover {
      background: color-mix(
        in srgb,
        var(--ve-bubble-divider-action-background)
          var(--ve-bubble-divider-action-hover-ratio),
        var(--ve-bubble-divider-action-hover-mix)
      );
    }

    .ve-bubble-divider__action:active {
      background: color-mix(
        in srgb,
        var(--ve-bubble-divider-action-background)
          var(--ve-bubble-divider-action-pressed-ratio),
        var(--ve-bubble-divider-action-pressed-mix)
      );
      transform: var(--ve-bubble-divider-action-pressed-transform);
    }

    .ve-bubble-divider__action:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--ve-bubble-divider-focus-ring-width)
        var(--ve-bubble-divider-focus-ring-color);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-bubble-divider__action {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _init_type, _init_extra__init_type, _init_showLabel, _init_extra__init_showLabel, _init_message, _init_extra__init_message, _init_actionText, _init_extra__init_actionText, _init_showAction, _init_extra__init_showAction, _init_status, _init_extra__init_status, _init_iconSlot, _init_extra__init_iconSlot, _init_messageSlot, _init_extra__init_messageSlot, _init_actionSlot, _init_extra__init_actionSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasMessageSlot, _init_extra__init_hasMessageSlot, _init_hasActionSlot, _init_extra__init_hasActionSlot;
const VE_BUBBLE_DIVIDER_TAG = 've-bubble-divider';
_dec = property({
    type: String,
    reflect: true
}), _dec1 = property({
    type: Boolean,
    reflect: true,
    attribute: 'show-label'
}), _dec2 = property({
    type: String
}), _dec3 = property({
    type: String,
    attribute: 'action-text'
}), _dec4 = property({
    type: Boolean,
    attribute: 'show-action'
}), _dec5 = property({
    type: String,
    reflect: true
}), _dec6 = query('slot[name="icon"]'), _dec7 = query('slot:not([name])'), _dec8 = query('slot[name="action"]'), _dec9 = state(), _dec10 = state(), _dec11 = state();
class VeBubbleDividerElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_showLabel, _init_extra__init_showLabel, _init_message, _init_extra__init_message, _init_actionText, _init_extra__init_actionText, _init_showAction, _init_extra__init_showAction, _init_status, _init_extra__init_status, _init_iconSlot, _init_extra__init_iconSlot, _init_messageSlot, _init_extra__init_messageSlot, _init_actionSlot, _init_extra__init_actionSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasMessageSlot, _init_extra__init_hasMessageSlot, _init_hasActionSlot, _init_extra__init_hasActionSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "type"
            ],
            [
                _dec1,
                1,
                "showLabel"
            ],
            [
                _dec2,
                1,
                "message"
            ],
            [
                _dec3,
                1,
                "actionText"
            ],
            [
                _dec4,
                1,
                "showAction"
            ],
            [
                _dec5,
                1,
                "status"
            ],
            [
                _dec6,
                1,
                "iconSlot"
            ],
            [
                _dec7,
                1,
                "messageSlot"
            ],
            [
                _dec8,
                1,
                "actionSlot"
            ],
            [
                _dec9,
                1,
                "hasIconSlot"
            ],
            [
                _dec10,
                1,
                "hasMessageSlot"
            ],
            [
                _dec11,
                1,
                "hasActionSlot"
            ]
        ]));
    }
    static styles = ve_bubble_divider;
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'divider');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_showLabel_2 = (()=>{
        const _value = _init_showLabel(this, true);
        _init_extra__init_showLabel(this);
        return _value;
    })();
    get showLabel() {
        return this.#___private_showLabel_2;
    }
    set showLabel(_v) {
        this.#___private_showLabel_2 = _v;
    }
    #___private_message_3 = (()=>{
        const _value = _init_message(this, '');
        _init_extra__init_message(this);
        return _value;
    })();
    get message() {
        return this.#___private_message_3;
    }
    set message(_v) {
        this.#___private_message_3 = _v;
    }
    #___private_actionText_4 = (()=>{
        const _value = _init_actionText(this, '');
        _init_extra__init_actionText(this);
        return _value;
    })();
    get actionText() {
        return this.#___private_actionText_4;
    }
    set actionText(_v) {
        this.#___private_actionText_4 = _v;
    }
    #___private_showAction_5 = (()=>{
        const _value = _init_showAction(this, false);
        _init_extra__init_showAction(this);
        return _value;
    })();
    get showAction() {
        return this.#___private_showAction_5;
    }
    set showAction(_v) {
        this.#___private_showAction_5 = _v;
    }
    #___private_status_6 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_6;
    }
    set status(_v) {
        this.#___private_status_6 = _v;
    }
    #___private_iconSlot_7 = (()=>{
        const _value = _init_iconSlot(this, null);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_7;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_7 = _v;
    }
    #___private_messageSlot_8 = (()=>{
        const _value = _init_messageSlot(this, null);
        _init_extra__init_messageSlot(this);
        return _value;
    })();
    get messageSlot() {
        return this.#___private_messageSlot_8;
    }
    set messageSlot(_v) {
        this.#___private_messageSlot_8 = _v;
    }
    #___private_actionSlot_9 = (()=>{
        const _value = _init_actionSlot(this, null);
        _init_extra__init_actionSlot(this);
        return _value;
    })();
    get actionSlot() {
        return this.#___private_actionSlot_9;
    }
    set actionSlot(_v) {
        this.#___private_actionSlot_9 = _v;
    }
    #___private_hasIconSlot_10 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_10;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_10 = _v;
    }
    #___private_hasMessageSlot_11 = (()=>{
        const _value = _init_hasMessageSlot(this, false);
        _init_extra__init_hasMessageSlot(this);
        return _value;
    })();
    get hasMessageSlot() {
        return this.#___private_hasMessageSlot_11;
    }
    set hasMessageSlot(_v) {
        this.#___private_hasMessageSlot_11 = _v;
    }
    #___private_hasActionSlot_12 = (()=>{
        const _value = _init_hasActionSlot(this, false);
        _init_extra__init_hasActionSlot(this);
        return _value;
    })();
    get hasActionSlot() {
        return this.#___private_hasActionSlot_12;
    }
    set hasActionSlot(_v) {
        this.#___private_hasActionSlot_12 = _v;
    }
    lightDomSlotObserver;
    get hasActionContent() {
        return this.hasActionSlot || hasNamedSlottedContent(this, 'action');
    }
    get shouldRenderAction() {
        return this.showAction || this.hasActionContent;
    }
    connectedCallback() {
        super.connectedCallback();
        this.observeLightDomSlots();
        this.syncSlotState();
    }
    firstUpdated() {
        this.syncSlotState();
    }
    disconnectedCallback() {
        this.lightDomSlotObserver?.disconnect();
        this.lightDomSlotObserver = void 0;
        super.disconnectedCallback();
    }
    observeLightDomSlots() {
        if ("u" < typeof MutationObserver || this.lightDomSlotObserver) return;
        this.lightDomSlotObserver = new MutationObserver(()=>{
            this.syncSlotState();
        });
        this.lightDomSlotObserver.observe(this, {
            childList: true,
            subtree: false,
            attributes: true,
            attributeFilter: [
                'slot'
            ]
        });
    }
    syncSlotState() {
        this.hasIconSlot = hasAssignedSlotContent(this.iconSlot) || hasNamedSlottedContent(this, 'icon');
        this.hasMessageSlot = hasAssignedSlotContent(this.messageSlot);
        this.hasActionSlot = hasAssignedSlotContent(this.actionSlot) || hasNamedSlottedContent(this, 'action');
    }
    handleSlotChange() {
        this.syncSlotState();
    }
    handleActionClick() {
        this.dispatchEvent(new CustomEvent('ve-bubble-divider-action', {
            bubbles: true,
            composed: true,
            detail: {
                type: this.type,
                status: this.status
            }
        }));
    }
    renderInfoIcon() {
        return svg`
      <svg
        class="ve-bubble-divider__icon-svg"
        viewBox="0 0 18 18"
        aria-hidden="true"
        focusable="false"
      >
        <circle cx="9" cy="9" r="6.5"></circle>
        <path d="M9 8.25v4.25"></path>
        <path d="M9 5.5h.01"></path>
      </svg>
    `;
    }
    renderDivider() {
        return html`
      <div class="ve-bubble-divider__line" part="line"></div>
      ${this.showLabel ? html`
            <span class="ve-bubble-divider__label" part="label">
              <slot name="label"></slot>
            </span>
          ` : nothing}
      <div class="ve-bubble-divider__line" part="line"></div>
    `;
    }
    renderNotice() {
        return html`
      <div class="ve-bubble-divider__body" part="body">
        <span class="ve-bubble-divider__icon" part="icon">
          <slot name="icon" @slotchange=${this.handleSlotChange}></slot>
          ${this.hasIconSlot ? nothing : this.renderInfoIcon()}
        </span>
        <span class="ve-bubble-divider__message" part="message">
          <slot @slotchange=${this.handleSlotChange}></slot>
          ${this.hasMessageSlot ? nothing : this.message}
        </span>
      </div>
      ${this.shouldRenderAction ? html`
            <button
              class="ve-bubble-divider__action"
              part="action"
              type="button"
              @click=${this.handleActionClick}
            >
              <slot name="action" @slotchange=${this.handleSlotChange}></slot>
              ${this.hasActionContent ? nothing : this.actionText}
            </button>
          ` : nothing}
    `;
    }
    render() {
        if ('divider' === this.type) return this.renderDivider();
        if ('notice' === this.type) return this.renderNotice();
        return nothing;
    }
}
defineCustomElement(VE_BUBBLE_DIVIDER_TAG, VeBubbleDividerElement);
export { VE_BUBBLE_DIVIDER_TAG, VeBubbleDividerElement };
