// @ve-design/web/web/dist/ve-checkbox.js
import { css, html } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
const veCheckboxTokenDefaults = css`
  :host {
    --ve-checkbox-font-family: var(--font-sans);
    --ve-checkbox-font-size: var(--text-body);

    --ve-checkbox-mask-border-width: var(--stroke-weight-strong);
    --ve-checkbox-mask-border-style: solid;
    --ve-checkbox-mask-border-radius: var(--radius-xs);

    --ve-checkbox-mask-size: 16px;
    --ve-checkbox-mask-surface-size: var(--ve-checkbox-mask-size);
    --ve-checkbox-offset-inline-start: calc(
      (var(--ve-checkbox-mask-surface-size) - var(--ve-checkbox-mask-size)) / 2
    );

    --ve-checkbox-mask-color-bg: var(--color-bg-base);
    --ve-checkbox-mask-color-bg-checked: var(--color-bg-primary);
    --ve-checkbox-mask-color-bg-disabled: var(--color-bg-muted);

    --ve-checkbox-mask-color-border: var(--color-border-strong);
    --ve-checkbox-mask-color-border-hover: var(--color-border-interactive);
    --ve-checkbox-mask-color-border-checked: transparent;
    --ve-checkbox-mask-color-border-disabled: var(--color-border-default);

    --ve-checkbox-text-color: var(--color-text-primary);

    --ve-checkbox-group-gap: var(--space-s);
    --ve-checkbox-text-mask-spacing: var(--space-xxs);
    --ve-checkbox-group-line-height-vertical: 32px;

    --ve-checkbox-check-icon-size: 12px;
    --ve-checkbox-check-icon-color: var(--color-text-foreground);

    --ve-checkbox-indeterminate-width: 10px;
    --ve-checkbox-indeterminate-height: 2px;
    --ve-checkbox-indeterminate-color: var(--color-text-foreground);

    --ve-checkbox-checked-hover-ratio: var(--state-hover-solid-ratio);
    --ve-checkbox-checked-hover-mix: var(--state-hover-solid-mix);
    --ve-checkbox-disabled-opacity: var(--state-disabled-ratio);
    --ve-checkbox-focus-ring-color: color-mix(
      var(--color-bg-primary) 24%,
      transparent
    );

    --ve-checkbox-transition-duration-fast: var(--motion-fast);
    --ve-checkbox-transition-duration-base: var(--motion-base);
    --ve-checkbox-transition-timing-linear: linear;
    --ve-checkbox-transition-timing-overshoot: cubic-bezier(
      0.34,
      1.56,
      0.64,
      1
    );
  }
`;
const token = veCheckboxTokenDefaults;
const ve_checkbox = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      align-items: center;
      vertical-align: middle;
      color: var(--ve-checkbox-text-color);
      font-size: var(--ve-checkbox-font-size);
      font-family: var(--ve-checkbox-font-family);
      user-select: none;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-checkbox {
      position: relative;
      display: inline-flex;
      align-items: center;
      box-sizing: border-box;
      cursor: pointer;
      padding-inline-start: var(--ve-checkbox-offset-inline-start);
      color: var(--ve-checkbox-text-color);
      font-size: var(--ve-checkbox-font-size);
      font-family: var(--ve-checkbox-font-family);
      line-height: 1.4;
      user-select: none;
      min-inline-size: 0;
    }

    .ve-checkbox-input {
      position: absolute;
      inset-block-start: 0;
      inset-inline-start: 0;
      inline-size: var(--ve-checkbox-mask-surface-size);
      block-size: var(--ve-checkbox-mask-surface-size);
      margin: 0;
      opacity: 0;
      cursor: inherit;
    }

    .ve-checkbox-text {
      color: var(--ve-checkbox-text-color);
      margin-inline-start: var(--ve-checkbox-text-mask-spacing);
      min-inline-size: 0;
      line-height: 1.4;
      overflow-wrap: anywhere;
    }

    .ve-checkbox-text--hidden {
      display: none;
    }

    .ve-checkbox-icon-hover {
      position: relative;
      inset-block-start: 0;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-checkbox-mask-surface-size);
      block-size: var(--ve-checkbox-mask-surface-size);
      line-height: 1;
      vertical-align: middle;
      cursor: inherit;
    }

    .ve-checkbox-mask {
      position: relative;
      box-sizing: border-box;
      inline-size: var(--ve-checkbox-mask-size);
      block-size: var(--ve-checkbox-mask-size);
      border: var(--ve-checkbox-mask-border-width)
        var(--ve-checkbox-mask-border-style)
        var(--ve-checkbox-mask-color-border);
      border-radius: var(--ve-checkbox-mask-border-radius);
      background-color: var(--ve-checkbox-mask-color-bg);
      user-select: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      transition:
        background-color var(--ve-checkbox-transition-duration-fast)
          var(--ve-checkbox-transition-timing-linear),
        border-color var(--ve-checkbox-transition-duration-fast)
          var(--ve-checkbox-transition-timing-linear);
    }

    .ve-checkbox-mask-icon {
      position: relative;
      display: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-checkbox-check-icon-size);
      block-size: var(--ve-checkbox-check-icon-size);
      margin: 0 auto;
      color: var(--ve-checkbox-check-icon-color);
      line-height: 0;
    }

    .ve-checkbox-mask-icon-default {
      inline-size: 100%;
      block-size: 100%;
    }

    .ve-checkbox-indeterminate-mark {
      position: absolute;
      inset-block-start: 50%;
      inset-inline-start: 50%;
      display: block;
      inline-size: var(--ve-checkbox-indeterminate-width);
      block-size: var(--ve-checkbox-indeterminate-height);
      border-radius: var(--radius-full);
      background: var(--ve-checkbox-indeterminate-color);
      transform: translate(-50%, -50%);
      visibility: hidden;
    }

    .ve-checkbox-input:focus-visible
      + .ve-checkbox-icon-hover
      .ve-checkbox-mask {
      box-shadow: 0 0 0 2px var(--ve-checkbox-focus-ring-color);
    }

    .ve-checkbox:not(.ve-checkbox--disabled):not(.ve-checkbox--checked):not(
        .ve-checkbox--indeterminate
      ):hover
      .ve-checkbox-mask {
      border-color: var(--ve-checkbox-mask-color-border-hover);
    }

    .ve-checkbox--checked .ve-checkbox-mask {
      border-color: var(--ve-checkbox-mask-color-border-checked);
      background-color: var(--ve-checkbox-mask-color-bg-checked);
    }

    .ve-checkbox--checked .ve-checkbox-mask-icon {
      display: inline-flex;
    }

    .ve-checkbox--indeterminate .ve-checkbox-mask {
      border-color: var(--ve-checkbox-mask-color-border-checked);
      background-color: var(--ve-checkbox-mask-color-bg-checked);
    }

    .ve-checkbox--indeterminate .ve-checkbox-mask-icon {
      display: none;
    }

    .ve-checkbox--indeterminate .ve-checkbox-indeterminate-mark {
      visibility: visible;
    }

    .ve-checkbox--checked:not(.ve-checkbox--disabled):hover .ve-checkbox-mask,
    .ve-checkbox--indeterminate:not(.ve-checkbox--disabled):hover
      .ve-checkbox-mask {
      background-color: color-mix(
        var(--ve-checkbox-mask-color-bg-checked)
          var(--ve-checkbox-checked-hover-ratio),
        var(--ve-checkbox-checked-hover-mix)
      );
    }

    .ve-checkbox--disabled {
      cursor: not-allowed;
      opacity: var(--ve-checkbox-disabled-opacity);
    }

    .ve-checkbox--disabled:hover .ve-checkbox-mask {
      border-color: var(--ve-checkbox-mask-color-border-disabled);
    }

    .ve-checkbox--checked:hover .ve-checkbox-mask,
    .ve-checkbox--indeterminate:hover .ve-checkbox-mask {
      border-color: var(--ve-checkbox-mask-color-border-checked);
    }

    .ve-checkbox--disabled .ve-checkbox-mask {
      border-color: var(--ve-checkbox-mask-color-border-disabled);
      background-color: var(--ve-checkbox-mask-color-bg-disabled);
    }

    .ve-checkbox--disabled.ve-checkbox--checked .ve-checkbox-mask,
    .ve-checkbox--disabled.ve-checkbox--checked:hover .ve-checkbox-mask,
    .ve-checkbox--disabled.ve-checkbox--indeterminate .ve-checkbox-mask,
    .ve-checkbox--disabled.ve-checkbox--indeterminate:hover .ve-checkbox-mask {
      border-color: var(--ve-checkbox-mask-color-border-disabled);
      background-color: var(--ve-checkbox-mask-color-bg-disabled);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-checkbox-mask,
      .ve-checkbox-mask-icon,
      .ve-checkbox-indeterminate-mark {
        transition: none;
      }
    }
  `
];
const ve_checkbox_group = [
    componentReset,
    token,
    css`
    :host {
      display: inline-block;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-checkbox-group {
      display: inline-block;
    }

    .ve-checkbox-group-direction-horizontal ::slotted(ve-checkbox),
    .ve-checkbox-group-direction-horizontal .ve-checkbox-group-option-item {
      margin-inline-end: var(--ve-checkbox-group-gap);
    }

    .ve-checkbox-group-direction-vertical ::slotted(ve-checkbox),
    .ve-checkbox-group-direction-vertical .ve-checkbox-group-option-item {
      display: block;
      margin-inline-end: 0;
      line-height: var(--ve-checkbox-group-line-height-vertical);
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _init_checked, _init_extra__init_checked, _init_defaultChecked, _init_extra__init_defaultChecked, _init_indeterminate, _init_extra__init_indeterminate, _init_disabled, _init_extra__init_disabled, _init_value, _init_extra__init_value, _init_groupDisabled, _init_extra__init_groupDisabled, _init_inputElement, _init_extra__init_inputElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_hasLabel, _init_extra__init_hasLabel, _dec9, _dec10, _dec11, _dec12, _dec13, _init_value1, _init_extra__init_value1, _init_defaultValue, _init_extra__init_defaultValue, _init_disabled1, _init_extra__init_disabled1, _init_direction, _init_extra__init_direction, _init_options, _init_extra__init_options;
const VE_CHECKBOX_TAG = 've-checkbox';
const VE_CHECKBOX_GROUP_TAG = 've-checkbox-group';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby'
];
const CHECKBOX_VALUE_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        return value ?? '';
    },
    toAttribute (value) {
        return String(value);
    }
};
const CHECKBOX_VALUE_ARRAY_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        if (null === value || 0 === value.trim().length) return [];
        try {
            const parsed = JSON.parse(value);
            if (!Array.isArray(parsed)) return [];
            return parsed.filter((item)=>isCheckboxValue(item));
        } catch  {
            return value.split(',').map((item)=>item.trim()).filter((item)=>item.length > 0);
        }
    },
    toAttribute (value) {
        if (0 === value.length) return '[]';
        return JSON.stringify(value);
    }
};
const CHECKBOX_GROUP_OPTIONS_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        if (null === value || 0 === value.trim().length) return [];
        try {
            const parsed = JSON.parse(value);
            return Array.isArray(parsed) ? parsed : [];
        } catch  {
            return [];
        }
    },
    toAttribute (value) {
        try {
            return JSON.stringify(value);
        } catch  {
            return null;
        }
    }
};
function isCheckboxValue(value) {
    return 'string' == typeof value || 'number' == typeof value;
}
function hasRenderableSlotNode(node) {
    if (node.nodeType === Node.ELEMENT_NODE) return true;
    if (node.nodeType !== Node.TEXT_NODE) return false;
    return !!node.textContent?.trim().length;
}
function isRenderableTemplateValue(value) {
    return null != value && 'boolean' != typeof value && 'function' != typeof value && 'symbol' != typeof value;
}
_dec = property({
    type: Boolean,
    reflect: true
}), _dec1 = property({
    type: Boolean,
    attribute: 'default-checked'
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    converter: CHECKBOX_VALUE_ATTRIBUTE_CONVERTER
}), _dec5 = property({
    attribute: false
}), _dec6 = query('.ve-checkbox-input'), _dec7 = query('slot:not([name])'), _dec8 = state();
class VeCheckboxElement extends VeElement {
    static{
        ({ e: [_init_checked, _init_extra__init_checked, _init_defaultChecked, _init_extra__init_defaultChecked, _init_indeterminate, _init_extra__init_indeterminate, _init_disabled, _init_extra__init_disabled, _init_value, _init_extra__init_value, _init_groupDisabled, _init_extra__init_groupDisabled, _init_inputElement, _init_extra__init_inputElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_hasLabel, _init_extra__init_hasLabel] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "checked"
            ],
            [
                _dec1,
                1,
                "defaultChecked"
            ],
            [
                _dec2,
                1,
                "indeterminate"
            ],
            [
                _dec3,
                1,
                "disabled"
            ],
            [
                _dec4,
                1,
                "value"
            ],
            [
                _dec5,
                1,
                "groupDisabled"
            ],
            [
                _dec6,
                1,
                "inputElement"
            ],
            [
                _dec7,
                1,
                "defaultSlot"
            ],
            [
                _dec8,
                1,
                "hasLabel"
            ]
        ]));
    }
    static styles = ve_checkbox;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    #___private_checked_1 = (()=>{
        const _value = _init_checked(this, void 0);
        _init_extra__init_checked(this);
        return _value;
    })();
    get checked() {
        return this.#___private_checked_1;
    }
    set checked(_v) {
        this.#___private_checked_1 = _v;
    }
    #___private_defaultChecked_2 = (()=>{
        const _value = _init_defaultChecked(this, false);
        _init_extra__init_defaultChecked(this);
        return _value;
    })();
    get defaultChecked() {
        return this.#___private_defaultChecked_2;
    }
    set defaultChecked(_v) {
        this.#___private_defaultChecked_2 = _v;
    }
    #___private_indeterminate_3 = (()=>{
        const _value = _init_indeterminate(this, false);
        _init_extra__init_indeterminate(this);
        return _value;
    })();
    get indeterminate() {
        return this.#___private_indeterminate_3;
    }
    set indeterminate(_v) {
        this.#___private_indeterminate_3 = _v;
    }
    #___private_disabled_4 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_4;
    }
    set disabled(_v) {
        this.#___private_disabled_4 = _v;
    }
    #___private_value_5 = (()=>{
        const _value = _init_value(this, '');
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_5;
    }
    set value(_v) {
        this.#___private_value_5 = _v;
    }
    #___private_groupDisabled_6 = (()=>{
        const _value = _init_groupDisabled(this, false);
        _init_extra__init_groupDisabled(this);
        return _value;
    })();
    get groupDisabled() {
        return this.#___private_groupDisabled_6;
    }
    set groupDisabled(_v) {
        this.#___private_groupDisabled_6 = _v;
    }
    #___private_inputElement_7 = (()=>{
        const _value = _init_inputElement(this, null);
        _init_extra__init_inputElement(this);
        return _value;
    })();
    get inputElement() {
        return this.#___private_inputElement_7;
    }
    set inputElement(_v) {
        this.#___private_inputElement_7 = _v;
    }
    #___private_defaultSlot_8 = (()=>{
        const _value = _init_defaultSlot(this, null);
        _init_extra__init_defaultSlot(this);
        return _value;
    })();
    get defaultSlot() {
        return this.#___private_defaultSlot_8;
    }
    set defaultSlot(_v) {
        this.#___private_defaultSlot_8 = _v;
    }
    #___private_hasLabel_9 = (()=>{
        const _value = _init_hasLabel(this, false);
        _init_extra__init_hasLabel(this);
        return _value;
    })();
    get hasLabel() {
        return this.#___private_hasLabel_9;
    }
    set hasLabel(_v) {
        this.#___private_hasLabel_9 = _v;
    }
    hasInitializedDefaultChecked = false;
    get mergedDisabled() {
        return this.disabled || this.groupDisabled;
    }
    get inputValue() {
        return String(this.value);
    }
    get inputAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get inputAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get inputAriaDescribedby() {
        return this.getAttribute('aria-describedby') ?? void 0;
    }
    connectedCallback() {
        super.connectedCallback();
        this.applyDefaultChecked();
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.syncNativeIndeterminateState();
        this.syncLabelState();
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
        this.syncNativeIndeterminateState();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    applyDefaultChecked() {
        if (this.hasInitializedDefaultChecked) return;
        if (void 0 === this.checked && this.defaultChecked) this.checked = true;
        this.hasInitializedDefaultChecked = true;
    }
    syncNativeIndeterminateState() {
        if (this.inputElement) this.inputElement.indeterminate = this.indeterminate;
    }
    syncLabelState() {
        if (!this.defaultSlot) {
            this.hasLabel = false;
            return;
        }
        this.hasLabel = this.defaultSlot.assignedNodes({
            flatten: true
        }).some((node)=>hasRenderableSlotNode(node));
    }
    emitVeChangeEvent() {
        const detail = {
            checked: true === this.checked
        };
        const resolvedValue = this.value;
        if ('number' == typeof resolvedValue || 'string' == typeof resolvedValue && resolvedValue.length > 0) detail.value = resolvedValue;
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    handleInputClick(event) {
        event.stopPropagation();
    }
    handleInputEvent(event) {
        event.stopPropagation();
    }
    handleInputChange(event) {
        const target = event.target;
        if (!(target instanceof HTMLInputElement)) return;
        event.stopPropagation();
        if (this.mergedDisabled) {
            target.checked = true === this.checked;
            return;
        }
        this.checked = target.checked;
        if (this.indeterminate) this.indeterminate = false;
        this.emitVeChangeEvent();
    }
    handleDefaultSlotChange() {
        this.syncLabelState();
    }
    renderCheckedIcon() {
        return html`
      <svg
        class="ve-checkbox-mask-icon-default"
        aria-hidden="true"
        focusable="false"
        width="12"
        height="12"
        viewBox="0 0 12 12"
        fill="none"
      >
        <path
          d="M2 6l3 3 5-5"
          stroke="currentColor"
          stroke-width="1.5"
          stroke-linecap="round"
          stroke-linejoin="round"
        ></path>
      </svg>
    `;
    }
    render() {
        const wrapperClass = classMap({
            've-checkbox': true,
            've-checkbox--checked': true === this.checked,
            've-checkbox--indeterminate': this.indeterminate,
            've-checkbox--disabled': this.mergedDisabled
        });
        const iconHoverClass = classMap({
            've-checkbox-icon-hover': true
        });
        const textClass = classMap({
            've-checkbox-text': true,
            've-checkbox-text--hidden': !this.hasLabel
        });
        return html`
      <label class=${wrapperClass} part="wrapper">
        <input
          class="ve-checkbox-input"
          part="input"
          type="checkbox"
          .checked=${true === this.checked}
          value=${this.inputValue}
          ?disabled=${this.mergedDisabled}
          aria-label=${ifDefined(this.inputAriaLabel)}
          aria-labelledby=${ifDefined(this.inputAriaLabelledby)}
          aria-describedby=${ifDefined(this.inputAriaDescribedby)}
          @click=${this.handleInputClick}
          @input=${this.handleInputEvent}
          @change=${this.handleInputChange}
        />

        <span class=${iconHoverClass} aria-hidden="true">
          <span class="ve-checkbox-mask" part="mask">
            <span class="ve-checkbox-mask-icon">
              ${this.renderCheckedIcon()}
            </span>
            <span class="ve-checkbox-indeterminate-mark"></span>
          </span>
        </span>

        <span class=${textClass} part="text">
          <slot @slotchange=${this.handleDefaultSlotChange}></slot>
        </span>
      </label>
    `;
    }
}
_dec9 = property({
    attribute: false
}), _dec10 = property({
    attribute: 'default-value',
    converter: CHECKBOX_VALUE_ARRAY_ATTRIBUTE_CONVERTER
}), _dec11 = property({
    type: Boolean,
    reflect: true
}), _dec12 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec13 = property({
    converter: CHECKBOX_GROUP_OPTIONS_ATTRIBUTE_CONVERTER
});
class VeCheckboxGroupElement extends VeElement {
    static{
        ({ e: [_init_value1, _init_extra__init_value1, _init_defaultValue, _init_extra__init_defaultValue, _init_disabled1, _init_extra__init_disabled1, _init_direction, _init_extra__init_direction, _init_options, _init_extra__init_options] } = _apply_decs_2311(this, [], [
            [
                _dec9,
                1,
                "value"
            ],
            [
                _dec10,
                1,
                "defaultValue"
            ],
            [
                _dec11,
                1,
                "disabled"
            ],
            [
                _dec12,
                1,
                "direction"
            ],
            [
                _dec13,
                1,
                "options"
            ]
        ]));
    }
    static styles = ve_checkbox_group;
    static dependencies = {
        [VE_CHECKBOX_TAG]: VeCheckboxElement
    };
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    #___private_value_10 = (()=>{
        const _value = _init_value1(this, []);
        _init_extra__init_value1(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_10;
    }
    set value(_v) {
        this.#___private_value_10 = _v;
    }
    #___private_defaultValue_11 = (()=>{
        const _value = _init_defaultValue(this, []);
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_11;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_11 = _v;
    }
    #___private_disabled_12 = (()=>{
        const _value = _init_disabled1(this, false);
        _init_extra__init_disabled1(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_12;
    }
    set disabled(_v) {
        this.#___private_disabled_12 = _v;
    }
    #___private_direction_13 = (()=>{
        const _value = _init_direction(this, 'horizontal');
        _init_extra__init_direction(this);
        return _value;
    })();
    get direction() {
        return this.#___private_direction_13;
    }
    set direction(_v) {
        this.#___private_direction_13 = _v;
    }
    #___private_options_14 = (()=>{
        const _value = _init_options(this, []);
        _init_extra__init_options(this);
        return _value;
    })();
    get options() {
        return this.#___private_options_14;
    }
    set options(_v) {
        this.#___private_options_14 = _v;
    }
    normalizedOptions = [];
    hasInitializedDefaultValue = false;
    slottedCheckboxChildren = [];
    handleGroupChildChange = (event)=>{
        if (!(event instanceof CustomEvent)) return;
        if (event.target === this) return;
        const target = event.target;
        if (!(target instanceof VeCheckboxElement)) return;
        if (!this.isManagedCheckbox(target)) return;
        event.stopPropagation();
        const changedValue = target.value;
        const checked = Boolean(event.detail?.checked);
        const normalizedValues = this.normalizeValueArray(this.value);
        let nextValue = normalizedValues;
        if (checked) {
            if (!normalizedValues.includes(changedValue)) nextValue = [
                ...normalizedValues,
                changedValue
            ];
        } else nextValue = normalizedValues.filter((value)=>value !== changedValue);
        this.value = nextValue;
        this.syncCheckboxStates();
        const detail = {
            value: this.filterAvailableValues(nextValue),
            checked
        };
        detail.changedValue = changedValue;
        this.dispatchEvent(new CustomEvent('ve-group-change', {
            detail,
            bubbles: true,
            composed: true
        }));
    };
    get hasOptions() {
        return this.normalizedOptions.length > 0;
    }
    get groupAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get groupAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get groupAriaDescribedby() {
        return this.getAttribute('aria-describedby') ?? void 0;
    }
    connectedCallback() {
        super.connectedCallback();
        this.normalizedOptions = this.normalizeOptions(this.options);
        this.applyDefaultValue();
        this.addEventListener('ve-change', this.handleGroupChildChange);
    }
    disconnectedCallback() {
        this.removeEventListener('ve-change', this.handleGroupChildChange);
        super.disconnectedCallback();
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.refreshSlottedCheckboxChildren();
        this.syncCheckboxStates();
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('options')) this.normalizedOptions = this.normalizeOptions(this.options);
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('options') && !this.hasOptions) this.refreshSlottedCheckboxChildren();
        if (changedProperties.has('value') || changedProperties.has('disabled') || changedProperties.has('options')) this.syncCheckboxStates();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    applyDefaultValue() {
        if (this.hasInitializedDefaultValue) return;
        const normalizedCurrentValue = this.normalizeValueArray(this.value);
        const normalizedDefaultValue = this.normalizeValueArray(this.defaultValue);
        if (0 === normalizedCurrentValue.length && normalizedDefaultValue.length > 0) this.value = normalizedDefaultValue;
        else if (!this.isSameValueArray(this.value, normalizedCurrentValue)) this.value = normalizedCurrentValue;
        this.hasInitializedDefaultValue = true;
    }
    normalizeValueArray(values) {
        if (!Array.isArray(values)) return [];
        const result = [];
        const set = new Set();
        for (const value of values){
            if (isCheckboxValue(value)) {
                if ('number' != typeof value || Number.isFinite(value)) {
                    if (!set.has(value)) {
                        set.add(value);
                        result.push(value);
                    }
                }
            }
        }
        return result;
    }
    normalizeOptions(options) {
        if (!Array.isArray(options)) return [];
        const result = [];
        const seen = new Set();
        for (const option of options){
            const normalizedOption = this.normalizeOption(option);
            if (normalizedOption) {
                if (!seen.has(normalizedOption.value)) {
                    seen.add(normalizedOption.value);
                    result.push(normalizedOption);
                }
            }
        }
        return result;
    }
    normalizeOption(option) {
        if (isCheckboxValue(option)) return {
            label: option,
            value: option,
            disabled: false
        };
        if (!option || 'object' != typeof option) return null;
        const optionLike = option;
        if (!isCheckboxValue(optionLike.value)) return null;
        return {
            label: optionLike.label ?? optionLike.value,
            value: optionLike.value,
            disabled: true === optionLike.disabled
        };
    }
    collectSlottedCheckboxChildren() {
        const checkboxSet = new Set();
        for (const element of this.querySelectorAll(VE_CHECKBOX_TAG))if (element instanceof VeCheckboxElement && element.closest(VE_CHECKBOX_GROUP_TAG) === this) checkboxSet.add(element);
        return [
            ...checkboxSet
        ];
    }
    refreshSlottedCheckboxChildren() {
        this.slottedCheckboxChildren = this.collectSlottedCheckboxChildren();
    }
    getSlottedCheckboxChildren() {
        return this.slottedCheckboxChildren;
    }
    getCheckboxChildren() {
        if (this.hasOptions) return Array.from(this.renderRoot.querySelectorAll('ve-checkbox[data-option-item]'));
        return this.getSlottedCheckboxChildren();
    }
    filterAvailableValues(values) {
        if (this.hasOptions) {
            const availableValues = new Set(this.normalizedOptions.map((option)=>option.value));
            return values.filter((value)=>availableValues.has(value));
        }
        const availableValues = new Set(this.getCheckboxChildren().map((checkbox)=>checkbox.value));
        return values.filter((value)=>availableValues.has(value));
    }
    syncCheckboxStates() {
        const normalizedValues = this.normalizeValueArray(this.value);
        if (!this.isSameValueArray(this.value, normalizedValues)) {
            this.value = normalizedValues;
            return;
        }
        if (this.hasOptions) return;
        const selectedSet = new Set(normalizedValues);
        for (const checkbox of this.getCheckboxChildren()){
            checkbox.groupDisabled = this.disabled;
            checkbox.checked = selectedSet.has(checkbox.value);
            checkbox.indeterminate = false;
        }
    }
    isSameValueArray(source, target) {
        if (!Array.isArray(source)) return 0 === target.length;
        if (source.length !== target.length) return false;
        return source.every((item, index)=>item === target[index]);
    }
    isManagedCheckbox(checkbox) {
        if (this.hasOptions) return checkbox.hasAttribute('data-option-item');
        return checkbox.closest(VE_CHECKBOX_GROUP_TAG) === this;
    }
    handleSlotChange() {
        this.refreshSlottedCheckboxChildren();
        this.syncCheckboxStates();
    }
    renderOptionLabel(option) {
        if (isRenderableTemplateValue(option.label)) return option.label;
        return String(option.value);
    }
    renderOptionNodes() {
        const selectedSet = new Set(this.normalizeValueArray(this.value));
        return this.normalizedOptions.map((option)=>html`
        <ve-checkbox
          data-option-item
          class="ve-checkbox-group-option-item"
          .value=${option.value}
          .disabled=${option.disabled}
          .groupDisabled=${this.disabled}
          .checked=${selectedSet.has(option.value)}
          .indeterminate=${false}
          >${this.renderOptionLabel(option)}</ve-checkbox
        >
      `);
    }
    render() {
        const groupClass = classMap({
            've-checkbox-group': true,
            've-checkbox-group--vertical': 'vertical' === this.direction,
            've-checkbox-group-direction-horizontal': 'horizontal' === this.direction,
            've-checkbox-group-direction-vertical': 'vertical' === this.direction
        });
        return html`
      <div
        class=${groupClass}
        part="group"
        role="group"
        aria-label=${ifDefined(this.groupAriaLabel)}
        aria-labelledby=${ifDefined(this.groupAriaLabelledby)}
        aria-describedby=${ifDefined(this.groupAriaDescribedby)}
      >
        ${this.hasOptions ? this.renderOptionNodes() : html`<slot @slotchange=${this.handleSlotChange}></slot>`}
      </div>
    `;
    }
}
defineCustomElement(VE_CHECKBOX_TAG, VeCheckboxElement);
defineCustomElement(VE_CHECKBOX_GROUP_TAG, VeCheckboxGroupElement);
export { VE_CHECKBOX_GROUP_TAG, VE_CHECKBOX_TAG, VeCheckboxElement, VeCheckboxGroupElement };
