// @ve-design/web/web/dist/ve-cascader.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createFloatingAutoUpdate, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
import { closeIcon } from "./icons/close.js";
import { chevronUpSmIcon } from "./icons/chevron-up-sm.js";
import { chevronDownSmIcon } from "./67641.js";
import { chevronRightSmIcon } from "./icons/chevron-right-sm.js";
import { checkIcon } from "./76270.js";
const VALUE_SEPARATOR = '__ve_cascader__';
const SHOW_PARENT = 'parent';
const PANEL_MODE = {
    cascader: 'cascader',
    select: 'select'
};
function isEmptyValue(value) {
    return !value || Array.isArray(value) && 0 === value.length;
}
function getStoreConfig(props) {
    const config = {
        showParent: 'multiple' === props.mode && !props.changeOnSelect && props.checkedStrategy === SHOW_PARENT
    };
    if (void 0 !== props.showEmptyChildren) config.showEmptyChildren = props.showEmptyChildren;
    if (void 0 !== props.changeOnSelect) config.changeOnSelect = props.changeOnSelect;
    if (void 0 !== props.filterOption) config.filterOption = props.filterOption;
    return config;
}
function getStore(props, value) {
    const normalizedValue = value ? Array.isArray(value[0]) ? value : [
        value
    ] : [];
    return new CascaderStore(props.options ?? [], normalizedValue, getStoreConfig(props));
}
function transformValuesToSet(values) {
    const normalizedValues = values ?? [];
    return normalizedValues.reduce((set, next)=>{
        set.add(next.join(VALUE_SEPARATOR));
        return set;
    }, new Set());
}
function pathToKey(path) {
    return path.join(VALUE_SEPARATOR);
}
function valueInSet(set, value) {
    const normalizedValue = value ?? [];
    return set.has(normalizedValue.join(VALUE_SEPARATOR));
}
function normalizePathValue(value) {
    if (!Array.isArray(value)) return [];
    return value.filter((item)=>'string' == typeof item || 'number' == typeof item).map((item)=>String(item));
}
function normalizePathValues(value, multiple) {
    if (null == value) return [];
    if (multiple) {
        if (!Array.isArray(value)) return [];
        if (0 === value.length) return [];
        if (Array.isArray(value[0])) return value.map((path)=>normalizePathValue(path)).filter((path)=>path.length > 0);
        const singlePath = normalizePathValue(value);
        return singlePath.length > 0 ? [
            singlePath
        ] : [];
    }
    if (!Array.isArray(value)) return [];
    if (0 === value.length) return [];
    if (Array.isArray(value[0])) {
        const path = normalizePathValue(value[0]);
        return path.length > 0 ? [
            path
        ] : [];
    }
    const singlePath = normalizePathValue(value);
    return singlePath.length > 0 ? [
        singlePath
    ] : [];
}
function formatValue(value, multiple, store) {
    const normalizedValue = normalizePathValues(value, multiple);
    if (store && store.config.showParent) {
        const checkedNodes = store.getCheckedNodes();
        const valuesSet = transformValuesToSet(checkedNodes.map((node)=>node.pathValue));
        const result = [];
        const cache = {};
        normalizedValue.forEach((path)=>{
            path.some((_, index, arr)=>{
                const currentPath = arr.slice(0, index + 1);
                const key = currentPath.join(VALUE_SEPARATOR);
                const pass = valueInSet(valuesSet, currentPath);
                if (pass && !cache[key]) {
                    result.push(currentPath);
                    cache[key] = true;
                }
                return pass;
            });
        });
        return result;
    }
    return normalizedValue;
}
function denyToChecked(option) {
    const deny = (options)=>{
        if (!Array.isArray(options)) return false;
        return options.every((item)=>{
            if (item._checked || item.disabled) return true;
            return deny(item.children);
        });
    };
    return option._halfChecked && deny(option.children);
}
function getMultipleCheckValue(propsValue, store, option, checked) {
    const nextChecked = checked && denyToChecked(option) ? false : checked;
    const beforeValueSet = store.getCheckedNodes().reduce((set, node)=>{
        set.add(node.pathValue.join(VALUE_SEPARATOR));
        return set;
    }, new Set());
    option.setCheckedState(nextChecked);
    const checkedNodes = store.getCheckedNodes();
    const currentValue = checkedNodes.map((node)=>node.pathValue);
    const currentValueSet = transformValuesToSet(currentValue);
    const newValueSet = new Set();
    return propsValue.filter((value)=>{
        if (!valueInSet(beforeValueSet, value) || valueInSet(currentValueSet, value)) {
            newValueSet.add(value.join(VALUE_SEPARATOR));
            return true;
        }
        return false;
    }).concat(currentValue.filter((value)=>!valueInSet(newValueSet, value)));
}
class CascaderNode {
    value = '';
    label = '';
    disabled = false;
    _level = 0;
    _index = 0;
    isLeaf = true;
    disableCheckbox = false;
    _checked = false;
    _halfChecked = false;
    parent = null;
    pathValue = [];
    pathLabel = [];
    pathLabelText = '';
    children;
    config = {};
    _data;
    constructor(data, config, parent){
        this.config = config ?? {};
        this.initNode(data, parent ?? null);
    }
    initNode(option, parent) {
        const { showEmptyChildren } = this.config;
        const optionRecord = option;
        const childrenValue = optionRecord.children;
        const children = Array.isArray(childrenValue) ? childrenValue : void 0;
        const explicitIsLeaf = 'boolean' == typeof optionRecord.isLeaf ? optionRecord.isLeaf : null;
        const isLeaf = explicitIsLeaf ?? (Array.isArray(children) ? showEmptyChildren ? false : 0 === children.length : true);
        const nodeValue = String(optionRecord.value ?? '');
        const nodeLabelRaw = optionRecord.label;
        const nodeLabel = 'string' == typeof nodeLabelRaw && nodeLabelRaw.length > 0 ? nodeLabelRaw : nodeValue;
        const level = parent ? (parent._level ?? 0) + 1 : 0;
        const optionWithoutChildren = {
            ...optionRecord
        };
        delete optionWithoutChildren.children;
        const nextData = {
            ...optionWithoutChildren,
            value: nodeValue,
            label: nodeLabel,
            isLeaf,
            disabled: Boolean(parent && parent.disabled || optionRecord.disabled),
            disableCheckbox: true === optionRecord.disableCheckbox,
            parent: parent?._data ?? null,
            pathValue: parent ? [
                ...parent.pathValue,
                nodeValue
            ] : [
                nodeValue
            ],
            pathLabel: parent ? [
                ...parent.pathLabel,
                nodeLabel
            ] : [
                nodeLabel
            ],
            _level: level,
            _index: 'number' == typeof optionRecord._index && Number.isFinite(optionRecord._index) ? optionRecord._index : 0,
            _checked: false,
            _halfChecked: false
        };
        this._data = nextData;
        this.value = nextData.value;
        this.label = nextData.label;
        this.isLeaf = nextData.isLeaf;
        this.disabled = nextData.disabled;
        this.disableCheckbox = nextData.disableCheckbox;
        this.parent = parent;
        this.pathValue = nextData.pathValue;
        this.pathLabel = nextData.pathLabel;
        this.pathLabelText = this.pathLabel.join(' / ');
        this._level = nextData._level;
        this._index = nextData._index;
        this._checked = nextData._checked;
        this._halfChecked = nextData._halfChecked;
        if (!isLeaf && children && children.length > 0) {
            this.children = children.map((child, index)=>new CascaderNode({
                    ...child,
                    _index: index
                }, this.config, this));
            if (this._data) this._data.children = this.children.map((node)=>node._data);
        }
    }
    isHalfChecked() {
        if (!this.children || 0 === this.children.length) return false;
        const checkedLen = this.children.reduce((total, child)=>{
            const score = child._halfChecked ? 0.5 : child._checked ? 1 : 0;
            return total + score;
        }, 0);
        return checkedLen !== this.children.length && checkedLen > 0;
    }
    setNodeChildrenChecked(checked, ignoreDisabled) {
        if (!ignoreDisabled && this.disabled) return;
        if (this.children && this.children.length > 0) {
            this.children.forEach((child)=>{
                if (child.disabled) {
                    if (ignoreDisabled) child.setCheckedStateIgnoreDisabled(checked);
                    return;
                }
                child.setCheckedState(checked);
            });
            this.updateHalfState(checked);
        }
    }
    getSelfChildrenValue() {
        const result = [];
        const collect = (pathValue, nodes)=>{
            if (!nodes || 0 === nodes.length) return void result.push(pathValue);
            nodes.forEach((node)=>{
                collect(node.pathValue, node.children);
            });
        };
        collect(this.pathValue, this.children);
        return result;
    }
    updateHalfState(checked) {
        this._halfChecked = this.isHalfChecked();
        if (this._halfChecked) {
            this._checked = false;
            return;
        }
        if (this.children && this.children.length > 0) {
            this._checked = this.children.every((child)=>child._checked);
            return;
        }
        this._checked = checked;
    }
    setCheckedProperty(checked) {
        this._checked = checked;
        this._halfChecked = false;
    }
    setCheckedState(checked) {
        const noNeedToUpdate = checked ? this._checked : !this._checked && !this._halfChecked;
        if (this.disabled || noNeedToUpdate) return;
        this.setCheckedProperty(checked);
        if (!this.config.changeOnSelect) {
            this.setNodeChildrenChecked(checked);
            let parent = this.parent;
            while(parent && !parent.disabled){
                parent.updateHalfState(checked);
                parent = parent.parent;
            }
        }
    }
    setCheckedStateIgnoreDisabled(checked) {
        if (checked === Boolean(this._checked)) return;
        this.setCheckedProperty(checked);
        if (!this.config.changeOnSelect) {
            this.setNodeChildrenChecked(checked, true);
            let parent = this.parent;
            while(parent){
                parent.updateHalfState(checked);
                parent = parent.parent;
            }
        }
    }
    getPathNodes() {
        const nodes = [
            this
        ];
        let parent = this.parent;
        while(parent){
            nodes.unshift(parent);
            parent = parent.parent;
        }
        return nodes;
    }
    getChildren() {
        return this.children;
    }
}
class CascaderStore {
    nodes = [];
    flatNodes = [];
    nodeMap = new Map();
    config;
    constructor(options, value, config){
        this.config = {
            ...config ?? {}
        };
        const values = Array.isArray(value) ? value : [];
        this.nodes = this.calcNodes(options, null);
        this.updateFlatNodes();
        this.setNodeCheckedByValue(values);
    }
    calcNodes(options, parent) {
        if (!Array.isArray(options)) return [];
        return options.map((option, index)=>new CascaderNode({
                ...option,
                _index: index
            }, this.config, parent ?? void 0));
    }
    updateFlatNodes() {
        const leafOnly = !this.config.changeOnSelect;
        this.flatNodes = [];
        this.nodeMap.clear();
        const visit = (option)=>{
            this.nodeMap.set(pathToKey(option.pathValue), option);
            if (!leafOnly || option.isLeaf) this.flatNodes.push(option);
            if (Array.isArray(option.children)) option.children.forEach((child)=>{
                visit(child);
            });
        };
        this.nodes.forEach((node)=>{
            visit(node);
        });
    }
    setNodeCheckedByValue(initValues) {
        const valuesSet = transformValuesToSet(initValues);
        this.flatNodes.forEach((node)=>{
            let checked = false;
            if (this.config.showParent) {
                if (node.pathValue.some((_, index, arr)=>valueInSet(valuesSet, arr.slice(0, index + 1)))) checked = true;
            } else if (valueInSet(valuesSet, node.pathValue)) checked = true;
            node.setCheckedStateIgnoreDisabled(checked);
        });
    }
    findNodeByValue(value) {
        if (!Array.isArray(value) || 0 === value.length) return null;
        let list = this.nodes;
        let targetNode = null;
        for (const pathValue of value){
            const item = list.find((node)=>node.value === pathValue);
            if (!item) break;
            targetNode = item;
            list = item.children ?? [];
        }
        return targetNode;
    }
    searchNodes(inputValue) {
        if (!inputValue) return this.flatNodes;
        const keyword = inputValue.trim().toLowerCase();
        if (!keyword) return this.flatNodes;
        const filterMethod = this.config.filterOption ?? ((keyword, node)=>[
                node.label,
                node.value
            ].some((text)=>String(text ?? '').toLowerCase().includes(keyword)));
        return this.flatNodes.filter((node)=>{
            if (!this.config.filterOption) return node.getPathNodes().some((pathNode)=>filterMethod(keyword, pathNode._data));
            return node.getPathNodes().some((pathNode)=>filterMethod(inputValue, pathNode._data));
        });
    }
    getOptions() {
        return this.nodes;
    }
    getNodeByKey(key) {
        if (!key) return null;
        return this.nodeMap.get(key) ?? null;
    }
    getCheckedNodes() {
        if (this.config.showParent) return this.getCheckedParentNodes();
        return this.flatNodes.filter((node)=>node._checked);
    }
    getCheckedParentNodes() {
        const result = new Set();
        this.flatNodes.forEach((node)=>{
            if (!node._checked) return;
            node.getPathNodes().some((pathNode)=>{
                if (!pathNode._checked) return false;
                if (!result.has(pathNode)) result.add(pathNode);
                return true;
            });
        });
        return [
            ...result
        ];
    }
}
const veCascaderTokenDefaults = css`
  :host {
    --ve-cascader-font-family: var(--font-sans);
    --ve-cascader-font-weight: var(--font-weight-normal);
    --ve-cascader-line-height: var(--line-height-body);

    --ve-cascader-font-size-small: var(--text-body-sm);
    --ve-cascader-font-size-default: var(--text-body);
    --ve-cascader-font-size-large: var(--text-body);

    --ve-cascader-height-small: var(--height-small);
    --ve-cascader-height-default: var(--height-default);
    --ve-cascader-height-large: var(--height-large);

    --ve-cascader-padding-inline-small: var(--space-xxs);
    --ve-cascader-padding-inline-default: var(--space-xxs);
    --ve-cascader-padding-inline-large: var(--space-xxs);
    --ve-cascader-padding-block-small: var(--space-xxxs);
    --ve-cascader-padding-block-default: var(--space-xxxs);
    --ve-cascader-padding-block-large: var(--space-xxs);
    --ve-cascader-multiple-padding-inline: var(--space-xxxs);

    --ve-cascader-radius: var(--radius-md);
    --ve-cascader-border-width: var(--stroke-weight-base);
    --ve-cascader-gap: var(--space-xxxs);
    --ve-cascader-prefix-margin: 0;
    --ve-cascader-suffix-margin: 0;
    --ve-cascader-tag-gap: var(--space-xxxs);
    --ve-cascader-tag-padding-inline: calc(var(--space-xxs) - 2px);
    --ve-cascader-tag-height: calc(var(--height-default) - var(--space-xxs));
    --ve-cascader-tag-radius: var(--radius-sm);
    --ve-cascader-action-size: var(--space-s);
    --ve-cascader-action-icon-size: var(--text-caption);
    --ve-cascader-arrow-icon-size: var(--space-s);
    --ve-cascader-search-input-min-width: var(--height-default);

    --ve-cascader-control-background: var(--color-bg-control);
    --ve-cascader-control-background-hover: var(--color-bg-base);
    --ve-cascader-control-background-focus: var(--color-bg-base);
    --ve-cascader-control-background-error: var(--color-bg-danger-secondary);
    --ve-cascader-control-background-error-hover: color-mix(
      var(--ve-cascader-control-background-error)
        var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-cascader-control-background-warning: var(--color-bg-warning-secondary);
    --ve-cascader-control-background-warning-hover: color-mix(
      var(--ve-cascader-control-background-warning)
        var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-cascader-control-background-disabled: var(--color-bg-control);
    --ve-cascader-control-text: var(--color-text-primary);
    --ve-cascader-control-text-disabled: var(--color-text-disable);
    --ve-cascader-control-placeholder: var(--color-text-disable);
    --ve-cascader-control-placeholder-disabled: var(--color-text-disable);
    --ve-cascader-control-border-default: var(--color-border-default);
    --ve-cascader-control-border-default-hover: var(--color-border-interactive);
    --ve-cascader-control-border-default-focus: var(
      --color-border-interactive-active
    );
    --ve-cascader-control-border-disabled: var(--color-border-default);
    --ve-cascader-control-border-warning: var(--color-border-warning);
    --ve-cascader-control-border-warning-hover: var(--color-text-warning);
    --ve-cascader-control-border-error: var(--color-border-danger);
    --ve-cascader-control-border-error-hover: var(--color-text-danger);
    --ve-cascader-focus-ring-default: var(
      --ve-cascader-control-border-default-focus
    );
    --ve-cascader-focus-ring-warning: var(--color-text-warning);
    --ve-cascader-focus-ring-error: var(--color-text-danger);
    --ve-cascader-focus-shadow-strong-color: rgb(
      from var(--color-border-interactive) r g b / 25%
    );
    --ve-cascader-focus-shadow-soft-color: rgb(
      from var(--color-border-interactive) r g b / 10%
    );
    --ve-cascader-shadow-focus-default:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        var(--ve-cascader-focus-shadow-strong-color),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        var(--ve-cascader-focus-shadow-soft-color);
    --ve-cascader-shadow-focus-warning: var(--ve-cascader-shadow-focus-default);
    --ve-cascader-shadow-focus-error: var(--ve-cascader-shadow-focus-default);

    --ve-cascader-prefix-color: var(--ve-cascader-control-text);
    --ve-cascader-arrow-color: var(--color-icon-disable);
    --ve-cascader-clear-color: var(--color-text-tertiary);
    --ve-cascader-clear-hover-color: var(--color-text-secondary);
    --ve-cascader-clear-hover-background: var(--color-bg-muted);

    --ve-cascader-panel-background: var(--color-bg-overlay);
    --ve-cascader-panel-border-color: var(--color-border-default);
    --ve-cascader-panel-shadow: var(--shadow-md);
    --ve-cascader-panel-radius: var(--radius-md);
    --ve-cascader-panel-z-index: 1000;
    --ve-cascader-panel-gap: var(--space-xxxs);
    --ve-cascader-panel-padding: var(--space-xxs);
    --ve-cascader-panel-padding-block: var(--ve-cascader-panel-padding);
    --ve-cascader-panel-max-height: 244px;

    --ve-cascader-popup-surface-hover: var(--color-bg-surface);

    --ve-cascader-item-height: 38px;
    --ve-cascader-item-min-width: 160px;
    --ve-cascader-item-font-size: var(--text-body);
    --ve-cascader-item-text: var(--color-text-primary);
    --ve-cascader-item-text-hover: var(--color-text-primary);
    --ve-cascader-item-text-active: var(--color-text-primary);
    --ve-cascader-item-text-disabled: var(--color-text-disable);
    --ve-cascader-item-text-disabled-active: var(--color-text-disable);
    --ve-cascader-item-background-hover: var(--ve-cascader-popup-surface-hover);
    --ve-cascader-item-background-active: var(--color-bg-surface);
    --ve-cascader-item-background-disabled: transparent;
    --ve-cascader-item-background-disabled-active: color-mix(
      var(--ve-cascader-item-background-active) var(--state-disabled-ratio),
      transparent
    );
    --ve-cascader-item-icon-color: var(--color-text-secondary);
    --ve-cascader-item-icon-selected: var(--color-text-primary);
    --ve-cascader-item-icon-selected-disabled: var(--color-text-disable);
    --ve-cascader-item-font-weight-active: var(--font-weight-normal);
    --ve-cascader-item-radius: var(--radius-md);

    --ve-cascader-item-padding-left: var(--space-xxs);
    --ve-cascader-item-padding-right: var(--space-xxs);
    --ve-cascader-item-icon-margin-left: var(--space-xxs);

    --ve-cascader-checkbox-size: var(--ve-checkbox-mask-size, 16px);
    --ve-cascader-checkbox-surface-size: var(
      --ve-checkbox-mask-surface-size,
      var(--ve-cascader-checkbox-size)
    );
    --ve-cascader-checkbox-icon-size: var(
      --ve-checkbox-check-icon-size,
      12px
    );
    --ve-cascader-checkbox-indeterminate-width: var(
      --ve-checkbox-indeterminate-width,
      10px
    );
    --ve-cascader-checkbox-indeterminate-height: var(
      --ve-checkbox-indeterminate-height,
      2px
    );
    --ve-cascader-checkbox-margin-right: var(--space-xxs);
    --ve-cascader-checkbox-radius: var(
      --ve-checkbox-mask-border-radius,
      var(--radius-xs)
    );
    --ve-cascader-checkbox-border-width: var(
      --ve-checkbox-mask-border-width,
      var(--stroke-weight-strong)
    );
    --ve-cascader-checkbox-border: var(
      --ve-checkbox-mask-color-border,
      var(--color-border-strong)
    );
    --ve-cascader-checkbox-border-hover: var(
      --ve-checkbox-mask-color-border-hover,
      var(--color-border-interactive)
    );
    --ve-cascader-checkbox-border-checked: var(
      --ve-checkbox-mask-color-border-checked,
      transparent
    );
    --ve-cascader-checkbox-border-disabled: var(
      --ve-checkbox-mask-color-border-disabled,
      var(--color-border-default)
    );
    --ve-cascader-checkbox-background: var(
      --ve-checkbox-mask-color-bg,
      var(--color-bg-base)
    );
    --ve-cascader-checkbox-checked: var(
      --ve-checkbox-mask-color-bg-checked,
      var(--color-bg-primary)
    );
    --ve-cascader-checkbox-checked-hover: color-mix(
      var(--ve-cascader-checkbox-checked) var(--state-hover-solid-ratio),
      var(--state-hover-solid-mix)
    );
    --ve-cascader-checkbox-disabled: var(
      --ve-checkbox-mask-color-bg-disabled,
      var(--color-bg-muted)
    );
    --ve-cascader-checkbox-mark-color: var(
      --ve-checkbox-check-icon-color,
      var(--color-text-foreground)
    );
    --ve-cascader-checkbox-indeterminate-color: var(
      --ve-checkbox-indeterminate-color,
      var(--color-text-foreground)
    );
    --ve-cascader-checkbox-disabled-opacity: var(
      --ve-checkbox-disabled-opacity,
      var(--state-disabled-ratio)
    );

    --ve-cascader-tag-background: var(--color-bg-surface);
    --ve-cascader-tag-color: var(--color-text-secondary);
    --ve-cascader-tag-remove-color: var(--color-text-tertiary);
    --ve-cascader-tag-remove-hover-background: var(--color-bg-surface);

    --ve-cascader-empty-color: var(--color-text-tertiary);
    --ve-cascader-empty-font-size: var(--text-body);

    --ve-cascader-disabled-opacity: var(--state-disabled-ratio);
    --ve-cascader-transition-duration: var(--motion-fast);
    --ve-cascader-transition-duration-active: var(--motion-fast);
    --ve-cascader-slide-duration: 300ms;
    --ve-cascader-transition-timing: var(--ease-standard);
    --ve-cascader-slide-timing: cubic-bezier(0.34, 0.69, 0.1, 1);

    --ve-cascader-current-height: var(--ve-cascader-height-default);
    --ve-cascader-current-padding-inline: var(
      --ve-cascader-padding-inline-default
    );
    --ve-cascader-current-padding-block: var(
      --ve-cascader-padding-block-default
    );
    --ve-cascader-current-font-size: var(--ve-cascader-font-size-default);
    --ve-cascader-current-background: var(--ve-cascader-control-background);
    --ve-cascader-current-background-hover: var(
      --ve-cascader-control-background-hover
    );
    --ve-cascader-current-border-color: var(
      --ve-cascader-control-border-default
    );
    --ve-cascader-current-border-color-hover: var(
      --ve-cascader-control-border-default-hover
    );
    --ve-cascader-current-focus-ring-color: var(
      --ve-cascader-focus-ring-default
    );
    --ve-cascader-current-shadow-focus: var(--ve-cascader-shadow-focus-default);
  }

  :host([size='small']) {
    --ve-cascader-current-height: var(--ve-cascader-height-small);
    --ve-cascader-current-padding-inline: var(
      --ve-cascader-padding-inline-small
    );
    --ve-cascader-current-padding-block: var(--ve-cascader-padding-block-small);
    --ve-cascader-current-font-size: var(--ve-cascader-font-size-small);
  }

  :host([size='large']) {
    --ve-cascader-current-height: var(--ve-cascader-height-large);
    --ve-cascader-current-padding-inline: var(
      --ve-cascader-padding-inline-large
    );
    --ve-cascader-current-padding-block: var(--ve-cascader-padding-block-large);
    --ve-cascader-current-font-size: var(--ve-cascader-font-size-large);
  }

  :host([status='warning']) {
    --ve-cascader-current-background: var(
      --ve-cascader-control-background-warning
    );
    --ve-cascader-current-background-hover: var(
      --ve-cascader-control-background-warning-hover
    );
    --ve-cascader-current-border-color: var(
      --ve-cascader-control-border-warning
    );
    --ve-cascader-current-border-color-hover: var(
      --ve-cascader-control-border-warning-hover
    );
    --ve-cascader-current-focus-ring-color: var(
      --ve-cascader-focus-ring-warning
    );
    --ve-cascader-current-shadow-focus: var(--ve-cascader-shadow-focus-warning);
  }

  :host([status='error']) {
    --ve-cascader-current-background: var(
      --ve-cascader-control-background-error
    );
    --ve-cascader-current-background-hover: var(
      --ve-cascader-control-background-error-hover
    );
    --ve-cascader-current-border-color: var(--ve-cascader-control-border-error);
    --ve-cascader-current-border-color-hover: var(
      --ve-cascader-control-border-error-hover
    );
    --ve-cascader-current-focus-ring-color: var(--ve-cascader-focus-ring-error);
    --ve-cascader-current-shadow-focus: var(--ve-cascader-shadow-focus-error);
  }
`;
const token = veCascaderTokenDefaults;
const ve_cascader = [
    componentReset,
    token,
    css`
    :host {
      position: relative;
      display: inline-flex;
      inline-size: 100%;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-cascader {
      position: relative;
      inline-size: 100%;
    }

    .ve-cascader-control {
      appearance: none;
      display: flex;
      align-items: center;
      gap: var(--ve-cascader-gap);
      inline-size: 100%;
      min-inline-size: 0;
      min-block-size: var(--ve-cascader-current-height);
      box-sizing: border-box;
      padding-inline: var(--ve-cascader-current-padding-inline);
      padding-block: var(--ve-cascader-current-padding-block);
      border: var(--ve-cascader-border-width) solid
        var(--ve-cascader-current-border-color);
      border-radius: var(--ve-cascader-radius);
      background: var(--ve-cascader-current-background);
      color: var(--ve-cascader-control-text);
      font-family: var(--ve-cascader-font-family);
      font-size: var(--ve-cascader-current-font-size);
      font-weight: var(--ve-cascader-font-weight);
      line-height: var(--ve-cascader-line-height);
      cursor: pointer;
      transition:
        border-color var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing),
        box-shadow var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing),
        background-color var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing),
        opacity var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing);
      overflow: visible;
    }

    :host([mode='single']) .ve-cascader-control {
      block-size: var(--ve-cascader-current-height);
      min-block-size: 0;
      line-height: var(--ve-cascader-line-height);
    }

    :host([mode='multiple']) .ve-cascader-control {
      block-size: auto;
      min-block-size: var(--ve-cascader-current-height);
      padding-inline: var(--ve-cascader-multiple-padding-inline);
      line-height: 0;
      cursor: text;
    }

    :host(:not([disabled])) .ve-cascader-control:hover {
      border-color: var(--ve-cascader-current-border-color-hover);
      background: var(--ve-cascader-current-background-hover);
    }

    :host(:not([disabled])) .ve-cascader-control:focus,
    :host([open]) .ve-cascader-control {
      outline: none;
      color: var(--ve-cascader-control-text);
      background: var(--ve-cascader-control-background-focus);
      border-color: var(--ve-cascader-current-focus-ring-color);
      box-shadow: var(--ve-cascader-current-shadow-focus);
    }

    :host([disabled]) .ve-cascader-control {
      opacity: var(--ve-cascader-disabled-opacity);
      color: var(--ve-cascader-control-text-disabled);
      background: var(--ve-cascader-control-background-disabled);
      border-color: var(--ve-cascader-control-border-disabled);
      cursor: not-allowed;
    }

    :host([disabled]) .ve-cascader-search-input::placeholder {
      color: var(--ve-cascader-control-placeholder-disabled);
    }

    .ve-cascader-prefix,
    .ve-cascader-arrow {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
    }

    .ve-cascader-prefix {
      color: var(--ve-cascader-prefix-color);
      margin-inline-end: var(--ve-cascader-prefix-margin);
      white-space: nowrap;
    }

    :host([disabled]) .ve-cascader-prefix {
      color: var(--ve-cascader-control-text-disabled);
    }

    .ve-cascader-arrow {
      color: var(--ve-cascader-arrow-color);
    }

    .ve-cascader-value {
      display: flex;
      align-items: center;
      gap: var(--ve-cascader-tag-gap);
      min-inline-size: 0;
      flex: 1;
      flex-wrap: wrap;
    }

    .ve-cascader-value-text {
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      min-inline-size: 0;
    }

    .ve-cascader-value-text-placeholder {
      color: var(--ve-cascader-control-placeholder);
    }

    .ve-cascader-search-input {
      min-inline-size: var(--ve-cascader-search-input-min-width);
      flex: 1;
      inline-size: 100%;
      border: none;
      background: transparent;
      color: inherit;
      font: inherit;
      outline: none;
      padding: 0;
      pointer-events: auto;
      cursor: inherit;
    }

    .ve-cascader-search-input::placeholder {
      color: var(--ve-cascader-control-placeholder);
    }

    .ve-cascader-tag {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      flex: none;
      max-inline-size: 100%;
      min-block-size: var(--ve-cascader-tag-height);
      padding-inline: var(--ve-cascader-tag-padding-inline);
      border-radius: var(--ve-cascader-tag-radius);
      background: var(--ve-cascader-tag-background);
      color: var(--ve-cascader-tag-color);
      font-size: var(--text-caption);
      line-height: var(--line-height-body);
      box-sizing: border-box;
      overflow: hidden;
    }

    .ve-cascader-tag-label {
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-cascader-tag-remove {
      appearance: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: var(--ve-cascader-action-size);
      block-size: var(--ve-cascader-action-size);
      border: none;
      border-radius: var(--radius-full);
      background: transparent;
      color: var(--ve-cascader-tag-remove-color);
      cursor: pointer;
      padding: 0;
    }

    .ve-cascader-tag-remove:hover {
      background: var(--ve-cascader-tag-remove-hover-background);
    }

    .ve-cascader-actions {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      margin-inline-start: auto;
      flex: none;
      pointer-events: none;
    }

    .ve-cascader-actions > * {
      pointer-events: auto;
    }

    .ve-cascader-clear {
      appearance: none;
      display: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-cascader-action-size);
      block-size: var(--ve-cascader-action-size);
      border: none;
      border-radius: var(--radius-full);
      background: transparent;
      color: var(--ve-cascader-clear-color);
      cursor: pointer;
      padding: 0;
    }

    :host(:not([disabled])) .ve-cascader-control:hover .ve-cascader-clear,
    :host([open]) .ve-cascader-clear {
      display: inline-flex;
    }

    .ve-cascader-clear:hover {
      color: var(--ve-cascader-clear-hover-color);
      background: var(--ve-cascader-clear-hover-background);
    }

    .ve-cascader-popup {
      position: fixed;
      inset: auto;
      top: 0;
      left: 0;
      margin: 0;
      padding: 0;
      z-index: var(--ve-cascader-panel-z-index);
      box-sizing: border-box;
      border: var(--ve-cascader-border-width) solid
        var(--ve-cascader-panel-border-color);
      border-radius: var(--ve-cascader-panel-radius);
      background: var(--ve-cascader-panel-background);
      box-shadow: var(--ve-cascader-panel-shadow);
      overflow: hidden;
      will-change: transform;
    }

    .ve-cascader-popup-inner {
      inline-size: 100%;
      block-size: var(--ve-cascader-panel-max-height);
      white-space: nowrap;
      list-style: none;
    }

    .ve-cascader-popup-trigger-hover .ve-cascader-list-item {
      transition: font-weight 0s;
    }

    .ve-cascader-list-column {
      position: relative;
      vertical-align: top;
      display: inline-block;
      background: var(--ve-cascader-panel-background);
      block-size: 100%;
      min-inline-size: var(--ve-cascader-item-min-width);
    }

    .ve-cascader-list-column:not(:last-of-type) {
      border-inline-end: var(--ve-cascader-border-width) solid
        var(--ve-cascader-panel-border-color);
    }

    .ve-cascader-list-wrapper {
      position: relative;
      white-space: nowrap;
      box-sizing: border-box;
      block-size: 100%;
      display: flex;
      padding-block: var(--ve-cascader-panel-padding-block);
      padding-inline: var(--ve-cascader-panel-padding);
      flex-direction: column;
    }

    .ve-cascader-list-wrapper-with-footer {
      padding-block-end: 0;
    }

    .ve-cascader-list-empty {
      block-size: 100%;
      display: flex;
      align-items: center;
      justify-content: center;
      min-inline-size: 120px;
      color: var(--ve-cascader-empty-color);
      box-sizing: border-box;
      font-size: var(--ve-cascader-empty-font-size);
    }

    .ve-cascader-list {
      padding: 0;
      margin: 0;
      list-style: none;
      box-sizing: border-box;
      overflow: auto;
      flex: 1;
    }

    .ve-cascader-list-item,
    .ve-cascader-list-search-item {
      position: relative;
      block-size: var(--ve-cascader-item-height);
      line-height: var(--ve-cascader-line-height);
      min-inline-size: var(--ve-cascader-item-min-width);
      font-size: var(--ve-cascader-item-font-size);
      color: var(--ve-cascader-item-text);
      box-sizing: border-box;
      display: flex;
      cursor: pointer;
      background: transparent;
      align-items: center;
      border-radius: var(--ve-cascader-item-radius);
      transition:
        background-color var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing),
        color var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing),
        font-weight var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing);
    }

    .ve-cascader-list-item-label {
      flex-grow: 1;
      min-inline-size: 0;
      padding-inline-start: var(--ve-cascader-item-padding-left);
      padding-inline-end: var(--ve-cascader-item-padding-right);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-cascader-list-search.ve-cascader-list-multiple
      .ve-cascader-list-item-label {
      display: flex;
      align-items: center;
      padding-inline-end: var(--ve-cascader-item-padding-left);
    }

    .ve-cascader-list-item-arrow,
    .ve-cascader-list-item-check {
      color: var(--ve-cascader-item-icon-color);
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-cascader-arrow-icon-size);
      block-size: var(--ve-cascader-arrow-icon-size);
      margin-inline-start: var(--ve-cascader-item-icon-margin-left);
      margin-inline-end: var(--ve-cascader-item-padding-right);
    }

    .ve-cascader-list-item-arrow ve-icon,
    .ve-cascader-list-item-check ve-icon {
      --ve-icon-size: var(--ve-cascader-arrow-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-cascader-list-item-arrow-icon,
    .ve-cascader-list-item-check-icon {
      display: block;
      flex: none;
      inline-size: var(--ve-cascader-arrow-icon-size);
      block-size: var(--ve-cascader-arrow-icon-size);
    }

    .ve-cascader-arrow ve-icon {
      --ve-icon-size: var(--ve-cascader-arrow-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-cascader-clear ve-icon {
      --ve-icon-size: var(--ve-cascader-action-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-cascader-tag-remove-icon {
      display: block;
      flex: none;
      inline-size: var(--ve-cascader-action-icon-size);
      block-size: var(--ve-cascader-action-icon-size);
    }

    .ve-cascader-list-item-check {
      color: var(--ve-cascader-item-icon-selected);
    }

    .ve-cascader-list-item:hover,
    .ve-cascader-list-search-item-hover {
      color: var(--ve-cascader-item-text-hover);
      background: var(--ve-cascader-item-background-hover);
    }

    .ve-cascader-list-item-disabled,
    .ve-cascader-list-item-disabled:hover,
    .ve-cascader-list-item-selection-disabled,
    .ve-cascader-list-item-selection-disabled:hover,
    .ve-cascader-list-search-item-disabled,
    .ve-cascader-list-search-item-disabled:hover,
    .ve-cascader-list-search-item-selection-disabled,
    .ve-cascader-list-search-item-selection-disabled.ve-cascader-list-search-item-hover {
      cursor: not-allowed;
      background: var(--ve-cascader-item-background-disabled);
      color: var(--ve-cascader-item-text-disabled);
    }

    .ve-cascader-list-item-disabled *,
    .ve-cascader-list-item-selection-disabled *,
    .ve-cascader-list-search-item-disabled *,
    .ve-cascader-list-search-item-selection-disabled * {
      cursor: not-allowed;
    }

    .ve-cascader-list-item-disabled .ve-cascader-list-item-arrow,
    .ve-cascader-list-item-selection-disabled .ve-cascader-list-item-arrow {
      color: inherit;
    }

    .ve-cascader-list-item-disabled .ve-cascader-list-item-check,
    .ve-cascader-list-item-selection-disabled .ve-cascader-list-item-check,
    .ve-cascader-list-search-item-disabled .ve-cascader-list-item-check,
    .ve-cascader-list-search-item-selection-disabled
      .ve-cascader-list-item-check {
      color: var(--ve-cascader-item-icon-selected-disabled);
    }

    .ve-cascader-list-item-active,
    .ve-cascader-list-search-item-active {
      background: var(--ve-cascader-item-background-active);
      color: var(--ve-cascader-item-text-active);
      font-weight: var(--ve-cascader-item-font-weight-active);
    }

    .ve-cascader-list-item-active:hover,
    .ve-cascader-list-search-item-active.ve-cascader-list-search-item-hover {
      background: var(--ve-cascader-item-background-active);
      color: var(--ve-cascader-item-text-active);
    }

    .ve-cascader-list-item-active.ve-cascader-list-item-disabled,
    .ve-cascader-list-item-active.ve-cascader-list-item-disabled:hover,
    .ve-cascader-list-item-active.ve-cascader-list-item-selection-disabled,
    .ve-cascader-list-item-active.ve-cascader-list-item-selection-disabled:hover,
    .ve-cascader-list-search-item-active.ve-cascader-list-search-item-disabled,
    .ve-cascader-list-search-item-active.ve-cascader-list-search-item-disabled.ve-cascader-list-search-item-hover,
    .ve-cascader-list-search-item-active.ve-cascader-list-search-item-selection-disabled,
    .ve-cascader-list-search-item-active.ve-cascader-list-search-item-selection-disabled.ve-cascader-list-search-item-hover {
      background: var(--ve-cascader-item-background-disabled-active);
      color: var(--ve-cascader-item-text-disabled-active);
      font-weight: var(--ve-cascader-item-font-weight-active);
    }

    .ve-cascader-list-multiple .ve-cascader-list-item {
      padding-inline-start: var(--ve-cascader-item-padding-left);
    }

    .ve-cascader-list-multiple .ve-cascader-list-item-label {
      padding-inline-start: 0;
    }

    .ve-cascader-checkbox {
      position: relative;
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-cascader-checkbox-surface-size);
      block-size: var(--ve-cascader-checkbox-surface-size);
      margin-inline-end: var(--ve-cascader-checkbox-margin-right);
      box-sizing: border-box;
      color: var(--ve-cascader-checkbox-mark-color);
      line-height: 1;
      cursor: inherit;
    }

    .ve-cascader-checkbox-mask {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-cascader-checkbox-size);
      block-size: var(--ve-cascader-checkbox-size);
      box-sizing: border-box;
      border: var(--ve-cascader-checkbox-border-width) solid
        var(--ve-cascader-checkbox-border);
      border-radius: var(--ve-cascader-checkbox-radius);
      background: var(--ve-cascader-checkbox-background);
      transition:
        background-color var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing),
        border-color var(--ve-cascader-transition-duration)
          var(--ve-cascader-transition-timing);
    }

    .ve-cascader-checkbox-mask-icon {
      position: relative;
      display: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-cascader-checkbox-icon-size);
      block-size: var(--ve-cascader-checkbox-icon-size);
      color: var(--ve-cascader-checkbox-mark-color);
      line-height: 0;
    }

    .ve-cascader-checkbox-mask-icon-default {
      display: block;
      inline-size: 100%;
      block-size: 100%;
    }

    .ve-cascader-checkbox-indeterminate-mark {
      position: absolute;
      inset-block-start: 50%;
      inset-inline-start: 50%;
      display: block;
      inline-size: var(--ve-cascader-checkbox-indeterminate-width);
      block-size: var(--ve-cascader-checkbox-indeterminate-height);
      border-radius: var(--radius-full);
      background: var(--ve-cascader-checkbox-indeterminate-color);
      transform: translate(-50%, -50%);
      visibility: hidden;
    }

    .ve-cascader-list-item:hover
      .ve-cascader-checkbox:not(.ve-cascader-checkbox-checked):not(
        .ve-cascader-checkbox-indeterminate
      ):not(.ve-cascader-checkbox-disabled)
      .ve-cascader-checkbox-mask,
    .ve-cascader-list-search-item-hover
      .ve-cascader-checkbox:not(.ve-cascader-checkbox-checked):not(
        .ve-cascader-checkbox-indeterminate
      ):not(.ve-cascader-checkbox-disabled)
      .ve-cascader-checkbox-mask {
      border-color: var(--ve-cascader-checkbox-border-hover);
    }

    .ve-cascader-checkbox-checked .ve-cascader-checkbox-mask,
    .ve-cascader-checkbox-indeterminate .ve-cascader-checkbox-mask {
      border-color: var(--ve-cascader-checkbox-border-checked);
      background: var(--ve-cascader-checkbox-checked);
    }

    .ve-cascader-checkbox-checked .ve-cascader-checkbox-mask-icon {
      display: inline-flex;
    }

    .ve-cascader-checkbox-indeterminate .ve-cascader-checkbox-mask-icon {
      display: none;
    }

    .ve-cascader-checkbox-indeterminate
      .ve-cascader-checkbox-indeterminate-mark {
      visibility: visible;
    }

    .ve-cascader-list-item:hover
      .ve-cascader-checkbox-checked:not(.ve-cascader-checkbox-disabled)
      .ve-cascader-checkbox-mask,
    .ve-cascader-list-item:hover
      .ve-cascader-checkbox-indeterminate:not(.ve-cascader-checkbox-disabled)
      .ve-cascader-checkbox-mask,
    .ve-cascader-list-search-item-hover
      .ve-cascader-checkbox-checked:not(.ve-cascader-checkbox-disabled)
      .ve-cascader-checkbox-mask,
    .ve-cascader-list-search-item-hover
      .ve-cascader-checkbox-indeterminate:not(.ve-cascader-checkbox-disabled)
      .ve-cascader-checkbox-mask {
      background: var(--ve-cascader-checkbox-checked-hover);
    }

    .ve-cascader-checkbox-disabled {
      cursor: not-allowed;
      opacity: var(--ve-cascader-checkbox-disabled-opacity);
    }

    .ve-cascader-checkbox-disabled .ve-cascader-checkbox-mask {
      border-color: var(--ve-cascader-checkbox-border-disabled);
      background: var(--ve-cascader-checkbox-disabled);
    }

    .ve-cascader-checkbox-disabled.ve-cascader-checkbox-checked,
    .ve-cascader-checkbox-disabled.ve-cascader-checkbox-indeterminate {
      opacity: var(--ve-cascader-checkbox-disabled-opacity);
    }

    .ve-cascader-checkbox-disabled.ve-cascader-checkbox-checked
      .ve-cascader-checkbox-mask,
    .ve-cascader-checkbox-disabled.ve-cascader-checkbox-indeterminate
      .ve-cascader-checkbox-mask {
      border-color: var(--ve-cascader-checkbox-border-disabled);
      background: var(--ve-cascader-checkbox-disabled);
    }

    .ve-cascader-list-search-item-active .ve-cascader-list-item-check {
      color: var(--ve-cascader-item-icon-selected);
    }

    .ve-cascader-list-footer {
      block-size: var(--ve-cascader-item-height);
      line-height: var(--ve-cascader-item-height);
      padding-inline-start: var(--ve-cascader-item-padding-left);
      border-block-start: var(--ve-cascader-border-width) solid
        var(--ve-cascader-panel-border-color);
      box-sizing: border-box;
      flex: none;
    }

    .ve-cascader-highlight {
      font-weight: var(--font-weight-medium);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-cascader-control,
      .ve-cascader-arrow,
      .ve-cascader-checkbox-mask,
      .ve-cascader-checkbox-mask-icon,
      .ve-cascader-checkbox-indeterminate-mark {
        transition: none;
        animation: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _dec24, _dec25, _dec26, _dec27, _dec28, _dec29, _dec30, _dec31, _dec32, _init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_options, _init_extra__init_options, _init_placeholder, _init_extra__init_placeholder, _init_mode, _init_extra__init_mode, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_expandTrigger, _init_extra__init_expandTrigger, _init_changeOnSelect, _init_extra__init_changeOnSelect, _init_showEmptyChildren, _init_extra__init_showEmptyChildren, _init_checkedStrategy, _init_extra__init_checkedStrategy, _init_showSearch, _init_extra__init_showSearch, _init_inputValue, _init_extra__init_inputValue, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_emptyText, _init_extra__init_emptyText, _init_defaultActiveFirstOption, _init_extra__init_defaultActiveFirstOption, _init_showSearchOptions, _init_extra__init_showSearchOptions, _init_renderFormat, _init_extra__init_renderFormat, _init_renderOption, _init_extra__init_renderOption, _init_renderFooter, _init_extra__init_renderFooter, _init_filterOption, _init_extra__init_filterOption, _init_controlElement, _init_extra__init_controlElement, _init_panelElement, _init_extra__init_panelElement, _init_searchInputElement, _init_extra__init_searchInputElement, _init_normalizedValue, _init_extra__init_normalizedValue, _init_activePath, _init_extra__init_activePath, _init_searchOptions, _init_extra__init_searchOptions, _init_searchHoverIndex, _init_extra__init_searchHoverIndex, _init_panelPlacement, _init_extra__init_panelPlacement;
const VE_CASCADER_TAG = 've-cascader';
const OPTIONS_CONVERTER = {
    fromAttribute (value) {
        if (null === value || 0 === value.trim().length) return [];
        try {
            const parsed = JSON.parse(value);
            return Array.isArray(parsed) ? parsed.filter((item)=>'object' == typeof item && null !== item) : [];
        } catch  {
            return [];
        }
    },
    toAttribute (value) {
        try {
            return JSON.stringify(value);
        } catch  {
            return null;
        }
    }
};
const VALUE_CONVERTER = {
    fromAttribute (value) {
        if (null === value || 0 === value.trim().length) return [];
        try {
            const parsed = JSON.parse(value);
            if (!Array.isArray(parsed)) return [];
            if (Array.isArray(parsed[0])) return parsed.map((path)=>normalizePathValue(path)).filter((path)=>path.length > 0);
            return normalizePathValue(parsed);
        } catch  {
            return value.split(',').map((item)=>item.trim()).filter((item)=>item.length > 0);
        }
    },
    toAttribute (value) {
        if (!Array.isArray(value) || 0 === value.length) return null;
        try {
            return JSON.stringify(value);
        } catch  {
            return null;
        }
    }
};
function pathEquals(a, b) {
    if (!a || !b) return false;
    if (a.length !== b.length) return false;
    return a.every((value, index)=>value === b[index]);
}
function valuesEquals(a, b) {
    if (a.length !== b.length) return false;
    return a.every((path, index)=>pathEquals(path, b[index]));
}
function getLegalIndex(currentIndex, maxIndex) {
    if (maxIndex < 0) return -1;
    if (currentIndex < 0) return maxIndex;
    if (currentIndex > maxIndex) return 0;
    return currentIndex;
}
function getLegalActiveNode(options) {
    return options.find((option)=>!option.disabled) ?? null;
}
function getBaseActiveNode(currentNode) {
    if (currentNode && currentNode.disabled) {
        let node = currentNode;
        while(node.parent)if (node.parent.disabled) node = node.parent;
        else break;
        return node;
    }
    return currentNode;
}
function normalizeMode(mode) {
    return 'multiple' === mode ? 'multiple' : 'single';
}
_dec = property({
    converter: VALUE_CONVERTER
}), _dec1 = property({
    attribute: 'default-value',
    converter: VALUE_CONVERTER
}), _dec2 = property({
    converter: OPTIONS_CONVERTER
}), _dec3 = property({
    type: String
}), _dec4 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec5 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec6 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: Boolean,
    reflect: true
}), _dec8 = property({
    type: Boolean,
    attribute: 'allow-clear',
    reflect: true
}), _dec9 = property({
    type: Boolean,
    reflect: true
}), _dec10 = property({
    type: Boolean,
    attribute: 'default-open'
}), _dec11 = property({
    type: String,
    attribute: 'expand-trigger',
    reflect: true,
    useDefault: true
}), _dec12 = property({
    type: Boolean,
    attribute: 'change-on-select'
}), _dec13 = property({
    type: Boolean,
    attribute: 'show-empty-children'
}), _dec14 = property({
    type: String,
    attribute: 'checked-strategy',
    reflect: true,
    useDefault: true
}), _dec15 = property({
    type: Boolean,
    attribute: 'show-search',
    reflect: true
}), _dec16 = property({
    type: String,
    attribute: 'input-value'
}), _dec17 = property({
    type: Boolean,
    attribute: 'popup-match-select-width'
}), _dec18 = property({
    type: String,
    attribute: 'empty-text'
}), _dec19 = property({
    type: Boolean,
    attribute: 'default-active-first-option'
}), _dec20 = property({
    attribute: false
}), _dec21 = property({
    attribute: false
}), _dec22 = property({
    attribute: false
}), _dec23 = property({
    attribute: false
}), _dec24 = property({
    attribute: false
}), _dec25 = query('.ve-cascader-control'), _dec26 = query('.ve-cascader-popup'), _dec27 = query('.ve-cascader-search-input'), _dec28 = state(), _dec29 = state(), _dec30 = state(), _dec31 = state(), _dec32 = state();
class VeCascaderElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_options, _init_extra__init_options, _init_placeholder, _init_extra__init_placeholder, _init_mode, _init_extra__init_mode, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_expandTrigger, _init_extra__init_expandTrigger, _init_changeOnSelect, _init_extra__init_changeOnSelect, _init_showEmptyChildren, _init_extra__init_showEmptyChildren, _init_checkedStrategy, _init_extra__init_checkedStrategy, _init_showSearch, _init_extra__init_showSearch, _init_inputValue, _init_extra__init_inputValue, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_emptyText, _init_extra__init_emptyText, _init_defaultActiveFirstOption, _init_extra__init_defaultActiveFirstOption, _init_showSearchOptions, _init_extra__init_showSearchOptions, _init_renderFormat, _init_extra__init_renderFormat, _init_renderOption, _init_extra__init_renderOption, _init_renderFooter, _init_extra__init_renderFooter, _init_filterOption, _init_extra__init_filterOption, _init_controlElement, _init_extra__init_controlElement, _init_panelElement, _init_extra__init_panelElement, _init_searchInputElement, _init_extra__init_searchInputElement, _init_normalizedValue, _init_extra__init_normalizedValue, _init_activePath, _init_extra__init_activePath, _init_searchOptions, _init_extra__init_searchOptions, _init_searchHoverIndex, _init_extra__init_searchHoverIndex, _init_panelPlacement, _init_extra__init_panelPlacement] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "value"
            ],
            [
                _dec1,
                1,
                "defaultValue"
            ],
            [
                _dec2,
                1,
                "options"
            ],
            [
                _dec3,
                1,
                "placeholder"
            ],
            [
                _dec4,
                1,
                "mode"
            ],
            [
                _dec5,
                1,
                "size"
            ],
            [
                _dec6,
                1,
                "status"
            ],
            [
                _dec7,
                1,
                "disabled"
            ],
            [
                _dec8,
                1,
                "allowClear"
            ],
            [
                _dec9,
                1,
                "open"
            ],
            [
                _dec10,
                1,
                "defaultOpen"
            ],
            [
                _dec11,
                1,
                "expandTrigger"
            ],
            [
                _dec12,
                1,
                "changeOnSelect"
            ],
            [
                _dec13,
                1,
                "showEmptyChildren"
            ],
            [
                _dec14,
                1,
                "checkedStrategy"
            ],
            [
                _dec15,
                1,
                "showSearch"
            ],
            [
                _dec16,
                1,
                "inputValue"
            ],
            [
                _dec17,
                1,
                "popupMatchSelectWidth"
            ],
            [
                _dec18,
                1,
                "emptyText"
            ],
            [
                _dec19,
                1,
                "defaultActiveFirstOption"
            ],
            [
                _dec20,
                1,
                "showSearchOptions"
            ],
            [
                _dec21,
                1,
                "renderFormat"
            ],
            [
                _dec22,
                1,
                "renderOption"
            ],
            [
                _dec23,
                1,
                "renderFooter"
            ],
            [
                _dec24,
                1,
                "filterOption"
            ],
            [
                _dec25,
                1,
                "controlElement"
            ],
            [
                _dec26,
                1,
                "panelElement"
            ],
            [
                _dec27,
                1,
                "searchInputElement"
            ],
            [
                _dec28,
                1,
                "normalizedValue"
            ],
            [
                _dec29,
                1,
                "activePath"
            ],
            [
                _dec30,
                1,
                "searchOptions"
            ],
            [
                _dec31,
                1,
                "searchHoverIndex"
            ],
            [
                _dec32,
                1,
                "panelPlacement"
            ]
        ]));
    }
    static styles = ve_cascader;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_value_1 = (()=>{
        const _value = _init_value(this, []);
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_defaultValue_2 = (()=>{
        const _value = _init_defaultValue(this, []);
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_2;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_2 = _v;
    }
    #___private_options_3 = (()=>{
        const _value = _init_options(this, []);
        _init_extra__init_options(this);
        return _value;
    })();
    get options() {
        return this.#___private_options_3;
    }
    set options(_v) {
        this.#___private_options_3 = _v;
    }
    #___private_placeholder_4 = (()=>{
        const _value = _init_placeholder(this, 'Please select');
        _init_extra__init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_4;
    }
    set placeholder(_v) {
        this.#___private_placeholder_4 = _v;
    }
    #___private_mode_5 = (()=>{
        const _value = _init_mode(this, 'single');
        _init_extra__init_mode(this);
        return _value;
    })();
    get mode() {
        return this.#___private_mode_5;
    }
    set mode(_v) {
        this.#___private_mode_5 = _v;
    }
    #___private_size_6 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_6;
    }
    set size(_v) {
        this.#___private_size_6 = _v;
    }
    #___private_status_7 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_7;
    }
    set status(_v) {
        this.#___private_status_7 = _v;
    }
    #___private_disabled_8 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_8;
    }
    set disabled(_v) {
        this.#___private_disabled_8 = _v;
    }
    #___private_allowClear_9 = (()=>{
        const _value = _init_allowClear(this, false);
        _init_extra__init_allowClear(this);
        return _value;
    })();
    get allowClear() {
        return this.#___private_allowClear_9;
    }
    set allowClear(_v) {
        this.#___private_allowClear_9 = _v;
    }
    #___private_open_10 = (()=>{
        const _value = _init_open(this, false);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_10;
    }
    set open(_v) {
        this.#___private_open_10 = _v;
    }
    #___private_defaultOpen_11 = (()=>{
        const _value = _init_defaultOpen(this, false);
        _init_extra__init_defaultOpen(this);
        return _value;
    })();
    get defaultOpen() {
        return this.#___private_defaultOpen_11;
    }
    set defaultOpen(_v) {
        this.#___private_defaultOpen_11 = _v;
    }
    #___private_expandTrigger_12 = (()=>{
        const _value = _init_expandTrigger(this, 'click');
        _init_extra__init_expandTrigger(this);
        return _value;
    })();
    get expandTrigger() {
        return this.#___private_expandTrigger_12;
    }
    set expandTrigger(_v) {
        this.#___private_expandTrigger_12 = _v;
    }
    #___private_changeOnSelect_13 = (()=>{
        const _value = _init_changeOnSelect(this, false);
        _init_extra__init_changeOnSelect(this);
        return _value;
    })();
    get changeOnSelect() {
        return this.#___private_changeOnSelect_13;
    }
    set changeOnSelect(_v) {
        this.#___private_changeOnSelect_13 = _v;
    }
    #___private_showEmptyChildren_14 = (()=>{
        const _value = _init_showEmptyChildren(this, false);
        _init_extra__init_showEmptyChildren(this);
        return _value;
    })();
    get showEmptyChildren() {
        return this.#___private_showEmptyChildren_14;
    }
    set showEmptyChildren(_v) {
        this.#___private_showEmptyChildren_14 = _v;
    }
    #___private_checkedStrategy_15 = (()=>{
        const _value = _init_checkedStrategy(this, 'child');
        _init_extra__init_checkedStrategy(this);
        return _value;
    })();
    get checkedStrategy() {
        return this.#___private_checkedStrategy_15;
    }
    set checkedStrategy(_v) {
        this.#___private_checkedStrategy_15 = _v;
    }
    #___private_showSearch_16 = (()=>{
        const _value = _init_showSearch(this, false);
        _init_extra__init_showSearch(this);
        return _value;
    })();
    get showSearch() {
        return this.#___private_showSearch_16;
    }
    set showSearch(_v) {
        this.#___private_showSearch_16 = _v;
    }
    #___private_inputValue_17 = (()=>{
        const _value = _init_inputValue(this, '');
        _init_extra__init_inputValue(this);
        return _value;
    })();
    get inputValue() {
        return this.#___private_inputValue_17;
    }
    set inputValue(_v) {
        this.#___private_inputValue_17 = _v;
    }
    #___private_popupMatchSelectWidth_18 = (()=>{
        const _value = _init_popupMatchSelectWidth(this, false);
        _init_extra__init_popupMatchSelectWidth(this);
        return _value;
    })();
    get popupMatchSelectWidth() {
        return this.#___private_popupMatchSelectWidth_18;
    }
    set popupMatchSelectWidth(_v) {
        this.#___private_popupMatchSelectWidth_18 = _v;
    }
    #___private_emptyText_19 = (()=>{
        const _value = _init_emptyText(this, 'No data');
        _init_extra__init_emptyText(this);
        return _value;
    })();
    get emptyText() {
        return this.#___private_emptyText_19;
    }
    set emptyText(_v) {
        this.#___private_emptyText_19 = _v;
    }
    #___private_defaultActiveFirstOption_20 = (()=>{
        const _value = _init_defaultActiveFirstOption(this, true);
        _init_extra__init_defaultActiveFirstOption(this);
        return _value;
    })();
    get defaultActiveFirstOption() {
        return this.#___private_defaultActiveFirstOption_20;
    }
    set defaultActiveFirstOption(_v) {
        this.#___private_defaultActiveFirstOption_20 = _v;
    }
    #___private_showSearchOptions_21 = (()=>{
        const _value = _init_showSearchOptions(this, void 0);
        _init_extra__init_showSearchOptions(this);
        return _value;
    })();
    get showSearchOptions() {
        return this.#___private_showSearchOptions_21;
    }
    set showSearchOptions(_v) {
        this.#___private_showSearchOptions_21 = _v;
    }
    #___private_renderFormat_22 = (()=>{
        const _value = _init_renderFormat(this, void 0);
        _init_extra__init_renderFormat(this);
        return _value;
    })();
    get renderFormat() {
        return this.#___private_renderFormat_22;
    }
    set renderFormat(_v) {
        this.#___private_renderFormat_22 = _v;
    }
    #___private_renderOption_23 = (()=>{
        const _value = _init_renderOption(this, void 0);
        _init_extra__init_renderOption(this);
        return _value;
    })();
    get renderOption() {
        return this.#___private_renderOption_23;
    }
    set renderOption(_v) {
        this.#___private_renderOption_23 = _v;
    }
    #___private_renderFooter_24 = (()=>{
        const _value = _init_renderFooter(this, void 0);
        _init_extra__init_renderFooter(this);
        return _value;
    })();
    get renderFooter() {
        return this.#___private_renderFooter_24;
    }
    set renderFooter(_v) {
        this.#___private_renderFooter_24 = _v;
    }
    #___private_filterOption_25 = (()=>{
        const _value = _init_filterOption(this, void 0);
        _init_extra__init_filterOption(this);
        return _value;
    })();
    get filterOption() {
        return this.#___private_filterOption_25;
    }
    set filterOption(_v) {
        this.#___private_filterOption_25 = _v;
    }
    #___private_controlElement_26 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_26;
    }
    set controlElement(_v) {
        this.#___private_controlElement_26 = _v;
    }
    #___private_panelElement_27 = (()=>{
        const _value = _init_panelElement(this, null);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_27;
    }
    set panelElement(_v) {
        this.#___private_panelElement_27 = _v;
    }
    #___private_searchInputElement_28 = (()=>{
        const _value = _init_searchInputElement(this, null);
        _init_extra__init_searchInputElement(this);
        return _value;
    })();
    get searchInputElement() {
        return this.#___private_searchInputElement_28;
    }
    set searchInputElement(_v) {
        this.#___private_searchInputElement_28 = _v;
    }
    #___private_normalizedValue_29 = (()=>{
        const _value = _init_normalizedValue(this, []);
        _init_extra__init_normalizedValue(this);
        return _value;
    })();
    get normalizedValue() {
        return this.#___private_normalizedValue_29;
    }
    set normalizedValue(_v) {
        this.#___private_normalizedValue_29 = _v;
    }
    #___private_activePath_30 = (()=>{
        const _value = _init_activePath(this, []);
        _init_extra__init_activePath(this);
        return _value;
    })();
    get activePath() {
        return this.#___private_activePath_30;
    }
    set activePath(_v) {
        this.#___private_activePath_30 = _v;
    }
    #___private_searchOptions_31 = (()=>{
        const _value = _init_searchOptions(this, []);
        _init_extra__init_searchOptions(this);
        return _value;
    })();
    get searchOptions() {
        return this.#___private_searchOptions_31;
    }
    set searchOptions(_v) {
        this.#___private_searchOptions_31 = _v;
    }
    #___private_searchHoverIndex_32 = (()=>{
        const _value = _init_searchHoverIndex(this, -1);
        _init_extra__init_searchHoverIndex(this);
        return _value;
    })();
    get searchHoverIndex() {
        return this.#___private_searchHoverIndex_32;
    }
    set searchHoverIndex(_v) {
        this.#___private_searchHoverIndex_32 = _v;
    }
    #___private_panelPlacement_33 = (()=>{
        const _value = _init_panelPlacement(this, 'bottom-start');
        _init_extra__init_panelPlacement(this);
        return _value;
    })();
    get panelPlacement() {
        return this.#___private_panelPlacement_33;
    }
    set panelPlacement(_v) {
        this.#___private_panelPlacement_33 = _v;
    }
    store = new CascaderStore([]);
    cleanupAutoUpdate = null;
    repositionScheduler = createRafScheduler();
    positionRequestId = 0;
    hideInputTimer = null;
    initialized = false;
    isSyncingValue = false;
    openChangeReason = 'api';
    stashNodeMap = new Map();
    lastSelectedKeySetValues = null;
    lastSelectedKeySet = new Set();
    lastSearchStore = null;
    lastSearchInputValue = '';
    lastSearchResults = [];
    getSelectedKeySet() {
        if (this.lastSelectedKeySetValues === this.normalizedValue) return this.lastSelectedKeySet;
        const next = new Set(this.normalizedValue.map((pathValue)=>pathToKey(pathValue)));
        this.lastSelectedKeySetValues = this.normalizedValue;
        this.lastSelectedKeySet = next;
        return next;
    }
    getNodeByKey(key) {
        if (!key) return null;
        return this.store.getNodeByKey(key) ?? this.stashNodeMap.get(key) ?? null;
    }
    popupId = `ve-cascader-popup-${Math.random().toString(36).slice(2, 10)}`;
    get activeDescendantId() {
        if (!this.open) return;
        if (this.shouldShowSearchPanel) {
            const activeSearchOption = this.searchOptions[this.searchHoverIndex];
            return activeSearchOption ? this.getOptionId(pathToKey(activeSearchOption.pathValue)) : void 0;
        }
        const activeNode = this.getActiveNode();
        return activeNode ? this.getOptionId(pathToKey(activeNode.pathValue)) : void 0;
    }
    getOptionId(key) {
        return `${this.popupId}-option-${encodeURIComponent(key)}`;
    }
    getBooleanAttr(value) {
        return value ? 'true' : 'false';
    }
    getOptionCheckedAttr(option) {
        if (option._halfChecked) return 'mixed';
        return this.getBooleanAttr(option._checked);
    }
    isCheckboxDisabled(option) {
        return option.disabled || option.disableCheckbox;
    }
    isOptionSelectionDisabled(option) {
        return option.disabled || this.isMultiple && option.disableCheckbox && option.isLeaf;
    }
    isSearchOptionSelectionDisabled(option) {
        return option.disabled || this.isMultiple && option.disableCheckbox;
    }
    handleDocumentPointerDown = (event)=>{
        if (!this.open) return;
        if (event.composedPath().includes(this)) return;
        this.setOpen(false, 'outside');
    };
    handleDocumentKeydown = (event)=>{
        if (!this.open || 'Escape' !== event.key) return;
        event.stopPropagation();
        event.preventDefault();
        this.setOpen(false, 'escape');
        this.focus();
    };
    handleHostKeydown = (event)=>{
        if (!this.open || this.disabled) return;
        if (event.isComposing || 229 === event.keyCode) return;
        if ('Tab' === event.key) return void this.setOpen(false, 'trigger');
        const handled = this.shouldShowSearchPanel ? this.handleSearchPanelKeydown(event) : this.handleCascaderPanelKeydown(event);
        if (handled) {
            event.stopPropagation();
            event.preventDefault();
        }
    };
    get isMultiple() {
        return 'multiple' === this.mode;
    }
    get mergedShowSearch() {
        return this.showSearch || void 0 !== this.showSearchOptions;
    }
    get shouldShowSearchInput() {
        return this.mergedShowSearch && (this.open || this.isMultiple);
    }
    get shouldShowSearchPanel() {
        const panelMode = this.showSearchOptions?.panelMode;
        if (panelMode === PANEL_MODE.select) return true;
        if (panelMode === PANEL_MODE.cascader) return false;
        return this.mergedShowSearch && (this.inputValue ?? '').trim().length > 0;
    }
    get shouldShowClear() {
        if (!this.allowClear || this.disabled) return false;
        return !isEmptyValue(this.normalizedValue);
    }
    get mergedShowEmptyChildren() {
        return this.showEmptyChildren;
    }
    get selectedPathOptions() {
        return this.getSelectedOptionsByValue(this.normalizedValue);
    }
    focus(options) {
        if (this.open && this.searchInputElement && this.shouldShowSearchInput) return void this.searchInputElement.focus(options);
        this.controlElement?.focus(options);
    }
    blur() {
        this.searchInputElement?.blur();
        this.controlElement?.blur();
    }
    show() {
        this.setOpen(true, 'api');
    }
    hide() {
        this.setOpen(false, 'api');
    }
    addEventListener(type, listener, options) {
        if (null === listener) return;
        super.addEventListener(type, listener, options);
    }
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('keydown', this.handleHostKeydown);
    }
    disconnectedCallback() {
        this.removeEventListener('keydown', this.handleHostKeydown);
        this.removeGlobalListeners();
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        this.teardownAutoUpdate();
        this.clearHideInputTimer();
        super.disconnectedCallback();
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.refreshStoreAndValue();
        if (!this.hasAttribute('value')) {
            const fromDefault = formatValue(this.defaultValue, this.isMultiple, this.store);
            if (!isEmptyValue(fromDefault)) {
                this.isSyncingValue = true;
                this.value = this.isMultiple ? fromDefault : fromDefault[0] ?? [];
                this.isSyncingValue = false;
                this.normalizedValue = fromDefault;
                this.store.setNodeCheckedByValue(fromDefault);
            }
        }
        if (!this.hasAttribute('open') && this.defaultOpen) this.open = true;
        this.ensureSearchOptions();
        if (this.open) {
            this.installGlobalListeners();
            this.updateComplete.then(()=>{
                if (!this.isConnected || !this.open || this.disabled) return;
                showPanelPopover(this.panelElement);
                this.setupAutoUpdate();
                this.scheduleUpdatePosition();
                this.ensureActiveNode();
            });
        }
        this.initialized = true;
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('mode')) {
            this.mode = normalizeMode(this.mode);
            const normalized = formatValue(this.value, 'multiple' === this.mode);
            this.isSyncingValue = true;
            this.value = 'multiple' === this.mode ? normalized : normalized[0] ?? [];
            this.isSyncingValue = false;
        }
        if (changedProperties.has('expandTrigger') && 'hover' !== this.expandTrigger) this.expandTrigger = 'click';
        if (changedProperties.has('checkedStrategy') && 'parent' !== this.checkedStrategy && 'child' !== this.checkedStrategy) this.checkedStrategy = 'child';
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        const shouldRefreshStore = changedProperties.has('options') || changedProperties.has('changeOnSelect') || changedProperties.has('showEmptyChildren') || changedProperties.has('checkedStrategy') || changedProperties.has('filterOption') || changedProperties.has('mode');
        if (shouldRefreshStore) this.refreshStoreAndValue();
        else if (changedProperties.has('value') && !this.isSyncingValue) {
            const nextValue = formatValue(this.value, this.isMultiple, this.store);
            this.normalizedValue = nextValue;
            this.store.setNodeCheckedByValue(nextValue);
            this.syncStashNodes(nextValue);
            this.ensureSearchOptions();
        }
        if (changedProperties.has('inputValue')) this.ensureSearchOptions();
        if (changedProperties.has('open')) {
            if (this.open && !this.disabled) {
                this.clearHideInputTimer();
                this.installGlobalListeners();
                this.updateComplete.then(()=>{
                    if (!this.isConnected || !this.open || this.disabled) return;
                    showPanelPopover(this.panelElement);
                    this.setupAutoUpdate();
                    this.scheduleUpdatePosition();
                    this.ensureActiveNode();
                    if (this.shouldShowSearchInput) this.searchInputElement?.focus();
                });
            } else {
                hidePanelPopover(this.panelElement);
                this.removeGlobalListeners();
                this.repositionScheduler.cancel();
                this.positionRequestId += 1;
                this.teardownAutoUpdate();
                if (this.inputValue && !this.shouldRetainInputValueOnHide()) this.hideInputTimer = window.setTimeout(()=>{
                    this.hideInputTimer = null;
                    this.updateInputValue('', 'optionListHide');
                }, 200);
            }
            if (this.initialized) {
                this.emitOpenChange(this.open, this.openChangeReason);
                this.openChangeReason = 'api';
            }
        }
        if (changedProperties.has('disabled') && this.disabled && this.open) this.setOpen(false, 'api');
        if (this.open && this.panelElement && this.controlElement) this.scheduleUpdatePosition();
        const changedKeys = changedProperties;
        if (this.open && (changedProperties.has('open') || changedKeys.has('activePath') || changedKeys.has('searchHoverIndex') || changedProperties.has('inputValue'))) this.scrollActiveOptionIntoView();
    }
    clearHideInputTimer() {
        if (null !== this.hideInputTimer) {
            window.clearTimeout(this.hideInputTimer);
            this.hideInputTimer = null;
        }
    }
    refreshStoreAndValue() {
        const storeSource = {
            options: this.options,
            showEmptyChildren: this.mergedShowEmptyChildren,
            changeOnSelect: this.changeOnSelect,
            checkedStrategy: this.checkedStrategy,
            mode: this.mode
        };
        if (void 0 !== this.filterOption) storeSource.filterOption = this.filterOption;
        this.store = getStore(storeSource, formatValue(this.value, this.isMultiple));
        const nextValue = formatValue(this.value, this.isMultiple, this.store);
        this.normalizedValue = nextValue;
        this.store.setNodeCheckedByValue(nextValue);
        this.syncStashNodes(nextValue);
        this.ensureSearchOptions();
        this.ensureActiveNode();
    }
    ensureSearchOptions() {
        if (!this.mergedShowSearch) {
            this.searchOptions = [];
            this.searchHoverIndex = -1;
            this.lastSearchStore = null;
            this.lastSearchInputValue = '';
            this.lastSearchResults = [];
            return;
        }
        const inputValue = this.inputValue;
        if (this.lastSearchStore === this.store && this.lastSearchInputValue === inputValue) this.searchOptions = this.lastSearchResults;
        else {
            const nextResults = this.store.searchNodes(inputValue);
            this.lastSearchStore = this.store;
            this.lastSearchInputValue = inputValue;
            this.lastSearchResults = nextResults;
            this.searchOptions = nextResults;
        }
        if (0 === this.searchOptions.length) {
            this.searchHoverIndex = -1;
            return;
        }
        const current = this.searchHoverIndex;
        if (current < 0 || current > this.searchOptions.length - 1) this.searchHoverIndex = this.defaultActiveFirstOption ? 0 : -1;
    }
    ensureActiveNode() {
        const currentActive = this.getActiveNode();
        if (currentActive) return;
        const valueNode = this.store.findNodeByValue(this.normalizedValue[this.normalizedValue.length - 1]);
        if (valueNode) {
            this.activePath = [
                ...valueNode.pathValue
            ];
            return;
        }
        if (this.defaultActiveFirstOption) {
            const first = getLegalActiveNode(this.store.getOptions());
            if (first) this.activePath = [
                ...first.pathValue
            ];
        }
    }
    installGlobalListeners() {
        document.addEventListener('pointerdown', this.handleDocumentPointerDown);
        document.addEventListener('keydown', this.handleDocumentKeydown);
    }
    removeGlobalListeners() {
        document.removeEventListener('pointerdown', this.handleDocumentPointerDown);
        document.removeEventListener('keydown', this.handleDocumentKeydown);
    }
    setupAutoUpdate() {
        if (!this.isConnected || !this.open || !this.controlElement || !this.panelElement) return;
        this.teardownAutoUpdate();
        this.cleanupAutoUpdate = createFloatingAutoUpdate(this.controlElement, this.panelElement, ()=>{
            this.scheduleUpdatePosition();
        });
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    async updatePosition() {
        if (!this.open || !this.controlElement || !this.panelElement || !this.isConnected) return;
        const requestId = ++this.positionRequestId;
        const triggerRect = this.controlElement.getBoundingClientRect();
        if (this.popupMatchSelectWidth || this.shouldShowSearchPanel) this.panelElement.style.minWidth = `${triggerRect.width}px`;
        else this.panelElement.style.minWidth = '';
        const gapValue = Number.parseFloat(window.getComputedStyle(this).getPropertyValue('--ve-cascader-panel-gap')) || 4;
        const result = await computeFloatingPanelPosition({
            reference: this.controlElement,
            floating: this.panelElement,
            placement: 'bottom-start',
            distance: gapValue
        });
        if (requestId !== this.positionRequestId || !this.open || !this.isConnected || !this.panelElement) return;
        this.panelPlacement = result.placement;
        Object.assign(this.panelElement.style, {
            left: '0',
            top: '0',
            transform: `translate3d(${Math.round(result.x)}px, ${Math.round(result.y)}px, 0)`
        });
    }
    scrollActiveOptionIntoView() {
        const panel = this.panelElement;
        if (!panel) return;
        const activeItem = this.shouldShowSearchPanel ? panel.querySelector('.ve-cascader-list-search-item-hover, .ve-cascader-list-search-item-active') : panel.querySelector('.ve-cascader-list-item-active');
        activeItem?.scrollIntoView({
            block: 'nearest'
        });
    }
    syncStashNodes(values) {
        const nextMap = new Map();
        const addByPath = (pathValue)=>{
            const key = pathToKey(pathValue);
            const node = this.store.findNodeByValue(pathValue) ?? this.stashNodeMap.get(key) ?? null;
            if (node) nextMap.set(key, node);
        };
        values.forEach((pathValue)=>{
            addByPath(pathValue);
        });
        this.store.getCheckedNodes().forEach((node)=>{
            nextMap.set(pathToKey(node.pathValue), node);
        });
        this.stashNodeMap = nextMap;
    }
    getSelectedOptionsByValue(values) {
        return values.map((pathValue)=>{
            const node = this.store.findNodeByValue(pathValue) ?? this.stashNodeMap.get(pathToKey(pathValue)) ?? null;
            if (!node) return pathValue.map((value)=>({
                    value,
                    label: value
                }));
            return node.getPathNodes().map((pathNode)=>{
                const data = pathNode._data;
                if (data) return {
                    ...data
                };
                return {
                    value: pathNode.value,
                    label: pathNode.label,
                    disabled: pathNode.disabled,
                    isLeaf: pathNode.isLeaf
                };
            });
        });
    }
    toNodeData(option) {
        return option._data ?? {
            value: option.value,
            label: option.label,
            isLeaf: option.isLeaf,
            disabled: option.disabled,
            disableCheckbox: option.disableCheckbox,
            _checked: option._checked,
            _halfChecked: option._halfChecked,
            _level: option._level,
            _index: option._index,
            pathValue: [
                ...option.pathValue
            ],
            pathLabel: [
                ...option.pathLabel
            ],
            parent: null
        };
    }
    getSingleValueText(pathOptions) {
        if (0 === pathOptions.length) return '';
        const labels = pathOptions.map((option)=>String(option.label ?? option.value ?? ''));
        const fromRender = this.renderFormat?.(labels, pathOptions);
        if ('string' == typeof fromRender) return fromRender;
        return labels.join(' / ');
    }
    getActiveNode() {
        return this.store.findNodeByValue(this.activePath);
    }
    setOpen(nextOpen, reason) {
        if (nextOpen && this.disabled) return;
        if (nextOpen === this.open) return;
        this.openChangeReason = reason;
        this.open = nextOpen;
    }
    emitChange(nextValue) {
        const selectedOptions = this.getSelectedOptionsByValue(nextValue);
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail: {
                value: this.isMultiple ? nextValue : nextValue[0] ?? [],
                selectedOptions: this.isMultiple ? selectedOptions : selectedOptions[0] ?? [],
                dropdownVisible: this.open
            },
            bubbles: true,
            composed: true
        }));
    }
    emitOpenChange(open, reason) {
        this.dispatchEvent(new CustomEvent('ve-open-change', {
            detail: {
                open,
                reason
            },
            bubbles: true,
            composed: true
        }));
    }
    emitClear() {
        this.dispatchEvent(new CustomEvent('ve-clear', {
            detail: {
                visible: this.open
            },
            bubbles: true,
            composed: true
        }));
    }
    emitSearch(inputValue, reason) {
        const detail = {
            inputValue,
            reason
        };
        this.dispatchEvent(new CustomEvent('ve-input-value-change', {
            detail,
            bubbles: true,
            composed: true
        }));
        if ('manual' !== reason && 'optionListHide' !== reason) return;
        this.dispatchEvent(new CustomEvent('ve-search', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    shouldRetainInputValueOnHide() {
        return false;
    }
    shouldRetainInputValueWhileSelect() {
        return this.showSearchOptions?.retainInputValueWhileSelect === true;
    }
    updateInputValue(value, reason) {
        if (this.inputValue === value) return;
        this.inputValue = value;
        this.emitSearch(value, reason);
    }
    commitValue(nextValue, trigger) {
        const normalizedNextValue = formatValue(nextValue, this.isMultiple, this.store);
        if (valuesEquals(this.normalizedValue, normalizedNextValue)) {
            if (!this.isMultiple && 'clear' !== trigger) this.setOpen(false, 'select');
            return;
        }
        this.normalizedValue = normalizedNextValue;
        this.store.setNodeCheckedByValue(normalizedNextValue);
        this.syncStashNodes(normalizedNextValue);
        this.isSyncingValue = true;
        this.value = this.isMultiple ? normalizedNextValue : normalizedNextValue[0] ?? [];
        this.isSyncingValue = false;
        if (!this.isMultiple) {
            const latestPath = normalizedNextValue[0];
            const latestNode = latestPath ? this.store.findNodeByValue(latestPath) : null;
            if ('search' === trigger || (this.inputValue ?? '').trim().length > 0) this.setOpen(false, 'select');
            else if (latestNode && latestNode.isLeaf || this.changeOnSelect && 'hover' === this.expandTrigger) this.setOpen(false, 'select');
        }
        if (this.mergedShowSearch && !this.shouldRetainInputValueWhileSelect()) this.updateInputValue('', 'optionChecked');
        this.emitChange(normalizedNextValue);
    }
    handleControlClick() {
        if (this.disabled) return;
        this.setOpen(!this.open, 'trigger');
    }
    handleControlKeydown(event) {
        if (this.disabled) return;
        if (event.isComposing || 229 === event.keyCode) return;
        if ('Enter' === event.key || ' ' === event.key || 'ArrowDown' === event.key) {
            this.setOpen(true, 'trigger');
            event.preventDefault();
            return;
        }
        if ('Escape' === event.key && this.open) {
            this.setOpen(false, 'escape');
            event.preventDefault();
        }
    }
    handleSearchInput(event) {
        const target = event.target;
        if (!(target instanceof HTMLInputElement)) return;
        this.updateInputValue(target.value, 'manual');
        if (!this.open) this.setOpen(true, 'trigger');
    }
    handleSearchInputClick(event) {
        event.stopPropagation();
        if (!this.open && !this.disabled) this.setOpen(true, 'trigger');
    }
    handleSearchPanelKeydown(event) {
        const key = event.key;
        if ('Escape' === key) {
            this.setOpen(false, 'escape');
            return true;
        }
        if ('ArrowDown' !== key && 'ArrowUp' !== key && 'Enter' !== key) return false;
        const options = this.searchOptions;
        if (0 === options.length) return true;
        if ('ArrowDown' === key || 'ArrowUp' === key) {
            const diff = 'ArrowDown' === key ? 1 : -1;
            let nextIndex = getLegalIndex(this.searchHoverIndex + diff, options.length - 1);
            while(nextIndex !== this.searchHoverIndex){
                const candidate = options[nextIndex];
                if (!candidate || this.isSearchOptionSelectionDisabled(candidate)) nextIndex = getLegalIndex(nextIndex + diff, options.length - 1);
                else break;
            }
            this.searchHoverIndex = nextIndex;
            return true;
        }
        if ('Enter' === key) {
            const activeItem = options[this.searchHoverIndex];
            if (activeItem) {
                const checked = this.isMultiple ? activeItem._checked : this.getSelectedKeySet().has(pathToKey(activeItem.pathValue));
                this.handleSearchOptionSelect(activeItem, !checked);
            }
            return true;
        }
        return false;
    }
    handleCascaderPanelKeydown(event) {
        const key = event.key;
        const options = this.store.getOptions();
        if ('Escape' === key) {
            this.setOpen(false, 'escape');
            return true;
        }
        if ('ArrowDown' !== key && 'ArrowUp' !== key && 'ArrowLeft' !== key && 'ArrowRight' !== key && 'Enter' !== key) return false;
        let nextActiveNode;
        const activeNode = this.getActiveNode();
        if ('ArrowDown' === key || 'ArrowUp' === key) {
            if (activeNode) {
                const baseActiveNode = getBaseActiveNode(activeNode);
                const list = baseActiveNode?.parent?.children ?? options;
                if (baseActiveNode) {
                    const diff = 'ArrowDown' === key ? 1 : -1;
                    let nextIndex = getLegalIndex((baseActiveNode._index ?? 0) + diff, list.length - 1);
                    while(nextIndex !== (baseActiveNode._index ?? 0)){
                        nextActiveNode = list[nextIndex];
                        if (nextActiveNode && this.isOptionSelectionDisabled(nextActiveNode)) nextIndex = getLegalIndex(nextIndex + diff, list.length - 1);
                        else break;
                    }
                }
            } else nextActiveNode = getLegalActiveNode(options);
            if (nextActiveNode) this.handleOptionClick(nextActiveNode, false);
            return true;
        }
        if ('ArrowRight' === key) {
            if (activeNode && !activeNode.disabled) nextActiveNode = activeNode.children?.[0] ?? activeNode;
            if (nextActiveNode) this.handleOptionClick(nextActiveNode, false);
            return true;
        }
        if ('ArrowLeft' === key) {
            if (activeNode) {
                const base = getBaseActiveNode(activeNode);
                nextActiveNode = base?.parent ?? base;
            }
            if (nextActiveNode) this.handleOptionClick(nextActiveNode, false);
            return true;
        }
        if ('Enter' === key) {
            if (activeNode) if (this.isMultiple) this.handleMultipleChecked(activeNode, !activeNode._checked);
            else this.handleOptionClick(activeNode, true);
            return true;
        }
        return false;
    }
    getMenus() {
        const rootOptions = this.store.getOptions();
        const list = [
            rootOptions
        ];
        const activeNode = this.getActiveNode();
        const pathNodes = activeNode ? activeNode.getPathNodes() : [];
        pathNodes.forEach((node)=>{
            if (node.children) list.push(node.children);
        });
        return list;
    }
    handleOptionClick(option, triggerChange = true) {
        if (option.disabled) return;
        this.activePath = [
            ...option.pathValue
        ];
        if (!this.isMultiple && triggerChange && (this.changeOnSelect || option.isLeaf)) this.commitValue([
            option.pathValue
        ], 'panel');
    }
    handleMultipleChecked(option, checked) {
        if (this.isCheckboxDisabled(option)) return;
        const nextValue = getMultipleCheckValue(this.normalizedValue, this.store, option, checked);
        this.activePath = [
            ...option.pathValue
        ];
        this.commitValue(nextValue, 'panel');
    }
    handleSearchOptionSelect(option, checked) {
        if (this.isSearchOptionSelectionDisabled(option)) return;
        if (this.isMultiple) {
            const nextValue = getMultipleCheckValue(this.normalizedValue, this.store, option, checked);
            this.commitValue(nextValue, 'search');
            return;
        }
        this.commitValue([
            option.pathValue
        ], 'search');
    }
    handleClear(event) {
        event?.preventDefault();
        event?.stopPropagation();
        if (this.isMultiple) {
            const nodes = this.store.getCheckedNodes();
            const nextValue = nodes.filter((node)=>node.disabled).map((node)=>node.pathValue);
            this.store.setNodeCheckedByValue(nextValue);
            this.commitValue(nextValue, 'clear');
        } else this.commitValue([], 'clear');
        this.emitClear();
    }
    handleRemoveTag(index, event) {
        event.preventDefault();
        event.stopPropagation();
        if (!this.isMultiple) return;
        const valuePath = this.normalizedValue[index];
        const node = valuePath ? this.store.findNodeByValue(valuePath) : null;
        if (node?.disabled) return;
        const nextValue = this.normalizedValue.filter((_, pathIndex)=>pathIndex !== index);
        this.store.setNodeCheckedByValue(nextValue);
        this.commitValue(nextValue, 'panel');
    }
    handleTagRemoveClick = (event)=>{
        const target = event.currentTarget;
        if (!(target instanceof HTMLElement)) return;
        const index = Number(target.dataset.index);
        if (!Number.isFinite(index)) return;
        this.handleRemoveTag(index, event);
    };
    handleFooterMouseDown = (event)=>{
        event.stopPropagation();
    };
    handlePopupInnerMouseDown = (event)=>{
        event.preventDefault();
    };
    handleCascaderListMouseOver = (event)=>{
        if ('hover' !== this.expandTrigger) return;
        const target = event.target;
        if (!(target instanceof Element)) return;
        const item = target.closest('li[data-key]');
        if (!item) return;
        const option = this.getNodeByKey(item.dataset.key);
        if (!option || this.isOptionSelectionDisabled(option)) return;
        if (pathEquals(this.activePath, option.pathValue)) return;
        this.activePath = [
            ...option.pathValue
        ];
    };
    handleCascaderListClick = (event)=>{
        const target = event.target;
        if (!(target instanceof Element)) return;
        const item = target.closest('li[data-key]');
        if (!item) return;
        const option = this.getNodeByKey(item.dataset.key);
        if (!option || option.disabled) return;
        const clickedCheckbox = Boolean(target.closest('.ve-cascader-checkbox'));
        if (this.isMultiple && clickedCheckbox) {
            event.stopPropagation();
            if (this.isCheckboxDisabled(option)) return;
            this.handleMultipleChecked(option, !option._checked);
            return;
        }
        if (this.isOptionSelectionDisabled(option)) return;
        if (option.isLeaf && this.isMultiple && !option.disableCheckbox) return void this.handleMultipleChecked(option, !option._checked);
        this.handleOptionClick(option, true);
    };
    handleCascaderListDblClick = (event)=>{
        if (this.isMultiple || !this.changeOnSelect) return;
        const target = event.target;
        if (!(target instanceof Element)) return;
        const item = target.closest('li[data-key]');
        if (!item) return;
        const option = this.getNodeByKey(item.dataset.key);
        if (!option || option.disabled) return;
        const clickedLabel = Boolean(target.closest('.ve-cascader-list-item-label'));
        if (!clickedLabel) return;
        this.setOpen(false, 'select');
    };
    handleSearchListMouseOver = (event)=>{
        const target = event.target;
        if (!(target instanceof Element)) return;
        const item = target.closest('li[data-index]');
        if (!item) return;
        const option = this.getNodeByKey(item.dataset.key);
        if (!option || this.isSearchOptionSelectionDisabled(option)) return;
        const index = Number(item.dataset.index);
        if (!Number.isFinite(index)) return;
        this.searchHoverIndex = index;
    };
    handleSearchListMouseLeave = ()=>{
        if (0 === this.searchOptions.length) {
            this.searchHoverIndex = -1;
            return;
        }
        this.searchHoverIndex = this.defaultActiveFirstOption ? 0 : -1;
    };
    handleSearchListClick = (event)=>{
        const target = event.target;
        if (!(target instanceof Element)) return;
        const item = target.closest('li[data-key]');
        if (!item) return;
        const option = this.getNodeByKey(item.dataset.key);
        if (!option || this.isSearchOptionSelectionDisabled(option)) return;
        event.stopPropagation();
        const selectedKeySet = this.getSelectedKeySet();
        const checked = this.isMultiple ? option._checked : selectedKeySet.has(pathToKey(option.pathValue));
        this.handleSearchOptionSelect(option, !checked);
    };
    renderPrefix() {
        return html`<span class="ve-cascader-prefix" part="prefix"
      ><slot name="prefix"></slot
    ></span>`;
    }
    renderMultipleTags() {
        const selected = this.selectedPathOptions;
        if (0 === selected.length && !this.shouldShowSearchInput) return html`<span
        class="ve-cascader-value-text ve-cascader-value-text-placeholder"
        >${this.placeholder}</span
      >`;
        return html`
      ${selected.map((pathOptions, index)=>{
            const text = this.getSingleValueText(pathOptions);
            return html`
          <span class="ve-cascader-tag" part="tag">
            <span class="ve-cascader-tag-label">${text}</span>
            <button
              type="button"
              part="tag-remove"
              class="ve-cascader-tag-remove"
              aria-label=${`Remove ${text}`}
              data-index=${String(index)}
              @click=${this.handleTagRemoveClick}
            >
              ${this.renderTagRemoveIcon()}
            </button>
          </span>
        `;
        })}
      ${this.shouldShowSearchInput ? html`<input
            class="ve-cascader-search-input"
            .value=${this.inputValue}
            placeholder=${0 === selected.length ? this.placeholder : ''}
            @click=${this.handleSearchInputClick}
            @input=${this.handleSearchInput}
          />` : nothing}
    `;
    }
    renderSingleValue() {
        const selected = this.selectedPathOptions[0] ?? [];
        if (this.shouldShowSearchInput) {
            const renderedValue = selected.length > 0 ? this.getSingleValueText(selected) : '';
            const inputValue = this.inputValue || (this.showSearchOptions?.retainInputValue ? renderedValue : '');
            return html`<input
        class="ve-cascader-search-input"
        .value=${inputValue}
        placeholder=${renderedValue || this.placeholder}
        @click=${this.handleSearchInputClick}
        @input=${this.handleSearchInput}
      />`;
        }
        if (0 === selected.length) return html`<span
        class="ve-cascader-value-text ve-cascader-value-text-placeholder"
        >${this.placeholder}</span
      >`;
        return html`<span class="ve-cascader-value-text"
      >${this.getSingleValueText(selected)}</span
    >`;
    }
    renderValue() {
        return html`<span class="ve-cascader-value" part="value"
      >${this.isMultiple ? this.renderMultipleTags() : this.renderSingleValue()}</span
    >`;
    }
    renderActions() {
        return html`
      <span class="ve-cascader-actions">
        ${this.shouldShowClear ? html`<button
              class="ve-cascader-clear"
              part="clear-button"
              type="button"
              aria-label="Clear selection"
              @click=${this.handleClear}
            >
              <ve-icon .definition=${closeIcon}></ve-icon>
            </button>` : nothing}
        <span class="ve-cascader-arrow" part="arrow" aria-hidden="true">
          <ve-icon
            .definition=${this.open ? chevronUpSmIcon : chevronDownSmIcon}
          ></ve-icon>
        </span>
      </span>
    `;
    }
    renderTagRemoveIcon() {
        return html`
      <svg
        class="ve-cascader-tag-remove-icon"
        aria-hidden="true"
        focusable="false"
        width="12"
        height="12"
        viewBox="0 0 12 12"
        fill="none"
      >
        <path
          d="M3.25 3.25L8.75 8.75M8.75 3.25L3.25 8.75"
          stroke="currentColor"
          stroke-width="1.5"
          stroke-linecap="round"
        ></path>
      </svg>
    `;
    }
    renderOptionLabel(option, level) {
        const rendered = this.renderOption?.(this.toNodeData(option), level);
        if (null != rendered) return rendered;
        return option.label;
    }
    isOptionExpandable(option) {
        return !option.isLeaf || Boolean(option.children?.length);
    }
    renderOptionSuffix(option, selected) {
        if (this.isOptionExpandable(option)) return html`<span class="ve-cascader-list-item-arrow" aria-hidden="true"
        >${this.renderOptionArrowIcon()}</span
      >`;
        if (!this.isMultiple && selected) return html`<span class="ve-cascader-list-item-check" aria-hidden="true"
        >${this.renderOptionCheckIcon()}</span
      >`;
        return nothing;
    }
    renderOptionArrowIcon() {
        return html`<ve-icon
      class="ve-cascader-list-item-arrow-icon"
      .definition=${chevronRightSmIcon}
    ></ve-icon>`;
    }
    renderOptionCheckIcon() {
        return html`<ve-icon
      class="ve-cascader-list-item-check-icon"
      .definition=${checkIcon}
    ></ve-icon>`;
    }
    renderCheckboxIndicator(option, checked = option._checked, indeterminate = option._halfChecked) {
        const disabled = this.isCheckboxDisabled(option);
        return html`
      <span
        part="checkbox"
        role="checkbox"
        aria-checked=${indeterminate ? 'mixed' : this.getBooleanAttr(checked)}
        aria-disabled=${this.getBooleanAttr(disabled)}
        class=${classMap({
            've-cascader-checkbox': true,
            've-cascader-checkbox-checked': checked,
            've-cascader-checkbox-indeterminate': indeterminate,
            've-cascader-checkbox-disabled': disabled
        })}
      >
        <span class="ve-cascader-checkbox-mask" aria-hidden="true">
          <span class="ve-cascader-checkbox-mask-icon">
            <svg
              class="ve-cascader-checkbox-mask-icon-default"
              aria-hidden="true"
              focusable="false"
              width="12"
              height="12"
              viewBox="0 0 12 12"
              fill="none"
            >
              <path
                d="M2 6l3 3 5-5"
                stroke="currentColor"
                stroke-width="1.5"
                stroke-linecap="round"
                stroke-linejoin="round"
              ></path>
            </svg>
          </span>
          <span class="ve-cascader-checkbox-indeterminate-mark"></span>
        </span>
      </span>
    `;
    }
    renderColumnList(list, level, menus) {
        const footer = this.renderFooter?.(level, this.getActiveNode()?._data ?? null);
        const selectedKeySet = this.getSelectedKeySet();
        const activeNode = this.getActiveNode();
        const activeValue = activeNode?.pathValue[level];
        if (0 === list.length && !this.mergedShowEmptyChildren) return nothing;
        return html`
      <div
        part="column"
        class=${classMap({
            've-cascader-list-column': true
        })}
        style=${`z-index:${menus.length - level};`}
      >
        <div
          class=${classMap({
            've-cascader-list-wrapper': true,
            've-cascader-list-wrapper-with-footer': null != footer
        })}
        >
          ${0 === list.length ? this.renderEmpty('100%') : html`<ul
                role="tree"
                class=${classMap({
            've-cascader-list': true,
            've-cascader-list-select': true,
            've-cascader-list-multiple': this.isMultiple
        })}
                @mouseover=${this.handleCascaderListMouseOver}
                @click=${this.handleCascaderListClick}
                @dblclick=${this.handleCascaderListDblClick}
              >
                ${list.map((option)=>{
            const key = pathToKey(option.pathValue);
            const isSelected = selectedKeySet.has(key);
            const isExpandable = this.isOptionExpandable(option);
            const isSelectionDisabled = this.isOptionSelectionDisabled(option);
            const isActive = Boolean(!isSelectionDisabled && void 0 !== activeValue && activeValue === option.value);
            return html`
                    <li
                      id=${this.getOptionId(key)}
                      tabindex="-1"
                      role="treeitem"
                      part="option"
                      data-key=${key}
                      aria-selected=${this.getBooleanAttr(isSelected)}
                      aria-checked=${ifDefined(this.isMultiple ? this.getOptionCheckedAttr(option) : void 0)}
                      aria-haspopup=${isExpandable ? 'tree' : 'false'}
                      aria-expanded=${ifDefined(isExpandable ? this.getBooleanAttr(isActive) : void 0)}
                      aria-disabled=${this.getBooleanAttr(isSelectionDisabled)}
                      class=${classMap({
                've-cascader-list-item': true,
                've-cascader-list-item-active': isActive,
                've-cascader-list-item-disabled': Boolean(option.disabled),
                've-cascader-list-item-selection-disabled': isSelectionDisabled && !option.disabled
            })}
                    >
                      ${this.isMultiple ? this.renderCheckboxIndicator(option) : nothing}
                      <div class="ve-cascader-list-item-label">
                        ${this.renderOptionLabel(option, level)}
                      </div>
                      ${this.renderOptionSuffix(option, isSelected)}
                    </li>
                  `;
        })}
              </ul>`}
          ${null != footer ? html`<div
                part="footer"
                class="ve-cascader-list-footer"
                @mousedown=${this.handleFooterMouseDown}
              >
                ${footer}
              </div>` : nothing}
        </div>
      </div>
    `;
    }
    renderEmpty(width) {
        return html`<div
      part="empty"
      class="ve-cascader-list-empty"
      style=${width ? `width:${width};` : ''}
    >
      ${this.emptyText}
    </div>`;
    }
    formatSearchLabel(option) {
        const pathNodes = option.getPathNodes();
        const inputValue = (this.inputValue ?? '').trim();
        const upperInputValue = inputValue.toUpperCase();
        return html`${pathNodes.map((pathNode, index)=>{
            const label = pathNode.label;
            const indexOfKeyword = label.toUpperCase().indexOf(upperInputValue);
            if (inputValue.length > 0 && indexOfKeyword > -1) {
                const prefix = label.slice(0, indexOfKeyword);
                const match = label.slice(indexOfKeyword, indexOfKeyword + inputValue.length);
                const suffix = label.slice(indexOfKeyword + inputValue.length);
                return html`<span>
          ${prefix}<span class="ve-cascader-highlight">${match}</span
          >${suffix}${index === pathNodes.length - 1 ? '' : ' / '}
        </span>`;
            }
            return html`<span
        >${label}${index === pathNodes.length - 1 ? '' : ' / '}</span
      >`;
        })}`;
    }
    renderSearchOption(option, index) {
        const selectedKeySet = this.getSelectedKeySet();
        const key = pathToKey(option.pathValue);
        const checked = this.isMultiple ? option._checked : selectedKeySet.has(key);
        const isSelectionDisabled = this.isSearchOptionSelectionDisabled(option);
        const renderOption = this.showSearchOptions?.renderOption?.(this.inputValue, this.toNodeData(option), {
            checked
        }) ?? null;
        const label = renderOption ?? this.formatSearchLabel(option);
        return html`
      <li
        id=${this.getOptionId(key)}
        role="option"
        part="search-option"
        data-key=${key}
        data-index=${String(index)}
        aria-selected=${this.getBooleanAttr(checked)}
        aria-checked=${ifDefined(this.isMultiple ? this.getOptionCheckedAttr(option) : void 0)}
        aria-disabled=${this.getBooleanAttr(isSelectionDisabled)}
        class=${classMap({
            've-cascader-list-search-item': true,
            've-cascader-list-search-item-active': checked,
            've-cascader-list-search-item-hover': !isSelectionDisabled && index === this.searchHoverIndex,
            've-cascader-list-search-item-disabled': Boolean(option.disabled),
            've-cascader-list-search-item-selection-disabled': isSelectionDisabled && !option.disabled
        })}
      >
        <div class="ve-cascader-list-item-label">
          ${renderOption ? label : this.isMultiple ? html`
                  ${this.renderCheckboxIndicator(option, checked, option._halfChecked)}
                  ${label}
                ` : label}
        </div>
        ${!this.isMultiple && checked ? html`<span class="ve-cascader-list-item-check" aria-hidden="true"
              >${this.renderOptionCheckIcon()}</span
            >` : nothing}
      </li>
    `;
    }
    renderSearchPanel() {
        if (0 === this.searchOptions.length) return this.renderEmpty();
        return html`
      <div class="ve-cascader-list-wrapper">
        <ul
          role="listbox"
          class=${classMap({
            've-cascader-list': true,
            've-cascader-list-search': true,
            've-cascader-list-multiple': this.isMultiple
        })}
          @mouseover=${this.handleSearchListMouseOver}
          @mouseleave=${this.handleSearchListMouseLeave}
          @click=${this.handleSearchListClick}
        >
          ${this.searchOptions.map((option, index)=>this.renderSearchOption(option, index))}
        </ul>
      </div>
    `;
    }
    renderCascaderPanel() {
        const menus = this.getMenus();
        if (0 === menus.length || (menus[0]?.length ?? 0) === 0) return this.renderEmpty();
        return html`${menus.map((list, level)=>this.renderColumnList(list, level, menus))}`;
    }
    renderPopupContent() {
        return this.shouldShowSearchPanel ? this.renderSearchPanel() : this.renderCascaderPanel();
    }
    renderPopup() {
        if (!this.open) return nothing;
        return html`
      <div
        id=${this.popupId}
        part="popup"
        popover=${"manual"}
        class=${classMap({
            've-cascader-popup': true,
            've-cascader-popup-trigger-hover': 'hover' === this.expandTrigger
        })}
      >
        <div
          part="panel"
          class="ve-cascader-popup-inner"
          @mousedown=${this.handlePopupInnerMouseDown}
        >
          ${this.renderPopupContent()}
        </div>
      </div>
    `;
    }
    render() {
        return html`
      <div class="ve-cascader" part="root">
        <div
          part="control"
          class="ve-cascader-control"
          tabindex=${this.disabled ? '-1' : '0'}
          role="combobox"
          aria-disabled=${this.getBooleanAttr(this.disabled)}
          aria-haspopup="listbox"
          aria-controls=${this.popupId}
          aria-expanded=${this.getBooleanAttr(this.open)}
          aria-activedescendant=${ifDefined(this.activeDescendantId)}
          @click=${this.handleControlClick}
          @keydown=${this.handleControlKeydown}
        >
          ${this.renderPrefix()} ${this.renderValue()} ${this.renderActions()}
        </div>
        ${this.renderPopup()}
      </div>
    `;
    }
}
defineCustomElement(VE_CASCADER_TAG, VeCascaderElement);
export { VE_CASCADER_TAG, VeCascaderElement };
