// @ve-design/web/web/dist/ve-chat-input.js
import { css, html, nothing, render as external_lit_render } from "lit";
import { property, query as decorators_js_query, state } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { repeat } from "lit/directives/repeat.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VE_SELECT_TAG, VE_SELECT_ITEM_TAG, VeSelectItemElement, VeSelectElement } from "./30129.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { registerIcons } from "./icons/registry.js";
import { chevronDownMdIcon } from "./icons/chevron-down-md.js";
import { attachmentIcon } from "./icons/attachment.js";
import { imageIcon } from "./icons/image.js";
import { fileIcon } from "./icons/file.js";
import { warningIcon } from "./icons/warning.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
import { createFloatingAutoUpdate, computeFloatingPanelPosition } from "./42896.js";
import { exitFullScreenIcon } from "./icons/exit-full-screen.js";
import { fullScreenIcon } from "./icons/full-screen.js";
import { closeIcon } from "./icons/close.js";
import { arrowUpMdIcon } from "./icons/arrow-up-md.js";
const veChatInputToken = css`
  :host {
    /* ── 字体 ── */
    --ve-chat-input-font-family: var(--font-sans);
    --ve-chat-input-font-size-body: var(--text-body, 14px);

    /* ── 容器 ── */
    --ve-chat-input-bg: var(--color-bg-base, #fff);
    --ve-chat-input-radius: var(--radius-2xl, 20px);
    --ve-chat-input-padding: 12px;
    --ve-chat-input-gap: 10px;
    --ve-chat-input-shadow: var(--shadow-md);

    /* ── 边框 ── */
    --ve-chat-input-border-width: var(--stroke-0_5, 0.5px);
    --ve-chat-input-border-color: var(--color-border-default, #e5e6eb);
    --ve-chat-input-border-color-active: var(--color-border-strong, #1d2129);

    /* ── Border glow ── */
    --ve-chat-input-border-glow-edge-proximity: 0;
    --ve-chat-input-border-glow-cursor-angle: 45deg;
    --ve-chat-input-border-glow-edge-sensitivity: 30;
    --ve-chat-input-border-glow-color-sensitivity: calc(
      var(--ve-chat-input-border-glow-edge-sensitivity) + 20
    );
    --ve-chat-input-border-glow-padding: 40px;
    --ve-chat-input-border-glow-cone-spread: 25;
    --ve-chat-input-border-glow-fill-opacity: 0.35;
    --ve-chat-input-border-glow-color: hsl(40deg 80% 80% / 100%);
    --ve-chat-input-border-glow-color-60: hsl(40deg 80% 80% / 60%);
    --ve-chat-input-border-glow-color-50: hsl(40deg 80% 80% / 50%);
    --ve-chat-input-border-glow-color-40: hsl(40deg 80% 80% / 40%);
    --ve-chat-input-border-glow-color-30: hsl(40deg 80% 80% / 30%);
    --ve-chat-input-border-glow-color-20: hsl(40deg 80% 80% / 20%);
    --ve-chat-input-border-glow-color-10: hsl(40deg 80% 80% / 10%);
    --ve-chat-input-border-glow-gradient-1: radial-gradient(
      at 80% 55%,
      #c084fc 0px,
      transparent 50%
    );
    --ve-chat-input-border-glow-gradient-2: radial-gradient(
      at 69% 34%,
      #f472b6 0px,
      transparent 50%
    );
    --ve-chat-input-border-glow-gradient-3: radial-gradient(
      at 8% 6%,
      #38bdf8 0px,
      transparent 50%
    );
    --ve-chat-input-border-glow-gradient-4: radial-gradient(
      at 41% 38%,
      #c084fc 0px,
      transparent 50%
    );
    --ve-chat-input-border-glow-gradient-5: radial-gradient(
      at 86% 85%,
      #f472b6 0px,
      transparent 50%
    );
    --ve-chat-input-border-glow-gradient-6: radial-gradient(
      at 82% 18%,
      #38bdf8 0px,
      transparent 50%
    );
    --ve-chat-input-border-glow-gradient-7: radial-gradient(
      at 51% 4%,
      #f472b6 0px,
      transparent 50%
    );
    --ve-chat-input-border-glow-gradient-base: linear-gradient(#c084fc 0 100%);

    /* ── 文本区 ── */
    --ve-chat-input-text-size: var(--text-body-lg, 16px);
    --ve-chat-input-text-color: var(--color-text-primary, #1d2129);
    --ve-chat-input-placeholder-color: var(--color-text-disable, #b8b7c3);
    --ve-chat-input-line-height: 1.6;
    --ve-chat-input-min-height: 72px;
    --ve-chat-input-max-height: 320px;
    --ve-chat-input-caret-color: var(--color-border-interactive, #165dff);

    /* ── 全屏 ── */
    --ve-chat-input-expanded-offset: 60px;

    /* ── header ── */
    --ve-chat-input-header-gap: 4px;
    --ve-chat-input-header-padding-inline: 4px;

    /* ── 工具栏 ── */
    --ve-chat-input-toolbar-height: 32px;
    --ve-chat-input-toolbar-gap: 4px;
    --ve-chat-input-action-gap: 8px;

    /* ── 发送 / 停止按钮 ── */
    --ve-chat-input-send-size: 32px;
    --ve-chat-input-send-radius: var(--radius-full, 99px);
    --ve-chat-input-send-bg: var(--color-bg-primary, #1d2129);
    --ve-chat-input-send-color: var(--color-text-foreground, #fff);
    --ve-chat-input-stop-bg: var(--color-bg-muted, #f2f3f5);
    --ve-chat-input-stop-color: var(--color-text-primary, #1d2129);

    /* ── Action 子组件 ── */
    --ve-chat-input-action-height: 32px;
    --ve-chat-input-action-padding-inline: 12px;
    --ve-chat-input-action-gap-inner: 4px;
    --ve-chat-input-action-radius: var(--radius-md, 8px);
    --ve-chat-input-action-icon-only-radius: var(--radius-lg, 12px);
    --ve-chat-input-action-icon-only-bg: var(--color-bg-secondary, #f7f8fa);
    --ve-chat-input-action-bg: transparent;
    --ve-chat-input-action-text: var(--color-text-primary, #1d2129);
    --ve-chat-input-action-active-bg: var(--color-bg-secondary, #f2f3f5);
    --ve-chat-input-action-active-text: var(--color-text-primary, #1d2129);

    /* ── 全屏切换按钮 ── */
    --ve-chat-input-expand-size: 22px;
    --ve-chat-input-expand-color: var(--color-text-secondary, #4e5969);
    --ve-chat-input-expand-hover-bg: var(--color-bg-muted, #f2f3f5);

    /* ── 状态 ── */
    --ve-chat-input-disabled-opacity: var(--state-disabled-ratio, 0.4);

    /* ── 动效 ── */
    --ve-chat-input-motion-duration: var(--motion-fast, 160ms);
    --ve-chat-input-motion-easing: var(--ease-standard, ease);
  }

  /* 聚焦 / 输入 / 完成：边框加深 */
  .ve-chat-input-wrapper[data-state='focused'],
  .ve-chat-input-wrapper[data-state='typing'],
  .ve-chat-input-wrapper[data-state='done'] {
    --ve-chat-input-border-color: var(--ve-chat-input-border-color-active);
  }
`;
const token = veChatInputToken;
const ve_chat_input = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      inline-size: 100%;
      color: inherit;
      font-family: var(--ve-chat-input-font-family);
    }

    :host([size='small']) .ve-chat-input-container {
      max-inline-size: 420px;
    }

    .ve-chat-input-container {
      inline-size: 100%;
      block-size: fit-content;
    }

    /* ── wrapper：与 ve-composer :host 对齐 ── */
    .ve-chat-input-wrapper {
      position: relative;
      display: flex;
      flex-direction: column;
      box-sizing: border-box;
      gap: var(--ve-chat-input-gap);
      padding: var(--ve-chat-input-padding);
      background: var(--ve-chat-input-bg);
      border: var(--ve-chat-input-border-width) solid
        var(--ve-chat-input-border-color);
      border-radius: var(--ve-chat-input-radius);
      box-shadow: var(--ve-chat-input-shadow);
      transition: border-color var(--ve-chat-input-motion-duration)
        var(--ve-chat-input-motion-easing);
      isolation: isolate;
      overflow: visible;
      transform: translate3d(0, 0, 0.01px);
    }

    .ve-chat-input-wrapper::before,
    .ve-chat-input-wrapper::after,
    .ve-chat-input-border-glow-edge {
      content: '';
      position: absolute;
      inset: 0;
      z-index: 0;
      border-radius: inherit;
      opacity: 0;
      pointer-events: none;
      transition: opacity 250ms ease-out;
    }

    :host([border-glow])
      .ve-chat-input-wrapper:hover:not(.is-disabled)::before {
      border: var(--ve-chat-input-border-width) solid transparent;
      background:
        linear-gradient(var(--ve-chat-input-bg) 0 100%) padding-box,
        linear-gradient(rgb(255 255 255 / 0%) 0 100%) border-box,
        var(--ve-chat-input-border-glow-gradient-1) border-box,
        var(--ve-chat-input-border-glow-gradient-2) border-box,
        var(--ve-chat-input-border-glow-gradient-3) border-box,
        var(--ve-chat-input-border-glow-gradient-4) border-box,
        var(--ve-chat-input-border-glow-gradient-5) border-box,
        var(--ve-chat-input-border-glow-gradient-6) border-box,
        var(--ve-chat-input-border-glow-gradient-7) border-box,
        var(--ve-chat-input-border-glow-gradient-base) border-box;
      opacity: calc(
        (
            var(--ve-chat-input-border-glow-edge-proximity) -
              var(--ve-chat-input-border-glow-color-sensitivity)
          ) /
          (
            100 -
              var(--ve-chat-input-border-glow-color-sensitivity)
          )
      );
      mask-image: conic-gradient(
        from var(--ve-chat-input-border-glow-cursor-angle) at center,
        black calc(var(--ve-chat-input-border-glow-cone-spread) * 1%),
        transparent
          calc((var(--ve-chat-input-border-glow-cone-spread) + 15) * 1%),
        transparent
          calc((100 - var(--ve-chat-input-border-glow-cone-spread) - 15) * 1%),
        black calc((100 - var(--ve-chat-input-border-glow-cone-spread)) * 1%)
      );
    }

    :host([border-glow])
      .ve-chat-input-wrapper:hover:not(.is-disabled)::after {
      border: var(--ve-chat-input-border-width) solid transparent;
      background:
        var(--ve-chat-input-border-glow-gradient-1) padding-box,
        var(--ve-chat-input-border-glow-gradient-2) padding-box,
        var(--ve-chat-input-border-glow-gradient-3) padding-box,
        var(--ve-chat-input-border-glow-gradient-4) padding-box,
        var(--ve-chat-input-border-glow-gradient-5) padding-box,
        var(--ve-chat-input-border-glow-gradient-6) padding-box,
        var(--ve-chat-input-border-glow-gradient-7) padding-box,
        var(--ve-chat-input-border-glow-gradient-base) padding-box;
      opacity: calc(
        var(--ve-chat-input-border-glow-fill-opacity) *
          (
            var(--ve-chat-input-border-glow-edge-proximity) -
              var(--ve-chat-input-border-glow-color-sensitivity)
          ) /
          (
            100 -
              var(--ve-chat-input-border-glow-color-sensitivity)
          )
      );
      mask-image:
        linear-gradient(to bottom, black, black),
        radial-gradient(ellipse at 50% 50%, black 40%, transparent 65%),
        radial-gradient(ellipse at 66% 66%, black 5%, transparent 40%),
        radial-gradient(ellipse at 33% 33%, black 5%, transparent 40%),
        radial-gradient(ellipse at 66% 33%, black 5%, transparent 40%),
        radial-gradient(ellipse at 33% 66%, black 5%, transparent 40%),
        conic-gradient(
          from var(--ve-chat-input-border-glow-cursor-angle) at center,
          transparent 5%,
          black 15%,
          black 85%,
          transparent 95%
        );
      mask-composite: subtract, add, add, add, add, add;
      mix-blend-mode: soft-light;
    }

    :host([border-glow])
      .ve-chat-input-wrapper:hover:not(.is-disabled)
      > .ve-chat-input-border-glow-edge {
      inset: calc(var(--ve-chat-input-border-glow-padding) * -1);
      z-index: 1;
      opacity: calc(
        (
            var(--ve-chat-input-border-glow-edge-proximity) -
              var(--ve-chat-input-border-glow-edge-sensitivity)
          ) /
          (
            100 -
              var(--ve-chat-input-border-glow-edge-sensitivity)
          )
      );
      mask-image: conic-gradient(
        from var(--ve-chat-input-border-glow-cursor-angle) at center,
        black 2.5%,
        transparent 10%,
        transparent 90%,
        black 97.5%
      );
      mix-blend-mode: plus-lighter;
    }

    .ve-chat-input-border-glow-edge::before {
      content: '';
      position: absolute;
      inset: var(--ve-chat-input-border-glow-padding);
      border-radius: inherit;
      /* 仅保留外发光；内侧不投射光晕，避免鼠标移过边框时内部偶现色圈 */
      box-shadow:
        0 0 1px 0 var(--ve-chat-input-border-glow-color-60),
        0 0 3px 0 var(--ve-chat-input-border-glow-color-50),
        0 0 6px 0 var(--ve-chat-input-border-glow-color-40),
        0 0 15px 0 var(--ve-chat-input-border-glow-color-30),
        0 0 25px 2px var(--ve-chat-input-border-glow-color-20),
        0 0 50px 2px var(--ve-chat-input-border-glow-color-10);
    }

    .ve-chat-input-wrapper.is-disabled {
      opacity: var(--ve-chat-input-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-chat-input-wrapper.is-disabled .ve-chat-input-textarea,
    .ve-chat-input-wrapper.is-disabled ::slotted(*) {
      pointer-events: none;
    }

    .ve-chat-input-box {
      position: relative;
      z-index: 2;
      display: flex;
      flex-direction: column;
      gap: var(--ve-chat-input-gap);
      inline-size: 100%;
    }

    /* ── header ── */
    .ve-chat-input-header {
      display: flex;
      flex-direction: column;
      gap: var(--ve-chat-input-header-gap);
      padding-inline-start: var(--ve-chat-input-header-padding-inline);
      padding-inline-end: var(--ve-chat-input-header-padding-inline);
    }

    .ve-chat-input-header-empty {
      display: none;
    }

    /* ── attachments ── */
    .ve-chat-input-attachments {
      display: grid;
      grid-template-columns: repeat(auto-fill, minmax(100px, 1fr));
      gap: var(--ve-chat-input-attachment-gap, 8px);
      padding-inline: var(--ve-chat-input-header-padding-inline);
      max-inline-size: 100%;
    }

    .ve-chat-input-attachment {
      position: relative;
      display: flex;
      flex-direction: column;
      border-radius: var(--ve-chat-input-attachment-radius, 8px);
      overflow: visible;
      background: var(--ve-chat-input-attachment-bg, var(--color-bg-secondary));
      transition: box-shadow var(--ve-chat-input-motion-duration)
        var(--ve-chat-input-motion-easing);
    }

    .ve-chat-input-attachment:hover {
      box-shadow: var(--shadow-sm);
    }

    .ve-chat-input-attachment-image {
      aspect-ratio: 1;
    }

    .ve-chat-input-attachment-image-wrapper {
      position: relative;
      inline-size: 100%;
      block-size: 100%;
      overflow: hidden;
      border-radius: inherit;
    }

    .ve-chat-input-attachment-image-content {
      inline-size: 100%;
      block-size: 100%;
      object-fit: cover;
      display: block;
    }

    .ve-chat-input-attachment-image.is-error
      .ve-chat-input-attachment-image-content {
      filter: grayscale(1);
    }

    .ve-chat-input-attachment-progress-overlay {
      position: absolute;
      inset: 0;
      display: flex;
      align-items: center;
      justify-content: center;
      background: rgb(0 0 0 / 50%);
      color: var(--color-text-foreground);
      font-size: var(--text-body-sm);
    }

    .ve-chat-input-attachment-image .ve-chat-input-attachment-error-meta {
      position: absolute;
      inset-inline: var(--space-xxs);
      bottom: var(--space-xxs);
      z-index: 1;
    }

    .ve-chat-input-attachment-file {
      min-block-size: 80px;
      padding: var(--space-xs);
    }

    .ve-chat-input-attachment-file-wrapper {
      display: flex;
      flex-direction: column;
      justify-content: space-between;
      height: 100%;
    }

    .ve-chat-input-attachment-file-icon {
      color: var(--color-text-accent);
    }

    .ve-chat-input-attachment-file-title {
      display: -webkit-box;
      -webkit-box-orient: vertical;
      -webkit-line-clamp: 2;
      overflow: hidden;
      font-size: 12px;
      line-height: 18px;
      color: var(--color-text-primary);
      word-break: break-word;
    }

    .ve-chat-input-attachment-file-footer {
      min-inline-size: 0;
      min-block-size: 18px;
    }

    .ve-chat-input-attachment-progress-text {
      font-size: 12px;
      color: var(--color-text-tertiary);
      font-variant-numeric: tabular-nums;
    }

    .ve-chat-input-attachment-error-meta {
      display: flex;
      align-items: center;
      gap: 2px;
      min-inline-size: 0;
      font-size: 12px;
      line-height: 18px;
      color: var(--color-text-danger);
      white-space: nowrap;
    }

    .ve-chat-input-attachment-error-meta span {
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ve-chat-input-attachment-error-icon {
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      color: var(--color-text-danger);
    }

    .ve-chat-input-attachment-remove {
      position: absolute;
      z-index: 2;
      top: -6px;
      right: -6px;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-upload-attachment-remove-size, 20px);
      block-size: var(--ve-upload-attachment-remove-size, 20px);
      border-radius: var(--radius-full);
      color: var(
        --ve-upload-attachment-remove-color,
        var(--color-text-foreground)
      );
      background: var(
        --ve-upload-attachment-remove-bg,
        var(--color-bg-inverse)
      );
      font-size: var(--ve-upload-attachment-remove-icon-size, 12px);
      line-height: 1;
      cursor: pointer;
      opacity: 0;
      pointer-events: none;
      transform: scale(0.92);
      transition:
        opacity var(--ve-chat-input-motion-duration)
          var(--ve-chat-input-motion-easing),
        transform var(--ve-chat-input-motion-duration)
          var(--ve-chat-input-motion-easing);
    }

    .ve-chat-input-attachment:hover .ve-chat-input-attachment-remove,
    .ve-chat-input-attachment:focus-within .ve-chat-input-attachment-remove,
    .ve-chat-input-attachment-remove:focus-visible {
      opacity: var(--ve-upload-attachment-remove-opacity, 0.82);
      pointer-events: auto;
      transform: scale(1);
    }

    .ve-chat-input-attachment-remove:hover {
      opacity: 1;
      transform: scale(1.04);
    }

    .ve-chat-input-attachment-remove:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px
        var(--ve-upload-color-icon-box-shadow-default, var(--color-border-focus));
    }

    /* ── body：与 ve-composer __body 对齐 ── */
    .ve-chat-input-body {
      display: flex;
      flex-direction: column;
      gap: 4px;
      padding: 4px 0 0 4px;
    }

    .ve-chat-input-textarea {
      display: block;
      inline-size: 100%;
      box-sizing: border-box;
      padding: 0;
      border: none;
      outline: none;
      resize: none;
      background: transparent;
      color: var(--ve-chat-input-text-color);
      font-family: var(--ve-chat-input-font-family);
      font-size: var(--ve-chat-input-text-size);
      line-height: var(--ve-chat-input-line-height);
      caret-color: var(--ve-chat-input-caret-color);
      min-block-size: var(--ve-chat-input-min-height);
      max-block-size: var(--ve-chat-input-max-height);
      overflow-y: auto;
      scrollbar-width: thin;
    }

    .ve-chat-input-textarea::placeholder {
      color: var(--ve-chat-input-placeholder-color);
    }

    .ve-chat-input-textarea.is-full-expanded {
      min-block-size: calc(100vh - var(--ve-chat-input-expanded-offset));
      max-block-size: calc(100vh - var(--ve-chat-input-expanded-offset));
    }

    .ve-chat-input-textarea[disabled] {
      cursor: not-allowed;
    }

    /* ── 工具栏：与 ve-composer __toolbar 对齐 ── */
    .ve-chat-input-toolbar {
      display: flex;
      align-items: center;
      justify-content: space-between;
      block-size: var(--ve-chat-input-toolbar-height);
    }

    .ve-chat-input-actions {
      display: flex;
      align-items: center;
      gap: var(--ve-chat-input-toolbar-gap);
      flex: 1 1 auto;
      min-inline-size: 0;
    }

    .ve-chat-input-submit {
      display: flex;
      align-items: center;
      gap: var(--ve-chat-input-action-gap);
      flex: none;
    }

    /* ── 发送 / 停止：与 ve-composer __send / __stop 对齐 ── */
    .ve-chat-input-send,
    .ve-chat-input-stop {
      flex: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-chat-input-send-size);
      block-size: var(--ve-chat-input-send-size);
      padding: 0;
      border: none;
      border-radius: var(--ve-chat-input-send-radius);
      cursor: pointer;
      transition: background-color var(--ve-chat-input-motion-duration)
        var(--ve-chat-input-motion-easing);
    }

    .ve-chat-input-send {
      background: var(--ve-chat-input-send-bg);
      color: var(--ve-chat-input-send-color);
    }

    .ve-chat-input-send:hover:not(:disabled) {
      background: color-mix(
        in srgb,
        var(--ve-chat-input-send-bg) var(--state-hover-solid-ratio, 90%),
        var(--state-hover-solid-mix, #fff)
      );
    }

    .ve-chat-input-send:disabled {
      opacity: var(--ve-chat-input-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-chat-input-stop {
      background: var(--ve-chat-input-stop-bg);
      color: var(--ve-chat-input-stop-color);
    }

    .ve-chat-input-stop:hover:not(:disabled) {
      background: color-mix(
        in srgb,
        var(--ve-chat-input-stop-bg) var(--state-hover-subtle-ratio, 97%),
        var(--state-hover-subtle-mix, #000)
      );
    }

    .ve-chat-input-stop-icon {
      inline-size: 8px;
      block-size: 8px;
      background: currentColor;
      border-radius: 1px;
      display: inline-block;
    }

    /* ── 全屏切换按钮 ── */
    .ve-chat-input-expand-btn {
      position: absolute;
      top: 12px;
      right: 12px;
      z-index: 1;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-chat-input-expand-size);
      block-size: var(--ve-chat-input-expand-size);
      padding: 0;
      border: none;
      border-radius: var(--radius-sm, 6px);
      background: transparent;
      color: var(--ve-chat-input-expand-color);
      cursor: pointer;
      transition: background-color var(--ve-chat-input-motion-duration)
        var(--ve-chat-input-motion-easing);
    }

    .ve-chat-input-expand-btn:hover {
      background: var(--ve-chat-input-expand-hover-bg);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-chat-input-wrapper,
      .ve-chat-input-send,
      .ve-chat-input-stop,
      .ve-chat-input-expand-btn,
      .ve-chat-input-wrapper::before,
      .ve-chat-input-wrapper::after,
      .ve-chat-input-border-glow-edge {
        transition: none;
      }
    }

    /* ── trigger panel (command palette container) ── */
    .ve-chat-input-trigger-panel {
      position: fixed;
      inset: auto;
      z-index: var(--ve-chat-input-trigger-panel-z, 1000);
      margin: 0;
      padding: 0;
      border: none;
      background: transparent;
      overflow: visible;
    }
  `
];
function isChatInputValue(value) {
    return 'object' == typeof value && null !== value && Array.isArray(value.segments);
}
function normalizeChatInputValue(value) {
    if (isChatInputValue(value)) return {
        segments: value.segments ?? [],
        attachments: value.attachments ?? []
    };
    const text = 'string' == typeof value ? value : '';
    const segments = text ? [
        {
            type: 'text',
            text
        }
    ] : [];
    return {
        segments,
        attachments: []
    };
}
function defaultNodeText(node) {
    const data = node.data ?? {};
    const candidate = data.text ?? data.value ?? data.label;
    return 'string' == typeof candidate ? candidate : '';
}
function chatInputValueToText(value, getNodeText = defaultNodeText) {
    return value.segments.map((segment)=>'text' === segment.type ? segment.text : getNodeText(segment)).join('');
}
function isChatInputValueEmpty(value) {
    if (value.attachments.length > 0) return false;
    return 0 === chatInputValueToText(value).trim().length;
}
function resolveChatInputNodeRender(renders, nodeType) {
    return renders[nodeType];
}
function isChatInputNodeEditable(renders, node) {
    return renders[node.nodeType]?.editable ?? false;
}
function getChatInputNodeText(renders, node) {
    const render = renders[node.nodeType];
    return render?.getText?.(node) ?? defaultNodeText(node);
}
function str(value, fallback = '') {
    return 'string' == typeof value ? value : fallback;
}
function createChipIcon(icon) {
    const el = document.createElement('ve-icon');
    el.className = 've-ci-node-icon';
    el.setAttribute('name', icon);
    el.setAttribute('size', '14');
    return el;
}
function createLabel(text) {
    const label = document.createElement('span');
    label.className = 've-ci-node-label';
    label.textContent = text;
    return label;
}
const tagRender = {
    editable: false,
    getText: (node)=>{
        const data = node.data ?? {};
        return str(data.value) || str(data.label);
    },
    render: ({ node })=>{
        const data = node.data ?? {};
        const label = str(data.label) || str(data.value);
        const icon = str(data.icon);
        const iconPosition = 'after' === data.iconPosition ? 'after' : 'before';
        const span = document.createElement('span');
        span.className = 've-ci-node ve-ci-tag';
        if (icon && 'before' === iconPosition) span.appendChild(createChipIcon(icon));
        span.appendChild(createLabel(label));
        if (icon && 'after' === iconPosition) span.appendChild(createChipIcon(icon));
        return span;
    }
};
const selectRender = {
    editable: false,
    getText: (node)=>str(node.data?.value),
    render: ({ node, disabled, updateData })=>{
        const data = node.data ?? {};
        const value = str(data.value);
        const placeholder = str(data.placeholder, 'Please select');
        const options = Array.isArray(data.options) ? data.options.map((opt)=>str(opt)) : [];
        const span = document.createElement('span');
        span.className = 've-ci-node ve-ci-select';
        const select = document.createElement('ve-select');
        select.className = 've-ci-select-control';
        select.setAttribute('size', 'small');
        select.setAttribute('popup-match-select-width', 'false');
        select.setAttribute('placeholder', placeholder);
        select.value = value;
        if (disabled) select.setAttribute('disabled', '');
        select.addEventListener('ve-change', (event)=>{
            const next = event.detail.value;
            updateData({
                value: 'string' == typeof next ? next : String(next)
            });
        });
        for (const opt of options){
            const item = document.createElement('ve-select-item');
            item.setAttribute('value', opt);
            item.textContent = opt;
            select.appendChild(item);
        }
        span.appendChild(select);
        return span;
    }
};
const variableRender = {
    editable: true,
    getText: (node)=>str(node.data?.value),
    render: ({ node })=>{
        const data = node.data ?? {};
        const value = str(data.value);
        const placeholder = str(data.placeholder);
        const span = document.createElement('span');
        span.className = 've-ci-node ve-ci-variable';
        span.dataset.empty = value ? 'false' : 'true';
        span.dataset.placeholder = placeholder;
        span.textContent = value;
        return span;
    }
};
const DEFAULT_CHAT_INPUT_NODE_RENDERS = {
    tag: tagRender,
    select: selectRender,
    variable: variableRender
};
function mergeChatInputNodeRenders(overrides) {
    if (!overrides) return DEFAULT_CHAT_INPUT_NODE_RENDERS;
    return {
        ...DEFAULT_CHAT_INPUT_NODE_RENDERS,
        ...overrides
    };
}
const style_editor = [
    componentReset,
    css`
    :host {
      display: block;
      width: 100%;
    }

    .ve-ci-editor {
      display: block;
      box-sizing: border-box;
      width: 100%;
      min-height: var(--ve-chat-input-min-height, 72px);
      max-height: var(--ve-chat-input-max-height, 320px);
      padding: 4px;
      overflow-y: auto;
      scrollbar-width: thin;
      outline: none;
      font-family: var(--ve-chat-input-font-family, inherit);
      font-size: var(--ve-chat-input-text-size, 16px);
      line-height: 24px;
      color: var(--ve-chat-input-text-color, var(--color-text-primary, #101013));
      caret-color: var(--ve-chat-input-caret-color, currentColor);
      /* Preserve trailing/consecutive spaces so caret landing points survive. */
      white-space: pre-wrap;
      word-wrap: break-word;
      overflow-wrap: break-word;
    }

    /* Placeholder, shown only when the editor is empty. */
    .ve-ci-editor.is-empty::before {
      content: attr(data-placeholder);
      color: var(--ve-chat-input-placeholder-color, rgba(16, 16, 19, 0.4));
      pointer-events: none;
    }

    .ve-ci-editor.is-full-expanded {
      min-height: calc(100vh - var(--ve-chat-input-expanded-offset, 60px));
      max-height: calc(100vh - var(--ve-chat-input-expanded-offset, 60px));
    }

    /* ── caret landing anchors around an island ── */
    .ve-ci-space {
      display: inline-block;
      width: 3px;
      line-height: inherit;
      user-select: none;
      vertical-align: middle;
    }

    /* ── inline island wrapper ── */
    .ve-ci-island {
      display: inline-flex;
      align-items: center;
      vertical-align: middle;
      /* The editor sets white-space: pre-wrap to preserve caret anchors; that
         property inherits across shadow boundaries into embedded controls
         (e.g. ve-select / ve-date-picker), wrapping their popup text
         char-by-char. Reset it for islands — the editable island re-declares
         pre-wrap for its own text. */
      white-space: normal;
    }

    /* ── editable island (variable) ── */
    .ve-ci-island.ve-ci-editable {
      display: inline-block;
      min-width: 1em;
      padding: 0 4px;
      outline: none;
      color: var(--ve-ci-node-text, var(--color-text-info));
      background: var(--ve-ci-node-bg, var(--color-bg-info-secondary));
      border-radius: var(--radius-sm, 4px);
      caret-color: var(--ve-ci-node-text, var(--color-text-info));
      white-space: pre-wrap;
      word-break: break-word;
      box-decoration-break: clone;
      -webkit-box-decoration-break: clone;
    }

    .ve-ci-island.ve-ci-editable[data-empty='true']::before {
      content: attr(data-placeholder);
      color: var(--ve-ci-node-placeholder, var(--color-text-tertiary));
      pointer-events: none;
    }

    /* ── chip-style nodes (tag / select) ── */
    .ve-ci-node {
      display: inline-flex;
      flex-direction: row;
      align-items: center;
      gap: 4px;
      height: 24px;
      padding: 0 4px;
      background: var(--ve-ci-node-bg, var(--color-bg-info-secondary));
      border-radius: var(--radius-sm, 4px);
      vertical-align: middle;
      transition: background-color 0.2s ease;
    }

    .ve-ci-node-icon {
      flex-shrink: 0;
      width: 14px;
      height: 14px;
      color: var(--ve-ci-node-text, var(--color-text-info));
    }

    .ve-ci-node-label {
      font-size: 16px;
      line-height: 24px;
      color: var(--ve-ci-node-text, var(--color-text-info));
      white-space: nowrap;
      overflow: hidden;
    }

    .ve-ci-tag {
      cursor: default;
    }

    .ve-ci-tag:hover {
      background: var(--ve-ci-node-bg-hover, var(--color-info-200));
    }

    /* ── select node (backed by ve-select) ── */
    .ve-ci-select {
      padding: 0;
      background: transparent;
      border-radius: 0;
    }

    .ve-ci-select:hover {
      background: transparent;
    }

    .ve-ci-select .ve-ci-select-control {
      display: inline-flex;
      align-items: center;
      vertical-align: middle;
      block-size: 24px;
      line-height: 24px;
    }

    /* ── React-bridged custom node ── */
    .ve-ci-react-node {
      display: inline-flex;
      align-items: center;
      vertical-align: middle;
    }

    /* ── disabled state ── */
    :host([disabled]) .ve-ci-editor {
      cursor: not-allowed;
      opacity: var(--ve-chat-input-disabled-opacity, 0.4);
    }

    :host([disabled]) .ve-ci-node,
    :host([disabled]) .ve-ci-select {
      cursor: not-allowed;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _init_segments, _init_extra__init_segments, _init_placeholder, _init_extra__init_placeholder, _init_disabled, _init_extra__init_disabled, _init_expanded, _init_extra__init_expanded, _init_triggers, _init_extra__init_triggers, _init_nodeRenders, _init_extra__init_nodeRenders, _init_isEmpty, _init_extra__init_isEmpty;
const VE_CHAT_INPUT_EDITOR_TAG = 've-chat-input-editor';
let nodeIdSeed = 0;
function nextNodeId() {
    nodeIdSeed += 1;
    return `ve-ci-node-${Date.now().toString(36)}-${nodeIdSeed}`;
}
_dec = property({
    attribute: false
}), _dec1 = property({
    type: String
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    attribute: false
}), _dec5 = property({
    attribute: false
}), _dec6 = state();
class VeChatInputEditorElement extends VeElement {
    static{
        ({ e: [_init_segments, _init_extra__init_segments, _init_placeholder, _init_extra__init_placeholder, _init_disabled, _init_extra__init_disabled, _init_expanded, _init_extra__init_expanded, _init_triggers, _init_extra__init_triggers, _init_nodeRenders, _init_extra__init_nodeRenders, _init_isEmpty, _init_extra__init_isEmpty] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "segments"
            ],
            [
                _dec1,
                1,
                "placeholder"
            ],
            [
                _dec2,
                1,
                "disabled"
            ],
            [
                _dec3,
                1,
                "expanded"
            ],
            [
                _dec4,
                1,
                "triggers"
            ],
            [
                _dec5,
                1,
                "nodeRenders"
            ],
            [
                _dec6,
                1,
                "isEmpty"
            ]
        ]));
    }
    static styles = style_editor;
    static dependencies = {
        [VE_SELECT_TAG]: VeSelectElement,
        [VE_SELECT_ITEM_TAG]: VeSelectItemElement,
        [VE_ICON_TAG]: VeIconElement
    };
    static{
        registerIcons([
            chevronDownMdIcon
        ]);
    }
    #___private_segments_1 = (()=>{
        const _value = _init_segments(this, []);
        _init_extra__init_segments(this);
        return _value;
    })();
    get segments() {
        return this.#___private_segments_1;
    }
    set segments(_v) {
        this.#___private_segments_1 = _v;
    }
    #___private_placeholder_2 = (()=>{
        const _value = _init_placeholder(this, '');
        _init_extra__init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_2;
    }
    set placeholder(_v) {
        this.#___private_placeholder_2 = _v;
    }
    #___private_disabled_3 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_3;
    }
    set disabled(_v) {
        this.#___private_disabled_3 = _v;
    }
    #___private_expanded_4 = (()=>{
        const _value = _init_expanded(this, false);
        _init_extra__init_expanded(this);
        return _value;
    })();
    get expanded() {
        return this.#___private_expanded_4;
    }
    set expanded(_v) {
        this.#___private_expanded_4 = _v;
    }
    #___private_triggers_5 = (()=>{
        const _value = _init_triggers(this, []);
        _init_extra__init_triggers(this);
        return _value;
    })();
    get triggers() {
        return this.#___private_triggers_5;
    }
    set triggers(_v) {
        this.#___private_triggers_5 = _v;
    }
    #___private_nodeRenders_6 = (()=>{
        const _value = _init_nodeRenders(this, DEFAULT_CHAT_INPUT_NODE_RENDERS);
        _init_extra__init_nodeRenders(this);
        return _value;
    })();
    get nodeRenders() {
        return this.#___private_nodeRenders_6;
    }
    set nodeRenders(_v) {
        this.#___private_nodeRenders_6 = _v;
    }
    #___private_isEmpty_7 = (()=>{
        const _value = _init_isEmpty(this, true);
        _init_extra__init_isEmpty(this);
        return _value;
    })();
    get isEmpty() {
        return this.#___private_isEmpty_7;
    }
    set isEmpty(_v) {
        this.#___private_isEmpty_7 = _v;
    }
    pendingDeleteNodeId = null;
    lastTrigger = null;
    savedRange = null;
    caretRafId = null;
    get value() {
        const editor = this.getEditorEl();
        if (!editor) return this.segments;
        return this.readSegmentsFromDom(editor);
    }
    disconnectedCallback() {
        if (null !== this.caretRafId) {
            cancelAnimationFrame(this.caretRafId);
            this.caretRafId = null;
        }
        super.disconnectedCallback();
    }
    focus(options) {
        const editor = this.getEditorEl();
        if (!editor) return;
        editor.focus(options);
        this.placeCaretAtEnd();
        queueMicrotask(()=>this.ensureCaretAtEnd());
        this.scheduleEnsureCaretAtEnd();
    }
    willUpdate(changed) {
        if (changed.has('segments')) this.isEmpty = this.computeIsEmptyFromSegments();
    }
    updated(changed) {
        if (!changed.has('segments')) return;
        const editor = this.getEditorEl();
        if (!editor) return;
        const next = serializeSegments(this.segments);
        const current = serializeSegments(this.readSegmentsFromDom(editor));
        if (next === current) return;
        const wasActive = this.isSelectionInsideEditor() || this.isEditorFocused();
        this.renderEditorContent();
        if (wasActive) {
            queueMicrotask(()=>this.ensureCaretAtEnd());
            this.scheduleEnsureCaretAtEnd();
        }
    }
    getEditorEl() {
        return this.shadowRoot?.querySelector('.ve-ci-editor') ?? null;
    }
    getSelection() {
        const shadowSel = this.shadowRoot?.getSelection?.();
        if (shadowSel) return shadowSel;
        return window.getSelection();
    }
    computeIsEmptyFromSegments() {
        return this.segments.every((segment)=>'text' === segment.type ? 0 === segment.text.length : false);
    }
    isEditorFocused() {
        const editor = this.getEditorEl();
        if (!editor) return false;
        const root = editor.getRootNode();
        const active = root instanceof ShadowRoot ? root.activeElement : editor.ownerDocument.activeElement;
        return active === editor || editor.ownerDocument.activeElement === this;
    }
    isSelectionInsideEditor() {
        const editor = this.getEditorEl();
        if (!editor) return false;
        const sel = this.getSelection();
        if (!sel || 0 === sel.rangeCount) return false;
        const { anchorNode } = sel;
        return !!anchorNode && editor.contains(anchorNode);
    }
    readSegmentsFromDom(editor) {
        const segments = [];
        const pushText = (text)=>{
            if (!text) return;
            const last = segments[segments.length - 1];
            if (last && 'text' === last.type) last.text += text;
            else segments.push({
                type: 'text',
                text
            });
        };
        for (const node of Array.from(editor.childNodes)){
            if (node.nodeType === Node.TEXT_NODE) {
                pushText(node.textContent ?? '');
                continue;
            }
            if (node instanceof HTMLElement) {
                if (!node.classList.contains('ve-ci-space')) {
                    if ('BR' === node.tagName) {
                        pushText('\n');
                        continue;
                    }
                    if (node.classList.contains('ve-ci-island')) {
                        const segment = this.readIslandSegment(node);
                        if (segment) segments.push(segment);
                        continue;
                    }
                    pushText(node.textContent ?? '');
                }
            }
        }
        return segments;
    }
    readIslandSegment(island) {
        const id = island.dataset.nodeId;
        if (!id) return null;
        const source = this.segments.find((segment)=>'node' === segment.type && segment.id === id);
        const nodeType = island.dataset.nodeType ?? source?.nodeType ?? '';
        const data = {
            ...source?.data ?? {}
        };
        if ('true' === island.dataset.nodeEditable) data.value = island.textContent ?? '';
        return {
            type: 'node',
            id,
            nodeType,
            data
        };
    }
    scheduleEnsureCaretAtEnd() {
        if (null !== this.caretRafId) cancelAnimationFrame(this.caretRafId);
        this.caretRafId = requestAnimationFrame(()=>{
            this.caretRafId = null;
            if (this.isConnected) this.ensureCaretAtEnd();
        });
    }
    ensureCaretAtEnd() {
        const editor = this.getEditorEl();
        if (!editor) return;
        const sel = this.getSelection();
        if (!sel || 0 === sel.rangeCount) return void this.placeCaretAtEnd();
        if (!sel.isCollapsed && editor.contains(sel.anchorNode)) return;
        if (!editor.contains(sel.anchorNode)) this.placeCaretAtEnd();
    }
    placeCaretAtEnd() {
        const editor = this.getEditorEl();
        if (!editor) return;
        const last = this.lastSignificantChild(editor);
        if (last && last.nodeType === Node.TEXT_NODE) return void this.setCaretInTextNode(last, last.textContent?.length ?? 0);
        if (last instanceof HTMLElement && last.classList.contains('ve-ci-island') && 'true' === last.dataset.nodeEditable) return void this.setCaretInElement(last, last.textContent?.length ?? 0);
        const anchor = document.createTextNode('');
        editor.appendChild(anchor);
        this.setCaretInTextNode(anchor, 0);
    }
    lastSignificantChild(editor) {
        for(let i = editor.childNodes.length - 1; i >= 0; i -= 1){
            const node = editor.childNodes[i];
            if (!this.isGapNode(node)) return node ?? null;
        }
        return null;
    }
    isGapNode(node) {
        if (!node) return false;
        if (node.nodeType === Node.COMMENT_NODE) return true;
        if (node.nodeType === Node.TEXT_NODE) return 0 === (node.textContent ?? '').length;
        return false;
    }
    setCaretInTextNode(node, offset) {
        const max = node.textContent?.length ?? 0;
        const range = document.createRange();
        range.setStart(node, Math.min(offset, max));
        range.collapse(true);
        const sel = this.getSelection();
        sel?.removeAllRanges();
        sel?.addRange(range);
    }
    setCaretInElement(el, offset) {
        const textNode = this.firstTextNodeIn(el);
        if (textNode) return void this.setCaretInTextNode(textNode, offset);
        const anchor = document.createTextNode('');
        el.appendChild(anchor);
        this.setCaretInTextNode(anchor, 0);
    }
    firstTextNodeIn(node) {
        if (node.nodeType === Node.TEXT_NODE) return node;
        for (const child of Array.from(node.childNodes)){
            const found = this.firstTextNodeIn(child);
            if (found) return found;
        }
        return null;
    }
    setCaretByChildIndex(index) {
        const editor = this.getEditorEl();
        if (!editor) return;
        const safe = Math.max(0, Math.min(index, editor.childNodes.length));
        const range = document.createRange();
        range.setStart(editor, safe);
        range.collapse(true);
        const sel = this.getSelection();
        sel?.removeAllRanges();
        sel?.addRange(range);
    }
    setCaretAfterChild(node) {
        const editor = this.getEditorEl();
        if (!editor || node.parentNode !== editor) return;
        const index = Array.prototype.indexOf.call(editor.childNodes, node);
        if (index >= 0) this.setCaretByChildIndex(index + 1);
    }
    setCaretBeforeChild(node) {
        const editor = this.getEditorEl();
        if (!editor || node.parentNode !== editor) return;
        const index = Array.prototype.indexOf.call(editor.childNodes, node);
        if (index >= 0) this.setCaretByChildIndex(index);
    }
    significantSibling(editor, boundary, direction) {
        const step = 'backward' === direction ? -1 : 1;
        let index = 'backward' === direction ? boundary - 1 : boundary;
        while(index >= 0 && index < editor.childNodes.length){
            const node = editor.childNodes[index];
            if (!this.isGapNode(node)) return node ?? null;
            index += step;
        }
        return null;
    }
    renderEditorContent() {
        const editor = this.getEditorEl();
        if (!editor) return;
        editor.innerHTML = '';
        delete editor['_$litPart$'];
        external_lit_render(html`${this.segments.map((segment)=>this.renderSegment(segment))}`, editor);
    }
    renderSegment(segment) {
        if ('text' === segment.type) return segment.text;
        return this.renderNode(segment);
    }
    renderNode(node) {
        const editable = isChatInputNodeEditable(this.nodeRenders, node);
        const before = html`<span
      class="ve-ci-space"
      contenteditable="false"
      data-space="before"
      data-node-id=${node.id}
      >\u00A0</span
    >`;
        const after = html`<span
      class="ve-ci-space"
      contenteditable="false"
      data-space="after"
      data-node-id=${node.id}
      >\u00A0</span
    >`;
        return html`${before}${this.renderIsland(node, editable)}${after}`;
    }
    renderIsland(node, editable) {
        if (editable) {
            const value = 'string' == typeof node.data?.value ? node.data.value : '';
            const placeholder = 'string' == typeof node.data?.placeholder ? node.data.placeholder : '';
            return html`<span
        class="ve-ci-island ve-ci-editable"
        data-node-id=${node.id}
        data-node-type=${node.nodeType}
        data-node-editable="true"
        data-empty=${value ? 'false' : 'true'}
        data-placeholder=${placeholder}
        contenteditable="true"
        spellcheck="false"
        >${value}</span
      >`;
        }
        const render = resolveChatInputNodeRender(this.nodeRenders, node.nodeType);
        const content = render ? render.render({
            node,
            disabled: this.disabled,
            updateData: (patch)=>this.handleNodeDataUpdate(node.id, patch),
            remove: ()=>this.deleteNodeGroupById(node.id)
        }) : getChatInputNodeText(this.nodeRenders, node);
        return html`<span
      class="ve-ci-island"
      data-node-id=${node.id}
      data-node-type=${node.nodeType}
      contenteditable="false"
      >${content}</span
    >`;
    }
    render() {
        return html`<div
        class=${classMap({
            've-ci-editor': true,
            'is-empty': this.isEmpty,
            'is-full-expanded': this.expanded
        })}
        contenteditable=${this.disabled ? 'false' : 'true'}
        role="textbox"
        aria-multiline="true"
        aria-placeholder=${this.placeholder}
        data-placeholder=${this.placeholder}
        @input=${this.handleInput}
        @focus=${this.handleFocus}
        @blur=${this.handleBlur}
        @keydown=${this.handleKeyDown}
        @click=${this.handleClick}
      ></div>`;
    }
    queryActiveTrigger() {
        return this.computeTriggerDetail();
    }
    getCaretRect() {
        const editor = this.getEditorEl();
        if (!editor) return null;
        const sel = this.getSelection();
        const anchorNode = sel?.anchorNode ?? null;
        const inside = !!sel && sel.rangeCount > 0 && !!anchorNode && editor.contains(anchorNode);
        if (inside && sel) {
            const range = sel.getRangeAt(0).cloneRange();
            range.collapse(true);
            const rects = range.getClientRects();
            if (rects.length > 0 && rects[0]) return rects[0];
            const marker = document.createElement('span');
            marker.textContent = '\u200b';
            range.insertNode(marker);
            const rect = marker.getBoundingClientRect();
            marker.remove();
            if (rect.width > 0 || rect.height > 0) return rect;
        }
        const editorRect = editor.getBoundingClientRect();
        return new DOMRect(editorRect.left + 4, editorRect.top + 4, 0, 20);
    }
    insertNode(node, options) {
        const editor = this.getEditorEl();
        if (!editor) return;
        this.restoreCaretIfNeeded(editor);
        editor.focus();
        this.restoreCaretIfNeeded(editor);
        if (options?.replaceTrigger !== false) this.removeActiveTriggerText();
        this.lastTrigger = null;
        const id = node.id || nextNodeId();
        const island = {
            ...node,
            id
        };
        const tpl = html`${this.renderNode(island)}`;
        const fragmentHost = document.createElement('span');
        external_lit_render(tpl, fragmentHost);
        const inserted = Array.from(fragmentHost.childNodes);
        this.insertNodesAtCaret(editor, inserted);
        const afterSpace = inserted[inserted.length - 1];
        if (afterSpace) this.setCaretAfterChild(afterSpace);
        this.savedRange = null;
        this.emitChange();
    }
    restoreCaretIfNeeded(editor) {
        const sel = this.getSelection();
        if (sel && sel.rangeCount > 0 && editor.contains(sel.anchorNode)) return;
        if (!this.savedRange) return;
        if (!editor.contains(this.savedRange.startContainer)) return;
        sel?.removeAllRanges();
        sel?.addRange(this.savedRange.cloneRange());
    }
    saveCaret() {
        const editor = this.getEditorEl();
        if (!editor) return;
        const sel = this.getSelection();
        if (!sel || 0 === sel.rangeCount || !sel.isCollapsed) return;
        if (!editor.contains(sel.anchorNode)) return;
        this.savedRange = sel.getRangeAt(0).cloneRange();
    }
    insertNodesAtCaret(editor, nodes) {
        const sel = this.getSelection();
        let range;
        if (sel && sel.rangeCount > 0 && editor.contains(sel.anchorNode)) {
            range = sel.getRangeAt(0);
            range.deleteContents();
        } else {
            range = document.createRange();
            range.selectNodeContents(editor);
            range.collapse(false);
        }
        const frag = document.createDocumentFragment();
        for (const node of nodes)frag.appendChild(node);
        range.insertNode(frag);
    }
    removeActiveTriggerText() {
        const detail = this.computeTriggerDetail();
        if (!detail.trigger) return;
        const sel = this.getSelection();
        if (!sel || 0 === sel.rangeCount) return;
        const { anchorNode, anchorOffset } = sel;
        if (!anchorNode || anchorNode.nodeType !== Node.TEXT_NODE) return;
        const removeLen = detail.trigger.prefix.length + detail.query.length;
        const start = Math.max(0, anchorOffset - removeLen);
        const range = document.createRange();
        range.setStart(anchorNode, start);
        range.setEnd(anchorNode, anchorOffset);
        range.deleteContents();
    }
    computeTriggerDetail() {
        const empty = {
            trigger: null,
            query: ''
        };
        if (0 === this.triggers.length) return empty;
        const sel = this.getSelection();
        if (!sel || 0 === sel.rangeCount || !sel.isCollapsed) return empty;
        const { anchorNode, anchorOffset } = sel;
        if (!anchorNode || anchorNode.nodeType !== Node.TEXT_NODE) return empty;
        const text = (anchorNode.textContent ?? '').slice(0, anchorOffset);
        for (const trigger of this.triggers){
            const idx = text.lastIndexOf(trigger.prefix);
            if (idx < 0) continue;
            const query = text.slice(idx + trigger.prefix.length);
            if (/\s/.test(query)) continue;
            const charBefore = idx > 0 ? text[idx - 1] : '';
            if (!charBefore || /\s/.test(charBefore)) return {
                trigger,
                query
            };
        }
        return empty;
    }
    emitTriggerIfChanged() {
        const detail = this.computeTriggerDetail();
        const changed = detail.trigger?.prefix !== this.lastTrigger?.trigger?.prefix || detail.query !== this.lastTrigger?.query;
        if (!changed) return;
        this.lastTrigger = detail;
        this.dispatchEvent(new CustomEvent('ve-ci-editor-trigger', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    handleNodeDataUpdate(id, patch) {
        const editor = this.getEditorEl();
        if (!editor) return;
        const source = this.segments.find((segment)=>'node' === segment.type && segment.id === id);
        if (source) source.data = {
            ...source.data ?? {},
            ...patch
        };
        const island = editor.querySelector(`.ve-ci-island[data-node-id="${id}"]`);
        if (island && source) {
            const editable = isChatInputNodeEditable(this.nodeRenders, source);
            const host = document.createElement('span');
            external_lit_render(html`${this.renderIsland(source, editable)}`, host);
            const next = host.querySelector('.ve-ci-island');
            if (next) island.replaceWith(next);
        }
        this.emitChange();
    }
    handleInput = ()=>{
        this.pendingDeleteNodeId = null;
        const editor = this.getEditorEl();
        if (editor) {
            this.isEmpty = 0 === (editor.textContent ?? '').trim().length;
            for (const island of editor.querySelectorAll('.ve-ci-island[data-node-editable="true"]')){
                const el = island;
                el.dataset.empty = (el.textContent ?? '').trim() ? 'false' : 'true';
            }
        }
        this.emitTriggerIfChanged();
        this.saveCaret();
        this.emitChange();
    };
    emitChange() {
        this.dispatchEvent(new CustomEvent('ve-ci-editor-change', {
            detail: {
                segments: this.value
            },
            bubbles: true,
            composed: true
        }));
    }
    handleFocus = ()=>{
        this.dispatchEvent(new CustomEvent('ve-ci-editor-focus', {
            bubbles: true,
            composed: true
        }));
        const sel = this.getSelection();
        const editor = this.getEditorEl();
        if (editor && sel && (0 === sel.rangeCount || !editor.contains(sel.anchorNode))) this.placeCaretAtEnd();
    };
    handleBlur = ()=>{
        this.dispatchEvent(new CustomEvent('ve-ci-editor-blur', {
            bubbles: true,
            composed: true
        }));
    };
    handleKeyDown = (event)=>{
        if (event.isComposing) return;
        if ('Backspace' !== event.key && 'Delete' !== event.key) this.pendingDeleteNodeId = null;
        if ('Enter' === event.key && !event.shiftKey) {
            event.preventDefault();
            this.dispatchEvent(new CustomEvent('ve-ci-editor-submit', {
                bubbles: true,
                composed: true
            }));
            return;
        }
        if ('ArrowLeft' === event.key || 'ArrowRight' === event.key) return void this.handleArrowKey(event);
        if ('Backspace' === event.key || 'Delete' === event.key) this.handleDeleteKey(event);
    };
    handleArrowKey(event) {
        const editor = this.getEditorEl();
        if (!editor) return;
        const sel = this.getSelection();
        if (!sel || 0 === sel.rangeCount || !sel.isCollapsed) return;
        const { anchorNode, anchorOffset } = sel;
        if (!anchorNode) return;
        const forward = 'ArrowRight' === event.key;
        let boundary = null;
        if (anchorNode === editor) boundary = anchorOffset;
        else if (anchorNode.nodeType === Node.TEXT_NODE && anchorNode.parentNode === editor) {
            const index = Array.prototype.indexOf.call(editor.childNodes, anchorNode);
            const len = anchorNode.textContent?.length ?? 0;
            if (forward && anchorOffset >= len) boundary = index + 1;
            else if (!forward && 0 === anchorOffset) boundary = index;
        }
        if (null === boundary) return;
        const adjacent = this.significantSibling(editor, boundary, forward ? 'forward' : 'backward');
        if (adjacent instanceof HTMLElement && adjacent.classList.contains('ve-ci-space')) {
            event.preventDefault();
            const island = forward ? adjacent.nextElementSibling : adjacent.previousElementSibling;
            const farSpace = forward ? island?.nextElementSibling : island?.previousElementSibling;
            const target = farSpace ?? adjacent;
            if (forward) this.setCaretAfterChild(target);
            else this.setCaretBeforeChild(target);
        }
    }
    handleDeleteKey(event) {
        if (this.deleteFromPendingClick(event.key)) return void event.preventDefault();
        if (this.deleteFromEventTarget(event)) return void event.preventDefault();
        const editor = this.getEditorEl();
        if (!editor) return;
        const sel = this.getSelection();
        if (!sel || 0 === sel.rangeCount || !sel.isCollapsed) return;
        const { anchorNode, anchorOffset } = sel;
        if (!anchorNode) return;
        const backspace = 'Backspace' === event.key;
        const island = (anchorNode.nodeType === Node.TEXT_NODE ? anchorNode.parentElement : anchorNode)?.closest?.('.ve-ci-island[data-node-editable="true"]') ?? null;
        if (island instanceof HTMLElement && 0 === (island.textContent ?? '').length) {
            const id = island.dataset.nodeId;
            if (id && this.deleteNodeGroupById(id)) return void event.preventDefault();
        }
        let boundary = null;
        if (anchorNode === editor) boundary = anchorOffset;
        else if (anchorNode.nodeType === Node.TEXT_NODE && anchorNode.parentNode === editor) {
            const index = Array.prototype.indexOf.call(editor.childNodes, anchorNode);
            const len = anchorNode.textContent?.length ?? 0;
            if (backspace && 0 === anchorOffset) boundary = index;
            else if (!backspace && anchorOffset >= len) boundary = index + 1;
        }
        if (null === boundary) return;
        const adjacent = this.significantSibling(editor, boundary, backspace ? 'backward' : 'forward');
        if (adjacent instanceof HTMLElement && adjacent.classList.contains('ve-ci-space')) {
            const id = adjacent.dataset.nodeId;
            if (id && this.deleteNodeGroupById(id)) event.preventDefault();
        }
    }
    deleteFromPendingClick(key) {
        const pending = this.pendingDeleteNodeId;
        if (!pending) return false;
        const matches = 'Backspace' === key && 'after' === pending.side || 'Delete' === key && 'before' === pending.side;
        if (!matches) return false;
        this.pendingDeleteNodeId = null;
        return this.deleteNodeGroupById(pending.id);
    }
    deleteFromEventTarget(event) {
        const island = event.composedPath().find((node)=>node instanceof HTMLElement && node.classList.contains('ve-ci-island') && 'true' !== node.dataset.nodeEditable);
        const id = island?.dataset.nodeId;
        if (!id) return false;
        return this.deleteNodeGroupById(id);
    }
    deleteNodeGroupById(id) {
        const editor = this.getEditorEl();
        if (!editor) return false;
        const group = Array.from(editor.children).filter((child)=>child instanceof HTMLElement && child.dataset.nodeId === id);
        if (0 === group.length) return false;
        const firstIndex = Array.prototype.indexOf.call(editor.childNodes, group[0]);
        for (const node of group)node.remove();
        this.setCaretByChildIndex(firstIndex);
        this.emitChange();
        return true;
    }
    handleClick = (event)=>{
        this.pendingDeleteNodeId = null;
        const target = event.target;
        if (target === this.getEditorEl()) {
            const sel = this.getSelection();
            if (!sel || 0 === sel.rangeCount || !target.contains(sel.anchorNode)) this.placeCaretAtEnd();
            return;
        }
        if (target.closest('.ve-ci-editable')) return;
        const slotTarget = target.closest('.ve-ci-island, .ve-ci-space');
        if (!slotTarget) return;
        const id = slotTarget.dataset.nodeId;
        if (!id) return;
        const editor = this.getEditorEl();
        if (!editor) return;
        const group = Array.from(editor.children).filter((child)=>child instanceof HTMLElement && child.dataset.nodeId === id);
        const beforeSpace = group.find((n)=>'before' === n.dataset.space);
        const island = group.find((n)=>n.classList.contains('ve-ci-island'));
        const afterSpace = group.find((n)=>'after' === n.dataset.space);
        if (!beforeSpace || !island || !afterSpace) return;
        const rect = island.getBoundingClientRect();
        if (event.clientX >= (rect.left + rect.right) / 2) {
            this.setCaretAfterChild(afterSpace);
            this.pendingDeleteNodeId = {
                id,
                side: 'after'
            };
        } else {
            this.setCaretBeforeChild(beforeSpace);
            this.pendingDeleteNodeId = {
                id,
                side: 'before'
            };
        }
    };
}
function serializeSegments(segments) {
    return JSON.stringify(segments.map((segment)=>'text' === segment.type ? [
            't',
            segment.text
        ] : [
            'n',
            segment.id,
            segment.nodeType,
            segment.data ?? {}
        ]));
}
function ve_chat_input_panel_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_chat_input_panel_element_toPropertyKey(arg) {
    var key = ve_chat_input_panel_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_chat_input_panel_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_chat_input_panel_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_chat_input_panel_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_chat_input_panel_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_chat_input_panel_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_chat_input_panel_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_chat_input_panel_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_chat_input_panel_element_dec, ve_chat_input_panel_element_dec1, ve_chat_input_panel_element_dec2, _init_trigger, _init_extra__init_trigger, _init_name, _init_extra__init_name, _init_placement, _init_extra__init_placement;
const VE_CHAT_INPUT_PANEL_TAG = 've-chat-input-panel';
ve_chat_input_panel_element_dec = property({
    type: String,
    reflect: true
}), ve_chat_input_panel_element_dec1 = property({
    type: String,
    reflect: true
}), ve_chat_input_panel_element_dec2 = property({
    type: String,
    reflect: true
});
class VeChatInputPanelElement extends VeElement {
    static{
        ({ e: [_init_trigger, _init_extra__init_trigger, _init_name, _init_extra__init_name, _init_placement, _init_extra__init_placement] } = ve_chat_input_panel_element_apply_decs_2311(this, [], [
            [
                ve_chat_input_panel_element_dec,
                1,
                "trigger"
            ],
            [
                ve_chat_input_panel_element_dec1,
                1,
                "name"
            ],
            [
                ve_chat_input_panel_element_dec2,
                1,
                "placement"
            ]
        ]));
    }
    static styles = css`
    :host {
      display: contents;
    }
  `;
    #___private_trigger_1 = (()=>{
        const _value = _init_trigger(this, '');
        _init_extra__init_trigger(this);
        return _value;
    })();
    get trigger() {
        return this.#___private_trigger_1;
    }
    set trigger(_v) {
        this.#___private_trigger_1 = _v;
    }
    #___private_name_2 = (()=>{
        const _value = _init_name(this, '');
        _init_extra__init_name(this);
        return _value;
    })();
    get name() {
        return this.#___private_name_2;
    }
    set name(_v) {
        this.#___private_name_2 = _v;
    }
    #___private_placement_3 = (()=>{
        const _value = _init_placement(this, 'top');
        _init_extra__init_placement(this);
        return _value;
    })();
    get placement() {
        return this.#___private_placement_3;
    }
    set placement(_v) {
        this.#___private_placement_3 = _v;
    }
    get panelKey() {
        return this.name || this.trigger;
    }
    connectedCallback() {
        super.connectedCallback();
        this.syncSlot();
        this.notifyParent();
    }
    updated(changed) {
        super.updated(changed);
        if (changed.has('trigger') || changed.has('name')) this.syncSlot();
        this.notifyParent();
    }
    syncSlot() {
        const key = this.panelKey;
        if (key) this.setAttribute('slot', `panel-${key}`);
        else this.removeAttribute('slot');
    }
    notifyParent() {
        this.dispatchEvent(new CustomEvent('ve-chat-input-panel-change', {
            bubbles: true,
            composed: true
        }));
    }
    render() {
        return html`<slot></slot>`;
    }
}
function ve_chat_input_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_chat_input_element_toPropertyKey(arg) {
    var key = ve_chat_input_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_chat_input_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_chat_input_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_chat_input_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_chat_input_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_chat_input_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_chat_input_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_chat_input_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_chat_input_element_dec, ve_chat_input_element_dec1, ve_chat_input_element_dec2, ve_chat_input_element_dec3, ve_chat_input_element_dec4, ve_chat_input_element_dec5, ve_chat_input_element_dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _init_value, _init_extra__init_value, ve_chat_input_element_init_placeholder, ve_chat_input_element_init_extra_init_placeholder, ve_chat_input_element_init_disabled, ve_chat_input_element_init_extra_init_disabled, _init_loading, _init_extra__init_loading, _init_size, _init_extra__init_size, _init_expandable, _init_extra__init_expandable, ve_chat_input_element_init_triggers, ve_chat_input_element_init_extra_init_triggers, _init_derivedTriggerPanels, _init_extra__init_derivedTriggerPanels, _init_derivedTriggers, _init_extra__init_derivedTriggers, ve_chat_input_element_init_nodeRenders, ve_chat_input_element_init_extra_init_nodeRenders, _init_borderGlow, _init_extra__init_borderGlow, _init_borderGlowColor, _init_extra__init_borderGlowColor, _init_borderGlowColors, _init_extra__init_borderGlowColors, _init_dataState, _init_extra__init_dataState, _init_normalized, _init_extra__init_normalized, _init_resolvedNodeRenders, _init_extra__init_resolvedNodeRenders, _init_isFocused, _init_extra__init_isFocused, ve_chat_input_element_init_expanded, ve_chat_input_element_init_extra_init_expanded, _init_hasHeader, _init_extra__init_hasHeader, _init_hasLeftAction, _init_extra__init_hasLeftAction, _init_hasFooter, _init_extra__init_hasFooter, _init_activePanelKey, _init_extra__init_activePanelKey, _init_editorEl, _init_extra__init_editorEl, _init_panelEl, _init_extra__init_panelEl;
const VE_CHAT_INPUT_TAG = 've-chat-input';
const BORDER_GLOW_GRADIENT_POSITIONS = [
    '80% 55%',
    '69% 34%',
    '8% 6%',
    '41% 38%',
    '86% 85%',
    '82% 18%',
    '51% 4%'
];
const BORDER_GLOW_GRADIENT_COLOR_MAP = [
    0,
    1,
    2,
    0,
    1,
    2,
    1
];
ve_chat_input_element_dec = property({
    attribute: false
}), ve_chat_input_element_dec1 = property({
    type: String
}), ve_chat_input_element_dec2 = property({
    type: Boolean,
    reflect: true
}), ve_chat_input_element_dec3 = property({
    type: Boolean,
    reflect: true
}), ve_chat_input_element_dec4 = property({
    type: String,
    reflect: true
}), ve_chat_input_element_dec5 = property({
    type: Boolean,
    reflect: true
}), ve_chat_input_element_dec6 = property({
    attribute: false
}), _dec7 = state(), _dec8 = state(), _dec9 = property({
    attribute: false
}), _dec10 = property({
    type: Boolean,
    reflect: true,
    attribute: 'border-glow'
}), _dec11 = property({
    type: String,
    attribute: 'border-glow-color'
}), _dec12 = property({
    type: String,
    attribute: 'border-glow-colors'
}), _dec13 = state(), _dec14 = state(), _dec15 = state(), _dec16 = state(), _dec17 = state(), _dec18 = state(), _dec19 = state(), _dec20 = state(), _dec21 = state(), _dec22 = decorators_js_query('ve-chat-input-editor'), _dec23 = decorators_js_query('.ve-chat-input-trigger-panel');
class VeChatInputElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, ve_chat_input_element_init_placeholder, ve_chat_input_element_init_extra_init_placeholder, ve_chat_input_element_init_disabled, ve_chat_input_element_init_extra_init_disabled, _init_loading, _init_extra__init_loading, _init_size, _init_extra__init_size, _init_expandable, _init_extra__init_expandable, ve_chat_input_element_init_triggers, ve_chat_input_element_init_extra_init_triggers, _init_derivedTriggerPanels, _init_extra__init_derivedTriggerPanels, _init_derivedTriggers, _init_extra__init_derivedTriggers, ve_chat_input_element_init_nodeRenders, ve_chat_input_element_init_extra_init_nodeRenders, _init_borderGlow, _init_extra__init_borderGlow, _init_borderGlowColor, _init_extra__init_borderGlowColor, _init_borderGlowColors, _init_extra__init_borderGlowColors, _init_dataState, _init_extra__init_dataState, _init_normalized, _init_extra__init_normalized, _init_resolvedNodeRenders, _init_extra__init_resolvedNodeRenders, _init_isFocused, _init_extra__init_isFocused, ve_chat_input_element_init_expanded, ve_chat_input_element_init_extra_init_expanded, _init_hasHeader, _init_extra__init_hasHeader, _init_hasLeftAction, _init_extra__init_hasLeftAction, _init_hasFooter, _init_extra__init_hasFooter, _init_activePanelKey, _init_extra__init_activePanelKey, _init_editorEl, _init_extra__init_editorEl, _init_panelEl, _init_extra__init_panelEl] } = ve_chat_input_element_apply_decs_2311(this, [], [
            [
                ve_chat_input_element_dec,
                1,
                "value"
            ],
            [
                ve_chat_input_element_dec1,
                1,
                "placeholder"
            ],
            [
                ve_chat_input_element_dec2,
                1,
                "disabled"
            ],
            [
                ve_chat_input_element_dec3,
                1,
                "loading"
            ],
            [
                ve_chat_input_element_dec4,
                1,
                "size"
            ],
            [
                ve_chat_input_element_dec5,
                1,
                "expandable"
            ],
            [
                ve_chat_input_element_dec6,
                1,
                "triggers"
            ],
            [
                _dec7,
                1,
                "derivedTriggerPanels"
            ],
            [
                _dec8,
                1,
                "derivedTriggers"
            ],
            [
                _dec9,
                1,
                "nodeRenders"
            ],
            [
                _dec10,
                1,
                "borderGlow"
            ],
            [
                _dec11,
                1,
                "borderGlowColor"
            ],
            [
                _dec12,
                1,
                "borderGlowColors"
            ],
            [
                _dec13,
                1,
                "dataState"
            ],
            [
                _dec14,
                1,
                "normalized"
            ],
            [
                _dec15,
                1,
                "resolvedNodeRenders"
            ],
            [
                _dec16,
                1,
                "isFocused"
            ],
            [
                _dec17,
                1,
                "expanded"
            ],
            [
                _dec18,
                1,
                "hasHeader"
            ],
            [
                _dec19,
                1,
                "hasLeftAction"
            ],
            [
                _dec20,
                1,
                "hasFooter"
            ],
            [
                _dec21,
                1,
                "activePanelKey"
            ],
            [
                _dec22,
                1,
                "editorEl"
            ],
            [
                _dec23,
                1,
                "panelEl"
            ]
        ]));
    }
    static styles = ve_chat_input;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement,
        [VE_CHAT_INPUT_EDITOR_TAG]: VeChatInputEditorElement,
        [VE_CHAT_INPUT_PANEL_TAG]: VeChatInputPanelElement
    };
    static{
        registerIcons([
            attachmentIcon,
            imageIcon,
            fileIcon,
            warningIcon
        ]);
    }
    #___private_value_1 = (()=>{
        const _value = _init_value(this, '');
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_placeholder_2 = (()=>{
        const _value = ve_chat_input_element_init_placeholder(this, '');
        ve_chat_input_element_init_extra_init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_2;
    }
    set placeholder(_v) {
        this.#___private_placeholder_2 = _v;
    }
    #___private_disabled_3 = (()=>{
        const _value = ve_chat_input_element_init_disabled(this, false);
        ve_chat_input_element_init_extra_init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_3;
    }
    set disabled(_v) {
        this.#___private_disabled_3 = _v;
    }
    #___private_loading_4 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_4;
    }
    set loading(_v) {
        this.#___private_loading_4 = _v;
    }
    #___private_size_5 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_5;
    }
    set size(_v) {
        this.#___private_size_5 = _v;
    }
    #___private_expandable_6 = (()=>{
        const _value = _init_expandable(this, false);
        _init_extra__init_expandable(this);
        return _value;
    })();
    get expandable() {
        return this.#___private_expandable_6;
    }
    set expandable(_v) {
        this.#___private_expandable_6 = _v;
    }
    #___private_triggers_7 = (()=>{
        const _value = ve_chat_input_element_init_triggers(this, []);
        ve_chat_input_element_init_extra_init_triggers(this);
        return _value;
    })();
    get triggers() {
        return this.#___private_triggers_7;
    }
    set triggers(_v) {
        this.#___private_triggers_7 = _v;
    }
    #___private_derivedTriggerPanels_8 = (()=>{
        const _value = _init_derivedTriggerPanels(this, []);
        _init_extra__init_derivedTriggerPanels(this);
        return _value;
    })();
    get derivedTriggerPanels() {
        return this.#___private_derivedTriggerPanels_8;
    }
    set derivedTriggerPanels(_v) {
        this.#___private_derivedTriggerPanels_8 = _v;
    }
    #___private_derivedTriggers_9 = (()=>{
        const _value = _init_derivedTriggers(this, []);
        _init_extra__init_derivedTriggers(this);
        return _value;
    })();
    get derivedTriggers() {
        return this.#___private_derivedTriggers_9;
    }
    set derivedTriggers(_v) {
        this.#___private_derivedTriggers_9 = _v;
    }
    #___private_nodeRenders_10 = (()=>{
        const _value = ve_chat_input_element_init_nodeRenders(this, void 0);
        ve_chat_input_element_init_extra_init_nodeRenders(this);
        return _value;
    })();
    get nodeRenders() {
        return this.#___private_nodeRenders_10;
    }
    set nodeRenders(_v) {
        this.#___private_nodeRenders_10 = _v;
    }
    #___private_borderGlow_11 = (()=>{
        const _value = _init_borderGlow(this, false);
        _init_extra__init_borderGlow(this);
        return _value;
    })();
    get borderGlow() {
        return this.#___private_borderGlow_11;
    }
    set borderGlow(_v) {
        this.#___private_borderGlow_11 = _v;
    }
    #___private_borderGlowColor_12 = (()=>{
        const _value = _init_borderGlowColor(this, '40 80 80');
        _init_extra__init_borderGlowColor(this);
        return _value;
    })();
    get borderGlowColor() {
        return this.#___private_borderGlowColor_12;
    }
    set borderGlowColor(_v) {
        this.#___private_borderGlowColor_12 = _v;
    }
    #___private_borderGlowColors_13 = (()=>{
        const _value = _init_borderGlowColors(this, '#c084fc,#f472b6,#38bdf8');
        _init_extra__init_borderGlowColors(this);
        return _value;
    })();
    get borderGlowColors() {
        return this.#___private_borderGlowColors_13;
    }
    set borderGlowColors(_v) {
        this.#___private_borderGlowColors_13 = _v;
    }
    #___private_dataState_14 = (()=>{
        const _value = _init_dataState(this, 'default');
        _init_extra__init_dataState(this);
        return _value;
    })();
    get dataState() {
        return this.#___private_dataState_14;
    }
    set dataState(_v) {
        this.#___private_dataState_14 = _v;
    }
    #___private_normalized_15 = (()=>{
        const _value = _init_normalized(this, {
            segments: [],
            attachments: []
        });
        _init_extra__init_normalized(this);
        return _value;
    })();
    get normalized() {
        return this.#___private_normalized_15;
    }
    set normalized(_v) {
        this.#___private_normalized_15 = _v;
    }
    #___private_resolvedNodeRenders_16 = (()=>{
        const _value = _init_resolvedNodeRenders(this, mergeChatInputNodeRenders());
        _init_extra__init_resolvedNodeRenders(this);
        return _value;
    })();
    get resolvedNodeRenders() {
        return this.#___private_resolvedNodeRenders_16;
    }
    set resolvedNodeRenders(_v) {
        this.#___private_resolvedNodeRenders_16 = _v;
    }
    #___private_isFocused_17 = (()=>{
        const _value = _init_isFocused(this, false);
        _init_extra__init_isFocused(this);
        return _value;
    })();
    get isFocused() {
        return this.#___private_isFocused_17;
    }
    set isFocused(_v) {
        this.#___private_isFocused_17 = _v;
    }
    #___private_expanded_18 = (()=>{
        const _value = ve_chat_input_element_init_expanded(this, false);
        ve_chat_input_element_init_extra_init_expanded(this);
        return _value;
    })();
    get expanded() {
        return this.#___private_expanded_18;
    }
    set expanded(_v) {
        this.#___private_expanded_18 = _v;
    }
    #___private_hasHeader_19 = (()=>{
        const _value = _init_hasHeader(this, false);
        _init_extra__init_hasHeader(this);
        return _value;
    })();
    get hasHeader() {
        return this.#___private_hasHeader_19;
    }
    set hasHeader(_v) {
        this.#___private_hasHeader_19 = _v;
    }
    #___private_hasLeftAction_20 = (()=>{
        const _value = _init_hasLeftAction(this, false);
        _init_extra__init_hasLeftAction(this);
        return _value;
    })();
    get hasLeftAction() {
        return this.#___private_hasLeftAction_20;
    }
    set hasLeftAction(_v) {
        this.#___private_hasLeftAction_20 = _v;
    }
    #___private_hasFooter_21 = (()=>{
        const _value = _init_hasFooter(this, false);
        _init_extra__init_hasFooter(this);
        return _value;
    })();
    get hasFooter() {
        return this.#___private_hasFooter_21;
    }
    set hasFooter(_v) {
        this.#___private_hasFooter_21 = _v;
    }
    #___private_activePanelKey_22 = (()=>{
        const _value = _init_activePanelKey(this, null);
        _init_extra__init_activePanelKey(this);
        return _value;
    })();
    get activePanelKey() {
        return this.#___private_activePanelKey_22;
    }
    set activePanelKey(_v) {
        this.#___private_activePanelKey_22 = _v;
    }
    #___private_editorEl_23 = (()=>{
        const _value = _init_editorEl(this, null);
        _init_extra__init_editorEl(this);
        return _value;
    })();
    get editorEl() {
        return this.#___private_editorEl_23;
    }
    set editorEl(_v) {
        this.#___private_editorEl_23 = _v;
    }
    #___private_panelEl_24 = (()=>{
        const _value = _init_panelEl(this, null);
        _init_extra__init_panelEl(this);
        return _value;
    })();
    get panelEl() {
        return this.#___private_panelEl_24;
    }
    set panelEl(_v) {
        this.#___private_panelEl_24 = _v;
    }
    panelAutoUpdate = null;
    get text() {
        return chatInputValueToText(this.normalized, (node)=>getChatInputNodeText(this.resolvedNodeRenders, node));
    }
    get isSendDisabled() {
        return isChatInputValueEmpty(this.normalized) || this.disabled;
    }
    focus(options) {
        this.updateComplete.then(async ()=>{
            const editor = this.editorEl;
            if (!editor) return;
            await editor.updateComplete;
            editor.focus(options);
        });
    }
    queryActiveTrigger() {
        return this.editorEl?.queryActiveTrigger() ?? null;
    }
    getCaretRect() {
        return this.editorEl?.getCaretRect() ?? null;
    }
    insertNode(node, options) {
        this.editorEl?.insertNode(node, options);
        this.closeTriggerPanel();
    }
    getChatInputInfo() {
        return {
            value: this.normalized,
            text: this.text,
            isFocus: this.isFocused,
            isLoading: this.loading,
            isExpanded: this.expanded
        };
    }
    panelObserver = null;
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('ve-chat-input-panel-change', this.handlePanelChange);
        this.panelObserver = new MutationObserver(()=>this.collectDerivedPanels());
        this.panelObserver.observe(this, {
            childList: true,
            subtree: true
        });
        this.collectDerivedPanels();
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('ve-chat-input-panel-change', this.handlePanelChange);
        this.panelObserver?.disconnect();
        this.panelObserver = null;
    }
    handlePanelChange = (event)=>{
        event.stopPropagation();
        this.collectDerivedPanels();
    };
    collectDerivedPanels() {
        const panels = Array.from(this.querySelectorAll(VE_CHAT_INPUT_PANEL_TAG)).filter((panel)=>panel.closest(VE_CHAT_INPUT_TAG) === this);
        const triggerPanels = [];
        const triggers = [];
        const seenTriggers = new Set();
        for (const panel of panels){
            const trigger = panel.trigger;
            const key = panel.panelKey;
            if (trigger && key) {
                triggerPanels.push({
                    key,
                    trigger,
                    placement: panel.placement
                });
                if (!seenTriggers.has(trigger)) {
                    seenTriggers.add(trigger);
                    triggers.push({
                        prefix: trigger
                    });
                }
            }
        }
        this.derivedTriggerPanels = triggerPanels;
        this.derivedTriggers = triggers;
    }
    willUpdate(changed) {
        if (changed.has('nodeRenders')) this.resolvedNodeRenders = mergeChatInputNodeRenders(this.nodeRenders);
        if (changed.has('value')) this.normalized = normalizeChatInputValue(this.value);
        if (changed.has('value') || changed.has('loading') || changed.has('disabled') || changed.has('normalized') || 'default' === this.dataState) this.dataState = this.computeDataState();
    }
    updated(changed) {
        super.updated(changed);
        if (changed.has('borderGlowColor') || changed.has('borderGlowColors')) this.syncBorderGlowVars();
    }
    syncBorderGlowVars() {
        const colors = this.parseBorderGlowColors(this.borderGlowColors);
        for(let i = 0; i < BORDER_GLOW_GRADIENT_POSITIONS.length; i += 1){
            const colorIndex = BORDER_GLOW_GRADIENT_COLOR_MAP[i] ?? 0;
            const color = colors[Math.min(colorIndex, colors.length - 1)] ?? colors[0];
            this.style.setProperty(`--ve-chat-input-border-glow-gradient-${i + 1}`, `radial-gradient(at ${BORDER_GLOW_GRADIENT_POSITIONS[i]}, ${color} 0px, transparent 50%)`);
        }
        this.style.setProperty('--ve-chat-input-border-glow-gradient-base', `linear-gradient(${colors[0]} 0 100%)`);
        const glowBase = this.normalizeBorderGlowHsl(this.borderGlowColor);
        const opacities = [
            100,
            60,
            50,
            40,
            30,
            20,
            10
        ];
        const suffixes = [
            '',
            '-60',
            '-50',
            '-40',
            '-30',
            '-20',
            '-10'
        ];
        for(let i = 0; i < opacities.length; i += 1)this.style.setProperty(`--ve-chat-input-border-glow-color${suffixes[i]}`, `hsl(${glowBase} / ${opacities[i]}%)`);
    }
    parseBorderGlowColors(value) {
        const colors = value.split(/[,\|]/).map((color)=>color.trim()).filter(Boolean);
        return colors.length > 0 ? colors : [
            '#c084fc',
            '#f472b6',
            '#38bdf8'
        ];
    }
    normalizeBorderGlowHsl(value) {
        const fallback = '40deg 80% 80%';
        const match = value.match(/([\d.]+)(deg)?\s+([\d.]+)%?\s+([\d.]+)%?/);
        if (!match) return fallback;
        const hue = match[1] ?? '40';
        const saturation = match[3] ?? '80';
        const lightness = match[4] ?? '80';
        return `${hue}deg ${saturation}% ${lightness}%`;
    }
    handleBorderGlowPointerMove = (event)=>{
        if (!this.borderGlow) return;
        const wrapper = event.currentTarget;
        if (!(wrapper instanceof HTMLElement)) return;
        const rect = wrapper.getBoundingClientRect();
        const x = event.clientX - rect.left;
        const y = event.clientY - rect.top;
        const centerX = rect.width / 2;
        const centerY = rect.height / 2;
        const dx = x - centerX;
        const dy = y - centerY;
        const kx = 0 === dx ? 1 / 0 : centerX / Math.abs(dx);
        const ky = 0 === dy ? 1 / 0 : centerY / Math.abs(dy);
        const edge = Math.min(Math.max(1 / Math.min(kx, ky), 0), 1);
        const radians = 0 === dx && 0 === dy ? 0 : Math.atan2(dy, dx);
        let angle = 180 / Math.PI * radians + 90;
        if (angle < 0) angle += 360;
        wrapper.style.setProperty('--ve-chat-input-border-glow-edge-proximity', `${(100 * edge).toFixed(3)}`);
        wrapper.style.setProperty('--ve-chat-input-border-glow-cursor-angle', `${angle.toFixed(3)}deg`);
    };
    computeDataState() {
        if (this.loading) return 'generating';
        const hasValue = !isChatInputValueEmpty(this.normalized);
        if (this.isFocused) return hasValue ? 'typing' : 'focused';
        return hasValue ? 'done' : 'default';
    }
    commitValue(segments) {
        const next = {
            segments,
            attachments: this.normalized.attachments
        };
        this.normalized = next;
        this.value = next;
        this.dataState = this.computeDataState();
        this.dispatchEvent(new CustomEvent('ve-chat-input-change', {
            detail: {
                value: this.text,
                data: next
            },
            bubbles: true,
            composed: true
        }));
    }
    handleEditorChange = (event)=>{
        event.stopPropagation();
        this.commitValue(event.detail.segments);
    };
    handleEditorTrigger = (event)=>{
        event.stopPropagation();
        this.syncTriggerPanel(event.detail);
        this.dispatchEvent(new CustomEvent('ve-chat-input-trigger', {
            detail: event.detail,
            bubbles: true,
            composed: true
        }));
    };
    handleEditorSubmit = ()=>{
        const pressEnterEvent = new CustomEvent('ve-chat-input-press-enter', {
            bubbles: true,
            composed: true,
            cancelable: true,
            detail: {
                originalEvent: null
            }
        });
        const continueDefault = this.dispatchEvent(pressEnterEvent);
        if (!continueDefault) return;
        if (!this.isSendDisabled && !this.loading) this.submit();
    };
    syncTriggerPanel(detail) {
        if (0 === this.derivedTriggerPanels.length) return;
        const prefix = detail.trigger?.prefix ?? null;
        const panel = prefix ? this.derivedTriggerPanels.find((item)=>item.trigger === prefix) : void 0;
        if (!panel) return void this.closeTriggerPanel();
        this.activePanelKey = panel.key;
        this.updateComplete.then(()=>this.positionTriggerPanel(panel));
    }
    closeTriggerPanel() {
        if (null === this.activePanelKey) return;
        this.activePanelKey = null;
        this.panelAutoUpdate?.();
        this.panelAutoUpdate = null;
        hidePanelPopover(this.panelEl);
    }
    positionTriggerPanel(panel) {
        const floating = this.panelEl;
        if (!floating) return;
        const placement = panel.placement ?? 'top';
        showPanelPopover(floating);
        if ('caret' === placement) {
            this.panelAutoUpdate?.();
            this.panelAutoUpdate = null;
            floating.style.width = '';
            const rect = this.getCaretRect();
            if (!rect) return;
            const reference = this.makeCaretReference(rect);
            computeFloatingPanelPosition({
                reference,
                floating,
                placement: 'bottom-start',
                distance: 6
            }).then((result)=>{
                floating.style.left = `${result.x}px`;
                floating.style.top = `${result.y}px`;
            });
            return;
        }
        const wrapper = this.shadowRoot?.querySelector('.ve-chat-input-wrapper');
        if (!wrapper) return;
        const update = ()=>{
            floating.style.width = `${wrapper.getBoundingClientRect().width}px`;
            computeFloatingPanelPosition({
                reference: wrapper,
                floating,
                placement: 'top-start',
                distance: 8
            }).then((result)=>{
                floating.style.left = `${result.x}px`;
                floating.style.top = `${result.y}px`;
            });
        };
        this.panelAutoUpdate?.();
        this.panelAutoUpdate = createFloatingAutoUpdate(wrapper, floating, update);
    }
    makeCaretReference(rect) {
        return {
            getBoundingClientRect: ()=>rect
        };
    }
    handleFocus = ()=>{
        this.isFocused = true;
        this.dataState = this.computeDataState();
        this.dispatchEvent(new CustomEvent('ve-chat-input-focus', {
            bubbles: true,
            composed: true
        }));
    };
    handleBlur = ()=>{
        this.isFocused = false;
        this.dataState = this.computeDataState();
        this.dispatchEvent(new CustomEvent('ve-chat-input-blur', {
            bubbles: true,
            composed: true
        }));
    };
    submit() {
        if (this.isSendDisabled || this.loading) return;
        this.dispatchEvent(new CustomEvent('ve-chat-input-submit', {
            detail: {
                value: this.text,
                data: this.normalized
            },
            bubbles: true,
            composed: true
        }));
    }
    handleSendClick = ()=>{
        if (this.isSendDisabled) return;
        this.submit();
    };
    handleStopClick = ()=>{
        if (this.disabled) return;
        this.dispatchEvent(new CustomEvent('ve-chat-input-cancel', {
            bubbles: true,
            composed: true
        }));
    };
    handleAttachmentRemove(attachment) {
        if (this.disabled) return;
        const attachments = this.normalized.attachments.filter((item)=>item.id !== attachment.id);
        const next = {
            segments: this.normalized.segments,
            attachments
        };
        this.normalized = next;
        this.value = next;
        this.dataState = this.computeDataState();
        this.dispatchEvent(new CustomEvent('ve-chat-input-attachment-remove', {
            detail: {
                id: attachment.id,
                attachment
            },
            bubbles: true,
            composed: true
        }));
        this.dispatchEvent(new CustomEvent('ve-chat-input-change', {
            detail: {
                value: this.text,
                data: next
            },
            bubbles: true,
            composed: true
        }));
    }
    handleExpandClick = ()=>{
        this.expanded = !this.expanded;
        this.dispatchEvent(new CustomEvent('ve-chat-input-expand-change', {
            detail: {
                expanded: this.expanded
            },
            bubbles: true,
            composed: true
        }));
    };
    handleHeaderSlotChange = (event)=>{
        const slot = event.target;
        this.hasHeader = this.hasAssignedContent(slot);
    };
    handleLeftActionSlotChange = (event)=>{
        this.hasLeftAction = this.hasAssignedContent(event.target);
    };
    handleFooterSlotChange = (event)=>{
        this.hasFooter = this.hasAssignedContent(event.target);
    };
    hasAssignedContent(slot) {
        return slot.assignedNodes({
            flatten: true
        }).some((node)=>{
            if (node.nodeType === Node.ELEMENT_NODE) return true;
            if (node.nodeType === Node.TEXT_NODE) return (node.textContent ?? '').trim().length > 0;
            return false;
        });
    }
    renderExpandButton() {
        if (!this.expandable) return nothing;
        return html`
      <button
        class="ve-chat-input-expand-btn"
        part="expand-btn"
        type="button"
        aria-label=${this.expanded ? '退出全屏' : '全屏'}
        @click=${this.handleExpandClick}
      >
        ${this.expanded ? html`<ve-icon .definition=${exitFullScreenIcon} size="14"></ve-icon>` : html`<ve-icon .definition=${fullScreenIcon} size="14"></ve-icon>`}
      </button>
    `;
    }
    attachmentIconName(attachment) {
        if ('image' === attachment.kind) return 'image';
        if ('video' === attachment.kind) return 'play-square';
        if ('file' === attachment.kind) return 'file';
        return 'attachment';
    }
    renderAttachments() {
        const attachments = this.normalized.attachments;
        if (0 === attachments.length) return nothing;
        return html`
      <div class="ve-chat-input-attachments" part="attachments">
        ${repeat(attachments, (attachment)=>attachment.id, (attachment)=>this.renderAttachmentChip(attachment))}
      </div>
    `;
    }
    renderAttachmentChip(attachment) {
        const uploading = 'uploading' === attachment.status;
        const error = 'error' === attachment.status;
        const previewUrl = attachment.thumbUrl || attachment.url;
        const isImage = 'image' === attachment.kind && Boolean(previewUrl);
        return html`
      <div
        class=${classMap({
            've-chat-input-attachment': true,
            've-chat-input-attachment-image': isImage,
            've-chat-input-attachment-file': !isImage,
            'is-uploading': uploading,
            'is-error': error
        })}
        title=${attachment.name}
      >
        ${isImage ? html`
              <div class="ve-chat-input-attachment-image-wrapper">
                <img
                  class="ve-chat-input-attachment-image-content"
                  src=${previewUrl}
                  alt=${attachment.name}
                />
                ${uploading ? html`<div class="ve-chat-input-attachment-progress-overlay">
                      <span class="ve-chat-input-attachment-progress-text"
                        >${attachment.percent ?? 0}%</span
                      >
                    </div>` : nothing}
                ${error ? html`<div class="ve-chat-input-attachment-error-meta">
                      <ve-icon
                        class="ve-chat-input-attachment-error-icon"
                        .definition=${warningIcon}
                        size="14"
                      ></ve-icon>
                      <span>上传失败</span>
                    </div>` : nothing}
              </div>` : html`
              <div class="ve-chat-input-attachment-file-wrapper">
                <div class="ve-chat-input-attachment-file-title">
                  ${attachment.name}
                </div>
                <div class="ve-chat-input-attachment-file-footer">
                  ${uploading ? html`<span class="ve-chat-input-attachment-progress-text"
                        >${attachment.percent ?? 0}%</span
                      >` : error ? html`<span class="ve-chat-input-attachment-error-meta">
                          <ve-icon
                            class="ve-chat-input-attachment-error-icon"
                            .definition=${warningIcon}
                            size="14"
                          ></ve-icon>
                          <span>上传失败</span>
                        </span>` : html`<ve-icon
                          class="ve-chat-input-attachment-file-icon"
                          name=${this.attachmentIconName(attachment)}
                          size="18"
                        ></ve-icon>`}
                </div>
              </div>`}
        <span
          class="ve-chat-input-attachment-remove"
          role="button"
          tabindex="0"
          aria-label="移除附件"
          @click=${()=>this.handleAttachmentRemove(attachment)}
          @keydown=${(event)=>{
            if ('Enter' === event.key || ' ' === event.key) {
                event.preventDefault();
                this.handleAttachmentRemove(attachment);
            }
        }}
        >
          <ve-icon .definition=${closeIcon} size="12"></ve-icon>
        </span>
      </div>
    `;
    }
    renderTriggerPanels() {
        const key = this.activePanelKey;
        if (null === key) return nothing;
        return html`
      <div
        class="ve-chat-input-trigger-panel"
        part="trigger-panel"
        popover="manual"
        role="listbox"
      >
        <slot name="panel-${key}"></slot>
      </div>
    `;
    }
    get effectiveTriggers() {
        if (0 === this.derivedTriggers.length) return this.triggers;
        return [
            ...this.triggers,
            ...this.derivedTriggers
        ];
    }
    renderEditor() {
        return html`
      <ve-chat-input-editor
        part="editor"
        .segments=${this.normalized.segments}
        .placeholder=${this.placeholder}
        .triggers=${this.effectiveTriggers}
        .nodeRenders=${this.resolvedNodeRenders}
        ?disabled=${this.disabled}
        ?expanded=${this.expanded}
        @ve-ci-editor-change=${this.handleEditorChange}
        @ve-ci-editor-submit=${this.handleEditorSubmit}
        @ve-ci-editor-trigger=${this.handleEditorTrigger}
        @ve-ci-editor-focus=${this.handleFocus}
        @ve-ci-editor-blur=${this.handleBlur}
      ></ve-chat-input-editor>
    `;
    }
    renderToolbar() {
        if (this.hasFooter) return html`
        <div class="ve-chat-input-toolbar" part="toolbar">
          <slot name="footer" @slotchange=${this.handleFooterSlotChange}></slot>
        </div>
      `;
        const defaultRight = this.loading ? html`
          <button
            class="ve-chat-input-stop"
            part="stop"
            type="button"
            aria-label="停止生成"
            ?disabled=${this.disabled}
            @click=${this.handleStopClick}
          >
            <span class="ve-chat-input-stop-icon"></span>
          </button>
        ` : html`
          <button
            class="ve-chat-input-send"
            part="send"
            type="button"
            aria-label="发送消息"
            ?disabled=${this.isSendDisabled}
            @click=${this.handleSendClick}
          >
            ${html`<ve-icon .definition=${arrowUpMdIcon} size="16"></ve-icon>`}
          </button>
        `;
        return html`
      <div class="ve-chat-input-toolbar" part="toolbar">
        <div class="ve-chat-input-actions" part="actions">
          <slot
            name="left-action"
            @slotchange=${this.handleLeftActionSlotChange}
          ></slot>
        </div>
        <div class="ve-chat-input-submit" part="submit">
          <slot name="right-action"></slot>
          ${defaultRight}
          <slot
            name="footer"
            hidden
            @slotchange=${this.handleFooterSlotChange}
          ></slot>
        </div>
      </div>
    `;
    }
    render() {
        return html`
      <div class="ve-chat-input-container" part="container">
        <div
          class=${classMap({
            've-chat-input-wrapper': true,
            'is-disabled': this.disabled,
            'is-loading': this.loading,
            'is-expanded': this.expanded
        })}
          data-state=${this.dataState}
          part="wrapper"
          @pointermove=${this.handleBorderGlowPointerMove}
        >
          <span class="ve-chat-input-border-glow-edge" aria-hidden="true"></span>
          <div class="ve-chat-input-box" part="box">
            <div
              class=${classMap({
            've-chat-input-header': true,
            've-chat-input-header-empty': !this.hasHeader
        })}
              part="header"
            >
              <slot
                name="header"
                @slotchange=${this.handleHeaderSlotChange}
              ></slot>
            </div>
            ${this.renderExpandButton()}
            ${this.renderAttachments()}
            <div class="ve-chat-input-body" part="body">
              ${this.renderEditor()}
            </div>
            ${this.renderToolbar()}
          </div>
          ${this.renderTriggerPanels()}
        </div>
      </div>
    `;
    }
}
const ve_chat_input_action = [
    componentReset,
    css`
    :host {
      display: inline-flex;
      align-items: stretch;
      box-sizing: border-box;
      vertical-align: middle;
    }

    /* delegatesFocus 会在 host 上按 :focus / :focus-visible 绘制 UA 默认焦点环
       （如下拉选完后焦点程序化回到触发器时），键盘可达性由内部按钮的
       :focus-visible 蓝色描边保证，故移除 host 焦点环。 */
    :host(:focus),
    :host(:focus-visible) {
      outline: none;
    }

    /* 与 ve-composer 工具栏 action 风格对齐：圆角按钮、active 切到 secondary 底色 */
    .ve-chat-input-action {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: var(--ve-chat-input-action-gap-inner, 4px);
      block-size: var(--ve-chat-input-action-height, 32px);
      padding: 0 var(--ve-chat-input-action-padding-inline, 12px);
      border: none;
      border-radius: var(--ve-chat-input-action-radius, 8px);
      background: var(--ve-chat-input-action-bg, transparent);
      color: var(--ve-chat-input-action-text, #1d2129);
      font-family: inherit;
      font-size: var(--text-body-sm, 13px);
      line-height: 1;
      cursor: pointer;
      white-space: nowrap;
      transition:
        background-color var(--motion-fast, 160ms) var(--ease-standard, ease),
        color var(--motion-fast, 160ms) var(--ease-standard, ease);
      -webkit-appearance: none;
      appearance: none;
    }

    :host([size='small']) .ve-chat-input-action {
      block-size: 24px;
      padding: 0 8px;
      font-size: var(--text-caption, 12px);
    }

    :host([icon-only]) .ve-chat-input-action {
      padding: 0;
      inline-size: var(--ve-chat-input-action-height, 32px);
      border-radius: var(--ve-chat-input-action-icon-only-radius, 12px);
    }

    :host([size='small'][icon-only]) .ve-chat-input-action {
      inline-size: 24px;
      border-radius: var(--radius-md, 8px);
    }

    :host([icon-only]) .ve-chat-input-action-label {
      display: none;
    }

    /* hover：与 ve-composer __stop 同款 subtle 混色 */
    .ve-chat-input-action:hover:not(:disabled) {
      background: color-mix(
        in srgb,
        var(--color-bg-base, #fff) var(--state-hover-subtle-ratio, 97%),
        var(--state-hover-subtle-mix, #000)
      );
    }

    :host([icon-only]) .ve-chat-input-action:hover:not(:disabled) {
      background: color-mix(
        in srgb,
        var(--ve-chat-input-action-icon-only-bg, #f7f8fa)
          var(--state-hover-subtle-ratio, 97%),
        var(--state-hover-subtle-mix, #000)
      );
    }

    /* 不绘制焦点环：delegatesFocus + 程序化 restoreTriggerFocus 会让浏览器
       在内部 button 上绘制 UA 默认/focus-visible 环（表现为黑色方框），
       这里彻底清除 :focus / :focus-visible 的 outline。 */
    .ve-chat-input-action:focus,
    .ve-chat-input-action:focus-visible {
      outline: none;
    }

    .ve-chat-input-action:disabled {
      opacity: var(--state-disabled-ratio, 0.4);
      cursor: not-allowed;
    }

    /* active：透明边框 + secondary 底色（与 sender-spec / composer toolbar action 一致） */
    :host([status='active']) .ve-chat-input-action {
      border-color: transparent;
      background: var(--ve-chat-input-action-active-bg, #f7f8fa);
      color: var(--ve-chat-input-action-active-text, #1d2129);
    }

    :host([status='active']) .ve-chat-input-action:hover:not(:disabled) {
      background: color-mix(
        in srgb,
        var(--ve-chat-input-action-active-bg, #f7f8fa)
          var(--state-hover-subtle-ratio, 97%),
        var(--state-hover-subtle-mix, #000)
      );
    }

    .ve-chat-input-action-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: 16px;
      block-size: 16px;
      flex: none;
    }

    .ve-chat-input-action-icon[hidden] {
      display: none;
    }

    .ve-chat-input-action-arrow {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: 12px;
      block-size: 12px;
      flex: none;
      transition: transform var(--motion-fast, 160ms)
        var(--ease-standard, ease);
    }

    :host([data-open='true']) .ve-chat-input-action-arrow {
      transform: rotate(180deg);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-chat-input-action,
      .ve-chat-input-action-arrow {
        transition: none;
      }
    }
  `
];
function ve_chat_input_action_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_chat_input_action_element_toPropertyKey(arg) {
    var key = ve_chat_input_action_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_chat_input_action_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_chat_input_action_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_chat_input_action_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_chat_input_action_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_chat_input_action_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_chat_input_action_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_chat_input_action_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_chat_input_action_element_dec, ve_chat_input_action_element_dec1, ve_chat_input_action_element_dec2, ve_chat_input_action_element_dec3, ve_chat_input_action_element_dec4, ve_chat_input_action_element_dec5, ve_chat_input_action_element_dec6, _init_type, _init_extra__init_type, _init_status, _init_extra__init_status, ve_chat_input_action_element_init_size, ve_chat_input_action_element_init_extra_init_size, ve_chat_input_action_element_init_disabled, ve_chat_input_action_element_init_extra_init_disabled, _init_open, _init_extra__init_open, _init_iconOnly, _init_extra__init_iconOnly, _init_hasIconSlot, _init_extra__init_hasIconSlot;
const VE_CHAT_INPUT_ACTION_TAG = 've-chat-input-action';
const arrowDownIcon = html`
  <svg
    viewBox="0 0 12 12"
    width="12"
    height="12"
    fill="none"
    aria-hidden="true"
  >
    <path
      d="m3 4.5 3 3 3-3"
      stroke="currentColor"
      stroke-width="1.4"
      stroke-linecap="round"
      stroke-linejoin="round"
    ></path>
  </svg>
`;
ve_chat_input_action_element_dec = property({
    type: String,
    reflect: true
}), ve_chat_input_action_element_dec1 = property({
    type: String,
    reflect: true
}), ve_chat_input_action_element_dec2 = property({
    type: String,
    reflect: true
}), ve_chat_input_action_element_dec3 = property({
    type: Boolean,
    reflect: true
}), ve_chat_input_action_element_dec4 = property({
    type: Boolean,
    reflect: true,
    attribute: 'data-open'
}), ve_chat_input_action_element_dec5 = property({
    type: Boolean,
    reflect: true,
    attribute: 'icon-only'
}), ve_chat_input_action_element_dec6 = state();
class VeChatInputActionElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_status, _init_extra__init_status, ve_chat_input_action_element_init_size, ve_chat_input_action_element_init_extra_init_size, ve_chat_input_action_element_init_disabled, ve_chat_input_action_element_init_extra_init_disabled, _init_open, _init_extra__init_open, _init_iconOnly, _init_extra__init_iconOnly, _init_hasIconSlot, _init_extra__init_hasIconSlot] } = ve_chat_input_action_element_apply_decs_2311(this, [], [
            [
                ve_chat_input_action_element_dec,
                1,
                "type"
            ],
            [
                ve_chat_input_action_element_dec1,
                1,
                "status"
            ],
            [
                ve_chat_input_action_element_dec2,
                1,
                "size"
            ],
            [
                ve_chat_input_action_element_dec3,
                1,
                "disabled"
            ],
            [
                ve_chat_input_action_element_dec4,
                1,
                "open"
            ],
            [
                ve_chat_input_action_element_dec5,
                1,
                "iconOnly"
            ],
            [
                ve_chat_input_action_element_dec6,
                1,
                "hasIconSlot"
            ]
        ]));
    }
    static styles = ve_chat_input_action;
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'button');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_status_2 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_2;
    }
    set status(_v) {
        this.#___private_status_2 = _v;
    }
    #___private_size_3 = (()=>{
        const _value = ve_chat_input_action_element_init_size(this, 'default');
        ve_chat_input_action_element_init_extra_init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_3;
    }
    set size(_v) {
        this.#___private_size_3 = _v;
    }
    #___private_disabled_4 = (()=>{
        const _value = ve_chat_input_action_element_init_disabled(this, false);
        ve_chat_input_action_element_init_extra_init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_4;
    }
    set disabled(_v) {
        this.#___private_disabled_4 = _v;
    }
    #___private_open_5 = (()=>{
        const _value = _init_open(this, false);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_5;
    }
    set open(_v) {
        this.#___private_open_5 = _v;
    }
    #___private_iconOnly_6 = (()=>{
        const _value = _init_iconOnly(this, false);
        _init_extra__init_iconOnly(this);
        return _value;
    })();
    get iconOnly() {
        return this.#___private_iconOnly_6;
    }
    set iconOnly(_v) {
        this.#___private_iconOnly_6 = _v;
    }
    #___private_hasIconSlot_7 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_7;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_7 = _v;
    }
    handleIconSlotChange = (event)=>{
        const slot = event.target;
        this.hasIconSlot = slot.assignedNodes({
            flatten: true
        }).some((node)=>{
            if (node.nodeType === Node.ELEMENT_NODE) return true;
            if (node.nodeType === Node.TEXT_NODE) return (node.textContent ?? '').trim().length > 0;
            return false;
        });
    };
    handleClick = (event)=>{
        if (this.disabled) {
            event.stopImmediatePropagation();
            event.preventDefault();
            return;
        }
        if ('dropdown' === this.type) this.open = !this.open;
        this.dispatchEvent(new CustomEvent('ve-chat-input-action-click', {
            bubbles: true,
            composed: true,
            detail: {
                type: this.type,
                open: this.open
            }
        }));
    };
    renderArrow() {
        if ('dropdown' !== this.type) return nothing;
        return html`<span class="ve-chat-input-action-arrow" part="arrow"
      >${arrowDownIcon}</span
    >`;
    }
    render() {
        return html`
      <button
        class="ve-chat-input-action"
        part="button"
        type="button"
        ?disabled=${this.disabled}
        aria-haspopup=${'dropdown' === this.type ? 'menu' : 'false'}
        aria-expanded=${'dropdown' === this.type ? String(this.open) : 'false'}
        @click=${this.handleClick}
      >
        <span
          class="ve-chat-input-action-icon"
          part="icon"
          ?hidden=${!this.hasIconSlot}
        >
          <slot name="icon" @slotchange=${this.handleIconSlotChange}></slot>
        </span>
        <span class="ve-chat-input-action-label" part="label">
          <slot></slot>
        </span>
        ${this.renderArrow()}
      </button>
    `;
    }
}
defineCustomElement(VE_CHAT_INPUT_TAG, VeChatInputElement);
defineCustomElement(VE_CHAT_INPUT_ACTION_TAG, VeChatInputActionElement);
defineCustomElement(VE_CHAT_INPUT_EDITOR_TAG, VeChatInputEditorElement);
defineCustomElement(VE_CHAT_INPUT_PANEL_TAG, VeChatInputPanelElement);
export { DEFAULT_CHAT_INPUT_NODE_RENDERS, VE_CHAT_INPUT_ACTION_TAG, VE_CHAT_INPUT_EDITOR_TAG, VE_CHAT_INPUT_PANEL_TAG, VE_CHAT_INPUT_TAG, VeChatInputActionElement, VeChatInputEditorElement, VeChatInputElement, VeChatInputPanelElement, chatInputValueToText, defaultNodeText, getChatInputNodeText, isChatInputNodeEditable, isChatInputValue, isChatInputValueEmpty, mergeChatInputNodeRenders, nextNodeId as createChatInputNodeId, normalizeChatInputValue, resolveChatInputNodeRender };
