// @ve-design/web/web/dist/ve-dropdown.js
import { css, html } from "lit";
import { property, query, state } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { getAssignedMenuNodes, VE_DROPDOWN_ITEM_TAG, VeDropdownItemElement, VeDropdownElement, VE_DROPDOWN_TAG, hasAssignedSlotContent, focusElement, token } from "./97605.js";
import { createFloatingAutoUpdate, getFloatingDistance, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
import { VE_DROPDOWN_GROUP_TAG, VeDropdownGroupElement } from "./16491.js";
const ve_dropdown_divider = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      block-size: var(--ve-dropdown-divider-width);
      border: none;
      background: var(--ve-dropdown-divider-color);
      margin-block: var(--ve-dropdown-divider-margin-block);
      margin-inline: var(--ve-dropdown-divider-margin-inline);
    }

    :host([hidden]) {
      display: none;
    }
  `
];
const VE_DROPDOWN_DIVIDER_TAG = 've-dropdown-divider';
class VeDropdownDividerElement extends VeElement {
    static styles = ve_dropdown_divider;
    connectedCallback() {
        super.connectedCallback();
        this.setAttribute('role', 'separator');
        this.setAttribute('aria-orientation', 'horizontal');
    }
}
const ve_dropdown_sub_item = [
    componentReset,
    token,
    css`
    :host {
      position: relative;
      display: block;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-dropdown-sub-item-trigger {
      appearance: none;
      display: flex;
      align-items: center;
      justify-content: space-between;
      inline-size: 100%;
      min-block-size: var(--ve-dropdown-item-min-height);
      box-sizing: border-box;
      padding-block: var(--ve-dropdown-item-padding-block);
      padding-inline: var(--ve-dropdown-item-padding-inline);
      gap: var(--ve-dropdown-item-gap);
      border: none;
      border-radius: var(--ve-dropdown-item-radius);
      background: transparent;
      color: var(--ve-dropdown-item-color);
      font-family: var(--font-sans);
      font-size: var(--ve-dropdown-item-font-size);
      font-weight: var(--ve-dropdown-item-font-weight);
      line-height: var(--ve-dropdown-item-label-line-height);
      text-align: left;
      cursor: pointer;
      transition:
        background-color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        color var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing);
    }

    .ve-dropdown-sub-item-trigger:hover,
    :host([open]) .ve-dropdown-sub-item-trigger {
      background: var(--ve-dropdown-item-hover-bg);
    }

    .ve-dropdown-sub-item-trigger:active {
      background: var(--ve-dropdown-item-active-bg);
    }

    .ve-dropdown-sub-item-trigger:focus-visible {
      outline: none;
      box-shadow: var(--ve-dropdown-item-focus-shadow);
    }

    :host([disabled]) .ve-dropdown-sub-item-trigger,
    :host([disabled]) .ve-dropdown-sub-item-trigger:hover,
    :host([disabled]) .ve-dropdown-sub-item-trigger:focus-visible,
    :host([disabled]) .ve-dropdown-sub-item-trigger:active {
      color: var(--ve-dropdown-item-disabled-color);
      background: transparent;
      cursor: not-allowed;
      opacity: var(--ve-dropdown-disabled-opacity);
      pointer-events: none;
    }

    .ve-dropdown-sub-item-label {
      min-inline-size: 0;
      display: block;
      flex: 1;
      overflow: hidden;
      text-align: left;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-dropdown-sub-item-arrow {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-dropdown-sub-item-arrow-size);
      block-size: var(--ve-dropdown-sub-item-arrow-size);
      flex: none;
      color: var(--ve-dropdown-sub-item-arrow-color);
      margin-inline-start: var(--ve-dropdown-item-extra-margin-inline-start);
      transform: translateY(var(--ve-dropdown-sub-item-arrow-offset-y));
    }

    .ve-dropdown-sub-item-arrow::before {
      content: '';
      inline-size: var(--ve-dropdown-sub-item-arrow-icon-size);
      block-size: var(--ve-dropdown-sub-item-arrow-icon-size);
      border-block-start: var(--ve-dropdown-sub-item-arrow-stroke) solid
        currentColor;
      border-inline-end: var(--ve-dropdown-sub-item-arrow-stroke) solid
        currentColor;
      transform: rotate(45deg);
    }

    .ve-dropdown-sub-item-panel {
      position: fixed;
      inset: auto;
      margin: 0;
      padding: 0;
      z-index: var(--ve-dropdown-z-index);
      box-sizing: border-box;
      min-inline-size: var(--ve-dropdown-min-width);
      max-inline-size: var(--ve-dropdown-max-width);
      border: var(--ve-dropdown-border-width) solid
        var(--ve-dropdown-border-color);
      border-radius: var(--ve-dropdown-radius);
      background: var(--ve-dropdown-background);
      box-shadow: var(--ve-dropdown-shadow);
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
      transform: scaleY(0.9)
        translate3d(calc(var(--ve-dropdown-enter-offset) * -1), 0, 0);
      transform-origin: left top;
      transition:
        opacity var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        transform var(--ve-dropdown-transition-duration)
          var(--ve-dropdown-transition-easing),
        visibility 0s linear var(--ve-dropdown-transition-duration);
    }

    .ve-dropdown-sub-item-panel.ve-dropdown-sub-item-placement-left,
    .ve-dropdown-sub-item-panel.ve-dropdown-sub-item-placement-left-start,
    .ve-dropdown-sub-item-panel.ve-dropdown-sub-item-placement-left-end {
      transform: scaleY(0.9) translate3d(var(--ve-dropdown-enter-offset), 0, 0);
      transform-origin: right top;
    }

    .ve-dropdown-sub-item-panel.ve-dropdown-sub-item-placement-right,
    .ve-dropdown-sub-item-panel.ve-dropdown-sub-item-placement-right-start,
    .ve-dropdown-sub-item-panel.ve-dropdown-sub-item-placement-right-end {
      transform: scaleY(0.9)
        translate3d(calc(var(--ve-dropdown-enter-offset) * -1), 0, 0);
      transform-origin: left top;
    }

    :host([open]) .ve-dropdown-sub-item-panel {
      opacity: 1;
      visibility: visible;
      pointer-events: auto;
      transform: none;
      transition-delay: 0s;
    }

    .ve-dropdown-sub-item-menu {
      display: flex;
      flex-direction: column;
      gap: var(--ve-dropdown-menu-gap);
      padding-block: var(--ve-dropdown-padding-block);
      padding-inline: var(--ve-dropdown-padding-inline);
      max-block-size: var(--ve-dropdown-max-height);
      overflow: auto;
      overscroll-behavior: contain;
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-dropdown-sub-item-panel {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _init_label, _init_extra__init_label, _init_disabled, _init_extra__init_disabled, _init_open, _init_extra__init_open, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_menuSlot, _init_extra__init_menuSlot;
const VE_DROPDOWN_SUB_ITEM_TAG = 've-dropdown-sub-item';
const HOVER_SAFE_CLOSE_DELAY = 300;
_dec = property({
    type: String
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = state(), _dec4 = query('.ve-dropdown-sub-item-trigger'), _dec5 = query('.ve-dropdown-sub-item-panel'), _dec6 = query('slot:not([name])');
class VeDropdownSubItemElement extends VeElement {
    static{
        ({ e: [_init_label, _init_extra__init_label, _init_disabled, _init_extra__init_disabled, _init_open, _init_extra__init_open, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_menuSlot, _init_extra__init_menuSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "label"
            ],
            [
                _dec1,
                1,
                "disabled"
            ],
            [
                _dec2,
                1,
                "open"
            ],
            [
                _dec3,
                1,
                "resolvedPlacement"
            ],
            [
                _dec4,
                1,
                "triggerElement"
            ],
            [
                _dec5,
                1,
                "panelElement"
            ],
            [
                _dec6,
                1,
                "menuSlot"
            ]
        ]));
    }
    static styles = ve_dropdown_sub_item;
    #___private_label_1 = (()=>{
        const _value = _init_label(this, '');
        _init_extra__init_label(this);
        return _value;
    })();
    get label() {
        return this.#___private_label_1;
    }
    set label(_v) {
        this.#___private_label_1 = _v;
    }
    #___private_disabled_2 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_2;
    }
    set disabled(_v) {
        this.#___private_disabled_2 = _v;
    }
    #___private_open_3 = (()=>{
        const _value = _init_open(this, false);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_3;
    }
    set open(_v) {
        this.#___private_open_3 = _v;
    }
    #___private_resolvedPlacement_4 = (()=>{
        const _value = _init_resolvedPlacement(this, 'right-start');
        _init_extra__init_resolvedPlacement(this);
        return _value;
    })();
    get resolvedPlacement() {
        return this.#___private_resolvedPlacement_4;
    }
    set resolvedPlacement(_v) {
        this.#___private_resolvedPlacement_4 = _v;
    }
    #___private_triggerElement_5 = (()=>{
        const _value = _init_triggerElement(this);
        _init_extra__init_triggerElement(this);
        return _value;
    })();
    get triggerElement() {
        return this.#___private_triggerElement_5;
    }
    set triggerElement(_v) {
        this.#___private_triggerElement_5 = _v;
    }
    #___private_panelElement_6 = (()=>{
        const _value = _init_panelElement(this);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_6;
    }
    set panelElement(_v) {
        this.#___private_panelElement_6 = _v;
    }
    #___private_menuSlot_7 = (()=>{
        const _value = _init_menuSlot(this);
        _init_extra__init_menuSlot(this);
        return _value;
    })();
    get menuSlot() {
        return this.#___private_menuSlot_7;
    }
    set menuSlot(_v) {
        this.#___private_menuSlot_7 = _v;
    }
    closeTimer = null;
    positionRequestId = 0;
    cleanupAutoUpdate = null;
    repositionScheduler = createRafScheduler();
    suppressFocusOpen = false;
    clearCloseTimer() {
        if (null !== this.closeTimer) {
            window.clearTimeout(this.closeTimer);
            this.closeTimer = null;
        }
    }
    setOpen(nextOpen) {
        if (this.disabled && nextOpen) return;
        if (this.open === nextOpen) return;
        if (nextOpen && !this.hasMenuContent()) return;
        this.open = nextOpen;
    }
    hasMenuContent() {
        return hasAssignedSlotContent(this.menuSlot);
    }
    getMenuItems() {
        return getAssignedMenuNodes(this.menuSlot);
    }
    focus(options) {
        this.triggerElement?.focus(options);
    }
    restoreTriggerFocus() {
        if (!this.triggerElement) {
            this.suppressFocusOpen = false;
            return;
        }
        this.suppressFocusOpen = true;
        focusElement(this.triggerElement);
    }
    focusFirstMenuItem() {
        focusElement(this.getMenuItems()[0]);
    }
    focusLastMenuItem() {
        const items = this.getMenuItems();
        focusElement(items.at(-1));
    }
    focusAdjacentMenuItem(step) {
        const items = this.getMenuItems();
        if (0 === items.length) return;
        const activeElement = document.activeElement;
        const activeItem = items.find((item)=>item === activeElement || item.shadowRoot?.activeElement === activeElement);
        const activeIndex = activeItem ? items.indexOf(activeItem) : -1;
        const nextIndex = -1 === activeIndex ? 1 === step ? 0 : items.length - 1 : (activeIndex + step + items.length) % items.length;
        focusElement(items[nextIndex]);
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    setupAutoUpdate() {
        if (!this.isConnected || !this.open || !this.triggerElement || !this.panelElement) return;
        this.teardownAutoUpdate();
        this.cleanupAutoUpdate = createFloatingAutoUpdate(this.triggerElement, this.panelElement, ()=>{
            this.scheduleUpdatePosition();
        });
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
    }
    async updatePosition() {
        if (!this.open || !this.isConnected || !this.triggerElement || !this.panelElement) return;
        const requestId = ++this.positionRequestId;
        const distance = getFloatingDistance(this, {
            offsetVarName: '--ve-dropdown-sub-item-gap',
            arrowVarName: '--ve-dropdown-arrow-size',
            fallbackOffset: 4,
            fallbackArrowSize: 0,
            arrowGap: 0
        });
        this.panelElement.style.position = 'fixed';
        this.panelElement.style.top = '0px';
        this.panelElement.style.left = '0px';
        const position = await computeFloatingPanelPosition({
            reference: this.triggerElement,
            floating: this.panelElement,
            placement: 'right-start',
            distance,
            viewportPadding: 8
        });
        if (requestId !== this.positionRequestId || !this.open || !this.isConnected) return;
        this.resolvedPlacement = position.placement;
        this.panelElement.style.left = `${position.x}px`;
        this.panelElement.style.top = `${position.y}px`;
        this.panelElement.style.visibility = position.hidden ? 'hidden' : 'visible';
    }
    handleMouseEnter = ()=>{
        if (this.disabled) return;
        this.clearCloseTimer();
        this.setOpen(true);
    };
    handleMouseLeave = ()=>{
        if (this.disabled) return;
        this.clearCloseTimer();
        this.closeTimer = window.setTimeout(()=>{
            this.setOpen(false);
            this.closeTimer = null;
        }, HOVER_SAFE_CLOSE_DELAY);
    };
    handleFocusIn = ()=>{
        if (this.disabled) return;
        this.clearCloseTimer();
        if (this.suppressFocusOpen) {
            this.suppressFocusOpen = false;
            return;
        }
        this.setOpen(true);
    };
    handleFocusOut = (event)=>{
        if (this.disabled) return;
        const nextTarget = event.relatedTarget;
        if (nextTarget && this.contains(nextTarget)) return;
        this.setOpen(false);
    };
    handleKeydown = (event)=>{
        if ('ArrowRight' === event.key || 'Enter' === event.key || ' ' === event.key) {
            event.preventDefault();
            this.setOpen(true);
            this.updateComplete.then(()=>{
                this.focusFirstMenuItem();
            });
            return;
        }
        if ('ArrowLeft' === event.key || 'Escape' === event.key) {
            event.preventDefault();
            event.stopPropagation();
            this.setOpen(false);
            this.restoreTriggerFocus();
        }
    };
    handlePanelKeydown = (event)=>{
        if ('ArrowDown' === event.key) {
            event.preventDefault();
            event.stopPropagation();
            this.focusAdjacentMenuItem(1);
            return;
        }
        if ('ArrowUp' === event.key) {
            event.preventDefault();
            event.stopPropagation();
            this.focusAdjacentMenuItem(-1);
            return;
        }
        if ('Home' === event.key) {
            event.preventDefault();
            event.stopPropagation();
            this.focusFirstMenuItem();
            return;
        }
        if ('End' === event.key) {
            event.preventDefault();
            event.stopPropagation();
            this.focusLastMenuItem();
            return;
        }
        if ('ArrowLeft' === event.key) {
            event.preventDefault();
            event.stopPropagation();
            this.setOpen(false);
            this.restoreTriggerFocus();
            return;
        }
        if ('Escape' === event.key) {
            event.preventDefault();
            event.stopPropagation();
            this.setOpen(false);
            this.restoreTriggerFocus();
        }
    };
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('disabled') && this.disabled) this.setOpen(false);
        if (changedProperties.has('open')) if (this.open) this.updateComplete.then(()=>{
            if (!this.isConnected || !this.open) return;
            showPanelPopover(this.panelElement);
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
        });
        else {
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            hidePanelPopover(this.panelElement);
            this.teardownAutoUpdate();
        }
    }
    disconnectedCallback() {
        this.clearCloseTimer();
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        hidePanelPopover(this.panelElement);
        this.teardownAutoUpdate();
        super.disconnectedCallback();
    }
    handleMenuSlotChange = ()=>{
        if (!this.hasMenuContent()) return void this.setOpen(false);
        if (this.open) this.scheduleUpdatePosition();
    };
    render() {
        const panelClass = classMap({
            've-dropdown-sub-item-panel': true,
            [`ve-dropdown-sub-item-placement-${this.resolvedPlacement}`]: true
        });
        return html`
      <div
        @mouseenter=${this.handleMouseEnter}
        @mouseleave=${this.handleMouseLeave}
        @focusin=${this.handleFocusIn}
        @focusout=${this.handleFocusOut}
      >
        <button
          class="ve-dropdown-sub-item-trigger"
          part="trigger"
          type="button"
          tabindex="-1"
          role="menuitem"
          aria-haspopup="menu"
          aria-expanded=${this.open ? 'true' : 'false'}
          ?disabled=${this.disabled}
          @keydown=${this.handleKeydown}
        >
          <span class="ve-dropdown-sub-item-label">
            ${(this.label ?? '').trim().length > 0 ? this.label : html`<slot name="label"></slot>`}
          </span>
          <span class="ve-dropdown-sub-item-arrow" aria-hidden="true"> </span>
        </button>

        <div
          class=${panelClass}
          part="panel"
          popover=${"manual"}
          role="menu"
          aria-hidden=${this.open ? 'false' : 'true'}
          @keydown=${this.handlePanelKeydown}
        >
          <div class="ve-dropdown-sub-item-menu" part="menu">
            <slot @slotchange=${this.handleMenuSlotChange}></slot>
          </div>
        </div>
      </div>
    `;
    }
}
defineCustomElement(VE_DROPDOWN_TAG, VeDropdownElement);
defineCustomElement(VE_DROPDOWN_ITEM_TAG, VeDropdownItemElement);
defineCustomElement(VE_DROPDOWN_GROUP_TAG, VeDropdownGroupElement);
defineCustomElement(VE_DROPDOWN_SUB_ITEM_TAG, VeDropdownSubItemElement);
defineCustomElement(VE_DROPDOWN_DIVIDER_TAG, VeDropdownDividerElement);
export { VE_DROPDOWN_GROUP_TAG, VeDropdownGroupElement } from "./16491.js";
export { VE_DROPDOWN_ITEM_SELECT_EVENT, VE_DROPDOWN_ITEM_TAG, VE_DROPDOWN_TAG, VeDropdownElement, VeDropdownItemElement } from "./97605.js";
export { VE_DROPDOWN_DIVIDER_TAG, VE_DROPDOWN_SUB_ITEM_TAG, VeDropdownDividerElement, VeDropdownSubItemElement };
