// @ve-design/web/web/dist/ve-date-picker.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createFloatingAutoUpdate, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
import { closeIcon } from "./icons/close.js";
import { loadingIcon } from "./42938.js";
import { dateCalendarIcon } from "./icons/date-calendar.js";
import { chevronLeftSmIcon } from "./icons/chevron-left-sm.js";
import { chevronRightSmIcon } from "./icons/chevron-right-sm.js";
const veDatePickerTokenDefaults = css`
  :host {
    --ve-date-picker-font-family: var(--font-sans);
    --ve-date-picker-font-weight: var(--font-weight-normal);
    --ve-date-picker-line-height: var(--line-height-body);
    --ve-date-picker-font-size-small: var(--text-body-sm);
    --ve-date-picker-font-size-default: var(--text-body);
    --ve-date-picker-font-size-large: var(--text-body);
    --ve-date-picker-cell-font-size: var(--text-body-sm);
    --ve-date-picker-height-small: var(--height-small);
    --ve-date-picker-height-default: var(--height-default);
    --ve-date-picker-height-large: var(--height-large);
    --ve-date-picker-padding-inline-small: var(--space-xxs);
    --ve-date-picker-padding-inline-default: var(--space-xxs);
    --ve-date-picker-padding-inline-large: var(--space-xs);
    --ve-date-picker-padding-block: var(--space-xxxs);
    --ve-date-picker-radius: var(--radius-md);
    --ve-date-picker-border-width: var(--stroke-weight-base);
    --ve-date-picker-gap: var(--space-xxxs);
    --ve-date-picker-label-padding-inline: var(--space-xxs);
    --ve-date-picker-label-color: var(--color-text-secondary);
    --ve-date-picker-control-background: var(--color-bg-control);
    --ve-date-picker-control-background-hover: var(--color-bg-base);
    --ve-date-picker-control-background-focus: var(--color-bg-base);
    --ve-date-picker-control-background-disabled: var(--color-bg-control);
    --ve-date-picker-control-text: var(--color-text-primary);
    --ve-date-picker-control-text-disabled: var(--color-text-disable);
    --ve-date-picker-control-placeholder: var(--color-text-disable);
    --ve-date-picker-control-border-default: var(--color-border-default);
    --ve-date-picker-control-border-hover: var(--color-border-interactive);
    --ve-date-picker-control-border-focus: var(
      --color-border-interactive-active
    );
    --ve-date-picker-control-border-disabled: var(--color-border-default);
    --ve-date-picker-focus-shadow-color: rgb(
      from var(--color-border-interactive) r g b / 18%
    );
    --ve-date-picker-shadow-focus: 0 0 0 var(--stroke-weight-bold)
      var(--ve-date-picker-focus-shadow-color);
    --ve-date-picker-status-warning-background: var(
      --color-bg-warning-secondary
    );
    --ve-date-picker-status-warning-background-hover: color-mix(
      var(--ve-date-picker-status-warning-background)
        var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-date-picker-status-warning-border: var(--color-border-warning);
    --ve-date-picker-status-warning-border-hover: var(--color-text-warning);
    --ve-date-picker-status-warning-border-focus: var(--color-text-warning);
    --ve-date-picker-status-warning-shadow-color: rgb(
      from var(--color-border-warning) r g b / 18%
    );
    --ve-date-picker-status-error-background: var(--color-bg-danger-secondary);
    --ve-date-picker-status-error-background-hover: color-mix(
      var(--ve-date-picker-status-error-background)
        var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-date-picker-status-error-border: var(--color-border-danger);
    --ve-date-picker-status-error-border-hover: var(--color-text-danger);
    --ve-date-picker-status-error-border-focus: var(--color-text-danger);
    --ve-date-picker-status-error-shadow-color: rgb(
      from var(--color-border-danger) r g b / 18%
    );
    --ve-date-picker-description-color: var(--color-text-tertiary);
    --ve-date-picker-status-warning-description-color: var(
      --color-text-warning
    );
    --ve-date-picker-status-error-description-color: var(--color-text-danger);
    --ve-date-picker-feedback-font-size: var(--text-caption);
    --ve-date-picker-feedback-line-height: var(--line-height-body);
    --ve-date-picker-prefix-color: var(--color-text-secondary);
    --ve-date-picker-arrow-color: var(--color-text-disable);
    --ve-date-picker-arrow-filled-color: var(--color-text-primary);
    --ve-date-picker-clear-color: var(--color-text-tertiary);
    --ve-date-picker-clear-hover-color: var(--color-text-secondary);
    --ve-date-picker-clear-hover-background: var(--color-bg-muted);
    --ve-date-picker-icon-size: var(--space-s);
    --ve-date-picker-clear-size: var(--space-s);
    --ve-date-picker-panel-background: var(--color-bg-overlay);
    --ve-date-picker-panel-border-color: var(--color-border-default);
    --ve-date-picker-panel-shadow: var(--shadow-md);
    --ve-date-picker-panel-radius: var(--radius-lg);
    --ve-date-picker-panel-z-index: 1000;
    --ve-date-picker-popup-gap: var(--space-xxxs);
    --ve-date-picker-panel-gap: 2px;
    --ve-date-picker-panel-padding-block: var(--space-s);
    --ve-date-picker-panel-padding-inline: var(--space-s);
    --ve-date-picker-panel-padding-inline-end-range: calc(
      var(--space-s) + var(--space-xxxs)
    );
    --ve-date-picker-panel-width: 270px;
    --ve-date-picker-panel-range-width: 536px;
    --ve-date-picker-panel-width-with-shortcuts: 388px;
    --ve-date-picker-panel-range-width-with-shortcuts: 646px;
    --ve-date-picker-panel-header-height: 32px;
    --ve-date-picker-navigation-size: var(--height-small);
    --ve-date-picker-panel-divider-color: var(--color-border-default);
    --ve-date-picker-weekday-text: var(--color-text-primary);
    --ve-date-picker-cell-size: 32px;
    --ve-date-picker-cell-radius: var(--radius-sm);
    --ve-date-picker-cell-text: var(--color-text-primary);
    --ve-date-picker-cell-muted-text: var(--color-text-secondary);
    --ve-date-picker-cell-disabled-text: var(--color-text-disable);
    --ve-date-picker-cell-hover-background: var(--color-bg-muted);
    --ve-date-picker-cell-active-background: color-mix(
      var(--color-bg-muted) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );
    --ve-date-picker-cell-selected-background: var(--color-bg-primary);
    --ve-date-picker-cell-selected-text: var(--color-text-foreground);
    --ve-date-picker-cell-range-background: var(--color-primary-50);
    --ve-date-picker-cell-range-inset: 3px;
    --ve-date-picker-cell-today-dot-offset: 1px;
    --ve-date-picker-cell-today-dot: var(--color-text-primary);
    --ve-date-picker-cell-selected-today-dot: var(--color-text-foreground);
    --ve-date-picker-footer-height: 34px;
    --ve-date-picker-footer-text: var(--color-text-accent);
    --ve-date-picker-footer-hover-background: var(--color-bg-muted);
    --ve-date-picker-shortcut-width: 112px;
    --ve-date-picker-shortcut-text: var(--color-text-secondary);
    --ve-date-picker-shortcut-hover-background: var(--color-bg-muted);
    --ve-date-picker-shortcut-selected-background: var(--color-bg-muted);
    --ve-date-picker-shortcut-selected-text: var(--color-text-primary);
    --ve-date-picker-time-select-background: var(--color-bg-control);
    --ve-date-picker-time-select-border: var(--color-border-default);
    --ve-date-picker-loading-color: var(--color-text-tertiary);
    --ve-date-picker-disabled-opacity: var(--state-disabled-ratio);
    --ve-date-picker-transition-duration: var(--motion-fast);
    --ve-date-picker-transition-timing: var(--ease-standard);

    --ve-date-picker-current-height: var(--ve-date-picker-height-default);
    --ve-date-picker-current-padding-inline: var(
      --ve-date-picker-padding-inline-default
    );
    --ve-date-picker-current-font-size: var(--ve-date-picker-font-size-default);
    --ve-date-picker-current-background: var(
      --ve-date-picker-control-background
    );
    --ve-date-picker-current-background-hover: var(
      --ve-date-picker-control-background-hover
    );
    --ve-date-picker-current-border-color: var(
      --ve-date-picker-control-border-default
    );
    --ve-date-picker-current-border-color-hover: var(
      --ve-date-picker-control-border-hover
    );
    --ve-date-picker-current-focus-ring-color: var(
      --ve-date-picker-control-border-focus
    );
    --ve-date-picker-current-description-color: var(
      --ve-date-picker-description-color
    );
    --ve-date-picker-current-arrow-color: var(--ve-date-picker-arrow-color);
  }

  :host([size='small']) {
    --ve-date-picker-current-height: var(--ve-date-picker-height-small);
    --ve-date-picker-current-padding-inline: var(
      --ve-date-picker-padding-inline-small
    );
    --ve-date-picker-current-font-size: var(--ve-date-picker-font-size-small);
  }

  :host([size='large']) {
    --ve-date-picker-current-height: var(--ve-date-picker-height-large);
    --ve-date-picker-current-padding-inline: var(
      --ve-date-picker-padding-inline-large
    );
    --ve-date-picker-current-font-size: var(--ve-date-picker-font-size-large);
  }

  :host([status='warning']) {
    --ve-date-picker-current-background: var(
      --ve-date-picker-status-warning-background
    );
    --ve-date-picker-current-background-hover: var(
      --ve-date-picker-status-warning-background-hover
    );
    --ve-date-picker-current-border-color: var(
      --ve-date-picker-status-warning-border
    );
    --ve-date-picker-current-border-color-hover: var(
      --ve-date-picker-status-warning-border-hover
    );
    --ve-date-picker-current-focus-ring-color: var(
      --ve-date-picker-status-warning-border-focus
    );
    --ve-date-picker-current-description-color: var(
      --ve-date-picker-status-warning-description-color
    );
    --ve-date-picker-focus-shadow-color: var(
      --ve-date-picker-status-warning-shadow-color
    );
  }

  :host([status='error']) {
    --ve-date-picker-current-background: var(
      --ve-date-picker-status-error-background
    );
    --ve-date-picker-current-background-hover: var(
      --ve-date-picker-status-error-background-hover
    );
    --ve-date-picker-current-border-color: var(
      --ve-date-picker-status-error-border
    );
    --ve-date-picker-current-border-color-hover: var(
      --ve-date-picker-status-error-border-hover
    );
    --ve-date-picker-current-focus-ring-color: var(
      --ve-date-picker-status-error-border-focus
    );
    --ve-date-picker-current-description-color: var(
      --ve-date-picker-status-error-description-color
    );
    --ve-date-picker-focus-shadow-color: var(
      --ve-date-picker-status-error-shadow-color
    );
  }
`;
const style_token = veDatePickerTokenDefaults;
const ve_date_picker = [
    componentReset,
    style_token,
    css`
    :host {
      position: relative;
      display: inline-flex;
      inline-size: 100%;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-date-picker {
      position: relative;
      inline-size: 100%;
    }

    .ve-date-picker-field {
      display: flex;
      align-items: stretch;
      inline-size: 100%;
    }

    .ve-date-picker-control {
      appearance: none;
      display: flex;
      align-items: center;
      gap: var(--ve-date-picker-gap);
      inline-size: 100%;
      block-size: var(--ve-date-picker-current-height);
      min-inline-size: 0;
      box-sizing: border-box;
      padding-inline: var(--ve-date-picker-current-padding-inline);
      padding-block: var(--ve-date-picker-padding-block);
      border: var(--ve-date-picker-border-width) solid
        var(--ve-date-picker-current-border-color);
      border-radius: var(--ve-date-picker-radius);
      background: var(--ve-date-picker-current-background);
      color: var(--ve-date-picker-control-text);
      font-family: var(--ve-date-picker-font-family);
      font-size: var(--ve-date-picker-current-font-size);
      font-weight: var(--ve-date-picker-font-weight);
      line-height: var(--ve-date-picker-line-height);
      cursor: pointer;
      transition:
        border-color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        box-shadow var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        background-color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        opacity var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing);
      overflow: hidden;
    }

    .ve-date-picker-label {
      display: inline-flex;
      align-items: center;
      flex: none;
      box-sizing: border-box;
      block-size: var(--ve-date-picker-current-height);
      min-inline-size: 0;
      padding-inline: var(--ve-date-picker-label-padding-inline);
      padding-block: var(--ve-date-picker-padding-block);
      border: var(--ve-date-picker-border-width) solid
        var(--ve-date-picker-current-border-color);
      border-start-start-radius: var(--ve-date-picker-radius);
      border-end-start-radius: var(--ve-date-picker-radius);
      border-start-end-radius: 0;
      border-end-end-radius: 0;
      background: var(--ve-date-picker-current-background);
      color: var(--ve-date-picker-label-color);
      font-family: var(--ve-date-picker-font-family);
      font-size: var(--ve-date-picker-current-font-size);
      font-weight: var(--ve-date-picker-font-weight);
      line-height: var(--ve-date-picker-line-height);
      white-space: nowrap;
      cursor: default;
      transition:
        border-color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        background-color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        opacity var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing);
    }

    .ve-date-picker-with-label .ve-date-picker-control {
      border-start-start-radius: 0;
      border-end-start-radius: 0;
      margin-inline-start: calc(var(--ve-date-picker-border-width) * -1);
    }

    .ve-date-picker-label-text,
    .ve-date-picker-value {
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-date-picker-value {
      flex: 1;
      text-align: start;
    }

    .ve-date-picker-range-value {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      flex: 1;
      min-inline-size: 0;
    }

    .ve-date-picker-range-segment {
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      border-radius: var(--radius-sm);
      padding-inline: var(--space-xxxs);
    }

    .ve-date-picker-range-segment-active {
      background: var(--color-bg-muted);
      color: var(--ve-date-picker-control-text);
    }

    .ve-date-picker-range-segment-placeholder {
      color: var(--ve-date-picker-control-placeholder);
    }

    .ve-date-picker-range-separator {
      flex: none;
      color: var(--color-text-secondary);
      white-space: nowrap;
    }

    .ve-date-picker-value-placeholder {
      color: var(--ve-date-picker-control-placeholder);
    }

    .ve-date-picker-prefix,
    .ve-date-picker-arrow,
    .ve-date-picker-loading {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
    }

    .ve-date-picker-prefix {
      color: var(--ve-date-picker-prefix-color);
    }

    .ve-date-picker-prefix-empty {
      display: none;
    }

    .ve-date-picker-arrow {
      color: var(--ve-date-picker-arrow-color);
      transition: transform var(--ve-date-picker-transition-duration)
        var(--ve-date-picker-transition-timing);
    }

    .ve-date-picker-control-filled .ve-date-picker-arrow {
      color: var(--ve-date-picker-arrow-filled-color);
    }

    .ve-date-picker-loading {
      color: var(--ve-date-picker-loading-color);
    }

    .ve-date-picker-prefix ve-icon,
    .ve-date-picker-arrow ve-icon,
    .ve-date-picker-loading ve-icon,
    .ve-date-picker-clear ve-icon,
    .ve-date-picker-navigation ve-icon {
      --ve-icon-size: var(--ve-date-picker-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-date-picker-loading ve-icon {
      animation: ve-date-picker-loading-spin var(--motion-slow) linear infinite;
    }

    .ve-date-picker-actions {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-date-picker-icon-size);
      block-size: var(--ve-date-picker-icon-size);
      margin-inline-start: auto;
      flex: none;
      pointer-events: auto;
    }

    .ve-date-picker-clear,
    .ve-date-picker-navigation {
      appearance: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      padding: 0;
      border: none;
      background: transparent;
      font: inherit;
      cursor: pointer;
    }

    .ve-date-picker-clear,
    .ve-date-picker-navigation {
      border-radius: var(--radius-sm);
      color: var(--ve-date-picker-clear-color);
    }

    .ve-date-picker-clear {
      position: absolute;
      inset: 0;
      z-index: 1;
      inline-size: var(--ve-date-picker-clear-size);
      block-size: var(--ve-date-picker-clear-size);
      opacity: 0;
      transform: scale(0.88);
      pointer-events: none;
      transition:
        opacity var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        transform var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        background-color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing);
    }

    .ve-date-picker-navigation {
      inline-size: var(--ve-date-picker-navigation-size);
      block-size: var(--ve-date-picker-navigation-size);
    }

    .ve-date-picker-arrow,
    .ve-date-picker-loading {
      pointer-events: none;
      transition:
        opacity var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        transform var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing);
    }

    :host([allow-clear]:not([disabled]):not([loading]))
      .ve-date-picker-control:hover
      .ve-date-picker-clear {
      opacity: 1;
      transform: scale(1);
      pointer-events: auto;
    }

    :host([allow-clear]:not([disabled]):not([loading]))
      .ve-date-picker-control:hover
      .ve-date-picker-arrow {
      opacity: 0;
      transform: scale(0.88);
    }

    .ve-date-picker-clear:hover,
    .ve-date-picker-navigation:hover {
      background: var(--ve-date-picker-clear-hover-background);
      color: var(--ve-date-picker-clear-hover-color);
    }

    :host(:not([disabled])) .ve-date-picker-control:hover {
      border-color: var(--ve-date-picker-current-border-color-hover);
      background: var(--ve-date-picker-current-background-hover);
    }

    :host(:not([disabled])) .ve-date-picker-control:focus,
    :host([open]) .ve-date-picker-control {
      outline: none;
      background: var(--ve-date-picker-control-background-focus);
      border-color: var(--ve-date-picker-current-focus-ring-color);
      box-shadow: var(--ve-date-picker-shadow-focus);
    }

    :host([open]) .ve-date-picker-label {
      background: var(--ve-date-picker-control-background-focus);
      border-color: var(--ve-date-picker-current-focus-ring-color);
    }

    :host([disabled]) .ve-date-picker-control,
    :host([disabled]) .ve-date-picker-label {
      color: var(--ve-date-picker-control-text-disabled);
      background: var(--ve-date-picker-control-background-disabled);
      border-color: var(--ve-date-picker-control-border-disabled);
      opacity: var(--ve-date-picker-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-date-picker-description {
      min-block-size: calc(
        var(--ve-date-picker-feedback-font-size) *
          var(--ve-date-picker-feedback-line-height)
      );
      margin-block-start: var(--space-xxxs);
      color: var(--ve-date-picker-current-description-color);
      font-family: var(--ve-date-picker-font-family);
      font-size: var(--ve-date-picker-feedback-font-size);
      line-height: var(--ve-date-picker-feedback-line-height);
    }

    .ve-date-picker-panel {
      position: fixed;
      inset: auto;
      inset-block-start: 0;
      inset-inline-start: 0;
      z-index: var(--ve-date-picker-panel-z-index);
      box-sizing: border-box;
      inline-size: var(--ve-date-picker-panel-width);
      margin: 0;
      padding-block: var(--ve-date-picker-panel-padding-block);
      padding-inline: var(--ve-date-picker-panel-padding-inline);
      border: var(--ve-date-picker-border-width) solid
        var(--ve-date-picker-panel-border-color);
      border-radius: var(--ve-date-picker-panel-radius);
      background: var(--ve-date-picker-panel-background);
      box-shadow: var(--ve-date-picker-panel-shadow);
      font-family: var(--ve-date-picker-font-family);
      font-size: var(--ve-date-picker-cell-font-size);
      line-height: var(--ve-date-picker-line-height);
      color: var(--ve-date-picker-cell-text);
    }

    :host([range]) .ve-date-picker-panel {
      inline-size: var(--ve-date-picker-panel-range-width);
      padding-inline-end: var(--ve-date-picker-panel-padding-inline-end-range);
    }

    .ve-date-picker-panel-with-left-shortcuts {
      display: flex;
      gap: var(--space-s);
      inline-size: var(--ve-date-picker-panel-width-with-shortcuts);
    }

    :host([range]) .ve-date-picker-panel-with-left-shortcuts {
      inline-size: var(--ve-date-picker-panel-range-width-with-shortcuts);
    }

    .ve-date-picker-panel-main {
      flex: 1;
      min-inline-size: 0;
    }

    .ve-date-picker-calendar-group {
      display: grid;
      grid-template-columns: repeat(2, minmax(0, 1fr));
      gap: var(--space-l);
    }

    .ve-date-picker-calendar {
      min-inline-size: 0;
    }

    .ve-date-picker-header {
      display: flex;
      align-items: center;
      justify-content: space-between;
      block-size: var(--ve-date-picker-panel-header-height);
      gap: var(--ve-date-picker-panel-gap);
      margin-block-end: var(--space-xxxs);
    }

    .ve-date-picker-header-label {
      min-inline-size: 0;
      text-align: start;
      color: var(--ve-date-picker-cell-text);
      font-weight: var(--font-weight-medium);
      font-size: var(--text-body);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-date-picker-navigation-group {
      display: flex;
      align-items: center;
      flex-shrink: 0;
      gap: var(--ve-date-picker-panel-gap);
    }

    .ve-date-picker-weekdays,
    .ve-date-picker-grid {
      display: grid;
      grid-template-columns: repeat(7, 1fr);
      gap: var(--ve-date-picker-panel-gap);
    }

    .ve-date-picker-weekdays {
      margin-block-end: 2px;
      color: var(--ve-date-picker-weekday-text);
      font-size: var(--text-caption);
    }

    .ve-date-picker-weekday {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      block-size: var(--ve-date-picker-cell-size);
      font-weight: var(--font-weight-normal);
    }

    .ve-date-picker-cell {
      appearance: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      position: relative;
      inline-size: 100%;
      block-size: var(--ve-date-picker-cell-size);
      box-sizing: border-box;
      padding: 0;
      border: var(--ve-date-picker-border-width) solid transparent;
      border-radius: var(--ve-date-picker-cell-radius);
      background: transparent;
      color: var(--ve-date-picker-cell-text);
      font: inherit;
      cursor: pointer;
      transition:
        background-color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        border-color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        color var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing),
        opacity var(--ve-date-picker-transition-duration)
          var(--ve-date-picker-transition-timing);
    }

    .ve-date-picker-cell::before {
      content: '';
      position: absolute;
      inset-block: var(--ve-date-picker-cell-range-inset);
      inset-inline: 0;
      background: transparent;
      pointer-events: none;
      z-index: 0;
    }

    .ve-date-picker-cell-label {
      position: relative;
      z-index: 1;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: calc(
        var(--ve-date-picker-cell-size) - var(
            --ve-date-picker-cell-range-inset
          ) *
          2
      );
      block-size: calc(
        var(--ve-date-picker-cell-size) - var(
            --ve-date-picker-cell-range-inset
          ) *
          2
      );
      border-radius: var(--radius-full);
    }

    .ve-date-picker-cell:hover,
    .ve-date-picker-cell:focus {
      outline: none;
      background: var(--ve-date-picker-cell-hover-background);
    }

    .ve-date-picker-cell:active {
      background: var(--ve-date-picker-cell-active-background);
    }

    .ve-date-picker-cell-outside {
      color: var(--ve-date-picker-cell-muted-text);
    }

    .ve-date-picker-cell-today .ve-date-picker-cell-label::after {
      content: '';
      position: absolute;
      inset-block-end: var(--ve-date-picker-cell-today-dot-offset);
      inset-inline-start: 50%;
      inline-size: 4px;
      block-size: 4px;
      border-radius: var(--radius-full);
      background: var(--ve-date-picker-cell-today-dot);
      transform: translateX(-50%);
    }

    .ve-date-picker-cell-in-range:not(.ve-date-picker-cell-selected) {
      background: var(--ve-date-picker-cell-range-background);
    }

    .ve-date-picker-grid-range {
      column-gap: 0;
      row-gap: var(--space-xxxs);
    }

    .ve-date-picker-grid-range .ve-date-picker-cell {
      border-radius: 0;
      z-index: 0;
    }

    .ve-date-picker-grid-range .ve-date-picker-cell::before {
      background: var(--ve-date-picker-cell-range-background);
    }

    .ve-date-picker-grid-range
      .ve-date-picker-cell:not(.ve-date-picker-cell-in-range):not(
        .ve-date-picker-cell-range-start
      ):not(.ve-date-picker-cell-range-end)::before {
      background: transparent;
    }

    .ve-date-picker-grid-range .ve-date-picker-cell-range-start::before {
      inset-inline-start: 50%;
    }

    .ve-date-picker-grid-range .ve-date-picker-cell-range-end::before {
      inset-inline-end: 50%;
    }

    .ve-date-picker-grid-range
      .ve-date-picker-cell-range-start.ve-date-picker-cell-range-end::before {
      background: transparent;
    }

    .ve-date-picker-grid-range .ve-date-picker-cell-selected {
      border-radius: var(--radius-full);
    }

    .ve-date-picker-cell-selected {
      border-radius: var(--radius-full);
      color: var(--ve-date-picker-cell-selected-text);
      font-weight: var(--font-weight-normal);
    }

    .ve-date-picker-cell-selected .ve-date-picker-cell-label {
      background: var(--ve-date-picker-cell-selected-background);
    }

    .ve-date-picker-cell-selected.ve-date-picker-cell-today
      .ve-date-picker-cell-label::after {
      background: var(--ve-date-picker-cell-selected-today-dot);
    }

    .ve-date-picker-cell-selected:hover,
    .ve-date-picker-cell-selected:focus,
    .ve-date-picker-cell-selected:active {
      background: transparent;
      color: var(--ve-date-picker-cell-selected-text);
    }

    .ve-date-picker-cell-selected:hover .ve-date-picker-cell-label,
    .ve-date-picker-cell-selected:focus .ve-date-picker-cell-label,
    .ve-date-picker-cell-selected:active .ve-date-picker-cell-label {
      background: var(--ve-date-picker-cell-selected-background);
    }

    .ve-date-picker-cell-disabled {
      color: var(--ve-date-picker-cell-disabled-text);
      cursor: not-allowed;
    }

    .ve-date-picker-grid-range .ve-date-picker-cell-outside {
      color: transparent;
      pointer-events: none;
    }

    .ve-date-picker-grid-range
      .ve-date-picker-cell-outside
      .ve-date-picker-cell-label::after {
      display: none;
    }

    .ve-date-picker-cell-disabled:hover,
    .ve-date-picker-cell-disabled:focus,
    .ve-date-picker-cell-disabled:active {
      background: transparent;
    }

    .ve-date-picker-shortcuts {
      display: flex;
      gap: var(--space-xxxs);
      padding-block-start: var(--space-xxs);
      margin-block-start: var(--space-xxs);
      border-block-start: var(--ve-date-picker-border-width) solid
        var(--ve-date-picker-panel-divider-color);
    }

    .ve-date-picker-panel-with-left-shortcuts .ve-date-picker-shortcuts {
      flex: none;
      flex-direction: column;
      inline-size: var(--ve-date-picker-shortcut-width);
      padding-block-start: 0;
      padding-inline-end: var(--space-xs);
      margin-block-start: 0;
      border-block-start: none;
      border-inline-end: var(--ve-date-picker-border-width) solid
        var(--ve-date-picker-panel-divider-color);
    }

    .ve-date-picker-shortcut {
      appearance: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      border: none;
      background: transparent;
      font: inherit;
      cursor: pointer;
      min-block-size: 28px;
      padding-inline: var(--space-xs);
      border-radius: var(--radius-md);
      color: var(--ve-date-picker-shortcut-text);
      font-weight: var(--font-weight-medium);
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-date-picker-shortcut {
      font-weight: var(--font-weight-normal);
    }

    .ve-date-picker-shortcut-selected {
      background: var(--ve-date-picker-shortcut-selected-background);
      color: var(--ve-date-picker-shortcut-selected-text);
      font-weight: var(--font-weight-medium);
    }

    .ve-date-picker-panel-with-left-shortcuts .ve-date-picker-shortcut {
      align-self: stretch;
      box-sizing: border-box;
      inline-size: 100%;
      justify-content: center;
      text-align: center;
    }

    .ve-date-picker-shortcut:hover,
    .ve-date-picker-shortcut:focus {
      outline: none;
      background: var(--ve-date-picker-shortcut-hover-background);
    }

    .ve-date-picker-shortcut-selected:hover,
    .ve-date-picker-shortcut-selected:focus {
      background: var(--ve-date-picker-shortcut-selected-background);
      color: var(--ve-date-picker-shortcut-selected-text);
    }

    @keyframes ve-date-picker-loading-spin {
      to {
        transform: rotate(360deg);
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _dec24, _dec25, _dec26, _dec27, _dec28, _dec29, _init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_placeholder, _init_extra__init_placeholder, _init_label, _init_extra__init_label, _init_description, _init_extra__init_description, _init_format, _init_extra__init_format, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_range, _init_extra__init_range, _init_loading, _init_extra__init_loading, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_min, _init_extra__init_min, _init_max, _init_extra__init_max, _init_disabledDate, _init_extra__init_disabledDate, _init_shortcuts, _init_extra__init_shortcuts, _init_shortcutPlacement, _init_extra__init_shortcutPlacement, _init_defaultPickerValue, _init_extra__init_defaultPickerValue, _init_pickerValue, _init_extra__init_pickerValue, _init_separator, _init_extra__init_separator, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_controlElement, _init_extra__init_controlElement, _init_fieldElement, _init_extra__init_fieldElement, _init_panelElement, _init_extra__init_panelElement, _init_panelPlacement, _init_extra__init_panelPlacement, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot, _init_viewYear, _init_extra__init_viewYear, _init_viewMonth, _init_extra__init_viewMonth, _init_activeRangeEnd, _init_extra__init_activeRangeEnd;
const VE_DATE_PICKER_TAG = 've-date-picker';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby',
    'aria-required',
    'aria-invalid'
];
const WEEKDAY_LABELS = [
    '日',
    '一',
    '二',
    '三',
    '四',
    '五',
    '六'
];
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        return null === value || 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
function normalizeAriaBooleanString(value) {
    if (null === value) return;
    return 'false' === value ? 'false' : 'true';
}
function toTwoDigit(value) {
    return String(value).padStart(2, '0');
}
function createDateFromParts(parts) {
    return new Date(parts.year, parts.month - 1, parts.day);
}
function getTodayParts() {
    const today = new Date();
    return {
        year: today.getFullYear(),
        month: today.getMonth() + 1,
        day: today.getDate()
    };
}
function addDateParts(parts, { days = 0, months = 0, years = 0 }) {
    const date = createDateFromParts(parts);
    date.setFullYear(date.getFullYear() + years);
    date.setMonth(date.getMonth() + months);
    date.setDate(date.getDate() + days);
    return {
        year: date.getFullYear(),
        month: date.getMonth() + 1,
        day: date.getDate()
    };
}
function getDateKey(parts) {
    return 10000 * parts.year + 100 * parts.month + parts.day;
}
function areSameDate(first, second) {
    return void 0 !== first && void 0 !== second && first.year === second.year && first.month === second.month && first.day === second.day;
}
function isValidDateParts(parts) {
    const date = createDateFromParts(parts);
    return date.getFullYear() === parts.year && date.getMonth() === parts.month - 1 && date.getDate() === parts.day;
}
function parseDateValue(value) {
    const trimmed = value.trim().split(/\s+/)[0] ?? '';
    if (!trimmed) return;
    let year;
    let month;
    let day;
    const yearFirstMatch = /^(\d{4})([-/])(\d{1,2})\2(\d{1,2})$/.exec(trimmed);
    if (yearFirstMatch) {
        year = Number(yearFirstMatch[1]);
        month = Number(yearFirstMatch[3]);
        day = Number(yearFirstMatch[4]);
    } else {
        const monthFirstMatch = /^(\d{1,2})\/(\d{1,2})\/(\d{4})$/.exec(trimmed);
        if (!monthFirstMatch) return;
        month = Number(monthFirstMatch[1]);
        day = Number(monthFirstMatch[2]);
        year = Number(monthFirstMatch[3]);
    }
    const parts = {
        year,
        month,
        day
    };
    return isValidDateParts(parts) ? parts : void 0;
}
function getDefaultFormat() {
    return 'YYYY-MM-DD';
}
function formatDateValue(parts, format) {
    const year = String(parts.year).padStart(4, '0');
    const month = toTwoDigit(parts.month);
    const day = toTwoDigit(parts.day);
    const result = format.replace('YYYY', year).replace('MM', month).replace('DD', day);
    return result || `${year}-${month}-${day}`;
}
function addMonths(year, month, offset) {
    const date = new Date(year, month - 1 + offset, 1);
    return {
        year: date.getFullYear(),
        month: date.getMonth() + 1
    };
}
function parseValue(value) {
    const trimmed = value.trim();
    if (!trimmed) return;
    const relativeParts = resolveRelativeShortcutParts(trimmed);
    if (relativeParts) return {
        parts: relativeParts
    };
    const parts = parseDateValue(trimmed);
    if (!parts) return;
    return {
        parts
    };
}
function resolveRelativeShortcutParts(token) {
    const normalizedToken = token.trim().toLowerCase();
    if (!normalizedToken) return;
    if ('today' === normalizedToken) return getTodayParts();
    if ('yesterday' === normalizedToken) return addDateParts(getTodayParts(), {
        days: -1
    });
    if ('tomorrow' === normalizedToken) return addDateParts(getTodayParts(), {
        days: 1
    });
    const offsetMatch = /^([+-]?\d+)([dwmy])$/.exec(normalizedToken);
    if (!offsetMatch) return;
    const amount = Number(offsetMatch[1]);
    const unit = offsetMatch[2];
    if ('d' === unit) return addDateParts(getTodayParts(), {
        days: amount
    });
    if ('w' === unit) return addDateParts(getTodayParts(), {
        days: 7 * amount
    });
    if ('m' === unit) return addDateParts(getTodayParts(), {
        months: amount
    });
    return addDateParts(getTodayParts(), {
        years: amount
    });
}
_dec = property({
    type: String
}), _dec1 = property({
    type: String,
    attribute: 'default-value'
}), _dec2 = property({
    type: String
}), _dec3 = property({
    type: String
}), _dec4 = property({
    type: String
}), _dec5 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec6 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec8 = property({
    type: Boolean,
    reflect: true
}), _dec9 = property({
    type: Boolean,
    reflect: true
}), _dec10 = property({
    type: Boolean,
    reflect: true
}), _dec11 = property({
    type: Boolean,
    attribute: 'allow-clear',
    reflect: true
}), _dec12 = property({
    type: Boolean,
    reflect: true
}), _dec13 = property({
    type: String
}), _dec14 = property({
    type: String
}), _dec15 = property({
    type: String,
    attribute: 'disabled-date'
}), _dec16 = property({
    type: String
}), _dec17 = property({
    type: String,
    attribute: 'shortcut-placement',
    reflect: true
}), _dec18 = property({
    type: String,
    attribute: 'default-picker-value'
}), _dec19 = property({
    type: String,
    attribute: 'picker-value'
}), _dec20 = property({
    type: String
}), _dec21 = property({
    attribute: 'popup-match-select-width',
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER
}), _dec22 = query('.ve-date-picker-control'), _dec23 = query('.ve-date-picker-field'), _dec24 = query('.ve-date-picker-panel'), _dec25 = state(), _dec26 = state(), _dec27 = state(), _dec28 = state(), _dec29 = state();
class VeDatePickerElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_placeholder, _init_extra__init_placeholder, _init_label, _init_extra__init_label, _init_description, _init_extra__init_description, _init_format, _init_extra__init_format, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_range, _init_extra__init_range, _init_loading, _init_extra__init_loading, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_min, _init_extra__init_min, _init_max, _init_extra__init_max, _init_disabledDate, _init_extra__init_disabledDate, _init_shortcuts, _init_extra__init_shortcuts, _init_shortcutPlacement, _init_extra__init_shortcutPlacement, _init_defaultPickerValue, _init_extra__init_defaultPickerValue, _init_pickerValue, _init_extra__init_pickerValue, _init_separator, _init_extra__init_separator, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_controlElement, _init_extra__init_controlElement, _init_fieldElement, _init_extra__init_fieldElement, _init_panelElement, _init_extra__init_panelElement, _init_panelPlacement, _init_extra__init_panelPlacement, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot, _init_viewYear, _init_extra__init_viewYear, _init_viewMonth, _init_extra__init_viewMonth, _init_activeRangeEnd, _init_extra__init_activeRangeEnd] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "value"
            ],
            [
                _dec1,
                1,
                "defaultValue"
            ],
            [
                _dec2,
                1,
                "placeholder"
            ],
            [
                _dec3,
                1,
                "label"
            ],
            [
                _dec4,
                1,
                "description"
            ],
            [
                _dec5,
                1,
                "format"
            ],
            [
                _dec6,
                1,
                "size"
            ],
            [
                _dec7,
                1,
                "status"
            ],
            [
                _dec8,
                1,
                "disabled"
            ],
            [
                _dec9,
                1,
                "range"
            ],
            [
                _dec10,
                1,
                "loading"
            ],
            [
                _dec11,
                1,
                "allowClear"
            ],
            [
                _dec12,
                1,
                "open"
            ],
            [
                _dec13,
                1,
                "min"
            ],
            [
                _dec14,
                1,
                "max"
            ],
            [
                _dec15,
                1,
                "disabledDate"
            ],
            [
                _dec16,
                1,
                "shortcuts"
            ],
            [
                _dec17,
                1,
                "shortcutPlacement"
            ],
            [
                _dec18,
                1,
                "defaultPickerValue"
            ],
            [
                _dec19,
                1,
                "pickerValue"
            ],
            [
                _dec20,
                1,
                "separator"
            ],
            [
                _dec21,
                1,
                "popupMatchSelectWidth"
            ],
            [
                _dec22,
                1,
                "controlElement"
            ],
            [
                _dec23,
                1,
                "fieldElement"
            ],
            [
                _dec24,
                1,
                "panelElement"
            ],
            [
                _dec25,
                1,
                "panelPlacement"
            ],
            [
                _dec26,
                1,
                "hasPrefixSlot"
            ],
            [
                _dec27,
                1,
                "viewYear"
            ],
            [
                _dec28,
                1,
                "viewMonth"
            ],
            [
                _dec29,
                1,
                "activeRangeEnd"
            ]
        ]));
    }
    static styles = ve_date_picker;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_value_1 = (()=>{
        const _value = _init_value(this, '');
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_defaultValue_2 = (()=>{
        const _value = _init_defaultValue(this, '');
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_2;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_2 = _v;
    }
    #___private_placeholder_3 = (()=>{
        const _value = _init_placeholder(this, 'Select date');
        _init_extra__init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_3;
    }
    set placeholder(_v) {
        this.#___private_placeholder_3 = _v;
    }
    #___private_label_4 = (()=>{
        const _value = _init_label(this, '');
        _init_extra__init_label(this);
        return _value;
    })();
    get label() {
        return this.#___private_label_4;
    }
    set label(_v) {
        this.#___private_label_4 = _v;
    }
    #___private_description_5 = (()=>{
        const _value = _init_description(this, '');
        _init_extra__init_description(this);
        return _value;
    })();
    get description() {
        return this.#___private_description_5;
    }
    set description(_v) {
        this.#___private_description_5 = _v;
    }
    #___private_format_6 = (()=>{
        const _value = _init_format(this, '');
        _init_extra__init_format(this);
        return _value;
    })();
    get format() {
        return this.#___private_format_6;
    }
    set format(_v) {
        this.#___private_format_6 = _v;
    }
    #___private_size_7 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_7;
    }
    set size(_v) {
        this.#___private_size_7 = _v;
    }
    #___private_status_8 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_8;
    }
    set status(_v) {
        this.#___private_status_8 = _v;
    }
    #___private_disabled_9 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_9;
    }
    set disabled(_v) {
        this.#___private_disabled_9 = _v;
    }
    #___private_range_10 = (()=>{
        const _value = _init_range(this, false);
        _init_extra__init_range(this);
        return _value;
    })();
    get range() {
        return this.#___private_range_10;
    }
    set range(_v) {
        this.#___private_range_10 = _v;
    }
    #___private_loading_11 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_11;
    }
    set loading(_v) {
        this.#___private_loading_11 = _v;
    }
    #___private_allowClear_12 = (()=>{
        const _value = _init_allowClear(this, false);
        _init_extra__init_allowClear(this);
        return _value;
    })();
    get allowClear() {
        return this.#___private_allowClear_12;
    }
    set allowClear(_v) {
        this.#___private_allowClear_12 = _v;
    }
    #___private_open_13 = (()=>{
        const _value = _init_open(this, false);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_13;
    }
    set open(_v) {
        this.#___private_open_13 = _v;
    }
    #___private_min_14 = (()=>{
        const _value = _init_min(this, '');
        _init_extra__init_min(this);
        return _value;
    })();
    get min() {
        return this.#___private_min_14;
    }
    set min(_v) {
        this.#___private_min_14 = _v;
    }
    #___private_max_15 = (()=>{
        const _value = _init_max(this, '');
        _init_extra__init_max(this);
        return _value;
    })();
    get max() {
        return this.#___private_max_15;
    }
    set max(_v) {
        this.#___private_max_15 = _v;
    }
    #___private_disabledDate_16 = (()=>{
        const _value = _init_disabledDate(this, '');
        _init_extra__init_disabledDate(this);
        return _value;
    })();
    get disabledDate() {
        return this.#___private_disabledDate_16;
    }
    set disabledDate(_v) {
        this.#___private_disabledDate_16 = _v;
    }
    #___private_shortcuts_17 = (()=>{
        const _value = _init_shortcuts(this, '');
        _init_extra__init_shortcuts(this);
        return _value;
    })();
    get shortcuts() {
        return this.#___private_shortcuts_17;
    }
    set shortcuts(_v) {
        this.#___private_shortcuts_17 = _v;
    }
    #___private_shortcutPlacement_18 = (()=>{
        const _value = _init_shortcutPlacement(this, 'bottom');
        _init_extra__init_shortcutPlacement(this);
        return _value;
    })();
    get shortcutPlacement() {
        return this.#___private_shortcutPlacement_18;
    }
    set shortcutPlacement(_v) {
        this.#___private_shortcutPlacement_18 = _v;
    }
    #___private_defaultPickerValue_19 = (()=>{
        const _value = _init_defaultPickerValue(this, '');
        _init_extra__init_defaultPickerValue(this);
        return _value;
    })();
    get defaultPickerValue() {
        return this.#___private_defaultPickerValue_19;
    }
    set defaultPickerValue(_v) {
        this.#___private_defaultPickerValue_19 = _v;
    }
    #___private_pickerValue_20 = (()=>{
        const _value = _init_pickerValue(this, '');
        _init_extra__init_pickerValue(this);
        return _value;
    })();
    get pickerValue() {
        return this.#___private_pickerValue_20;
    }
    set pickerValue(_v) {
        this.#___private_pickerValue_20 = _v;
    }
    #___private_separator_21 = (()=>{
        const _value = _init_separator(this, ' ~ ');
        _init_extra__init_separator(this);
        return _value;
    })();
    get separator() {
        return this.#___private_separator_21;
    }
    set separator(_v) {
        this.#___private_separator_21 = _v;
    }
    #___private_popupMatchSelectWidth_22 = (()=>{
        const _value = _init_popupMatchSelectWidth(this, false);
        _init_extra__init_popupMatchSelectWidth(this);
        return _value;
    })();
    get popupMatchSelectWidth() {
        return this.#___private_popupMatchSelectWidth_22;
    }
    set popupMatchSelectWidth(_v) {
        this.#___private_popupMatchSelectWidth_22 = _v;
    }
    #___private_controlElement_23 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_23;
    }
    set controlElement(_v) {
        this.#___private_controlElement_23 = _v;
    }
    #___private_fieldElement_24 = (()=>{
        const _value = _init_fieldElement(this, null);
        _init_extra__init_fieldElement(this);
        return _value;
    })();
    get fieldElement() {
        return this.#___private_fieldElement_24;
    }
    set fieldElement(_v) {
        this.#___private_fieldElement_24 = _v;
    }
    #___private_panelElement_25 = (()=>{
        const _value = _init_panelElement(this, null);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_25;
    }
    set panelElement(_v) {
        this.#___private_panelElement_25 = _v;
    }
    #___private_panelPlacement_26 = (()=>{
        const _value = _init_panelPlacement(this, 'bottom-start');
        _init_extra__init_panelPlacement(this);
        return _value;
    })();
    get panelPlacement() {
        return this.#___private_panelPlacement_26;
    }
    set panelPlacement(_v) {
        this.#___private_panelPlacement_26 = _v;
    }
    #___private_hasPrefixSlot_27 = (()=>{
        const _value = _init_hasPrefixSlot(this, false);
        _init_extra__init_hasPrefixSlot(this);
        return _value;
    })();
    get hasPrefixSlot() {
        return this.#___private_hasPrefixSlot_27;
    }
    set hasPrefixSlot(_v) {
        this.#___private_hasPrefixSlot_27 = _v;
    }
    #___private_viewYear_28 = (()=>{
        const _value = _init_viewYear(this, getTodayParts().year);
        _init_extra__init_viewYear(this);
        return _value;
    })();
    get viewYear() {
        return this.#___private_viewYear_28;
    }
    set viewYear(_v) {
        this.#___private_viewYear_28 = _v;
    }
    #___private_viewMonth_29 = (()=>{
        const _value = _init_viewMonth(this, getTodayParts().month);
        _init_extra__init_viewMonth(this);
        return _value;
    })();
    get viewMonth() {
        return this.#___private_viewMonth_29;
    }
    set viewMonth(_v) {
        this.#___private_viewMonth_29 = _v;
    }
    #___private_activeRangeEnd_30 = (()=>{
        const _value = _init_activeRangeEnd(this, 'start');
        _init_extra__init_activeRangeEnd(this);
        return _value;
    })();
    get activeRangeEnd() {
        return this.#___private_activeRangeEnd_30;
    }
    set activeRangeEnd(_v) {
        this.#___private_activeRangeEnd_30 = _v;
    }
    hasInitializedDefaultValue = false;
    hasInitializedDefaultPickerValue = false;
    visibilityReason = 'api';
    repositionScheduler = createRafScheduler();
    positionRequestId = 0;
    cleanupAutoUpdate = null;
    gridId = '';
    descriptionId = '';
    constructor(){
        super();
        this.gridId = `ve-date-picker-grid-${Math.random().toString(36).slice(2)}`;
        this.descriptionId = `ve-date-picker-description-${Math.random().toString(36).slice(2)}`;
    }
    addEventListener(type, listener, options) {
        super.addEventListener(type, listener, options);
    }
    removeEventListener(type, listener, options) {
        super.removeEventListener(type, listener, options);
    }
    handleDocumentPointerDown = (event)=>{
        if (!this.open || event.composedPath().includes(this)) return;
        this.setOpen(false, 'outside');
    };
    handleDocumentKeydown = (event)=>{
        if (!this.open || 'Escape' !== event.key) return;
        event.stopPropagation();
        this.setOpen(false, 'escape');
        this.focus();
    };
    get normalizedFormat() {
        return (this.format ?? '').trim() || getDefaultFormat();
    }
    get normalizedValue() {
        return this.value ?? '';
    }
    get normalizedPickerValue() {
        return this.pickerValue ?? '';
    }
    get parsedValue() {
        return parseValue(this.primaryValue);
    }
    get parsedRangeValue() {
        const [start = '', end = ''] = this.rangeValues;
        return [
            parseValue(start),
            parseValue(end)
        ];
    }
    get rangeStartValue() {
        return this.rangeValues[0] ?? '';
    }
    get rangeEndValue() {
        return this.rangeValues[1] ?? '';
    }
    get primaryValue() {
        return this.range ? this.rangeValues[0] ?? '' : this.normalizedValue;
    }
    get rangeValues() {
        return this.normalizedValue.split(',').map((part)=>part.trim()).filter((part)=>part.length > 0).slice(0, 2);
    }
    get displayValue() {
        if (this.range) {
            if (!this.rangeStartValue && !this.rangeEndValue) return '';
            return `${this.rangeStartValue}${this.separator}${this.rangeEndValue}`;
        }
        const parsedValue = this.parsedValue;
        return parsedValue ? formatDateValue(parsedValue.parts, this.normalizedFormat) : '';
    }
    get hasDisplayValue() {
        return this.range ? Boolean(this.formattedRangeStartValue || this.formattedRangeEndValue) : Boolean(this.displayValue);
    }
    get formattedRangeStartValue() {
        const start = this.parsedRangeValue[0];
        return start ? formatDateValue(start.parts, this.normalizedFormat) : '';
    }
    get formattedRangeEndValue() {
        const end = this.parsedRangeValue[1];
        return end ? formatDateValue(end.parts, this.normalizedFormat) : '';
    }
    get minDateKey() {
        const value = parseDateValue(this.min);
        return value ? getDateKey(value) : void 0;
    }
    get maxDateKey() {
        const value = parseDateValue(this.max);
        return value ? getDateKey(value) : void 0;
    }
    get parsedShortcuts() {
        const trimmed = (this.shortcuts ?? '').trim();
        if (!trimmed) return [];
        try {
            const parsed = JSON.parse(trimmed);
            if (Array.isArray(parsed)) return parsed.map((item)=>{
                if ('object' != typeof item || null === item) return;
                const record = item;
                const label = record.label ?? record.text;
                const value = record.value;
                const endValue = record.endValue ?? record.end;
                if ('string' != typeof label || 'string' != typeof value) return;
                return this.createResolvedShortcut('string' == typeof label ? label : '', value, 'string' == typeof endValue ? endValue : void 0);
            }).filter((item)=>void 0 !== item);
        } catch  {}
        return trimmed.split(';').map((item)=>{
            const [label = '', rawValue = ''] = item.split(':');
            const [value = '', endValue = ''] = rawValue.split(',');
            return this.createResolvedShortcut(label.trim(), value.trim(), endValue.trim() || void 0);
        }).filter((item)=>void 0 !== item);
    }
    get shouldShowClear() {
        return this.allowClear && !this.disabled && !this.loading && '' !== this.value;
    }
    get hasLeadingLabel() {
        return (this.label ?? '').trim().length > 0;
    }
    get hasDescriptionFeedback() {
        return (this.description ?? '').trim().length > 0;
    }
    createResolvedShortcut(label, value, endValue) {
        const resolvedValue = this.normalizeShortcutValue(value);
        if (!label || !resolvedValue) return;
        const resolvedEndValue = endValue ? this.normalizeShortcutValue(endValue) : '';
        if (resolvedEndValue) return endValue ? {
            label,
            value,
            endValue,
            resolvedValue,
            resolvedEndValue
        } : {
            label,
            value,
            resolvedValue,
            resolvedEndValue
        };
        return {
            label,
            value,
            resolvedValue
        };
    }
    normalizeShortcutValue(value) {
        const resolvedParts = resolveRelativeShortcutParts(value);
        if (resolvedParts) return formatDateValue(resolvedParts, this.normalizedFormat);
        return this.normalizeSingleValue(value);
    }
    isShortcutSelected(shortcut) {
        if (this.range) return this.rangeValues[0] === shortcut.resolvedValue && (this.rangeValues[1] ?? '') === (shortcut.resolvedEndValue ?? '');
        return this.value === shortcut.resolvedValue;
    }
    get controlAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get controlAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get controlAriaDescribedby() {
        const describedBy = this.getAttribute('aria-describedby');
        const ids = [
            describedBy,
            this.hasDescriptionFeedback ? this.descriptionId : void 0
        ].filter((id)=>'string' == typeof id && id.length > 0);
        return ids.length > 0 ? ids.join(' ') : void 0;
    }
    get controlAriaRequired() {
        return normalizeAriaBooleanString(this.getAttribute('aria-required'));
    }
    get controlAriaInvalid() {
        return normalizeAriaBooleanString(this.getAttribute('aria-invalid')) ?? ('error' === this.status ? 'true' : void 0);
    }
    focus(options) {
        this.controlElement?.focus(options);
    }
    blur() {
        this.controlElement?.blur();
    }
    show() {
        this.setOpen(true, 'api');
    }
    hide() {
        this.setOpen(false, 'api');
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.applyDefaultValueIfNeeded();
        this.applyDefaultPickerValueIfNeeded();
        this.syncViewToSelectedDate();
        if (this.open) this.installGlobalListeners();
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('value') || changedProperties.has('format') || changedProperties.has('range')) {
            const normalizedValue = this.normalizeValue(this.normalizedValue);
            if (this.value !== normalizedValue) this.value = normalizedValue;
        }
        if (changedProperties.has('shortcutPlacement')) {
            const normalizedShortcutPlacement = 'left' === this.shortcutPlacement ? 'left' : 'bottom';
            if (this.shortcutPlacement !== normalizedShortcutPlacement) this.shortcutPlacement = normalizedShortcutPlacement;
        }
        if (changedProperties.has('pickerValue')) this.syncViewToPickerValue();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('disabled') && this.disabled && this.open) return void this.setOpen(false, 'api');
        if (changedProperties.has('open')) this.handleOpenChanged();
        if (changedProperties.has('value') && this.parsedValue) this.syncViewToSelectedDate();
        if (this.open) this.scheduleUpdatePosition();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    disconnectedCallback() {
        this.removeGlobalListeners();
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        this.teardownAutoUpdate();
        super.disconnectedCallback();
    }
    applyDefaultValueIfNeeded() {
        if (this.hasInitializedDefaultValue) return;
        this.hasInitializedDefaultValue = true;
        if (this.value || !this.defaultValue) return;
        this.value = this.normalizeValue(this.defaultValue);
    }
    applyDefaultPickerValueIfNeeded() {
        if (this.hasInitializedDefaultPickerValue) return;
        this.hasInitializedDefaultPickerValue = true;
        this.syncViewToPickerValue(this.defaultPickerValue);
    }
    normalizeValue(value) {
        const normalizedValue = value ?? '';
        if (this.range) return normalizedValue.split(',').map((part)=>this.normalizeSingleValue(part)).filter((part)=>part.length > 0).slice(0, 2).join(',');
        return this.normalizeSingleValue(normalizedValue);
    }
    normalizeSingleValue(value) {
        const parsedValue = parseValue(value);
        return parsedValue ? formatDateValue(parsedValue.parts, this.normalizedFormat) : '';
    }
    syncViewToSelectedDate() {
        const parts = (this.range ? this.parsedRangeValue[0]?.parts : this.parsedValue?.parts) ?? getTodayParts();
        this.viewYear = parts.year;
        this.viewMonth = parts.month;
    }
    syncViewToOpenDate() {
        const parsedPickerValue = parseValue(this.normalizedPickerValue || this.defaultPickerValue);
        const parts = parsedPickerValue?.parts ?? getTodayParts();
        this.viewYear = parts.year;
        this.viewMonth = parts.month;
    }
    syncViewToPickerValue(value = this.normalizedPickerValue) {
        const parsed = parseValue(value);
        if (!parsed) return;
        this.viewYear = parsed.parts.year;
        this.viewMonth = parsed.parts.month;
    }
    getUnitKey(parts) {
        return getDateKey(parts);
    }
    getDisabledRuleKey(value) {
        const parsed = parseValue(value);
        return parsed ? this.getUnitKey(parsed.parts) : void 0;
    }
    isDateDisabled(parts) {
        const dateKey = this.getUnitKey(parts);
        const minDateKey = this.getDisabledRuleKey(this.min);
        const maxDateKey = this.getDisabledRuleKey(this.max);
        const ruleDisabled = this.isDateDisabledByRules(parts);
        return void 0 !== minDateKey && dateKey < minDateKey || void 0 !== maxDateKey && dateKey > maxDateKey || void 0 !== minDateKey && void 0 !== maxDateKey && minDateKey > maxDateKey || ruleDisabled;
    }
    isDateDisabledByRules(parts) {
        const rules = (this.disabledDate ?? '').trim();
        if (!rules) return false;
        const key = this.getUnitKey(parts);
        return rules.split(';').some((rule)=>{
            const [name = '', rawValue = ''] = rule.split(':');
            const values = rawValue.split(',').map((value)=>this.getDisabledRuleKey(value.trim())).filter((value)=>void 0 !== value);
            if ('before' === name.trim()) return void 0 !== values[0] && key < values[0];
            if ('after' === name.trim()) return void 0 !== values[0] && key > values[0];
            if ('dates' === name.trim() || 'values' === name.trim()) return values.includes(key);
            return false;
        });
    }
    isPartsSelected(parts) {
        const selected = this.range ? this.parsedRangeValue : [
            this.parsedValue
        ];
        const key = this.getUnitKey(parts);
        return selected.some((value)=>void 0 !== value && this.getUnitKey(value.parts) === key);
    }
    isPartsInRange(parts) {
        if (!this.range) return false;
        const [start, end] = this.parsedRangeValue;
        if (!start || !end) return false;
        const key = this.getUnitKey(parts);
        const startKey = this.getUnitKey(start.parts);
        const endKey = this.getUnitKey(end.parts);
        return key >= Math.min(startKey, endKey) && key <= Math.max(startKey, endKey);
    }
    createDateCells(viewYear = this.viewYear, viewMonth = this.viewMonth) {
        const firstDay = new Date(viewYear, viewMonth - 1, 1);
        const leadingDays = firstDay.getDay();
        const daysInMonth = new Date(viewYear, viewMonth, 0).getDate();
        const rowCount = Math.ceil((leadingDays + daysInMonth) / 7);
        const cellCount = 7 * rowCount;
        const gridStart = new Date(viewYear, viewMonth - 1, 1 - leadingDays);
        const today = getTodayParts();
        const [start, end] = this.parsedRangeValue;
        const startKey = start ? this.getUnitKey(start.parts) : void 0;
        const endKey = end ? this.getUnitKey(end.parts) : void 0;
        const cells = [];
        for(let index = 0; index < cellCount; index += 1){
            const date = new Date(gridStart.getFullYear(), gridStart.getMonth(), gridStart.getDate() + index);
            const parts = {
                year: date.getFullYear(),
                month: date.getMonth() + 1,
                day: date.getDate()
            };
            const outside = parts.month !== viewMonth;
            cells.push({
                value: formatDateValue(parts, this.normalizedFormat),
                date,
                parts,
                label: this.range && outside ? '' : String(parts.day),
                outside,
                today: areSameDate(parts, today),
                selected: outside && this.range ? false : this.isPartsSelected(parts),
                inRange: outside && this.range ? false : this.isPartsInRange(parts),
                disabled: this.isDateDisabled(parts) || this.range && outside,
                rangeStart: (!outside || !this.range) && void 0 !== startKey && this.getUnitKey(parts) === startKey,
                rangeEnd: (!outside || !this.range) && void 0 !== endKey && this.getUnitKey(parts) === endKey
            });
        }
        return cells;
    }
    selectDate(parts, source) {
        if (this.disabled || this.loading || this.isDateDisabled(parts)) return;
        const nextValue = formatDateValue(parts, this.normalizedFormat);
        if (this.range) return void this.selectRangeValue(nextValue, source);
        this.value = nextValue;
        this.emitChange(parts, source);
        this.setOpen(false, 'select');
    }
    selectRangeValue(nextValue, source) {
        const values = this.rangeValues;
        if ('start' === this.activeRangeEnd || 0 === values.length) {
            this.value = nextValue;
            this.activeRangeEnd = 'end';
            return;
        }
        const nextValues = [
            values[0] ?? nextValue,
            nextValue
        ];
        const [start, end] = nextValues.map((value)=>parseValue(value));
        if (start && end && this.getUnitKey(start.parts) > this.getUnitKey(end.parts)) nextValues.reverse();
        this.value = nextValues.join(',');
        this.activeRangeEnd = 'start';
        this.emitChange(this.parsedRangeValue, source);
        this.setOpen(false, 'select');
    }
    clearValue(event) {
        event?.preventDefault();
        event?.stopPropagation();
        if (this.disabled || this.loading) return;
        this.value = '';
        this.activeRangeEnd = 'start';
        this.dispatchEvent(new CustomEvent('ve-clear', {
            detail: {
                value: ''
            },
            bubbles: true,
            composed: true
        }));
    }
    emitChange(value, source) {
        const date = Array.isArray(value) ? value.map((item)=>item?.parts).filter((item)=>void 0 !== item).map((parts)=>createDateFromParts(parts)) : createDateFromParts(value);
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail: {
                value: this.value,
                date,
                source
            },
            bubbles: true,
            composed: true
        }));
    }
    setOpen(open, reason) {
        if (this.disabled || this.loading) open = false;
        if (this.open === open) return;
        this.visibilityReason = reason;
        this.open = open;
    }
    handleOpenChanged() {
        if (this.open) {
            this.syncViewToOpenDate();
            this.installGlobalListeners();
            this.updateComplete.then(()=>{
                if (!this.isConnected || !this.open || this.disabled || this.loading) return;
                showPanelPopover(this.panelElement);
                this.setupAutoUpdate();
                this.scheduleUpdatePosition();
            });
        } else {
            hidePanelPopover(this.panelElement);
            this.removeGlobalListeners();
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            this.teardownAutoUpdate();
        }
        this.dispatchEvent(new CustomEvent('ve-visible-change', {
            detail: {
                open: this.open,
                reason: this.visibilityReason
            },
            bubbles: true,
            composed: true
        }));
    }
    installGlobalListeners() {
        document.addEventListener('pointerdown', this.handleDocumentPointerDown);
        document.addEventListener('keydown', this.handleDocumentKeydown);
    }
    removeGlobalListeners() {
        document.removeEventListener('pointerdown', this.handleDocumentPointerDown);
        document.removeEventListener('keydown', this.handleDocumentKeydown);
    }
    setupAutoUpdate() {
        if (!this.isConnected || !this.open || !this.controlElement || !this.panelElement) return;
        this.teardownAutoUpdate();
        this.cleanupAutoUpdate = createFloatingAutoUpdate(this.fieldElement ?? this.controlElement, this.panelElement, ()=>this.scheduleUpdatePosition());
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    async updatePosition() {
        if (!this.open || !this.isConnected || !this.controlElement || !this.panelElement) return;
        const requestId = ++this.positionRequestId;
        const referenceElement = this.fieldElement ?? this.controlElement;
        if (this.popupMatchSelectWidth) {
            const rect = referenceElement.getBoundingClientRect();
            this.panelElement.style.minWidth = `${rect.width}px`;
        } else this.panelElement.style.minWidth = '';
        const styles = window.getComputedStyle(this);
        const gapValue = Number.parseFloat(styles.getPropertyValue('--ve-date-picker-popup-gap')) || 4;
        const result = await computeFloatingPanelPosition({
            reference: referenceElement,
            floating: this.panelElement,
            placement: 'bottom-start',
            distance: gapValue
        });
        if (requestId !== this.positionRequestId || !this.open || !this.isConnected || !this.panelElement) return;
        this.panelPlacement = result.placement;
        Object.assign(this.panelElement.style, {
            left: '0',
            top: '0',
            transform: `translate3d(${Math.round(result.x)}px, ${Math.round(result.y)}px, 0)`
        });
    }
    handleControlClick = ()=>{
        if (this.disabled) return;
        this.setOpen(!this.open, 'trigger');
    };
    handleControlKeydown = (event)=>{
        if (this.disabled || event.target !== this.controlElement) return;
        switch(event.key){
            case 'Enter':
            case ' ':
            case 'ArrowDown':
                event.preventDefault();
                this.setOpen(true, 'trigger');
                break;
            case 'Escape':
                if (this.open) {
                    event.preventDefault();
                    this.setOpen(false, 'escape');
                }
                break;
            default:
                break;
        }
    };
    handlePrefixSlotChange = (event)=>{
        const slot = event.currentTarget;
        this.hasPrefixSlot = (slot?.assignedNodes({
            flatten: true
        }).filter((node)=>node.nodeType !== Node.TEXT_NODE || (node.textContent?.trim().length ?? 0) > 0).length ?? 0) > 0;
    };
    handleClearClick = (event)=>{
        this.clearValue(event);
    };
    handleClearPointerDown = (event)=>{
        event.preventDefault();
        event.stopPropagation();
    };
    createRangeSegmentClickHandler(rangeEnd) {
        return (event)=>{
            event.stopPropagation();
            if (this.disabled) return;
            this.activeRangeEnd = rangeEnd;
            this.setOpen(true, 'trigger');
        };
    }
    createCellClickHandler(cell) {
        return ()=>this.selectDate(cell.parts, 'select');
    }
    createCellKeydownHandler(cell) {
        return (event)=>{
            if ('Enter' !== event.key && ' ' !== event.key) return;
            event.preventDefault();
            this.selectDate(cell.parts, 'select');
        };
    }
    createMonthNavigationHandler(offset) {
        return ()=>{
            const next = addMonths(this.viewYear, this.viewMonth, offset);
            this.viewYear = next.year;
            this.viewMonth = next.month;
        };
    }
    createShortcutClickHandler(shortcut) {
        return ()=>{
            const start = shortcut.resolvedValue;
            const end = shortcut.resolvedEndValue ?? '';
            if (!start) return;
            this.value = this.range && end ? `${start},${end}` : start;
            this.activeRangeEnd = 'start';
            if (this.range) this.emitChange(this.parsedRangeValue, 'shortcut');
            else if (this.parsedValue) this.emitChange(this.parsedValue.parts, 'shortcut');
            this.setOpen(false, 'select');
        };
    }
    renderLabel() {
        const label = (this.label ?? '').trim();
        if (!label) return nothing;
        return html`
      <span class="ve-date-picker-label" part="label">
        <span class="ve-date-picker-label-text">${label}</span>
      </span>
    `;
    }
    renderPrefix() {
        return html`<span
      class=${classMap({
            've-date-picker-prefix': true,
            've-date-picker-prefix-empty': !this.hasPrefixSlot
        })}
      ><slot name="prefix" @slotchange=${this.handlePrefixSlotChange}></slot
    ></span>`;
    }
    renderValue() {
        if (this.range) {
            const startValue = this.formattedRangeStartValue;
            const endValue = this.formattedRangeEndValue;
            const hasValue = Boolean(startValue || endValue);
            if (!hasValue) return html`<span
          class="ve-date-picker-value ve-date-picker-value-placeholder"
          >${this.placeholder}</span
        >`;
            return html`
        <span class="ve-date-picker-range-value">
          <span
            class=${classMap({
                've-date-picker-range-segment': true,
                've-date-picker-range-segment-active': this.open && 'start' === this.activeRangeEnd,
                've-date-picker-range-segment-placeholder': !startValue
            })}
            @click=${this.createRangeSegmentClickHandler('start')}
          >
            ${startValue || 'Start date'}
          </span>
          <span class="ve-date-picker-range-separator">${this.separator}</span>
          <span
            class=${classMap({
                've-date-picker-range-segment': true,
                've-date-picker-range-segment-active': this.open && 'end' === this.activeRangeEnd,
                've-date-picker-range-segment-placeholder': !endValue
            })}
            @click=${this.createRangeSegmentClickHandler('end')}
          >
            ${endValue || 'End date'}
          </span>
        </span>
      `;
        }
        const displayValue = this.displayValue;
        return html`<span
      class=${classMap({
            've-date-picker-value': true,
            've-date-picker-value-placeholder': !displayValue
        })}
      >${displayValue || this.placeholder}</span
    >`;
    }
    renderActions() {
        return html`
      <span class="ve-date-picker-actions">
        ${this.shouldShowClear ? html`<button
              class="ve-date-picker-clear"
              part="clear"
              type="button"
              aria-label="Clear"
              @pointerdown=${this.handleClearPointerDown}
              @click=${this.handleClearClick}
            >
              <ve-icon .definition=${closeIcon}></ve-icon>
            </button>` : nothing}
        ${this.loading ? html`<span class="ve-date-picker-loading" aria-hidden="true">
              <ve-icon .definition=${loadingIcon}></ve-icon>
            </span>` : html`<span class="ve-date-picker-arrow" aria-hidden="true">
              <ve-icon .definition=${dateCalendarIcon}></ve-icon>
            </span>`}
      </span>
    `;
    }
    renderPanelHeader(viewYear = this.viewYear, viewMonth = this.viewMonth) {
        const headerLabel = `${viewYear}年${toTwoDigit(viewMonth)}月`;
        return html`
      <div class="ve-date-picker-header">
        <span class="ve-date-picker-header-label"> ${headerLabel} </span>
        <div class="ve-date-picker-navigation-group">
          <button
            class="ve-date-picker-navigation"
            type="button"
            aria-label="Previous month"
            @click=${this.createMonthNavigationHandler(-1)}
          >
            <ve-icon .definition=${chevronLeftSmIcon}></ve-icon>
          </button>
          <button
            class="ve-date-picker-navigation"
            type="button"
            aria-label="Next month"
            @click=${this.createMonthNavigationHandler(1)}
          >
            <ve-icon .definition=${chevronRightSmIcon}></ve-icon>
          </button>
        </div>
      </div>
    `;
    }
    renderWeekdays() {
        return html`<div class="ve-date-picker-weekdays" aria-hidden="true">
      ${WEEKDAY_LABELS.map((weekday)=>html`<span class="ve-date-picker-weekday">${weekday}</span>`)}
    </div>`;
    }
    renderGrid(viewYear = this.viewYear, viewMonth = this.viewMonth, gridId = this.gridId) {
        const cells = this.createDateCells(viewYear, viewMonth);
        return html`
      <div
        id=${gridId}
        class=${classMap({
            've-date-picker-grid': true,
            've-date-picker-grid-range': this.range
        })}
        role="grid"
      >
        ${cells.map((cell)=>html`<button
            class=${classMap({
                've-date-picker-cell': true,
                've-date-picker-cell-outside': cell.outside,
                've-date-picker-cell-today': cell.today,
                've-date-picker-cell-selected': cell.selected,
                've-date-picker-cell-in-range': cell.inRange,
                've-date-picker-cell-disabled': cell.disabled,
                've-date-picker-cell-range-start': cell.rangeStart,
                've-date-picker-cell-range-end': cell.rangeEnd
            })}
            part="cell"
            type="button"
            role="gridcell"
            aria-label=${cell.value}
            aria-selected=${cell.selected ? 'true' : 'false'}
            aria-disabled=${cell.disabled ? 'true' : 'false'}
            title=${cell.label ? cell.value : ''}
            ?disabled=${cell.disabled}
            @click=${this.createCellClickHandler(cell)}
            @keydown=${this.createCellKeydownHandler(cell)}
          >
            <span class="ve-date-picker-cell-label">${cell.label}</span>
          </button>`)}
      </div>
    `;
    }
    renderCalendar(viewYear = this.viewYear, viewMonth = this.viewMonth, gridId = this.gridId) {
        return html`
      <div class="ve-date-picker-calendar">
        ${this.renderPanelHeader(viewYear, viewMonth)} ${this.renderWeekdays()}
        ${this.renderGrid(viewYear, viewMonth, gridId)}
      </div>
    `;
    }
    renderShortcuts() {
        const shortcuts = this.parsedShortcuts;
        if (0 === shortcuts.length) return nothing;
        return html`<div class="ve-date-picker-shortcuts">
      ${shortcuts.map((shortcut)=>html`<button
            class=${classMap({
                've-date-picker-shortcut': true,
                've-date-picker-shortcut-selected': this.isShortcutSelected(shortcut)
            })}
            type="button"
            title=${shortcut.label}
            aria-pressed=${this.isShortcutSelected(shortcut) ? 'true' : 'false'}
            @click=${this.createShortcutClickHandler(shortcut)}
          >
            ${shortcut.label}
          </button>`)}
    </div>`;
    }
    renderPanel() {
        if (!this.open) return nothing;
        return html`
      <div
        class=${classMap({
            've-date-picker-panel': true,
            've-date-picker-panel-with-left-shortcuts': 'left' === this.shortcutPlacement && this.parsedShortcuts.length > 0
        })}
        part="panel"
        role="dialog"
        data-placement=${this.panelPlacement}
      >
        ${'left' === this.shortcutPlacement ? this.renderShortcuts() : nothing}
        <div class="ve-date-picker-panel-main">
          ${this.range ? html`<div class="ve-date-picker-calendar-group">
                ${this.renderCalendar(this.viewYear, this.viewMonth, this.gridId)}
                ${(()=>{
            const next = addMonths(this.viewYear, this.viewMonth, 1);
            return this.renderCalendar(next.year, next.month, `${this.gridId}-end`);
        })()}
              </div>` : this.renderCalendar()}
          ${'bottom' === this.shortcutPlacement ? this.renderShortcuts() : nothing}
        </div>
      </div>
    `;
    }
    renderDescription() {
        const description = (this.description ?? '').trim();
        if (!description) return nothing;
        return html`<div
      id=${this.descriptionId}
      class="ve-date-picker-description"
      part="description"
    >
      ${description}
    </div>`;
    }
    render() {
        return html`
      <div
        class=${classMap({
            've-date-picker': true,
            've-date-picker-with-label': this.hasLeadingLabel
        })}
      >
        <div class="ve-date-picker-field">
          ${this.renderLabel()}
          <div
            class=${classMap({
            've-date-picker-control': true,
            've-date-picker-control-filled': this.hasDisplayValue
        })}
            part="control"
            role="combobox"
            tabindex=${this.disabled ? '-1' : '0'}
            aria-haspopup="dialog"
            aria-expanded=${this.open ? 'true' : 'false'}
            aria-controls=${ifDefined(this.open ? this.gridId : void 0)}
            aria-label=${ifDefined(this.controlAriaLabel)}
            aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
            aria-describedby=${ifDefined(this.controlAriaDescribedby)}
            aria-required=${ifDefined(this.controlAriaRequired)}
            aria-invalid=${ifDefined(this.controlAriaInvalid)}
            @click=${this.handleControlClick}
            @keydown=${this.handleControlKeydown}
          >
            ${this.renderPrefix()} ${this.renderValue()} ${this.renderActions()}
          </div>
        </div>
        ${this.renderDescription()}
      </div>
      ${this.renderPanel()}
    `;
    }
}
defineCustomElement(VE_DATE_PICKER_TAG, VeDatePickerElement);
export { VE_DATE_PICKER_TAG, VeDatePickerElement };
