// @ve-design/web/web/dist/ve-input.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { styleMap } from "lit/directives/style-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { closeIcon } from "./icons/close.js";
const veInputTokenDefaults = css`
  :host {
    --ve-input-font-family: var(--font-sans);
    --ve-input-font-weight: var(--font-weight-normal);
    --ve-input-line-height-default: var(--line-height-body);
    --ve-input-transition-duration: var(--motion-fast);
    --ve-input-transition-timing: var(--ease-standard);

    --ve-input-height-small: var(--height-small);
    --ve-input-height-default: var(--height-default);
    --ve-input-height-large: var(--height-large);

    --ve-input-font-size-small: var(--text-body-sm);
    --ve-input-font-size-default: var(--text-body);
    --ve-input-font-size-large: var(--text-body);

    --ve-input-padding-inline-small: var(--space-xxs);
    --ve-input-padding-inline-default: var(--space-xxs);
    --ve-input-padding-inline-large: var(--space-xxs);
    --ve-input-padding-block-small: var(--space-xxxs);
    --ve-input-padding-block-default: var(--space-xxxs);
    --ve-input-padding-block-large: var(--space-xxxs);

    --ve-input-radius: var(--radius-md);
    --ve-input-border-width: var(--stroke-weight-base);

    --ve-input-bg: var(--color-bg-base);
    --ve-input-bg-hover: var(--color-bg-base);
    --ve-input-bg-focus: var(--color-bg-base);

    --ve-input-border-color: var(--color-border-strong);
    --ve-input-border-color-hover: var(--color-border-interactive);
    --ve-input-border-color-focus: var(--color-border-interactive);
    --ve-input-shadow-focus-strong-color: rgb(
      from var(--color-border-interactive) r g b / 25%
    );
    --ve-input-shadow-focus-soft-color: rgb(
      from var(--color-border-interactive) r g b / 10%
    );
    --ve-input-shadow-focus:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-shadow-focus-strong-color),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-shadow-focus-soft-color);
    --ve-input-focus-ring-color: var(--ve-input-shadow-focus-strong-color);
    --ve-input-focus-shadow-size: var(--stroke-weight-bold);

    --ve-input-status-error-bg: var(--color-bg-danger-secondary);
    --ve-input-status-error-bg-hover: color-mix(
      var(--ve-input-status-error-bg) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-input-status-error-border: var(--color-border-danger);
    --ve-input-status-error-border-hover: var(--color-text-danger);
    --ve-input-status-error-border-focus: var(--color-text-danger);
    --ve-input-status-error-shadow-strong-color: rgb(
      from var(--color-text-danger) r g b / 25%
    );
    --ve-input-status-error-shadow-soft-color: rgb(
      from var(--color-text-danger) r g b / 10%
    );
    --ve-input-status-error-shadow-focus:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-status-error-shadow-strong-color),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-status-error-shadow-soft-color);

    --ve-input-status-warning-bg: var(--color-bg-warning-secondary);
    --ve-input-status-warning-bg-hover: color-mix(
      var(--ve-input-status-warning-bg) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-input-status-warning-border: var(--color-border-warning);
    --ve-input-status-warning-border-hover: var(--color-text-warning);
    --ve-input-status-warning-border-focus: var(--color-text-warning);
    --ve-input-status-warning-shadow-strong-color: rgb(
      from var(--color-text-warning) r g b / 25%
    );
    --ve-input-status-warning-shadow-soft-color: rgb(
      from var(--color-text-warning) r g b / 10%
    );
    --ve-input-status-warning-shadow-focus:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-status-warning-shadow-strong-color),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-status-warning-shadow-soft-color);

    --ve-input-status-success-bg: var(--color-bg-success-secondary);
    --ve-input-status-success-bg-hover: color-mix(
      var(--ve-input-status-success-bg) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-input-status-success-border: var(--color-border-success);
    --ve-input-status-success-border-hover: var(--color-text-success);
    --ve-input-status-success-border-focus: var(--color-text-success);
    --ve-input-status-success-shadow-strong-color: rgb(
      from var(--color-text-success) r g b / 25%
    );
    --ve-input-status-success-shadow-soft-color: rgb(
      from var(--color-text-success) r g b / 10%
    );
    --ve-input-status-success-shadow-focus:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-status-success-shadow-strong-color),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        var(--ve-input-status-success-shadow-soft-color);

    --ve-input-disabled-bg: var(--color-bg-control);
    --ve-input-disabled-border-color: var(--color-border-default);
    --ve-input-disabled-text: var(--color-text-disable);
    --ve-input-disabled-opacity: var(--state-disabled-ratio);

    --ve-input-text-color: var(--color-text-primary);
    --ve-input-placeholder-color: var(--color-text-disable);

    --ve-input-prefix-color: var(--color-text-tertiary);
    --ve-input-suffix-color: var(--color-text-tertiary);
    --ve-input-affix-gap: var(--space-xxxs);

    --ve-input-icon-size: var(--text-caption);
    --ve-input-action-size: calc(var(--height-small) - var(--space-xxs));
    --ve-input-icon-color: var(--color-text-secondary);
    --ve-input-icon-hover-bg: color-mix(
      var(--ve-input-bg) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );

    --ve-input-word-limit-color: var(--color-text-tertiary);
    --ve-input-word-limit-error-color: var(--color-text-danger);
    --ve-input-word-limit-font-size: var(--text-caption);

    --ve-input-textarea-padding-inline: var(--space-xs);
    --ve-input-textarea-padding-block: var(--space-xxxs);
    --ve-input-textarea-word-limit-offset-inline: var(--space-xs);
    --ve-input-textarea-word-limit-offset-block: var(--space-xxs);
    --ve-input-textarea-clear-offset-inline: var(--space-xs);
    --ve-input-textarea-clear-offset-block: var(--space-xs);
    --ve-input-current-height: var(--ve-input-height-default);
    --ve-input-current-font-size: var(--ve-input-font-size-default);
    --ve-input-current-line-height: var(--ve-input-line-height-default);
    --ve-input-current-padding-inline: var(--ve-input-padding-inline-default);
    --ve-input-current-padding-block: var(--ve-input-padding-block-default);
  }

  :host([size='small']) {
    --ve-input-current-height: var(--ve-input-height-small);
    --ve-input-current-font-size: var(--ve-input-font-size-small);
    --ve-input-current-padding-inline: var(--ve-input-padding-inline-small);
    --ve-input-current-padding-block: var(--ve-input-padding-block-small);
  }

  :host([size='large']) {
    --ve-input-current-height: var(--ve-input-height-large);
    --ve-input-current-font-size: var(--ve-input-font-size-large);
    --ve-input-current-padding-inline: var(--ve-input-padding-inline-large);
    --ve-input-current-padding-block: var(--ve-input-padding-block-large);
  }
`;
const token = veInputTokenDefaults;
const ve_input = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      inline-size: 100%;
      vertical-align: top;
      box-sizing: border-box;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-input-wrapper {
      position: relative;
      display: inline-flex;
      align-items: center;
      inline-size: 100%;
      min-inline-size: 0;
      block-size: var(--ve-input-current-height);
      box-sizing: border-box;
      padding-inline: var(--ve-input-current-padding-inline);
      border: var(--ve-input-border-width) solid var(--ve-input-border-color);
      border-radius: var(--ve-input-radius);
      background: var(--ve-input-bg);
      color: var(--ve-input-text-color);
      font-family: var(--ve-input-font-family);
      font-size: var(--ve-input-current-font-size);
      font-weight: var(--ve-input-font-weight);
      line-height: var(--ve-input-current-line-height);
      transition:
        border-color var(--ve-input-transition-duration)
          var(--ve-input-transition-timing),
        background-color var(--ve-input-transition-duration)
          var(--ve-input-transition-timing),
        box-shadow var(--ve-input-transition-duration)
          var(--ve-input-transition-timing),
        color var(--ve-input-transition-duration)
          var(--ve-input-transition-timing);
    }

    .ve-input-wrapper--auto-width {
      inline-size: auto;
      max-inline-size: 100%;
      overflow: hidden;
    }

    .ve-input-wrapper:not(.ve-input-wrapper--disabled):hover {
      border-color: var(--ve-input-border-color-hover);
      background: var(--ve-input-bg-hover);
    }

    .ve-input-wrapper:not(.ve-input-wrapper--disabled):focus-within {
      border-color: var(--ve-input-border-color-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-shadow-focus);
    }

    .ve-input-wrapper--error {
      background: var(--ve-input-status-error-bg);
      border-color: var(--ve-input-status-error-border);
    }

    .ve-input-wrapper--error:not(.ve-input-wrapper--disabled):hover {
      background: var(--ve-input-status-error-bg-hover);
      border-color: var(--ve-input-status-error-border-hover);
    }

    .ve-input-wrapper--error:not(.ve-input-wrapper--disabled):focus-within {
      border-color: var(--ve-input-status-error-border-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-status-error-shadow-focus);
    }

    .ve-input-wrapper--warning {
      background: var(--ve-input-status-warning-bg);
      border-color: var(--ve-input-status-warning-border);
    }

    .ve-input-wrapper--warning:not(.ve-input-wrapper--disabled):hover {
      background: var(--ve-input-status-warning-bg-hover);
      border-color: var(--ve-input-status-warning-border-hover);
    }

    .ve-input-wrapper--warning:not(.ve-input-wrapper--disabled):focus-within {
      border-color: var(--ve-input-status-warning-border-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-status-warning-shadow-focus);
    }

    .ve-input-wrapper--success {
      background: var(--ve-input-status-success-bg);
      border-color: var(--ve-input-status-success-border);
    }

    .ve-input-wrapper--success:not(.ve-input-wrapper--disabled):hover {
      background: var(--ve-input-status-success-bg-hover);
      border-color: var(--ve-input-status-success-border-hover);
    }

    .ve-input-wrapper--success:not(.ve-input-wrapper--disabled):focus-within {
      border-color: var(--ve-input-status-success-border-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-status-success-shadow-focus);
    }

    .ve-input-wrapper--disabled,
    .ve-input-wrapper--disabled:hover,
    .ve-input-wrapper--disabled:focus-within {
      background: var(--ve-input-disabled-bg);
      border-color: var(--ve-input-disabled-border-color);
      color: var(--ve-input-disabled-text);
      box-shadow: none;
      cursor: not-allowed;
    }

    .ve-input-control {
      inline-size: 100%;
      min-inline-size: 0;
      border: none;
      outline: none;
      background: transparent;
      color: inherit;
      font: inherit;
      line-height: inherit;
      padding: 0;
      margin: 0;
      flex: 1 1 auto;
      box-sizing: border-box;
      appearance: none;
      -webkit-tap-highlight-color: transparent;
    }

    .ve-input-control::placeholder {
      color: var(--ve-input-placeholder-color);
    }

    .ve-input-control:disabled {
      color: var(--ve-input-disabled-text);
      -webkit-text-fill-color: var(--ve-input-disabled-text);
      cursor: not-allowed;
    }

    .ve-input-wrapper--disabled .ve-input-control::placeholder {
      color: var(--ve-input-disabled-text);
    }

    .ve-input-field {
      padding-block: var(--ve-input-current-padding-block);
      padding-inline: 0;
      line-height: var(--ve-input-current-line-height);
      text-overflow: ellipsis;
    }

    .ve-input-wrapper--auto-width .ve-input-field {
      flex: 1 1 auto;
      max-inline-size: 100%;
    }

    .ve-input-prefix,
    .ve-input-suffix {
      display: inline-flex;
      align-items: center;
      block-size: 100%;
      white-space: nowrap;
      color: var(--ve-input-prefix-color);
      user-select: none;
      flex: none;
    }

    .ve-input-wrapper--has-prefix > .ve-input-field {
      padding-inline-start: var(--ve-input-affix-gap);
    }

    .ve-input-suffix {
      color: var(--ve-input-suffix-color);
    }

    .ve-input-word-limit {
      font-size: var(--ve-input-word-limit-font-size);
      color: var(--ve-input-word-limit-color);
      font-variant-numeric: tabular-nums;
      padding-inline-start: var(--space-xxs);
    }

    .ve-input-word-limit--error {
      color: var(--ve-input-word-limit-error-color);
    }

    .ve-input-clear {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-input-action-size);
      block-size: var(--ve-input-action-size);
      border: none;
      border-radius: var(--radius-full);
      background: transparent;
      color: var(--ve-input-icon-color);
      cursor: pointer;
      padding: 0;
      margin-inline-start: 0;
      flex: none;
      visibility: hidden;
      transition:
        color var(--ve-input-transition-duration)
          var(--ve-input-transition-timing),
        background-color var(--ve-input-transition-duration)
          var(--ve-input-transition-timing);
    }

    .ve-input-wrapper:hover .ve-input-clear,
    .ve-input-wrapper:focus-within .ve-input-clear {
      visibility: visible;
    }

    .ve-input-clear:hover {
      background: var(--ve-input-icon-hover-bg);
    }

    .ve-input-clear:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--ve-input-focus-shadow-size)
        var(--ve-input-focus-ring-color);
    }

    .ve-input-clear ve-icon {
      --ve-icon-size: var(--ve-input-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-input-mirror {
      position: absolute;
      inset-block-start: 0;
      inset-inline-start: 0;
      white-space: pre;
      visibility: hidden;
      pointer-events: none;
      user-select: none;
      font: inherit;
      letter-spacing: inherit;
      text-transform: inherit;
      text-indent: inherit;
    }

    .ve-textarea-wrapper {
      position: relative;
      display: inline-block;
      inline-size: 100%;
      vertical-align: top;
      box-sizing: border-box;
    }

    .ve-textarea-wrapper--word-limit-outside {
      display: inline-flex;
      flex-direction: column;
      align-items: stretch;
    }

    .ve-input-textarea {
      display: block;
      inline-size: 100%;
      min-block-size: var(--ve-input-current-height);
      box-sizing: border-box;
      border: var(--ve-input-border-width) solid var(--ve-input-border-color);
      border-radius: var(--ve-input-radius);
      padding: var(--ve-input-textarea-padding-block)
        var(--ve-input-textarea-padding-inline);
      background: var(--ve-input-bg);
      color: var(--ve-input-text-color);
      font-family: var(--ve-input-font-family);
      font-size: var(--ve-input-current-font-size);
      font-weight: var(--ve-input-font-weight);
      line-height: var(--ve-input-line-height-default);
      outline: none;
      resize: vertical;
      transition:
        border-color var(--ve-input-transition-duration)
          var(--ve-input-transition-timing),
        background-color var(--ve-input-transition-duration)
          var(--ve-input-transition-timing),
        box-shadow var(--ve-input-transition-duration)
          var(--ve-input-transition-timing);
    }

    .ve-input-textarea::placeholder {
      color: var(--ve-input-placeholder-color);
    }

    .ve-input-textarea:not(:disabled):hover {
      border-color: var(--ve-input-border-color-hover);
      background: var(--ve-input-bg-hover);
    }

    .ve-input-textarea:not(:disabled):focus {
      border-color: var(--ve-input-border-color-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-shadow-focus);
    }

    .ve-textarea-wrapper--error .ve-input-textarea {
      background: var(--ve-input-status-error-bg);
      border-color: var(--ve-input-status-error-border);
    }

    .ve-textarea-wrapper--error .ve-input-textarea:not(:disabled):hover {
      background: var(--ve-input-status-error-bg-hover);
      border-color: var(--ve-input-status-error-border-hover);
    }

    .ve-textarea-wrapper--error .ve-input-textarea:not(:disabled):focus {
      border-color: var(--ve-input-status-error-border-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-status-error-shadow-focus);
    }

    .ve-textarea-wrapper--warning .ve-input-textarea {
      background: var(--ve-input-status-warning-bg);
      border-color: var(--ve-input-status-warning-border);
    }

    .ve-textarea-wrapper--warning .ve-input-textarea:not(:disabled):hover {
      background: var(--ve-input-status-warning-bg-hover);
      border-color: var(--ve-input-status-warning-border-hover);
    }

    .ve-textarea-wrapper--warning .ve-input-textarea:not(:disabled):focus {
      border-color: var(--ve-input-status-warning-border-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-status-warning-shadow-focus);
    }

    .ve-textarea-wrapper--success .ve-input-textarea {
      background: var(--ve-input-status-success-bg);
      border-color: var(--ve-input-status-success-border);
    }

    .ve-textarea-wrapper--success .ve-input-textarea:not(:disabled):hover {
      background: var(--ve-input-status-success-bg-hover);
      border-color: var(--ve-input-status-success-border-hover);
    }

    .ve-textarea-wrapper--success .ve-input-textarea:not(:disabled):focus {
      border-color: var(--ve-input-status-success-border-focus);
      background: var(--ve-input-bg-focus);
      box-shadow: var(--ve-input-status-success-shadow-focus);
    }

    .ve-textarea-wrapper--disabled .ve-input-textarea,
    .ve-textarea-wrapper--disabled .ve-input-textarea:hover,
    .ve-textarea-wrapper--disabled .ve-input-textarea:focus {
      background: var(--ve-input-disabled-bg);
      border-color: var(--ve-input-disabled-border-color);
      color: var(--ve-input-disabled-text);
      box-shadow: none;
      cursor: not-allowed;
      resize: none;
    }

    .ve-textarea-wrapper--disabled .ve-input-textarea::placeholder {
      color: var(--ve-input-disabled-text);
    }

    .ve-textarea-wrapper--clearable .ve-input-textarea {
      padding-inline-end: calc(
        var(--ve-input-textarea-padding-inline) + var(--ve-input-action-size) +
          var(--space-xxs)
      );
    }

    .ve-input-clear--textarea {
      position: absolute;
      inset-inline-end: var(--ve-input-textarea-clear-offset-inline);
      inset-block-start: var(--ve-input-textarea-clear-offset-block);
      margin-inline-start: 0;
      display: none;
      visibility: visible;
    }

    .ve-textarea-wrapper:hover .ve-input-clear--textarea,
    .ve-textarea-wrapper:focus-within .ve-input-clear--textarea {
      display: inline-flex;
    }

    .ve-textarea-word-limit {
      position: absolute;
      inset-inline-end: var(--ve-input-textarea-word-limit-offset-inline);
      inset-block-end: var(--ve-input-textarea-word-limit-offset-block);
      font-size: var(--ve-input-word-limit-font-size);
      line-height: var(--ve-input-line-height-default);
      color: var(--ve-input-word-limit-color);
      font-variant-numeric: tabular-nums;
      user-select: none;
      white-space: nowrap;
    }

    .ve-textarea-wrapper--clearable .ve-textarea-word-limit {
      inset-inline-end: calc(
        var(--ve-input-textarea-word-limit-offset-inline) +
          var(--ve-input-action-size) + var(--space-xxs)
      );
    }

    .ve-textarea-word-limit--outside {
      position: static;
      align-self: flex-end;
      margin-block-start: var(--space-xxxs);
    }

    .ve-textarea-word-limit--error {
      color: var(--ve-input-word-limit-error-color);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-input-wrapper,
      .ve-input-textarea,
      .ve-input-clear {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_type, _init_extra__init_type, _init_placeholder, _init_extra__init_placeholder, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_clearable, _init_extra__init_clearable, _init_maxLength, _init_extra__init_maxLength, _init_wordLimit, _init_extra__init_wordLimit, _init_rows, _init_extra__init_rows, _init_autoSize, _init_extra__init_autoSize, _init_autoWidth, _init_extra__init_autoWidth, _init_controlElement, _init_extra__init_controlElement, _init_mirrorElement, _init_extra__init_mirrorElement, _init_prefixSlotElement, _init_extra__init_prefixSlotElement, _init_suffixSlotElement, _init_extra__init_suffixSlotElement, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot, _init_hasSuffixSlot, _init_extra__init_hasSuffixSlot;
const VE_INPUT_TAG = 've-input';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby'
];
const CLEAR_BUTTON_LABEL = 'Clear';
function isRecord(value) {
    return 'object' == typeof value && null !== value;
}
function hasRenderableAssignedNode(slot) {
    if (!slot) return false;
    return slot.assignedNodes({
        flatten: true
    }).some(hasRenderableNode);
}
function hasRenderableNode(node) {
    if (node.nodeType === Node.ELEMENT_NODE) return true;
    if (node.nodeType !== Node.TEXT_NODE) return false;
    return !!node.textContent?.trim().length;
}
function toPositiveInteger(value) {
    if ('number' != typeof value || !Number.isFinite(value) || value <= 0) return;
    return Math.floor(value);
}
function normalizeCssSizeValue(value) {
    if (void 0 === value) return;
    if ('number' == typeof value) return Number.isFinite(value) ? `${value}px` : void 0;
    const trimmed = value.trim();
    if (!trimmed) return;
    if (/^-?\d+(\.\d+)?$/.test(trimmed)) return `${trimmed}px`;
    return trimmed;
}
function sliceByCharacterCount(value, maxLength) {
    const characters = Array.from(value);
    if (characters.length <= maxLength) return value;
    return characters.slice(0, maxLength).join('');
}
function getCharacterCount(value) {
    return Array.from(value).length;
}
const MAX_LENGTH_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return;
        const trimmed = value.trim();
        if (!trimmed) return;
        try {
            const parsed = JSON.parse(trimmed);
            if ('number' == typeof parsed) return toPositiveInteger(parsed);
            if (isRecord(parsed)) {
                const length = toPositiveInteger(parsed.length);
                if (!length) return;
                return {
                    length,
                    errorOnly: true === parsed.errorOnly
                };
            }
        } catch  {}
        return toPositiveInteger(Number(trimmed));
    },
    toAttribute (value) {
        if (void 0 === value) return null;
        if ('number' == typeof value) return String(value);
        return JSON.stringify(value);
    }
};
const AUTO_SIZE_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return false;
        const trimmed = value.trim();
        if (!trimmed || 'true' === trimmed) return true;
        if ('false' === trimmed) return false;
        try {
            const parsed = JSON.parse(trimmed);
            if ('boolean' == typeof parsed) return parsed;
            if (isRecord(parsed)) {
                const minRows = toPositiveInteger(parsed.minRows);
                const maxRows = toPositiveInteger(parsed.maxRows);
                return {
                    ...void 0 !== minRows ? {
                        minRows
                    } : {},
                    ...void 0 !== maxRows ? {
                        maxRows
                    } : {}
                };
            }
        } catch  {}
        return true;
    },
    toAttribute (value) {
        if (false === value) return null;
        if (true === value) return '';
        return JSON.stringify(value);
    }
};
const AUTO_WIDTH_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return false;
        const trimmed = value.trim();
        if (!trimmed || 'true' === trimmed) return true;
        if ('false' === trimmed) return false;
        try {
            const parsed = JSON.parse(trimmed);
            if ('boolean' == typeof parsed) return parsed;
            if (isRecord(parsed)) {
                const minWidth = 'number' == typeof parsed.minWidth || 'string' == typeof parsed.minWidth ? parsed.minWidth : void 0;
                const maxWidth = 'number' == typeof parsed.maxWidth || 'string' == typeof parsed.maxWidth ? parsed.maxWidth : void 0;
                return {
                    ...void 0 !== minWidth ? {
                        minWidth
                    } : {},
                    ...void 0 !== maxWidth ? {
                        maxWidth
                    } : {}
                };
            }
        } catch  {}
        return true;
    },
    toAttribute (value) {
        if (false === value) return null;
        if (true === value) return '';
        return JSON.stringify(value);
    }
};
_dec = property({
    type: String
}), _dec1 = property({
    type: String,
    attribute: 'default-value'
}), _dec2 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec3 = property({
    type: String
}), _dec4 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec5 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec6 = property({
    type: Boolean,
    reflect: true
}), _dec7 = property({
    type: Boolean,
    reflect: true
}), _dec8 = property({
    attribute: 'max-length',
    converter: MAX_LENGTH_ATTRIBUTE_CONVERTER
}), _dec9 = property({
    type: String,
    attribute: 'word-limit',
    reflect: true,
    useDefault: true
}), _dec10 = property({
    type: Number,
    reflect: true,
    useDefault: true
}), _dec11 = property({
    attribute: 'auto-size',
    converter: AUTO_SIZE_ATTRIBUTE_CONVERTER
}), _dec12 = property({
    attribute: 'auto-width',
    converter: AUTO_WIDTH_ATTRIBUTE_CONVERTER
}), _dec13 = query('.ve-input-control'), _dec14 = query('.ve-input-mirror'), _dec15 = query('slot[name="prefix"]'), _dec16 = query('slot[name="suffix"]'), _dec17 = state(), _dec18 = state();
class VeInputElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_type, _init_extra__init_type, _init_placeholder, _init_extra__init_placeholder, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_clearable, _init_extra__init_clearable, _init_maxLength, _init_extra__init_maxLength, _init_wordLimit, _init_extra__init_wordLimit, _init_rows, _init_extra__init_rows, _init_autoSize, _init_extra__init_autoSize, _init_autoWidth, _init_extra__init_autoWidth, _init_controlElement, _init_extra__init_controlElement, _init_mirrorElement, _init_extra__init_mirrorElement, _init_prefixSlotElement, _init_extra__init_prefixSlotElement, _init_suffixSlotElement, _init_extra__init_suffixSlotElement, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot, _init_hasSuffixSlot, _init_extra__init_hasSuffixSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "value"
            ],
            [
                _dec1,
                1,
                "defaultValue"
            ],
            [
                _dec2,
                1,
                "type"
            ],
            [
                _dec3,
                1,
                "placeholder"
            ],
            [
                _dec4,
                1,
                "size"
            ],
            [
                _dec5,
                1,
                "status"
            ],
            [
                _dec6,
                1,
                "disabled"
            ],
            [
                _dec7,
                1,
                "clearable"
            ],
            [
                _dec8,
                1,
                "maxLength"
            ],
            [
                _dec9,
                1,
                "wordLimit"
            ],
            [
                _dec10,
                1,
                "rows"
            ],
            [
                _dec11,
                1,
                "autoSize"
            ],
            [
                _dec12,
                1,
                "autoWidth"
            ],
            [
                _dec13,
                1,
                "controlElement"
            ],
            [
                _dec14,
                1,
                "mirrorElement"
            ],
            [
                _dec15,
                1,
                "prefixSlotElement"
            ],
            [
                _dec16,
                1,
                "suffixSlotElement"
            ],
            [
                _dec17,
                1,
                "hasPrefixSlot"
            ],
            [
                _dec18,
                1,
                "hasSuffixSlot"
            ]
        ]));
    }
    static styles = ve_input;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    #___private_value_1 = (()=>{
        const _value = _init_value(this, '');
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_defaultValue_2 = (()=>{
        const _value = _init_defaultValue(this, '');
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_2;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_2 = _v;
    }
    #___private_type_3 = (()=>{
        const _value = _init_type(this, 'text');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_3;
    }
    set type(_v) {
        this.#___private_type_3 = _v;
    }
    #___private_placeholder_4 = (()=>{
        const _value = _init_placeholder(this, '');
        _init_extra__init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_4;
    }
    set placeholder(_v) {
        this.#___private_placeholder_4 = _v;
    }
    #___private_size_5 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_5;
    }
    set size(_v) {
        this.#___private_size_5 = _v;
    }
    #___private_status_6 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_6;
    }
    set status(_v) {
        this.#___private_status_6 = _v;
    }
    #___private_disabled_7 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_7;
    }
    set disabled(_v) {
        this.#___private_disabled_7 = _v;
    }
    #___private_clearable_8 = (()=>{
        const _value = _init_clearable(this, false);
        _init_extra__init_clearable(this);
        return _value;
    })();
    get clearable() {
        return this.#___private_clearable_8;
    }
    set clearable(_v) {
        this.#___private_clearable_8 = _v;
    }
    #___private_maxLength_9 = (()=>{
        const _value = _init_maxLength(this, void 0);
        _init_extra__init_maxLength(this);
        return _value;
    })();
    get maxLength() {
        return this.#___private_maxLength_9;
    }
    set maxLength(_v) {
        this.#___private_maxLength_9 = _v;
    }
    #___private_wordLimit_10 = (()=>{
        const _value = _init_wordLimit(this, 'hidden');
        _init_extra__init_wordLimit(this);
        return _value;
    })();
    get wordLimit() {
        return this.#___private_wordLimit_10;
    }
    set wordLimit(_v) {
        this.#___private_wordLimit_10 = _v;
    }
    #___private_rows_11 = (()=>{
        const _value = _init_rows(this, 2);
        _init_extra__init_rows(this);
        return _value;
    })();
    get rows() {
        return this.#___private_rows_11;
    }
    set rows(_v) {
        this.#___private_rows_11 = _v;
    }
    #___private_autoSize_12 = (()=>{
        const _value = _init_autoSize(this, false);
        _init_extra__init_autoSize(this);
        return _value;
    })();
    get autoSize() {
        return this.#___private_autoSize_12;
    }
    set autoSize(_v) {
        this.#___private_autoSize_12 = _v;
    }
    #___private_autoWidth_13 = (()=>{
        const _value = _init_autoWidth(this, false);
        _init_extra__init_autoWidth(this);
        return _value;
    })();
    get autoWidth() {
        return this.#___private_autoWidth_13;
    }
    set autoWidth(_v) {
        this.#___private_autoWidth_13 = _v;
    }
    #___private_controlElement_14 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_14;
    }
    set controlElement(_v) {
        this.#___private_controlElement_14 = _v;
    }
    #___private_mirrorElement_15 = (()=>{
        const _value = _init_mirrorElement(this, null);
        _init_extra__init_mirrorElement(this);
        return _value;
    })();
    get mirrorElement() {
        return this.#___private_mirrorElement_15;
    }
    set mirrorElement(_v) {
        this.#___private_mirrorElement_15 = _v;
    }
    #___private_prefixSlotElement_16 = (()=>{
        const _value = _init_prefixSlotElement(this, null);
        _init_extra__init_prefixSlotElement(this);
        return _value;
    })();
    get prefixSlotElement() {
        return this.#___private_prefixSlotElement_16;
    }
    set prefixSlotElement(_v) {
        this.#___private_prefixSlotElement_16 = _v;
    }
    #___private_suffixSlotElement_17 = (()=>{
        const _value = _init_suffixSlotElement(this, null);
        _init_extra__init_suffixSlotElement(this);
        return _value;
    })();
    get suffixSlotElement() {
        return this.#___private_suffixSlotElement_17;
    }
    set suffixSlotElement(_v) {
        this.#___private_suffixSlotElement_17 = _v;
    }
    #___private_hasPrefixSlot_18 = (()=>{
        const _value = _init_hasPrefixSlot(this, false);
        _init_extra__init_hasPrefixSlot(this);
        return _value;
    })();
    get hasPrefixSlot() {
        return this.#___private_hasPrefixSlot_18;
    }
    set hasPrefixSlot(_v) {
        this.#___private_hasPrefixSlot_18 = _v;
    }
    #___private_hasSuffixSlot_19 = (()=>{
        const _value = _init_hasSuffixSlot(this, false);
        _init_extra__init_hasSuffixSlot(this);
        return _value;
    })();
    get hasSuffixSlot() {
        return this.#___private_hasSuffixSlot_19;
    }
    set hasSuffixSlot(_v) {
        this.#___private_hasSuffixSlot_19 = _v;
    }
    autoWidthFrame = 0;
    textareaAutoSizeFrame = 0;
    cachedAutoWidthFontSignature = '';
    lightDomObserver = null;
    hasInitializedDefaultValue = false;
    lastCommittedValue = '';
    isComposing = false;
    lastWordCountValue = '';
    lastWordCount = 0;
    get controlAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get controlAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get controlAriaDescribedby() {
        return this.getAttribute('aria-describedby') ?? void 0;
    }
    get normalizedSize() {
        return 'small' === this.size || 'large' === this.size ? this.size : 'default';
    }
    get isTextarea() {
        return 'textarea' === this.type;
    }
    get resolvedMaxLengthConfig() {
        if ('number' == typeof this.maxLength) {
            const length = toPositiveInteger(this.maxLength);
            if (!length) return;
            return {
                length,
                errorOnly: false
            };
        }
        if (isRecord(this.maxLength)) {
            const length = toPositiveInteger(this.maxLength.length);
            if (!length) return;
            return {
                length,
                errorOnly: true === this.maxLength.errorOnly
            };
        }
    }
    get nativeMaxLength() {
        const config = this.resolvedMaxLengthConfig;
        if (!config || config.errorOnly) return;
        return config.length;
    }
    get wordCount() {
        if (this.value === this.lastWordCountValue) return this.lastWordCount;
        const count = getCharacterCount(this.value);
        this.lastWordCountValue = this.value;
        this.lastWordCount = count;
        return count;
    }
    get hasLengthError() {
        const config = this.resolvedMaxLengthConfig;
        if (!config || !config.errorOnly) return false;
        return this.wordCount > config.length;
    }
    get mergedStatus() {
        if (this.hasLengthError) return 'error';
        if ('error' === this.status || 'warning' === this.status || 'success' === this.status) return this.status;
        return 'default';
    }
    get hasPrefixContent() {
        return this.hasPrefixSlot;
    }
    get hasSuffixContent() {
        if (this.shouldShowWordLimitInInput) return false;
        return this.hasSuffixSlot;
    }
    get resolvedWordLimitMode() {
        if (!this.resolvedMaxLengthConfig) return 'hidden';
        if (!this.isTextarea) return 'hidden' === this.wordLimit ? 'hidden' : 'inside';
        return this.wordLimit;
    }
    get shouldShowWordLimitInInput() {
        return !this.isTextarea && 'inside' === this.resolvedWordLimitMode;
    }
    get shouldShowWordLimitInsideTextarea() {
        return this.isTextarea && 'inside' === this.resolvedWordLimitMode;
    }
    get shouldShowWordLimitOutsideTextarea() {
        return this.isTextarea && 'outside' === this.resolvedWordLimitMode;
    }
    get resolvedWordLimitText() {
        const length = this.resolvedMaxLengthConfig?.length;
        if (!length) return '';
        return `${this.wordCount}/${length}`;
    }
    get shouldShowClearButton() {
        return this.clearable && !this.disabled && this.value.length > 0;
    }
    get resolvedRows() {
        if (!Number.isFinite(this.rows) || this.rows <= 0) return 2;
        return Math.floor(this.rows);
    }
    get resolvedAutoWidthConfig() {
        if (!this.autoWidth || this.isTextarea) return null;
        if (true === this.autoWidth) return {
            minWidth: 0,
            maxWidth: '100%'
        };
        return {
            minWidth: this.autoWidth.minWidth ?? 0,
            maxWidth: this.autoWidth.maxWidth ?? '100%'
        };
    }
    get nativeAriaInvalid() {
        return 'error' === this.mergedStatus ? 'true' : void 0;
    }
    focus(options) {
        this.controlElement?.focus(options);
    }
    blur() {
        this.controlElement?.blur();
    }
    connectedCallback() {
        super.connectedCallback();
        this.lightDomObserver = new MutationObserver(()=>{
            this.syncSlotState();
        });
        this.lightDomObserver.observe(this, {
            childList: true,
            subtree: true,
            characterData: true,
            attributes: true,
            attributeFilter: [
                'slot'
            ]
        });
        if (!this.hasInitializedDefaultValue) {
            if (!this.hasAttribute('value') && 0 === this.value.length && this.defaultValue) this.value = this.applyMaxLengthConstraint(this.defaultValue);
            this.hasInitializedDefaultValue = true;
        }
        this.lastCommittedValue = this.value;
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.syncSlotState();
        this.syncControlValue();
        this.scheduleAutoWidthUpdate();
        this.scheduleTextareaAutoSizeUpdate();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('value') || changedProperties.has('maxLength')) {
            const constrained = this.applyMaxLengthConstraint(this.value);
            if (constrained !== this.value) {
                this.value = constrained;
                return;
            }
        }
        if (changedProperties.has('value')) this.syncControlValue();
        if (changedProperties.has('value') || changedProperties.has('placeholder') || changedProperties.has('autoWidth') || changedProperties.has('size') || changedProperties.has('type')) this.scheduleAutoWidthUpdate();
        if (changedProperties.has('value') || changedProperties.has('autoSize') || changedProperties.has('rows') || changedProperties.has('type') || changedProperties.has('disabled')) this.scheduleTextareaAutoSizeUpdate();
    }
    disconnectedCallback() {
        this.lightDomObserver?.disconnect();
        this.lightDomObserver = null;
        if ("u" < typeof window) return void super.disconnectedCallback();
        if (this.autoWidthFrame) {
            window.cancelAnimationFrame(this.autoWidthFrame);
            this.autoWidthFrame = 0;
        }
        if (this.textareaAutoSizeFrame) {
            window.cancelAnimationFrame(this.textareaAutoSizeFrame);
            this.textareaAutoSizeFrame = 0;
        }
        super.disconnectedCallback();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    applyMaxLengthConstraint(rawValue) {
        const nativeLength = this.nativeMaxLength;
        if (!nativeLength) return rawValue;
        return sliceByCharacterCount(rawValue, nativeLength);
    }
    syncControlValue() {
        if (!this.controlElement) return;
        if (this.controlElement.value !== this.value) this.controlElement.value = this.value;
    }
    syncSlotState() {
        this.hasPrefixSlot = this.hasNamedSlotContent('prefix', this.prefixSlotElement);
        this.hasSuffixSlot = this.hasNamedSlotContent('suffix', this.suffixSlotElement);
    }
    hasNamedSlotContent(slotName, slotElement) {
        if (hasRenderableAssignedNode(slotElement)) return true;
        return Array.from(this.childNodes).some((node)=>node instanceof HTMLElement && node.slot === slotName && hasRenderableNode(node));
    }
    scheduleTextareaAutoSizeUpdate() {
        if ("u" < typeof window) return;
        if (this.textareaAutoSizeFrame) window.cancelAnimationFrame(this.textareaAutoSizeFrame);
        this.textareaAutoSizeFrame = window.requestAnimationFrame(()=>{
            this.textareaAutoSizeFrame = 0;
            this.updateTextareaAutoSize();
        });
    }
    scheduleAutoWidthUpdate() {
        if ("u" < typeof window) return;
        if (this.autoWidthFrame) window.cancelAnimationFrame(this.autoWidthFrame);
        this.autoWidthFrame = window.requestAnimationFrame(()=>{
            this.autoWidthFrame = 0;
            this.updateAutoWidth();
        });
    }
    updateTextareaAutoSize() {
        if (!this.isTextarea || "u" < typeof window) return;
        if (!(this.controlElement instanceof HTMLTextAreaElement)) return;
        const textarea = this.controlElement;
        if (!this.autoSize) {
            textarea.style.removeProperty('overflow-y');
            textarea.style.removeProperty('height');
            return;
        }
        textarea.style.height = 'auto';
        const computedStyle = window.getComputedStyle(textarea);
        const fontSize = Number.parseFloat(computedStyle.fontSize) || 14;
        const lineHeightValue = Number.parseFloat(computedStyle.lineHeight);
        const lineHeight = Number.isFinite(lineHeightValue) ? lineHeightValue : 1.5 * fontSize;
        const paddingTop = Number.parseFloat(computedStyle.paddingTop) || 0;
        const paddingBottom = Number.parseFloat(computedStyle.paddingBottom) || 0;
        const borderTop = Number.parseFloat(computedStyle.borderTopWidth) || 0;
        const borderBottom = Number.parseFloat(computedStyle.borderBottomWidth) || 0;
        const chromeHeight = paddingTop + paddingBottom + borderTop + borderBottom;
        const config = true === this.autoSize ? void 0 : this.autoSize;
        const minRows = toPositiveInteger(config?.minRows);
        const maxRows = toPositiveInteger(config?.maxRows);
        const minHeight = minRows ? minRows * lineHeight + chromeHeight : void 0;
        const maxHeight = maxRows ? maxRows * lineHeight + chromeHeight : void 0;
        let nextHeight = textarea.scrollHeight;
        if (void 0 !== minHeight) nextHeight = Math.max(nextHeight, minHeight);
        let overflowY = 'hidden';
        if (void 0 !== maxHeight && nextHeight > maxHeight) {
            nextHeight = maxHeight;
            overflowY = 'auto';
        }
        textarea.style.height = `${nextHeight}px`;
        textarea.style.overflowY = overflowY;
    }
    updateAutoWidth() {
        const config = this.resolvedAutoWidthConfig;
        if (!config || "u" < typeof window || this.isTextarea) {
            if (this.controlElement instanceof HTMLInputElement) {
                this.controlElement.style.removeProperty('width');
                this.controlElement.style.removeProperty('min-width');
                this.controlElement.style.removeProperty('max-width');
            }
            return;
        }
        if (!(this.controlElement instanceof HTMLInputElement)) return;
        if (!this.mirrorElement) return;
        const input = this.controlElement;
        const mirror = this.mirrorElement;
        const computedStyle = window.getComputedStyle(input);
        const fontSignature = [
            computedStyle.font,
            computedStyle.letterSpacing,
            computedStyle.textIndent,
            computedStyle.textTransform
        ].join('|');
        if (fontSignature !== this.cachedAutoWidthFontSignature) {
            this.cachedAutoWidthFontSignature = fontSignature;
            mirror.style.font = computedStyle.font;
            mirror.style.letterSpacing = computedStyle.letterSpacing;
            mirror.style.textIndent = computedStyle.textIndent;
            mirror.style.textTransform = computedStyle.textTransform;
        }
        mirror.textContent = this.value || this.placeholder || '\u00a0';
        const width = Math.ceil(mirror.getBoundingClientRect().width + 2);
        input.style.width = `${width}px`;
        input.style.minWidth = '0';
        input.style.maxWidth = '100%';
    }
    dispatchInputEvent(detail) {
        this.dispatchEvent(new CustomEvent('ve-input', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    dispatchChangeEvent(detail) {
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    handleInput(event) {
        const target = event.target;
        if (!(target instanceof HTMLInputElement || target instanceof HTMLTextAreaElement)) return;
        event.stopPropagation();
        const nextValue = event.isComposing || this.isComposing ? target.value : this.applyMaxLengthConstraint(target.value);
        if (target.value !== nextValue) target.value = nextValue;
        const oldValue = this.value;
        this.value = nextValue;
        if (oldValue !== nextValue) this.dispatchInputEvent({
            value: nextValue,
            oldValue,
            source: 'input',
            sourceEvent: event
        });
    }
    handleChange(event) {
        const target = event.target;
        if (!(target instanceof HTMLInputElement || target instanceof HTMLTextAreaElement)) return;
        event.stopPropagation();
        const nextValue = this.applyMaxLengthConstraint(target.value);
        if (target.value !== nextValue) target.value = nextValue;
        this.value = nextValue;
        if (this.lastCommittedValue !== nextValue) {
            this.dispatchChangeEvent({
                value: nextValue,
                oldValue: this.lastCommittedValue,
                source: 'input',
                sourceEvent: event
            });
            this.lastCommittedValue = nextValue;
        }
    }
    handleClearClick(event) {
        event.stopPropagation();
        const oldValue = this.value;
        this.value = '';
        this.syncControlValue();
        if ('' !== oldValue) this.dispatchInputEvent({
            value: '',
            oldValue,
            source: 'clear',
            sourceEvent: event
        });
        if ('' !== this.lastCommittedValue) {
            this.dispatchChangeEvent({
                value: '',
                oldValue: this.lastCommittedValue,
                source: 'clear',
                sourceEvent: event
            });
            this.lastCommittedValue = '';
        }
        this.controlElement?.focus();
    }
    handleKeepFocusMouseDown(event) {
        event.preventDefault();
    }
    isInteractiveEventPath(event) {
        const eventPath = 'function' == typeof event.composedPath ? event.composedPath() : [
            event.target
        ];
        for (const item of eventPath){
            if (item === event.currentTarget || item === this) break;
            if (item instanceof HTMLElement) {
                if (item.matches('a[href],button,input,textarea,select,summary,[role="button"],[role="link"],[role="menuitem"],[role="option"],[role="switch"],[role="checkbox"],[role="radio"],[role="tab"],[contenteditable=""],[contenteditable="true"],[tabindex]:not([tabindex="-1"])')) return true;
            }
        }
        return false;
    }
    handleWrapperMouseDown(event) {
        if (this.disabled || this.isInteractiveEventPath(event)) return;
        event.preventDefault();
    }
    handleWrapperClick(event) {
        if (this.disabled || this.isInteractiveEventPath(event)) return;
        this.controlElement?.focus();
    }
    handleCompositionStart() {
        this.isComposing = true;
    }
    handleCompositionEnd(event) {
        this.isComposing = false;
        const target = event.target;
        if (!(target instanceof HTMLInputElement || target instanceof HTMLTextAreaElement)) return;
        const nextValue = this.applyMaxLengthConstraint(target.value);
        if (target.value === nextValue && this.value === nextValue) return;
        const oldValue = this.value;
        target.value = nextValue;
        this.value = nextValue;
        this.dispatchInputEvent({
            value: nextValue,
            oldValue,
            source: 'input',
            sourceEvent: event
        });
    }
    handleSlotChange() {
        this.syncSlotState();
    }
    renderClearButton() {
        if (!this.shouldShowClearButton) return nothing;
        return html`
      <button
        class=${classMap({
            've-input-clear': true,
            've-input-clear--textarea': this.isTextarea
        })}
        part="clear"
        type="button"
        tabindex=${ifDefined(this.disabled ? '-1' : void 0)}
        aria-label=${CLEAR_BUTTON_LABEL}
        @mousedown=${this.handleKeepFocusMouseDown}
        @click=${this.handleClearClick}
      >
        <ve-icon .definition=${closeIcon}></ve-icon>
      </button>
    `;
    }
    renderPrefix() {
        if (!this.hasPrefixContent) return nothing;
        return html`
      <span class="ve-input-prefix" part="prefix">
        <slot name="prefix" @slotchange=${this.handleSlotChange}></slot>
      </span>
    `;
    }
    renderSuffix() {
        if (this.shouldShowWordLimitInInput) return html`
        <span
          class=${classMap({
            've-input-suffix': true,
            've-input-word-limit': true,
            've-input-word-limit--error': this.hasLengthError
        })}
          part="word-limit"
        >
          ${this.resolvedWordLimitText}
        </span>
      `;
        if (!this.hasSuffixContent) return nothing;
        return html`
      <span class="ve-input-suffix" part="suffix">
        <slot name="suffix" @slotchange=${this.handleSlotChange}></slot>
      </span>
    `;
    }
    getInputWrapperStyles() {
        const styles = {};
        const autoWidthConfig = this.resolvedAutoWidthConfig;
        if (autoWidthConfig) {
            styles.width = 'auto';
            const minWidth = normalizeCssSizeValue(autoWidthConfig.minWidth);
            const maxWidth = normalizeCssSizeValue(autoWidthConfig.maxWidth);
            if (minWidth) styles.minWidth = minWidth;
            if (maxWidth) styles.maxWidth = maxWidth;
        }
        return styles;
    }
    getTextareaStyles() {
        const styles = {};
        if (this.autoSize || this.disabled) styles.resize = 'none';
        return styles;
    }
    renderControl() {
        if (this.isTextarea) return html`
        <textarea
          class="ve-input-control ve-input-textarea"
          part="input"
          .value=${this.value}
          rows=${this.resolvedRows}
          style=${styleMap(this.getTextareaStyles())}
          placeholder=${ifDefined(this.placeholder || void 0)}
          maxlength=${ifDefined(this.nativeMaxLength)}
          aria-label=${ifDefined(this.controlAriaLabel)}
          aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
          aria-describedby=${ifDefined(this.controlAriaDescribedby)}
          aria-invalid=${ifDefined(this.nativeAriaInvalid)}
          ?disabled=${this.disabled}
          @input=${this.handleInput}
          @change=${this.handleChange}
          @compositionstart=${this.handleCompositionStart}
          @compositionend=${this.handleCompositionEnd}
        ></textarea>
      `;
        return html`
      <input
        class="ve-input-control ve-input-field"
        part="input"
        .value=${this.value}
        type="text"
        placeholder=${ifDefined(this.placeholder || void 0)}
        maxlength=${ifDefined(this.nativeMaxLength)}
        aria-label=${ifDefined(this.controlAriaLabel)}
        aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
        aria-describedby=${ifDefined(this.controlAriaDescribedby)}
        aria-invalid=${ifDefined(this.nativeAriaInvalid)}
        ?disabled=${this.disabled}
        @input=${this.handleInput}
        @change=${this.handleChange}
        @compositionstart=${this.handleCompositionStart}
        @compositionend=${this.handleCompositionEnd}
      />
    `;
    }
    render() {
        if (this.isTextarea) {
            const textareaWrapperClass = classMap({
                've-textarea-wrapper': true,
                [`ve-textarea-wrapper--${this.normalizedSize}`]: true,
                've-textarea-wrapper--disabled': this.disabled,
                've-textarea-wrapper--error': 'error' === this.mergedStatus,
                've-textarea-wrapper--warning': 'warning' === this.mergedStatus,
                've-textarea-wrapper--success': 'success' === this.mergedStatus,
                've-textarea-wrapper--word-limit-outside': this.shouldShowWordLimitOutsideTextarea,
                've-textarea-wrapper--clearable': this.clearable
            });
            return html`
        <div
          class=${textareaWrapperClass}
          style=${styleMap(this.getInputWrapperStyles())}
          part="wrapper"
          @mousedown=${this.handleWrapperMouseDown}
          @click=${this.handleWrapperClick}
        >
          ${this.renderControl()} ${this.renderClearButton()}
          ${this.shouldShowWordLimitInsideTextarea ? html`
                <span
                  class=${classMap({
                've-textarea-word-limit': true,
                've-textarea-word-limit--error': this.hasLengthError
            })}
                  part="word-limit"
                >
                  ${this.resolvedWordLimitText}
                </span>
              ` : nothing}
          ${this.shouldShowWordLimitOutsideTextarea ? html`
                <span
                  class=${classMap({
                've-textarea-word-limit': true,
                've-textarea-word-limit--outside': true,
                've-textarea-word-limit--error': this.hasLengthError
            })}
                  part="word-limit"
                >
                  ${this.resolvedWordLimitText}
                </span>
              ` : nothing}
        </div>
      `;
        }
        const wrapperClass = classMap({
            've-input-wrapper': true,
            [`ve-input-wrapper--${this.normalizedSize}`]: true,
            've-input-wrapper--disabled': this.disabled,
            've-input-wrapper--error': 'error' === this.mergedStatus,
            've-input-wrapper--warning': 'warning' === this.mergedStatus,
            've-input-wrapper--success': 'success' === this.mergedStatus,
            've-input-wrapper--auto-width': null !== this.resolvedAutoWidthConfig,
            've-input-wrapper--has-prefix': this.hasPrefixContent,
            've-input-wrapper--clearable': this.clearable
        });
        return html`
      <div
        class=${wrapperClass}
        style=${styleMap(this.getInputWrapperStyles())}
        part="wrapper"
        @mousedown=${this.handleWrapperMouseDown}
        @click=${this.handleWrapperClick}
      >
        ${this.renderPrefix()} ${this.renderControl()}
        ${this.renderClearButton()} ${this.renderSuffix()}
        ${null !== this.resolvedAutoWidthConfig ? html`<span class="ve-input-mirror" aria-hidden="true"></span>` : nothing}
      </div>
    `;
    }
}
defineCustomElement(VE_INPUT_TAG, VeInputElement);
export { VE_INPUT_TAG, VeInputElement };
