// @ve-design/web/web/dist/ve-message.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
import { closeIcon } from "./icons/close.js";
const veMessageTokenDefaults = css`
  :host {
    --ve-message-wrapper-margin-top: var(--space-xxl);
    --ve-message-wrapper-margin-bottom: var(--space-xxl);
    --ve-message-wrapper-padding-inline: var(--space-xs);
    --ve-message-z-index: var(--z-index-message, 1003);

    --ve-message-max-width: none;
    --ve-message-padding-block: var(--space-xs);
    --ve-message-padding-inline: var(--space-s);
    --ve-message-margin-bottom: var(--space-s);
    --ve-message-border-radius: var(--radius-lg);
    --ve-message-font-family: var(--font-cn);
    --ve-message-font-size-icon: 18px;
    --ve-message-font-size-content: var(--text-body);
    --ve-message-font-weight-content: var(--font-weight-medium);
    --ve-message-line-height-content: 22px;
    --ve-message-gap: var(--space-xxs);
    --ve-message-border-width: var(--stroke-0_5);
    --ve-message-border-style: solid;
    --ve-message-box-shadow: var(--shadow-md);

    --ve-message-close-icon-color: var(--color-text-primary);
    --ve-message-close-icon-size: var(--text-ui-strong);
    --ve-message-close-icon-padding: var(--space-xxxs);
    --ve-message-close-icon-hover-base: var(--color-bg-base);
    --ve-message-close-icon-hover-radius: var(--radius-xs);

    --ve-message-info-color-bg: var(--color-bg-base);
    --ve-message-info-color-icon: var(--color-bg-info);
    --ve-message-info-color-content: var(--color-text-primary);
    --ve-message-info-color-border: var(--color-border-default);

    --ve-message-success-color-bg: var(--color-bg-base);
    --ve-message-success-color-icon: var(--color-bg-success);
    --ve-message-success-color-content: var(--color-text-primary);
    --ve-message-success-color-border: var(--color-border-default);

    --ve-message-warning-color-bg: var(--color-bg-base);
    --ve-message-warning-color-icon: var(--color-bg-warning);
    --ve-message-warning-color-content: var(--color-text-primary);
    --ve-message-warning-color-border: var(--color-border-default);

    --ve-message-error-color-bg: var(--color-bg-base);
    --ve-message-error-color-icon: var(--color-bg-danger);
    --ve-message-error-color-content: var(--color-text-primary);
    --ve-message-error-color-border: var(--color-border-default);

    --ve-message-loading-color-bg: var(--color-bg-base);
    --ve-message-loading-color-icon: var(--color-text-primary);
    --ve-message-loading-color-content: var(--color-text-primary);
    --ve-message-loading-color-border: var(--color-border-default);

    --ve-message-current-color-bg: var(--ve-message-info-color-bg);
    --ve-message-current-color-icon: var(--ve-message-info-color-icon);
    --ve-message-current-color-content: var(--ve-message-info-color-content);
    --ve-message-current-color-border: var(--ve-message-info-color-border);

    --ve-message-enter-duration: 100ms;
    --ve-message-fade-duration: 100ms;
    --ve-message-exit-duration: 300ms;
    --ve-message-icon-scale-duration: 400ms;
    --ve-message-loading-duration: 1s;
    --ve-message-linear-easing: cubic-bezier(0, 0, 1, 1);
    --ve-message-standard-easing: cubic-bezier(0.34, 0.69, 0.1, 1);
    --ve-message-overshoot-easing: cubic-bezier(0.3, 1.3, 0.3, 1);
  }
`;
const token = veMessageTokenDefaults;
const ve_message = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      width: 100%;
      text-align: center;
      pointer-events: none;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-message {
      position: relative;
      display: inline-flex;
      align-items: center;
      box-sizing: border-box;
      max-width: var(--ve-message-max-width);
      margin-bottom: var(--ve-message-margin-bottom);
      padding: var(--ve-message-padding-block) var(--ve-message-padding-inline);
      gap: var(--ve-message-gap);
      color: var(--ve-message-current-color-content);
      font-family: var(--ve-message-font-family);
      font-size: var(--ve-message-font-size-content);
      font-weight: var(--ve-message-font-weight-content);
      line-height: var(--ve-message-line-height-content);
      letter-spacing: var(--letter-spacing-normal);
      text-align: start;
      background-color: var(--ve-message-current-color-bg);
      border: var(--ve-message-border-width) var(--ve-message-border-style)
        var(--ve-message-current-color-border);
      border-radius: var(--ve-message-border-radius);
      box-shadow: var(--ve-message-box-shadow);
      opacity: 0;
      pointer-events: auto;
      transition: opacity var(--ve-message-enter-duration)
        var(--ve-message-linear-easing);
    }

    .ve-message-info {
      --ve-message-current-color-bg: var(--ve-message-info-color-bg);
      --ve-message-current-color-icon: var(--ve-message-info-color-icon);
      --ve-message-current-color-content: var(--ve-message-info-color-content);
      --ve-message-current-color-border: var(--ve-message-info-color-border);
    }

    .ve-message-success {
      --ve-message-current-color-bg: var(--ve-message-success-color-bg);
      --ve-message-current-color-icon: var(--ve-message-success-color-icon);
      --ve-message-current-color-content: var(
        --ve-message-success-color-content
      );
      --ve-message-current-color-border: var(--ve-message-success-color-border);
    }

    .ve-message-warning {
      --ve-message-current-color-bg: var(--ve-message-warning-color-bg);
      --ve-message-current-color-icon: var(--ve-message-warning-color-icon);
      --ve-message-current-color-content: var(
        --ve-message-warning-color-content
      );
      --ve-message-current-color-border: var(--ve-message-warning-color-border);
    }

    .ve-message-error {
      --ve-message-current-color-bg: var(--ve-message-error-color-bg);
      --ve-message-current-color-icon: var(--ve-message-error-color-icon);
      --ve-message-current-color-content: var(--ve-message-error-color-content);
      --ve-message-current-color-border: var(--ve-message-error-color-border);
    }

    .ve-message-loading {
      --ve-message-current-color-bg: var(--ve-message-loading-color-bg);
      --ve-message-current-color-icon: var(--ve-message-loading-color-icon);
      --ve-message-current-color-content: var(
        --ve-message-loading-color-content
      );
      --ve-message-current-color-border: var(--ve-message-loading-color-border);
    }

    .ve-message-entered {
      opacity: 1;
    }

    .ve-message-leaving {
      opacity: 0;
      transition: opacity var(--ve-message-fade-duration)
        var(--ve-message-linear-easing);
    }

    .ve-message-multiline {
      align-items: flex-start;
    }

    .ve-message-icon {
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      width: var(--ve-message-font-size-icon);
      height: var(--ve-message-line-height-content);
      color: var(--ve-message-current-color-icon);
      font-size: var(--ve-message-font-size-icon);
      line-height: 1;
      animation:
        ve-message-fade var(--ve-message-fade-duration)
          var(--ve-message-linear-easing),
        ve-message-scale var(--ve-message-icon-scale-duration)
          var(--ve-message-overshoot-easing);
    }

    .ve-message-icon svg,
    .ve-message-icon ::slotted(svg) {
      display: inline-block;
      width: 1em;
      height: 1em;
      color: inherit;
      flex: none;
    }

    .ve-message-icon ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-width: 1em;
      max-height: 1em;
      color: inherit;
      line-height: 1;
    }

    .ve-message-loading .ve-message-icon svg,
    .ve-message-loading .ve-message-icon ::slotted(svg) {
      animation: ve-message-loading-circle var(--ve-message-loading-duration)
        infinite var(--ve-message-linear-easing);
    }

    .ve-message-content {
      flex: 1 1 auto;
      min-width: 0;
      color: var(--ve-message-current-color-content);
      font-size: var(--ve-message-font-size-content);
      font-weight: var(--ve-message-font-weight-content);
      line-height: var(--ve-message-line-height-content);
      word-break: break-word;
    }

    .ve-message-close-btn {
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      width: calc(
        var(--ve-message-close-icon-size) +
          var(--ve-message-close-icon-padding) * 2
      );
      height: calc(
        var(--ve-message-close-icon-size) +
          var(--ve-message-close-icon-padding) * 2
      );
      padding: var(--ve-message-close-icon-padding);
      color: var(--ve-message-close-icon-color);
      font: inherit;
      font-size: var(--ve-message-close-icon-size);
      line-height: 1;
      background: transparent;
      border: 0;
      border-radius: var(--ve-message-close-icon-hover-radius);
      cursor: pointer;
      transition:
        background-color var(--ve-message-fade-duration)
          var(--ve-message-linear-easing),
        color var(--ve-message-fade-duration) var(--ve-message-linear-easing);
    }

    .ve-message-multiline .ve-message-close-btn {
      margin-top: calc(
        (
            var(--ve-message-line-height-content) -
              (
                var(--ve-message-close-icon-size) +
                  var(--ve-message-close-icon-padding) * 2
              )
          ) /
          2
      );
    }

    .ve-message-close-btn:hover,
    .ve-message-close-btn:focus-visible {
      background-color: color-mix(
        var(--ve-message-close-icon-hover-base) var(--state-hover-subtle-ratio),
        var(--state-hover-subtle-mix)
      );
    }

    .ve-message-close-btn slot,
    .ve-message-close-btn ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      width: 100%;
      height: 100%;
      color: inherit;
      line-height: 1;
    }

    .ve-message-close-btn ve-icon,
    .ve-message-close-btn svg,
    .ve-message-close-btn ::slotted(svg) {
      width: 1em;
      height: 1em;
      color: inherit;
    }

    @keyframes ve-message-fade {
      from {
        opacity: 0;
      }

      to {
        opacity: 1;
      }
    }

    @keyframes ve-message-scale {
      from {
        transform: scale(0);
      }

      to {
        transform: scale(1);
      }
    }

    @keyframes ve-message-loading-circle {
      100% {
        transform: rotate(360deg);
      }
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-message,
      .ve-message-leaving,
      .ve-message-close-btn {
        transition: none;
      }

      .ve-message-icon,
      .ve-message-loading .ve-message-icon svg,
      .ve-message-loading .ve-message-icon ::slotted(svg) {
        animation: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _init_type, _init_extra__init_type, _init_content, _init_extra__init_content, _init_messageId, _init_extra__init_messageId, _init_position, _init_extra__init_position, _init_showIcon, _init_extra__init_showIcon, _init_closable, _init_extra__init_closable, _init_duration, _init_extra__init_duration, _init_enterDuration, _init_extra__init_enterDuration, _init_exitDuration, _init_extra__init_exitDuration, _init_iconSlot, _init_extra__init_iconSlot, _init_closeIconSlot, _init_extra__init_closeIconSlot, _init_contentElement, _init_extra__init_contentElement, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasCloseIconSlot, _init_extra__init_hasCloseIconSlot, _init_contentMultiline, _init_extra__init_contentMultiline, _init_phase, _init_extra__init_phase;
const VE_MESSAGE_TAG = 've-message';
const VE_MESSAGE_CLOSE_EVENT = 've-close';
const VE_MESSAGE_AFTER_CLOSE_EVENT = 've-after-close';
const MESSAGE_TYPES = [
    'info',
    'success',
    'warning',
    'error',
    'loading'
];
const MESSAGE_DEFAULT_ENTER_DURATION = 100;
const MESSAGE_DEFAULT_EXIT_DURATION = 300;
const MESSAGE_DEFAULT_DURATION = 3000;
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
function normalizeDuration(value, fallback) {
    return Number.isFinite(value) && value >= 0 ? value : fallback;
}
function shouldReduceMotion() {
    return "u" > typeof window && 'function' == typeof window.matchMedia && window.matchMedia('(prefers-reduced-motion: reduce)').matches;
}
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: String,
    reflect: true
}), _dec2 = property({
    type: String,
    attribute: 'message-id',
    reflect: true
}), _dec3 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec4 = property({
    attribute: 'show-icon',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = property({
    type: Number,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: Number,
    attribute: 'enter-duration',
    reflect: true,
    useDefault: true
}), _dec8 = property({
    type: Number,
    attribute: 'exit-duration',
    reflect: true,
    useDefault: true
}), _dec9 = query('slot[name="icon"]'), _dec10 = query('slot[name="close-icon"]'), _dec11 = query('.ve-message-content'), _dec12 = state(), _dec13 = state(), _dec14 = state(), _dec15 = state();
class VeMessageElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_content, _init_extra__init_content, _init_messageId, _init_extra__init_messageId, _init_position, _init_extra__init_position, _init_showIcon, _init_extra__init_showIcon, _init_closable, _init_extra__init_closable, _init_duration, _init_extra__init_duration, _init_enterDuration, _init_extra__init_enterDuration, _init_exitDuration, _init_extra__init_exitDuration, _init_iconSlot, _init_extra__init_iconSlot, _init_closeIconSlot, _init_extra__init_closeIconSlot, _init_contentElement, _init_extra__init_contentElement, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasCloseIconSlot, _init_extra__init_hasCloseIconSlot, _init_contentMultiline, _init_extra__init_contentMultiline, _init_phase, _init_extra__init_phase] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "type"
            ],
            [
                _dec1,
                1,
                "content"
            ],
            [
                _dec2,
                1,
                "messageId"
            ],
            [
                _dec3,
                1,
                "position"
            ],
            [
                _dec4,
                1,
                "showIcon"
            ],
            [
                _dec5,
                1,
                "closable"
            ],
            [
                _dec6,
                1,
                "duration"
            ],
            [
                _dec7,
                1,
                "enterDuration"
            ],
            [
                _dec8,
                1,
                "exitDuration"
            ],
            [
                _dec9,
                1,
                "iconSlot"
            ],
            [
                _dec10,
                1,
                "closeIconSlot"
            ],
            [
                _dec11,
                1,
                "contentElement"
            ],
            [
                _dec12,
                1,
                "hasIconSlot"
            ],
            [
                _dec13,
                1,
                "hasCloseIconSlot"
            ],
            [
                _dec14,
                1,
                "contentMultiline"
            ],
            [
                _dec15,
                1,
                "phase"
            ]
        ]));
    }
    static styles = ve_message;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'info');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_content_2 = (()=>{
        const _value = _init_content(this, '');
        _init_extra__init_content(this);
        return _value;
    })();
    get content() {
        return this.#___private_content_2;
    }
    set content(_v) {
        this.#___private_content_2 = _v;
    }
    #___private_messageId_3 = (()=>{
        const _value = _init_messageId(this, '');
        _init_extra__init_messageId(this);
        return _value;
    })();
    get messageId() {
        return this.#___private_messageId_3;
    }
    set messageId(_v) {
        this.#___private_messageId_3 = _v;
    }
    #___private_position_4 = (()=>{
        const _value = _init_position(this, 'top');
        _init_extra__init_position(this);
        return _value;
    })();
    get position() {
        return this.#___private_position_4;
    }
    set position(_v) {
        this.#___private_position_4 = _v;
    }
    #___private_showIcon_5 = (()=>{
        const _value = _init_showIcon(this, true);
        _init_extra__init_showIcon(this);
        return _value;
    })();
    get showIcon() {
        return this.#___private_showIcon_5;
    }
    set showIcon(_v) {
        this.#___private_showIcon_5 = _v;
    }
    #___private_closable_6 = (()=>{
        const _value = _init_closable(this, false);
        _init_extra__init_closable(this);
        return _value;
    })();
    get closable() {
        return this.#___private_closable_6;
    }
    set closable(_v) {
        this.#___private_closable_6 = _v;
    }
    #___private_duration_7 = (()=>{
        const _value = _init_duration(this, MESSAGE_DEFAULT_DURATION);
        _init_extra__init_duration(this);
        return _value;
    })();
    get duration() {
        return this.#___private_duration_7;
    }
    set duration(_v) {
        this.#___private_duration_7 = _v;
    }
    #___private_enterDuration_8 = (()=>{
        const _value = _init_enterDuration(this, MESSAGE_DEFAULT_ENTER_DURATION);
        _init_extra__init_enterDuration(this);
        return _value;
    })();
    get enterDuration() {
        return this.#___private_enterDuration_8;
    }
    set enterDuration(_v) {
        this.#___private_enterDuration_8 = _v;
    }
    #___private_exitDuration_9 = (()=>{
        const _value = _init_exitDuration(this, MESSAGE_DEFAULT_EXIT_DURATION);
        _init_extra__init_exitDuration(this);
        return _value;
    })();
    get exitDuration() {
        return this.#___private_exitDuration_9;
    }
    set exitDuration(_v) {
        this.#___private_exitDuration_9 = _v;
    }
    #___private_iconSlot_10 = (()=>{
        const _value = _init_iconSlot(this, null);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_10;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_10 = _v;
    }
    #___private_closeIconSlot_11 = (()=>{
        const _value = _init_closeIconSlot(this, null);
        _init_extra__init_closeIconSlot(this);
        return _value;
    })();
    get closeIconSlot() {
        return this.#___private_closeIconSlot_11;
    }
    set closeIconSlot(_v) {
        this.#___private_closeIconSlot_11 = _v;
    }
    #___private_contentElement_12 = (()=>{
        const _value = _init_contentElement(this, null);
        _init_extra__init_contentElement(this);
        return _value;
    })();
    get contentElement() {
        return this.#___private_contentElement_12;
    }
    set contentElement(_v) {
        this.#___private_contentElement_12 = _v;
    }
    #___private_hasIconSlot_13 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_13;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_13 = _v;
    }
    #___private_hasCloseIconSlot_14 = (()=>{
        const _value = _init_hasCloseIconSlot(this, false);
        _init_extra__init_hasCloseIconSlot(this);
        return _value;
    })();
    get hasCloseIconSlot() {
        return this.#___private_hasCloseIconSlot_14;
    }
    set hasCloseIconSlot(_v) {
        this.#___private_hasCloseIconSlot_14 = _v;
    }
    #___private_contentMultiline_15 = (()=>{
        const _value = _init_contentMultiline(this, false);
        _init_extra__init_contentMultiline(this);
        return _value;
    })();
    get contentMultiline() {
        return this.#___private_contentMultiline_15;
    }
    set contentMultiline(_v) {
        this.#___private_contentMultiline_15 = _v;
    }
    #___private_phase_16 = (()=>{
        const _value = _init_phase(this, 'entering');
        _init_extra__init_phase(this);
        return _value;
    })();
    get phase() {
        return this.#___private_phase_16;
    }
    set phase(_v) {
        this.#___private_phase_16 = _v;
    }
    autoCloseTimer = null;
    enterFrame = null;
    closeOpacityTimer = null;
    closeCompleteTimer = null;
    hovering = false;
    focusWithin = false;
    lightDomSlotObserver;
    contentResizeObserver = null;
    observedContentElement = null;
    contentMeasureFrame = null;
    get normalizedType() {
        return MESSAGE_TYPES.includes(this.type) ? this.type : 'info';
    }
    get normalizedEnterDuration() {
        return normalizeDuration(this.enterDuration, MESSAGE_DEFAULT_ENTER_DURATION);
    }
    get normalizedExitDuration() {
        return normalizeDuration(this.exitDuration, MESSAGE_DEFAULT_EXIT_DURATION);
    }
    get normalizedDuration() {
        return normalizeDuration(this.duration, MESSAGE_DEFAULT_DURATION);
    }
    get shouldRenderIcon() {
        if (!this.showIcon) return false;
        return true;
    }
    get messageRole() {
        return 'error' === this.normalizedType || 'warning' === this.normalizedType ? 'alert' : 'status';
    }
    get messageAriaLive() {
        return 'status' === this.messageRole ? 'polite' : 'assertive';
    }
    get hasIconContent() {
        return this.hasIconSlot || hasNamedSlottedContent(this, 'icon');
    }
    get hasCloseIconContent() {
        return this.hasCloseIconSlot || hasNamedSlottedContent(this, 'close-icon');
    }
    connectedCallback() {
        super.connectedCallback();
        this.observeLightDomSlots();
        this.syncSlotState();
        this.applyTransitionVariables();
    }
    firstUpdated() {
        this.syncSlotState();
        this.observeContentElement();
        this.scheduleEnter();
        this.startTimer();
    }
    updated(changedProperties) {
        this.observeContentElement();
        if (changedProperties.has('content')) this.scheduleContentMeasure();
        if (changedProperties.has('duration') && 'entered' === this.phase) this.restartTimer();
        if (changedProperties.has('enterDuration') || changedProperties.has('exitDuration')) this.applyTransitionVariables();
    }
    disconnectedCallback() {
        this.lightDomSlotObserver?.disconnect();
        this.lightDomSlotObserver = void 0;
        this.contentResizeObserver?.disconnect();
        this.contentResizeObserver = null;
        this.observedContentElement = null;
        this.cancelContentMeasure();
        this.clearTimers();
        super.disconnectedCallback();
    }
    restartTimer() {
        this.clearAutoCloseTimer();
        this.syncAutoCloseTimer();
    }
    close(sourceEvent) {
        if ('leaving' === this.phase || 'closed' === this.phase) return;
        this.clearAutoCloseTimer();
        this.clearCloseTimers();
        if (!this.dispatchMessageEvent(VE_MESSAGE_CLOSE_EVENT, sourceEvent)) return void this.syncAutoCloseTimer();
        this.phase = 'leaving';
        if (shouldReduceMotion()) return void this.finishClose(sourceEvent);
        const hostHeight = this.scrollHeight || this.getBoundingClientRect().height;
        this.style.height = `${hostHeight}px`;
        this.style.overflow = 'hidden';
        this.style.transition = 'none';
        this.getBoundingClientRect();
        const fadeDuration = this.normalizedEnterDuration;
        this.closeOpacityTimer = window.setTimeout(()=>{
            this.closeOpacityTimer = null;
            this.style.transition = `height ${this.normalizedExitDuration}ms var(--ve-message-standard-easing)`;
            this.style.height = '0px';
        }, fadeDuration);
        this.closeCompleteTimer = window.setTimeout(()=>{
            this.closeCompleteTimer = null;
            this.finishClose(sourceEvent);
        }, fadeDuration + this.normalizedExitDuration);
    }
    applyTransitionVariables() {
        this.style.setProperty('--ve-message-enter-duration', `${this.normalizedEnterDuration}ms`);
        this.style.setProperty('--ve-message-fade-duration', `${this.normalizedEnterDuration}ms`);
        this.style.setProperty('--ve-message-exit-duration', `${this.normalizedExitDuration}ms`);
    }
    scheduleEnter() {
        this.cancelEnterFrame();
        this.phase = 'entering';
        this.enterFrame = window.requestAnimationFrame(()=>{
            this.enterFrame = null;
            if ('entering' === this.phase) this.phase = 'entered';
        });
    }
    startTimer() {
        if (0 === this.duration || 'leaving' === this.phase || 'closed' === this.phase) return;
        if (null !== this.autoCloseTimer) return;
        const duration = this.normalizedDuration;
        this.autoCloseTimer = window.setTimeout(()=>{
            this.autoCloseTimer = null;
            this.close();
        }, duration);
    }
    clearAutoCloseTimer() {
        if (null === this.autoCloseTimer) return;
        window.clearTimeout(this.autoCloseTimer);
        this.autoCloseTimer = null;
    }
    clearCloseTimers() {
        if (null !== this.closeOpacityTimer) {
            window.clearTimeout(this.closeOpacityTimer);
            this.closeOpacityTimer = null;
        }
        if (null !== this.closeCompleteTimer) {
            window.clearTimeout(this.closeCompleteTimer);
            this.closeCompleteTimer = null;
        }
    }
    cancelEnterFrame() {
        if (null === this.enterFrame) return;
        window.cancelAnimationFrame(this.enterFrame);
        this.enterFrame = null;
    }
    clearTimers() {
        this.clearAutoCloseTimer();
        this.clearCloseTimers();
        this.cancelEnterFrame();
    }
    syncAutoCloseTimer() {
        if (this.hovering || this.focusWithin) return void this.clearAutoCloseTimer();
        this.startTimer();
    }
    isMessageNode(target) {
        return target instanceof Node && (this.contains(target) || this.renderRoot.contains(target));
    }
    syncSlotState() {
        this.hasIconSlot = hasAssignedSlotContent(this.iconSlot) || hasNamedSlottedContent(this, 'icon');
        this.hasCloseIconSlot = hasAssignedSlotContent(this.closeIconSlot) || hasNamedSlottedContent(this, 'close-icon');
    }
    observeContentElement() {
        if ("u" < typeof ResizeObserver) return void this.scheduleContentMeasure();
        this.contentResizeObserver ??= new ResizeObserver(()=>{
            this.syncContentMultiline();
        });
        if (this.observedContentElement === this.contentElement) return;
        if (this.observedContentElement) this.contentResizeObserver.unobserve(this.observedContentElement);
        this.observedContentElement = this.contentElement;
        if (this.observedContentElement) this.contentResizeObserver.observe(this.observedContentElement);
        this.scheduleContentMeasure();
    }
    scheduleContentMeasure() {
        if ("u" < typeof window || 'function' != typeof window.requestAnimationFrame) return void this.syncContentMultiline();
        if (null !== this.contentMeasureFrame) return;
        this.contentMeasureFrame = window.requestAnimationFrame(()=>{
            this.contentMeasureFrame = null;
            this.syncContentMultiline();
        });
    }
    cancelContentMeasure() {
        if (null === this.contentMeasureFrame || "u" < typeof window || 'function' != typeof window.cancelAnimationFrame) {
            this.contentMeasureFrame = null;
            return;
        }
        window.cancelAnimationFrame(this.contentMeasureFrame);
        this.contentMeasureFrame = null;
    }
    syncContentMultiline() {
        if (!this.contentElement || "u" < typeof getComputedStyle) {
            this.contentMultiline = false;
            return;
        }
        const lineHeight = Number.parseFloat(getComputedStyle(this.contentElement).lineHeight);
        const contentHeight = this.contentElement.getBoundingClientRect().height;
        const nextContentMultiline = Number.isFinite(lineHeight) && lineHeight > 0 && contentHeight > 1.5 * lineHeight;
        if (this.contentMultiline !== nextContentMultiline) this.contentMultiline = nextContentMultiline;
    }
    observeLightDomSlots() {
        if ("u" < typeof MutationObserver || this.lightDomSlotObserver) return;
        this.lightDomSlotObserver = new MutationObserver(()=>{
            this.syncSlotState();
            this.scheduleContentMeasure();
        });
        this.lightDomSlotObserver.observe(this, {
            childList: true,
            subtree: false,
            attributes: true,
            attributeFilter: [
                'slot'
            ]
        });
    }
    handleSlotChange() {
        this.syncSlotState();
        this.scheduleContentMeasure();
    }
    handleMouseEnter() {
        this.hovering = true;
        this.syncAutoCloseTimer();
    }
    handleMouseLeave() {
        this.hovering = false;
        this.syncAutoCloseTimer();
    }
    handleFocusIn() {
        this.focusWithin = true;
        this.syncAutoCloseTimer();
    }
    handleFocusOut(event) {
        if (this.isMessageNode(event.relatedTarget)) return;
        this.focusWithin = false;
        this.syncAutoCloseTimer();
    }
    handleClose(event) {
        this.close(event);
    }
    createCloseDetail(sourceEvent) {
        const detail = {
            type: this.normalizedType
        };
        if (this.messageId) detail.id = this.messageId;
        if (sourceEvent) detail.sourceEvent = sourceEvent;
        return detail;
    }
    dispatchMessageEvent(type, sourceEvent) {
        return this.dispatchEvent(new CustomEvent(type, {
            detail: this.createCloseDetail(sourceEvent),
            bubbles: true,
            composed: true,
            cancelable: type === VE_MESSAGE_CLOSE_EVENT
        }));
    }
    finishClose(sourceEvent) {
        this.phase = 'closed';
        this.style.removeProperty('height');
        this.style.removeProperty('overflow');
        this.style.removeProperty('transition');
        this.dispatchMessageEvent(VE_MESSAGE_AFTER_CLOSE_EVENT, sourceEvent);
    }
    renderDefaultIcon() {
        switch(this.normalizedType){
            case 'success':
                return html`
          <svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
            <path
              d="m7.95 12 2.7 2.7 5.4-5.4M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            ></path>
          </svg>
        `;
            case 'warning':
                return html`
          <svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
            <path
              d="M12 8v5m0 3v.1M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            ></path>
          </svg>
        `;
            case 'error':
                return html`
          <svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
            <path
              d="m14.7 9.3-5.4 5.4m0-5.4 5.4 5.4M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            ></path>
          </svg>
        `;
            case 'loading':
                return html`
          <svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
            <path
              d="M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0"
              opacity=".3"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            ></path>
            <path
              d="M21 12a9 9 0 0 1-9 9"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            ></path>
          </svg>
        `;
            case 'info':
            default:
                return html`
          <svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
            <path
              d="M12 11v5m0 5a9 9 0 1 1 0-18 9 9 0 0 1 0 18m.05-13v.1h-.1V8z"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="1.5"
            ></path>
          </svg>
        `;
        }
    }
    renderIcon() {
        if (!this.shouldRenderIcon) return nothing;
        return html`
      <span class="ve-message-icon" part="icon" aria-hidden="true">
        <slot name="icon" @slotchange=${this.handleSlotChange}></slot>
        ${this.hasIconContent ? nothing : this.renderDefaultIcon()}
      </span>
    `;
    }
    renderCloseIcon() {
        return html`
      <slot name="close-icon" @slotchange=${this.handleSlotChange}></slot>
      ${this.hasCloseIconContent ? nothing : html`<ve-icon
            .definition=${closeIcon}
            size="var(--ve-message-close-icon-size)"
          ></ve-icon>`}
    `;
    }
    render() {
        if ('closed' === this.phase) return nothing;
        const messageClass = classMap({
            've-message': true,
            [`ve-message-${this.normalizedType}`]: true,
            've-message-multiline': this.contentMultiline,
            've-message-closable': this.closable,
            've-message-entered': 'entered' === this.phase,
            've-message-leaving': 'leaving' === this.phase
        });
        return html`<div
      class=${messageClass}
      part="message"
      role=${this.messageRole}
      aria-live=${this.messageAriaLive}
      @mouseenter=${this.handleMouseEnter}
      @mouseleave=${this.handleMouseLeave}
      @focusin=${this.handleFocusIn}
      @focusout=${this.handleFocusOut}
    >
      ${this.renderIcon()}
      <div class="ve-message-content" part="content">
        <slot @slotchange=${this.handleSlotChange}>${this.content}</slot>
      </div>
      ${this.closable ? html`<button
            type="button"
            class="ve-message-close-btn"
            part="close-button"
            aria-label="Close message"
            @click=${this.handleClose}
          >
            ${this.renderCloseIcon()}
          </button>` : nothing}
    </div>`;
    }
}
defineCustomElement(VE_MESSAGE_TAG, VeMessageElement);
let holders = {};
const messageOnCloseCallbacks = new WeakMap();
const messageContentCleanups = new WeakMap();
const messageRecordByElement = new WeakMap();
function isBrowser() {
    return "u" > typeof document;
}
function isNode(value) {
    return "u" > typeof Node && value instanceof Node;
}
function isContentRenderer(value) {
    return 'function' == typeof value;
}
function isMessageContent(value) {
    return 'string' == typeof value || 'function' == typeof value || isNode(value);
}
function normalizeInput(type, input) {
    const options = isMessageContent(input) ? {
        content: input
    } : input;
    return {
        ...options,
        type
    };
}
function generateMessageId() {
    if ("u" > typeof crypto && 'function' == typeof crypto.randomUUID) return `ve_notice_id_${crypto.randomUUID()}`;
    return `ve_notice_id_${Math.random().toFixed(10).slice(2)}`;
}
function getContainerElement() {
    if (!isBrowser()) return null;
    return document.body;
}
function applyHolderStyle(holder, position) {
    holder.className = `ve-message-wrapper ve-message-wrapper-${position}`;
    holder.dataset.veMessagePosition = position;
    holder.style.width = '100%';
    holder.style.position = 'fixed';
    holder.style.zIndex = 'var(--ve-message-z-index, var(--z-index-message, 1003))';
    holder.style.padding = '0 var(--ve-message-wrapper-padding-inline, 12px)';
    holder.style.textAlign = 'center';
    holder.style.pointerEvents = 'none';
    holder.style.boxSizing = 'border-box';
    holder.style.left = '0';
    if ('top' === position) {
        holder.style.top = 'var(--ve-message-wrapper-margin-top, 40px)';
        holder.style.bottom = '';
    } else {
        holder.style.top = '';
        holder.style.bottom = 'var(--ve-message-wrapper-margin-bottom, 40px)';
    }
}
function getHolder(position) {
    const container = getContainerElement();
    if (!container) return null;
    const existing = holders[position];
    if (existing && existing.container === container) return existing;
    const holder = document.createElement('div');
    applyHolderStyle(holder, position);
    container.appendChild(holder);
    const record = {
        holder,
        messages: [],
        container,
        position
    };
    holders[position] = record;
    return record;
}
function getNodeSlotName(node) {
    if (node instanceof Element) return node.getAttribute('slot') ?? '';
    return '';
}
function removeSlottedChildren(element, slotName) {
    for (const child of Array.from(element.childNodes))if (getNodeSlotName(child) === slotName) child.remove();
}
function toSlottedNode(value, slotName) {
    if (!isNode(value)) {
        const span = document.createElement('span');
        span.textContent = value;
        if (slotName) span.slot = slotName;
        return span;
    }
    if (!slotName) {
        if (value instanceof Element) value.removeAttribute('slot');
        return value;
    }
    if (value instanceof Element) {
        value.setAttribute('slot', slotName);
        return value;
    }
    const span = document.createElement('span');
    span.slot = slotName;
    span.appendChild(value);
    return span;
}
function setSlottedContent(element, slotName, value) {
    removeSlottedChildren(element, slotName);
    if (void 0 === value) return;
    element.appendChild(toSlottedNode(value, slotName));
}
function cleanupRenderedContent(element) {
    const cleanup = messageContentCleanups.get(element);
    if (!cleanup) return;
    messageContentCleanups.delete(element);
    cleanup();
}
function applyContentToElement(element, content) {
    cleanupRenderedContent(element);
    if ('string' == typeof content) {
        removeSlottedChildren(element, '');
        element.content = content;
        return;
    }
    element.content = '';
    removeSlottedChildren(element, '');
    if (!isContentRenderer(content)) return void element.appendChild(toSlottedNode(content, ''));
    const container = document.createElement('div');
    element.appendChild(container);
    const result = content(container, {
        element,
        close: ()=>element.close()
    });
    if ('function' == typeof result) return void messageContentCleanups.set(element, result);
    if ('string' == typeof result) {
        container.textContent = result;
        return;
    }
    if (isNode(result)) container.appendChild(result);
}
function applyOptionsToElement(element, options, fallbackId) {
    element.type = options.type ?? 'info';
    element.position = options.position ?? 'top';
    element.messageId = options.id ?? fallbackId;
    element.duration = options.duration ?? MESSAGE_DEFAULT_DURATION;
    element.showIcon = options.showIcon ?? true;
    element.closable = options.closable ?? false;
    element.enterDuration = MESSAGE_DEFAULT_ENTER_DURATION;
    element.exitDuration = MESSAGE_DEFAULT_EXIT_DURATION;
    applyContentToElement(element, options.content);
    setSlottedContent(element, 'icon', options.icon);
    setSlottedContent(element, 'close-icon', options.closeIcon);
}
function cleanupHolder(record) {
    if (record.holder.childElementCount > 0) return;
    record.holder.remove();
    if (holders[record.position] === record) delete holders[record.position];
}
function detachMessageFromRecord(record, element) {
    record.messages = record.messages.filter((message)=>message !== element);
}
function removeMessageFromRecord(record, element) {
    detachMessageFromRecord(record, element);
    messageRecordByElement.delete(element);
    cleanupRenderedContent(element);
    element.remove();
    cleanupHolder(record);
}
function findMessageById(id) {
    for (const record of Object.values(holders)){
        const message = record.messages.find((item)=>item.messageId === id);
        if (message) return {
            record,
            message
        };
    }
    return null;
}
function moveMessageToRecord(element, currentRecord, nextRecord) {
    if (currentRecord === nextRecord) return;
    detachMessageFromRecord(currentRecord, element);
    nextRecord.messages.push(element);
    messageRecordByElement.set(element, nextRecord);
    nextRecord.holder.appendChild(element);
    cleanupHolder(currentRecord);
}
function addInstance(options) {
    const position = options.position ?? 'top';
    const record = getHolder(position);
    if (!record) return ()=>void 0;
    const existing = options.id ? findMessageById(options.id) : null;
    if (existing) {
        moveMessageToRecord(existing.message, existing.record, record);
        applyOptionsToElement(existing.message, options, existing.message.messageId);
        messageOnCloseCallbacks.set(existing.message, options.onClose);
        existing.message.restartTimer();
        return ()=>existing.message.close();
    }
    const element = document.createElement(VE_MESSAGE_TAG);
    const id = options.id ?? generateMessageId();
    applyOptionsToElement(element, options, id);
    messageOnCloseCallbacks.set(element, options.onClose);
    element.addEventListener(VE_MESSAGE_AFTER_CLOSE_EVENT, ()=>{
        const currentRecord = messageRecordByElement.get(element);
        if (currentRecord) removeMessageFromRecord(currentRecord, element);
        else element.remove();
        const onClose = messageOnCloseCallbacks.get(element);
        messageOnCloseCallbacks.delete(element);
        onClose?.();
    }, {
        once: true
    });
    record.messages.push(element);
    messageRecordByElement.set(element, record);
    record.holder.appendChild(element);
    return ()=>element.close();
}
function clearMessages() {
    for (const record of Object.values(holders))for (const message of [
        ...record.messages
    ])message.close();
}
const Message = {
    info (config) {
        return addInstance(normalizeInput('info', config));
    },
    success (config) {
        return addInstance(normalizeInput('success', config));
    },
    warning (config) {
        return addInstance(normalizeInput('warning', config));
    },
    error (config) {
        return addInstance(normalizeInput('error', config));
    },
    loading (config) {
        return addInstance(normalizeInput('loading', config));
    },
    addInstance (config) {
        return addInstance(config);
    },
    clear () {
        clearMessages();
    }
};
const ve_message_ve_message = Message;
export default ve_message_ve_message;
export { Message, VE_MESSAGE_AFTER_CLOSE_EVENT, VE_MESSAGE_CLOSE_EVENT, VE_MESSAGE_TAG, VeMessageElement };
