// @ve-design/web/web/dist/ve-resource-preview.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { volumeIcon } from "./icons/volume.js";
import { zoomInIcon } from "./icons/zoom-in.js";
import { zoomOutIcon } from "./icons/zoom-out.js";
import { VeButtonElement, VE_BUTTON_TAG } from "./68533.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { downloadIcon } from "./icons/download.js";
import { closeIcon } from "./icons/close.js";
import { chevronLeftMdIcon } from "./icons/chevron-left-md.js";
import { chevronRightMdIcon } from "./icons/chevron-right-md.js";
const veResourcePreviewTokenDefaults = css`
  :host {
    /* ===== 卡片尺寸 ===== */
    --ve-resource-preview-image-width: 194px;
    --ve-resource-preview-video-width: 396px;

    --ve-resource-preview-image-aspect-ratio: 1 / 1;
    --ve-resource-preview-video-aspect-ratio: 16 / 9;

    --ve-resource-preview-current-width: var(--ve-resource-preview-image-width);
    --ve-resource-preview-current-aspect-ratio: var(
      --ve-resource-preview-image-aspect-ratio
    );

    /* ===== 卡片外观 ===== */
    --ve-resource-preview-radius: var(--radius-lg);
    --ve-resource-preview-background: var(--color-bg-muted);
    --ve-resource-preview-border-color: var(--color-border-default);
    --ve-resource-preview-border-width: var(--stroke-weight-base);
    --ve-resource-preview-text-color: var(--color-text-primary);
    --ve-resource-preview-focus-ring: 0 0 0 var(--stroke-weight-bold)
      color-mix(var(--color-text-accent) 28%, transparent);

    /* ===== 下载按钮浮层 ===== */
    --ve-resource-preview-action-offset: var(--space-xxs);

    /* ===== 角标 ===== */
    --ve-resource-preview-tip-offset: var(--space-xxs);
    --ve-resource-preview-tip-height: 18px;
    --ve-resource-preview-tip-padding-inline: var(--space-xxxs);
    --ve-resource-preview-tip-radius: var(--radius-sm);
    --ve-resource-preview-tip-color: var(--color-white, #FFF);
    --ve-resource-preview-tip-background: rgb(0 0 0 / 28%);
    --ve-resource-preview-tip-border: rgb(255 255 255 / 48%);
    --ve-resource-preview-tip-font-size: 12px;
    --ve-resource-preview-tip-font-weight: 400;

    /* ===== 媒体播放控制条 ===== */
    --ve-resource-preview-control-height: 36px;
    --ve-resource-preview-control-icon-size: 16px;
    --ve-resource-preview-control-track-height: 4px;
    --ve-resource-preview-control-padding-block: var(--space-xxs);
    --ve-resource-preview-control-padding-inline: var(--space-xs);
    --ve-resource-preview-control-gap: var(--space-xxs);
    --ve-resource-preview-control-row-gap: var(--space-xs);
    --ve-resource-preview-control-background: linear-gradient(
      to top,
      color-mix(var(--color-bg-inverse) 55%, transparent),
      transparent
    );
    --ve-resource-preview-control-text-color: var(--color-white, #FFF);
    --ve-resource-preview-control-track-color: color-mix(
      var(--color-white, #FFF) 34%,
      transparent
    );
    --ve-resource-preview-control-progress-color: var(--color-white, #FFF);
    --ve-resource-preview-control-time-font-size: var(--text-body-sm);
    --ve-resource-preview-control-time-font-family: var(--font-en);
    --ve-resource-preview-control-time-line-height: var(--line-height-body);

    /* ===== 大预览层 ===== */
    --ve-resource-preview-viewer-overlay-background: color-mix(
      var(--color-bg-inverse) 60%,
      transparent
    );
    --ve-resource-preview-viewer-background: var(
      --ve-resource-preview-viewer-overlay-background
    );
    --ve-resource-preview-viewer-scrim: color-mix(
      var(--color-bg-inverse) 45%,
      transparent
    );
    --ve-resource-preview-viewer-stage-background: transparent;
    --ve-resource-preview-viewer-toolbar-background: var(
      --ve-resource-preview-viewer-overlay-background
    );
    --ve-resource-preview-viewer-toolbar-border: transparent;
    --ve-resource-preview-viewer-toolbar-shadow: var(--shadow-md);
    --ve-resource-preview-viewer-media-shadow: none;
    --ve-resource-preview-viewer-count-color: var(--color-text-foreground);
    --ve-resource-preview-viewer-action-color: var(--color-text-foreground);
    --ve-resource-preview-viewer-action-size: var(--height-default);
    --ve-resource-preview-viewer-toolbar-icon-size: 20px;
    --ve-resource-preview-viewer-toolbar-gap: var(--space-s);
    --ve-resource-preview-viewer-toolbar-padding-block: var(--space-xs);
    --ve-resource-preview-viewer-toolbar-padding-inline: var(--space-s);
    --ve-resource-preview-viewer-toolbar-bottom: var(--space-xxl);
    --ve-resource-preview-viewer-close-offset: calc(var(--space-xxl) * -1);
    --ve-resource-preview-viewer-image-max-size: 560px;
    --ve-resource-preview-viewer-video-max-width: 998px;
    --ve-resource-preview-viewer-video-max-height: 560px;
    --ve-resource-preview-viewer-video-control-track-height: 6px;
    --ve-resource-preview-viewer-video-control-padding-inline: var(--space-l);
    --ve-resource-preview-viewer-video-control-padding-block: var(--space-xs);
    --ve-resource-preview-viewer-video-control-gap: var(--space-s);

    /* ===== 分组间距（白名单枚举） ===== */
    --ve-resource-preview-group-gap: var(--space-xs);

    /* ===== 运动 ===== */
    --ve-resource-preview-transition-duration: var(--motion-fast);
    --ve-resource-preview-transition-timing: ease;
  }

  :host([type='video']) {
    --ve-resource-preview-current-width: var(--ve-resource-preview-video-width);
    --ve-resource-preview-current-aspect-ratio: var(
      --ve-resource-preview-video-aspect-ratio
    );
  }

  :host([gap='small']) {
    --ve-resource-preview-group-gap: var(--space-xxs);
  }

  :host([gap='medium']) {
    --ve-resource-preview-group-gap: var(--space-xs);
  }

  :host([gap='large']) {
    --ve-resource-preview-group-gap: var(--space-m);
  }
`;
const token = veResourcePreviewTokenDefaults;
const viewerStyles = css`
  .ve-resource-preview-icon {
    display: block;
    inline-size: 1em;
    block-size: 1em;
    color: inherit;
    fill: none;
  }

  /* Icons placed as slotted children of ve-button inherit host color,
     which is usually not the same as the button's internal text color.
     Force an explicit color based on the enclosing context. */
  .ve-resource-preview-action .ve-resource-preview-icon {
    color: #FFF;
    font-size: 16px;
  }

  .ve-resource-preview-controls .ve-resource-preview-icon {
    color: var(--ve-resource-preview-control-text-color);
    font-size: var(--ve-resource-preview-control-icon-size);
  }

  .ve-resource-preview-controls-viewer .ve-resource-preview-icon,
  .ve-resource-preview-viewer-close .ve-resource-preview-icon,
  .ve-resource-preview-viewer-toolbar .ve-resource-preview-icon {
    color: var(--ve-resource-preview-viewer-action-color);
    font-size: var(--ve-resource-preview-viewer-toolbar-icon-size);
  }

  .ve-resource-preview-controls-viewer .ve-resource-preview-icon {
    color: var(--ve-resource-preview-control-text-color);
  }

  .ve-resource-preview-control-button {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    flex: none;
    inline-size: var(--ve-resource-preview-control-icon-size);
    block-size: var(--ve-resource-preview-control-icon-size);
    color: currentColor;
    font-size: var(--ve-resource-preview-control-icon-size);
  }

  .ve-resource-preview-control-button ve-button {
    --ve-button-height-large: var(--ve-resource-preview-control-icon-size);
    --ve-button-current-height: var(--ve-resource-preview-control-icon-size);
    --ve-button-current-font-size: var(--ve-resource-preview-control-icon-size);
    --ve-button-text-text: var(--ve-resource-preview-control-text-color);
    --ve-button-current-background: transparent;
    --ve-button-current-hover-mix: transparent;
    --ve-button-current-pressed-mix: transparent;
  }

  .ve-resource-preview-controls {
    position: absolute;
    left: 0;
    right: 0;
    bottom: 0;
    z-index: 2;
    display: flex;
    align-items: center;
    gap: var(--ve-resource-preview-control-gap);
    min-block-size: var(--ve-resource-preview-control-height);
    padding-inline: var(--ve-resource-preview-control-padding-inline);
    box-sizing: border-box;
    color: var(--ve-resource-preview-control-text-color);
    background: var(--ve-resource-preview-control-background);
  }

  .ve-resource-preview-controls-video {
    display: flex;
    flex-direction: column;
    align-items: stretch;
    gap: 0;
    min-block-size: auto;
    padding-inline: 0;
    background: transparent;
  }

  .ve-resource-preview-controls-row {
    order: 1;
    display: flex;
    align-items: center;
    justify-content: space-between;
    min-block-size: var(--ve-resource-preview-control-height);
    padding-block: var(--ve-resource-preview-control-padding-block);
    padding-inline: var(--space-s) var(--space-xxs);
    box-sizing: border-box;
    background: var(--ve-resource-preview-control-background);
  }

  .ve-resource-preview-controls-main {
    display: inline-flex;
    align-items: center;
    gap: var(--ve-resource-preview-control-row-gap);
    min-inline-size: 0;
  }

  .ve-resource-preview-controls-video ve-button {
    --ve-button-text-text: var(--ve-resource-preview-control-text-color);
  }

  .ve-resource-preview-controls-viewer {
    position: absolute;
    padding: var(--ve-resource-preview-viewer-video-control-padding-block)
      var(--ve-resource-preview-viewer-video-control-padding-inline);
    min-block-size: auto;
  }

  .ve-resource-preview-progress {
    appearance: none;
    flex: 1 1 auto;
    min-inline-size: 42px;
    block-size: 12px;
    border-radius: var(--radius-full);
    background: transparent;
    cursor: pointer;
  }

  .ve-resource-preview-progress:focus-visible {
    outline: none;
    box-shadow: var(--ve-resource-preview-focus-ring);
  }

  .ve-resource-preview-progress::-webkit-slider-runnable-track {
    block-size: var(--ve-resource-preview-control-track-height);
    border-radius: var(--radius-full);
    background: linear-gradient(
      to right,
      var(--ve-resource-preview-control-progress-color) 0%,
      var(--ve-resource-preview-control-progress-color)
        var(--ve-resource-preview-progress, 0%),
      var(--ve-resource-preview-control-track-color)
        var(--ve-resource-preview-progress, 0%),
      var(--ve-resource-preview-control-track-color) 100%
    );
  }

  .ve-resource-preview-progress::-moz-range-track {
    block-size: var(--ve-resource-preview-control-track-height);
    border-radius: var(--radius-full);
    background: var(--ve-resource-preview-control-track-color);
  }

  .ve-resource-preview-progress::-moz-range-progress {
    block-size: var(--ve-resource-preview-control-track-height);
    border-radius: var(--radius-full);
    background: var(--ve-resource-preview-control-progress-color);
  }

  .ve-resource-preview-progress::-webkit-slider-thumb {
    appearance: none;
    inline-size: 0;
    block-size: 0;
  }

  .ve-resource-preview-progress::-moz-range-thumb {
    inline-size: 0;
    block-size: 0;
    border: none;
  }

  .ve-resource-preview-controls-video .ve-resource-preview-progress {
    order: 2;
    flex: none;
    inline-size: 100%;
    min-inline-size: 0;
    block-size: var(--ve-resource-preview-control-track-height);
  }

  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer {
    gap: var(--ve-resource-preview-viewer-video-control-gap);
    background: var(--ve-resource-preview-control-background);
  }

  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer
    .ve-resource-preview-controls-row {
    order: 2;
    min-block-size: var(--ve-resource-preview-viewer-toolbar-icon-size);
    padding: 0;
    background: transparent;
  }

  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer
    .ve-resource-preview-progress {
    order: 1;
    block-size: var(--ve-resource-preview-viewer-video-control-track-height);
  }

  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer
    .ve-resource-preview-progress::-webkit-slider-runnable-track {
    block-size: var(--ve-resource-preview-viewer-video-control-track-height);
  }

  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer
    .ve-resource-preview-progress::-moz-range-track,
  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer
    .ve-resource-preview-progress::-moz-range-progress {
    block-size: var(--ve-resource-preview-viewer-video-control-track-height);
  }

  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer
    .ve-resource-preview-control-button,
  .ve-resource-preview-controls-video.ve-resource-preview-controls-viewer
    .ve-resource-preview-control-button
    ve-button {
    --ve-button-height-large: var(
      --ve-resource-preview-viewer-toolbar-icon-size
    );
    --ve-button-current-height: var(
      --ve-resource-preview-viewer-toolbar-icon-size
    );
    --ve-button-current-font-size: var(
      --ve-resource-preview-viewer-toolbar-icon-size
    );
    inline-size: var(--ve-resource-preview-viewer-toolbar-icon-size);
    block-size: var(--ve-resource-preview-viewer-toolbar-icon-size);
  }

  .ve-resource-preview-time {
    flex: none;
    color: currentColor;
    font-family: var(--ve-resource-preview-control-time-font-family);
    font-size: var(--ve-resource-preview-control-time-font-size);
    line-height: var(--ve-resource-preview-control-time-line-height);
    white-space: nowrap;
  }

  dialog.ve-resource-preview-viewer {
    position: fixed;
    inset: 0;
    inline-size: 100vw;
    block-size: 100vh;
    max-inline-size: 100vw;
    max-block-size: 100vh;
    margin: 0;
    padding: 0;
    border: none;
    z-index: 1000;
    display: flex;
    flex-direction: column;
    box-sizing: border-box;
    background: var(--ve-resource-preview-viewer-background);
    color: var(--ve-resource-preview-viewer-count-color);
    outline: none;
  }

  dialog.ve-resource-preview-viewer:not([open]) {
    display: none;
  }

  dialog.ve-resource-preview-viewer::backdrop {
    background: transparent;
  }

  .ve-resource-preview-viewer-body {
    position: relative;
    display: flex;
    align-items: center;
    justify-content: center;
    flex: 1 1 auto;
    min-block-size: 0;
    margin: 0;
    padding: var(--space-xxl);
    overflow: auto;
    box-sizing: border-box;
    background: var(--ve-resource-preview-viewer-stage-background);
  }

  .ve-resource-preview-viewer-content {
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    max-inline-size: 100%;
    max-block-size: 100%;
  }

  .ve-resource-preview-viewer-close {
    position: absolute;
    inset-block-start: 0;
    inset-inline-end: var(--ve-resource-preview-viewer-close-offset);
    inline-size: var(--ve-resource-preview-viewer-action-size);
    block-size: var(--ve-resource-preview-viewer-action-size);
    color: var(--ve-resource-preview-viewer-action-color);
  }

  .ve-resource-preview-viewer-close ve-button,
  .ve-resource-preview-viewer-toolbar ve-button {
    --ve-button-height-large: var(
      --ve-resource-preview-viewer-toolbar-icon-size
    );
    --ve-button-current-height: var(
      --ve-resource-preview-viewer-toolbar-icon-size
    );
    --ve-button-current-font-size: var(
      --ve-resource-preview-viewer-toolbar-icon-size
    );
    --ve-button-text-text: var(--ve-resource-preview-viewer-action-color);
    --ve-button-current-background: transparent;
    --ve-button-current-hover-mix: transparent;
    --ve-button-current-pressed-mix: transparent;
  }

  .ve-resource-preview-viewer-close ve-button {
    --ve-button-height-large: var(--ve-resource-preview-viewer-action-size);
    --ve-button-current-height: var(--ve-resource-preview-viewer-action-size);
  }

  .ve-resource-preview-viewer-image {
    display: block;
    max-inline-size: min(
      72vw,
      var(--ve-resource-preview-viewer-image-max-size)
    );
    max-block-size: min(78vh, var(--ve-resource-preview-viewer-image-max-size));
    object-fit: contain;
    box-shadow: var(--ve-resource-preview-viewer-media-shadow);
    background: var(--color-bg-overlay);
    transform: scale(var(--ve-resource-preview-viewer-scale, 1));
    transform-origin: center;
    transition: transform var(--ve-resource-preview-transition-duration)
      var(--ve-resource-preview-transition-timing);
  }

  .ve-resource-preview-viewer-media-wrap {
    position: relative;
    inline-size: min(72vw, var(--ve-resource-preview-viewer-video-max-width));
    max-block-size: var(--ve-resource-preview-viewer-video-max-height);
    aspect-ratio: var(--ve-resource-preview-video-aspect-ratio);
    overflow: visible;
    background: var(--color-bg-inverse);
  }

  .ve-resource-preview-viewer-media {
    display: block;
    inline-size: 100%;
    block-size: 100%;
    object-fit: contain;
    transform: scale(var(--ve-resource-preview-viewer-scale, 1));
    transform-origin: center;
    transition: transform var(--ve-resource-preview-transition-duration)
      var(--ve-resource-preview-transition-timing);
  }

  .ve-resource-preview-viewer-toolbar {
    position: fixed;
    left: 50%;
    bottom: var(--ve-resource-preview-viewer-toolbar-bottom);
    display: inline-flex;
    align-items: center;
    gap: var(--ve-resource-preview-viewer-toolbar-gap);
    min-block-size: auto;
    padding: var(--ve-resource-preview-viewer-toolbar-padding-block)
      var(--ve-resource-preview-viewer-toolbar-padding-inline);
    border: var(--ve-resource-preview-border-width) solid
      var(--ve-resource-preview-viewer-toolbar-border);
    border-radius: var(--radius-md);
    background: var(--ve-resource-preview-viewer-toolbar-background);
    box-shadow: var(--ve-resource-preview-viewer-toolbar-shadow);
    transform: translateX(-50%);
  }

  .ve-resource-preview-viewer-count {
    min-inline-size: 48px;
    color: var(--ve-resource-preview-viewer-count-color);
    font-family: var(--ve-resource-preview-control-time-font-family);
    font-size: var(--text-body);
    font-weight: var(--font-weight-normal);
    line-height: 22px;
    text-align: center;
    white-space: nowrap;
  }

  @media (max-width: 640px) {
    .ve-resource-preview-viewer-body {
      padding: var(--space-s);
    }

    .ve-resource-preview-viewer-close {
      inset-inline-end: 0;
      inset-block-start: calc(
        var(--ve-resource-preview-viewer-action-size) * -1
      );
    }

    .ve-resource-preview-viewer-image,
    .ve-resource-preview-viewer-media-wrap {
      inline-size: 100%;
      max-inline-size: 100%;
    }
  }

  @media (prefers-reduced-motion: reduce) {
    .ve-resource-preview-viewer-image,
    .ve-resource-preview-viewer-media {
      transition: none;
    }
  }
`;
const viewer = viewerStyles;
const cardStyles = css`
  :host {
    display: inline-block;
    inline-size: var(--ve-resource-preview-current-width);
    max-inline-size: 100%;
    vertical-align: top;
    font-family: var(--font-sans);
  }

  :host([hidden]) {
    display: none;
  }

    .ve-resource-preview {
      position: relative;
      inline-size: 100%;
      max-inline-size: 100%;
      box-sizing: border-box;
      border-radius: var(--ve-resource-preview-radius);
      color: var(--ve-resource-preview-text-color);
    }

    :host([type='video']) .ve-resource-preview {
      overflow: hidden;
      background: var(--ve-resource-preview-background);
      clip-path: inset(0 round var(--ve-resource-preview-radius));
    }

  .ve-resource-preview-surface {
    position: relative;
    display: block;
    inline-size: 100%;
    aspect-ratio: var(--ve-resource-preview-current-aspect-ratio);
    overflow: hidden;
    box-sizing: border-box;
    border-radius: var(--ve-resource-preview-radius);
    background: var(--ve-resource-preview-background);
    clip-path: inset(0 round var(--ve-resource-preview-radius));
    cursor: pointer;
    outline: none;
    transition:
      box-shadow var(--ve-resource-preview-transition-duration)
        var(--ve-resource-preview-transition-timing),
      transform var(--ve-resource-preview-transition-duration)
        var(--ve-resource-preview-transition-timing);
  }

  :host([preview-disabled]) .ve-resource-preview-surface,
  .ve-resource-preview-no-source .ve-resource-preview-surface {
    cursor: default;
  }

  .ve-resource-preview-surface:focus-visible {
    box-shadow:
      inset 0 0 0 var(--ve-resource-preview-border-width)
        var(--ve-resource-preview-border-color),
      var(--ve-resource-preview-focus-ring);
  }

  .ve-resource-preview-image,
  .ve-resource-preview-video {
    display: block;
    inline-size: 100%;
    block-size: 100%;
    border-radius: inherit;
    object-fit: cover;
    user-select: none;
  }

  .ve-resource-preview-video {
    background: var(--color-bg-inverse);
  }

  .ve-resource-preview-tip {
    position: absolute;
    top: var(--ve-resource-preview-tip-offset);
    left: var(--ve-resource-preview-tip-offset);
    z-index: 3;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    box-sizing: border-box;
    block-size: var(--ve-resource-preview-tip-height);
    max-inline-size: calc(
      100% - var(--ve-resource-preview-tip-offset) * 2
    );
    padding-inline: var(--ve-resource-preview-tip-padding-inline);
    border: var(--ve-resource-preview-border-width) solid
      var(--ve-resource-preview-tip-border);
    border-radius: var(--ve-resource-preview-tip-radius);
    color: var(--ve-resource-preview-tip-color);
    background: var(--ve-resource-preview-tip-background);
    font-size: var(--ve-resource-preview-tip-font-size);
    font-weight: var(--ve-resource-preview-tip-font-weight);
    line-height: var(--ve-resource-preview-tip-height);
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    pointer-events: none;
  }

  .ve-resource-preview-action {
    position: absolute;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    right: var(--ve-resource-preview-action-offset);
    bottom: var(--ve-resource-preview-action-offset);
    z-index: 3;
    opacity: 0;
    transform: translateY(2px);
    pointer-events: none;
    transition:
      opacity var(--ve-resource-preview-transition-duration)
        var(--ve-resource-preview-transition-timing),
      transform var(--ve-resource-preview-transition-duration)
        var(--ve-resource-preview-transition-timing);
  }

  .ve-resource-preview-action ve-button {
    --ve-button-current-background: rgba(0, 0, 0, 0.60);
    --ve-button-current-background-hover: rgba(0, 0, 0, 0.60);
    --ve-button-current-background-pressed: rgba(0, 0, 0, 0.60);
    --ve-button-current-text: #FFF;
    --ve-button-current-border: transparent;
    --ve-button-current-border-hover: transparent;
    --ve-button-current-border-pressed: transparent;
  }

  .ve-resource-preview:hover .ve-resource-preview-action,
  .ve-resource-preview:focus-within .ve-resource-preview-action,
  .ve-resource-preview-action:focus-visible {
    opacity: 1;
    transform: translateY(0);
    pointer-events: auto;
  }

  @media (prefers-reduced-motion: reduce) {
    .ve-resource-preview-surface,
    .ve-resource-preview-action {
      transition: none;
    }
  }
`;
const ve_resource_preview = [
    componentReset,
    token,
    cardStyles,
    viewer
];
const RESOURCE_PREVIEW_TYPES = [
    'image',
    'video'
];
const RESOURCE_PREVIEW_PRELOAD_VALUES = [
    'none',
    'metadata',
    'auto'
];
const RESOURCE_PREVIEW_GAP_VALUES = [
    'small',
    'medium',
    'large'
];
const VIEWER_SCALE_DEFAULT = 1;
const VIEWER_SCALE_MIN = 0.5;
const VIEWER_SCALE_MAX = 3;
const VIEWER_SCALE_STEP = 0.25;
function normalizeEnum(value, allowed, fallback) {
    return allowed.includes(value) ? value : fallback;
}
function clampScale(value) {
    return Math.min(VIEWER_SCALE_MAX, Math.max(VIEWER_SCALE_MIN, value));
}
function nextScale(current, delta) {
    return clampScale(Math.round((current + delta) * 100) / 100);
}
function formatMediaTime(value) {
    if (!Number.isFinite(value) || value <= 0) return '00:00';
    const totalSeconds = Math.floor(value);
    const minutes = Math.floor(totalSeconds / 60);
    const seconds = totalSeconds % 60;
    return `${String(minutes).padStart(2, '0')}:${String(seconds).padStart(2, '0')}`;
}
function triggerResourceDownload(item) {
    if (!item.src || "u" < typeof document) return;
    const anchor = document.createElement('a');
    anchor.href = item.src;
    anchor.download = item.name;
    anchor.target = '_blank';
    anchor.rel = 'noopener';
    anchor.click();
}
function renderIcon(definition) {
    return html`<ve-icon
    class="ve-resource-preview-icon"
    .definition=${definition}
  ></ve-icon>`;
}
function renderPlayIcon() {
    return html`<svg
    class="ve-resource-preview-icon"
    viewBox="0 0 24 24"
    fill="none"
    aria-hidden="true"
    focusable="false"
  >
    <path
      d="M8.5 6.8C8.5 6.1 9.3 5.7 9.9 6.1L17.4 10.8C18 11.2 18 12.1 17.4 12.5L9.9 17.2C9.3 17.6 8.5 17.2 8.5 16.5V6.8Z"
      fill="currentColor"
    ></path>
  </svg>`;
}
function renderPauseIcon() {
    return html`<svg
    class="ve-resource-preview-icon"
    viewBox="0 0 24 24"
    fill="none"
    aria-hidden="true"
    focusable="false"
  >
    <path
      d="M8 6.5V17.5M16 6.5V17.5"
      stroke="currentColor"
      stroke-width="2.4"
      stroke-linecap="round"
    ></path>
  </svg>`;
}
function renderVolumeOffIcon() {
    return html`<svg
    class="ve-resource-preview-icon"
    viewBox="0 0 24 24"
    fill="none"
    aria-hidden="true"
    focusable="false"
  >
    <path
      d="M4.5 9.5H8L13 5.5V18.5L8 14.5H4.5V9.5Z"
      stroke="currentColor"
      stroke-width="1.5"
      stroke-linecap="round"
      stroke-linejoin="round"
    ></path>
    <path
      d="M17 10L20 13M20 10L17 13"
      stroke="currentColor"
      stroke-width="1.5"
      stroke-linecap="round"
    ></path>
  </svg>`;
}
function renderMediaControls(opts) {
    const progress = opts.duration > 0 ? Math.min(opts.currentTime / opts.duration * 1000, 1000) : 0;
    const progressStyle = `--ve-resource-preview-progress:${progress / 10}%;`;
    return html`
    <div
      class=${classMap({
        've-resource-preview-controls': true,
        've-resource-preview-controls-video': true,
        've-resource-preview-controls-viewer': 'viewer' === opts.scope
    })}
      part="controls"
      @click=${(event)=>event.stopPropagation()}
    >
      <div class="ve-resource-preview-controls-row">
        <div class="ve-resource-preview-controls-main">
          <span class="ve-resource-preview-control-button">
            <ve-button
              type="text"
              size="large"
              shape="circle"
              aria-label=${opts.playing ? 'Pause' : 'Play'}
              @click=${opts.onTogglePlay}
            >
              ${opts.playing ? renderPauseIcon() : renderPlayIcon()}
            </ve-button>
          </span>
          <span class="ve-resource-preview-control-button">
            <ve-button
              type="text"
              size="large"
              shape="circle"
              aria-label=${opts.muted ? 'Unmute' : 'Mute'}
              @click=${opts.onToggleMute}
            >
              ${opts.muted ? renderVolumeOffIcon() : renderIcon(volumeIcon)}
            </ve-button>
          </span>
          <span class="ve-resource-preview-time">
            ${formatMediaTime(opts.currentTime)} /
            ${formatMediaTime(opts.duration)}
          </span>
        </div>
      </div>
      <input
        class="ve-resource-preview-progress"
        type="range"
        min="0"
        max="1000"
        step="1"
        .value=${String(progress)}
        aria-label="Media progress"
        style=${progressStyle}
        @input=${opts.onSeek}
        @keydown=${(event)=>event.stopPropagation()}
      />
    </div>
  `;
}
function renderViewerContent(opts) {
    const { item } = opts;
    if ('video' === item.type) return html`
      <div class="ve-resource-preview-viewer-media-wrap">
        <video
          class="ve-resource-preview-viewer-media"
          src=${ifDefined(item.src || void 0)}
          poster=${ifDefined(item.poster || void 0)}
          preload=${item.preload}
          style=${opts.scaleStyle}
          playsinline
          @loadedmetadata=${opts.onMediaEvent}
          @timeupdate=${opts.onMediaEvent}
          @play=${opts.onMediaEvent}
          @pause=${opts.onMediaEvent}
          @ended=${opts.onMediaEvent}
          @volumechange=${opts.onMediaEvent}
        ></video>
        ${opts.mediaControls}
      </div>
    `;
    return html`
    <img
      class="ve-resource-preview-viewer-image"
      src=${item.src}
      alt=${item.name || ''}
      style=${opts.scaleStyle}
      draggable="false"
    />
  `;
}
function renderZoomControls(scale, onZoom) {
    return html`
    <ve-button
      type="text"
      size="large"
      shape="circle"
      aria-label="Zoom in"
      ?disabled=${scale >= VIEWER_SCALE_MAX}
      @click=${()=>onZoom(VIEWER_SCALE_STEP)}
    >
      ${renderIcon(zoomInIcon)}
    </ve-button>
    <ve-button
      type="text"
      size="large"
      shape="circle"
      aria-label="Zoom out"
      ?disabled=${scale <= VIEWER_SCALE_MIN}
      @click=${()=>onZoom(-VIEWER_SCALE_STEP)}
    >
      ${renderIcon(zoomOutIcon)}
    </ve-button>
  `;
}
let bodyScrollLockCount = 0;
let previousBodyOverflow = '';
function lockBodyScroll() {
    if ("u" < typeof document) return;
    if (0 === bodyScrollLockCount) {
        previousBodyOverflow = document.body.style.overflow;
        document.body.style.overflow = 'hidden';
    }
    bodyScrollLockCount += 1;
}
function unlockBodyScroll() {
    if ("u" < typeof document || 0 === bodyScrollLockCount) return;
    bodyScrollLockCount -= 1;
    if (0 === bodyScrollLockCount) {
        document.body.style.overflow = previousBodyOverflow;
        previousBodyOverflow = '';
    }
}
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _init_type, _init_extra__init_type, _init_src, _init_extra__init_src, _init_name, _init_extra__init_name, _init_tip, _init_extra__init_tip, _init_poster, _init_extra__init_poster, _init_preload, _init_extra__init_preload, _init_previewDisabled, _init_extra__init_previewDisabled, _init_hideDownload, _init_extra__init_hideDownload, _init_inlineMediaElement, _init_extra__init_inlineMediaElement, _init_viewerMediaElement, _init_extra__init_viewerMediaElement, _init_viewerDialog, _init_extra__init_viewerDialog, _init_previewVisible, _init_extra__init_previewVisible, _init_playing, _init_extra__init_playing, _init_muted, _init_extra__init_muted, _init_currentTime, _init_extra__init_currentTime, _init_duration, _init_extra__init_duration, _init_mediaAspectRatio, _init_extra__init_mediaAspectRatio, _init_viewerScale, _init_extra__init_viewerScale;
const VE_RESOURCE_PREVIEW_TAG = 've-resource-preview';
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: String
}), _dec2 = property({
    type: String
}), _dec3 = property({
    type: String
}), _dec4 = property({
    type: String
}), _dec5 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec6 = property({
    type: Boolean,
    attribute: 'preview-disabled',
    reflect: true
}), _dec7 = property({
    type: Boolean,
    attribute: 'hide-download',
    reflect: true
}), _dec8 = query('.ve-resource-preview-inline-media'), _dec9 = query('.ve-resource-preview-viewer-media'), _dec10 = query('dialog.ve-resource-preview-viewer'), _dec11 = state(), _dec12 = state(), _dec13 = state(), _dec14 = state(), _dec15 = state(), _dec16 = state(), _dec17 = state();
class VeResourcePreviewElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_src, _init_extra__init_src, _init_name, _init_extra__init_name, _init_tip, _init_extra__init_tip, _init_poster, _init_extra__init_poster, _init_preload, _init_extra__init_preload, _init_previewDisabled, _init_extra__init_previewDisabled, _init_hideDownload, _init_extra__init_hideDownload, _init_inlineMediaElement, _init_extra__init_inlineMediaElement, _init_viewerMediaElement, _init_extra__init_viewerMediaElement, _init_viewerDialog, _init_extra__init_viewerDialog, _init_previewVisible, _init_extra__init_previewVisible, _init_playing, _init_extra__init_playing, _init_muted, _init_extra__init_muted, _init_currentTime, _init_extra__init_currentTime, _init_duration, _init_extra__init_duration, _init_mediaAspectRatio, _init_extra__init_mediaAspectRatio, _init_viewerScale, _init_extra__init_viewerScale] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "type"
            ],
            [
                _dec1,
                1,
                "src"
            ],
            [
                _dec2,
                1,
                "name"
            ],
            [
                _dec3,
                1,
                "tip"
            ],
            [
                _dec4,
                1,
                "poster"
            ],
            [
                _dec5,
                1,
                "preload"
            ],
            [
                _dec6,
                1,
                "previewDisabled"
            ],
            [
                _dec7,
                1,
                "hideDownload"
            ],
            [
                _dec8,
                1,
                "inlineMediaElement"
            ],
            [
                _dec9,
                1,
                "viewerMediaElement"
            ],
            [
                _dec10,
                1,
                "viewerDialog"
            ],
            [
                _dec11,
                1,
                "previewVisible"
            ],
            [
                _dec12,
                1,
                "playing"
            ],
            [
                _dec13,
                1,
                "muted"
            ],
            [
                _dec14,
                1,
                "currentTime"
            ],
            [
                _dec15,
                1,
                "duration"
            ],
            [
                _dec16,
                1,
                "mediaAspectRatio"
            ],
            [
                _dec17,
                1,
                "viewerScale"
            ]
        ]));
    }
    static styles = ve_resource_preview;
    static dependencies = {
        [VE_BUTTON_TAG]: VeButtonElement,
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'image');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_src_2 = (()=>{
        const _value = _init_src(this, '');
        _init_extra__init_src(this);
        return _value;
    })();
    get src() {
        return this.#___private_src_2;
    }
    set src(_v) {
        this.#___private_src_2 = _v;
    }
    #___private_name_3 = (()=>{
        const _value = _init_name(this, '');
        _init_extra__init_name(this);
        return _value;
    })();
    get name() {
        return this.#___private_name_3;
    }
    set name(_v) {
        this.#___private_name_3 = _v;
    }
    #___private_tip_4 = (()=>{
        const _value = _init_tip(this, '');
        _init_extra__init_tip(this);
        return _value;
    })();
    get tip() {
        return this.#___private_tip_4;
    }
    set tip(_v) {
        this.#___private_tip_4 = _v;
    }
    #___private_poster_5 = (()=>{
        const _value = _init_poster(this, '');
        _init_extra__init_poster(this);
        return _value;
    })();
    get poster() {
        return this.#___private_poster_5;
    }
    set poster(_v) {
        this.#___private_poster_5 = _v;
    }
    #___private_preload_6 = (()=>{
        const _value = _init_preload(this, 'metadata');
        _init_extra__init_preload(this);
        return _value;
    })();
    get preload() {
        return this.#___private_preload_6;
    }
    set preload(_v) {
        this.#___private_preload_6 = _v;
    }
    #___private_previewDisabled_7 = (()=>{
        const _value = _init_previewDisabled(this, false);
        _init_extra__init_previewDisabled(this);
        return _value;
    })();
    get previewDisabled() {
        return this.#___private_previewDisabled_7;
    }
    set previewDisabled(_v) {
        this.#___private_previewDisabled_7 = _v;
    }
    #___private_hideDownload_8 = (()=>{
        const _value = _init_hideDownload(this, false);
        _init_extra__init_hideDownload(this);
        return _value;
    })();
    get hideDownload() {
        return this.#___private_hideDownload_8;
    }
    set hideDownload(_v) {
        this.#___private_hideDownload_8 = _v;
    }
    #___private_inlineMediaElement_9 = (()=>{
        const _value = _init_inlineMediaElement(this, null);
        _init_extra__init_inlineMediaElement(this);
        return _value;
    })();
    get inlineMediaElement() {
        return this.#___private_inlineMediaElement_9;
    }
    set inlineMediaElement(_v) {
        this.#___private_inlineMediaElement_9 = _v;
    }
    #___private_viewerMediaElement_10 = (()=>{
        const _value = _init_viewerMediaElement(this, null);
        _init_extra__init_viewerMediaElement(this);
        return _value;
    })();
    get viewerMediaElement() {
        return this.#___private_viewerMediaElement_10;
    }
    set viewerMediaElement(_v) {
        this.#___private_viewerMediaElement_10 = _v;
    }
    #___private_viewerDialog_11 = (()=>{
        const _value = _init_viewerDialog(this, null);
        _init_extra__init_viewerDialog(this);
        return _value;
    })();
    get viewerDialog() {
        return this.#___private_viewerDialog_11;
    }
    set viewerDialog(_v) {
        this.#___private_viewerDialog_11 = _v;
    }
    #___private_previewVisible_12 = (()=>{
        const _value = _init_previewVisible(this, false);
        _init_extra__init_previewVisible(this);
        return _value;
    })();
    get previewVisible() {
        return this.#___private_previewVisible_12;
    }
    set previewVisible(_v) {
        this.#___private_previewVisible_12 = _v;
    }
    #___private_playing_13 = (()=>{
        const _value = _init_playing(this, false);
        _init_extra__init_playing(this);
        return _value;
    })();
    get playing() {
        return this.#___private_playing_13;
    }
    set playing(_v) {
        this.#___private_playing_13 = _v;
    }
    #___private_muted_14 = (()=>{
        const _value = _init_muted(this, false);
        _init_extra__init_muted(this);
        return _value;
    })();
    get muted() {
        return this.#___private_muted_14;
    }
    set muted(_v) {
        this.#___private_muted_14 = _v;
    }
    #___private_currentTime_15 = (()=>{
        const _value = _init_currentTime(this, 0);
        _init_extra__init_currentTime(this);
        return _value;
    })();
    get currentTime() {
        return this.#___private_currentTime_15;
    }
    set currentTime(_v) {
        this.#___private_currentTime_15 = _v;
    }
    #___private_duration_16 = (()=>{
        const _value = _init_duration(this, 0);
        _init_extra__init_duration(this);
        return _value;
    })();
    get duration() {
        return this.#___private_duration_16;
    }
    set duration(_v) {
        this.#___private_duration_16 = _v;
    }
    #___private_mediaAspectRatio_17 = (()=>{
        const _value = _init_mediaAspectRatio(this, '');
        _init_extra__init_mediaAspectRatio(this);
        return _value;
    })();
    get mediaAspectRatio() {
        return this.#___private_mediaAspectRatio_17;
    }
    set mediaAspectRatio(_v) {
        this.#___private_mediaAspectRatio_17 = _v;
    }
    #___private_viewerScale_18 = (()=>{
        const _value = _init_viewerScale(this, VIEWER_SCALE_DEFAULT);
        _init_extra__init_viewerScale(this);
        return _value;
    })();
    get viewerScale() {
        return this.#___private_viewerScale_18;
    }
    set viewerScale(_v) {
        this.#___private_viewerScale_18 = _v;
    }
    get normalizedType() {
        return normalizeEnum(this.type, RESOURCE_PREVIEW_TYPES, 'image');
    }
    get normalizedPreload() {
        return normalizeEnum(this.preload, RESOURCE_PREVIEW_PRELOAD_VALUES, 'metadata');
    }
    get hasSource() {
        return (this.src ?? '').trim().length > 0;
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('previewVisible')) {
            if (this.previewVisible && this.viewerDialog && !this.viewerDialog.open) this.viewerDialog.showModal();
            else if (!this.previewVisible && this.viewerDialog?.open) this.viewerDialog.close();
        }
    }
    disconnectedCallback() {
        if (this.previewVisible) unlockBodyScroll();
        this.pauseMedia('inline');
        this.pauseMedia('viewer');
        super.disconnectedCallback();
    }
    getResourceItem() {
        return {
            type: this.normalizedType,
            src: this.src,
            name: this.name,
            tip: this.tip,
            poster: this.poster,
            preload: this.normalizedPreload,
            downloadable: !this.hideDownload
        };
    }
    openPreview() {
        if (this.previewDisabled || !this.hasSource || this.previewVisible) return;
        this.pauseMedia('inline');
        this.viewerScale = VIEWER_SCALE_DEFAULT;
        this.previewVisible = true;
        lockBodyScroll();
    }
    closePreview() {
        if (!this.previewVisible) return;
        this.pauseMedia('viewer');
        this.previewVisible = false;
        this.viewerScale = VIEWER_SCALE_DEFAULT;
        unlockBodyScroll();
        this.dispatchEvent(new CustomEvent('ve-preview-close', {
            detail: {
                item: this.getResourceItem(),
                source: this
            },
            bubbles: true,
            composed: true
        }));
    }
    getMediaElement(scope) {
        return 'viewer' === scope ? this.viewerMediaElement : this.inlineMediaElement;
    }
    syncMediaState(media) {
        this.currentTime = media.currentTime;
        this.duration = Number.isFinite(media.duration) ? media.duration : 0;
        this.playing = !media.paused && !media.ended;
        this.muted = media.muted;
        if (media instanceof HTMLVideoElement && media.videoWidth > 0 && media.videoHeight > 0) this.mediaAspectRatio = `${media.videoWidth} / ${media.videoHeight}`;
    }
    handleImageLoad = (event)=>{
        const image = event.target instanceof HTMLImageElement ? event.target : null;
        if (!image || image.naturalWidth <= 0 || image.naturalHeight <= 0) return;
        this.mediaAspectRatio = `${image.naturalWidth} / ${image.naturalHeight}`;
    };
    handleMediaEvent = (event)=>{
        const media = event.currentTarget;
        if (media instanceof HTMLMediaElement) this.syncMediaState(media);
    };
    handleSurfaceClick = (event)=>{
        this.requestPreview(event);
    };
    handleKeyboardPreview = (event)=>{
        if ('Enter' !== event.key && ' ' !== event.key) return;
        event.preventDefault();
        this.requestPreview(event);
    };
    handleDialogClose = ()=>{
        if (this.previewVisible) this.closePreview();
    };
    handleDialogCancel = (event)=>{
        event.preventDefault();
        this.closePreview();
    };
    handleZoom = (delta)=>{
        this.viewerScale = nextScale(this.viewerScale, delta);
    };
    get viewerScaleStyle() {
        return `--ve-resource-preview-viewer-scale:${clampScale(this.viewerScale)};`;
    }
    requestPreview(nativeEvent) {
        if (this.previewDisabled || !this.hasSource) return;
        const previewEvent = new CustomEvent('ve-preview', {
            detail: {
                item: this.getResourceItem(),
                source: this,
                nativeEvent
            },
            bubbles: true,
            composed: true,
            cancelable: true
        });
        this.dispatchEvent(previewEvent);
        if (!previewEvent.defaultPrevented) this.openPreview();
    }
    pauseMedia(scope) {
        this.getMediaElement(scope)?.pause();
        this.playing = false;
    }
    async toggleMedia(scope) {
        const media = this.getMediaElement(scope);
        if (!media) return;
        if (media.paused || media.ended) {
            try {
                await media.play();
            } catch  {
                this.playing = false;
            }
            return;
        }
        media.pause();
    }
    toggleMuted(scope) {
        const media = this.getMediaElement(scope);
        if (!media) return;
        media.muted = !media.muted;
        this.muted = media.muted;
    }
    seekMedia(event, scope) {
        const input = event.currentTarget instanceof HTMLInputElement ? event.currentTarget : null;
        const media = this.getMediaElement(scope);
        if (!input || !media || !Number.isFinite(media.duration) || media.duration <= 0) return;
        const progress = Number(input.value);
        if (!Number.isFinite(progress)) return;
        media.currentTime = progress / 1000 * media.duration;
        this.syncMediaState(media);
    }
    emitDownload(event, item) {
        event?.stopPropagation();
        const downloadEvent = new CustomEvent('ve-download', {
            detail: {
                item,
                source: this,
                nativeEvent: event
            },
            bubbles: true,
            composed: true,
            cancelable: true
        });
        this.dispatchEvent(downloadEvent);
        if (!downloadEvent.defaultPrevented) triggerResourceDownload(item);
    }
    renderCardMedia() {
        const type = this.normalizedType;
        if ('video' === type) return html`
        <video
          class="ve-resource-preview-video ve-resource-preview-inline-media"
          src=${ifDefined(this.src || void 0)}
          poster=${ifDefined(this.poster || void 0)}
          preload=${this.normalizedPreload}
          playsinline
          @loadedmetadata=${this.handleMediaEvent}
          @timeupdate=${this.handleMediaEvent}
          @play=${this.handleMediaEvent}
          @pause=${this.handleMediaEvent}
          @ended=${this.handleMediaEvent}
          @volumechange=${this.handleMediaEvent}
        ></video>
        ${renderMediaControls({
            scope: 'inline',
            variant: 'video',
            playing: this.playing,
            muted: this.muted,
            currentTime: this.currentTime,
            duration: this.duration,
            onTogglePlay: ()=>void this.toggleMedia('inline'),
            onToggleMute: ()=>this.toggleMuted('inline'),
            onSeek: (event)=>this.seekMedia(event, 'inline')
        })}
      `;
        return html`
      <img
        class="ve-resource-preview-image"
        src=${ifDefined(this.src || void 0)}
        alt=${this.name || ''}
        draggable="false"
        @load=${this.handleImageLoad}
      />
    `;
    }
    renderTip() {
        const tip = (this.tip ?? '').trim();
        if (!tip) return nothing;
        return html`
      <span class="ve-resource-preview-tip" part="tip">${tip}</span>
    `;
    }
    renderDownloadButton(item) {
        if (!item.downloadable || !item.src) return nothing;
        return html`
      <span class="ve-resource-preview-action" part="action">
        <ve-button
          size="default"
          shape="default"
          aria-label="Download"
          @click=${(event)=>this.emitDownload(event, item)}
        >
          ${renderIcon(downloadIcon)}
        </ve-button>
      </span>
    `;
    }
    renderViewer() {
        const item = this.getResourceItem();
        const showToolbar = this.previewVisible;
        const mediaControls = 'video' === item.type ? renderMediaControls({
            scope: 'viewer',
            variant: 'video',
            playing: this.playing,
            muted: this.muted,
            currentTime: this.currentTime,
            duration: this.duration,
            onTogglePlay: ()=>void this.toggleMedia('viewer'),
            onToggleMute: ()=>this.toggleMuted('viewer'),
            onSeek: (event)=>this.seekMedia(event, 'viewer')
        }) : html``;
        return html`
      <dialog
        class="ve-resource-preview-viewer"
        part="viewer"
        aria-label=${item.name || 'Resource preview'}
        @close=${this.handleDialogClose}
        @cancel=${this.handleDialogCancel}
      >
        ${this.previewVisible ? html`
              <div class="ve-resource-preview-viewer-body">
                <div class="ve-resource-preview-viewer-content">
                  ${renderViewerContent({
            item,
            scaleStyle: this.viewerScaleStyle,
            onMediaEvent: this.handleMediaEvent,
            mediaControls
        })}
                  <span class="ve-resource-preview-viewer-close">
                    <ve-button
                      type="text"
                      size="large"
                      shape="circle"
                      aria-label="Close"
                      @click=${()=>this.closePreview()}
                    >
                      ${renderIcon(closeIcon)}
                    </ve-button>
                  </span>
                </div>
              </div>
              ${showToolbar ? html`
                    <div class="ve-resource-preview-viewer-toolbar">
                      ${renderZoomControls(this.viewerScale, this.handleZoom)}
                    </div>
                  ` : nothing}
            ` : nothing}
      </dialog>
    `;
    }
    render() {
        const surfaceStyle = this.mediaAspectRatio ? `--ve-resource-preview-current-aspect-ratio:${this.mediaAspectRatio};` : '';
        const interactive = !this.previewDisabled && this.hasSource;
        const item = this.getResourceItem();
        return html`
      <div
        class=${classMap({
            've-resource-preview': true,
            [`ve-resource-preview-${this.normalizedType}`]: true,
            've-resource-preview-no-source': !this.hasSource
        })}
        part="root"
      >
        <div
          class="ve-resource-preview-surface"
          part="surface"
          role=${interactive ? 'button' : 'img'}
          tabindex=${interactive ? '0' : '-1'}
          aria-label=${this.name || 'Resource preview'}
          style=${surfaceStyle}
          @click=${this.handleSurfaceClick}
          @keydown=${this.handleKeyboardPreview}
        >
          ${this.renderCardMedia()}
          ${this.renderTip()}
        </div>
        ${this.renderDownloadButton(item)}
      </div>
      ${this.renderViewer()}
    `;
    }
}
const groupStyles = css`
  :host {
    display: inline-flex;
    inline-size: auto;
    max-inline-size: 100%;
    vertical-align: top;
    font-family: var(--font-sans);
  }

  :host([hidden]) {
    display: none;
  }

  .ve-resource-preview-group {
    display: inline-flex;
    align-items: flex-start;
    gap: var(--ve-resource-preview-group-gap);
    max-inline-size: 100%;
    box-sizing: border-box;
    flex-wrap: wrap;
  }

  :host([no-wrap]) .ve-resource-preview-group {
    flex-wrap: nowrap;
    overflow-x: auto;
    overflow-y: hidden;
  }

  .ve-resource-preview-group ::slotted(ve-resource-preview) {
    flex: none;
  }
`;
const ve_resource_preview_group = [
    componentReset,
    token,
    groupStyles,
    viewer
];
function ve_resource_preview_group_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_resource_preview_group_element_toPropertyKey(arg) {
    var key = ve_resource_preview_group_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_resource_preview_group_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_resource_preview_group_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_resource_preview_group_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_resource_preview_group_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_resource_preview_group_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_resource_preview_group_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_resource_preview_group_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_resource_preview_group_element_dec, ve_resource_preview_group_element_dec1, ve_resource_preview_group_element_dec2, ve_resource_preview_group_element_dec3, ve_resource_preview_group_element_dec4, ve_resource_preview_group_element_dec5, ve_resource_preview_group_element_dec6, ve_resource_preview_group_element_dec7, ve_resource_preview_group_element_dec8, ve_resource_preview_group_element_dec9, ve_resource_preview_group_element_dec10, ve_resource_preview_group_element_dec11, ve_resource_preview_group_element_dec12, _init_gap, _init_extra__init_gap, _init_noWrap, _init_extra__init_noWrap, ve_resource_preview_group_element_init_viewerDialog, ve_resource_preview_group_element_init_extra_init_viewerDialog, ve_resource_preview_group_element_init_viewerMediaElement, ve_resource_preview_group_element_init_extra_init_viewerMediaElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_entries, _init_extra__init_entries, ve_resource_preview_group_element_init_previewVisible, ve_resource_preview_group_element_init_extra_init_previewVisible, _init_currentIndex, _init_extra__init_currentIndex, ve_resource_preview_group_element_init_playing, ve_resource_preview_group_element_init_extra_init_playing, ve_resource_preview_group_element_init_muted, ve_resource_preview_group_element_init_extra_init_muted, ve_resource_preview_group_element_init_currentTime, ve_resource_preview_group_element_init_extra_init_currentTime, ve_resource_preview_group_element_init_duration, ve_resource_preview_group_element_init_extra_init_duration, ve_resource_preview_group_element_init_viewerScale, ve_resource_preview_group_element_init_extra_init_viewerScale;
const VE_RESOURCE_PREVIEW_GROUP_TAG = 've-resource-preview-group';
ve_resource_preview_group_element_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), ve_resource_preview_group_element_dec1 = property({
    type: Boolean,
    attribute: 'no-wrap',
    reflect: true
}), ve_resource_preview_group_element_dec2 = query('dialog.ve-resource-preview-viewer'), ve_resource_preview_group_element_dec3 = query('.ve-resource-preview-viewer-media'), ve_resource_preview_group_element_dec4 = query('slot'), ve_resource_preview_group_element_dec5 = state(), ve_resource_preview_group_element_dec6 = state(), ve_resource_preview_group_element_dec7 = state(), ve_resource_preview_group_element_dec8 = state(), ve_resource_preview_group_element_dec9 = state(), ve_resource_preview_group_element_dec10 = state(), ve_resource_preview_group_element_dec11 = state(), ve_resource_preview_group_element_dec12 = state();
class VeResourcePreviewGroupElement extends VeElement {
    static{
        ({ e: [_init_gap, _init_extra__init_gap, _init_noWrap, _init_extra__init_noWrap, ve_resource_preview_group_element_init_viewerDialog, ve_resource_preview_group_element_init_extra_init_viewerDialog, ve_resource_preview_group_element_init_viewerMediaElement, ve_resource_preview_group_element_init_extra_init_viewerMediaElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_entries, _init_extra__init_entries, ve_resource_preview_group_element_init_previewVisible, ve_resource_preview_group_element_init_extra_init_previewVisible, _init_currentIndex, _init_extra__init_currentIndex, ve_resource_preview_group_element_init_playing, ve_resource_preview_group_element_init_extra_init_playing, ve_resource_preview_group_element_init_muted, ve_resource_preview_group_element_init_extra_init_muted, ve_resource_preview_group_element_init_currentTime, ve_resource_preview_group_element_init_extra_init_currentTime, ve_resource_preview_group_element_init_duration, ve_resource_preview_group_element_init_extra_init_duration, ve_resource_preview_group_element_init_viewerScale, ve_resource_preview_group_element_init_extra_init_viewerScale] } = ve_resource_preview_group_element_apply_decs_2311(this, [], [
            [
                ve_resource_preview_group_element_dec,
                1,
                "gap"
            ],
            [
                ve_resource_preview_group_element_dec1,
                1,
                "noWrap"
            ],
            [
                ve_resource_preview_group_element_dec2,
                1,
                "viewerDialog"
            ],
            [
                ve_resource_preview_group_element_dec3,
                1,
                "viewerMediaElement"
            ],
            [
                ve_resource_preview_group_element_dec4,
                1,
                "defaultSlot"
            ],
            [
                ve_resource_preview_group_element_dec5,
                1,
                "entries"
            ],
            [
                ve_resource_preview_group_element_dec6,
                1,
                "previewVisible"
            ],
            [
                ve_resource_preview_group_element_dec7,
                1,
                "currentIndex"
            ],
            [
                ve_resource_preview_group_element_dec8,
                1,
                "playing"
            ],
            [
                ve_resource_preview_group_element_dec9,
                1,
                "muted"
            ],
            [
                ve_resource_preview_group_element_dec10,
                1,
                "currentTime"
            ],
            [
                ve_resource_preview_group_element_dec11,
                1,
                "duration"
            ],
            [
                ve_resource_preview_group_element_dec12,
                1,
                "viewerScale"
            ]
        ]));
    }
    static styles = ve_resource_preview_group;
    static dependencies = {
        [VE_BUTTON_TAG]: VeButtonElement,
        [VE_ICON_TAG]: VeIconElement,
        [VE_RESOURCE_PREVIEW_TAG]: VeResourcePreviewElement
    };
    #___private_gap_1 = (()=>{
        const _value = _init_gap(this, 'medium');
        _init_extra__init_gap(this);
        return _value;
    })();
    get gap() {
        return this.#___private_gap_1;
    }
    set gap(_v) {
        this.#___private_gap_1 = _v;
    }
    #___private_noWrap_2 = (()=>{
        const _value = _init_noWrap(this, false);
        _init_extra__init_noWrap(this);
        return _value;
    })();
    get noWrap() {
        return this.#___private_noWrap_2;
    }
    set noWrap(_v) {
        this.#___private_noWrap_2 = _v;
    }
    #___private_viewerDialog_3 = (()=>{
        const _value = ve_resource_preview_group_element_init_viewerDialog(this, null);
        ve_resource_preview_group_element_init_extra_init_viewerDialog(this);
        return _value;
    })();
    get viewerDialog() {
        return this.#___private_viewerDialog_3;
    }
    set viewerDialog(_v) {
        this.#___private_viewerDialog_3 = _v;
    }
    #___private_viewerMediaElement_4 = (()=>{
        const _value = ve_resource_preview_group_element_init_viewerMediaElement(this, null);
        ve_resource_preview_group_element_init_extra_init_viewerMediaElement(this);
        return _value;
    })();
    get viewerMediaElement() {
        return this.#___private_viewerMediaElement_4;
    }
    set viewerMediaElement(_v) {
        this.#___private_viewerMediaElement_4 = _v;
    }
    #___private_defaultSlot_5 = (()=>{
        const _value = _init_defaultSlot(this, null);
        _init_extra__init_defaultSlot(this);
        return _value;
    })();
    get defaultSlot() {
        return this.#___private_defaultSlot_5;
    }
    set defaultSlot(_v) {
        this.#___private_defaultSlot_5 = _v;
    }
    #___private_entries_6 = (()=>{
        const _value = _init_entries(this, []);
        _init_extra__init_entries(this);
        return _value;
    })();
    get entries() {
        return this.#___private_entries_6;
    }
    set entries(_v) {
        this.#___private_entries_6 = _v;
    }
    #___private_previewVisible_7 = (()=>{
        const _value = ve_resource_preview_group_element_init_previewVisible(this, false);
        ve_resource_preview_group_element_init_extra_init_previewVisible(this);
        return _value;
    })();
    get previewVisible() {
        return this.#___private_previewVisible_7;
    }
    set previewVisible(_v) {
        this.#___private_previewVisible_7 = _v;
    }
    #___private_currentIndex_8 = (()=>{
        const _value = _init_currentIndex(this, 0);
        _init_extra__init_currentIndex(this);
        return _value;
    })();
    get currentIndex() {
        return this.#___private_currentIndex_8;
    }
    set currentIndex(_v) {
        this.#___private_currentIndex_8 = _v;
    }
    #___private_playing_9 = (()=>{
        const _value = ve_resource_preview_group_element_init_playing(this, false);
        ve_resource_preview_group_element_init_extra_init_playing(this);
        return _value;
    })();
    get playing() {
        return this.#___private_playing_9;
    }
    set playing(_v) {
        this.#___private_playing_9 = _v;
    }
    #___private_muted_10 = (()=>{
        const _value = ve_resource_preview_group_element_init_muted(this, false);
        ve_resource_preview_group_element_init_extra_init_muted(this);
        return _value;
    })();
    get muted() {
        return this.#___private_muted_10;
    }
    set muted(_v) {
        this.#___private_muted_10 = _v;
    }
    #___private_currentTime_11 = (()=>{
        const _value = ve_resource_preview_group_element_init_currentTime(this, 0);
        ve_resource_preview_group_element_init_extra_init_currentTime(this);
        return _value;
    })();
    get currentTime() {
        return this.#___private_currentTime_11;
    }
    set currentTime(_v) {
        this.#___private_currentTime_11 = _v;
    }
    #___private_duration_12 = (()=>{
        const _value = ve_resource_preview_group_element_init_duration(this, 0);
        ve_resource_preview_group_element_init_extra_init_duration(this);
        return _value;
    })();
    get duration() {
        return this.#___private_duration_12;
    }
    set duration(_v) {
        this.#___private_duration_12 = _v;
    }
    #___private_viewerScale_13 = (()=>{
        const _value = ve_resource_preview_group_element_init_viewerScale(this, VIEWER_SCALE_DEFAULT);
        ve_resource_preview_group_element_init_extra_init_viewerScale(this);
        return _value;
    })();
    get viewerScale() {
        return this.#___private_viewerScale_13;
    }
    set viewerScale(_v) {
        this.#___private_viewerScale_13 = _v;
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('gap')) {
            const normalized = normalizeEnum(this.gap, RESOURCE_PREVIEW_GAP_VALUES, 'medium');
            if (normalized !== this.gap) this.gap = normalized;
        }
        if (changedProperties.has('previewVisible')) {
            if (this.previewVisible && this.viewerDialog && !this.viewerDialog.open) this.viewerDialog.showModal();
            else if (!this.previewVisible && this.viewerDialog?.open) this.viewerDialog.close();
        }
    }
    disconnectedCallback() {
        if (this.previewVisible) unlockBodyScroll();
        this.viewerMediaElement?.pause();
        super.disconnectedCallback();
    }
    get currentEntry() {
        return this.entries[this.currentIndex] ?? this.entries.at(-1);
    }
    refreshEntries() {
        const assigned = this.defaultSlot?.assignedElements({
            flatten: true
        }) ?? [];
        this.entries = assigned.filter((element)=>element.localName === VE_RESOURCE_PREVIEW_TAG).map((child)=>({
                child,
                item: child.getResourceItem()
            })).filter(({ item })=>(item.src ?? '').trim().length > 0);
    }
    handleSlotChange = ()=>{
        this.refreshEntries();
        if (this.previewVisible && 0 === this.entries.length) this.closeSharedPreview();
    };
    handleChildPreview = (event)=>{
        if (event.defaultPrevented) return;
        const source = event.detail.source;
        if (!(source instanceof VeResourcePreviewElement)) return;
        this.refreshEntries();
        const index = this.entries.findIndex((entry)=>entry.child === source);
        if (index < 0) return;
        event.preventDefault();
        this.openSharedPreview(index);
    };
    handleChildDownload = (event)=>{};
    handleChildClose = (event)=>{
        event.stopPropagation();
    };
    openSharedPreview(index) {
        if (0 === this.entries.length) return;
        this.viewerMediaElement?.pause();
        this.currentIndex = clampIndex(index, this.entries.length);
        this.resetViewerMediaState();
        this.previewVisible = true;
        lockBodyScroll();
        this.emitPreviewChange();
    }
    closeSharedPreview = ()=>{
        if (!this.previewVisible) return;
        const entry = this.currentEntry;
        this.viewerMediaElement?.pause();
        this.previewVisible = false;
        this.currentIndex = 0;
        this.resetViewerMediaState();
        unlockBodyScroll();
        this.dispatchEvent(new CustomEvent('ve-preview-group-close', {
            detail: {
                item: entry?.item ?? null,
                source: entry?.child ?? null
            },
            bubbles: true,
            composed: true
        }));
    };
    showByOffset(offset) {
        if (0 === this.entries.length) return void this.closeSharedPreview();
        this.viewerMediaElement?.pause();
        this.currentIndex = (this.currentIndex + offset + this.entries.length) % this.entries.length;
        this.resetViewerMediaState();
        this.emitPreviewChange();
    }
    resetViewerMediaState() {
        this.viewerScale = VIEWER_SCALE_DEFAULT;
        this.currentTime = 0;
        this.duration = 0;
        this.playing = false;
    }
    emitPreviewChange() {
        const entry = this.currentEntry;
        const total = this.entries.length;
        if (!entry || 0 === total) return;
        this.dispatchEvent(new CustomEvent('ve-preview-change', {
            detail: {
                item: entry.item,
                index: this.currentIndex,
                total,
                source: entry.child
            },
            bubbles: true,
            composed: true
        }));
    }
    syncMediaState(media) {
        this.currentTime = media.currentTime;
        this.duration = Number.isFinite(media.duration) ? media.duration : 0;
        this.playing = !media.paused && !media.ended;
        this.muted = media.muted;
    }
    handleMediaEvent = (event)=>{
        const media = event.currentTarget;
        if (media instanceof HTMLMediaElement) this.syncMediaState(media);
    };
    handleDialogClose = ()=>{
        if (this.previewVisible) this.closeSharedPreview();
    };
    handleDialogCancel = (event)=>{
        event.preventDefault();
        this.closeSharedPreview();
    };
    handleViewerKeydown = (event)=>{
        if (event.target instanceof HTMLInputElement || event.target instanceof HTMLTextAreaElement) return;
        if ('ArrowLeft' === event.key) {
            event.preventDefault();
            this.showByOffset(-1);
        } else if ('ArrowRight' === event.key) {
            event.preventDefault();
            this.showByOffset(1);
        }
    };
    handleZoom = (delta)=>{
        this.viewerScale = nextScale(this.viewerScale, delta);
    };
    get viewerScaleStyle() {
        return `--ve-resource-preview-viewer-scale:${clampScale(this.viewerScale)};`;
    }
    async toggleViewerMedia() {
        const media = this.viewerMediaElement;
        if (!media) return;
        if (media.paused || media.ended) {
            try {
                await media.play();
            } catch  {
                this.playing = false;
            }
            return;
        }
        media.pause();
    }
    toggleMuted = ()=>{
        const media = this.viewerMediaElement;
        if (!media) return;
        media.muted = !media.muted;
        this.muted = media.muted;
    };
    seekViewerMedia = (event)=>{
        const input = event.currentTarget instanceof HTMLInputElement ? event.currentTarget : null;
        const media = this.viewerMediaElement;
        if (!input || !media || !Number.isFinite(media.duration) || media.duration <= 0) return;
        const progress = Number(input.value);
        if (!Number.isFinite(progress)) return;
        media.currentTime = progress / 1000 * media.duration;
        this.syncMediaState(media);
    };
    renderViewer() {
        const entry = this.previewVisible ? this.currentEntry : void 0;
        const total = this.entries.length;
        const item = entry?.item ?? null;
        const variant = item?.type === 'video' ? 'video' : null;
        const mediaControls = variant ? renderMediaControls({
            scope: 'viewer',
            variant,
            playing: this.playing,
            muted: this.muted,
            currentTime: this.currentTime,
            duration: this.duration,
            onTogglePlay: ()=>void this.toggleViewerMedia(),
            onToggleMute: this.toggleMuted,
            onSeek: this.seekViewerMedia
        }) : html``;
        const showToolbar = Boolean(item);
        return html`
      <dialog
        class="ve-resource-preview-viewer"
        part="viewer"
        aria-label=${item?.name || 'Resource preview'}
        @close=${this.handleDialogClose}
        @cancel=${this.handleDialogCancel}
        @keydown=${this.handleViewerKeydown}
      >
        ${item ? html`
              <div class="ve-resource-preview-viewer-body">
                <div class="ve-resource-preview-viewer-content">
                  ${renderViewerContent({
            item,
            scaleStyle: this.viewerScaleStyle,
            onMediaEvent: this.handleMediaEvent,
            mediaControls
        })}
                  <span class="ve-resource-preview-viewer-close">
                    <ve-button
                      type="text"
                      size="large"
                      shape="circle"
                      aria-label="Close"
                      @click=${this.closeSharedPreview}
                    >
                      ${renderIcon(closeIcon)}
                    </ve-button>
                  </span>
                </div>
              </div>
              ${showToolbar ? html`
                    <div class="ve-resource-preview-viewer-toolbar">
                      ${total > 1 ? html`
                            <ve-button
                              type="text"
                              size="large"
                              shape="circle"
                              aria-label="Previous"
                              @click=${()=>this.showByOffset(-1)}
                            >
                              ${renderIcon(chevronLeftMdIcon)}
                            </ve-button>
                            <span class="ve-resource-preview-viewer-count">
                              ${this.currentIndex + 1} / ${total}
                            </span>
                            <ve-button
                              type="text"
                              size="large"
                              shape="circle"
                              aria-label="Next"
                              @click=${()=>this.showByOffset(1)}
                            >
                              ${renderIcon(chevronRightMdIcon)}
                            </ve-button>
                          ` : nothing}
                      ${renderZoomControls(this.viewerScale, this.handleZoom)}
                    </div>
                  ` : nothing}
            ` : nothing}
      </dialog>
    `;
    }
    render() {
        return html`
      <div
        class="ve-resource-preview-group"
        part="group"
        @ve-preview=${this.handleChildPreview}
        @ve-preview-close=${this.handleChildClose}
        @ve-download=${this.handleChildDownload}
      >
        <slot @slotchange=${this.handleSlotChange}></slot>
      </div>
      ${this.renderViewer()}
    `;
    }
}
function clampIndex(index, length) {
    if (length <= 0) return 0;
    return Math.min(Math.max(index, 0), length - 1);
}
defineCustomElement(VE_RESOURCE_PREVIEW_TAG, VeResourcePreviewElement);
defineCustomElement(VE_RESOURCE_PREVIEW_GROUP_TAG, VeResourcePreviewGroupElement);
export { VE_RESOURCE_PREVIEW_GROUP_TAG, VE_RESOURCE_PREVIEW_TAG, VeResourcePreviewElement, VeResourcePreviewGroupElement };
