// @ve-design/web/web/dist/ve-radio.js
import { css, html } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, queryAssignedElements, state } from "lit/decorators.js";
import { repeat } from "lit/directives/repeat.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
const veRadioTokenDefaults = css`
  :host {
    --ve-radio-font-family: var(--font-sans);
    --ve-radio-font-size: var(--text-body);
    --ve-radio-font-weight: var(--font-weight-normal);

    --ve-radio-mask-size: 16px;
    --ve-radio-mask-border-width: var(--stroke-weight-strong);
    --ve-radio-mask-border-style: solid;
    --ve-radio-mask-border-radius: var(--radius-full);

    --ve-radio-mask-color-bg: var(--color-bg-base);
    --ve-radio-mask-color-bg-checked: var(--color-bg-primary);
    --ve-radio-mask-color-bg-checked-hover: color-mix(
      var(--ve-radio-mask-color-bg-checked) var(--state-hover-solid-ratio),
      var(--state-hover-solid-mix)
    );
    --ve-radio-mask-color-border: var(--color-border-strong);
    --ve-radio-mask-color-border-hover: var(--color-border-interactive);
    --ve-radio-mask-color-border-checked: transparent;

    --ve-radio-dot-size: 6px;
    --ve-radio-dot-color: var(--color-text-foreground);

    --ve-radio-text-color: var(--color-text-primary);
    --ve-radio-text-mask-spacing: var(--space-xxs);

    --ve-radio-group-gap-horizontal: var(--space-s);
    --ve-radio-group-gap-vertical: var(--space-xs);
    --ve-radio-group-line-height-vertical: 32px;

    --ve-radio-disabled-subtle-opacity: var(--state-disabled-ratio);
    --ve-radio-disabled-solid-opacity: var(--state-disabled-ratio);
    --ve-radio-focus-ring-color: color-mix(
      var(--color-bg-primary) 24%,
      transparent
    );

    --ve-radio-transition-duration-fast: var(--motion-fast);
    --ve-radio-transition-duration-base: var(--motion-base);
    --ve-radio-transition-timing-linear: linear;
    --ve-radio-transition-timing-overshoot: cubic-bezier(0.34, 1.56, 0.64, 1);
  }
`;
const token = veRadioTokenDefaults;
const ve_radio = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      color: var(--ve-radio-text-color);
      font-family: var(--ve-radio-font-family);
      font-size: var(--ve-radio-font-size);
      font-weight: var(--ve-radio-font-weight);
      line-height: var(--line-height-body);
      vertical-align: middle;
    }

    .ve-radio {
      position: relative;
      box-sizing: border-box;
      display: inline-flex;
      align-items: center;
      gap: var(--ve-radio-text-mask-spacing);
      color: var(--ve-radio-text-color);
      cursor: pointer;
      font: inherit;
      user-select: none;
      transition:
        color var(--ve-radio-transition-duration-fast)
          var(--ve-radio-transition-timing-linear),
        background-color var(--ve-radio-transition-duration-fast)
          var(--ve-radio-transition-timing-linear),
        border-color var(--ve-radio-transition-duration-fast)
          var(--ve-radio-transition-timing-linear),
        opacity var(--ve-radio-transition-duration-fast)
          var(--ve-radio-transition-timing-linear);
    }

    .ve-radio--disabled {
      cursor: not-allowed;
      opacity: var(--ve-radio-disabled-subtle-opacity);
    }

    .ve-radio-input {
      position: absolute;
      inline-size: 1px;
      block-size: 1px;
      margin: -1px;
      padding: 0;
      overflow: hidden;
      border: 0;
      clip: rect(0 0 0 0);
      clip-path: inset(50%);
      white-space: nowrap;
    }

    .ve-radio-indicator {
      position: relative;
      box-sizing: border-box;
      display: inline-flex;
      flex: 0 0 auto;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-radio-mask-size);
      block-size: var(--ve-radio-mask-size);
      border: var(--ve-radio-mask-border-width)
        var(--ve-radio-mask-border-style) var(--ve-radio-mask-color-border);
      border-radius: var(--ve-radio-mask-border-radius);
      background: var(--ve-radio-mask-color-bg);
      transition:
        background-color var(--ve-radio-transition-duration-fast)
          var(--ve-radio-transition-timing-linear),
        border-color var(--ve-radio-transition-duration-fast)
          var(--ve-radio-transition-timing-linear),
        box-shadow var(--ve-radio-transition-duration-fast)
          var(--ve-radio-transition-timing-linear);
    }

    .ve-radio:not(.ve-radio--disabled):not(.ve-radio--checked):hover
      .ve-radio-indicator {
      border-color: var(--ve-radio-mask-color-border-hover);
    }

    .ve-radio--checked .ve-radio-indicator {
      border-color: var(--ve-radio-mask-color-border-checked);
      background: var(--ve-radio-mask-color-bg-checked);
    }

    .ve-radio--checked:not(.ve-radio--disabled):hover .ve-radio-indicator {
      background: var(--ve-radio-mask-color-bg-checked-hover);
    }

    .ve-radio-input:focus-visible + .ve-radio-indicator {
      box-shadow: 0 0 0 3px var(--ve-radio-focus-ring-color);
    }

    .ve-radio-dot {
      position: relative;
      inline-size: var(--ve-radio-dot-size);
      block-size: var(--ve-radio-dot-size);
      border-radius: var(--radius-full);
      background: var(--ve-radio-dot-color);
      transform: scale(0);
      transition: transform var(--ve-radio-transition-duration-fast)
        var(--ve-radio-transition-timing-overshoot);
    }

    .ve-radio--checked .ve-radio-dot {
      transform: scale(1);
    }

    .ve-radio-label {
      display: inline-flex;
      align-items: center;
      min-inline-size: 0;
      line-height: 1.4;
      white-space: nowrap;
    }

    .ve-radio-input:focus-visible + .ve-radio-label,
    .ve-radio-input:focus-visible ~ .ve-radio-label {
      outline: none;
    }
  `
];
const ve_radio_group = [
    componentReset,
    token,
    css`
    :host {
      display: inline-block;
      font-family: var(--ve-radio-font-family);
      font-size: var(--ve-radio-font-size);
      line-height: var(--line-height-body);
      vertical-align: middle;
    }

    .ve-radio-group {
      display: inline-flex;
      align-items: center;
      color: var(--ve-radio-text-color);
    }

    .ve-radio-group-direction-horizontal {
      flex-direction: row;
      gap: var(--ve-radio-group-gap-horizontal);
    }

    .ve-radio-group-direction-vertical {
      flex-direction: column;
      align-items: flex-start;
      gap: var(--ve-radio-group-gap-vertical);
    }

    .ve-radio-group-direction-vertical ::slotted(ve-radio) {
      min-block-size: var(--ve-radio-group-line-height-vertical);
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _init_checked, _init_extra__init_checked, _init_defaultChecked, _init_extra__init_defaultChecked, _init_disabled, _init_extra__init_disabled, _init_value, _init_extra__init_value, _init_groupDisabled, _init_extra__init_groupDisabled, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _init_value1, _init_extra__init_value1, _init_defaultValue, _init_extra__init_defaultValue, _init_disabled1, _init_extra__init_disabled1, _init_direction, _init_extra__init_direction, _init_options, _init_extra__init_options, _init_assignedRadioElements, _init_extra__init_assignedRadioElements;
const VE_RADIO_TAG = 've-radio';
const VE_RADIO_GROUP_TAG = 've-radio-group';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby'
];
const RADIO_VALUE_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        return value ?? '';
    },
    toAttribute (value) {
        return String(value);
    }
};
const RADIO_GROUP_OPTIONS_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        if (null === value || 0 === value.trim().length) return [];
        try {
            const parsed = JSON.parse(value);
            return Array.isArray(parsed) ? parsed : [];
        } catch  {
            return [];
        }
    },
    toAttribute (value) {
        try {
            return JSON.stringify(value);
        } catch  {
            return null;
        }
    }
};
function isRadioValue(value) {
    return 'string' == typeof value || 'number' == typeof value && Number.isFinite(value);
}
function isRenderableTemplateValue(value) {
    return null != value && 'boolean' != typeof value && 'function' != typeof value && 'symbol' != typeof value;
}
function createObservedAttributes(baseObservedAttributes) {
    return [
        ...new Set([
            ...baseObservedAttributes,
            ...CONTROL_ACCESSIBILITY_ATTRIBUTES
        ])
    ];
}
function getForwardedAriaAttribute(element, attributeName) {
    return element.getAttribute(attributeName) ?? void 0;
}
_dec = property({
    type: Boolean,
    reflect: true
}), _dec1 = property({
    type: Boolean,
    attribute: 'default-checked'
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    converter: RADIO_VALUE_ATTRIBUTE_CONVERTER
}), _dec4 = state();
class VeRadioElement extends VeElement {
    static{
        ({ e: [_init_checked, _init_extra__init_checked, _init_defaultChecked, _init_extra__init_defaultChecked, _init_disabled, _init_extra__init_disabled, _init_value, _init_extra__init_value, _init_groupDisabled, _init_extra__init_groupDisabled] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "checked"
            ],
            [
                _dec1,
                1,
                "defaultChecked"
            ],
            [
                _dec2,
                1,
                "disabled"
            ],
            [
                _dec3,
                1,
                "value"
            ],
            [
                _dec4,
                1,
                "groupDisabled"
            ]
        ]));
    }
    static styles = ve_radio;
    static get observedAttributes() {
        return createObservedAttributes(super.observedAttributes);
    }
    #___private_checked_1 = (()=>{
        const _value = _init_checked(this, void 0);
        _init_extra__init_checked(this);
        return _value;
    })();
    get checked() {
        return this.#___private_checked_1;
    }
    set checked(_v) {
        this.#___private_checked_1 = _v;
    }
    #___private_defaultChecked_2 = (()=>{
        const _value = _init_defaultChecked(this, false);
        _init_extra__init_defaultChecked(this);
        return _value;
    })();
    get defaultChecked() {
        return this.#___private_defaultChecked_2;
    }
    set defaultChecked(_v) {
        this.#___private_defaultChecked_2 = _v;
    }
    #___private_disabled_3 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_3;
    }
    set disabled(_v) {
        this.#___private_disabled_3 = _v;
    }
    #___private_value_4 = (()=>{
        const _value = _init_value(this, '');
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_4;
    }
    set value(_v) {
        this.#___private_value_4 = _v;
    }
    #___private_groupDisabled_5 = (()=>{
        const _value = _init_groupDisabled(this, false);
        _init_extra__init_groupDisabled(this);
        return _value;
    })();
    get groupDisabled() {
        return this.#___private_groupDisabled_5;
    }
    set groupDisabled(_v) {
        this.#___private_groupDisabled_5 = _v;
    }
    hasInitializedDefaultChecked = false;
    get mergedDisabled() {
        return this.disabled || this.groupDisabled;
    }
    get inputValue() {
        return String(this.value);
    }
    connectedCallback() {
        super.connectedCallback();
        this.applyDefaultChecked();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    applyDefaultChecked() {
        if (this.hasInitializedDefaultChecked) return;
        if (void 0 === this.checked && this.defaultChecked) this.checked = true;
        this.hasInitializedDefaultChecked = true;
    }
    emitVeChangeEvent() {
        const detail = {
            checked: true === this.checked
        };
        if (isRadioValue(this.value) && String(this.value).length > 0) detail.value = this.value;
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    handleInputClick(event) {
        event.stopPropagation();
    }
    handleInputChange(event) {
        const target = event.target;
        if (!(target instanceof HTMLInputElement)) return;
        event.stopPropagation();
        if (this.mergedDisabled) {
            target.checked = true === this.checked;
            return;
        }
        this.checked = target.checked;
        this.emitVeChangeEvent();
    }
    render() {
        const wrapperClass = classMap({
            've-radio': true,
            've-radio--checked': true === this.checked,
            've-radio--disabled': this.mergedDisabled
        });
        return html`
      <label class=${wrapperClass} part="wrapper">
        <input
          class="ve-radio-input"
          part="input"
          type="radio"
          .checked=${true === this.checked}
          ?disabled=${this.mergedDisabled}
          .value=${this.inputValue}
          aria-label=${ifDefined(getForwardedAriaAttribute(this, 'aria-label'))}
          aria-labelledby=${ifDefined(getForwardedAriaAttribute(this, 'aria-labelledby'))}
          aria-describedby=${ifDefined(getForwardedAriaAttribute(this, 'aria-describedby'))}
          @click=${this.handleInputClick}
          @change=${this.handleInputChange}
        />
        <span class="ve-radio-indicator" part="indicator" aria-hidden="true">
          <span class="ve-radio-dot" part="dot"></span>
        </span>
        <span class="ve-radio-label" part="label">
          <slot></slot>
        </span>
      </label>
    `;
    }
}
_dec5 = property({
    converter: RADIO_VALUE_ATTRIBUTE_CONVERTER,
    reflect: true
}), _dec6 = property({
    attribute: 'default-value',
    converter: RADIO_VALUE_ATTRIBUTE_CONVERTER
}), _dec7 = property({
    type: Boolean,
    reflect: true
}), _dec8 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec9 = property({
    converter: RADIO_GROUP_OPTIONS_ATTRIBUTE_CONVERTER
}), _dec10 = queryAssignedElements({
    selector: VE_RADIO_TAG,
    flatten: true
});
class VeRadioGroupElement extends VeElement {
    static{
        ({ e: [_init_value1, _init_extra__init_value1, _init_defaultValue, _init_extra__init_defaultValue, _init_disabled1, _init_extra__init_disabled1, _init_direction, _init_extra__init_direction, _init_options, _init_extra__init_options, _init_assignedRadioElements, _init_extra__init_assignedRadioElements] } = _apply_decs_2311(this, [], [
            [
                _dec5,
                1,
                "value"
            ],
            [
                _dec6,
                1,
                "defaultValue"
            ],
            [
                _dec7,
                1,
                "disabled"
            ],
            [
                _dec8,
                1,
                "direction"
            ],
            [
                _dec9,
                1,
                "options"
            ],
            [
                _dec10,
                1,
                "assignedRadioElements"
            ]
        ]));
    }
    static styles = ve_radio_group;
    static dependencies = {
        [VE_RADIO_TAG]: VeRadioElement
    };
    static get observedAttributes() {
        return createObservedAttributes(super.observedAttributes);
    }
    #___private_value_6 = (()=>{
        const _value = _init_value1(this, '');
        _init_extra__init_value1(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_6;
    }
    set value(_v) {
        this.#___private_value_6 = _v;
    }
    #___private_defaultValue_7 = (()=>{
        const _value = _init_defaultValue(this, '');
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_7;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_7 = _v;
    }
    #___private_disabled_8 = (()=>{
        const _value = _init_disabled1(this, false);
        _init_extra__init_disabled1(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_8;
    }
    set disabled(_v) {
        this.#___private_disabled_8 = _v;
    }
    #___private_direction_9 = (()=>{
        const _value = _init_direction(this, 'horizontal');
        _init_extra__init_direction(this);
        return _value;
    })();
    get direction() {
        return this.#___private_direction_9;
    }
    set direction(_v) {
        this.#___private_direction_9 = _v;
    }
    #___private_options_10 = (()=>{
        const _value = _init_options(this, []);
        _init_extra__init_options(this);
        return _value;
    })();
    get options() {
        return this.#___private_options_10;
    }
    set options(_v) {
        this.#___private_options_10 = _v;
    }
    #___private_assignedRadioElements_11 = (()=>{
        const _value = _init_assignedRadioElements(this);
        _init_extra__init_assignedRadioElements(this);
        return _value;
    })();
    get assignedRadioElements() {
        return this.#___private_assignedRadioElements_11;
    }
    set assignedRadioElements(_v) {
        this.#___private_assignedRadioElements_11 = _v;
    }
    normalizedOptions = [];
    hasInitializedDefaultValue = false;
    handleGroupChildChange = (event)=>{
        if (!(event instanceof CustomEvent)) return;
        const target = this.resolveEventRadioTarget(event);
        if (!target) return;
        if (!this.isManagedRadio(target)) return;
        event.stopPropagation();
        const nextValue = target.value;
        if (!isRadioValue(nextValue)) return;
        if (this.value !== nextValue) {
            this.value = nextValue;
            this.syncRadioStates();
            this.dispatchEvent(new CustomEvent('ve-group-change', {
                detail: {
                    value: this.value
                },
                bubbles: true,
                composed: true
            }));
        } else this.syncRadioStates();
    };
    get hasOptions() {
        return this.normalizedOptions.length > 0;
    }
    connectedCallback() {
        super.connectedCallback();
        this.normalizedOptions = this.normalizeOptions(this.options);
        this.applyDefaultValue();
        this.addEventListener('ve-change', this.handleGroupChildChange);
    }
    disconnectedCallback() {
        this.removeEventListener('ve-change', this.handleGroupChildChange);
        super.disconnectedCallback();
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.syncRadioStates();
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('options')) this.normalizedOptions = this.normalizeOptions(this.options);
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('value') || changedProperties.has('disabled') || changedProperties.has('options')) this.syncRadioStates();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    applyDefaultValue() {
        if (this.hasInitializedDefaultValue) return;
        if (0 === String(this.value).length && String(this.defaultValue).length > 0) this.value = this.defaultValue;
        this.hasInitializedDefaultValue = true;
    }
    normalizeOptions(options) {
        if (!Array.isArray(options)) return [];
        const result = [];
        const seen = new Set();
        for (const option of options){
            const normalizedOption = this.normalizeOption(option);
            if (normalizedOption) {
                if (!seen.has(normalizedOption.value)) {
                    seen.add(normalizedOption.value);
                    result.push(normalizedOption);
                }
            }
        }
        return result;
    }
    normalizeOption(option) {
        if (isRadioValue(option)) return {
            label: option,
            value: option,
            disabled: false
        };
        if (!option || 'object' != typeof option) return null;
        const optionLike = option;
        if (!isRadioValue(optionLike.value)) return null;
        return {
            label: optionLike.label ?? optionLike.value,
            value: optionLike.value,
            disabled: true === optionLike.disabled
        };
    }
    getSlottedRadioChildren() {
        return this.assignedRadioElements.filter((element)=>element instanceof VeRadioElement);
    }
    getRadioChildren() {
        if (this.hasOptions) return Array.from(this.renderRoot.querySelectorAll('ve-radio[data-option-item]'));
        return this.getSlottedRadioChildren();
    }
    syncManagedRadio(radio) {
        radio.groupDisabled = this.disabled;
        radio.checked = radio.value === this.value;
    }
    syncRadioStates() {
        for (const radio of this.getRadioChildren())this.syncManagedRadio(radio);
    }
    isManagedRadio(radio) {
        if (this.hasOptions) return radio.hasAttribute('data-option-item');
        return radio.closest(VE_RADIO_GROUP_TAG) === this;
    }
    resolveEventRadioTarget(event) {
        if (event.target instanceof VeRadioElement) return event.target;
        return event.composedPath().find((node)=>node instanceof VeRadioElement);
    }
    handleSlotChange() {
        this.syncRadioStates();
    }
    renderOptionLabel(option) {
        if (isRenderableTemplateValue(option.label)) return option.label;
        return String(option.value);
    }
    renderOptionNodes() {
        return repeat(this.normalizedOptions, (option)=>option.value, (option)=>html`
        <ve-radio
          data-option-item
          class="ve-radio-group-option-item"
          .value=${option.value}
          .disabled=${option.disabled}
          .groupDisabled=${this.disabled}
          .checked=${option.value === this.value}
        >
          ${this.renderOptionLabel(option)}
        </ve-radio>
      `);
    }
    render() {
        const groupClass = classMap({
            've-radio-group': true,
            've-radio-group-direction-horizontal': 'horizontal' === this.direction,
            've-radio-group-direction-vertical': 'vertical' === this.direction
        });
        return html`
      <div
        class=${groupClass}
        part="group"
        role="radiogroup"
        aria-label=${ifDefined(getForwardedAriaAttribute(this, 'aria-label'))}
        aria-labelledby=${ifDefined(getForwardedAriaAttribute(this, 'aria-labelledby'))}
        aria-describedby=${ifDefined(getForwardedAriaAttribute(this, 'aria-describedby'))}
      >
        ${this.hasOptions ? this.renderOptionNodes() : html`<slot @slotchange=${this.handleSlotChange}></slot>`}
      </div>
    `;
    }
}
defineCustomElement(VE_RADIO_TAG, VeRadioElement);
defineCustomElement(VE_RADIO_GROUP_TAG, VeRadioGroupElement);
export { VE_RADIO_GROUP_TAG, VE_RADIO_TAG, VeRadioElement, VeRadioGroupElement };
