// @ve-design/web/web/dist/ve-switch.js
import { css, html } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
const veSwitchTokenDefaults = css`
  :host {
    --ve-switch-track-width-small: 24px;
    --ve-switch-track-height-small: 14px;
    --ve-switch-thumb-size-small: 10px;

    --ve-switch-track-width-default: 32px;
    --ve-switch-track-height-default: 18px;
    --ve-switch-thumb-size-default: 14px;

    --ve-switch-track-width-large: 40px;
    --ve-switch-track-height-large: 22px;
    --ve-switch-thumb-size-large: 18px;

    --ve-switch-thumb-offset: 1px;
    --ve-switch-track-radius: var(--radius-full);
    --ve-switch-thumb-radius: var(--radius-full);
    --ve-switch-transition-duration: 200ms;
    --ve-switch-disabled-opacity: var(--state-disabled-ratio);

    --ve-switch-track-background-unchecked: var(--color-bg-muted);
    --ve-switch-track-background-checked: var(--color-bg-primary);
    --ve-switch-track-border-unchecked: var(--color-border-default);
    --ve-switch-track-border-checked: transparent;
    --ve-switch-track-border-width: var(--stroke-weight-strong);
    --ve-switch-thumb-background: var(--color-white);
    --ve-switch-thumb-shadow: 0 1px 3px rgba(0, 0, 0, 0.2);

    --ve-switch-current-track-width: var(--ve-switch-track-width-default);
    --ve-switch-current-track-height: var(--ve-switch-track-height-default);
    --ve-switch-current-thumb-size: var(--ve-switch-thumb-size-default);
    --ve-switch-current-track-background: var(
      --ve-switch-track-background-unchecked
    );
    --ve-switch-current-track-border: var(--ve-switch-track-border-unchecked);
    --ve-switch-current-hover-ratio: var(--state-hover-subtle-ratio);
    --ve-switch-current-hover-mix: var(--state-hover-subtle-mix);
  }

  :host([checked]) {
    --ve-switch-current-track-background: var(--ve-switch-track-background-checked);
    --ve-switch-current-track-border: var(--ve-switch-track-border-checked);
    --ve-switch-current-hover-ratio: var(--state-hover-solid-ratio);
    --ve-switch-current-hover-mix: var(--state-hover-solid-mix);
  }

  :host([size='small']) {
    --ve-switch-current-track-height: var(--ve-switch-track-height-small);
    --ve-switch-current-track-width: var(--ve-switch-track-width-small);
    --ve-switch-current-thumb-size: var(--ve-switch-thumb-size-small);
  }

  :host([size='large']) {
    --ve-switch-current-track-height: var(--ve-switch-track-height-large);
    --ve-switch-current-track-width: var(--ve-switch-track-width-large);
    --ve-switch-current-thumb-size: var(--ve-switch-thumb-size-large);
  }
`;
const token = veSwitchTokenDefaults;
const ve_switch = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      align-items: center;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    :host([disabled]) {
      cursor: not-allowed;
      opacity: var(--ve-switch-disabled-opacity);
    }

    .ve-switch {
      position: relative;
      display: inline-flex;
      align-items: center;
      box-sizing: border-box;
      inline-size: var(--ve-switch-current-track-width);
      block-size: var(--ve-switch-current-track-height);
      margin: 0;
      padding: 0;
      border: var(--ve-switch-track-border-width) solid
        var(--ve-switch-current-track-border);
      border-radius: var(--ve-switch-track-radius);
      background: var(--ve-switch-current-track-background);
      appearance: none;
      cursor: pointer;
      user-select: none;
      transition:
        background-color var(--ve-switch-transition-duration) ease,
        border-color var(--ve-switch-transition-duration) ease;
    }

    .ve-switch:focus-visible {
      outline: var(--stroke-2) solid var(--color-border-interactive-active);
      outline-offset: var(--stroke-2);
    }

    :host([checked]) .ve-switch {
      background: var(--ve-switch-current-track-background);
      border-color: var(--ve-switch-current-track-border);
    }

    :host(:not([disabled])) .ve-switch:hover {
      background: color-mix(
        var(--ve-switch-current-track-background)
          var(--ve-switch-current-hover-ratio),
        var(--ve-switch-current-hover-mix)
      );
    }

    :host([disabled]) .ve-switch {
      cursor: not-allowed;
      pointer-events: none;
    }

    .ve-switch-thumb {
      position: absolute;
      top: 50%;
      left: var(--ve-switch-thumb-offset);
      inline-size: var(--ve-switch-current-thumb-size);
      block-size: var(--ve-switch-current-thumb-size);
      border-radius: var(--ve-switch-thumb-radius);
      background: var(--ve-switch-thumb-background);
      box-shadow: var(--ve-switch-thumb-shadow);
      transform: translateY(-50%);
      transition: left var(--ve-switch-transition-duration) ease;
    }

    :host([checked]) .ve-switch-thumb {
      left: calc(
        var(--ve-switch-current-track-width) -
          var(--ve-switch-current-thumb-size) - var(--ve-switch-thumb-offset) -
          var(--ve-switch-track-border-width) -
          var(--ve-switch-track-border-width)
      );
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-switch,
      .ve-switch-thumb {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _init_checked, _init_extra__init_checked, _init_defaultChecked, _init_extra__init_defaultChecked, _init_disabled, _init_extra__init_disabled, _init_size, _init_extra__init_size, _init_controlElement, _init_extra__init_controlElement;
const VE_SWITCH_TAG = 've-switch';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby'
];
_dec = property({
    type: Boolean,
    reflect: true
}), _dec1 = property({
    type: Boolean,
    attribute: 'default-checked'
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec4 = query('.ve-switch');
class VeSwitchElement extends VeElement {
    static{
        ({ e: [_init_checked, _init_extra__init_checked, _init_defaultChecked, _init_extra__init_defaultChecked, _init_disabled, _init_extra__init_disabled, _init_size, _init_extra__init_size, _init_controlElement, _init_extra__init_controlElement] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "checked"
            ],
            [
                _dec1,
                1,
                "defaultChecked"
            ],
            [
                _dec2,
                1,
                "disabled"
            ],
            [
                _dec3,
                1,
                "size"
            ],
            [
                _dec4,
                1,
                "controlElement"
            ]
        ]));
    }
    static styles = ve_switch;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    #___private_checked_1 = (()=>{
        const _value = _init_checked(this, void 0);
        _init_extra__init_checked(this);
        return _value;
    })();
    get checked() {
        return this.#___private_checked_1;
    }
    set checked(_v) {
        this.#___private_checked_1 = _v;
    }
    #___private_defaultChecked_2 = (()=>{
        const _value = _init_defaultChecked(this, false);
        _init_extra__init_defaultChecked(this);
        return _value;
    })();
    get defaultChecked() {
        return this.#___private_defaultChecked_2;
    }
    set defaultChecked(_v) {
        this.#___private_defaultChecked_2 = _v;
    }
    #___private_disabled_3 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_3;
    }
    set disabled(_v) {
        this.#___private_disabled_3 = _v;
    }
    #___private_size_4 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_4;
    }
    set size(_v) {
        this.#___private_size_4 = _v;
    }
    #___private_controlElement_5 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_5;
    }
    set controlElement(_v) {
        this.#___private_controlElement_5 = _v;
    }
    get controlAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get controlAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get controlAriaDescribedby() {
        return this.getAttribute('aria-describedby') ?? void 0;
    }
    syncControlAccessibilityAttribute(name, value) {
        const control = this.controlElement;
        if (!control) return;
        if (null === value) return void control.removeAttribute(name);
        control.setAttribute(name, value);
    }
    connectedCallback() {
        super.connectedCallback();
        if (void 0 === this.checked && this.defaultChecked) this.checked = true;
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.syncControlAccessibilityAttribute(name, newValue);
    }
    click() {
        if (this.controlElement) return void this.controlElement.click();
        super.click();
    }
    emitChangeEvents(nextChecked) {
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail: {
                checked: nextChecked
            },
            bubbles: true,
            composed: true
        }));
    }
    setChecked(nextChecked, emitEvents = false) {
        if (this.checked === nextChecked) return;
        this.checked = nextChecked;
        if (emitEvents) this.emitChangeEvents(nextChecked);
    }
    toggleChecked() {
        if (this.disabled) return;
        this.setChecked(true !== this.checked, true);
    }
    render() {
        return html`
      <button
        class="ve-switch"
        part="control track"
        type="button"
        role="switch"
        ?disabled=${this.disabled}
        .ariaChecked=${true === this.checked ? 'true' : 'false'}
        aria-disabled=${ifDefined(this.disabled ? 'true' : void 0)}
        aria-label=${ifDefined(this.controlAriaLabel)}
        aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
        aria-describedby=${ifDefined(this.controlAriaDescribedby)}
        @click=${this.toggleChecked}
      >
        <span class="ve-switch-thumb" part="thumb"></span>
      </button>
    `;
    }
}
defineCustomElement(VE_SWITCH_TAG, VeSwitchElement);
export { VE_SWITCH_TAG, VeSwitchElement };
