// @ve-design/web/web/dist/ve-suggestion.js
import { css, html, nothing } from "lit";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { VeDropdownElement, VeDropdownItemElement, VE_DROPDOWN_TAG, VE_DROPDOWN_ITEM_TAG } from "./97605.js";
import { VE_DROPDOWN_GROUP_TAG, VeDropdownGroupElement } from "./16491.js";
import { cornerDownLeftIcon } from "./icons/corner-down-left.js";
const acSuggestionTokenDefaults = css`
  :host {
    --ve-suggestion-min-width: 0px;
    --ve-suggestion-max-width: none;
    --ve-suggestion-panel-width: auto;
    --ve-suggestion-item-gap: var(--space-xs, 8px);
    --ve-suggestion-item-extra-color: var(--color-text-tertiary);
    --ve-suggestion-item-extra-font-size: var(--text-caption);
    --ve-suggestion-item-icon-size: 16px;
    --ve-suggestion-item-icon-color: var(--color-text-secondary);
    --ve-suggestion-enter-icon-size: 14px;
    --ve-suggestion-enter-icon-color: var(--color-text-tertiary);
  }
`;
const token = acSuggestionTokenDefaults;
const ve_suggestion = [
    componentReset,
    token,
    css`
    :host {
      display: contents;
    }

    :host([hidden]) {
      display: none;
    }

    ve-dropdown {
      display: block;
      inline-size: 100%;
      /* 关闭 dropdown-item 聚焦时的 inset 黑框 */
      --ve-dropdown-item-focus-shadow: none;
    }

    ve-dropdown::part(panel) {
      min-inline-size: var(--ve-suggestion-min-width);
      max-inline-size: var(--ve-suggestion-max-width);
      inline-size: var(--ve-suggestion-panel-width);
    }

    .ve-suggestion-item {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-suggestion-item-gap);
      inline-size: 100%;
    }

    .ve-suggestion-item-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: var(--ve-suggestion-item-icon-size);
      block-size: var(--ve-suggestion-item-icon-size);
      color: var(--ve-suggestion-item-icon-color);
    }

    .ve-suggestion-item-icon ve-icon {
      inline-size: 100%;
      block-size: 100%;
    }

    .ve-suggestion-item-label {
      flex: 1;
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-suggestion-item-extra {
      flex: none;
      color: var(--ve-suggestion-item-extra-color);
      font-size: var(--ve-suggestion-item-extra-font-size);
      white-space: nowrap;
    }

    .ve-suggestion-item-enter {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: var(--ve-suggestion-enter-icon-size);
      block-size: var(--ve-suggestion-enter-icon-size);
      color: var(--ve-suggestion-enter-icon-color);
    }

    .ve-suggestion-item-enter ve-icon {
      inline-size: 100%;
      block-size: 100%;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _init_items, _init_extra__init_items, _init_open, _init_extra__init_open, _init_placement, _init_extra__init_placement, _init_block, _init_extra__init_block, _init_disabled, _init_extra__init_disabled, _init_activeIndex, _init_extra__init_activeIndex, _init_dropdownElement, _init_extra__init_dropdownElement;
const VE_SUGGESTION_TAG = 've-suggestion';
function isGroupArray(items) {
    const first = items[0];
    return items.length > 0 && 'object' == typeof first && null !== first && 'items' in first;
}
function flattenItems(items) {
    if (0 === items.length) return [];
    if (isGroupArray(items)) return items.flatMap((group)=>group.items);
    return items;
}
function normalizeToGroups(items) {
    if (0 === items.length) return [];
    if (isGroupArray(items)) return items;
    return [
        {
            items: items
        }
    ];
}
_dec = property({
    attribute: false
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    type: String,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = state(), _dec6 = query('ve-dropdown');
class VeSuggestionElement extends VeElement {
    static{
        ({ e: [_init_items, _init_extra__init_items, _init_open, _init_extra__init_open, _init_placement, _init_extra__init_placement, _init_block, _init_extra__init_block, _init_disabled, _init_extra__init_disabled, _init_activeIndex, _init_extra__init_activeIndex, _init_dropdownElement, _init_extra__init_dropdownElement] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "items"
            ],
            [
                _dec1,
                1,
                "open"
            ],
            [
                _dec2,
                1,
                "placement"
            ],
            [
                _dec3,
                1,
                "block"
            ],
            [
                _dec4,
                1,
                "disabled"
            ],
            [
                _dec5,
                1,
                "activeIndex"
            ],
            [
                _dec6,
                1,
                "dropdownElement"
            ]
        ]));
    }
    static styles = ve_suggestion;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement,
        [VE_DROPDOWN_TAG]: VeDropdownElement,
        [VE_DROPDOWN_ITEM_TAG]: VeDropdownItemElement,
        [VE_DROPDOWN_GROUP_TAG]: VeDropdownGroupElement
    };
    #___private_items_1 = (()=>{
        const _value = _init_items(this, []);
        _init_extra__init_items(this);
        return _value;
    })();
    get items() {
        return this.#___private_items_1;
    }
    set items(_v) {
        this.#___private_items_1 = _v;
    }
    #___private_open_2 = (()=>{
        const _value = _init_open(this, false);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_2;
    }
    set open(_v) {
        this.#___private_open_2 = _v;
    }
    #___private_placement_3 = (()=>{
        const _value = _init_placement(this, 'top-start');
        _init_extra__init_placement(this);
        return _value;
    })();
    get placement() {
        return this.#___private_placement_3;
    }
    set placement(_v) {
        this.#___private_placement_3 = _v;
    }
    #___private_block_4 = (()=>{
        const _value = _init_block(this, true);
        _init_extra__init_block(this);
        return _value;
    })();
    get block() {
        return this.#___private_block_4;
    }
    set block(_v) {
        this.#___private_block_4 = _v;
    }
    #___private_disabled_5 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_5;
    }
    set disabled(_v) {
        this.#___private_disabled_5 = _v;
    }
    #___private_activeIndex_6 = (()=>{
        const _value = _init_activeIndex(this, 0);
        _init_extra__init_activeIndex(this);
        return _value;
    })();
    get activeIndex() {
        return this.#___private_activeIndex_6;
    }
    set activeIndex(_v) {
        this.#___private_activeIndex_6 = _v;
    }
    #___private_dropdownElement_7 = (()=>{
        const _value = _init_dropdownElement(this);
        _init_extra__init_dropdownElement(this);
        return _value;
    })();
    get dropdownElement() {
        return this.#___private_dropdownElement_7;
    }
    set dropdownElement(_v) {
        this.#___private_dropdownElement_7 = _v;
    }
    resizeObserver = null;
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('keydown', this.handleHostKeyDownCapture, true);
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.observeTriggerWidth();
    }
    disconnectedCallback() {
        this.removeEventListener('keydown', this.handleHostKeyDownCapture, true);
        this.resizeObserver?.disconnect();
        this.resizeObserver = null;
        super.disconnectedCallback();
    }
    handleHostKeyDownCapture = (event)=>{
        if (!this.open || this.disabled) return;
        const key = event.key;
        if ('ArrowDown' !== key && 'ArrowUp' !== key && 'Enter' !== key && 'Escape' !== key) return;
        const consumed = this.handleKeyDown(event);
        if (consumed) {
            event.stopPropagation();
            event.stopImmediatePropagation();
        }
    };
    observeTriggerWidth() {
        if ("u" < typeof ResizeObserver || !this.dropdownElement) return;
        const syncWidth = ()=>{
            const width = this.dropdownElement?.offsetWidth ?? 0;
            if (width > 0) this.style.setProperty('--ve-suggestion-panel-width', `${width}px`);
        };
        this.resizeObserver = new ResizeObserver(syncWidth);
        this.resizeObserver.observe(this.dropdownElement);
        syncWidth();
    }
    get flattenedItems() {
        return flattenItems(this.items).filter((item)=>!item.disabled);
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('items') || changedProperties.has('open')) {
            const total = this.flattenedItems.length;
            if (0 === total) this.activeIndex = 0;
            else if (this.activeIndex >= total) this.activeIndex = 0;
            else if (this.activeIndex < 0) this.activeIndex = 0;
        }
    }
    handleKeyDown = (event)=>{
        if (this.disabled || !this.open) return false;
        const items = this.flattenedItems;
        if (0 === items.length) {
            if ('Escape' === event.key) {
                event.preventDefault();
                this.setOpenInternal(false);
                return true;
            }
            return false;
        }
        if ('ArrowDown' === event.key) {
            event.preventDefault();
            this.activeIndex = (this.activeIndex + 1) % items.length;
            return true;
        }
        if ('ArrowUp' === event.key) {
            event.preventDefault();
            this.activeIndex = (this.activeIndex - 1 + items.length) % items.length;
            return true;
        }
        if ('Enter' === event.key) {
            event.preventDefault();
            const item = items[this.activeIndex];
            if (item) this.selectItem(item);
            return true;
        }
        if ('Escape' === event.key) {
            event.preventDefault();
            this.setOpenInternal(false);
            return true;
        }
        return false;
    };
    setOpen(open) {
        this.setOpenInternal(open);
    }
    setOpenInternal(open) {
        if (this.open === open) return;
        this.open = open;
        this.dispatchEvent(new CustomEvent('ve-suggestion-open-change', {
            detail: {
                open
            },
            bubbles: true,
            composed: true
        }));
    }
    selectItem(item) {
        if (item.disabled) return;
        this.dispatchEvent(new CustomEvent('ve-suggestion-select', {
            detail: {
                value: item.value,
                item
            },
            bubbles: true,
            composed: true
        }));
        this.setOpenInternal(false);
    }
    handleDropdownOpenChange = (event)=>{
        event.stopPropagation();
        if (!event.detail.open && this.open && 'trigger' !== event.detail.source) this.setOpenInternal(false);
    };
    handleItemClick = (item)=>(event)=>{
            event.stopPropagation();
            this.selectItem(item);
        };
    renderItem(item, activeItem) {
        const isActive = item === activeItem;
        return html`
      <ve-dropdown-item
        value=${item.value}
        ?disabled=${item.disabled ?? false}
        ?selected=${isActive}
        @click=${this.handleItemClick(item)}
      >
        <span class="ve-suggestion-item">
          ${item.icon ? html`<span class="ve-suggestion-item-icon" aria-hidden="true">
                <ve-icon name=${item.icon}></ve-icon>
              </span>` : nothing}
          <span class="ve-suggestion-item-label">${item.label}</span>
          ${item.extra ? html`<span class="ve-suggestion-item-extra">${item.extra}</span>` : nothing}
          ${isActive ? html`<span class="ve-suggestion-item-enter" aria-hidden="true">
                <ve-icon .definition=${cornerDownLeftIcon}></ve-icon>
              </span>` : nothing}
        </span>
      </ve-dropdown-item>
    `;
    }
    renderMenu() {
        const groups = normalizeToGroups(this.items);
        const flattened = this.flattenedItems;
        const activeItem = flattened[this.activeIndex];
        return groups.map((group)=>{
            const itemsTemplate = group.items.map((item)=>this.renderItem(item, activeItem));
            if (group.label) return html`
          <ve-dropdown-group title=${group.label}>
            ${itemsTemplate}
          </ve-dropdown-group>
        `;
            return html`${itemsTemplate}`;
        });
    }
    render() {
        const placement = this.placement;
        return html`
      <ve-dropdown
        trigger="click"
        .open=${this.open}
        .position=${placement}
        ?block=${this.block}
        ?disabled=${this.disabled}
        @ve-open-change=${this.handleDropdownOpenChange}
      >
        <slot slot="trigger"></slot>
        ${this.renderMenu()}
      </ve-dropdown>
    `;
    }
}
defineCustomElement(VE_SUGGESTION_TAG, VeSuggestionElement);
export { VE_SUGGESTION_TAG, VeSuggestionElement };
