// @ve-design/web/web/dist/ve-time-picker.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createFloatingAutoUpdate, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { closeIcon } from "./icons/close.js";
import { loadingIcon } from "./42938.js";
import { timeClockIcon } from "./icons/time-clock.js";
const veTimePickerTokenDefaults = css`
  :host {
    --ve-time-picker-font-family: var(--font-sans);
    --ve-time-picker-font-weight: var(--font-weight-normal);
    --ve-time-picker-line-height: var(--line-height-body);
    --ve-time-picker-font-size-small: var(--text-body-sm);
    --ve-time-picker-font-size-default: var(--text-body);
    --ve-time-picker-font-size-large: var(--text-body);
    --ve-time-picker-width: calc(var(--height-default) * 4.5);
    --ve-time-picker-range-width: calc(var(--height-default) * 7);
    --ve-time-picker-current-width: var(--ve-time-picker-width);
    --ve-time-picker-height-small: var(--height-small);
    --ve-time-picker-height-default: var(--height-default);
    --ve-time-picker-height-large: var(--height-large);
    --ve-time-picker-padding-inline-small: var(--space-xxs);
    --ve-time-picker-padding-inline-default: var(--space-xxs);
    --ve-time-picker-padding-inline-large: var(--space-xs);
    --ve-time-picker-padding-block: var(--space-xxxs);
    --ve-time-picker-radius: var(--radius-md);
    --ve-time-picker-border-width: var(--stroke-weight-base);
    --ve-time-picker-gap: var(--space-xxxs);
    --ve-time-picker-label-padding-inline: var(--space-xxs);
    --ve-time-picker-label-color: var(--color-text-secondary);
    --ve-time-picker-control-background: var(--color-bg-control);
    --ve-time-picker-control-background-hover: var(--color-bg-base);
    --ve-time-picker-control-background-focus: var(--color-bg-base);
    --ve-time-picker-control-background-disabled: var(--color-bg-control);
    --ve-time-picker-control-text: var(--color-text-primary);
    --ve-time-picker-control-text-disabled: var(--color-text-disable);
    --ve-time-picker-control-placeholder: var(--color-text-tertiary);
    --ve-time-picker-control-border-default: var(--color-border-default);
    --ve-time-picker-control-border-hover: var(--color-border-interactive);
    --ve-time-picker-control-border-focus: var(
      --color-border-interactive-active
    );
    --ve-time-picker-control-border-disabled: var(--color-border-default);
    --ve-time-picker-focus-border-width: var(--stroke-weight-base);
    --ve-time-picker-shadow-focus: none;
    --ve-time-picker-status-warning-background: var(
      --color-bg-warning-secondary
    );
    --ve-time-picker-status-warning-background-hover: color-mix(
      var(--ve-time-picker-status-warning-background)
        var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-time-picker-status-warning-border: var(--color-border-warning);
    --ve-time-picker-status-warning-border-hover: var(--color-text-warning);
    --ve-time-picker-status-warning-border-focus: var(--color-text-warning);
    --ve-time-picker-status-error-background: var(--color-bg-danger-secondary);
    --ve-time-picker-status-error-background-hover: color-mix(
      var(--ve-time-picker-status-error-background)
        var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-time-picker-status-error-border: var(--color-border-danger);
    --ve-time-picker-status-error-border-hover: var(--color-text-danger);
    --ve-time-picker-status-error-border-focus: var(--color-text-danger);
    --ve-time-picker-description-color: var(--color-text-tertiary);
    --ve-time-picker-status-warning-description-color: var(
      --color-text-warning
    );
    --ve-time-picker-status-error-description-color: var(--color-text-danger);
    --ve-time-picker-feedback-font-size: var(--text-caption);
    --ve-time-picker-feedback-line-height: var(--line-height-body);
    --ve-time-picker-prefix-color: var(--color-text-secondary);
    --ve-time-picker-arrow-color: var(--color-icon-secondary);
    --ve-time-picker-clear-color: var(--color-text-tertiary);
    --ve-time-picker-clear-hover-color: var(--color-text-secondary);
    --ve-time-picker-clear-hover-background: var(--color-bg-muted);
    --ve-time-picker-icon-size: var(--space-s);
    --ve-time-picker-panel-background: var(--color-bg-overlay);
    --ve-time-picker-panel-border-color: var(--color-border-default);
    --ve-time-picker-panel-shadow: var(--shadow-md);
    --ve-time-picker-panel-radius: var(--radius-lg);
    --ve-time-picker-panel-z-index: 1000;
    --ve-time-picker-panel-gap: var(--space-xxxs);
    --ve-time-picker-panel-padding: var(--space-xxs);
    --ve-time-picker-panel-column-gap: var(--space-xxxs);
    --ve-time-picker-panel-column-width: calc(
      var(--height-default) + var(--space-xs)
    );
    --ve-time-picker-panel-column-height: calc(var(--height-default) * 6);
    --ve-time-picker-range-segment-text: var(--color-text-secondary);
    --ve-time-picker-range-segment-text-active: var(--color-text-primary);
    --ve-time-picker-range-segment-placeholder: var(--color-text-tertiary);
    --ve-time-picker-range-segment-background-active: var(--color-bg-surface);
    --ve-time-picker-range-segment-padding-inline: var(--space-xxxs);
    --ve-time-picker-range-segment-radius: var(--radius-md);
    --ve-time-picker-option-height: var(--height-default);
    --ve-time-picker-option-padding-inline: var(--space-xxs);
    --ve-time-picker-option-radius: var(--radius-md);
    --ve-time-picker-option-text: var(--color-text-tertiary);
    --ve-time-picker-option-text-disabled: var(--color-text-disable);
    --ve-time-picker-option-hover-background: var(--color-bg-surface);
    --ve-time-picker-option-active-background: color-mix(
      var(--color-bg-overlay) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );
    --ve-time-picker-option-selected-background: transparent;
    --ve-time-picker-option-selected-text: var(--color-text-primary);
    --ve-time-picker-column-border-color: var(--color-border-default);
    --ve-time-picker-loading-color: var(--color-text-tertiary);
    --ve-time-picker-disabled-opacity: var(--state-disabled-ratio);
    --ve-time-picker-transition-duration: var(--motion-fast);
    --ve-time-picker-transition-timing: var(--ease-standard);

    --ve-time-picker-current-height: var(--ve-time-picker-height-default);
    --ve-time-picker-current-padding-inline: var(
      --ve-time-picker-padding-inline-default
    );
    --ve-time-picker-current-font-size: var(--ve-time-picker-font-size-default);
    --ve-time-picker-current-background: var(
      --ve-time-picker-control-background
    );
    --ve-time-picker-current-background-hover: var(
      --ve-time-picker-control-background-hover
    );
    --ve-time-picker-current-border-color: var(
      --ve-time-picker-control-border-default
    );
    --ve-time-picker-current-border-color-hover: var(
      --ve-time-picker-control-border-hover
    );
    --ve-time-picker-current-focus-ring-color: var(
      --ve-time-picker-control-border-focus
    );
    --ve-time-picker-current-description-color: var(
      --ve-time-picker-description-color
    );
  }

  :host([size='small']) {
    --ve-time-picker-current-height: var(--ve-time-picker-height-small);
    --ve-time-picker-current-padding-inline: var(
      --ve-time-picker-padding-inline-small
    );
    --ve-time-picker-current-font-size: var(--ve-time-picker-font-size-small);
  }

  :host([range]) {
    --ve-time-picker-current-width: var(--ve-time-picker-range-width);
  }

  :host([size='large']) {
    --ve-time-picker-current-height: var(--ve-time-picker-height-large);
    --ve-time-picker-current-padding-inline: var(
      --ve-time-picker-padding-inline-large
    );
    --ve-time-picker-current-font-size: var(--ve-time-picker-font-size-large);
  }

  :host([status='warning']) {
    --ve-time-picker-current-background: var(
      --ve-time-picker-status-warning-background
    );
    --ve-time-picker-current-background-hover: var(
      --ve-time-picker-status-warning-background-hover
    );
    --ve-time-picker-current-border-color: var(
      --ve-time-picker-status-warning-border
    );
    --ve-time-picker-current-border-color-hover: var(
      --ve-time-picker-status-warning-border-hover
    );
    --ve-time-picker-current-focus-ring-color: var(
      --ve-time-picker-status-warning-border-focus
    );
    --ve-time-picker-current-description-color: var(
      --ve-time-picker-status-warning-description-color
    );
  }

  :host([status='error']) {
    --ve-time-picker-current-background: var(
      --ve-time-picker-status-error-background
    );
    --ve-time-picker-current-background-hover: var(
      --ve-time-picker-status-error-background-hover
    );
    --ve-time-picker-current-border-color: var(
      --ve-time-picker-status-error-border
    );
    --ve-time-picker-current-border-color-hover: var(
      --ve-time-picker-status-error-border-hover
    );
    --ve-time-picker-current-focus-ring-color: var(
      --ve-time-picker-status-error-border-focus
    );
    --ve-time-picker-current-description-color: var(
      --ve-time-picker-status-error-description-color
    );
  }
`;
const token = veTimePickerTokenDefaults;
const ve_time_picker = [
    componentReset,
    token,
    css`
    :host {
      position: relative;
      display: inline-flex;
      inline-size: var(--ve-time-picker-current-width);
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-time-picker {
      position: relative;
      inline-size: 100%;
    }

    .ve-time-picker-field {
      display: flex;
      align-items: stretch;
      inline-size: 100%;
    }

    .ve-time-picker-control {
      appearance: none;
      display: flex;
      align-items: center;
      gap: var(--ve-time-picker-gap);
      inline-size: 100%;
      block-size: var(--ve-time-picker-current-height);
      min-inline-size: 0;
      box-sizing: border-box;
      padding-inline: var(--ve-time-picker-current-padding-inline);
      padding-block: var(--ve-time-picker-padding-block);
      border: var(--ve-time-picker-border-width) solid
        var(--ve-time-picker-current-border-color);
      border-radius: var(--ve-time-picker-radius);
      background: var(--ve-time-picker-current-background);
      color: var(--ve-time-picker-control-text);
      font-family: var(--ve-time-picker-font-family);
      font-size: var(--ve-time-picker-current-font-size);
      font-weight: var(--ve-time-picker-font-weight);
      line-height: var(--ve-time-picker-line-height);
      cursor: pointer;
      transition:
        border-color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        box-shadow var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        background-color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        opacity var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing);
      overflow: hidden;
    }

    .ve-time-picker-label {
      display: inline-flex;
      align-items: center;
      flex: none;
      box-sizing: border-box;
      block-size: var(--ve-time-picker-current-height);
      min-inline-size: 0;
      padding-inline: var(--ve-time-picker-label-padding-inline);
      padding-block: var(--ve-time-picker-padding-block);
      border: var(--ve-time-picker-border-width) solid
        var(--ve-time-picker-current-border-color);
      border-start-start-radius: var(--ve-time-picker-radius);
      border-end-start-radius: var(--ve-time-picker-radius);
      border-start-end-radius: 0;
      border-end-end-radius: 0;
      background: var(--ve-time-picker-current-background);
      color: var(--ve-time-picker-label-color);
      font-family: var(--ve-time-picker-font-family);
      font-size: var(--ve-time-picker-current-font-size);
      font-weight: var(--ve-time-picker-font-weight);
      line-height: var(--ve-time-picker-line-height);
      white-space: nowrap;
      cursor: pointer;
      transition:
        border-color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        background-color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        opacity var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing);
    }

    .ve-time-picker-with-label .ve-time-picker-control {
      border-start-start-radius: 0;
      border-end-start-radius: 0;
      margin-inline-start: calc(var(--ve-time-picker-border-width) * -1);
    }

    .ve-time-picker-label-text,
    .ve-time-picker-input,
    .ve-time-picker-range-value,
    .ve-time-picker-range-segment {
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-time-picker-input {
      appearance: none;
      flex: 1;
      inline-size: 100%;
      padding: 0;
      border: none;
      outline: none;
      background: transparent;
      color: inherit;
      font: inherit;
      text-align: start;
      cursor: text;
    }

    .ve-time-picker-input::placeholder {
      color: var(--ve-time-picker-control-placeholder);
    }

    .ve-time-picker-range-value {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      flex: 1;
    }

    .ve-time-picker-range-segment {
      appearance: none;
      min-inline-size: 0;
      padding-inline: var(--ve-time-picker-range-segment-padding-inline);
      border: none;
      border-radius: var(--ve-time-picker-range-segment-radius);
      background: transparent;
      color: var(--ve-time-picker-range-segment-text);
      font: inherit;
      text-align: start;
      cursor: pointer;
      transition:
        background-color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing);
    }

    .ve-time-picker-range-segment-placeholder {
      color: var(--ve-time-picker-range-segment-placeholder);
    }

    :host([open]) .ve-time-picker-range-segment-active {
      color: var(--ve-time-picker-range-segment-text-active);
      background: var(--ve-time-picker-range-segment-background-active);
    }

    .ve-time-picker-range-separator {
      color: var(--color-icon-secondary);
      flex: none;
    }

    .ve-time-picker-prefix,
    .ve-time-picker-arrow,
    .ve-time-picker-loading {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
    }

    .ve-time-picker-prefix {
      color: var(--ve-time-picker-prefix-color);
    }

    .ve-time-picker-prefix-empty {
      display: none;
    }

    .ve-time-picker-arrow {
      color: var(--ve-time-picker-arrow-color);
      transition: transform var(--ve-time-picker-transition-duration)
        var(--ve-time-picker-transition-timing);
    }

    .ve-time-picker-loading {
      color: var(--ve-time-picker-loading-color);
    }

    .ve-time-picker-prefix ve-icon,
    .ve-time-picker-arrow ve-icon,
    .ve-time-picker-loading ve-icon,
    .ve-time-picker-clear ve-icon {
      --ve-icon-size: var(--ve-time-picker-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-time-picker-loading ve-icon {
      animation: ve-time-picker-loading-spin var(--motion-slow) linear infinite;
    }

    .ve-time-picker-actions {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-time-picker-icon-size);
      block-size: var(--ve-time-picker-icon-size);
      margin-inline-start: auto;
      flex: none;
      pointer-events: auto;
    }

    .ve-time-picker-clear {
      appearance: none;
      position: absolute;
      inset: 0;
      z-index: 1;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: 100%;
      block-size: 100%;
      padding: 0;
      border: none;
      border-radius: var(--radius-full);
      background: transparent;
      color: var(--ve-time-picker-clear-color);
      cursor: pointer;
      opacity: 0;
      transform: scale(0.85);
      pointer-events: none;
      transition:
        opacity var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        transform var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        background-color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing);
    }

    :host(:not([disabled])) .ve-time-picker-control:hover .ve-time-picker-clear,
    :host([open]) .ve-time-picker-clear {
      opacity: 1;
      transform: scale(1);
      pointer-events: auto;
    }

    .ve-time-picker-clear:hover {
      background: var(--ve-time-picker-clear-hover-background);
      color: var(--ve-time-picker-clear-hover-color);
    }

    .ve-time-picker-arrow,
    .ve-time-picker-loading {
      pointer-events: none;
      transition:
        opacity var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        transform var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing);
    }

    :host(:not([disabled]))
      .ve-time-picker-control:hover
      .ve-time-picker-clear
      + .ve-time-picker-arrow,
    :host([open]) .ve-time-picker-clear + .ve-time-picker-arrow {
      opacity: 0;
      transform: scale(0.85);
    }

    :host(:not([disabled])) .ve-time-picker-control:hover,
    :host(:not([disabled])) .ve-time-picker-label:hover {
      border-color: var(--ve-time-picker-current-border-color-hover);
      background: var(--ve-time-picker-current-background-hover);
    }

    :host(:not([disabled])) .ve-time-picker-control:focus-within,
    :host([open]) .ve-time-picker-control {
      outline: none;
      background: var(--ve-time-picker-control-background-focus);
      border-color: var(--ve-time-picker-current-focus-ring-color);
      border-width: var(--ve-time-picker-focus-border-width);
      box-shadow: var(--ve-time-picker-shadow-focus);
    }

    :host([open]) .ve-time-picker-label {
      background: var(--ve-time-picker-control-background-focus);
      border-color: var(--ve-time-picker-current-focus-ring-color);
      border-width: var(--ve-time-picker-focus-border-width);
    }

    :host([disabled]) .ve-time-picker-control,
    :host([disabled]) .ve-time-picker-label {
      color: var(--ve-time-picker-control-text-disabled);
      background: var(--ve-time-picker-control-background-disabled);
      border-color: var(--ve-time-picker-control-border-disabled);
      opacity: var(--ve-time-picker-disabled-opacity);
      cursor: not-allowed;
    }

    :host([disabled]) .ve-time-picker-input,
    :host([disabled]) .ve-time-picker-range-segment {
      color: var(--ve-time-picker-control-text-disabled);
      cursor: not-allowed;
    }

    .ve-time-picker-description {
      max-inline-size: var(--ve-time-picker-current-width);
      min-block-size: calc(
        var(--ve-time-picker-feedback-font-size) *
          var(--ve-time-picker-feedback-line-height)
      );
      margin-block-start: var(--space-xxxs);
      color: var(--ve-time-picker-current-description-color);
      font-family: var(--ve-time-picker-font-family);
      font-size: var(--ve-time-picker-feedback-font-size);
      line-height: var(--ve-time-picker-feedback-line-height);
    }

    .ve-time-picker-panel {
      position: fixed;
      inset: auto;
      inset-block-start: 0;
      inset-inline-start: 0;
      z-index: var(--ve-time-picker-panel-z-index);
      box-sizing: border-box;
      inline-size: max-content;
      min-inline-size: calc(var(--ve-time-picker-panel-column-width) * 2);
      margin: 0;
      padding: var(--ve-time-picker-panel-padding);
      border: var(--ve-time-picker-border-width) solid
        var(--ve-time-picker-panel-border-color);
      border-radius: var(--ve-time-picker-panel-radius);
      background: var(--ve-time-picker-panel-background);
      box-shadow: var(--ve-time-picker-panel-shadow);
      display: flex;
      gap: var(--ve-time-picker-panel-column-gap);
      font-family: var(--ve-time-picker-font-family);
      font-size: var(--ve-time-picker-current-font-size);
      line-height: var(--ve-time-picker-line-height);
      color: var(--ve-time-picker-option-text);
    }

    .ve-time-picker-range-section {
      display: flex;
      flex-direction: column;
      gap: var(--ve-time-picker-panel-gap);
    }

    .ve-time-picker-range-columns {
      display: flex;
      gap: var(--ve-time-picker-panel-column-gap);
    }

    .ve-time-picker-column {
      display: flex;
      flex-direction: column;
      gap: var(--space-xxxs);
      inline-size: var(--ve-time-picker-panel-column-width);
      max-block-size: var(--ve-time-picker-panel-column-height);
      box-sizing: border-box;
      overflow: auto;
      overflow-x: hidden;
      scrollbar-width: none;
    }

    .ve-time-picker-column::-webkit-scrollbar {
      display: none;
      inline-size: 0;
      block-size: 0;
    }

    .ve-time-picker-option {
      appearance: none;
      display: flex;
      align-items: center;
      justify-content: center;
      inline-size: 100%;
      min-block-size: var(--ve-time-picker-option-height);
      box-sizing: border-box;
      padding-inline: var(--ve-time-picker-option-padding-inline);
      border: none;
      border-radius: var(--ve-time-picker-option-radius);
      background: transparent;
      color: var(--ve-time-picker-option-text);
      font: inherit;
      cursor: pointer;
      transition:
        background-color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        color var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing),
        opacity var(--ve-time-picker-transition-duration)
          var(--ve-time-picker-transition-timing);
    }

    .ve-time-picker-option:hover,
    .ve-time-picker-option:focus {
      outline: none;
      background: var(--ve-time-picker-option-hover-background);
    }

    .ve-time-picker-option:active {
      background: var(--ve-time-picker-option-active-background);
    }

    .ve-time-picker-option-selected {
      background: var(--ve-time-picker-option-selected-background);
      color: var(--ve-time-picker-option-selected-text);
      font-weight: var(--font-weight-medium);
    }

    .ve-time-picker-option-disabled {
      color: var(--ve-time-picker-option-text-disabled);
      opacity: var(--ve-time-picker-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-time-picker-option-disabled:hover,
    .ve-time-picker-option-disabled:focus,
    .ve-time-picker-option-disabled:active {
      background: transparent;
    }

    @keyframes ve-time-picker-loading-spin {
      to {
        transform: rotate(360deg);
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _dec24, _dec25, _dec26, _dec27, _dec28, _dec29, _init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_placeholder, _init_extra__init_placeholder, _init_label, _init_extra__init_label, _init_description, _init_extra__init_description, _init_width, _init_extra__init_width, _init_format, _init_extra__init_format, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_range, _init_extra__init_range, _init_loading, _init_extra__init_loading, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_hourStep, _init_extra__init_hourStep, _init_minuteStep, _init_extra__init_minuteStep, _init_secondStep, _init_extra__init_secondStep, _init_min, _init_extra__init_min, _init_max, _init_extra__init_max, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_controlElement, _init_extra__init_controlElement, _init_inputElement, _init_extra__init_inputElement, _init_rangeStartElement, _init_extra__init_rangeStartElement, _init_rangeEndElement, _init_extra__init_rangeEndElement, _init_fieldElement, _init_extra__init_fieldElement, _init_panelElement, _init_extra__init_panelElement, _init_panelPlacement, _init_extra__init_panelPlacement, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot, _init_activeRangeEnd, _init_extra__init_activeRangeEnd, _init_inputValue, _init_extra__init_inputValue;
const VE_TIME_PICKER_TAG = 've-time-picker';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby',
    'aria-required',
    'aria-invalid'
];
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        return null === value || 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
function normalizeAriaBooleanString(value) {
    if (null === value) return;
    return 'false' === value ? 'false' : 'true';
}
function toTwoDigit(value) {
    return String(value).padStart(2, '0');
}
function clampStep(value) {
    return Number.isFinite(value) && value > 0 ? Math.floor(value) : 1;
}
function isSupportedFormat(value) {
    return 'HH:mm:ss' === value || 'HH:mm' === value;
}
function parseTimeValue(value) {
    const trimmed = value.trim();
    if (!trimmed) return;
    const match = /^(\d{1,2}):(\d{1,2})(?::(\d{1,2}))?$/.exec(trimmed);
    if (!match) return;
    const hour = Number(match[1]);
    const minute = Number(match[2]);
    const second = void 0 === match[3] ? 0 : Number(match[3]);
    if (!Number.isInteger(hour) || !Number.isInteger(minute) || !Number.isInteger(second) || hour < 0 || hour > 23 || minute < 0 || minute > 59 || second < 0 || second > 59) return;
    return {
        hour,
        minute,
        second
    };
}
function getTimeValue(parts) {
    return 3600 * parts.hour + 60 * parts.minute + parts.second;
}
function formatTimeValue(parts, format) {
    const base = `${toTwoDigit(parts.hour)}:${toTwoDigit(parts.minute)}`;
    return 'HH:mm:ss' === format ? `${base}:${toTwoDigit(parts.second)}` : base;
}
function normalizeCssSize(value) {
    const trimmedValue = value.trim();
    if (!trimmedValue) return '';
    return /^\d+(?:\.\d+)?$/.test(trimmedValue) ? `${trimmedValue}px` : trimmedValue;
}
_dec = property({
    type: String
}), _dec1 = property({
    type: String,
    attribute: 'default-value'
}), _dec2 = property({
    type: String
}), _dec3 = property({
    type: String
}), _dec4 = property({
    type: String
}), _dec5 = property({
    type: String
}), _dec6 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec8 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec9 = property({
    type: Boolean,
    reflect: true
}), _dec10 = property({
    type: Boolean,
    reflect: true
}), _dec11 = property({
    type: Boolean,
    reflect: true
}), _dec12 = property({
    type: Boolean,
    attribute: 'allow-clear',
    reflect: true
}), _dec13 = property({
    type: Boolean,
    reflect: true
}), _dec14 = property({
    type: Number,
    attribute: 'hour-step'
}), _dec15 = property({
    type: Number,
    attribute: 'minute-step'
}), _dec16 = property({
    type: Number,
    attribute: 'second-step'
}), _dec17 = property({
    type: String
}), _dec18 = property({
    type: String
}), _dec19 = property({
    attribute: 'popup-match-select-width',
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER
}), _dec20 = query('.ve-time-picker-control'), _dec21 = query('.ve-time-picker-input'), _dec22 = query('.ve-time-picker-range-segment-start'), _dec23 = query('.ve-time-picker-range-segment-end'), _dec24 = query('.ve-time-picker-field'), _dec25 = query('.ve-time-picker-panel'), _dec26 = state(), _dec27 = state(), _dec28 = state(), _dec29 = state();
class VeTimePickerElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_placeholder, _init_extra__init_placeholder, _init_label, _init_extra__init_label, _init_description, _init_extra__init_description, _init_width, _init_extra__init_width, _init_format, _init_extra__init_format, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_range, _init_extra__init_range, _init_loading, _init_extra__init_loading, _init_allowClear, _init_extra__init_allowClear, _init_open, _init_extra__init_open, _init_hourStep, _init_extra__init_hourStep, _init_minuteStep, _init_extra__init_minuteStep, _init_secondStep, _init_extra__init_secondStep, _init_min, _init_extra__init_min, _init_max, _init_extra__init_max, _init_popupMatchSelectWidth, _init_extra__init_popupMatchSelectWidth, _init_controlElement, _init_extra__init_controlElement, _init_inputElement, _init_extra__init_inputElement, _init_rangeStartElement, _init_extra__init_rangeStartElement, _init_rangeEndElement, _init_extra__init_rangeEndElement, _init_fieldElement, _init_extra__init_fieldElement, _init_panelElement, _init_extra__init_panelElement, _init_panelPlacement, _init_extra__init_panelPlacement, _init_hasPrefixSlot, _init_extra__init_hasPrefixSlot, _init_activeRangeEnd, _init_extra__init_activeRangeEnd, _init_inputValue, _init_extra__init_inputValue] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "value"
            ],
            [
                _dec1,
                1,
                "defaultValue"
            ],
            [
                _dec2,
                1,
                "placeholder"
            ],
            [
                _dec3,
                1,
                "label"
            ],
            [
                _dec4,
                1,
                "description"
            ],
            [
                _dec5,
                1,
                "width"
            ],
            [
                _dec6,
                1,
                "format"
            ],
            [
                _dec7,
                1,
                "size"
            ],
            [
                _dec8,
                1,
                "status"
            ],
            [
                _dec9,
                1,
                "disabled"
            ],
            [
                _dec10,
                1,
                "range"
            ],
            [
                _dec11,
                1,
                "loading"
            ],
            [
                _dec12,
                1,
                "allowClear"
            ],
            [
                _dec13,
                1,
                "open"
            ],
            [
                _dec14,
                1,
                "hourStep"
            ],
            [
                _dec15,
                1,
                "minuteStep"
            ],
            [
                _dec16,
                1,
                "secondStep"
            ],
            [
                _dec17,
                1,
                "min"
            ],
            [
                _dec18,
                1,
                "max"
            ],
            [
                _dec19,
                1,
                "popupMatchSelectWidth"
            ],
            [
                _dec20,
                1,
                "controlElement"
            ],
            [
                _dec21,
                1,
                "inputElement"
            ],
            [
                _dec22,
                1,
                "rangeStartElement"
            ],
            [
                _dec23,
                1,
                "rangeEndElement"
            ],
            [
                _dec24,
                1,
                "fieldElement"
            ],
            [
                _dec25,
                1,
                "panelElement"
            ],
            [
                _dec26,
                1,
                "panelPlacement"
            ],
            [
                _dec27,
                1,
                "hasPrefixSlot"
            ],
            [
                _dec28,
                1,
                "activeRangeEnd"
            ],
            [
                _dec29,
                1,
                "inputValue"
            ]
        ]));
    }
    static styles = ve_time_picker;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_value_1 = (()=>{
        const _value = _init_value(this, '');
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_defaultValue_2 = (()=>{
        const _value = _init_defaultValue(this, '');
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_2;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_2 = _v;
    }
    #___private_placeholder_3 = (()=>{
        const _value = _init_placeholder(this, 'Select time');
        _init_extra__init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_3;
    }
    set placeholder(_v) {
        this.#___private_placeholder_3 = _v;
    }
    #___private_label_4 = (()=>{
        const _value = _init_label(this, '');
        _init_extra__init_label(this);
        return _value;
    })();
    get label() {
        return this.#___private_label_4;
    }
    set label(_v) {
        this.#___private_label_4 = _v;
    }
    #___private_description_5 = (()=>{
        const _value = _init_description(this, '');
        _init_extra__init_description(this);
        return _value;
    })();
    get description() {
        return this.#___private_description_5;
    }
    set description(_v) {
        this.#___private_description_5 = _v;
    }
    #___private_width_6 = (()=>{
        const _value = _init_width(this, '');
        _init_extra__init_width(this);
        return _value;
    })();
    get width() {
        return this.#___private_width_6;
    }
    set width(_v) {
        this.#___private_width_6 = _v;
    }
    #___private_format_7 = (()=>{
        const _value = _init_format(this, 'HH:mm');
        _init_extra__init_format(this);
        return _value;
    })();
    get format() {
        return this.#___private_format_7;
    }
    set format(_v) {
        this.#___private_format_7 = _v;
    }
    #___private_size_8 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_8;
    }
    set size(_v) {
        this.#___private_size_8 = _v;
    }
    #___private_status_9 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_9;
    }
    set status(_v) {
        this.#___private_status_9 = _v;
    }
    #___private_disabled_10 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_10;
    }
    set disabled(_v) {
        this.#___private_disabled_10 = _v;
    }
    #___private_range_11 = (()=>{
        const _value = _init_range(this, false);
        _init_extra__init_range(this);
        return _value;
    })();
    get range() {
        return this.#___private_range_11;
    }
    set range(_v) {
        this.#___private_range_11 = _v;
    }
    #___private_loading_12 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_12;
    }
    set loading(_v) {
        this.#___private_loading_12 = _v;
    }
    #___private_allowClear_13 = (()=>{
        const _value = _init_allowClear(this, false);
        _init_extra__init_allowClear(this);
        return _value;
    })();
    get allowClear() {
        return this.#___private_allowClear_13;
    }
    set allowClear(_v) {
        this.#___private_allowClear_13 = _v;
    }
    #___private_open_14 = (()=>{
        const _value = _init_open(this, false);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_14;
    }
    set open(_v) {
        this.#___private_open_14 = _v;
    }
    #___private_hourStep_15 = (()=>{
        const _value = _init_hourStep(this, 1);
        _init_extra__init_hourStep(this);
        return _value;
    })();
    get hourStep() {
        return this.#___private_hourStep_15;
    }
    set hourStep(_v) {
        this.#___private_hourStep_15 = _v;
    }
    #___private_minuteStep_16 = (()=>{
        const _value = _init_minuteStep(this, 1);
        _init_extra__init_minuteStep(this);
        return _value;
    })();
    get minuteStep() {
        return this.#___private_minuteStep_16;
    }
    set minuteStep(_v) {
        this.#___private_minuteStep_16 = _v;
    }
    #___private_secondStep_17 = (()=>{
        const _value = _init_secondStep(this, 1);
        _init_extra__init_secondStep(this);
        return _value;
    })();
    get secondStep() {
        return this.#___private_secondStep_17;
    }
    set secondStep(_v) {
        this.#___private_secondStep_17 = _v;
    }
    #___private_min_18 = (()=>{
        const _value = _init_min(this, '');
        _init_extra__init_min(this);
        return _value;
    })();
    get min() {
        return this.#___private_min_18;
    }
    set min(_v) {
        this.#___private_min_18 = _v;
    }
    #___private_max_19 = (()=>{
        const _value = _init_max(this, '');
        _init_extra__init_max(this);
        return _value;
    })();
    get max() {
        return this.#___private_max_19;
    }
    set max(_v) {
        this.#___private_max_19 = _v;
    }
    #___private_popupMatchSelectWidth_20 = (()=>{
        const _value = _init_popupMatchSelectWidth(this, false);
        _init_extra__init_popupMatchSelectWidth(this);
        return _value;
    })();
    get popupMatchSelectWidth() {
        return this.#___private_popupMatchSelectWidth_20;
    }
    set popupMatchSelectWidth(_v) {
        this.#___private_popupMatchSelectWidth_20 = _v;
    }
    #___private_controlElement_21 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_21;
    }
    set controlElement(_v) {
        this.#___private_controlElement_21 = _v;
    }
    #___private_inputElement_22 = (()=>{
        const _value = _init_inputElement(this, null);
        _init_extra__init_inputElement(this);
        return _value;
    })();
    get inputElement() {
        return this.#___private_inputElement_22;
    }
    set inputElement(_v) {
        this.#___private_inputElement_22 = _v;
    }
    #___private_rangeStartElement_23 = (()=>{
        const _value = _init_rangeStartElement(this, null);
        _init_extra__init_rangeStartElement(this);
        return _value;
    })();
    get rangeStartElement() {
        return this.#___private_rangeStartElement_23;
    }
    set rangeStartElement(_v) {
        this.#___private_rangeStartElement_23 = _v;
    }
    #___private_rangeEndElement_24 = (()=>{
        const _value = _init_rangeEndElement(this, null);
        _init_extra__init_rangeEndElement(this);
        return _value;
    })();
    get rangeEndElement() {
        return this.#___private_rangeEndElement_24;
    }
    set rangeEndElement(_v) {
        this.#___private_rangeEndElement_24 = _v;
    }
    #___private_fieldElement_25 = (()=>{
        const _value = _init_fieldElement(this, null);
        _init_extra__init_fieldElement(this);
        return _value;
    })();
    get fieldElement() {
        return this.#___private_fieldElement_25;
    }
    set fieldElement(_v) {
        this.#___private_fieldElement_25 = _v;
    }
    #___private_panelElement_26 = (()=>{
        const _value = _init_panelElement(this, null);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_26;
    }
    set panelElement(_v) {
        this.#___private_panelElement_26 = _v;
    }
    #___private_panelPlacement_27 = (()=>{
        const _value = _init_panelPlacement(this, 'bottom-start');
        _init_extra__init_panelPlacement(this);
        return _value;
    })();
    get panelPlacement() {
        return this.#___private_panelPlacement_27;
    }
    set panelPlacement(_v) {
        this.#___private_panelPlacement_27 = _v;
    }
    #___private_hasPrefixSlot_28 = (()=>{
        const _value = _init_hasPrefixSlot(this, false);
        _init_extra__init_hasPrefixSlot(this);
        return _value;
    })();
    get hasPrefixSlot() {
        return this.#___private_hasPrefixSlot_28;
    }
    set hasPrefixSlot(_v) {
        this.#___private_hasPrefixSlot_28 = _v;
    }
    #___private_activeRangeEnd_29 = (()=>{
        const _value = _init_activeRangeEnd(this, 'start');
        _init_extra__init_activeRangeEnd(this);
        return _value;
    })();
    get activeRangeEnd() {
        return this.#___private_activeRangeEnd_29;
    }
    set activeRangeEnd(_v) {
        this.#___private_activeRangeEnd_29 = _v;
    }
    #___private_inputValue_30 = (()=>{
        const _value = _init_inputValue(this, '');
        _init_extra__init_inputValue(this);
        return _value;
    })();
    get inputValue() {
        return this.#___private_inputValue_30;
    }
    set inputValue(_v) {
        this.#___private_inputValue_30 = _v;
    }
    hasInitializedDefaultValue = false;
    visibilityReason = 'api';
    repositionScheduler = createRafScheduler();
    positionRequestId = 0;
    cleanupAutoUpdate = null;
    listboxId = '';
    descriptionId = '';
    constructor(){
        super();
        this.listboxId = `ve-time-picker-listbox-${Math.random().toString(36).slice(2)}`;
        this.descriptionId = `ve-time-picker-description-${Math.random().toString(36).slice(2)}`;
    }
    addEventListener(type, listener, options) {
        super.addEventListener(type, listener, options);
    }
    removeEventListener(type, listener, options) {
        super.removeEventListener(type, listener, options);
    }
    connectedCallback() {
        super.connectedCallback();
        this.syncInitialAttributes();
    }
    handleDocumentPointerDown = (event)=>{
        if (!this.open || event.composedPath().includes(this)) return;
        this.setOpen(false, 'outside');
    };
    handleDocumentKeydown = (event)=>{
        if (!this.open || 'Escape' !== event.key) return;
        event.stopPropagation();
        this.setOpen(false, 'escape');
        this.focus();
    };
    get normalizedFormat() {
        return isSupportedFormat(this.format) ? this.format : 'HH:mm';
    }
    get shouldShowSecondColumn() {
        return 'HH:mm:ss' === this.normalizedFormat;
    }
    get normalizedValue() {
        return this.value ?? '';
    }
    get parsedValue() {
        return parseTimeValue(this.primaryValue);
    }
    get parsedRangeValue() {
        const [start = '', end = ''] = this.rangeValues;
        return [
            parseTimeValue(start),
            parseTimeValue(end)
        ];
    }
    get minTimeValue() {
        const value = parseTimeValue(this.min);
        return value ? getTimeValue(value) : void 0;
    }
    get maxTimeValue() {
        const value = parseTimeValue(this.max);
        return value ? getTimeValue(value) : void 0;
    }
    get rangeStartValue() {
        return this.rangeValues[0] ?? '';
    }
    get rangeEndValue() {
        return this.rangeValues[1] ?? '';
    }
    get primaryValue() {
        return this.range ? this.rangeStartValue : this.normalizedValue;
    }
    get rangeValues() {
        return this.normalizedValue.split(',').map((part)=>part.trim()).filter((part)=>part.length > 0).slice(0, 2);
    }
    getSelectedParts(rangeEnd) {
        if (!this.range || !rangeEnd) return this.parsedValue ?? {
            hour: 0,
            minute: 0,
            second: 0
        };
        const [start, end] = this.parsedRangeValue;
        return ('start' === rangeEnd ? start : end) ?? {
            hour: 0,
            minute: 0,
            second: 0
        };
    }
    get displayValue() {
        const parsedValue = this.parsedValue;
        return parsedValue ? formatTimeValue(parsedValue, this.normalizedFormat) : '';
    }
    get displayRangeStartValue() {
        const parsedValue = parseTimeValue(this.rangeStartValue);
        return parsedValue ? formatTimeValue(parsedValue, this.normalizedFormat) : '';
    }
    get displayRangeEndValue() {
        const parsedValue = parseTimeValue(this.rangeEndValue);
        return parsedValue ? formatTimeValue(parsedValue, this.normalizedFormat) : '';
    }
    get rangeStartPlaceholder() {
        const placeholder = (this.placeholder ?? '').trim();
        return placeholder && 'Select time' !== placeholder ? placeholder : '开始时间';
    }
    get rangeEndPlaceholder() {
        return '结束时间';
    }
    get displayInputValue() {
        return this.inputValue;
    }
    get shouldShowClear() {
        return this.allowClear && !this.disabled && !this.loading && '' !== this.value;
    }
    get hasLeadingLabel() {
        return (this.label ?? '').trim().length > 0;
    }
    get hasDescriptionFeedback() {
        return (this.description ?? '').trim().length > 0;
    }
    get controlAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get controlAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get controlAriaDescribedby() {
        const describedBy = this.getAttribute('aria-describedby');
        const ids = [
            describedBy,
            this.hasDescriptionFeedback ? this.descriptionId : void 0
        ].filter((id)=>'string' == typeof id && id.length > 0);
        return ids.length > 0 ? ids.join(' ') : void 0;
    }
    get controlAriaRequired() {
        return normalizeAriaBooleanString(this.getAttribute('aria-required'));
    }
    get controlAriaInvalid() {
        return normalizeAriaBooleanString(this.getAttribute('aria-invalid')) ?? ('error' === this.status ? 'true' : void 0);
    }
    focus(options) {
        if (this.range) return void ('start' === this.activeRangeEnd ? this.rangeStartElement : this.rangeEndElement)?.focus(options);
        this.inputElement?.focus(options);
    }
    blur() {
        this.inputElement?.blur();
        this.rangeStartElement?.blur();
        this.rangeEndElement?.blur();
    }
    show() {
        this.setOpen(true, 'api');
    }
    hide() {
        this.setOpen(false, 'api');
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.applyDefaultValueIfNeeded();
        this.applyDimensionProperties();
        if (this.open) this.installGlobalListeners();
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('format')) {
            const normalizedFormat = this.normalizedFormat;
            if (this.format !== normalizedFormat) this.format = normalizedFormat;
        }
        if (changedProperties.has('value') || changedProperties.has('format') || changedProperties.has('range')) {
            const normalizedValue = this.normalizeValue(this.normalizedValue);
            if (this.value !== normalizedValue) this.value = normalizedValue;
            if (!this.range) this.inputValue = normalizedValue;
        }
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('disabled') && this.disabled && this.open) return void this.setOpen(false, 'api');
        if (changedProperties.has('open')) this.handleOpenChanged();
        if (changedProperties.has('width')) this.applyDimensionProperties();
        if (this.open) {
            this.scheduleUpdatePosition();
            if (changedProperties.has('value') || changedProperties.has('format') || changedProperties.has('min') || changedProperties.has('max')) this.updateComplete.then(()=>{
                if (!this.isConnected || !this.open) return;
                this.scrollSelectedOptionsIntoView();
            });
        }
    }
    applyDimensionProperties() {
        this.applyCssSizeProperty('--ve-time-picker-current-width', this.width);
    }
    applyCssSizeProperty(name, value) {
        const normalizedValue = normalizeCssSize(value);
        if (normalizedValue) return void this.style.setProperty(name, normalizedValue);
        this.style.removeProperty(name);
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    disconnectedCallback() {
        this.removeGlobalListeners();
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        this.teardownAutoUpdate();
        super.disconnectedCallback();
    }
    syncInitialAttributes() {
        const attrValue = this.getAttribute('value');
        const attrDefaultValue = this.getAttribute('default-value');
        if (!this.value && attrValue) this.value = attrValue;
        if (!this.defaultValue && attrDefaultValue) this.defaultValue = attrDefaultValue;
        if (!this.range && this.hasAttribute('range')) this.range = true;
    }
    applyDefaultValueIfNeeded() {
        if (this.hasInitializedDefaultValue) return;
        this.hasInitializedDefaultValue = true;
        if (this.value || !this.defaultValue) return;
        this.value = this.normalizeValue(this.defaultValue);
    }
    normalizeValue(value) {
        const normalizedValue = value ?? '';
        if (this.range) return normalizedValue.split(',').map((part)=>this.normalizeSingleValue(part)).filter((part)=>part.length > 0).slice(0, 2).join(',');
        return this.normalizeSingleValue(normalizedValue);
    }
    normalizeSingleValue(value) {
        const parsedValue = parseTimeValue(value);
        return parsedValue ? formatTimeValue(parsedValue, this.normalizedFormat) : '';
    }
    isTimePartsInRange(parts) {
        const timeValue = getTimeValue(parts);
        const minTimeValue = this.minTimeValue ?? 0;
        const maxTimeValue = this.maxTimeValue ?? 86399;
        return minTimeValue <= maxTimeValue && timeValue >= minTimeValue && timeValue <= maxTimeValue;
    }
    createOptions(unit, rangeEnd) {
        return this.createUnitValues(unit).map((value)=>({
                value,
                label: toTwoDigit(value),
                disabled: this.isUnitValueDisabled(unit, value, rangeEnd)
            }));
    }
    createUnitValues(unit) {
        const max = 'hour' === unit ? 23 : 59;
        const step = clampStep('hour' === unit ? this.hourStep : 'minute' === unit ? this.minuteStep : this.secondStep);
        const values = [];
        for(let value = 0; value <= max; value += step)values.push(value);
        return values;
    }
    isUnitValueDisabled(unit, value, rangeEnd) {
        return void 0 === this.findSelectableParts(unit, value, rangeEnd);
    }
    findSelectableParts(unit, value, rangeEnd) {
        const minTimeValue = this.minTimeValue ?? 0;
        const maxTimeValue = this.maxTimeValue ?? 86399;
        if (minTimeValue > maxTimeValue) return;
        const selected = this.range && rangeEnd ? 'start' === rangeEnd ? this.parsedRangeValue[0] : this.parsedRangeValue[1] : this.parsedValue;
        const hours = this.createCandidateValues('hour', unit, value, selected?.hour);
        const minutes = this.createCandidateValues('minute', unit, value, selected?.minute);
        const seconds = this.shouldShowSecondColumn ? this.createCandidateValues('second', unit, value, selected?.second) : [
            0
        ];
        for (const hour of hours)for (const minute of minutes)for (const second of seconds){
            const parts = {
                hour,
                minute,
                second
            };
            const timeValue = getTimeValue(parts);
            if (timeValue >= minTimeValue && timeValue <= maxTimeValue) return parts;
        }
    }
    createCandidateValues(candidateUnit, selectedUnit, selectedValue, currentValue) {
        if (candidateUnit === selectedUnit) return [
            selectedValue
        ];
        const values = this.createUnitValues(candidateUnit);
        if (void 0 === currentValue) return values;
        return [
            currentValue,
            ...values.filter((item)=>item !== currentValue)
        ];
    }
    selectUnitValue(unit, value, rangeEnd) {
        if (this.disabled || this.loading) return;
        const nextParts = this.findSelectableParts(unit, value, rangeEnd);
        if (!nextParts) return;
        if (this.range) return void this.selectRangeUnitValue(unit, nextParts, rangeEnd ?? this.activeRangeEnd);
        const nextValue = formatTimeValue(nextParts, this.normalizedFormat);
        this.value = nextValue;
        this.emitChange(nextParts, 'select', false);
        if ('second' === unit || !this.shouldShowSecondColumn && 'minute' === unit) this.setOpen(false, 'select');
    }
    selectRangeUnitValue(unit, nextParts, rangeEnd) {
        const values = [
            ...this.rangeValues
        ];
        const nextValue = formatTimeValue(nextParts, this.normalizedFormat);
        const index = 'start' === rangeEnd ? 0 : 1;
        values[index] = nextValue;
        const isFinalUnit = 'second' === unit || !this.shouldShowSecondColumn && 'minute' === unit;
        if (!isFinalUnit) {
            this.value = values.filter((item)=>Boolean(item)).slice(0, 2).join(',');
            return;
        }
        const [start, end] = values.map((item)=>parseTimeValue(item ?? ''));
        const sortedValues = [
            ...values
        ];
        if (start && end && getTimeValue(start) > getTimeValue(end)) sortedValues.reverse();
        this.value = sortedValues.filter((item)=>Boolean(item)).slice(0, 2).join(',');
        if (start && end) this.emitChange(this.parsedRangeValue, 'select', true);
        if ('start' === rangeEnd) {
            this.activeRangeEnd = 'end';
            return;
        }
        this.activeRangeEnd = 'start';
        this.setOpen(false, 'select');
    }
    commitInputValue() {
        if (this.range) return;
        const parsedValue = parseTimeValue(this.inputValue);
        if (!parsedValue || !this.isTimePartsInRange(parsedValue)) {
            this.inputValue = this.displayValue;
            return;
        }
        const nextValue = formatTimeValue(parsedValue, this.normalizedFormat);
        if (this.value !== nextValue) {
            this.value = nextValue;
            this.emitChange(parsedValue, 'input', false);
        }
        this.inputValue = nextValue;
    }
    clearValue(event) {
        event?.preventDefault();
        event?.stopPropagation();
        if (this.disabled || this.loading) return;
        this.value = '';
        this.inputValue = '';
        this.activeRangeEnd = 'start';
        this.dispatchEvent(new CustomEvent('ve-clear', {
            detail: {
                value: ''
            },
            bubbles: true,
            composed: true
        }));
    }
    handleClearPointerDown = (event)=>{
        event.preventDefault();
        event.stopPropagation();
    };
    emitChange(parts, source, range) {
        const primaryParts = Array.isArray(parts) ? parts[0] : parts;
        if (!primaryParts) return;
        const value = Array.isArray(parts) ? parts.map((part)=>part ? formatTimeValue(part, this.normalizedFormat) : '').filter((part)=>part.length > 0).join(',') : formatTimeValue(parts, this.normalizedFormat);
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail: {
                value,
                hour: primaryParts.hour,
                minute: primaryParts.minute,
                second: this.shouldShowSecondColumn ? primaryParts.second : void 0,
                time: Array.isArray(parts) ? parts.filter((part)=>void 0 !== part) : parts,
                range,
                source
            },
            bubbles: true,
            composed: true
        }));
    }
    setOpen(open, reason) {
        if (this.disabled || this.loading) open = false;
        if (this.open === open) return;
        this.visibilityReason = reason;
        this.open = open;
    }
    handleOpenChanged() {
        if (this.open) {
            this.installGlobalListeners();
            this.updateComplete.then(()=>{
                if (!this.isConnected || !this.open || this.disabled || this.loading) return;
                this.setupAutoUpdate();
                this.scheduleUpdatePosition();
                this.scrollSelectedOptionsIntoView();
            });
        } else {
            this.removeGlobalListeners();
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            this.teardownAutoUpdate();
        }
        this.dispatchEvent(new CustomEvent('ve-visible-change', {
            detail: {
                open: this.open,
                reason: this.visibilityReason
            },
            bubbles: true,
            composed: true
        }));
    }
    installGlobalListeners() {
        document.addEventListener('pointerdown', this.handleDocumentPointerDown);
        document.addEventListener('keydown', this.handleDocumentKeydown);
    }
    removeGlobalListeners() {
        document.removeEventListener('pointerdown', this.handleDocumentPointerDown);
        document.removeEventListener('keydown', this.handleDocumentKeydown);
    }
    setupAutoUpdate() {
        if (!this.isConnected || !this.open || !this.controlElement || !this.panelElement) return;
        this.teardownAutoUpdate();
        this.cleanupAutoUpdate = createFloatingAutoUpdate(this.fieldElement ?? this.controlElement, this.panelElement, ()=>this.scheduleUpdatePosition());
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    async updatePosition() {
        if (!this.open || !this.isConnected || !this.controlElement || !this.panelElement) return;
        const requestId = ++this.positionRequestId;
        const referenceElement = this.fieldElement ?? this.controlElement;
        if (this.popupMatchSelectWidth) {
            const rect = referenceElement.getBoundingClientRect();
            this.panelElement.style.minWidth = `${rect.width}px`;
        } else this.panelElement.style.minWidth = '';
        const styles = window.getComputedStyle(this);
        const gapValue = Number.parseFloat(styles.getPropertyValue('--ve-time-picker-panel-gap')) || 4;
        const result = await computeFloatingPanelPosition({
            reference: referenceElement,
            floating: this.panelElement,
            placement: 'bottom-start',
            distance: gapValue
        });
        if (requestId !== this.positionRequestId || !this.open || !this.isConnected || !this.panelElement) return;
        this.panelPlacement = result.placement;
        Object.assign(this.panelElement.style, {
            left: '0',
            top: '0',
            transform: `translate3d(${Math.round(result.x)}px, ${Math.round(result.y)}px, 0)`
        });
    }
    scrollSelectedOptionsIntoView() {
        if (!this.panelElement) return;
        this.panelElement.querySelectorAll('.ve-time-picker-option-selected').forEach((option)=>{
            const column = option.closest('.ve-time-picker-column');
            if (!column) return;
            const centeredScrollTop = option.offsetTop - (column.clientHeight - option.offsetHeight) / 2;
            column.scrollTo({
                top: centeredScrollTop
            });
        });
    }
    handleControlClick = ()=>{
        if (this.disabled) return;
        this.setOpen(true, 'trigger');
    };
    handleControlKeydown = (event)=>{
        if (this.disabled) return;
        switch(event.key){
            case 'Enter':
                if (!this.range && event.target === this.inputElement) {
                    event.preventDefault();
                    this.commitInputValue();
                }
                break;
            case 'ArrowDown':
                event.preventDefault();
                this.setOpen(true, 'trigger');
                break;
            case 'Escape':
                if (this.open) {
                    event.preventDefault();
                    this.setOpen(false, 'escape');
                }
                break;
            default:
                break;
        }
    };
    handleInputFocus = ()=>{
        if (!this.disabled) this.setOpen(true, 'trigger');
    };
    handleInput = (event)=>{
        this.inputValue = event.target.value;
    };
    handleInputBlur = ()=>{
        this.commitInputValue();
    };
    handleRangeSegmentClick = (event)=>{
        if (this.disabled) return;
        this.activeRangeEnd = event.currentTarget.dataset.rangeEnd;
        this.setOpen(true, 'trigger');
    };
    handleLabelClick = ()=>{
        if (this.disabled) return;
        this.focus();
        this.setOpen(true, 'trigger');
    };
    handlePrefixSlotChange = (event)=>{
        const slot = event.currentTarget;
        this.hasPrefixSlot = (slot?.assignedNodes({
            flatten: true
        }).filter((node)=>node.nodeType !== Node.TEXT_NODE || (node.textContent?.trim().length ?? 0) > 0).length ?? 0) > 0;
    };
    handleClearClick = (event)=>{
        this.clearValue(event);
    };
    createOptionClickHandler(unit, value, rangeEnd) {
        return ()=>this.selectUnitValue(unit, value, rangeEnd);
    }
    createOptionKeydownHandler(unit, value, rangeEnd) {
        return (event)=>{
            if ('Enter' !== event.key && ' ' !== event.key) return;
            event.preventDefault();
            this.selectUnitValue(unit, value, rangeEnd);
        };
    }
    renderLabel() {
        const label = (this.label ?? '').trim();
        if (!label) return nothing;
        return html`
      <span
        class="ve-time-picker-label"
        part="label"
        @click=${this.handleLabelClick}
      >
        <span class="ve-time-picker-label-text">${label}</span>
      </span>
    `;
    }
    renderPrefix() {
        return html`<span
      class=${classMap({
            've-time-picker-prefix': true,
            've-time-picker-prefix-empty': !this.hasPrefixSlot
        })}
      ><slot name="prefix" @slotchange=${this.handlePrefixSlotChange}></slot
    ></span>`;
    }
    renderValue() {
        if (this.range) {
            const startValue = this.displayRangeStartValue;
            const endValue = this.displayRangeEndValue;
            return html`<span class="ve-time-picker-range-value">
        <button
          class=${classMap({
                've-time-picker-range-segment': true,
                've-time-picker-range-segment-start': true,
                've-time-picker-range-segment-active': 'start' === this.activeRangeEnd,
                've-time-picker-range-segment-placeholder': !startValue
            })}
          type="button"
          data-range-end="start"
          ?disabled=${this.disabled}
          @click=${this.handleRangeSegmentClick}
        >
          ${startValue || this.rangeStartPlaceholder}
        </button>
        <span class="ve-time-picker-range-separator" aria-hidden="true">-</span>
        <button
          class=${classMap({
                've-time-picker-range-segment': true,
                've-time-picker-range-segment-end': true,
                've-time-picker-range-segment-active': 'end' === this.activeRangeEnd,
                've-time-picker-range-segment-placeholder': !endValue
            })}
          type="button"
          data-range-end="end"
          ?disabled=${this.disabled}
          @click=${this.handleRangeSegmentClick}
        >
          ${endValue || this.rangeEndPlaceholder}
        </button>
      </span>`;
        }
        return html`<input
      class="ve-time-picker-input"
      .value=${this.displayInputValue}
      placeholder=${this.placeholder}
      ?disabled=${this.disabled}
      @focus=${this.handleInputFocus}
      @input=${this.handleInput}
      @blur=${this.handleInputBlur}
    />`;
    }
    renderActions() {
        return html`
      <span class="ve-time-picker-actions">
        ${this.shouldShowClear ? html`<button
              class="ve-time-picker-clear"
              part="clear"
              type="button"
              aria-label="Clear"
              @pointerdown=${this.handleClearPointerDown}
              @click=${this.handleClearClick}
            >
              <ve-icon .definition=${closeIcon}></ve-icon>
            </button>` : nothing}
        ${this.loading ? html`<span class="ve-time-picker-loading" aria-hidden="true">
              <ve-icon .definition=${loadingIcon}></ve-icon>
            </span>` : html`<span class="ve-time-picker-arrow" aria-hidden="true">
              <ve-icon .definition=${timeClockIcon}></ve-icon>
            </span>`}
      </span>
    `;
    }
    renderColumn(unit, options, rangeEnd) {
        const selectedValue = this.getSelectedParts(rangeEnd)[unit];
        const optionIdPrefix = rangeEnd ? `${this.listboxId}-${rangeEnd}-${unit}` : `${this.listboxId}-${unit}`;
        return html`
      <div class="ve-time-picker-column" role="listbox" aria-label=${unit}>
        ${options.map((option)=>{
            const selected = option.value === selectedValue;
            const optionId = `${optionIdPrefix}-${option.value}`;
            return html`<button
            id=${optionId}
            class=${classMap({
                've-time-picker-option': true,
                've-time-picker-option-selected': selected,
                've-time-picker-option-disabled': option.disabled
            })}
            part="option"
            type="button"
            role="option"
            aria-selected=${selected ? 'true' : 'false'}
            aria-disabled=${option.disabled ? 'true' : 'false'}
            ?disabled=${option.disabled}
            @click=${this.createOptionClickHandler(unit, option.value, rangeEnd)}
            @keydown=${this.createOptionKeydownHandler(unit, option.value, rangeEnd)}
          >
            ${option.label}
          </button>`;
        })}
      </div>
    `;
    }
    renderRangePanelColumns() {
        const rangeEnd = this.activeRangeEnd;
        return html`
      <div class="ve-time-picker-range-section">
        <div class="ve-time-picker-range-columns">
          ${this.renderColumn('hour', this.createOptions('hour', rangeEnd), rangeEnd)}
          ${this.renderColumn('minute', this.createOptions('minute', rangeEnd), rangeEnd)}
          ${this.shouldShowSecondColumn ? this.renderColumn('second', this.createOptions('second', rangeEnd), rangeEnd) : nothing}
        </div>
      </div>
    `;
    }
    renderPanel() {
        if (!this.open) return nothing;
        return html`
      <div
        id=${this.listboxId}
        class="ve-time-picker-panel"
        part="panel"
        popover=${"manual"}
        role="group"
        data-placement=${this.panelPlacement}
      >
        ${this.range ? this.renderRangePanelColumns() : html`
              ${this.renderColumn('hour', this.createOptions('hour'))}
              ${this.renderColumn('minute', this.createOptions('minute'))}
              ${this.shouldShowSecondColumn ? this.renderColumn('second', this.createOptions('second')) : nothing}
            `}
      </div>
    `;
    }
    renderDescription() {
        const description = (this.description ?? '').trim();
        if (!description) return nothing;
        return html`<div
      id=${this.descriptionId}
      class="ve-time-picker-description"
      part="description"
    >
      ${description}
    </div>`;
    }
    render() {
        return html`
      <div
        class=${classMap({
            've-time-picker': true,
            've-time-picker-with-label': this.hasLeadingLabel
        })}
      >
        <div class="ve-time-picker-field">
          ${this.renderLabel()}
          <div
            class="ve-time-picker-control"
            part="control"
            role="combobox"
            aria-haspopup="dialog"
            aria-expanded=${this.open ? 'true' : 'false'}
            aria-controls=${ifDefined(this.open ? this.listboxId : void 0)}
            aria-label=${ifDefined(this.controlAriaLabel)}
            aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
            aria-describedby=${ifDefined(this.controlAriaDescribedby)}
            aria-required=${ifDefined(this.controlAriaRequired)}
            aria-invalid=${ifDefined(this.controlAriaInvalid)}
            @click=${this.handleControlClick}
            @keydown=${this.handleControlKeydown}
          >
            ${this.renderPrefix()} ${this.renderValue()} ${this.renderActions()}
          </div>
        </div>
        ${this.renderDescription()}
      </div>
      ${this.renderPanel()}
    `;
    }
}
defineCustomElement(VE_TIME_PICKER_TAG, VeTimePickerElement);
export { VE_TIME_PICKER_TAG, VeTimePickerElement };
