// @ve-design/web/web/dist/ve-tabs.js
import { css, html } from "lit";
import { property, queryAssignedElements, state } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { repeat } from "lit/directives/repeat.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _init_key, _init_extra__init_key, _init_disabled, _init_extra__init_disabled, _init_tabSlotElements, _init_extra__init_tabSlotElements;
const VE_TAB_PANE_TAG = 've-tab-pane';
const VE_TAB_PANE_CHANGE_EVENT = 've-tab-pane-change';
_dec = property({
    type: String,
    reflect: true
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = queryAssignedElements({
    slot: 'tab',
    flatten: true
});
class VeTabPaneElement extends VeElement {
    static{
        ({ e: [_init_key, _init_extra__init_key, _init_disabled, _init_extra__init_disabled, _init_tabSlotElements, _init_extra__init_tabSlotElements] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "key"
            ],
            [
                _dec1,
                1,
                "disabled"
            ],
            [
                _dec2,
                1,
                "tabSlotElements"
            ]
        ]));
    }
    static styles = [
        componentReset,
        css`
      :host {
        display: block;
      }

      :host([hidden]) {
        display: none;
      }
    `
    ];
    #___private_key_1 = (()=>{
        const _value = _init_key(this, '');
        _init_extra__init_key(this);
        return _value;
    })();
    get key() {
        return this.#___private_key_1;
    }
    set key(_v) {
        this.#___private_key_1 = _v;
    }
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_disabled_2 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_2;
    }
    set disabled(_v) {
        this.#___private_disabled_2 = _v;
    }
    #___private_tabSlotElements_3 = (()=>{
        const _value = _init_tabSlotElements(this);
        _init_extra__init_tabSlotElements(this);
        return _value;
    })();
    get tabSlotElements() {
        return this.#___private_tabSlotElements_3;
    }
    set tabSlotElements(_v) {
        this.#___private_tabSlotElements_3 = _v;
    }
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    get panelId() {
        return `ve-tabs-panel-${this.key}`;
    }
    get tabId() {
        return `ve-tabs-tab-${this.key}`;
    }
    getTabNodes() {
        return this.tabSlotElements;
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('key') || changedProperties.has('title') || changedProperties.has('disabled')) this.dispatchPaneChange();
    }
    dispatchPaneChange() {
        this.dispatchEvent(new CustomEvent(VE_TAB_PANE_CHANGE_EVENT, {
            bubbles: true,
            composed: true
        }));
    }
    render() {
        return html`
      <div hidden aria-hidden="true">
        <slot name="tab" @slotchange=${this.dispatchPaneChange}></slot>
      </div>
      <slot @slotchange=${this.dispatchPaneChange}></slot>
    `;
    }
}
function isSelectablePane(pane) {
    return Boolean(pane.key) && !pane.disabled;
}
function resolveTabsState({ panes, activeTab, defaultActiveTab, innerActiveKey, focusKey }) {
    const selectablePanes = panes.filter(isSelectablePane);
    if (0 === selectablePanes.length) return {
        resolvedActiveKey: '',
        nextInnerActiveKey: '',
        nextFocusKey: ''
    };
    const isControlled = 'string' == typeof activeTab;
    const currentKey = isControlled ? activeTab ?? '' : innerActiveKey;
    const currentPane = selectablePanes.find((pane)=>pane.key === currentKey);
    const defaultPane = defaultActiveTab ? selectablePanes.find((pane)=>pane.key === defaultActiveTab) : void 0;
    const fallbackPane = selectablePanes[0];
    const resolvedActiveKey = currentPane?.key ?? defaultPane?.key ?? fallbackPane?.key ?? '';
    const nextInnerActiveKey = isControlled ? innerActiveKey : resolvedActiveKey;
    const nextFocusKey = selectablePanes.some((pane)=>pane.key === focusKey) ? focusKey : resolvedActiveKey;
    return {
        resolvedActiveKey,
        nextInnerActiveKey,
        nextFocusKey
    };
}
const veTabsTokenDefaults = css`
  :host {
    --ve-tabs-text-color: var(--color-text-tertiary);
    --ve-tabs-font-family: var(--font-sans);
    --ve-tabs-nav-gap: var(--space-m);
    --ve-tabs-nav-padding-inline: 0px;
    --ve-tabs-nav-padding-block: 0px;
    --ve-tabs-nav-border-color: var(--color-border-default);
    --ve-tabs-nav-border-width: var(--stroke-weight-base);
    --ve-tabs-nav-background: transparent;
    --ve-tabs-nav-radius: 0px;
    --ve-tabs-layout-gap-inline: var(--space-m);
    --ve-tabs-content-padding-block: var(--space-s);
    --ve-tabs-content-padding-inline: 0px;
    --ve-tabs-pane-radius: 0px;
    --ve-tabs-pane-text-color: var(--color-text-primary);
    --ve-tabs-pane-font-size: var(--text-body);
    --ve-tabs-panels-background: transparent;

    --ve-tabs-font-weight: var(--font-weight-normal);
    --ve-tabs-font-weight-active: var(--font-weight-medium);
    --ve-tabs-focus-ring-width: var(--stroke-2);
    --ve-tabs-disabled-opacity: 1;
    --ve-tabs-transition-duration: var(--motion-base);
    --ve-tabs-transition-easing: var(--ease-standard);

    --ve-tabs-height-small: 28px;
    --ve-tabs-height-default: 32px;
    --ve-tabs-height-large: 36px;

    --ve-tabs-pipe-inner-height-small: 24px;
    --ve-tabs-pipe-inner-height-default: 28px;
    --ve-tabs-pipe-inner-height-large: 32px;

    --ve-tabs-padding-inline-small: var(--space-xxs);
    --ve-tabs-padding-inline-default: var(--space-xs);
    --ve-tabs-padding-inline-large: var(--space-xs);

    --ve-tabs-font-size-small: var(--text-body-sm);
    --ve-tabs-font-size-default: var(--text-body);
    --ve-tabs-font-size-large: var(--text-body);

    --ve-tabs-radius-line: 0px;
    --ve-tabs-radius-text: 0px;
    --ve-tabs-radius-rounded: var(--radius-lg);
    --ve-tabs-radius-capsule: var(--radius-lg);
    --ve-tabs-radius-capsule-small: 10px;

    --ve-tabs-indicator-thickness: var(--stroke-weight-bold);
    --ve-tabs-indicator-color: var(--color-border-interactive-active);

    --ve-tabs-item-text: var(--color-text-tertiary);
    --ve-tabs-item-text-hover: var(--color-text-secondary);
    --ve-tabs-item-text-active: var(--color-text-primary);
    --ve-tabs-item-text-disabled: var(--color-text-disable);

    --ve-tabs-item-background: transparent;
    --ve-tabs-item-background-hover: transparent;
    --ve-tabs-item-background-active: transparent;

    --ve-tabs-focus-ring-color: color-mix(
      var(--ve-tabs-current-item-text-active) 20%,
      transparent
    );
    --ve-tabs-active-inner-border-width: var(--stroke-weight-base);
    --ve-tabs-active-inner-border-color: transparent;

    --ve-tabs-item-padding-block-line-vertical: var(--space-xxs);
    --ve-tabs-item-padding-inline-line-vertical-start: 0px;
    --ve-tabs-item-padding-inline-line-vertical-end: var(--space-xxs);

    --ve-tabs-current-height: var(--ve-tabs-height-default);
    --ve-tabs-current-padding-inline: var(--ve-tabs-padding-inline-default);
    --ve-tabs-current-font-size: var(--ve-tabs-font-size-default);
    --ve-tabs-current-radius: var(--ve-tabs-radius-line);
    --ve-tabs-current-nav-gap: var(--ve-tabs-nav-gap);
    --ve-tabs-current-nav-padding-inline: var(--ve-tabs-nav-padding-inline);
    --ve-tabs-current-nav-padding-block: var(--ve-tabs-nav-padding-block);
    --ve-tabs-current-nav-background: var(--ve-tabs-nav-background);
    --ve-tabs-current-nav-radius: var(--ve-tabs-nav-radius);
    --ve-tabs-current-nav-border-width: var(--ve-tabs-nav-border-width);
    --ve-tabs-current-item-background: var(--ve-tabs-item-background);
    --ve-tabs-current-item-background-hover: var(
      --ve-tabs-item-background-hover
    );
    --ve-tabs-current-item-background-active: var(
      --ve-tabs-item-background-active
    );
    --ve-tabs-current-item-text: var(--ve-tabs-item-text);
    --ve-tabs-current-item-text-hover: var(--ve-tabs-item-text-hover);
    --ve-tabs-current-item-text-active: var(--ve-tabs-item-text-active);
    --ve-tabs-current-item-text-disabled: var(--ve-tabs-item-text-disabled);
    --ve-tabs-current-active-inner-border-color: var(
      --ve-tabs-active-inner-border-color
    );
  }

  :host([size='small']) {
    --ve-tabs-current-height: var(--ve-tabs-height-small);
    --ve-tabs-current-padding-inline: var(--ve-tabs-padding-inline-small);
    --ve-tabs-current-font-size: var(--ve-tabs-font-size-small);
  }

  :host([size='large']) {
    --ve-tabs-current-height: var(--ve-tabs-height-large);
    --ve-tabs-current-padding-inline: var(--ve-tabs-padding-inline-large);
    --ve-tabs-current-font-size: var(--ve-tabs-font-size-large);
  }

  :host([type='text']) {
    --ve-tabs-current-nav-gap: var(--space-l);
    --ve-tabs-current-radius: var(--ve-tabs-radius-text);
    --ve-tabs-current-nav-border-width: 0px;
  }

  :host([type='rounded']) {
    --ve-tabs-current-nav-gap: var(--space-xxxs);
    --ve-tabs-current-radius: var(--ve-tabs-radius-rounded);
    --ve-tabs-current-nav-border-width: 0px;
    --ve-tabs-current-item-background-hover: var(--color-bg-surface);
    --ve-tabs-current-item-background-active: var(--color-bg-secondary);
    --ve-tabs-current-item-text-hover: var(--color-text-tertiary);
  }

  :host([type='capsule']) {
    --ve-tabs-current-nav-gap: 2px;
    --ve-tabs-current-nav-padding-inline: 2px;
    --ve-tabs-current-nav-padding-block: 2px;
    --ve-tabs-current-nav-background: var(--color-bg-surface);
    --ve-tabs-current-nav-radius: var(--radius-lg);
    --ve-tabs-current-nav-border-width: 0px;
    --ve-tabs-current-radius: var(--ve-tabs-radius-capsule);
    --ve-tabs-current-height: var(--ve-tabs-pipe-inner-height-default);
    --ve-tabs-current-item-background-active: var(--color-bg-overlay);
    --ve-tabs-current-active-inner-border-color: var(--color-bg-muted);
  }

  :host([type='capsule'][size='small']) {
    --ve-tabs-current-height: var(--ve-tabs-pipe-inner-height-small);
    --ve-tabs-current-radius: var(--ve-tabs-radius-capsule-small);
    --ve-tabs-current-nav-radius: var(--ve-tabs-radius-capsule-small);
  }

  :host([type='capsule'][size='large']) {
    --ve-tabs-current-height: var(--ve-tabs-pipe-inner-height-large);
  }

  :host([tab-position='left']),
  :host([tab-position='right']) {
    --ve-tabs-nav-gap: var(--space-xxs);
    --ve-tabs-content-padding-inline: var(--space-s);
    --ve-tabs-content-padding-block: 0px;
  }
`;
const token = veTabsTokenDefaults;
const ve_tabs = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      color: var(--ve-tabs-text-color);
    }

    :host([hidden]) {
      display: none;
    }

    .ve-tabs {
      display: flex;
      flex-direction: column;
      gap: 0;
      min-inline-size: 0;
      min-block-size: 0;
    }

    :host([tab-position='bottom']) .ve-tabs {
      flex-direction: column-reverse;
    }

    :host([tab-position='left']) .ve-tabs {
      flex-direction: row;
      align-items: stretch;
      gap: var(--ve-tabs-layout-gap-inline);
    }

    :host([tab-position='right']) .ve-tabs {
      flex-direction: row-reverse;
      align-items: stretch;
      gap: var(--ve-tabs-layout-gap-inline);
    }

    .ve-tabs-nav {
      position: relative;
      display: inline-flex;
      align-items: center;
      gap: var(--ve-tabs-current-nav-gap);
      min-inline-size: 0;
      inline-size: fit-content;
      max-inline-size: 100%;
      align-self: flex-start;
      padding-inline: var(--ve-tabs-current-nav-padding-inline);
      padding-block: var(--ve-tabs-current-nav-padding-block);
      border-radius: var(--ve-tabs-current-nav-radius);
      background: var(--ve-tabs-current-nav-background);
      box-sizing: border-box;
      overflow-x: auto;
      overflow-y: hidden;
      scrollbar-width: none;
    }

    .ve-tabs-nav::-webkit-scrollbar {
      display: none;
    }

    :host(:not([type]):not([tab-position])) .ve-tabs-nav,
    :host(:not([type])[tab-position='top']) .ve-tabs-nav,
    :host([type='line']:not([tab-position])) .ve-tabs-nav,
    :host([type='line'][tab-position='top']) .ve-tabs-nav {
      inline-size: 100%;
      align-self: stretch;
      border-block-end: var(--ve-tabs-current-nav-border-width) solid
        var(--ve-tabs-nav-border-color);
    }

    :host(:not([type])[tab-position='bottom']) .ve-tabs-nav,
    :host([type='line'][tab-position='bottom']) .ve-tabs-nav {
      inline-size: 100%;
      align-self: stretch;
      border-block-start: var(--ve-tabs-current-nav-border-width) solid
        var(--ve-tabs-nav-border-color);
    }

    :host([tab-position='left']) .ve-tabs-nav,
    :host([tab-position='right']) .ve-tabs-nav {
      flex-direction: column;
      align-items: stretch;
      inline-size: max-content;
      min-inline-size: 0;
      flex: none;
    }

    :host(:not([type])[tab-position='left']) .ve-tabs-nav,
    :host([type='line'][tab-position='left']) .ve-tabs-nav {
      border-inline-end: var(--ve-tabs-current-nav-border-width) solid
        var(--ve-tabs-nav-border-color);
    }

    :host(:not([type])[tab-position='right']) .ve-tabs-nav,
    :host([type='line'][tab-position='right']) .ve-tabs-nav {
      border-inline-start: var(--ve-tabs-current-nav-border-width) solid
        var(--ve-tabs-nav-border-color);
    }

    .ve-tabs-tab {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      min-inline-size: 0;
      max-inline-size: 200px;
      min-block-size: var(--ve-tabs-current-height);
      padding-inline: var(--ve-tabs-current-padding-inline);
      padding-block: 0;
      border: none;
      border-radius: var(--ve-tabs-current-radius);
      background: var(--ve-tabs-current-item-background);
      color: var(--ve-tabs-current-item-text);
      font-family: var(--ve-tabs-font-family);
      font-size: var(--ve-tabs-current-font-size);
      font-weight: var(--ve-tabs-font-weight);
      line-height: var(--line-height-body);
      box-shadow: none;
      cursor: pointer;
      user-select: none;
      white-space: nowrap;
      overflow: hidden;
      transition:
        background-color var(--ve-tabs-transition-duration)
          var(--ve-tabs-transition-easing),
        color var(--ve-tabs-transition-duration)
          var(--ve-tabs-transition-easing),
        opacity var(--ve-tabs-transition-duration)
          var(--ve-tabs-transition-easing),
        box-shadow var(--ve-tabs-transition-duration)
          var(--ve-tabs-transition-easing);
    }

    :host(:not([type])) .ve-tabs-tab,
    :host([type='line']) .ve-tabs-tab {
      min-block-size: var(--ve-tabs-current-height);
      padding-inline: 0;
      border-radius: 0;
      background: transparent;
    }

    :host(:not([type]):not([tab-position])) .ve-tabs-tab-content::after,
    :host(:not([type])[tab-position='top']) .ve-tabs-tab-content::after,
    :host([type='line']:not([tab-position])) .ve-tabs-tab-content::after,
    :host([type='line'][tab-position='top']) .ve-tabs-tab-content::after,
    :host(:not([type])[tab-position='bottom']) .ve-tabs-tab-content::after,
    :host([type='line'][tab-position='bottom']) .ve-tabs-tab-content::after {
      content: '';
      position: absolute;
      pointer-events: none;
      background: var(--ve-tabs-indicator-color);
      opacity: 0;
      border-radius: var(--ve-tabs-indicator-thickness);
      transition: opacity var(--ve-tabs-transition-duration)
        var(--ve-tabs-transition-easing);
    }

    :host(:not([type]):not([tab-position])) .ve-tabs-tab,
    :host(:not([type])[tab-position='top']) .ve-tabs-tab,
    :host([type='line']:not([tab-position])) .ve-tabs-tab,
    :host([type='line'][tab-position='top']) .ve-tabs-tab {
      margin-block-end: calc(var(--ve-tabs-current-nav-border-width) * -1);
    }

    :host(:not([type]):not([tab-position])) .ve-tabs-tab-content::after,
    :host(:not([type])[tab-position='top']) .ve-tabs-tab-content::after,
    :host([type='line']:not([tab-position])) .ve-tabs-tab-content::after,
    :host([type='line'][tab-position='top']) .ve-tabs-tab-content::after {
      inset-inline: 0;
      inset-block-end: 0;
      block-size: var(--ve-tabs-indicator-thickness);
    }

    :host(:not([type])[tab-position='bottom']) .ve-tabs-tab,
    :host([type='line'][tab-position='bottom']) .ve-tabs-tab {
      margin-block-start: calc(var(--ve-tabs-current-nav-border-width) * -1);
    }

    :host(:not([type])[tab-position='bottom']) .ve-tabs-tab-content::after,
    :host([type='line'][tab-position='bottom']) .ve-tabs-tab-content::after {
      inset-inline: 0;
      inset-block-start: 0;
      block-size: var(--ve-tabs-indicator-thickness);
    }

    :host(:not([type])[tab-position='left']) .ve-tabs-tab,
    :host(:not([type])[tab-position='right']) .ve-tabs-tab,
    :host([type='line'][tab-position='left']) .ve-tabs-tab,
    :host([type='line'][tab-position='right']) .ve-tabs-tab {
      inline-size: 100%;
      min-block-size: auto;
      padding-block: var(--ve-tabs-item-padding-block-line-vertical);
      justify-content: flex-start;
    }

    :host(:not([type])[tab-position='left']) .ve-tabs-tab,
    :host([type='line'][tab-position='left']) .ve-tabs-tab {
      padding-inline: var(--ve-tabs-item-padding-inline-line-vertical-start)
        var(--ve-tabs-item-padding-inline-line-vertical-end);
      margin-inline-end: calc(var(--ve-tabs-current-nav-border-width) * -1);
    }

    :host(:not([type])[tab-position='left']) .ve-tabs-tab::after,
    :host([type='line'][tab-position='left']) .ve-tabs-tab::after,
    :host(:not([type])[tab-position='right']) .ve-tabs-tab::after,
    :host([type='line'][tab-position='right']) .ve-tabs-tab::after {
      content: '';
      position: absolute;
      pointer-events: none;
      background: var(--ve-tabs-indicator-color);
      opacity: 0;
      border-radius: var(--ve-tabs-indicator-thickness);
      transition: opacity var(--ve-tabs-transition-duration)
        var(--ve-tabs-transition-easing);
    }

    :host(:not([type])[tab-position='left']) .ve-tabs-tab::after,
    :host([type='line'][tab-position='left']) .ve-tabs-tab::after {
      inset-block-start: 0;
      inset-block-end: 0;
      inset-inline-end: 0;
      inline-size: var(--ve-tabs-indicator-thickness);
    }

    :host(:not([type])[tab-position='right']) .ve-tabs-tab,
    :host([type='line'][tab-position='right']) .ve-tabs-tab {
      padding-inline: var(--ve-tabs-item-padding-inline-line-vertical-end)
        var(--ve-tabs-item-padding-inline-line-vertical-start);
      margin-inline-start: calc(var(--ve-tabs-current-nav-border-width) * -1);
    }

    :host(:not([type])[tab-position='right']) .ve-tabs-tab::after,
    :host([type='line'][tab-position='right']) .ve-tabs-tab::after {
      inset-block: 0;
      inset-inline-start: 0;
      inline-size: var(--ve-tabs-indicator-thickness);
    }

    :host([type='text']) .ve-tabs-nav {
      border: none;
    }

    :host([type='text']) .ve-tabs-tab {
      padding-inline: 0;
    }

    :host([type='rounded']) .ve-tabs-nav {
      border: none;
    }

    :host([type='capsule']) .ve-tabs-nav {
      border: none;
    }

    :host([tab-position='left']) .ve-tabs-tab,
    :host([tab-position='right']) .ve-tabs-tab {
      justify-content: flex-start;
    }

    .ve-tabs-tab:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--ve-tabs-focus-ring-width)
        var(--ve-tabs-focus-ring-color);
    }

    .ve-tabs-tab-selected {
      background: var(--ve-tabs-current-item-background-active);
      color: var(--ve-tabs-current-item-text-active);
      box-shadow: inset 0 0 0 var(--ve-tabs-active-inner-border-width)
        var(--ve-tabs-current-active-inner-border-color);
      font-weight: var(--ve-tabs-font-weight-active);
    }

    .ve-tabs-tab-selected .ve-tabs-tab-content {
      font-weight: var(--ve-tabs-font-weight-active);
    }

    :host(:not([type])) .ve-tabs-tab-selected::after,
    :host([type='line']) .ve-tabs-tab-selected::after {
      opacity: 1;
    }

    :host(:not([type]))
      .ve-tabs-tab-selected
      .ve-tabs-tab-content::after,
    :host([type='line']) .ve-tabs-tab-selected .ve-tabs-tab-content::after {
      opacity: 1;
    }

    :host(:not([type])) .ve-tabs-tab-selected,
    :host([type='line']) .ve-tabs-tab-selected {
      background: transparent;
      color: var(--ve-tabs-current-item-text-active);
      z-index: 1;
    }

    .ve-tabs-tab-disabled {
      opacity: var(--ve-tabs-disabled-opacity);
      color: var(--ve-tabs-current-item-text-disabled);
      cursor: not-allowed;
      pointer-events: none;
    }

    .ve-tabs-tab-disabled.ve-tabs-tab-selected {
      background: var(--ve-tabs-current-item-background);
      color: var(--ve-tabs-current-item-text-disabled);
      box-shadow: none;
      font-weight: var(--ve-tabs-font-weight);
    }

    :host(:not([type])) .ve-tabs-tab-disabled.ve-tabs-tab-selected::after,
    :host([type='line']) .ve-tabs-tab-disabled.ve-tabs-tab-selected::after {
      opacity: 0;
    }

    :host(:not([type]))
      .ve-tabs-tab-disabled.ve-tabs-tab-selected
      .ve-tabs-tab-content::after,
    :host([type='line'])
      .ve-tabs-tab-disabled.ve-tabs-tab-selected
      .ve-tabs-tab-content::after {
      opacity: 0;
    }

    .ve-tabs-tab:not(.ve-tabs-tab-disabled):hover {
      background: var(--ve-tabs-current-item-background-hover);
      color: var(--ve-tabs-current-item-text-hover);
    }

    .ve-tabs-tab-selected:not(.ve-tabs-tab-disabled):hover {
      background: var(--ve-tabs-current-item-background-active);
      color: var(--ve-tabs-current-item-text-active);
    }

    .ve-tabs-tab:not(.ve-tabs-tab-disabled):active {
      color: var(--ve-tabs-current-item-text-active);
    }

    .ve-tabs-tab-content {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: var(--space-xxxs);
      min-inline-size: 0;
      max-inline-size: 200px;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      color: inherit;
      font-weight: inherit;
    }

    :host(:not([type])) .ve-tabs-tab-content,
    :host([type='line']) .ve-tabs-tab-content {
      flex: 1 1 auto;
      padding-inline: 0;
    }

    :host(:not([type]):not([tab-position])) .ve-tabs-tab-content,
    :host(:not([type])[tab-position='top']) .ve-tabs-tab-content,
    :host([type='line']:not([tab-position])) .ve-tabs-tab-content,
    :host([type='line'][tab-position='top']) .ve-tabs-tab-content,
    :host(:not([type])[tab-position='bottom']) .ve-tabs-tab-content,
    :host([type='line'][tab-position='bottom']) .ve-tabs-tab-content {
      min-block-size: var(--ve-tabs-current-height);
    }

    .ve-tabs-panels {
      min-inline-size: 0;
      min-block-size: 0;
      flex: 1 1 auto;
      background: var(--ve-tabs-panels-background);
    }

    :host(:not([tab-position])) .ve-tabs-panels,
    :host([tab-position='top']) .ve-tabs-panels,
    :host([tab-position='bottom']) .ve-tabs-panels {
      padding-inline: 0;
    }

    ::slotted(ve-tab-pane) {
      box-sizing: border-box;
      padding-block: var(--ve-tabs-content-padding-block);
      padding-inline: var(--ve-tabs-content-padding-inline);
      border-radius: var(--ve-tabs-pane-radius);
      color: var(--ve-tabs-pane-text-color);
      font-size: var(--ve-tabs-pane-font-size);
    }
  `
];
function ve_tabs_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_tabs_element_toPropertyKey(arg) {
    var key = ve_tabs_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_tabs_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_tabs_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_tabs_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_tabs_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_tabs_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_tabs_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_tabs_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_tabs_element_dec, ve_tabs_element_dec1, ve_tabs_element_dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _init_activeTab, _init_extra__init_activeTab, _init_defaultActiveTab, _init_extra__init_defaultActiveTab, _init_size, _init_extra__init_size, _init_type, _init_extra__init_type, _init_tabPosition, _init_extra__init_tabPosition, _init_paneElements, _init_extra__init_paneElements, _init_innerActiveKey, _init_extra__init_innerActiveKey, _init_focusKey, _init_extra__init_focusKey, _init_resolvedActiveKey, _init_extra__init_resolvedActiveKey;
const VE_TABS_TAG = 've-tabs';
ve_tabs_element_dec = property({
    type: String,
    attribute: 'active-tab'
}), ve_tabs_element_dec1 = property({
    type: String,
    attribute: 'default-active-tab'
}), ve_tabs_element_dec2 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec3 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec4 = property({
    type: String,
    attribute: 'tab-position',
    reflect: true,
    useDefault: true
}), _dec5 = queryAssignedElements({
    flatten: true,
    selector: VE_TAB_PANE_TAG
}), _dec6 = state(), _dec7 = state(), _dec8 = state();
class VeTabsElement extends VeElement {
    static{
        ({ e: [_init_activeTab, _init_extra__init_activeTab, _init_defaultActiveTab, _init_extra__init_defaultActiveTab, _init_size, _init_extra__init_size, _init_type, _init_extra__init_type, _init_tabPosition, _init_extra__init_tabPosition, _init_paneElements, _init_extra__init_paneElements, _init_innerActiveKey, _init_extra__init_innerActiveKey, _init_focusKey, _init_extra__init_focusKey, _init_resolvedActiveKey, _init_extra__init_resolvedActiveKey] } = ve_tabs_element_apply_decs_2311(this, [], [
            [
                ve_tabs_element_dec,
                1,
                "activeTab"
            ],
            [
                ve_tabs_element_dec1,
                1,
                "defaultActiveTab"
            ],
            [
                ve_tabs_element_dec2,
                1,
                "size"
            ],
            [
                _dec3,
                1,
                "type"
            ],
            [
                _dec4,
                1,
                "tabPosition"
            ],
            [
                _dec5,
                1,
                "paneElements"
            ],
            [
                _dec6,
                1,
                "innerActiveKey"
            ],
            [
                _dec7,
                1,
                "focusKey"
            ],
            [
                _dec8,
                1,
                "resolvedActiveKey"
            ]
        ]));
    }
    static styles = ve_tabs;
    static dependencies = {
        [VE_TAB_PANE_TAG]: VeTabPaneElement
    };
    #___private_activeTab_1 = (()=>{
        const _value = _init_activeTab(this, void 0);
        _init_extra__init_activeTab(this);
        return _value;
    })();
    get activeTab() {
        return this.#___private_activeTab_1;
    }
    set activeTab(_v) {
        this.#___private_activeTab_1 = _v;
    }
    #___private_defaultActiveTab_2 = (()=>{
        const _value = _init_defaultActiveTab(this, '');
        _init_extra__init_defaultActiveTab(this);
        return _value;
    })();
    get defaultActiveTab() {
        return this.#___private_defaultActiveTab_2;
    }
    set defaultActiveTab(_v) {
        this.#___private_defaultActiveTab_2 = _v;
    }
    #___private_size_3 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_3;
    }
    set size(_v) {
        this.#___private_size_3 = _v;
    }
    #___private_type_4 = (()=>{
        const _value = _init_type(this, 'line');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_4;
    }
    set type(_v) {
        this.#___private_type_4 = _v;
    }
    #___private_tabPosition_5 = (()=>{
        const _value = _init_tabPosition(this, 'top');
        _init_extra__init_tabPosition(this);
        return _value;
    })();
    get tabPosition() {
        return this.#___private_tabPosition_5;
    }
    set tabPosition(_v) {
        this.#___private_tabPosition_5 = _v;
    }
    #___private_paneElements_6 = (()=>{
        const _value = _init_paneElements(this);
        _init_extra__init_paneElements(this);
        return _value;
    })();
    get paneElements() {
        return this.#___private_paneElements_6;
    }
    set paneElements(_v) {
        this.#___private_paneElements_6 = _v;
    }
    #___private_innerActiveKey_7 = (()=>{
        const _value = _init_innerActiveKey(this, '');
        _init_extra__init_innerActiveKey(this);
        return _value;
    })();
    get innerActiveKey() {
        return this.#___private_innerActiveKey_7;
    }
    set innerActiveKey(_v) {
        this.#___private_innerActiveKey_7 = _v;
    }
    #___private_focusKey_8 = (()=>{
        const _value = _init_focusKey(this, '');
        _init_extra__init_focusKey(this);
        return _value;
    })();
    get focusKey() {
        return this.#___private_focusKey_8;
    }
    set focusKey(_v) {
        this.#___private_focusKey_8 = _v;
    }
    #___private_resolvedActiveKey_9 = (()=>{
        const _value = _init_resolvedActiveKey(this, '');
        _init_extra__init_resolvedActiveKey(this);
        return _value;
    })();
    get resolvedActiveKey() {
        return this.#___private_resolvedActiveKey_9;
    }
    set resolvedActiveKey(_v) {
        this.#___private_resolvedActiveKey_9 = _v;
    }
    paneCache = [];
    enabledPaneCache = [];
    get panes() {
        return this.paneCache;
    }
    get enabledPanes() {
        return this.enabledPaneCache;
    }
    get isControlled() {
        return 'string' == typeof this.activeTab;
    }
    addEventListener(type, listener, options) {
        super.addEventListener(type, listener, options);
    }
    removeEventListener(type, listener, options) {
        super.removeEventListener(type, listener, options);
    }
    connectedCallback() {
        super.connectedCallback();
        this.innerActiveKey = this.defaultActiveTab;
        this.addEventListener(VE_TAB_PANE_CHANGE_EVENT, this.handlePaneChange);
    }
    disconnectedCallback() {
        this.removeEventListener(VE_TAB_PANE_CHANGE_EVENT, this.handlePaneChange);
        super.disconnectedCallback();
    }
    firstUpdated() {
        this.syncFromPaneStructure(false);
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('activeTab') || changedProperties.has('defaultActiveTab') || changedProperties.has('innerActiveKey') || changedProperties.has('resolvedActiveKey')) {
            this.syncActiveState();
            this.syncPanePresentation();
        }
    }
    handleSlotChange() {
        this.syncFromPaneStructure();
    }
    handlePaneChange = (event)=>{
        if (!(event.target instanceof VeTabPaneElement)) return;
        this.syncFromPaneStructure();
    };
    syncFromPaneStructure(requestRender = true) {
        this.refreshPaneCache();
        this.warnInvalidPanes();
        this.syncActiveState();
        this.syncPanePresentation();
        if (requestRender) this.requestUpdate();
    }
    refreshPaneCache() {
        this.paneCache = [
            ...this.paneElements
        ];
        this.enabledPaneCache = this.paneCache.filter((pane)=>Boolean(pane.key) && !pane.disabled);
    }
    warnInvalidPanes() {
        const paneKeys = new Set();
        for (const pane of this.panes){
            if (!pane.key) {
                console.warn(`[${VE_TABS_TAG}] Each [${pane.tagName.toLowerCase()}] requires a non-empty key attribute.`, pane);
                continue;
            }
            if (paneKeys.has(pane.key)) {
                console.warn(`[${VE_TABS_TAG}] Duplicate pane key "${pane.key}" detected. Keys must be unique.`, pane);
                continue;
            }
            paneKeys.add(pane.key);
        }
    }
    syncActiveState() {
        const nextState = resolveTabsState({
            panes: this.panes,
            activeTab: this.activeTab,
            defaultActiveTab: this.defaultActiveTab,
            innerActiveKey: this.innerActiveKey,
            focusKey: this.focusKey
        });
        if (this.resolvedActiveKey !== nextState.resolvedActiveKey) this.resolvedActiveKey = nextState.resolvedActiveKey;
        if (this.innerActiveKey !== nextState.nextInnerActiveKey) this.innerActiveKey = nextState.nextInnerActiveKey;
        if (this.focusKey !== nextState.nextFocusKey) this.focusKey = nextState.nextFocusKey;
    }
    syncPanePresentation() {
        const activeKey = this.resolvedActiveKey;
        for (const pane of this.panes){
            const selected = pane.key === activeKey && !pane.disabled;
            pane.id = pane.panelId;
            pane.hidden = !selected;
            pane.setAttribute('role', 'tabpanel');
            pane.setAttribute('aria-labelledby', pane.tabId);
            pane.setAttribute('aria-hidden', selected ? 'false' : 'true');
            pane.setAttribute('data-active', selected ? 'true' : 'false');
            pane.setAttribute('part', 'panel');
        }
    }
    setActiveKey(nextKey, emit = false) {
        if (!nextKey || this.resolvedActiveKey === nextKey) return;
        if (!this.isControlled) {
            this.innerActiveKey = nextKey;
            this.resolvedActiveKey = nextKey;
        }
        this.focusKey = nextKey;
        if (emit) this.dispatchEvent(new CustomEvent('ve-change', {
            detail: {
                activeTab: nextKey
            },
            bubbles: true,
            composed: true
        }));
    }
    focusTabByKey(key) {
        this.updateComplete.then(()=>{
            const button = this.renderRoot.querySelector(`[data-key="${CSS.escape(key)}"]`);
            button?.focus();
        });
    }
    moveFocus(currentKey, step) {
        const enabledPanes = this.enabledPanes;
        if (0 === enabledPanes.length) return;
        const currentIndex = Math.max(0, enabledPanes.findIndex((pane)=>pane.key === currentKey));
        const nextIndex = (currentIndex + step + enabledPanes.length) % enabledPanes.length;
        const nextPane = enabledPanes[nextIndex];
        if (!nextPane) return;
        this.focusKey = nextPane.key;
        this.focusTabByKey(nextPane.key);
    }
    handleTabClick(pane) {
        if (pane.disabled) return;
        this.setActiveKey(pane.key, true);
    }
    handleTabKeydown(event, pane) {
        const isHorizontal = 'top' === this.tabPosition || 'bottom' === this.tabPosition;
        switch(event.key){
            case 'ArrowRight':
                if (isHorizontal) {
                    event.preventDefault();
                    this.moveFocus(pane.key, 1);
                }
                break;
            case 'ArrowLeft':
                if (isHorizontal) {
                    event.preventDefault();
                    this.moveFocus(pane.key, -1);
                }
                break;
            case 'ArrowDown':
                if (!isHorizontal) {
                    event.preventDefault();
                    this.moveFocus(pane.key, 1);
                }
                break;
            case 'ArrowUp':
                if (!isHorizontal) {
                    event.preventDefault();
                    this.moveFocus(pane.key, -1);
                }
                break;
            case 'Home':
                {
                    event.preventDefault();
                    const firstPane = this.enabledPanes[0];
                    if (!firstPane) return;
                    this.focusKey = firstPane.key;
                    this.focusTabByKey(firstPane.key);
                    break;
                }
            case 'End':
                {
                    event.preventDefault();
                    const enabledPanes = this.enabledPanes;
                    const lastPane = enabledPanes[enabledPanes.length - 1];
                    if (!lastPane) return;
                    this.focusKey = lastPane.key;
                    this.focusTabByKey(lastPane.key);
                    break;
                }
            case 'Enter':
            case ' ':
                event.preventDefault();
                this.handleTabClick(pane);
                break;
            default:
                break;
        }
    }
    cloneTabNode(node) {
        const clonedNode = node.cloneNode(true);
        if (clonedNode instanceof Element) clonedNode.removeAttribute('slot');
        return clonedNode;
    }
    renderTabLabel(pane) {
        const tabNodes = pane.getTabNodes();
        if (0 === tabNodes.length) return pane.title;
        return repeat(tabNodes, (_node, index)=>`${pane.key}-${index}`, (node)=>this.cloneTabNode(node));
    }
    renderTabs() {
        return repeat(this.panes, (pane)=>pane.key, (pane)=>{
            const selected = this.resolvedActiveKey === pane.key;
            const focused = this.focusKey ? this.focusKey === pane.key : selected;
            return html`
          <button
            class=${classMap({
                've-tabs-tab': true,
                've-tabs-tab-selected': selected,
                've-tabs-tab-disabled': pane.disabled
            })}
            part="tab"
            id=${pane.tabId}
            type="button"
            role="tab"
            data-key=${pane.key}
            aria-selected=${selected ? 'true' : 'false'}
            aria-controls=${pane.panelId}
            aria-disabled=${pane.disabled ? 'true' : 'false'}
            tabindex=${focused && !pane.disabled ? '0' : '-1'}
            @click=${()=>this.handleTabClick(pane)}
            @keydown=${(event)=>this.handleTabKeydown(event, pane)}
          >
            <span class="ve-tabs-tab-content" part="tab-content"
              >${this.renderTabLabel(pane)}</span
            >
          </button>
        `;
        });
    }
    renderPanels() {
        return html` <slot @slotchange=${this.handleSlotChange}></slot> `;
    }
    render() {
        return html`
      <div class="ve-tabs">
        <div
          class="ve-tabs-nav"
          part="nav"
          role="tablist"
          aria-orientation=${'left' === this.tabPosition || 'right' === this.tabPosition ? 'vertical' : 'horizontal'}
        >
          ${this.renderTabs()}
        </div>
        <div class="ve-tabs-panels" part="panels">${this.renderPanels()}</div>
      </div>
    `;
    }
}
defineCustomElement(VE_TABS_TAG, VeTabsElement);
defineCustomElement(VE_TAB_PANE_TAG, VeTabPaneElement);
export { VE_TABS_TAG, VE_TAB_PANE_TAG, VeTabPaneElement, VeTabsElement };
