// @ve-design/web/web/dist/ve-thought-chain.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { styleMap } from "lit/directives/style-map.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { checkCircleIcon } from "./icons/check-circle.js";
import { xCircleIcon } from "./icons/x-circle.js";
import { loadingIcon } from "./42938.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
const veThoughtChainTokenDefaults = css`
  :host {
    --ve-thought-chain-text-color: var(--color-text-secondary);
    --ve-thought-chain-title-text-color: var(--color-text-tertiary);
    --ve-thought-chain-content-text-color: var(--color-text-secondary);
    --ve-thought-chain-tertiary-text-color: var(--color-text-tertiary);
    --ve-thought-chain-border-color: var(--color-border-default);
    --ve-thought-chain-border-strong-color: var(--color-border-strong);
    --ve-thought-chain-surface-color: var(--color-bg-base);
    --ve-thought-chain-disabled-opacity: var(--state-disabled-ratio);

    --ve-thought-chain-gap: var(--space-xxs);
    --ve-thought-chain-header-gap: var(--space-xxs);
    --ve-thought-chain-list-gap: var(--ve-thought-chain-gap);
    --ve-thought-chain-item-gap: var(--space-xxs);
    --ve-thought-chain-item-content-gap: var(--space-xxs);
    --ve-thought-chain-item-icon-box-size: var(--space-s);
    --ve-thought-chain-item-icon-align-padding: calc(var(--space-xxxs) - 1px);
    --ve-thought-chain-title-icon-gap: var(--space-xxxs);
    --ve-thought-chain-text-line-height: calc(
      var(--text-body) + var(--space-xxs)
    );
    --ve-thought-chain-toggle-size: var(--space-s);
    --ve-thought-chain-connector-width: var(--stroke-weight-base);
    --ve-thought-chain-icon-stroke-width: var(--stroke-weight-strong);
    --ve-thought-chain-status-icon-stroke-width: var(
      --ve-thought-chain-icon-stroke-width
    );
    --ve-thought-chain-connector-offset: calc(
      var(--ve-thought-chain-item-icon-box-size) / 2
    );
    --ve-thought-chain-motion-duration: var(--motion-base);
    --ve-thought-chain-motion-duration-fast: var(--motion-fast);
    --ve-thought-chain-collapse-shift: var(--space-xxs);
    --ve-thought-chain-scroll-fade-size: calc(
      var(--space-xxl) + var(--space-xs) - var(--stroke-weight-strong)
    );
    --ve-thought-chain-scrollbar-size: var(--space-xxxs);
    --ve-thought-chain-scrollbar-color: var(--color-border-strong);
    --ve-thought-chain-shimmer-base-color: var(--color-text-tertiary);
    --ve-thought-chain-shimmer-highlight-color: var(--color-bg-muted);
    --ve-thought-chain-shimmer-duration: 1800ms;

    --ve-thought-chain-status-default-color: var(--color-text-tertiary);
    --ve-thought-chain-status-error-color: var(--color-text-danger);

    --ve-thought-chain-current-status-color: var(
      --ve-thought-chain-status-default-color
    );
  }
`;
const token = veThoughtChainTokenDefaults;
function toAriaBoolean(value) {
    return value ? 'true' : 'false';
}
function hasMeaningfulAssignedNodes(slot) {
    const assignedNodes = slot?.assignedNodes({
        flatten: true
    }) ?? [];
    return assignedNodes.some((node)=>{
        if (node.nodeType === Node.TEXT_NODE) return (node.textContent?.trim().length ?? 0) > 0;
        return node.nodeType === Node.ELEMENT_NODE;
    });
}
function renderChevronIcon() {
    return html`<svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
    <path
      d="m19 9-7 7-7-7"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      style="stroke-width: var(--ve-thought-chain-icon-stroke-width);"
    ></path>
  </svg>`;
}
const collapseTriggerResetStyles = css`
  .ve-thought-chain-trigger-reset {
    appearance: none;
    border: 0;
    background: transparent;
    color: inherit;
    padding: 0;
    font: inherit;
    text-align: left;
    width: 100%;
    min-width: 0;
  }

  .ve-thought-chain-trigger-reset:focus-visible {
    outline: var(--stroke-weight-bold) solid
      var(--ve-thought-chain-border-strong-color);
    outline-offset: var(--space-xxxs);
    border-radius: var(--radius-md);
  }

  .ve-thought-chain-trigger-enabled {
    cursor: pointer;
  }
`;
const collapseMotionStyles = css`
  .ve-thought-chain-motion {
    display: grid;
    grid-template-rows: 1fr;
    min-width: 0;
    overflow: hidden;
    opacity: 1;
    transform: translateY(0);
    transition:
      grid-template-rows var(--ve-thought-chain-motion-duration) ease,
      opacity var(--ve-thought-chain-motion-duration-fast) ease,
      transform var(--ve-thought-chain-motion-duration) ease;
  }

  .ve-thought-chain-motion-collapsed {
    grid-template-rows: 0fr;
    opacity: 0;
    transform: translateY(calc(var(--ve-thought-chain-collapse-shift) * -1));
  }

  .ve-thought-chain-motion[hidden] {
    display: none;
  }

  @media (prefers-reduced-motion: reduce) {
    .ve-thought-chain-motion {
      transition: none;
      transform: none;
    }
  }
`;
const chevronStyles = css`
  .ve-thought-chain-chevron {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    flex: none;
    width: var(--ve-thought-chain-toggle-size);
    height: var(--ve-thought-chain-text-line-height);
    color: var(--ve-thought-chain-tertiary-text-color);
    transition: transform var(--ve-thought-chain-motion-duration-fast) ease;
  }

  .ve-thought-chain-chevron svg,
  .ve-thought-chain-chevron ve-icon {
    width: var(--ve-thought-chain-toggle-size);
    height: var(--ve-thought-chain-toggle-size);
    --ve-icon-size: var(--ve-thought-chain-toggle-size);
  }

  .ve-thought-chain-chevron-collapsed {
    transform: rotate(-90deg);
  }

  @media (prefers-reduced-motion: reduce) {
    .ve-thought-chain-chevron {
      transition: none;
    }
  }
`;
const ve_thought_chain_item = [
    componentReset,
    token,
    collapseTriggerResetStyles,
    collapseMotionStyles,
    chevronStyles,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      width: 100%;
      font-family: var(--font-sans);
      color: var(--ve-thought-chain-text-color);
    }

    :host([hidden]) {
      display: none;
    }

    :host([status='default']) {
      --ve-thought-chain-current-status-color: var(
        --ve-thought-chain-status-default-color
      );
    }

    :host([status='error']) {
      --ve-thought-chain-current-status-color: var(
        --ve-thought-chain-status-error-color
      );
    }

    .ve-thought-chain-item {
      position: relative;
      display: grid;
      grid-template-columns: var(--ve-thought-chain-item-icon-box-size) minmax(
          0,
          1fr
        );
      column-gap: var(--ve-thought-chain-item-gap);
      min-width: 0;
      box-sizing: border-box;
    }

    .ve-thought-chain-item-disabled {
      opacity: var(--ve-thought-chain-disabled-opacity);
    }

    .ve-thought-chain-item-connector {
      position: absolute;
      top: calc(
        var(--ve-thought-chain-item-icon-box-size) +
          var(--ve-thought-chain-item-icon-align-padding) * 2
      );
      bottom: 0;
      left: calc(
        var(--ve-thought-chain-connector-offset) - var(
            --ve-thought-chain-connector-width
          ) /
          2
      );
      width: var(--ve-thought-chain-connector-width);
      min-height: var(--space-xxs);
      background: var(--ve-thought-chain-border-color);
      pointer-events: none;
    }

    :host(:not(:last-of-type)) .ve-thought-chain-item-connector {
      bottom: calc(var(--ve-thought-chain-list-gap) * -1);
    }

    :host([chain-line='dashed']) .ve-thought-chain-item-connector {
      width: 0;
      border-left: var(--ve-thought-chain-connector-width) dashed
        var(--ve-thought-chain-border-color);
      background: transparent;
    }

    :host([chain-line='dotted']) .ve-thought-chain-item-connector {
      width: 0;
      border-left: var(--ve-thought-chain-connector-width) dotted
        var(--ve-thought-chain-border-color);
      background: transparent;
    }

    :host([chain-line='none']) .ve-thought-chain-item-connector {
      display: none;
    }

    .ve-thought-chain-item-icon {
      position: relative;
      z-index: 1;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      width: var(--ve-thought-chain-item-icon-box-size);
      height: var(--ve-thought-chain-item-icon-box-size);
      margin-block-start: var(--ve-thought-chain-item-icon-align-padding);
      color: var(--ve-thought-chain-current-status-color);
      background: var(--ve-thought-chain-surface-color);
      border-radius: var(--radius-full);
    }

    .ve-thought-chain-item-icon ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-width: 100%;
      max-height: 100%;
      line-height: 1;
      flex: none;
    }

    .ve-thought-chain-item-icon svg,
    .ve-thought-chain-item-icon ve-icon,
    .ve-thought-chain-item-icon ::slotted(svg),
    .ve-thought-chain-item-icon ::slotted(ve-icon) {
      width: var(--ve-thought-chain-item-icon-box-size);
      height: var(--ve-thought-chain-item-icon-box-size);
      --ve-icon-size: var(--ve-thought-chain-item-icon-box-size);
      flex: none;
    }

    :host([status='loading']) .ve-thought-chain-item-icon svg,
    :host([status='loading']) .ve-thought-chain-item-icon ve-icon {
      animation: ve-thought-chain-spin 1.6s linear infinite;
    }

    .ve-thought-chain-item-main {
      display: grid;
      row-gap: var(--ve-thought-chain-item-content-gap);
      min-width: 0;
      transition: row-gap var(--ve-thought-chain-motion-duration) ease;
    }

    .ve-thought-chain-item-collapsed .ve-thought-chain-item-main {
      row-gap: 0;
    }

    .ve-thought-chain-item-header {
      display: flex;
      align-items: flex-start;
      gap: var(--space-xxs);
      min-width: 0;
    }

    .ve-thought-chain-item-toggle {
      display: inline-flex;
      align-items: flex-start;
      gap: 0;
      width: auto;
      max-width: 100%;
    }

    .ve-thought-chain-item-toggle-enabled:hover .ve-thought-chain-item-title {
      color: var(--ve-thought-chain-text-color);
    }

    .ve-thought-chain-item-heading {
      display: block;
      min-width: 0;
    }

    .ve-thought-chain-item-title {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-thought-chain-title-icon-gap);
      min-width: 0;
      color: var(--ve-thought-chain-title-text-color);
      font-size: var(--text-body);
      font-weight: var(--font-weight-normal);
      line-height: var(--ve-thought-chain-text-line-height);
    }

    :host([status='loading']) .ve-thought-chain-item-title,
    :host([status='loading']) .ve-thought-chain-item-title ::slotted(*) {
      background-image: linear-gradient(
        90deg,
        var(--ve-thought-chain-shimmer-base-color) 6%,
        var(--ve-thought-chain-shimmer-highlight-color) 39%,
        var(--ve-thought-chain-shimmer-base-color) 72%
      );
      background-size: 220% 100%;
      background-clip: text;
      -webkit-background-clip: text;
      -webkit-text-fill-color: transparent;
      color: transparent;
      animation: ve-thought-chain-shimmer var(--ve-thought-chain-shimmer-duration)
        linear infinite;
    }

    .ve-thought-chain-item-title[hidden] {
      display: none;
    }

    .ve-thought-chain-item-content {
      min-width: 0;
      color: var(--ve-thought-chain-content-text-color);
      font-size: var(--text-body);
      line-height: var(--ve-thought-chain-text-line-height);
    }

    .ve-thought-chain-item-scroll-frame {
      min-width: 0;
      min-height: 0;
    }

    .ve-thought-chain-item-scroll-clip {
      position: relative;
      min-width: 0;
      min-height: 0;
    }

    .ve-thought-chain-item-scroll {
      display: grid;
      min-width: 0;
      min-height: 0;
      scrollbar-width: thin;
      scrollbar-color: var(--ve-thought-chain-scrollbar-color) transparent;
    }

    .ve-thought-chain-item-scroll-limited {
      --ve-thought-chain-scroll-mask-start: black;
      --ve-thought-chain-scroll-mask-end: black;
      overflow-y: auto;
      -webkit-mask-image: linear-gradient(
        to bottom,
        var(--ve-thought-chain-scroll-mask-start) 0,
        black var(--ve-thought-chain-scroll-fade-size),
        black calc(100% - var(--ve-thought-chain-scroll-fade-size)),
        var(--ve-thought-chain-scroll-mask-end) 100%
      );
      mask-image: linear-gradient(
        to bottom,
        var(--ve-thought-chain-scroll-mask-start) 0,
        black var(--ve-thought-chain-scroll-fade-size),
        black calc(100% - var(--ve-thought-chain-scroll-fade-size)),
        var(--ve-thought-chain-scroll-mask-end) 100%
      );
    }

    .ve-thought-chain-item-scroll-mask-start {
      --ve-thought-chain-scroll-mask-start: transparent;
    }

    .ve-thought-chain-item-scroll-mask-end {
      --ve-thought-chain-scroll-mask-end: transparent;
    }

    .ve-thought-chain-item-scroll-collapsible.ve-thought-chain-item-scroll-limited {
      overflow: hidden;
    }

    .ve-thought-chain-item-scroll-expanded {
      max-height: none;
      overflow: visible;
    }

    .ve-thought-chain-item-overflow-toggle {
      appearance: none;
      display: inline-flex;
      align-items: center;
      margin: var(--space-xxs) 0 0;
      padding: 0;
      border: 0;
      background: transparent;
      color: var(--ve-thought-chain-title-text-color);
      font: inherit;
      line-height: var(--ve-thought-chain-text-line-height);
      cursor: pointer;
    }

    .ve-thought-chain-item-overflow-toggle:hover {
      color: var(--color-text-accent);
    }

    .ve-thought-chain-item-scroll-limited::-webkit-scrollbar {
      width: var(--ve-thought-chain-scrollbar-size);
      height: var(--ve-thought-chain-scrollbar-size);
    }

    .ve-thought-chain-item-scroll-limited::-webkit-scrollbar-track {
      background: transparent;
    }

    .ve-thought-chain-item-scroll-limited::-webkit-scrollbar-thumb {
      border-radius: var(--radius-full);
      background: var(--ve-thought-chain-scrollbar-color);
    }


    @keyframes ve-thought-chain-spin {
      to {
        transform: rotate(360deg);
      }
    }

    @keyframes ve-thought-chain-shimmer {
      0% {
        background-position: 100% 0;
      }
      100% {
        background-position: -100% 0;
      }
    }

    :host([dir='rtl']) .ve-thought-chain-item,
    :host-context([dir='rtl']) .ve-thought-chain-item {
      direction: rtl;
      grid-template-columns: minmax(0, 1fr) var(
          --ve-thought-chain-item-icon-box-size
        );
    }

    :host([dir='rtl']) .ve-thought-chain-item-icon,
    :host-context([dir='rtl']) .ve-thought-chain-item-icon {
      grid-column: 2;
      grid-row: 1;
    }

    :host([dir='rtl']) .ve-thought-chain-item-main,
    :host-context([dir='rtl']) .ve-thought-chain-item-main {
      grid-column: 1;
      grid-row: 1;
    }

    :host([dir='rtl']) .ve-thought-chain-item-connector,
    :host-context([dir='rtl']) .ve-thought-chain-item-connector {
      left: auto;
      right: calc(
        var(--ve-thought-chain-connector-offset) - var(
            --ve-thought-chain-connector-width
          ) /
          2
      );
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-thought-chain-item-main {
        transition: none;
      }

      :host([status='loading']) .ve-thought-chain-item-icon svg,
      :host([status='loading']) .ve-thought-chain-item-icon ve-icon,
      :host([status='loading']) .ve-thought-chain-item-title {
        animation: none;
      }

      :host([status='loading']) .ve-thought-chain-item-title,
      :host([status='loading']) .ve-thought-chain-item-title ::slotted(*) {
        background: none;
        -webkit-text-fill-color: var(--ve-thought-chain-title-text-color);
        color: var(--ve-thought-chain-title-text-color);
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _init_key, _init_extra__init_key, _init_status, _init_extra__init_status, _init_collapsible, _init_extra__init_collapsible, _init_defaultExpanded, _init_extra__init_defaultExpanded, _init_disabled, _init_extra__init_disabled, _init_maxHeight, _init_extra__init_maxHeight, _init_overflow, _init_extra__init_overflow, _init_expanded, _init_extra__init_expanded, _init_managedByChain, _init_extra__init_managedByChain, _init_chainLine, _init_extra__init_chainLine, _init_contentSlot, _init_extra__init_contentSlot, _init_titleSlot, _init_extra__init_titleSlot, _init_iconSlot, _init_extra__init_iconSlot, _init_scrollContainer, _init_extra__init_scrollContainer, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasContentSlot, _init_extra__init_hasContentSlot, _init_internalExpanded, _init_extra__init_internalExpanded, _init_chainExpanded, _init_extra__init_chainExpanded, _init_canScrollStart, _init_extra__init_canScrollStart, _init_canScrollEnd, _init_extra__init_canScrollEnd, _init_hasOverflow, _init_extra__init_hasOverflow, _init_overflowExpanded, _init_extra__init_overflowExpanded;
const VE_THOUGHT_CHAIN_ITEM_TAG = 've-thought-chain-item';
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        return null === value ? true : 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
const STATUS_ICONS = {
    success: html`<ve-icon .definition=${checkCircleIcon}></ve-icon>`,
    error: html`<ve-icon .definition=${xCircleIcon}></ve-icon>`,
    loading: html`<ve-icon .definition=${loadingIcon}></ve-icon>`,
    abort: html`<svg
    class="ve-thought-chain-status-icon"
    viewBox="0 0 24 24"
    aria-hidden="true"
    fill="none"
  >
    <path
      d="M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      style="stroke-width: var(--ve-thought-chain-status-icon-stroke-width);"
    ></path>
    <path
      d="M8 12h8"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      style="stroke-width: var(--ve-thought-chain-status-icon-stroke-width);"
    ></path>
  </svg>`,
    default: html`<svg
    class="ve-thought-chain-status-icon"
    viewBox="0 0 24 24"
    aria-hidden="true"
    fill="none"
  >
    <circle
      cx="12"
      cy="12"
      r="9"
      stroke="currentColor"
      style="stroke-width: var(--ve-thought-chain-status-icon-stroke-width);"
    ></circle>
  </svg>`
};
let nextThoughtChainItemId = 0;
_dec = property({
    attribute: 'key',
    reflect: true
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec3 = property({
    attribute: 'default-expanded',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = property({
    attribute: 'max-height',
    reflect: true
}), _dec6 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    attribute: false
}), _dec8 = state(), _dec9 = state(), _dec10 = query('slot:not([name])'), _dec11 = query('slot[name="title"]'), _dec12 = query('slot[name="icon"]'), _dec13 = query('.ve-thought-chain-item-scroll'), _dec14 = state(), _dec15 = state(), _dec16 = state(), _dec17 = state(), _dec18 = state(), _dec19 = state(), _dec20 = state(), _dec21 = state(), _dec22 = state();
class VeThoughtChainItemElement extends VeElement {
    static{
        ({ e: [_init_key, _init_extra__init_key, _init_status, _init_extra__init_status, _init_collapsible, _init_extra__init_collapsible, _init_defaultExpanded, _init_extra__init_defaultExpanded, _init_disabled, _init_extra__init_disabled, _init_maxHeight, _init_extra__init_maxHeight, _init_overflow, _init_extra__init_overflow, _init_expanded, _init_extra__init_expanded, _init_managedByChain, _init_extra__init_managedByChain, _init_chainLine, _init_extra__init_chainLine, _init_contentSlot, _init_extra__init_contentSlot, _init_titleSlot, _init_extra__init_titleSlot, _init_iconSlot, _init_extra__init_iconSlot, _init_scrollContainer, _init_extra__init_scrollContainer, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasContentSlot, _init_extra__init_hasContentSlot, _init_internalExpanded, _init_extra__init_internalExpanded, _init_chainExpanded, _init_extra__init_chainExpanded, _init_canScrollStart, _init_extra__init_canScrollStart, _init_canScrollEnd, _init_extra__init_canScrollEnd, _init_hasOverflow, _init_extra__init_hasOverflow, _init_overflowExpanded, _init_extra__init_overflowExpanded] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "key"
            ],
            [
                _dec1,
                1,
                "status"
            ],
            [
                _dec2,
                1,
                "collapsible"
            ],
            [
                _dec3,
                1,
                "defaultExpanded"
            ],
            [
                _dec4,
                1,
                "disabled"
            ],
            [
                _dec5,
                1,
                "maxHeight"
            ],
            [
                _dec6,
                1,
                "overflow"
            ],
            [
                _dec7,
                1,
                "expanded"
            ],
            [
                _dec8,
                1,
                "managedByChain"
            ],
            [
                _dec9,
                1,
                "chainLine"
            ],
            [
                _dec10,
                1,
                "contentSlot"
            ],
            [
                _dec11,
                1,
                "titleSlot"
            ],
            [
                _dec12,
                1,
                "iconSlot"
            ],
            [
                _dec13,
                1,
                "scrollContainer"
            ],
            [
                _dec14,
                1,
                "hasTitleSlot"
            ],
            [
                _dec15,
                1,
                "hasIconSlot"
            ],
            [
                _dec16,
                1,
                "hasContentSlot"
            ],
            [
                _dec17,
                1,
                "internalExpanded"
            ],
            [
                _dec18,
                1,
                "chainExpanded"
            ],
            [
                _dec19,
                1,
                "canScrollStart"
            ],
            [
                _dec20,
                1,
                "canScrollEnd"
            ],
            [
                _dec21,
                1,
                "hasOverflow"
            ],
            [
                _dec22,
                1,
                "overflowExpanded"
            ]
        ]));
    }
    static styles = ve_thought_chain_item;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_key_1 = (()=>{
        const _value = _init_key(this, '');
        _init_extra__init_key(this);
        return _value;
    })();
    get key() {
        return this.#___private_key_1;
    }
    set key(_v) {
        this.#___private_key_1 = _v;
    }
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_status_2 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_2;
    }
    set status(_v) {
        this.#___private_status_2 = _v;
    }
    #___private_collapsible_3 = (()=>{
        const _value = _init_collapsible(this, true);
        _init_extra__init_collapsible(this);
        return _value;
    })();
    get collapsible() {
        return this.#___private_collapsible_3;
    }
    set collapsible(_v) {
        this.#___private_collapsible_3 = _v;
    }
    #___private_defaultExpanded_4 = (()=>{
        const _value = _init_defaultExpanded(this, true);
        _init_extra__init_defaultExpanded(this);
        return _value;
    })();
    get defaultExpanded() {
        return this.#___private_defaultExpanded_4;
    }
    set defaultExpanded(_v) {
        this.#___private_defaultExpanded_4 = _v;
    }
    #___private_disabled_5 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_5;
    }
    set disabled(_v) {
        this.#___private_disabled_5 = _v;
    }
    #___private_maxHeight_6 = (()=>{
        const _value = _init_maxHeight(this, '');
        _init_extra__init_maxHeight(this);
        return _value;
    })();
    get maxHeight() {
        return this.#___private_maxHeight_6;
    }
    set maxHeight(_v) {
        this.#___private_maxHeight_6 = _v;
    }
    #___private_overflow_7 = (()=>{
        const _value = _init_overflow(this, 'scroll');
        _init_extra__init_overflow(this);
        return _value;
    })();
    get overflow() {
        return this.#___private_overflow_7;
    }
    set overflow(_v) {
        this.#___private_overflow_7 = _v;
    }
    #___private_expanded_8 = (()=>{
        const _value = _init_expanded(this);
        _init_extra__init_expanded(this);
        return _value;
    })();
    get expanded() {
        return this.#___private_expanded_8;
    }
    set expanded(_v) {
        this.#___private_expanded_8 = _v;
    }
    #___private_managedByChain_9 = (()=>{
        const _value = _init_managedByChain(this, false);
        _init_extra__init_managedByChain(this);
        return _value;
    })();
    get managedByChain() {
        return this.#___private_managedByChain_9;
    }
    set managedByChain(_v) {
        this.#___private_managedByChain_9 = _v;
    }
    #___private_chainLine_10 = (()=>{
        const _value = _init_chainLine(this, 'solid');
        _init_extra__init_chainLine(this);
        return _value;
    })();
    get chainLine() {
        return this.#___private_chainLine_10;
    }
    set chainLine(_v) {
        this.#___private_chainLine_10 = _v;
    }
    #___private_contentSlot_11 = (()=>{
        const _value = _init_contentSlot(this, null);
        _init_extra__init_contentSlot(this);
        return _value;
    })();
    get contentSlot() {
        return this.#___private_contentSlot_11;
    }
    set contentSlot(_v) {
        this.#___private_contentSlot_11 = _v;
    }
    #___private_titleSlot_12 = (()=>{
        const _value = _init_titleSlot(this, null);
        _init_extra__init_titleSlot(this);
        return _value;
    })();
    get titleSlot() {
        return this.#___private_titleSlot_12;
    }
    set titleSlot(_v) {
        this.#___private_titleSlot_12 = _v;
    }
    #___private_iconSlot_13 = (()=>{
        const _value = _init_iconSlot(this, null);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_13;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_13 = _v;
    }
    #___private_scrollContainer_14 = (()=>{
        const _value = _init_scrollContainer(this, null);
        _init_extra__init_scrollContainer(this);
        return _value;
    })();
    get scrollContainer() {
        return this.#___private_scrollContainer_14;
    }
    set scrollContainer(_v) {
        this.#___private_scrollContainer_14 = _v;
    }
    #___private_hasTitleSlot_15 = (()=>{
        const _value = _init_hasTitleSlot(this, false);
        _init_extra__init_hasTitleSlot(this);
        return _value;
    })();
    get hasTitleSlot() {
        return this.#___private_hasTitleSlot_15;
    }
    set hasTitleSlot(_v) {
        this.#___private_hasTitleSlot_15 = _v;
    }
    #___private_hasIconSlot_16 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_16;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_16 = _v;
    }
    #___private_hasContentSlot_17 = (()=>{
        const _value = _init_hasContentSlot(this, false);
        _init_extra__init_hasContentSlot(this);
        return _value;
    })();
    get hasContentSlot() {
        return this.#___private_hasContentSlot_17;
    }
    set hasContentSlot(_v) {
        this.#___private_hasContentSlot_17 = _v;
    }
    #___private_internalExpanded_18 = (()=>{
        const _value = _init_internalExpanded(this, false);
        _init_extra__init_internalExpanded(this);
        return _value;
    })();
    get internalExpanded() {
        return this.#___private_internalExpanded_18;
    }
    set internalExpanded(_v) {
        this.#___private_internalExpanded_18 = _v;
    }
    #___private_chainExpanded_19 = (()=>{
        const _value = _init_chainExpanded(this);
        _init_extra__init_chainExpanded(this);
        return _value;
    })();
    get chainExpanded() {
        return this.#___private_chainExpanded_19;
    }
    set chainExpanded(_v) {
        this.#___private_chainExpanded_19 = _v;
    }
    #___private_canScrollStart_20 = (()=>{
        const _value = _init_canScrollStart(this, false);
        _init_extra__init_canScrollStart(this);
        return _value;
    })();
    get canScrollStart() {
        return this.#___private_canScrollStart_20;
    }
    set canScrollStart(_v) {
        this.#___private_canScrollStart_20 = _v;
    }
    #___private_canScrollEnd_21 = (()=>{
        const _value = _init_canScrollEnd(this, false);
        _init_extra__init_canScrollEnd(this);
        return _value;
    })();
    get canScrollEnd() {
        return this.#___private_canScrollEnd_21;
    }
    set canScrollEnd(_v) {
        this.#___private_canScrollEnd_21 = _v;
    }
    #___private_hasOverflow_22 = (()=>{
        const _value = _init_hasOverflow(this, false);
        _init_extra__init_hasOverflow(this);
        return _value;
    })();
    get hasOverflow() {
        return this.#___private_hasOverflow_22;
    }
    set hasOverflow(_v) {
        this.#___private_hasOverflow_22 = _v;
    }
    #___private_overflowExpanded_23 = (()=>{
        const _value = _init_overflowExpanded(this, false);
        _init_extra__init_overflowExpanded(this);
        return _value;
    })();
    get overflowExpanded() {
        return this.#___private_overflowExpanded_23;
    }
    set overflowExpanded(_v) {
        this.#___private_overflowExpanded_23 = _v;
    }
    contentId = `ve-thought-chain-item-content-${++nextThoughtChainItemId}`;
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    scrollResizeObserver;
    scrollFrame = 0;
    seededInitialState = false;
    get actualExpanded() {
        if (!this.collapsible) return true;
        return this.expanded ?? this.chainExpanded ?? this.internalExpanded;
    }
    get hasTitleContent() {
        return (this.title ?? '').trim().length > 0 || this.hasTitleSlot;
    }
    get isOverflowLimited() {
        return Boolean(this.maxHeight) && this.actualExpanded;
    }
    get isScrollActive() {
        return this.isOverflowLimited && 'scroll' === this.overflow;
    }
    get isCollapsibleOverflowActive() {
        return this.isOverflowLimited && 'collapsible' === this.overflow;
    }
    get statusIcon() {
        return Object.hasOwn(STATUS_ICONS, this.status) ? STATUS_ICONS[this.status] : STATUS_ICONS.default;
    }
    setChainLine(line) {
        if (this.chainLine !== line) {
            this.chainLine = line;
            this.setAttribute('chain-line', line);
        }
    }
    setManagedByChain(managed) {
        if (this.managedByChain !== managed) this.managedByChain = managed;
    }
    setChainExpanded(expanded) {
        if (this.chainExpanded !== expanded) this.chainExpanded = expanded;
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (!this.seededInitialState) {
            this.internalExpanded = this.defaultExpanded;
            this.seededInitialState = true;
        }
        if (changedProperties.has('defaultExpanded') && void 0 === this.expanded && this.seededInitialState) this.internalExpanded = this.defaultExpanded;
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.syncSlotState();
        this.refreshScrollObserver();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('maxHeight') || changedProperties.has('overflow') || changedProperties.has('expanded') || changedProperties.has('chainExpanded') || changedProperties.has('collapsible') || changedProperties.has('internalExpanded')) this.refreshScrollObserver();
    }
    disconnectedCallback() {
        this.teardownScrollObserver();
        super.disconnectedCallback();
    }
    syncSlotState() {
        this.hasTitleSlot = hasMeaningfulAssignedNodes(this.titleSlot);
        this.hasIconSlot = hasMeaningfulAssignedNodes(this.iconSlot);
        this.hasContentSlot = hasMeaningfulAssignedNodes(this.contentSlot);
    }
    handleSlotChange() {
        this.syncSlotState();
        this.refreshScrollObserver();
    }
    refreshScrollObserver() {
        if (!this.isOverflowLimited) {
            this.teardownScrollObserver();
            this.canScrollStart = false;
            this.canScrollEnd = false;
            this.hasOverflow = false;
            this.overflowExpanded = false;
            return;
        }
        const scrollContainer = this.scrollContainer;
        if (!scrollContainer) return;
        if (!this.scrollResizeObserver) this.scrollResizeObserver = new ResizeObserver(()=>{
            this.scheduleScrollStateUpdate();
        });
        this.scrollResizeObserver.disconnect();
        this.scrollResizeObserver.observe(scrollContainer);
        Array.from(scrollContainer.children).forEach((child)=>{
            this.scrollResizeObserver?.observe(child);
        });
        this.scheduleScrollStateUpdate();
    }
    teardownScrollObserver() {
        this.scrollResizeObserver?.disconnect();
        this.scrollResizeObserver = void 0;
        window.cancelAnimationFrame(this.scrollFrame);
        this.scrollFrame = 0;
    }
    scheduleScrollStateUpdate() {
        window.cancelAnimationFrame(this.scrollFrame);
        this.scrollFrame = window.requestAnimationFrame(()=>{
            this.updateScrollState();
        });
    }
    updateScrollState() {
        const scrollContainer = this.scrollContainer;
        if (!scrollContainer || !this.isOverflowLimited) {
            this.canScrollStart = false;
            this.canScrollEnd = false;
            this.hasOverflow = false;
            return;
        }
        if (this.isCollapsibleOverflowActive && this.overflowExpanded) {
            this.canScrollStart = false;
            this.canScrollEnd = false;
            return;
        }
        const scrollTolerance = 1;
        const maxScrollTop = scrollContainer.scrollHeight - scrollContainer.clientHeight;
        this.hasOverflow = maxScrollTop > scrollTolerance;
        this.canScrollStart = this.isScrollActive && scrollContainer.scrollTop > scrollTolerance;
        this.canScrollEnd = this.isScrollActive && maxScrollTop - scrollContainer.scrollTop > scrollTolerance;
    }
    handleScroll() {
        this.updateScrollState();
    }
    handleOverflowToggle() {
        this.overflowExpanded = !this.overflowExpanded;
        this.scheduleScrollStateUpdate();
    }
    handleToggle() {
        if (!this.collapsible || this.disabled) return;
        const expanded = !this.actualExpanded;
        if (!this.managedByChain && void 0 === this.expanded) this.internalExpanded = expanded;
        this.dispatchEvent(new CustomEvent('ve-thought-chain-item-toggle', {
            detail: {
                itemKey: this.key,
                expanded
            },
            bubbles: true,
            composed: true
        }));
    }
    renderHeaderContent() {
        const collapsed = this.collapsible && !this.actualExpanded;
        return html`
      <span class="ve-thought-chain-item-heading">
        <span
          class="ve-thought-chain-item-title"
          part="title"
          ?hidden=${!this.hasTitleContent}
        >
          <slot name="title" @slotchange=${this.handleSlotChange}
            >${this.title}</slot
          >
          ${this.collapsible && this.hasContentSlot ? this.renderChevron(collapsed) : nothing}
        </span>
      </span>
    `;
    }
    renderChevron(collapsed) {
        const chevronClass = classMap({
            've-thought-chain-chevron': true,
            've-thought-chain-chevron-collapsed': collapsed
        });
        return html`<span class=${chevronClass} aria-hidden="true"
      >${renderChevronIcon()}</span
    >`;
    }
    renderHeader() {
        const toggleClass = classMap({
            've-thought-chain-item-toggle': true,
            've-thought-chain-trigger-reset': this.collapsible,
            've-thought-chain-item-toggle-enabled': this.collapsible && !this.disabled,
            've-thought-chain-trigger-enabled': this.collapsible && !this.disabled
        });
        if (!this.collapsible) return html`<div class=${toggleClass} part="toggle">
        ${this.renderHeaderContent()}
      </div>`;
        return html`<button
      class=${toggleClass}
      part="toggle"
      type="button"
      ?disabled=${this.disabled}
      aria-expanded=${toAriaBoolean(this.actualExpanded)}
      aria-controls=${this.contentId}
      @click=${this.handleToggle}
    >
      ${this.renderHeaderContent()}
    </button>`;
    }
    render() {
        const collapsed = this.collapsible && !this.actualExpanded;
        const itemClass = classMap({
            've-thought-chain-item': true,
            've-thought-chain-item-collapsed': collapsed,
            've-thought-chain-item-disabled': this.disabled
        });
        const contentClass = classMap({
            've-thought-chain-item-content': true,
            've-thought-chain-motion': true,
            've-thought-chain-motion-collapsed': collapsed
        });
        const scrollClass = classMap({
            've-thought-chain-item-scroll': true,
            've-thought-chain-item-scroll-limited': this.isOverflowLimited && !this.overflowExpanded,
            've-thought-chain-item-scroll-collapsible': this.isCollapsibleOverflowActive,
            've-thought-chain-item-scroll-expanded': this.overflowExpanded,
            've-thought-chain-item-scroll-mask-start': this.canScrollStart,
            've-thought-chain-item-scroll-mask-end': this.canScrollEnd
        });
        const scrollStyle = this.maxHeight && !this.overflowExpanded ? {
            maxHeight: this.maxHeight
        } : {};
        return html`
      <div
        class=${itemClass}
        part="item"
        role="listitem"
        aria-disabled=${toAriaBoolean(this.disabled)}
      >
        <span
          class="ve-thought-chain-item-connector"
          part="connector"
          aria-hidden="true"
        ></span>
        <span
          class="ve-thought-chain-item-icon"
          part="icon"
          aria-hidden="true"
        >
          <slot name="icon" @slotchange=${this.handleSlotChange}></slot>
          ${this.hasIconSlot ? nothing : this.statusIcon}
        </span>
        <div class="ve-thought-chain-item-main">
          <div class="ve-thought-chain-item-header" part="header">
            ${this.renderHeader()}
          </div>
          <div
            id=${this.contentId}
            class=${contentClass}
            part="content"
            ?hidden=${!this.hasContentSlot}
            aria-hidden=${toAriaBoolean(collapsed)}
            ?inert=${collapsed}
          >
            <div class="ve-thought-chain-item-scroll-frame">
              <div class="ve-thought-chain-item-scroll-clip">
                <div
                  class=${scrollClass}
                  style=${styleMap(scrollStyle)}
                  @scroll=${this.handleScroll}
                >
                  <slot @slotchange=${this.handleSlotChange}></slot>
                </div>
              </div>
              ${this.isCollapsibleOverflowActive && this.hasOverflow ? html`<button
                    class="ve-thought-chain-item-overflow-toggle"
                    type="button"
                    @click=${this.handleOverflowToggle}
                  >
                    ${this.overflowExpanded ? '收起' : '查看更多'}
                  </button>` : nothing}
            </div>
          </div>
        </div>
      </div>
    `;
    }
}
const ve_thought_chain = [
    componentReset,
    token,
    collapseTriggerResetStyles,
    collapseMotionStyles,
    chevronStyles,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      width: 100%;
      font-family: var(--font-sans);
      color: var(--ve-thought-chain-text-color);
    }

    :host([hidden]) {
      display: none;
    }

    .ve-thought-chain {
      box-sizing: border-box;
      display: grid;
      row-gap: var(--ve-thought-chain-gap);
      width: 100%;
      transition: row-gap var(--ve-thought-chain-motion-duration) ease;
    }

    .ve-thought-chain-collapsed {
      row-gap: 0;
    }

    .ve-thought-chain-header {
      box-sizing: border-box;
      display: flex;
      align-items: flex-start;
      gap: var(--ve-thought-chain-header-gap);
      min-width: 0;
      color: var(--ve-thought-chain-text-color);
    }

    .ve-thought-chain-trigger {
      display: flex;
      align-items: center;
      gap: var(--ve-thought-chain-header-gap);
    }

    .ve-thought-chain-root-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      width: var(--ve-thought-chain-item-icon-box-size);
      height: var(--ve-thought-chain-text-line-height);
      color: var(--ve-thought-chain-tertiary-text-color);
    }

    .ve-thought-chain-root-icon ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-width: 100%;
      max-height: 100%;
      line-height: 1;
      flex: none;
    }

    .ve-thought-chain-root-icon svg,
    .ve-thought-chain-root-icon ve-icon,
    .ve-thought-chain-root-icon ::slotted(svg),
    .ve-thought-chain-root-icon ::slotted(ve-icon) {
      width: var(--ve-thought-chain-item-icon-box-size);
      height: var(--ve-thought-chain-item-icon-box-size);
      --ve-icon-size: var(--ve-thought-chain-item-icon-box-size);
      flex: none;
    }

    .ve-thought-chain-heading {
      min-width: 0;
      flex: 1;
    }

    .ve-thought-chain-hidden-slot {
      display: none;
    }

    .ve-thought-chain-title {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-thought-chain-title-icon-gap);
      min-width: 0;
      color: var(--ve-thought-chain-title-text-color);
      font-size: var(--text-body);
      font-weight: var(--font-weight-normal);
      line-height: var(--ve-thought-chain-text-line-height);
    }

    :host([loading]) .ve-thought-chain-title {
      background-image: linear-gradient(
        90deg,
        var(--ve-thought-chain-shimmer-base-color) 6%,
        var(--ve-thought-chain-shimmer-highlight-color) 39%,
        var(--ve-thought-chain-shimmer-base-color) 72%
      );
      background-size: 220% 100%;
      background-clip: text;
      -webkit-background-clip: text;
      -webkit-text-fill-color: transparent;
      color: transparent;
      animation: ve-thought-chain-shimmer var(--ve-thought-chain-shimmer-duration)
        linear infinite;
    }

    .ve-thought-chain-title[hidden] {
      display: none;
    }

    .ve-thought-chain-list {
      display: grid;
      gap: var(--ve-thought-chain-list-gap);
      margin: 0;
      padding: 0;
      list-style: none;
      min-width: 0;
      min-height: 0;
      overflow: hidden;
    }

    :host([dir='rtl']) .ve-thought-chain-trigger,
    :host-context([dir='rtl']) .ve-thought-chain-trigger {
      direction: rtl;
      text-align: right;
    }

    @keyframes ve-thought-chain-shimmer {
      0% {
        background-position: 100% 0;
      }
      100% {
        background-position: -100% 0;
      }
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-thought-chain {
        transition: none;
      }

      :host([loading]) .ve-thought-chain-title {
        background: none;
        -webkit-text-fill-color: var(--ve-thought-chain-title-text-color);
        color: var(--ve-thought-chain-title-text-color);
        animation: none;
      }
    }
  `
];
function ve_thought_chain_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_thought_chain_element_toPropertyKey(arg) {
    var key = ve_thought_chain_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_thought_chain_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_thought_chain_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_thought_chain_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_thought_chain_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_thought_chain_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_thought_chain_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_thought_chain_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_thought_chain_element_dec, ve_thought_chain_element_dec1, ve_thought_chain_element_dec2, ve_thought_chain_element_dec3, ve_thought_chain_element_dec4, ve_thought_chain_element_dec5, ve_thought_chain_element_dec6, ve_thought_chain_element_dec7, ve_thought_chain_element_dec8, ve_thought_chain_element_dec9, ve_thought_chain_element_dec10, ve_thought_chain_element_dec11, ve_thought_chain_element_dec12, ve_thought_chain_element_dec13, _init_line, _init_extra__init_line, _init_loading, _init_extra__init_loading, ve_thought_chain_element_init_collapsible, ve_thought_chain_element_init_extra_init_collapsible, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_expandedKeys, _init_extra__init_expandedKeys, _init_defaultExpandedKeys, _init_extra__init_defaultExpandedKeys, _init_itemSlot, _init_extra__init_itemSlot, ve_thought_chain_element_init_titleSlot, ve_thought_chain_element_init_extra_init_titleSlot, ve_thought_chain_element_init_iconSlot, ve_thought_chain_element_init_extra_init_iconSlot, ve_thought_chain_element_init_hasTitleSlot, ve_thought_chain_element_init_extra_init_hasTitleSlot, ve_thought_chain_element_init_hasIconSlot, ve_thought_chain_element_init_extra_init_hasIconSlot, _init_internalOpen, _init_extra__init_internalOpen, _init_internalExpandedKeys, _init_extra__init_internalExpandedKeys;
const VE_THOUGHT_CHAIN_TAG = 've-thought-chain';
const ve_thought_chain_element_TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        return null === value ? true : 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
function shouldWarnThoughtChainUsage() {
    const environment = globalThis.process?.env?.NODE_ENV;
    return 'production' !== environment && "u" > typeof console;
}
let nextThoughtChainId = 0;
ve_thought_chain_element_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), ve_thought_chain_element_dec1 = property({
    type: Boolean,
    reflect: true
}), ve_thought_chain_element_dec2 = property({
    reflect: true,
    converter: ve_thought_chain_element_TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), ve_thought_chain_element_dec3 = property({
    type: Boolean,
    reflect: true
}), ve_thought_chain_element_dec4 = property({
    attribute: 'default-open',
    reflect: true,
    converter: ve_thought_chain_element_TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), ve_thought_chain_element_dec5 = property({
    attribute: false
}), ve_thought_chain_element_dec6 = property({
    attribute: false
}), ve_thought_chain_element_dec7 = query('slot:not([name])'), ve_thought_chain_element_dec8 = query('slot[name="title"]'), ve_thought_chain_element_dec9 = query('slot[name="icon"]'), ve_thought_chain_element_dec10 = state(), ve_thought_chain_element_dec11 = state(), ve_thought_chain_element_dec12 = state(), ve_thought_chain_element_dec13 = state();
class VeThoughtChainElement extends VeElement {
    static{
        ({ e: [_init_line, _init_extra__init_line, _init_loading, _init_extra__init_loading, ve_thought_chain_element_init_collapsible, ve_thought_chain_element_init_extra_init_collapsible, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_expandedKeys, _init_extra__init_expandedKeys, _init_defaultExpandedKeys, _init_extra__init_defaultExpandedKeys, _init_itemSlot, _init_extra__init_itemSlot, ve_thought_chain_element_init_titleSlot, ve_thought_chain_element_init_extra_init_titleSlot, ve_thought_chain_element_init_iconSlot, ve_thought_chain_element_init_extra_init_iconSlot, ve_thought_chain_element_init_hasTitleSlot, ve_thought_chain_element_init_extra_init_hasTitleSlot, ve_thought_chain_element_init_hasIconSlot, ve_thought_chain_element_init_extra_init_hasIconSlot, _init_internalOpen, _init_extra__init_internalOpen, _init_internalExpandedKeys, _init_extra__init_internalExpandedKeys] } = ve_thought_chain_element_apply_decs_2311(this, [], [
            [
                ve_thought_chain_element_dec,
                1,
                "line"
            ],
            [
                ve_thought_chain_element_dec1,
                1,
                "loading"
            ],
            [
                ve_thought_chain_element_dec2,
                1,
                "collapsible"
            ],
            [
                ve_thought_chain_element_dec3,
                1,
                "open"
            ],
            [
                ve_thought_chain_element_dec4,
                1,
                "defaultOpen"
            ],
            [
                ve_thought_chain_element_dec5,
                1,
                "expandedKeys"
            ],
            [
                ve_thought_chain_element_dec6,
                1,
                "defaultExpandedKeys"
            ],
            [
                ve_thought_chain_element_dec7,
                1,
                "itemSlot"
            ],
            [
                ve_thought_chain_element_dec8,
                1,
                "titleSlot"
            ],
            [
                ve_thought_chain_element_dec9,
                1,
                "iconSlot"
            ],
            [
                ve_thought_chain_element_dec10,
                1,
                "hasTitleSlot"
            ],
            [
                ve_thought_chain_element_dec11,
                1,
                "hasIconSlot"
            ],
            [
                ve_thought_chain_element_dec12,
                1,
                "internalOpen"
            ],
            [
                ve_thought_chain_element_dec13,
                1,
                "internalExpandedKeys"
            ]
        ]));
    }
    static styles = ve_thought_chain;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement,
        [VE_THOUGHT_CHAIN_ITEM_TAG]: VeThoughtChainItemElement
    };
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_line_1 = (()=>{
        const _value = _init_line(this, 'solid');
        _init_extra__init_line(this);
        return _value;
    })();
    get line() {
        return this.#___private_line_1;
    }
    set line(_v) {
        this.#___private_line_1 = _v;
    }
    #___private_loading_2 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_2;
    }
    set loading(_v) {
        this.#___private_loading_2 = _v;
    }
    #___private_collapsible_3 = (()=>{
        const _value = ve_thought_chain_element_init_collapsible(this, true);
        ve_thought_chain_element_init_extra_init_collapsible(this);
        return _value;
    })();
    get collapsible() {
        return this.#___private_collapsible_3;
    }
    set collapsible(_v) {
        this.#___private_collapsible_3 = _v;
    }
    #___private_open_4 = (()=>{
        const _value = _init_open(this);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_4;
    }
    set open(_v) {
        this.#___private_open_4 = _v;
    }
    #___private_defaultOpen_5 = (()=>{
        const _value = _init_defaultOpen(this, true);
        _init_extra__init_defaultOpen(this);
        return _value;
    })();
    get defaultOpen() {
        return this.#___private_defaultOpen_5;
    }
    set defaultOpen(_v) {
        this.#___private_defaultOpen_5 = _v;
    }
    #___private_expandedKeys_6 = (()=>{
        const _value = _init_expandedKeys(this);
        _init_extra__init_expandedKeys(this);
        return _value;
    })();
    get expandedKeys() {
        return this.#___private_expandedKeys_6;
    }
    set expandedKeys(_v) {
        this.#___private_expandedKeys_6 = _v;
    }
    #___private_defaultExpandedKeys_7 = (()=>{
        const _value = _init_defaultExpandedKeys(this, []);
        _init_extra__init_defaultExpandedKeys(this);
        return _value;
    })();
    get defaultExpandedKeys() {
        return this.#___private_defaultExpandedKeys_7;
    }
    set defaultExpandedKeys(_v) {
        this.#___private_defaultExpandedKeys_7 = _v;
    }
    #___private_itemSlot_8 = (()=>{
        const _value = _init_itemSlot(this, null);
        _init_extra__init_itemSlot(this);
        return _value;
    })();
    get itemSlot() {
        return this.#___private_itemSlot_8;
    }
    set itemSlot(_v) {
        this.#___private_itemSlot_8 = _v;
    }
    #___private_titleSlot_9 = (()=>{
        const _value = ve_thought_chain_element_init_titleSlot(this, null);
        ve_thought_chain_element_init_extra_init_titleSlot(this);
        return _value;
    })();
    get titleSlot() {
        return this.#___private_titleSlot_9;
    }
    set titleSlot(_v) {
        this.#___private_titleSlot_9 = _v;
    }
    #___private_iconSlot_10 = (()=>{
        const _value = ve_thought_chain_element_init_iconSlot(this, null);
        ve_thought_chain_element_init_extra_init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_10;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_10 = _v;
    }
    #___private_hasTitleSlot_11 = (()=>{
        const _value = ve_thought_chain_element_init_hasTitleSlot(this, false);
        ve_thought_chain_element_init_extra_init_hasTitleSlot(this);
        return _value;
    })();
    get hasTitleSlot() {
        return this.#___private_hasTitleSlot_11;
    }
    set hasTitleSlot(_v) {
        this.#___private_hasTitleSlot_11 = _v;
    }
    #___private_hasIconSlot_12 = (()=>{
        const _value = ve_thought_chain_element_init_hasIconSlot(this, false);
        ve_thought_chain_element_init_extra_init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_12;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_12 = _v;
    }
    #___private_internalOpen_13 = (()=>{
        const _value = _init_internalOpen(this, true);
        _init_extra__init_internalOpen(this);
        return _value;
    })();
    get internalOpen() {
        return this.#___private_internalOpen_13;
    }
    set internalOpen(_v) {
        this.#___private_internalOpen_13 = _v;
    }
    #___private_internalExpandedKeys_14 = (()=>{
        const _value = _init_internalExpandedKeys(this, []);
        _init_extra__init_internalExpandedKeys(this);
        return _value;
    })();
    get internalExpandedKeys() {
        return this.#___private_internalExpandedKeys_14;
    }
    set internalExpandedKeys(_v) {
        this.#___private_internalExpandedKeys_14 = _v;
    }
    seededDefaultExpandedItems = new WeakSet();
    warnedMissingKeyItems = new WeakSet();
    warnedDuplicateItemKeys = new Set();
    seededInitialState = false;
    contentId = `ve-thought-chain-content-${++nextThoughtChainId}`;
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    get actualOpen() {
        if (!this.shouldRenderHeader || !this.collapsible) return true;
        return this.open ?? this.internalOpen;
    }
    get actualExpandedKeys() {
        return this.expandedKeys ?? this.internalExpandedKeys;
    }
    get hasTitleContent() {
        return (this.title ?? '').trim().length > 0 || this.hasTitleSlot;
    }
    get shouldRenderHeader() {
        return this.hasTitleContent;
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (!this.seededInitialState) {
            this.internalOpen = this.defaultOpen;
            this.internalExpandedKeys = [
                ...this.defaultExpandedKeys
            ];
            this.seededInitialState = true;
        }
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.syncSlotState();
        this.syncChildItems();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('line')) this.syncChildLineOnly();
        if (changedProperties.has('expandedKeys') || changedProperties.has('defaultExpandedKeys')) this.syncChildItems();
    }
    getAssignedItems() {
        return this.itemSlot?.assignedElements({
            flatten: true
        }).filter((element)=>element.localName === VE_THOUGHT_CHAIN_ITEM_TAG) ?? [];
    }
    syncSlotState() {
        this.hasTitleSlot = hasMeaningfulAssignedNodes(this.titleSlot);
        this.hasIconSlot = hasMeaningfulAssignedNodes(this.iconSlot);
    }
    syncChildLineOnly() {
        for (const item of this.getAssignedItems())item.setChainLine(this.line);
    }
    warnItemKeyIssues(items) {
        if (!shouldWarnThoughtChainUsage()) return;
        const seenKeys = new Set();
        for (const item of items){
            if (!item.collapsible) continue;
            const itemKey = (item.key ?? '').trim();
            if (!itemKey) {
                if (!this.warnedMissingKeyItems.has(item)) {
                    console.warn("[ve-thought-chain] A collapsible ve-thought-chain-item is missing a key. Provide a stable unique key so expansion state can be managed correctly.");
                    this.warnedMissingKeyItems.add(item);
                }
                continue;
            }
            if (seenKeys.has(itemKey) && !this.warnedDuplicateItemKeys.has(itemKey)) {
                console.warn(`[ve-thought-chain] Duplicate item key "${itemKey}" found among direct children. Items with the same key share expansion state; provide unique keys for each collapsible item.`);
                this.warnedDuplicateItemKeys.add(itemKey);
                continue;
            }
            seenKeys.add(itemKey);
        }
    }
    syncChildItems() {
        const items = this.getAssignedItems();
        const expandedKeys = new Set(this.actualExpandedKeys);
        let changedDefaultKeys = false;
        this.warnItemKeyIssues(items);
        for (const item of items){
            const itemKey = (item.key ?? '').trim();
            const shouldManageExpansion = item.collapsible && itemKey.length > 0;
            item.setManagedByChain(shouldManageExpansion);
            item.setChainLine(this.line);
            if (shouldManageExpansion && void 0 === this.expandedKeys && void 0 === item.expanded && item.defaultExpanded && !this.seededDefaultExpandedItems.has(item)) {
                expandedKeys.add(itemKey);
                this.seededDefaultExpandedItems.add(item);
                changedDefaultKeys = true;
            }
            if (shouldManageExpansion) item.setChainExpanded(expandedKeys.has(itemKey));
            else item.setChainExpanded(void 0);
        }
        if (changedDefaultKeys) this.internalExpandedKeys = [
            ...expandedKeys
        ];
    }
    handleSlotChange() {
        this.syncSlotState();
        this.syncChildItems();
    }
    handleRootToggle() {
        if (!this.shouldRenderHeader || !this.collapsible) return;
        const open = !this.actualOpen;
        if (void 0 === this.open) this.internalOpen = open;
        this.dispatchEvent(new CustomEvent('ve-thought-chain-open-change', {
            detail: {
                open
            },
            bubbles: true,
            composed: true
        }));
    }
    handleItemToggle(event) {
        event.stopPropagation();
        const { itemKey, expanded } = event.detail;
        if (!itemKey) return;
        const nextExpandedKeys = new Set(this.actualExpandedKeys);
        if (expanded) nextExpandedKeys.add(itemKey);
        else nextExpandedKeys.delete(itemKey);
        const expandedKeys = [
            ...nextExpandedKeys
        ];
        if (void 0 === this.expandedKeys) {
            this.internalExpandedKeys = expandedKeys;
            this.syncChildItems();
        }
        this.dispatchEvent(new CustomEvent('ve-thought-chain-expand', {
            detail: {
                itemKey,
                expanded,
                expandedKeys
            },
            bubbles: true,
            composed: true
        }));
    }
    renderHeaderContent() {
        const collapsed = !this.actualOpen;
        const chevronClass = classMap({
            've-thought-chain-chevron': true,
            've-thought-chain-chevron-collapsed': collapsed
        });
        return html`
      ${this.hasIconSlot ? html`<span
            class="ve-thought-chain-root-icon"
            part="icon"
            aria-hidden="true"
          >
            <slot name="icon" @slotchange=${this.handleSlotChange}></slot>
          </span>` : html`<slot
            name="icon"
            class="ve-thought-chain-hidden-slot"
            @slotchange=${this.handleSlotChange}
          ></slot>`}
      <span class="ve-thought-chain-heading">
        <span
          class="ve-thought-chain-title"
          part="title"
          ?hidden=${!this.hasTitleContent}
        >
          <slot name="title" @slotchange=${this.handleSlotChange}
            >${this.title}</slot
          >
          ${this.collapsible ? html`<span class=${chevronClass} aria-hidden="true"
                >${renderChevronIcon()}</span
              >` : nothing}
        </span>
      </span>
    `;
    }
    renderHeader() {
        const triggerClass = classMap({
            've-thought-chain-trigger': true,
            've-thought-chain-trigger-reset': this.collapsible,
            've-thought-chain-trigger-enabled': this.collapsible
        });
        if (!this.collapsible) return html`<div class=${triggerClass} part="trigger">
        ${this.renderHeaderContent()}
      </div>`;
        return html`<button
      class=${triggerClass}
      part="trigger"
      type="button"
      aria-expanded=${toAriaBoolean(this.actualOpen)}
      aria-controls=${this.contentId}
      @click=${this.handleRootToggle}
    >
      ${this.renderHeaderContent()}
    </button>`;
    }
    render() {
        const collapsed = !this.actualOpen;
        const rootClass = classMap({
            've-thought-chain': true,
            've-thought-chain-collapsed': collapsed
        });
        const contentClass = classMap({
            've-thought-chain-motion': true,
            've-thought-chain-motion-collapsed': collapsed
        });
        return html`
      <section class=${rootClass} part="root">
        ${this.shouldRenderHeader ? html`<header class="ve-thought-chain-header" part="header">
              ${this.renderHeader()}
            </header>` : nothing}
        <div
          id=${this.contentId}
          class=${contentClass}
          part="content"
          aria-hidden=${toAriaBoolean(collapsed)}
          ?inert=${collapsed}
        >
          <div
            class="ve-thought-chain-list"
            part="list"
            role="list"
            @ve-thought-chain-item-toggle=${this.handleItemToggle}
          >
            <slot @slotchange=${this.handleSlotChange}></slot>
          </div>
        </div>
      </section>
    `;
    }
}
defineCustomElement(VE_THOUGHT_CHAIN_ITEM_TAG, VeThoughtChainItemElement);
defineCustomElement(VE_THOUGHT_CHAIN_TAG, VeThoughtChainElement);
export { VE_THOUGHT_CHAIN_ITEM_TAG, VE_THOUGHT_CHAIN_TAG, VeThoughtChainElement, VeThoughtChainItemElement };
