// @ve-design/web/dist/24968.js
function setComponentTitle(host, currentTitle, nextValue, applyTitle) {
    const nextTitle = null == nextValue ? '' : String(nextValue);
    if (currentTitle !== nextTitle) {
        applyTitle(nextTitle);
        host.requestUpdate('title', currentTitle);
    }
    if (null !== host.getAttribute('title')) host.removeAttribute('title');
}
function createNativeTitleAttributeController(host) {
    let observer;
    const consume = ()=>{
        const titleAttribute = host.getAttribute('title');
        if (null !== titleAttribute) host.title = titleAttribute;
    };
    const controller = {
        hostConnected () {
            consume();
            if (observer || "u" < typeof MutationObserver) return;
            observer = new MutationObserver(consume);
            observer.observe(host, {
                attributes: true,
                attributeFilter: [
                    'title'
                ]
            });
        },
        hostDisconnected () {
            observer?.disconnect();
            observer = void 0;
        }
    };
    host.addController(controller);
}
export { createNativeTitleAttributeController, setComponentTitle };
