// @ve-design/web/dist/29989.js
import { css, html, nothing } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { unsafeSVG } from "lit/directives/unsafe-svg.js";
import { property } from "lit/decorators.js";
import { componentReset, VeElement } from "./78928.js";
import { getIcon, addIconRegistryChangeListener } from "./icons/registry.js";
const iconToken = css`
  :host {
    --ve-icon-size: 1.15em;
    --ve-icon-stroke-width: 1.5;
  }
`;
const token = iconToken;
const ve_icon = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      color: inherit;
      line-height: 0;
      vertical-align: -0.125em;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-icon {
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-icon-size);
      block-size: var(--ve-icon-size);
      color: inherit;
      line-height: 0;
    }

    .ve-icon svg {
      display: block;
      flex: none;
      inline-size: 100%;
      block-size: 100%;
      color: inherit;
      overflow: visible;
    }

    .ve-icon svg [stroke] {
      stroke-width: var(--ve-icon-stroke-width);
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _init_name, _init_extra__init_name, _init_definition, _init_extra__init_definition, _init_label, _init_extra__init_label, _init_size, _init_extra__init_size;
const VE_ICON_TAG = 've-icon';
_dec = property({
    type: String,
    reflect: true
}), _dec1 = property({
    attribute: false
}), _dec2 = property({
    type: String
}), _dec3 = property({
    reflect: true
});
class VeIconElement extends VeElement {
    static{
        ({ e: [_init_name, _init_extra__init_name, _init_definition, _init_extra__init_definition, _init_label, _init_extra__init_label, _init_size, _init_extra__init_size] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "name"
            ],
            [
                _dec1,
                1,
                "definition"
            ],
            [
                _dec2,
                1,
                "label"
            ],
            [
                _dec3,
                1,
                "size"
            ]
        ]));
    }
    static styles = ve_icon;
    #___private_name_1 = (()=>{
        const _value = _init_name(this, '');
        _init_extra__init_name(this);
        return _value;
    })();
    get name() {
        return this.#___private_name_1;
    }
    set name(_v) {
        this.#___private_name_1 = _v;
    }
    #___private_definition_2 = (()=>{
        const _value = _init_definition(this, void 0);
        _init_extra__init_definition(this);
        return _value;
    })();
    get definition() {
        return this.#___private_definition_2;
    }
    set definition(_v) {
        this.#___private_definition_2 = _v;
    }
    #___private_label_3 = (()=>{
        const _value = _init_label(this, '');
        _init_extra__init_label(this);
        return _value;
    })();
    get label() {
        return this.#___private_label_3;
    }
    set label(_v) {
        this.#___private_label_3 = _v;
    }
    #___private_size_4 = (()=>{
        const _value = _init_size(this, '');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_4;
    }
    set size(_v) {
        this.#___private_size_4 = _v;
    }
    removeIconRegistryChangeListener = null;
    connectedCallback() {
        super.connectedCallback();
        this.removeIconRegistryChangeListener = addIconRegistryChangeListener((names)=>{
            if (this.name && names.includes(this.name)) this.requestUpdate();
        });
        this.loadCurrentIcon();
    }
    disconnectedCallback() {
        this.removeIconRegistryChangeListener?.();
        this.removeIconRegistryChangeListener = null;
        super.disconnectedCallback();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('name') || changedProperties.has('definition')) this.loadCurrentIcon();
    }
    loadCurrentIcon() {
        if (this.definition) return;
        const name = (this.name ?? '').trim();
        if (!name || getIcon(name)) return;
        import("./23467.js").then((m)=>m.loader_namespaceObject).then(({ isBuiltInIconName, loadIcon })=>{
            if ((this.name ?? '').trim() !== name || getIcon(name) || !isBuiltInIconName(name)) return;
            return loadIcon(name);
        }).catch(()=>{});
    }
    get resolvedIcon() {
        if (this.definition) return this.definition;
        return this.name ? getIcon(this.name) : void 0;
    }
    get normalizedSize() {
        if ('number' == typeof this.size) return Number.isFinite(this.size) ? `${this.size}px` : void 0;
        const size = (this.size ?? '').trim();
        if (!size) return;
        return /^\d+(\.\d+)?$/.test(size) ? `${size}px` : size;
    }
    get iconStyle() {
        const size = this.normalizedSize;
        return size ? `--ve-icon-size: ${size};` : void 0;
    }
    render() {
        const icon = this.resolvedIcon;
        if (!icon) return nothing;
        const label = (this.label ?? '').trim();
        return html`
      <span
        class="ve-icon"
        part="icon"
        style=${ifDefined(this.iconStyle)}
        role=${ifDefined(label ? 'img' : void 0)}
        aria-label=${ifDefined(label || void 0)}
        aria-hidden=${ifDefined(label ? void 0 : 'true')}
      >
        ${unsafeSVG(icon.svg)}
      </span>
    `;
    }
}
export { VE_ICON_TAG, VeIconElement };
