// @ve-design/web/dist/42896.js
import { arrow, autoUpdate, computePosition, flip, hide, offset, shift } from "@floating-ui/dom";
function readCssPixelValue(styles, name, fallback) {
    const value = Number.parseFloat(styles.getPropertyValue(name));
    return Number.isNaN(value) ? fallback : value;
}
function getFloatingDistance(host, options) {
    const { offsetVarName, arrowVarName, fallbackOffset = 8, fallbackArrowSize = 6, arrowGap = 2 } = options;
    const styles = window.getComputedStyle(host);
    const offsetValue = readCssPixelValue(styles, offsetVarName, fallbackOffset);
    const arrowSize = readCssPixelValue(styles, arrowVarName, fallbackArrowSize);
    return offsetValue + arrowSize / 2 + arrowGap;
}
async function computeFloatingPanelPosition(options) {
    const { reference, floating, arrowElement, placement, distance, viewportPadding = 8, boundary } = options;
    const boundaryOptions = boundary ? {
        boundary
    } : {};
    const middleware = [
        offset(distance),
        flip({
            padding: viewportPadding,
            ...boundaryOptions
        }),
        shift({
            padding: viewportPadding,
            ...boundaryOptions
        }),
        hide(boundaryOptions)
    ];
    if (arrowElement) middleware.push(arrow({
        element: arrowElement,
        padding: 8
    }));
    const result = await computePosition(reference, floating, {
        placement,
        strategy: 'fixed',
        middleware
    });
    const arrowData = result.middlewareData.arrow;
    const hideData = result.middlewareData.hide;
    return {
        x: result.x,
        y: result.y,
        placement: result.placement,
        arrowX: arrowData?.x ?? null,
        arrowY: arrowData?.y ?? null,
        hidden: Boolean(hideData?.referenceHidden || hideData?.escaped)
    };
}
function createFloatingAutoUpdate(reference, floating, update) {
    return autoUpdate(reference, floating, update, {
        ancestorScroll: true,
        ancestorResize: true,
        elementResize: true,
        layoutShift: true,
        animationFrame: false
    });
}
function createRafScheduler() {
    let frameId = null;
    return {
        schedule (task) {
            if (null !== frameId) return;
            frameId = window.requestAnimationFrame(()=>{
                frameId = null;
                task();
            });
        },
        cancel () {
            if (null === frameId) return;
            window.cancelAnimationFrame(frameId);
            frameId = null;
        }
    };
}
export { computeFloatingPanelPosition, createFloatingAutoUpdate, createRafScheduler, getFloatingDistance };
