// @ve-design/web/dist/67885.js
import { css, html, nothing } from "lit";
import { property, query, state } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { componentReset, VeElement } from "./78928.js";
import { createFloatingAutoUpdate, getFloatingDistance, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
const veTooltipTokenDefaults = css`
  :host {
    --ve-tooltip-z-index: 1000;
    --ve-tooltip-padding-block: var(--space-xxxs);
    --ve-tooltip-padding-inline: var(--space-xxs);
    --ve-tooltip-content-gap: 2px;
    --ve-tooltip-trigger-focus-radius: var(--radius-sm);
    --ve-tooltip-trigger-focus-ring-width: var(--stroke-2);
    --ve-tooltip-trigger-focus-ring-color: color-mix(
      currentColor 24%,
      transparent
    );
    --ve-tooltip-border-radius: var(--radius-md);
    --ve-tooltip-border-width: var(--stroke-weight-base);
    --ve-tooltip-border-color: transparent;
    --ve-tooltip-background: var(--color-bg-inverse);
    --ve-tooltip-text-color: var(--color-text-foreground);
    --ve-tooltip-font-size: var(--text-caption);
    --ve-tooltip-font-weight: var(--font-weight-normal);
    --ve-tooltip-line-height: var(--line-height-body);
    --ve-tooltip-title-font-weight: var(--font-weight-medium);
    --ve-tooltip-shadow: var(--shadow-sm);
    --ve-tooltip-arrow-size: var(--space-xs);
    --ve-tooltip-current-arrow-size: 0px;
    --ve-tooltip-offset: var(--space-xxs);
    --ve-tooltip-arrow-offset: calc(
      var(--ve-tooltip-current-arrow-size) * -0.5
    );
    --ve-tooltip-arrow-background: var(--ve-tooltip-background);
    --ve-tooltip-disabled-opacity: var(--state-disabled-ratio);
    --ve-tooltip-transition-duration: var(--motion-base, 220ms);
    --ve-tooltip-transition-easing: var(--ease-standard, ease);
    --ve-tooltip-enter-easing: var(--ease-enter, cubic-bezier(0.16, 1, 0.3, 1));
    --ve-tooltip-exit-easing: var(--ease-exit, cubic-bezier(0.7, 0, 0.84, 0));
    --ve-tooltip-enter-transform-y: var(--space-xxs);
    --ve-tooltip-enter-transform-x: var(--space-xxs);

    --ve-tooltip-current-enter-transform: translate3d(
      0,
      var(--ve-tooltip-enter-transform-y),
      0
    );
  }

  :host([arrow]) {
    --ve-tooltip-current-arrow-size: var(--ve-tooltip-arrow-size);
  }

  :host([position^='top']) {
    --ve-tooltip-current-enter-transform: translate3d(
      0,
      var(--ve-tooltip-enter-transform-y),
      0
    );
  }

  :host([position^='bottom']) {
    --ve-tooltip-current-enter-transform: translate3d(
      0,
      calc(var(--ve-tooltip-enter-transform-y) * -1),
      0
    );
  }

  :host([position^='left']) {
    --ve-tooltip-current-enter-transform: translate3d(
      var(--ve-tooltip-enter-transform-x),
      0,
      0
    );
  }

  :host([position^='right']) {
    --ve-tooltip-current-enter-transform: translate3d(
      calc(var(--ve-tooltip-enter-transform-x) * -1),
      0,
      0
    );
  }
`;
const style_token = veTooltipTokenDefaults;
const ve_tooltip = [
    componentReset,
    style_token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-tooltip {
      position: relative;
      display: inline-flex;
      inline-size: fit-content;
      max-inline-size: 100%;
    }

    .ve-tooltip-trigger {
      display: inline-flex;
      align-items: inherit;
      justify-content: inherit;
      inline-size: fit-content;
      max-inline-size: 100%;
      outline: none;
    }

    .ve-tooltip-disabled .ve-tooltip-trigger {
      cursor: not-allowed;
    }

    .ve-tooltip-trigger:focus-visible {
      border-radius: var(--ve-tooltip-trigger-focus-radius);
      box-shadow: 0 0 0 var(--ve-tooltip-trigger-focus-ring-width)
        var(--ve-tooltip-trigger-focus-ring-color);
    }

    .ve-tooltip-panel {
      position: fixed;
      inset: auto;
      margin: 0;
      overflow: visible;
      z-index: var(--ve-tooltip-z-index);
      box-sizing: border-box;
      padding-block: var(--ve-tooltip-padding-block);
      padding-inline: var(--ve-tooltip-padding-inline);
      border: var(--ve-tooltip-border-width) solid
        var(--ve-tooltip-border-color);
      border-radius: var(--ve-tooltip-border-radius);
      background: var(--ve-tooltip-background);
      color: var(--ve-tooltip-text-color);
      box-shadow: var(--ve-tooltip-shadow);
      font-family: var(--font-sans);
      font-size: var(--ve-tooltip-font-size);
      font-weight: var(--ve-tooltip-font-weight);
      line-height: var(--ve-tooltip-line-height);
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
      transform: var(--ve-tooltip-current-enter-transform);
      transform-origin: center;
      transition:
        opacity var(--ve-tooltip-transition-duration)
          var(--ve-tooltip-exit-easing),
        transform var(--ve-tooltip-transition-duration)
          var(--ve-tooltip-exit-easing),
        visibility 0s linear var(--ve-tooltip-transition-duration);
    }

    .ve-tooltip-panel.is-open {
      opacity: 1;
      visibility: visible;
      pointer-events: auto;
      transform: translate3d(0, 0, 0);
      transition:
        opacity var(--ve-tooltip-transition-duration)
          var(--ve-tooltip-enter-easing),
        transform var(--ve-tooltip-transition-duration)
          var(--ve-tooltip-enter-easing),
        visibility 0s linear 0s;
    }

    .ve-tooltip-panel.is-closed {
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
    }

    .ve-tooltip-inner {
      display: flex;
      flex-direction: column;
      gap: var(--ve-tooltip-content-gap);
      inline-size: max-content;
    }

    .ve-tooltip-inner ::slotted(*) {
      margin: 0;
    }

    .ve-tooltip-inner ::slotted(strong),
    .ve-tooltip-inner ::slotted(b) {
      font-weight: var(--ve-tooltip-title-font-weight);
    }

    .ve-tooltip-panel.is-open::before {
      content: '';
      position: absolute;
      pointer-events: auto;
      background: transparent;
    }

    .ve-tooltip-placement-top.is-open::before,
    .ve-tooltip-placement-top-start.is-open::before,
    .ve-tooltip-placement-top-end.is-open::before {
      inset-block-start: 100%;
      inset-inline: 0;
      block-size: calc(
        var(--ve-tooltip-offset) + var(--ve-tooltip-current-arrow-size)
      );
    }

    .ve-tooltip-placement-bottom.is-open::before,
    .ve-tooltip-placement-bottom-start.is-open::before,
    .ve-tooltip-placement-bottom-end.is-open::before {
      inset-block-end: 100%;
      inset-inline: 0;
      block-size: calc(
        var(--ve-tooltip-offset) + var(--ve-tooltip-current-arrow-size)
      );
    }

    .ve-tooltip-placement-left.is-open::before,
    .ve-tooltip-placement-left-start.is-open::before,
    .ve-tooltip-placement-left-end.is-open::before {
      inset-block: 0;
      inset-inline-start: 100%;
      inline-size: calc(
        var(--ve-tooltip-offset) + var(--ve-tooltip-current-arrow-size)
      );
    }

    .ve-tooltip-placement-right.is-open::before,
    .ve-tooltip-placement-right-start.is-open::before,
    .ve-tooltip-placement-right-end.is-open::before {
      inset-block: 0;
      inset-inline-end: 100%;
      inline-size: calc(
        var(--ve-tooltip-offset) + var(--ve-tooltip-current-arrow-size)
      );
    }

    .ve-tooltip-arrow {
      position: absolute;
      inline-size: var(--ve-tooltip-current-arrow-size);
      block-size: var(--ve-tooltip-current-arrow-size);
      background: var(--ve-tooltip-arrow-background);
      transform: rotate(45deg);
      pointer-events: none;
    }

    .ve-tooltip-placement-top .ve-tooltip-arrow,
    .ve-tooltip-placement-top-start .ve-tooltip-arrow,
    .ve-tooltip-placement-top-end .ve-tooltip-arrow {
      inset-block-end: var(--ve-tooltip-arrow-offset);
    }

    .ve-tooltip-placement-bottom .ve-tooltip-arrow,
    .ve-tooltip-placement-bottom-start .ve-tooltip-arrow,
    .ve-tooltip-placement-bottom-end .ve-tooltip-arrow {
      inset-block-start: var(--ve-tooltip-arrow-offset);
    }

    .ve-tooltip-placement-left .ve-tooltip-arrow,
    .ve-tooltip-placement-left-start .ve-tooltip-arrow,
    .ve-tooltip-placement-left-end .ve-tooltip-arrow {
      inset-inline-end: var(--ve-tooltip-arrow-offset);
    }

    .ve-tooltip-placement-right .ve-tooltip-arrow,
    .ve-tooltip-placement-right-start .ve-tooltip-arrow,
    .ve-tooltip-placement-right-end .ve-tooltip-arrow {
      inset-inline-start: var(--ve-tooltip-arrow-offset);
    }

    :host([disabled]) {
      opacity: var(--ve-tooltip-disabled-opacity);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-tooltip-panel {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _init_position, _init_extra__init_position, _init_trigger, _init_extra__init_trigger, _init_arrow, _init_extra__init_arrow, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_disabled, _init_extra__init_disabled, _init_delay, _init_extra__init_delay, _init_innerOpen, _init_extra__init_innerOpen, _init_hasContent, _init_extra__init_hasContent, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_arrowElement, _init_extra__init_arrowElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_contentSlot, _init_extra__init_contentSlot;
const VE_TOOLTIP_TAG = 've-tooltip';
const DEFAULT_DELAY = 120;
const HOVER_SAFE_CLOSE_DELAY = 160;
const VIEWPORT_PADDING = 8;
const FALLBACK_OFFSET = 8;
const FALLBACK_ARROW_SIZE = 6;
const ARROW_GAP = 2;
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    type: Boolean,
    attribute: 'default-open'
}), _dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = property({
    type: Number
}), _dec7 = state(), _dec8 = state(), _dec9 = state(), _dec10 = query('.ve-tooltip-trigger'), _dec11 = query('.ve-tooltip-panel'), _dec12 = query('.ve-tooltip-arrow'), _dec13 = query('slot:not([name])'), _dec14 = query('slot[name="content"]');
class VeTooltipElement extends VeElement {
    static{
        ({ e: [_init_position, _init_extra__init_position, _init_trigger, _init_extra__init_trigger, _init_arrow, _init_extra__init_arrow, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_disabled, _init_extra__init_disabled, _init_delay, _init_extra__init_delay, _init_innerOpen, _init_extra__init_innerOpen, _init_hasContent, _init_extra__init_hasContent, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_arrowElement, _init_extra__init_arrowElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_contentSlot, _init_extra__init_contentSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "position"
            ],
            [
                _dec1,
                1,
                "trigger"
            ],
            [
                _dec2,
                1,
                "arrow"
            ],
            [
                _dec3,
                1,
                "open"
            ],
            [
                _dec4,
                1,
                "defaultOpen"
            ],
            [
                _dec5,
                1,
                "disabled"
            ],
            [
                _dec6,
                1,
                "delay"
            ],
            [
                _dec7,
                1,
                "innerOpen"
            ],
            [
                _dec8,
                1,
                "hasContent"
            ],
            [
                _dec9,
                1,
                "resolvedPlacement"
            ],
            [
                _dec10,
                1,
                "triggerElement"
            ],
            [
                _dec11,
                1,
                "panelElement"
            ],
            [
                _dec12,
                1,
                "arrowElement"
            ],
            [
                _dec13,
                1,
                "defaultSlot"
            ],
            [
                _dec14,
                1,
                "contentSlot"
            ]
        ]));
    }
    static styles = ve_tooltip;
    #___private_position_1 = (()=>{
        const _value = _init_position(this, 'top');
        _init_extra__init_position(this);
        return _value;
    })();
    get position() {
        return this.#___private_position_1;
    }
    set position(_v) {
        this.#___private_position_1 = _v;
    }
    #___private_trigger_2 = (()=>{
        const _value = _init_trigger(this, 'hover');
        _init_extra__init_trigger(this);
        return _value;
    })();
    get trigger() {
        return this.#___private_trigger_2;
    }
    set trigger(_v) {
        this.#___private_trigger_2 = _v;
    }
    #___private_arrow_3 = (()=>{
        const _value = _init_arrow(this, false);
        _init_extra__init_arrow(this);
        return _value;
    })();
    get arrow() {
        return this.#___private_arrow_3;
    }
    set arrow(_v) {
        this.#___private_arrow_3 = _v;
    }
    #___private_open_4 = (()=>{
        const _value = _init_open(this, void 0);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_4;
    }
    set open(_v) {
        this.#___private_open_4 = _v;
    }
    #___private_defaultOpen_5 = (()=>{
        const _value = _init_defaultOpen(this, false);
        _init_extra__init_defaultOpen(this);
        return _value;
    })();
    get defaultOpen() {
        return this.#___private_defaultOpen_5;
    }
    set defaultOpen(_v) {
        this.#___private_defaultOpen_5 = _v;
    }
    #___private_disabled_6 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_6;
    }
    set disabled(_v) {
        this.#___private_disabled_6 = _v;
    }
    #___private_delay_7 = (()=>{
        const _value = _init_delay(this, DEFAULT_DELAY);
        _init_extra__init_delay(this);
        return _value;
    })();
    get delay() {
        return this.#___private_delay_7;
    }
    set delay(_v) {
        this.#___private_delay_7 = _v;
    }
    #___private_innerOpen_8 = (()=>{
        const _value = _init_innerOpen(this, false);
        _init_extra__init_innerOpen(this);
        return _value;
    })();
    get innerOpen() {
        return this.#___private_innerOpen_8;
    }
    set innerOpen(_v) {
        this.#___private_innerOpen_8 = _v;
    }
    #___private_hasContent_9 = (()=>{
        const _value = _init_hasContent(this, false);
        _init_extra__init_hasContent(this);
        return _value;
    })();
    get hasContent() {
        return this.#___private_hasContent_9;
    }
    set hasContent(_v) {
        this.#___private_hasContent_9 = _v;
    }
    #___private_resolvedPlacement_10 = (()=>{
        const _value = _init_resolvedPlacement(this, this.position);
        _init_extra__init_resolvedPlacement(this);
        return _value;
    })();
    get resolvedPlacement() {
        return this.#___private_resolvedPlacement_10;
    }
    set resolvedPlacement(_v) {
        this.#___private_resolvedPlacement_10 = _v;
    }
    #___private_triggerElement_11 = (()=>{
        const _value = _init_triggerElement(this);
        _init_extra__init_triggerElement(this);
        return _value;
    })();
    get triggerElement() {
        return this.#___private_triggerElement_11;
    }
    set triggerElement(_v) {
        this.#___private_triggerElement_11 = _v;
    }
    #___private_panelElement_12 = (()=>{
        const _value = _init_panelElement(this);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_12;
    }
    set panelElement(_v) {
        this.#___private_panelElement_12 = _v;
    }
    #___private_arrowElement_13 = (()=>{
        const _value = _init_arrowElement(this);
        _init_extra__init_arrowElement(this);
        return _value;
    })();
    get arrowElement() {
        return this.#___private_arrowElement_13;
    }
    set arrowElement(_v) {
        this.#___private_arrowElement_13 = _v;
    }
    #___private_defaultSlot_14 = (()=>{
        const _value = _init_defaultSlot(this);
        _init_extra__init_defaultSlot(this);
        return _value;
    })();
    get defaultSlot() {
        return this.#___private_defaultSlot_14;
    }
    set defaultSlot(_v) {
        this.#___private_defaultSlot_14 = _v;
    }
    #___private_contentSlot_15 = (()=>{
        const _value = _init_contentSlot(this);
        _init_extra__init_contentSlot(this);
        return _value;
    })();
    get contentSlot() {
        return this.#___private_contentSlot_15;
    }
    set contentSlot(_v) {
        this.#___private_contentSlot_15 = _v;
    }
    panelId = `ve-tooltip-panel-${Math.random().toString(36).slice(2)}`;
    openTimer = null;
    closeTimer = null;
    repositionScheduler = createRafScheduler();
    cleanupAutoUpdate = null;
    autoUpdateReference = null;
    documentBound = false;
    positionRequestId = 0;
    describedElement = null;
    get referenceElement() {
        return this.getSlottedTriggerElement() ?? this.triggerElement;
    }
    get isOpen() {
        return 'boolean' == typeof this.open ? this.open : this.innerOpen;
    }
    get isControlled() {
        return 'boolean' == typeof this.open;
    }
    get canShowPanel() {
        return this.isOpen && this.hasContent && !this.disabled;
    }
    get isInteractiveTrigger() {
        return !this.disabled;
    }
    connectedCallback() {
        super.connectedCallback();
        this.innerOpen = this.open ?? this.defaultOpen;
    }
    firstUpdated() {
        this.syncContentState();
        this.syncOverlayState();
    }
    disconnectedCallback() {
        this.clearDelayTimers();
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        this.teardownAutoUpdate();
        this.removeDocumentListener();
        this.clearTriggerDescription();
        super.disconnectedCallback();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('open') && 'boolean' == typeof this.open) this.innerOpen = this.open;
        if (changedProperties.has('defaultOpen') && 1 === changedProperties.size && 'boolean' != typeof this.open) this.innerOpen = this.defaultOpen;
        if (changedProperties.has('disabled') && this.disabled && !this.isControlled) this.innerOpen = false;
        if (changedProperties.has('open') || changedProperties.has('innerOpen') || changedProperties.has('disabled') || changedProperties.has('trigger') || changedProperties.has('position') || changedProperties.has('arrow') || changedProperties.has('hasContent')) this.syncOverlayState();
    }
    dispatchOpenChange(nextOpen) {
        this.dispatchEvent(new CustomEvent('ve-open-change', {
            detail: {
                open: nextOpen
            },
            bubbles: true,
            composed: true
        }));
    }
    setOpen(nextOpen) {
        if (nextOpen && this.disabled) return;
        if (this.isOpen === nextOpen) return;
        if (nextOpen && !this.hasContent) return;
        if (!this.isControlled) this.innerOpen = nextOpen;
        this.dispatchOpenChange(nextOpen);
    }
    syncContentState() {
        const assignedNodes = this.contentSlot?.assignedNodes({
            flatten: true
        }) ?? [];
        this.hasContent = assignedNodes.some((node)=>{
            if (node.nodeType === Node.ELEMENT_NODE) return true;
            if (node.nodeType === Node.TEXT_NODE) return (node.textContent ?? '').trim().length > 0;
            return false;
        });
        if (!this.hasContent && this.isOpen && !this.isControlled) this.setOpen(false);
        this.syncOverlayState();
    }
    getSlottedTriggerElement() {
        const assignedElements = this.defaultSlot?.assignedElements({
            flatten: true
        }) ?? [];
        return assignedElements.find((element)=>element instanceof HTMLElement) ?? null;
    }
    clearDelayTimers() {
        if (null !== this.openTimer) {
            window.clearTimeout(this.openTimer);
            this.openTimer = null;
        }
        if (null !== this.closeTimer) {
            window.clearTimeout(this.closeTimer);
            this.closeTimer = null;
        }
    }
    scheduleOpen() {
        this.clearDelayTimers();
        if (this.delay <= 0) return void this.setOpen(true);
        this.openTimer = window.setTimeout(()=>{
            this.setOpen(true);
            this.openTimer = null;
        }, Math.max(0, this.delay));
    }
    scheduleClose(delay = this.delay) {
        this.clearDelayTimers();
        if (delay <= 0) return void this.setOpen(false);
        this.closeTimer = window.setTimeout(()=>{
            this.setOpen(false);
            this.closeTimer = null;
        }, Math.max(0, delay));
    }
    handleTriggerMouseEnter = ()=>{
        if (!this.isInteractiveTrigger || 'hover' !== this.trigger) return;
        this.scheduleOpen();
    };
    handleTriggerMouseLeave = (event)=>{
        if (!this.isInteractiveTrigger || 'hover' !== this.trigger) return;
        const nextTarget = event.relatedTarget;
        if (nextTarget && this.panelElement?.contains(nextTarget)) return;
        this.scheduleClose(Math.max(this.delay, HOVER_SAFE_CLOSE_DELAY));
    };
    handleTriggerClick = ()=>{
        if (!this.isInteractiveTrigger || 'click' !== this.trigger) return;
        this.setOpen(!this.isOpen);
    };
    handleTriggerFocusIn = ()=>{
        if (!this.isInteractiveTrigger || 'focus' !== this.trigger && 'hover' !== this.trigger) return;
        this.setOpen(true);
    };
    handleTriggerFocusOut = (event)=>{
        if (!this.isInteractiveTrigger || 'focus' !== this.trigger && 'hover' !== this.trigger) return;
        const nextTarget = event.relatedTarget;
        if (nextTarget && (this.contains(nextTarget) || this.panelElement?.contains(nextTarget))) return;
        this.scheduleClose();
    };
    handlePanelMouseEnter = ()=>{
        if (!this.isInteractiveTrigger || 'hover' !== this.trigger) return;
        this.clearDelayTimers();
    };
    handlePanelMouseLeave = (event)=>{
        if (!this.isInteractiveTrigger || 'hover' !== this.trigger) return;
        const nextTarget = event.relatedTarget;
        if (nextTarget && this.referenceElement?.contains(nextTarget)) return;
        this.scheduleClose(Math.max(this.delay, HOVER_SAFE_CLOSE_DELAY));
    };
    handleKeydown = (event)=>{
        if ('Escape' !== event.key) return;
        event.stopPropagation();
        if (!this.isInteractiveTrigger) return;
        this.setOpen(false);
        this.referenceElement?.focus();
    };
    handleDocumentPointerDown = (event)=>{
        if (!this.canShowPanel || 'click' !== this.trigger || !this.isInteractiveTrigger) return;
        const path = event.composedPath();
        const clickedInside = path.includes(this) || path.includes(this.referenceElement) || path.includes(this.panelElement);
        if (!clickedInside) this.setOpen(false);
    };
    syncDocumentListener() {
        if (!this.isConnected) return void this.removeDocumentListener();
        if ('click' === this.trigger && this.canShowPanel && !this.documentBound) {
            document.addEventListener('pointerdown', this.handleDocumentPointerDown, true);
            this.documentBound = true;
            return;
        }
        if (('click' !== this.trigger || !this.canShowPanel) && this.documentBound) this.removeDocumentListener();
    }
    syncTriggerDescription() {
        const referenceElement = this.getSlottedTriggerElement();
        if (this.describedElement && this.describedElement !== referenceElement) {
            this.removeDescribedByToken(this.describedElement);
            this.describedElement = null;
        }
        if (!referenceElement || !this.canShowPanel) return void this.clearTriggerDescription();
        this.addDescribedByToken(referenceElement);
        this.describedElement = referenceElement;
    }
    clearTriggerDescription() {
        if (!this.describedElement) return;
        this.removeDescribedByToken(this.describedElement);
        this.describedElement = null;
    }
    addDescribedByToken(element) {
        const tokens = new Set((element.getAttribute('aria-describedby') ?? '').split(/\s+/).filter(Boolean));
        tokens.add(this.panelId);
        element.setAttribute('aria-describedby', [
            ...tokens
        ].join(' '));
    }
    removeDescribedByToken(element) {
        const tokens = (element.getAttribute('aria-describedby') ?? '').split(/\s+/).filter((token)=>token && token !== this.panelId);
        if (tokens.length > 0) return void element.setAttribute('aria-describedby', tokens.join(' '));
        element.removeAttribute('aria-describedby');
    }
    syncOverlayState() {
        this.syncDocumentListener();
        this.syncTriggerDescription();
        if (!this.canShowPanel) {
            hidePanelPopover(this.panelElement);
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            this.teardownAutoUpdate();
            return;
        }
        this.updateComplete.then(()=>{
            if (!this.canShowPanel || !this.isConnected) return;
            showPanelPopover(this.panelElement);
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
            this.syncDocumentListener();
            this.syncTriggerDescription();
        });
    }
    removeDocumentListener() {
        if (!this.documentBound) return;
        document.removeEventListener('pointerdown', this.handleDocumentPointerDown, true);
        this.documentBound = false;
    }
    setupAutoUpdate() {
        const referenceElement = this.referenceElement;
        if (!this.isConnected || !this.canShowPanel || !referenceElement || !this.panelElement) return;
        if (this.cleanupAutoUpdate && this.autoUpdateReference === referenceElement) return;
        this.teardownAutoUpdate();
        this.autoUpdateReference = referenceElement;
        this.cleanupAutoUpdate = createFloatingAutoUpdate(referenceElement, this.panelElement, ()=>{
            this.scheduleUpdatePosition();
        });
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
        this.autoUpdateReference = null;
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    async updatePosition() {
        const referenceElement = this.referenceElement;
        if (!this.canShowPanel || !this.isConnected || !referenceElement || !this.panelElement) return;
        const requestId = ++this.positionRequestId;
        const distance = getFloatingDistance(this, {
            offsetVarName: '--ve-tooltip-offset',
            arrowVarName: '--ve-tooltip-current-arrow-size',
            fallbackOffset: FALLBACK_OFFSET,
            fallbackArrowSize: this.arrow ? FALLBACK_ARROW_SIZE : 0,
            arrowGap: this.arrow ? ARROW_GAP : 0
        });
        this.panelElement.style.position = 'fixed';
        this.panelElement.style.top = '0px';
        this.panelElement.style.left = '0px';
        const position = await computeFloatingPanelPosition({
            reference: referenceElement,
            floating: this.panelElement,
            arrowElement: this.arrow ? this.arrowElement : null,
            placement: this.position,
            distance,
            viewportPadding: VIEWPORT_PADDING
        });
        if (requestId !== this.positionRequestId || !this.canShowPanel || !this.isConnected) return;
        this.resolvedPlacement = position.placement;
        this.panelElement.style.left = `${position.x}px`;
        this.panelElement.style.top = `${position.y}px`;
        if (this.arrowElement) {
            this.arrowElement.style.removeProperty('left');
            this.arrowElement.style.removeProperty('top');
            if (null !== position.arrowX) this.arrowElement.style.left = `${position.arrowX}px`;
            if (null !== position.arrowY) this.arrowElement.style.top = `${position.arrowY}px`;
        }
        this.panelElement.style.visibility = position.hidden ? 'hidden' : 'visible';
    }
    handleContentSlotChange = ()=>{
        this.syncContentState();
    };
    handleDefaultSlotChange = ()=>{
        if (this.canShowPanel) {
            this.teardownAutoUpdate();
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
            this.syncTriggerDescription();
        }
    };
    render() {
        const hostClass = classMap({
            've-tooltip': true,
            've-tooltip-disabled': this.disabled
        });
        const panelClass = classMap({
            've-tooltip-panel': true,
            [`ve-tooltip-placement-${this.resolvedPlacement}`]: true,
            'is-open': this.canShowPanel,
            'is-closed': !this.canShowPanel
        });
        return html`
      <span class=${hostClass}>
        <span
          class="ve-tooltip-trigger"
          part="trigger"
          aria-disabled=${this.disabled ? 'true' : nothing}
          @mouseenter=${this.handleTriggerMouseEnter}
          @mouseleave=${this.handleTriggerMouseLeave}
          @click=${this.handleTriggerClick}
          @focusin=${this.handleTriggerFocusIn}
          @focusout=${this.handleTriggerFocusOut}
          @keydown=${this.handleKeydown}
        >
          <slot @slotchange=${this.handleDefaultSlotChange}></slot>
        </span>

        <div
          id=${this.panelId}
          class=${panelClass}
          part="panel"
          popover=${"manual"}
          role="tooltip"
          aria-hidden=${this.canShowPanel ? 'false' : 'true'}
          @mouseenter=${this.handlePanelMouseEnter}
          @mouseleave=${this.handlePanelMouseLeave}
          @keydown=${this.handleKeydown}
        >
          ${this.arrow ? html`<div class="ve-tooltip-arrow" part="arrow"></div>` : nothing}
          <div class="ve-tooltip-inner" part="content">
            <slot name="content" @slotchange=${this.handleContentSlotChange}></slot>
          </div>
        </div>
      </span>
    `;
    }
}
export { VE_TOOLTIP_TAG, VeTooltipElement };
