// @ve-design/web/dist/68533.js
import { css, html, nothing } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { loadingIcon } from "./42938.js";
const veButtonTokenDefaults = css`
  :host {
    --ve-button-gap: var(--space-xxxs);

    --ve-button-height-small: var(--height-small);
    --ve-button-height-default: var(--height-default);
    --ve-button-height-large: var(--height-large);

    --ve-button-padding-inline-small: var(--space-xxs);
    --ve-button-padding-inline-default: var(--space-xs);
    --ve-button-padding-inline-large: var(--space-xs);

    --ve-button-radius-default: var(--radius-md);
    --ve-button-radius-round: var(--radius-full);
    --ve-button-radius-circle: var(--radius-full);

    --ve-button-font-size-small: var(--text-body-sm);
    --ve-button-font-size-default: var(--text-body);
    --ve-button-font-size-large: var(--text-body);
    --ve-button-font-weight: var(--font-weight-normal);

    --ve-button-background-primary: var(--color-bg-primary);
    --ve-button-text-primary: var(--color-text-foreground);
    --ve-button-border-primary: transparent;

    --ve-button-background-secondary: var(--color-bg-secondary);
    --ve-button-text-secondary: var(--color-text-primary);
    --ve-button-border-secondary: transparent;

    --ve-button-background-outline: transparent;
    --ve-button-text-outline: var(--color-text-primary);
    --ve-button-border-outline: var(--color-border-strong);
    --ve-button-border-outline-hover: var(--color-border-interactive);
    --ve-button-border-outline-pressed: var(--color-border-interactive-active);

    --ve-button-background-text: transparent;
    --ve-button-text-text: var(--color-text-accent);
    --ve-button-border-text: transparent;

    --ve-button-status-surface-success: var(--color-bg-success);
    --ve-button-status-surface-warning: var(--color-bg-warning);
    --ve-button-status-surface-danger: var(--color-bg-danger);

    --ve-button-status-surface-secondary-success: var(
      --color-bg-success-secondary
    );
    --ve-button-status-surface-secondary-warning: var(
      --color-bg-warning-secondary
    );
    --ve-button-status-surface-secondary-danger: var(
      --color-bg-danger-secondary
    );

    --ve-button-status-text-success: var(--color-text-success);
    --ve-button-status-text-warning: var(--color-text-warning);
    --ve-button-status-text-danger: var(--color-text-danger);

    --ve-button-status-border-success: var(--color-border-success);
    --ve-button-status-border-warning: var(--color-border-warning);
    --ve-button-status-border-danger: var(--color-border-danger);

    --ve-button-loading-icon-size: var(--ve-button-current-font-size);
    --ve-button-disabled-opacity: var(--state-disabled-ratio);
    --ve-button-press-transform: translateY(0.4px) scale(0.995);

    --ve-button-current-height: var(--ve-button-height-default);
    --ve-button-current-padding-inline: var(--ve-button-padding-inline-default);
    --ve-button-current-radius: var(--ve-button-radius-default);
    --ve-button-current-font-size: var(--ve-button-font-size-default);

    --ve-button-current-background: var(--ve-button-background-primary);
    --ve-button-current-background-hover: color-mix(
      var(--ve-button-current-background) var(--ve-button-current-hover-ratio),
      var(--ve-button-current-hover-mix)
    );
    --ve-button-current-background-pressed: color-mix(
      var(--ve-button-current-background) var(--ve-button-current-pressed-ratio),
      var(--ve-button-current-pressed-mix)
    );
    --ve-button-current-text: var(--ve-button-text-primary);
    --ve-button-current-text-hover: var(--ve-button-current-text);
    --ve-button-current-text-pressed: var(--ve-button-current-text);
    --ve-button-current-border: var(--ve-button-border-primary);
    --ve-button-current-border-hover: color-mix(
      var(--ve-button-current-border) var(--ve-button-current-hover-ratio),
      var(--ve-button-current-hover-mix)
    );
    --ve-button-current-border-pressed: color-mix(
      var(--ve-button-current-border) var(--ve-button-current-pressed-ratio),
      var(--ve-button-current-pressed-mix)
    );
    --ve-button-current-border-width: 0px;

    --ve-button-current-hover-ratio: var(--state-hover-solid-ratio);
    --ve-button-current-hover-mix: var(--state-hover-solid-mix);
    --ve-button-current-pressed-ratio: var(--state-pressed-solid-ratio);
    --ve-button-current-pressed-mix: var(--state-pressed-solid-mix);

    --ve-button-current-status-surface: transparent;
    --ve-button-current-status-surface-secondary: transparent;
    --ve-button-current-status-text: currentColor;
    --ve-button-current-status-border: currentColor;
  }

  :host([type='secondary']) {
    --ve-button-current-background: var(--ve-button-background-secondary);
    --ve-button-current-text: var(--ve-button-text-secondary);
    --ve-button-current-border: var(--ve-button-border-secondary);
    --ve-button-current-border-width: 0px;

    --ve-button-current-hover-ratio: var(--state-hover-subtle-ratio);
    --ve-button-current-hover-mix: var(--state-hover-subtle-mix);
    --ve-button-current-pressed-ratio: var(--state-pressed-subtle-ratio);
    --ve-button-current-pressed-mix: var(--state-pressed-subtle-mix);
  }

  :host([type='outline']),
  :host([type='outline-filled']) {
    --ve-button-current-background: var(--ve-button-background-outline);
    --ve-button-current-text: var(--ve-button-text-outline);
    --ve-button-current-border: var(--ve-button-border-outline);
    --ve-button-current-border-hover: var(--ve-button-border-outline-hover);
    --ve-button-current-border-pressed: var(--ve-button-border-outline-pressed);
    --ve-button-current-border-width: var(--stroke-weight-base);

    --ve-button-current-hover-ratio: var(--state-hover-subtle-ratio);
    --ve-button-current-hover-mix: var(--state-hover-subtle-mix);
    --ve-button-current-pressed-ratio: var(--state-pressed-subtle-ratio);
    --ve-button-current-pressed-mix: var(--state-pressed-subtle-mix);
  }

  :host([type='outline']) {
    --ve-button-current-background-hover: var(--ve-button-current-background);
    --ve-button-current-background-pressed: var(
      --ve-button-current-background
    );
  }

  :host([type='outline-filled']) {
    --ve-button-current-border-hover: var(--ve-button-current-border);
    --ve-button-current-border-pressed: var(--ve-button-current-border);
  }

  :host([type='text']) {
    --ve-button-current-background: var(--ve-button-background-text);
    --ve-button-current-text: var(--ve-button-text-text);
    --ve-button-current-border: var(--ve-button-border-text);
    --ve-button-current-border-width: 0px;

    --ve-button-current-hover-ratio: var(--state-hover-subtle-ratio);
    --ve-button-current-hover-mix: var(--state-hover-subtle-mix);
    --ve-button-current-pressed-ratio: var(--state-pressed-subtle-ratio);
    --ve-button-current-pressed-mix: var(--state-pressed-subtle-mix);
  }

  :host([size='small']) {
    --ve-button-current-height: var(--ve-button-height-small);
    --ve-button-current-padding-inline: var(--ve-button-padding-inline-small);
    --ve-button-current-font-size: var(--ve-button-font-size-small);
  }

  :host([size='large']) {
    --ve-button-current-height: var(--ve-button-height-large);
    --ve-button-current-padding-inline: var(--ve-button-padding-inline-large);
    --ve-button-current-font-size: var(--ve-button-font-size-large);
  }

  :host([shape='round']) {
    --ve-button-current-radius: var(--ve-button-radius-round);
  }

  :host([shape='circle']) {
    --ve-button-current-radius: var(--ve-button-radius-circle);
    --ve-button-current-padding-inline: 0px;
  }

  :host([status='success']) {
    --ve-button-current-status-surface: var(--ve-button-status-surface-success);
    --ve-button-current-status-surface-secondary: var(
      --ve-button-status-surface-secondary-success
    );
    --ve-button-current-status-text: var(--ve-button-status-text-success);
    --ve-button-current-status-border: var(--ve-button-status-border-success);
  }

  :host([status='warning']) {
    --ve-button-current-status-surface: var(--ve-button-status-surface-warning);
    --ve-button-current-status-surface-secondary: var(
      --ve-button-status-surface-secondary-warning
    );
    --ve-button-current-status-text: var(--ve-button-status-text-warning);
    --ve-button-current-status-border: var(--ve-button-status-border-warning);
  }

  :host([status='danger']) {
    --ve-button-current-status-surface: var(--ve-button-status-surface-danger);
    --ve-button-current-status-surface-secondary: var(
      --ve-button-status-surface-secondary-danger
    );
    --ve-button-current-status-text: var(--ve-button-status-text-danger);
    --ve-button-current-status-border: var(--ve-button-status-border-danger);
  }

  :host([status='success'][type='primary']),
  :host([status='warning'][type='primary']),
  :host([status='danger'][type='primary']) {
    --ve-button-current-background: var(--ve-button-current-status-surface);
    --ve-button-current-text: var(--ve-button-text-primary);
  }

  :host([status='success'][type='secondary']),
  :host([status='warning'][type='secondary']),
  :host([status='danger'][type='secondary']) {
    --ve-button-current-background: var(
      --ve-button-current-status-surface-secondary
    );
    --ve-button-current-text: var(--ve-button-current-status-text);
  }

  :host([status='success'][type='outline']),
  :host([status='warning'][type='outline']),
  :host([status='danger'][type='outline']),
  :host([status='success'][type='outline-filled']),
  :host([status='warning'][type='outline-filled']),
  :host([status='danger'][type='outline-filled']) {
    --ve-button-current-text: var(--ve-button-current-status-text);
    --ve-button-current-border: var(--ve-button-current-status-border);
    --ve-button-current-text-hover: var(--ve-button-current-status-text);
    --ve-button-current-text-pressed: color-mix(
      var(--ve-button-current-status-text)
        var(--ve-button-current-pressed-ratio),
      var(--ve-button-current-pressed-mix)
    );
    --ve-button-current-border-hover: var(--ve-button-current-text-hover);
    --ve-button-current-border-pressed: color-mix(
      var(--ve-button-current-status-border)
        var(--ve-button-current-pressed-ratio),
      var(--ve-button-current-pressed-mix)
    );
  }

  :host([status='success'][type='outline-filled']),
  :host([status='warning'][type='outline-filled']),
  :host([status='danger'][type='outline-filled']) {
    --ve-button-current-border-hover: var(--ve-button-current-border);
    --ve-button-current-border-pressed: var(--ve-button-current-border);
    --ve-button-current-background-hover: color-mix(
      var(--ve-button-current-status-surface-secondary)
        var(--ve-button-current-hover-ratio),
      var(--ve-button-current-hover-mix)
    );
    --ve-button-current-background-pressed: color-mix(
      var(--ve-button-current-status-surface-secondary)
        var(--ve-button-current-pressed-ratio),
      var(--ve-button-current-pressed-mix)
    );
  }

  :host([status='success'][type='text']),
  :host([status='warning'][type='text']),
  :host([status='danger'][type='text']) {
    --ve-button-current-background-hover: color-mix(
      var(--ve-button-current-status-surface-secondary)
        var(--ve-button-current-hover-ratio),
      var(--ve-button-current-hover-mix)
    );
    --ve-button-current-background-pressed: color-mix(
      var(--ve-button-current-status-surface-secondary)
        var(--ve-button-current-pressed-ratio),
      var(--ve-button-current-pressed-mix)
    );
    --ve-button-current-text: var(--ve-button-current-status-text);
  }
`;
const token = veButtonTokenDefaults;
const ve_button = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    :host([disabled]) {
      cursor: not-allowed;
    }

    :host([loading]) {
      cursor: wait;
    }

    :host([long]) {
      display: flex;
      width: 100%;
    }

    .ve-btn {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: var(--ve-button-gap);
      appearance: none;
      box-sizing: border-box;
      margin: 0;
      min-inline-size: fit-content;
      block-size: var(--ve-button-current-height);
      padding-block: 0;
      padding-inline: var(--ve-button-current-padding-inline);
      border-style: solid;
      border-width: var(--ve-button-current-border-width);
      border-color: var(--ve-button-current-border);
      border-radius: var(--ve-button-current-radius);
      background: var(--ve-button-current-background);
      color: var(--ve-button-current-text);
      font-family: var(--font-sans);
      font-size: var(--ve-button-current-font-size);
      font-weight: var(--ve-button-font-weight);
      line-height: 1;
      text-align: center;
      text-decoration: none;
      white-space: nowrap;
      user-select: none;
      cursor: pointer;
      transition:
        background-color 160ms ease,
        border-color 160ms ease,
        color 160ms ease,
        transform 90ms ease-out,
        opacity 160ms ease,
        box-shadow 160ms ease;
    }

    :host([long]) .ve-btn {
      width: 100%;
    }

    .ve-btn--compact {
      inline-size: var(--ve-button-current-height);
      min-inline-size: var(--ve-button-current-height);
      padding-inline: 0;
    }

    :host([shape='circle']) .ve-btn {
      inline-size: var(--ve-button-current-height);
    }

    .ve-btn:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--stroke-2)
        color-mix(in oklch, currentColor 24%, transparent);
    }

    :host(:not([disabled]):not([loading])) .ve-btn:hover {
      background: var(--ve-button-current-background-hover);
      border-color: var(--ve-button-current-border-hover);
      color: var(--ve-button-current-text-hover);
    }

    :host(:not([disabled]):not([loading])) .ve-btn:active {
      background: var(--ve-button-current-background-pressed);
      border-color: var(--ve-button-current-border-pressed);
      color: var(--ve-button-current-text-pressed);
      transform: var(--ve-button-press-transform);
    }

    :host([disabled]) .ve-btn {
      opacity: var(--ve-button-disabled-opacity);
      cursor: not-allowed;
      pointer-events: none;
    }

    :host([loading]) .ve-btn {
      cursor: wait;
      pointer-events: none;
    }

    .ve-btn-content {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: var(--ve-button-gap);
    }

    .ve-btn-loading-icon {
      --ve-icon-size: var(--ve-button-loading-icon-size);

      animation: ve-button-spin 0.8s linear infinite;
      flex: none;
    }

    @keyframes ve-button-spin {
      to {
        transform: rotate(360deg);
      }
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-btn {
        transition: none;
      }

      .ve-btn-loading-icon {
        animation: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _init_type, _init_extra__init_type, _init_size, _init_extra__init_size, _init_shape, _init_extra__init_shape, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_loading, _init_extra__init_loading, _init_long, _init_extra__init_long, _init_href, _init_extra__init_href, _init_target, _init_extra__init_target, _init_controlElement, _init_extra__init_controlElement, _init_contentElement, _init_extra__init_contentElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_compact, _init_extra__init_compact;
const VE_BUTTON_TAG = 've-button';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby'
];
const COMPACT_SYMBOL_TEXT = /^[^\p{L}\p{N}\p{Script=Han}]{1,2}$/u;
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec3 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = property({
    type: Boolean,
    reflect: true
}), _dec7 = property({
    type: String
}), _dec8 = property({
    type: String
}), _dec9 = query('.ve-btn'), _dec10 = query('.ve-btn-content'), _dec11 = query('slot'), _dec12 = state();
class VeButtonElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_size, _init_extra__init_size, _init_shape, _init_extra__init_shape, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_loading, _init_extra__init_loading, _init_long, _init_extra__init_long, _init_href, _init_extra__init_href, _init_target, _init_extra__init_target, _init_controlElement, _init_extra__init_controlElement, _init_contentElement, _init_extra__init_contentElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_compact, _init_extra__init_compact] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "type"
            ],
            [
                _dec1,
                1,
                "size"
            ],
            [
                _dec2,
                1,
                "shape"
            ],
            [
                _dec3,
                1,
                "status"
            ],
            [
                _dec4,
                1,
                "disabled"
            ],
            [
                _dec5,
                1,
                "loading"
            ],
            [
                _dec6,
                1,
                "long"
            ],
            [
                _dec7,
                1,
                "href"
            ],
            [
                _dec8,
                1,
                "target"
            ],
            [
                _dec9,
                1,
                "controlElement"
            ],
            [
                _dec10,
                1,
                "contentElement"
            ],
            [
                _dec11,
                1,
                "defaultSlot"
            ],
            [
                _dec12,
                1,
                "compact"
            ]
        ]));
    }
    static styles = ve_button;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'primary');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_size_2 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_2;
    }
    set size(_v) {
        this.#___private_size_2 = _v;
    }
    #___private_shape_3 = (()=>{
        const _value = _init_shape(this, 'default');
        _init_extra__init_shape(this);
        return _value;
    })();
    get shape() {
        return this.#___private_shape_3;
    }
    set shape(_v) {
        this.#___private_shape_3 = _v;
    }
    #___private_status_4 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_4;
    }
    set status(_v) {
        this.#___private_status_4 = _v;
    }
    #___private_disabled_5 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_5;
    }
    set disabled(_v) {
        this.#___private_disabled_5 = _v;
    }
    #___private_loading_6 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_6;
    }
    set loading(_v) {
        this.#___private_loading_6 = _v;
    }
    #___private_long_7 = (()=>{
        const _value = _init_long(this, false);
        _init_extra__init_long(this);
        return _value;
    })();
    get long() {
        return this.#___private_long_7;
    }
    set long(_v) {
        this.#___private_long_7 = _v;
    }
    #___private_href_8 = (()=>{
        const _value = _init_href(this, '');
        _init_extra__init_href(this);
        return _value;
    })();
    get href() {
        return this.#___private_href_8;
    }
    set href(_v) {
        this.#___private_href_8 = _v;
    }
    #___private_target_9 = (()=>{
        const _value = _init_target(this, '_self');
        _init_extra__init_target(this);
        return _value;
    })();
    get target() {
        return this.#___private_target_9;
    }
    set target(_v) {
        this.#___private_target_9 = _v;
    }
    #___private_controlElement_10 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_10;
    }
    set controlElement(_v) {
        this.#___private_controlElement_10 = _v;
    }
    #___private_contentElement_11 = (()=>{
        const _value = _init_contentElement(this, null);
        _init_extra__init_contentElement(this);
        return _value;
    })();
    get contentElement() {
        return this.#___private_contentElement_11;
    }
    set contentElement(_v) {
        this.#___private_contentElement_11 = _v;
    }
    #___private_defaultSlot_12 = (()=>{
        const _value = _init_defaultSlot(this, null);
        _init_extra__init_defaultSlot(this);
        return _value;
    })();
    get defaultSlot() {
        return this.#___private_defaultSlot_12;
    }
    set defaultSlot(_v) {
        this.#___private_defaultSlot_12 = _v;
    }
    #___private_compact_13 = (()=>{
        const _value = _init_compact(this, false);
        _init_extra__init_compact(this);
        return _value;
    })();
    get compact() {
        return this.#___private_compact_13;
    }
    set compact(_v) {
        this.#___private_compact_13 = _v;
    }
    contentResizeObserver = null;
    observedContentElement = null;
    constructor(){
        super();
        this.addEventListener('click', this.handleHostClick, {
            capture: true
        });
    }
    get isLinkMode() {
        return (this.href ?? '').trim().length > 0;
    }
    get isUnavailable() {
        return this.disabled || this.loading;
    }
    get controlAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get controlAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get controlAriaDescribedby() {
        return this.getAttribute('aria-describedby') ?? void 0;
    }
    focus(options) {
        this.controlElement?.focus(options);
    }
    click() {
        if (this.isUnavailable) return;
        this.controlElement?.click();
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.contentResizeObserver = new ResizeObserver(()=>{
            this.syncCompactState();
        });
        this.observeContentElement();
        this.syncCompactState();
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
        this.observeContentElement();
        this.syncCompactState();
    }
    disconnectedCallback() {
        this.contentResizeObserver?.disconnect();
        this.observedContentElement = null;
        super.disconnectedCallback();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    handleHostClick(event) {
        if (this.isUnavailable) {
            event.preventDefault();
            event.stopImmediatePropagation();
        }
    }
    handleClick(event) {
        if (this.isUnavailable) {
            event.preventDefault();
            event.stopPropagation();
        }
    }
    handleSlotChange() {
        this.syncCompactState();
    }
    observeContentElement() {
        if (!this.contentResizeObserver) return;
        if (this.observedContentElement === this.contentElement) return;
        if (this.observedContentElement) this.contentResizeObserver.unobserve(this.observedContentElement);
        if (this.contentElement) this.contentResizeObserver.observe(this.contentElement);
        this.observedContentElement = this.contentElement;
    }
    syncCompactState() {
        this.compact = this.shouldRenderCompactSquare();
    }
    shouldRenderCompactSquare() {
        if ('default' !== this.shape || this.long || this.loading) return false;
        const controlElement = this.controlElement;
        const contentElement = this.contentElement;
        if (!controlElement || !contentElement || !this.hasCompactContent()) return false;
        const contentWidth = contentElement.getBoundingClientRect().width;
        const controlHeight = controlElement.getBoundingClientRect().height;
        return contentWidth > 0 && controlHeight > 0 && contentWidth <= controlHeight;
    }
    hasCompactContent() {
        const assignedNodes = this.defaultSlot?.assignedNodes({
            flatten: true
        }) ?? [];
        const meaningfulNodes = assignedNodes.filter((node)=>{
            if (node.nodeType === Node.TEXT_NODE) return (node.textContent?.trim().length ?? 0) > 0;
            return node.nodeType === Node.ELEMENT_NODE;
        });
        return meaningfulNodes.length > 0 && meaningfulNodes.every((node)=>this.isCompactContentNode(node));
    }
    isCompactContentNode(node) {
        if (node.nodeType === Node.TEXT_NODE) return this.isCompactText(node.textContent ?? '');
        if (node.nodeType !== Node.ELEMENT_NODE) return false;
        const element = node;
        const textContent = element.textContent?.trim() ?? '';
        const tagName = element.tagName.toLowerCase();
        const isIconElement = 'svg' === tagName || 'img' === tagName || tagName.includes('icon');
        'true' === element.getAttribute('aria-hidden') || 'presentation' === element.getAttribute('role') || element.getAttribute('role');
        if (isIconElement && 0 === textContent.length) return true;
        if (0 === textContent.length) return false;
        return this.isCompactText(textContent);
    }
    isCompactText(text) {
        return COMPACT_SYMBOL_TEXT.test(text.trim());
    }
    renderContent() {
        return html`
      ${this.loading ? html`<ve-icon
            class="ve-btn-loading-icon"
            part="spinner"
            .definition=${loadingIcon}
          ></ve-icon>` : nothing}
      <span class="ve-btn-content" part="content">
        <slot @slotchange=${this.handleSlotChange}></slot>
      </span>
    `;
    }
    render() {
        const controlClass = this.compact ? 've-btn ve-btn--compact' : 've-btn';
        if (this.isLinkMode) return html`
        <a
          class=${controlClass}
          part="control"
          href=${ifDefined(this.isUnavailable ? void 0 : this.href)}
          target=${ifDefined(this.isUnavailable ? void 0 : this.target)}
          rel=${ifDefined(!this.isUnavailable && '_blank' === this.target ? 'noreferrer noopener' : void 0)}
          aria-label=${ifDefined(this.controlAriaLabel)}
          aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
          aria-describedby=${ifDefined(this.controlAriaDescribedby)}
          aria-disabled=${ifDefined(this.isUnavailable ? 'true' : void 0)}
          aria-busy=${ifDefined(this.loading ? 'true' : void 0)}
          tabindex=${ifDefined(this.isUnavailable ? '-1' : void 0)}
          @click=${this.handleClick}
        >
          ${this.renderContent()}
        </a>
      `;
        return html`
      <button
        class=${controlClass}
        part="control"
        type="button"
        ?disabled=${this.isUnavailable}
        aria-label=${ifDefined(this.controlAriaLabel)}
        aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
        aria-describedby=${ifDefined(this.controlAriaDescribedby)}
        aria-busy=${ifDefined(this.loading ? 'true' : void 0)}
        @click=${this.handleClick}
      >
        ${this.renderContent()}
      </button>
    `;
    }
}
export { VE_BUTTON_TAG, VeButtonElement };
