// @ve-design/web/dist/ve-actions.js
import { css, html, nothing } from "lit";
import { property, query, state } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { moreHorizontalIcon } from "./icons/more-horizontal.js";
const veActionsToken = css`
  :host {
    /* ── layout ── */
    --ve-actions-gap: var(--space-xxxs);
    --ve-actions-extra-padding-inline: var(--space-xxxs);
    --ve-actions-btn-size: 28px;
    --ve-actions-icon-size: 16px;

    /* ── button surface ── */
    --ve-actions-btn-radius: var(--radius-md);
    --ve-actions-btn-radius-sm: var(--radius-sm);
    --ve-actions-btn-bg-hover: var(--color-bg-surface);
    --ve-actions-btn-bg-active: var(--color-bg-muted);

    /* ── icon ── */
    --ve-actions-icon-color: var(--color-text-tertiary);

    /* ── disabled ── */
    --ve-actions-disabled-opacity: var(--state-disabled-ratio);

    /* ── divider ── */
    --ve-actions-divider-color: var(--ve-actions-icon-color);
    --ve-actions-divider-height: 12px;
  }
`;
const token = veActionsToken;
const ve_actions = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      align-items: center;
      position: relative;
    }

    /* ── root bar ── */
    .ve-actions {
      display: flex;
      align-items: center;
      justify-content: space-between;
      inline-size: 100%;
      gap: var(--ve-actions-gap);
    }

    .ve-actions-children {
      display: flex;
      align-items: center;
      gap: var(--ve-actions-gap);
      flex: 1 1 auto;
      min-inline-size: 0;
    }

    .ve-actions-extra {
      display: flex;
      align-items: center;
      flex: none;
      padding-inline: var(--ve-actions-extra-padding-inline);
    }

    .ve-actions-extra-empty {
      display: none;
    }

    /* ── icon button base ── */
    .ve-actions-item {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      width: var(--ve-actions-btn-size);
      height: var(--ve-actions-btn-size);
      border-radius: var(--ve-actions-btn-radius);
      border: none;
      background: transparent;
      padding: 0;
      margin: 0;
      cursor: pointer;
      flex: none;
      color: var(--ve-actions-icon-color);
      transition: background-color 150ms ease;
      box-sizing: border-box;
      -webkit-appearance: none;
      appearance: none;
    }

    .ve-actions-item:hover {
      background: var(--ve-actions-btn-bg-hover);
    }

    .ve-actions-item:active {
      background: var(--ve-actions-btn-bg-active);
    }

    .ve-actions-item:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px color-mix(in srgb, currentColor 20%, transparent);
    }

    .ve-actions-item:disabled,
    .ve-actions-item[aria-disabled='true'] {
      opacity: var(--ve-actions-disabled-opacity);
      cursor: not-allowed;
      pointer-events: none;
    }

    /* ── icon wrapper ── */
    .ve-actions-item-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-actions-icon-size);
      block-size: var(--ve-actions-icon-size);
      line-height: 1;
      color: inherit;
    }

    .ve-actions-item-icon svg {
      inline-size: 100%;
      block-size: 100%;
    }

    /* ── text item ── */
    .ve-actions-item--text {
      width: auto;
      padding-inline: 8px;
      cursor: default;
    }

    .ve-actions-item--text:hover {
      background: transparent;
    }

    .ve-actions-item-text {
      display: inline-flex;
      align-items: center;
      white-space: nowrap;
      font-size: 13px;
      font-weight: 400;
      line-height: 150%;
      letter-spacing: 0;
      color: inherit;
    }

    /* ── divider ── */
    .ve-actions-divider {
      display: inline-block;
      vertical-align: middle;
      border-inline-start: 1px solid var(--ve-actions-divider-color);
      min-inline-size: 1px;
      max-inline-size: 1px;
      block-size: var(--ve-actions-divider-height);
      flex: none;
    }

    /* ── more item ── */
    .ve-actions-more {
      display: inline-flex;
      align-items: center;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _init_items, _init_extra__init_items, _init_extraSlot, _init_extra__init_extraSlot, _init_hasExtra, _init_extra__init_hasExtra;
const VE_ACTIONS_TAG = 've-actions';
_dec = property({
    attribute: false
}), _dec1 = query('slot[name="extra"]'), _dec2 = state();
class VeActionsElement extends VeElement {
    static{
        ({ e: [_init_items, _init_extra__init_items, _init_extraSlot, _init_extra__init_extraSlot, _init_hasExtra, _init_extra__init_hasExtra] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "items"
            ],
            [
                _dec1,
                1,
                "extraSlot"
            ],
            [
                _dec2,
                1,
                "hasExtra"
            ]
        ]));
    }
    static styles = ve_actions;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_items_1 = (()=>{
        const _value = _init_items(this, []);
        _init_extra__init_items(this);
        return _value;
    })();
    get items() {
        return this.#___private_items_1;
    }
    set items(_v) {
        this.#___private_items_1 = _v;
    }
    #___private_extraSlot_2 = (()=>{
        const _value = _init_extraSlot(this);
        _init_extra__init_extraSlot(this);
        return _value;
    })();
    get extraSlot() {
        return this.#___private_extraSlot_2;
    }
    set extraSlot(_v) {
        this.#___private_extraSlot_2 = _v;
    }
    #___private_hasExtra_3 = (()=>{
        const _value = _init_hasExtra(this, false);
        _init_extra__init_hasExtra(this);
        return _value;
    })();
    get hasExtra() {
        return this.#___private_hasExtra_3;
    }
    set hasExtra(_v) {
        this.#___private_hasExtra_3 = _v;
    }
    syncSlots = ()=>{
        const nodes = this.extraSlot?.assignedNodes({
            flatten: true
        }) ?? [];
        this.hasExtra = nodes.some((node)=>{
            if (node.nodeType === Node.ELEMENT_NODE) return true;
            if (node.nodeType === Node.TEXT_NODE) return (node.textContent ?? '').trim().length > 0;
            return false;
        });
    };
    firstUpdated() {
        this.syncSlots();
    }
    isTextIcon(icon) {
        if ('string' != typeof icon) return false;
        const template = document.createElement('template');
        template.innerHTML = icon.trim();
        return !template.content.firstElementChild;
    }
    renderIcon(icon) {
        if (null == icon) return nothing;
        if ('string' == typeof icon) {
            const template = document.createElement('template');
            template.innerHTML = icon.trim();
            const node = template.content.firstElementChild;
            if (node) return html`<span class="ve-actions-item-icon">${node}</span>`;
            return html`<span class="ve-actions-item-text">${icon}</span>`;
        }
        return html`<span class="ve-actions-item-icon">${icon}</span>`;
    }
    handleItemClick(item, event) {
        if (item.disabled) return;
        if (item.onClick) item.onClick(item, event);
        this.dispatchEvent(new CustomEvent('ve-actions-click', {
            detail: {
                item
            },
            bubbles: true,
            composed: true
        }));
    }
    renderItem(item) {
        if (item.divider) return html`<span
        class="ve-actions-divider"
        part="divider"
        role="separator"
        aria-orientation="vertical"
      ></span>`;
        if (item.text) return html`<span
        class="ve-actions-item ve-actions-item--text"
        part="item"
      >
        <span class="ve-actions-item-text">${item.text}</span>
      </span>`;
        if ('more' === item.type) {
            const menu = item.menu?.cloneNode(true);
            const moreIconContent = item.icon ? this.renderIcon(item.icon) : html`<span class="ve-actions-item-icon"><ve-icon .definition=${moreHorizontalIcon}></ve-icon></span>`;
            return html`<ve-dropdown
        class="ve-actions-more"
        part="more"
        trigger="click"
        position="bottom-start"
      >
        <button
          slot="trigger"
          class=${classMap({
                've-actions-item': true
            })}
          type="button"
          part="item"
          ?disabled=${item.disabled ?? false}
        >
          ${moreIconContent}
        </button>
        ${menu ?? nothing}
      </ve-dropdown>`;
        }
        return html`<button
      class=${classMap({
            've-actions-item': true,
            've-actions-item--text': this.isTextIcon(item.icon)
        })}
      part="item"
      type="button"
      ?disabled=${item.disabled ?? false}
      @click=${(event)=>this.handleItemClick(item, event)}
    >
      ${this.renderIcon(item.icon)}
    </button>`;
    }
    render() {
        return html`<div class="ve-actions" part="root">
      <div class="ve-actions-children" part="children">
        ${this.items.map((item)=>this.renderItem(item))}
      </div>
      <div
        class=${classMap({
            've-actions-extra': true,
            've-actions-extra-empty': !this.hasExtra
        })}
        part="extra"
      >
        <slot name="extra" @slotchange=${this.syncSlots}></slot>
      </div>
    </div>`;
    }
}
defineCustomElement(VE_ACTIONS_TAG, VeActionsElement);
export { VE_ACTIONS_TAG, VeActionsElement };
