// @ve-design/web/dist/ve-bubble.js
import { css, html, nothing } from "lit";
import { property, query, state } from "lit/decorators.js";
import { classMap } from "lit/directives/class-map.js";
import { styleMap } from "lit/directives/style-map.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
const veBubbleTokenDefaults = css`
  :host {
    /* ── 容器布局 ── */
    --ve-bubble-gap: var(--space-xxs, 8px);
    --ve-bubble-list-gap: var(--space-xl, 32px);

    /* ── 头像 ── */
    --ve-bubble-avatar-size: 32px;
    --ve-bubble-avatar-radius: var(--radius-full, 99px);

    /* ── header / footer ── */
    --ve-bubble-header-font-size: 14px;
    --ve-bubble-header-line-height: 22px;
    --ve-bubble-header-color: var(--color-text-tertiary, #7a7880);
    --ve-bubble-header-margin-block-start: var(--space-xxxs, 4px);
    --ve-bubble-header-margin-block-end: var(--space-xxs, 8px);

    --ve-bubble-footer-font-size: 14px;
    --ve-bubble-footer-line-height: 22px;
    --ve-bubble-footer-color: var(--color-text-tertiary, #7a7880);
    --ve-bubble-footer-margin-block-start: var(--space-xxxs, 4px);

    /* ── 内容 ── */
    --ve-bubble-content-font-size: 15px;
    --ve-bubble-content-line-height: var(--line-height-body, 1.5);
    --ve-bubble-content-text-color: var(--color-text-primary, #08080c);
    --ve-bubble-content-padding-block: var(--space-xs, 12px);
    --ve-bubble-content-padding-inline: var(--space-s, 16px);
    --ve-bubble-content-max-width: 450px;

    /* ── filled ── */
    --ve-bubble-content-filled-bg: var(--color-bg-surface, #f4f4fb);
    /* ── outlined ── */
    --ve-bubble-content-outlined-border: var(--color-border-default, #d2d1e0);
    --ve-bubble-content-outlined-border-width: 1px;

    /* ── shape 半径 ── */
    --ve-bubble-content-radius-default: var(--radius-lg, 12px);
    --ve-bubble-content-radius-round: 52px;
    --ve-bubble-content-radius-corner: 2px;

    /* ── loading ── */
    --ve-bubble-loading-dot-size: 6px;
    --ve-bubble-loading-dot-gap: 4px;
    --ve-bubble-loading-dot-color: var(--color-text-tertiary, #7a7880);

    /* ── expand/collapse ── */
    --ve-bubble-collapsed-max-height: 216px;
    --ve-bubble-overlay-height: 76px;
    --ve-bubble-overlay-padding-block-end: var(--space-s, 16px);
    --ve-bubble-overlay-surface: var(--ve-bubble-content-filled-bg);
    --ve-bubble-overlay-fade: transparent;
    --ve-bubble-toggle-height: 28px;
    --ve-bubble-toggle-font-size: 13px;
    --ve-bubble-toggle-color: var(--color-text-tertiary, #7a7880);
    --ve-bubble-toggle-color-hover: var(--color-text-secondary, #4e5969);

    /* ── quote（引用） ── */
    --ve-bubble-quote-gap: var(--space-xxs, 8px);
    --ve-bubble-quote-inner-gap: var(--space-xxxs, 4px);
    --ve-bubble-quote-bar-width: 2px;
    --ve-bubble-quote-bar-height: 14px;
    --ve-bubble-quote-bar-radius: var(--radius-xs, 2px);
    --ve-bubble-quote-bar-color: var(--color-border-default, #d2d1e0);
    --ve-bubble-quote-text: var(--color-text-tertiary, #7a7880);
    --ve-bubble-quote-font-size: var(--text-body, 14px);
    --ve-bubble-quote-line-height: 22px;
    --ve-bubble-quote-margin-block-end: var(--space-xxs, 8px);

    /* ── edit actions ── */
    --ve-bubble-edit-action-min-width: 48px;
    --ve-bubble-edit-action-font-weight: var(--font-weight-normal, 400);
  }
`;
const token = veBubbleTokenDefaults;
const ve_bubble = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      box-sizing: border-box;
      max-inline-size: 100%;
      inline-size: fit-content;
      align-self: flex-start;
      vertical-align: top;
      font-size: var(--ve-bubble-content-font-size);
      line-height: var(--ve-bubble-content-line-height);
      font-family: var(
        --font-sans,
        'PingFang SC',
        -apple-system,
        BlinkMacSystemFont,
        'Segoe UI',
        'Hiragino Sans GB',
        'Microsoft YaHei',
        sans-serif
      );
      color: var(--ve-bubble-content-text-color);
    }

    .ve-bubble {
      display: flex;
      gap: var(--ve-bubble-gap);
      max-inline-size: 100%;
      align-items: flex-start;
    }

    /* loading 时让 avatar 与 loading 占位垂直居中 */
    :host([loading]) .ve-bubble {
      align-items: center;
    }

    /* ── placement ── */
    :host([placement='start']) .ve-bubble {
      flex-direction: row;
    }

    :host([placement='end']) {
      align-self: flex-end;
    }

    :host([placement='end']) .ve-bubble {
      flex-direction: row-reverse;
    }

    /* ── avatar ── */
    .ve-bubble-avatar {
      flex: 0 0 auto;
      min-inline-size: var(--ve-bubble-avatar-size);
      block-size: var(--ve-bubble-avatar-size);
      border-radius: var(--ve-bubble-avatar-radius);
      overflow: hidden;
      display: inline-flex;
      align-items: center;
      justify-content: center;
    }

    /* 没有 avatar 时折叠占位 */
    .ve-bubble-avatar-empty {
      display: none;
    }

    /* ── body ── */
    .ve-bubble-body {
      display: flex;
      flex-direction: column;
      align-items: flex-start;
      inline-size: fit-content;
      max-inline-size: 100%;
      min-inline-size: 0;
    }

    :host([placement='end']) .ve-bubble-body {
      align-items: flex-end;
    }

    /* placement=end 时 header 内部翻转 */
    :host([placement='end']) .ve-bubble-header {
      flex-direction: row-reverse;
    }

    /* ── header ── */
    .ve-bubble-header {
      display: flex;
      align-items: center;
      gap: var(--space-xxs, 8px);
      font-size: var(--ve-bubble-header-font-size);
      line-height: var(--ve-bubble-header-line-height);
      color: var(--ve-bubble-header-color);
      margin: var(--ve-bubble-header-margin-block-start) 0
        var(--ve-bubble-header-margin-block-end) 0;
    }

    .ve-bubble-header-empty {
      display: none;
    }

    /* ── footer ── */
    .ve-bubble-footer {
      font-size: var(--ve-bubble-footer-font-size);
      line-height: var(--ve-bubble-footer-line-height);
      color: var(--ve-bubble-footer-color);
      margin-block-start: var(--ve-bubble-footer-margin-block-start);
    }

    .ve-bubble-footer-empty {
      display: none;
    }

    /* ── content ── */
    .ve-bubble-content {
      font-size: var(--ve-bubble-content-font-size);
      line-height: var(--ve-bubble-content-line-height);
      color: var(--ve-bubble-content-text-color);
      padding-block: var(--ve-bubble-content-padding-block);
      padding-inline: var(--ve-bubble-content-padding-inline);
      word-break: break-all;
      box-sizing: border-box;
    }

    /* variant: filled */
    .ve-bubble-content-filled {
      background-color: var(--ve-bubble-content-filled-bg);
    }

    /* variant: outlined */
    .ve-bubble-content-outlined {
      background-color: transparent;
      border: var(--ve-bubble-content-outlined-border-width) solid
        var(--ve-bubble-content-outlined-border);
    }

    /* variant: text */
    .ve-bubble-content-text {
      background-color: transparent;
      border: 0;
      padding: 0;
    }

    /* shape: default */
    .ve-bubble-content-default {
      border-radius: var(--ve-bubble-content-radius-default);
      max-inline-size: var(--ve-bubble-content-max-width);
    }

    /* shape: round */
    .ve-bubble-content-round {
      border-radius: var(--ve-bubble-content-radius-round);
    }

    /* shape: corner（固定右上 2px 锚角，与 React 版本保持一致） */
    .ve-bubble-content-corner {
      border-start-start-radius: var(--ve-bubble-content-radius-default);
      border-start-end-radius: var(--ve-bubble-content-radius-corner);
      border-end-end-radius: var(--ve-bubble-content-radius-default);
      border-end-start-radius: var(--ve-bubble-content-radius-default);
    }

    /* ── loading ── */
    .ve-bubble-loading {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-bubble-loading-dot-gap);
      padding-block: var(--ve-bubble-content-padding-block);
      padding-inline: var(--ve-bubble-content-padding-inline);
      color: var(--ve-bubble-loading-dot-color);
      font-size: var(--ve-bubble-content-font-size);
      line-height: var(--ve-bubble-content-line-height);
    }

    .ve-bubble-loading-dots {
      display: inline-flex;
      gap: var(--ve-bubble-loading-dot-gap);
    }

    .ve-bubble-loading-dot {
      inline-size: var(--ve-bubble-loading-dot-size);
      block-size: var(--ve-bubble-loading-dot-size);
      border-radius: 50%;
      background: currentColor;
      animation: ve-bubble-loading-bounce 1.2s infinite ease-in-out both;
    }

    .ve-bubble-loading-dot:nth-child(1) {
      animation-delay: -0.32s;
    }

    .ve-bubble-loading-dot:nth-child(2) {
      animation-delay: -0.16s;
    }

    @keyframes ve-bubble-loading-bounce {
      0%,
      80%,
      100% {
        transform: scale(0.6);
        opacity: 0.5;
      }
      40% {
        transform: scale(1);
        opacity: 1;
      }
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-bubble-loading-dot {
        animation: none;
        opacity: 0.7;
      }
    }

    /* ── expand/collapse ── */
    .ve-bubble-content-overflowing {
      position: relative;
    }

    .ve-bubble-content-collapsed {
      block-size: var(
        --ve-bubble-current-max-height,
        var(--ve-bubble-collapsed-max-height)
      );
      overflow: hidden;
    }

    .ve-bubble-content-text.ve-bubble-content-collapsed {
      border-end-start-radius: var(--ve-bubble-content-radius-default);
      border-end-end-radius: var(--ve-bubble-content-radius-default);
    }

    .ve-bubble-content-inner {
      box-sizing: border-box;
    }

    .ve-bubble-content-collapsed .ve-bubble-content-inner {
      -webkit-mask-image: linear-gradient(
        to bottom,
        black 0,
        black calc(100% - var(--ve-bubble-overlay-height)),
        transparent
          calc(
            100% - var(--ve-bubble-toggle-height) -
              var(--ve-bubble-overlay-padding-block-end)
          ),
        transparent 100%
      );
      mask-image: linear-gradient(
        to bottom,
        black 0,
        black calc(100% - var(--ve-bubble-overlay-height)),
        transparent
          calc(
            100% - var(--ve-bubble-toggle-height) -
              var(--ve-bubble-overlay-padding-block-end)
          ),
        transparent 100%
      );
    }

    /* ── quote（引用，参考 ve-chat-bubble） ── */
    .ve-bubble-quote {
      display: flex;
      align-items: center;
      gap: var(--ve-bubble-quote-gap);
      width: 100%;
      margin-block-end: var(--ve-bubble-quote-margin-block-end);
    }

    .ve-bubble-quote-bar {
      flex: none;
      inline-size: var(--ve-bubble-quote-bar-width);
      block-size: var(--ve-bubble-quote-bar-height);
      border-radius: var(--ve-bubble-quote-bar-radius);
      background: var(--ve-bubble-quote-bar-color);
    }

    .ve-bubble-quote-inner {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-bubble-quote-inner-gap);
      min-width: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      color: var(--ve-bubble-quote-text);
      font-size: var(--ve-bubble-quote-font-size);
      font-weight: var(--font-weight-normal, 400);
      line-height: var(--ve-bubble-quote-line-height);
    }

    /* ── toggle button 通用样式 ── */
    .ve-bubble-toggle {
      display: flex;
      align-items: center;
      block-size: var(--ve-bubble-toggle-height);
      padding: 0;
      border: none;
      background: none;
      font-size: var(--ve-bubble-toggle-font-size);
      color: var(--ve-bubble-toggle-color);
      cursor: pointer;
      transition: color var(--motion-fast, 150ms) var(--ease-standard, ease);
    }

    .ve-bubble-toggle:hover {
      color: var(--ve-bubble-toggle-color-hover);
    }

    /* 收起态：渐变遮罩覆盖在 content 底部 */
    .ve-bubble-toggle-overlay {
      position: absolute;
      inset-inline-start: 0;
      inset-block-end: 0;
      inline-size: 100%;
      block-size: var(--ve-bubble-overlay-height);
      align-items: flex-end;
      justify-content: flex-start;
      padding-block-end: var(--ve-bubble-overlay-padding-block-end);
      padding-inline-start: var(--ve-bubble-content-padding-inline);
      border-radius: 0;
      background: linear-gradient(
        to top,
        var(--ve-bubble-overlay-surface) 0%,
        var(--ve-bubble-overlay-surface) 54%,
        var(--ve-bubble-overlay-fade) 100%
      );
    }

    /* 展开态：按钮在内容下方 */
    .ve-bubble-content-expanded .ve-bubble-toggle {
      margin-block-start: 4px;
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-bubble-toggle {
        transition: none;
      }
    }

    /* ── edit mode（编辑态） ── */
    .ve-bubble-edit {
      display: flex;
      flex-direction: column;
      gap: 12px;
      inline-size: 100%;
      min-inline-size: 300px;
      padding: var(--ve-bubble-content-padding-block)
        var(--ve-bubble-content-padding-inline);
      border-radius: var(--ve-bubble-content-radius-default);
      background-color: var(--ve-bubble-content-filled-bg);
      box-sizing: border-box;
    }

    .ve-bubble-edit-textarea {
      --ve-input-border-color: transparent;
      --ve-input-bg: transparent;
      --ve-input-padding-inline: 0;
      --ve-input-padding-block: 0;
    }

    .ve-bubble-edit-actions {
      display: flex;
      justify-content: flex-end;
      gap: 8px;
    }

    .ve-bubble-edit-actions ve-button {
      min-inline-size: var(--ve-bubble-edit-action-min-width);
      --ve-button-font-weight: var(--ve-bubble-edit-action-font-weight);
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _dec24, _dec25, _dec26, _dec27, _dec28, _dec29, _init_placement, _init_extra__init_placement, _init_variant, _init_extra__init_variant, _init_shape, _init_extra__init_shape, _init_content, _init_extra__init_content, _init_loading, _init_extra__init_loading, _init_maxHeight, _init_extra__init_maxHeight, _init_seeMoreLabel, _init_extra__init_seeMoreLabel, _init_seeLessLabel, _init_extra__init_seeLessLabel, _init_hasQuote, _init_extra__init_hasQuote, _init_quoteLabel, _init_extra__init_quoteLabel, _init_editable, _init_extra__init_editable, _init_editConfirmLabel, _init_extra__init_editConfirmLabel, _init_editCancelLabel, _init_extra__init_editCancelLabel, _init__editValue, _init_extra__init__editValue, _init__editMinWidth, _init_extra__init__editMinWidth, _init__expanded, _init_extra__init__expanded, _init__overflowing, _init_extra__init__overflowing, _init_contentEl, _init_extra__init_contentEl, _init_contentInnerEl, _init_extra__init_contentInnerEl, _init_editInputEl, _init_extra__init_editInputEl, _init_avatarSlot, _init_extra__init_avatarSlot, _init_headerSlot, _init_extra__init_headerSlot, _init_footerSlot, _init_extra__init_footerSlot, _init_defaultSlot, _init_extra__init_defaultSlot, _init_loadingSlot, _init_extra__init_loadingSlot, _init_hasAvatar, _init_extra__init_hasAvatar, _init_hasHeader, _init_extra__init_hasHeader, _init_hasFooter, _init_extra__init_hasFooter, _init_hasDefaultSlotContent, _init_extra__init_hasDefaultSlotContent, _init_hasLoadingSlotContent, _init_extra__init_hasLoadingSlotContent;
const VE_BUBBLE_TAG = 've-bubble';
const DEFAULT_MAX_HEIGHT = '216px';
_dec = property({
    type: String,
    reflect: true
}), _dec1 = property({
    type: String,
    reflect: true
}), _dec2 = property({
    type: String,
    reflect: true
}), _dec3 = property({
    type: String
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = property({
    type: String,
    attribute: 'max-height'
}), _dec6 = property({
    type: String,
    attribute: 'see-more-label'
}), _dec7 = property({
    type: String,
    attribute: 'see-less-label'
}), _dec8 = property({
    type: Boolean,
    reflect: true,
    attribute: 'has-quote'
}), _dec9 = property({
    type: String,
    attribute: 'quote-label'
}), _dec10 = property({
    type: Boolean,
    reflect: true
}), _dec11 = property({
    type: String,
    attribute: 'edit-confirm-label'
}), _dec12 = property({
    type: String,
    attribute: 'edit-cancel-label'
}), _dec13 = state(), _dec14 = state(), _dec15 = state(), _dec16 = state(), _dec17 = query('.ve-bubble-content'), _dec18 = query('.ve-bubble-content-inner'), _dec19 = query('.ve-bubble-edit-textarea'), _dec20 = query('slot[name="avatar"]'), _dec21 = query('slot[name="header"]'), _dec22 = query('slot[name="footer"]'), _dec23 = query('slot:not([name])'), _dec24 = query('slot[name="loading"]'), _dec25 = state(), _dec26 = state(), _dec27 = state(), _dec28 = state(), _dec29 = state();
class VeBubbleElement extends VeElement {
    static{
        ({ e: [_init_placement, _init_extra__init_placement, _init_variant, _init_extra__init_variant, _init_shape, _init_extra__init_shape, _init_content, _init_extra__init_content, _init_loading, _init_extra__init_loading, _init_maxHeight, _init_extra__init_maxHeight, _init_seeMoreLabel, _init_extra__init_seeMoreLabel, _init_seeLessLabel, _init_extra__init_seeLessLabel, _init_hasQuote, _init_extra__init_hasQuote, _init_quoteLabel, _init_extra__init_quoteLabel, _init_editable, _init_extra__init_editable, _init_editConfirmLabel, _init_extra__init_editConfirmLabel, _init_editCancelLabel, _init_extra__init_editCancelLabel, _init__editValue, _init_extra__init__editValue, _init__editMinWidth, _init_extra__init__editMinWidth, _init__expanded, _init_extra__init__expanded, _init__overflowing, _init_extra__init__overflowing, _init_contentEl, _init_extra__init_contentEl, _init_contentInnerEl, _init_extra__init_contentInnerEl, _init_editInputEl, _init_extra__init_editInputEl, _init_avatarSlot, _init_extra__init_avatarSlot, _init_headerSlot, _init_extra__init_headerSlot, _init_footerSlot, _init_extra__init_footerSlot, _init_defaultSlot, _init_extra__init_defaultSlot, _init_loadingSlot, _init_extra__init_loadingSlot, _init_hasAvatar, _init_extra__init_hasAvatar, _init_hasHeader, _init_extra__init_hasHeader, _init_hasFooter, _init_extra__init_hasFooter, _init_hasDefaultSlotContent, _init_extra__init_hasDefaultSlotContent, _init_hasLoadingSlotContent, _init_extra__init_hasLoadingSlotContent] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "placement"
            ],
            [
                _dec1,
                1,
                "variant"
            ],
            [
                _dec2,
                1,
                "shape"
            ],
            [
                _dec3,
                1,
                "content"
            ],
            [
                _dec4,
                1,
                "loading"
            ],
            [
                _dec5,
                1,
                "maxHeight"
            ],
            [
                _dec6,
                1,
                "seeMoreLabel"
            ],
            [
                _dec7,
                1,
                "seeLessLabel"
            ],
            [
                _dec8,
                1,
                "hasQuote"
            ],
            [
                _dec9,
                1,
                "quoteLabel"
            ],
            [
                _dec10,
                1,
                "editable"
            ],
            [
                _dec11,
                1,
                "editConfirmLabel"
            ],
            [
                _dec12,
                1,
                "editCancelLabel"
            ],
            [
                _dec13,
                1,
                "_editValue"
            ],
            [
                _dec14,
                1,
                "_editMinWidth"
            ],
            [
                _dec15,
                1,
                "_expanded"
            ],
            [
                _dec16,
                1,
                "_overflowing"
            ],
            [
                _dec17,
                1,
                "contentEl"
            ],
            [
                _dec18,
                1,
                "contentInnerEl"
            ],
            [
                _dec19,
                1,
                "editInputEl"
            ],
            [
                _dec20,
                1,
                "avatarSlot"
            ],
            [
                _dec21,
                1,
                "headerSlot"
            ],
            [
                _dec22,
                1,
                "footerSlot"
            ],
            [
                _dec23,
                1,
                "defaultSlot"
            ],
            [
                _dec24,
                1,
                "loadingSlot"
            ],
            [
                _dec25,
                1,
                "hasAvatar"
            ],
            [
                _dec26,
                1,
                "hasHeader"
            ],
            [
                _dec27,
                1,
                "hasFooter"
            ],
            [
                _dec28,
                1,
                "hasDefaultSlotContent"
            ],
            [
                _dec29,
                1,
                "hasLoadingSlotContent"
            ]
        ]));
    }
    static styles = ve_bubble;
    #___private_placement_1 = (()=>{
        const _value = _init_placement(this, 'start');
        _init_extra__init_placement(this);
        return _value;
    })();
    get placement() {
        return this.#___private_placement_1;
    }
    set placement(_v) {
        this.#___private_placement_1 = _v;
    }
    #___private_variant_2 = (()=>{
        const _value = _init_variant(this, 'filled');
        _init_extra__init_variant(this);
        return _value;
    })();
    get variant() {
        return this.#___private_variant_2;
    }
    set variant(_v) {
        this.#___private_variant_2 = _v;
    }
    #___private_shape_3 = (()=>{
        const _value = _init_shape(this, 'default');
        _init_extra__init_shape(this);
        return _value;
    })();
    get shape() {
        return this.#___private_shape_3;
    }
    set shape(_v) {
        this.#___private_shape_3 = _v;
    }
    #___private_content_4 = (()=>{
        const _value = _init_content(this, null);
        _init_extra__init_content(this);
        return _value;
    })();
    get content() {
        return this.#___private_content_4;
    }
    set content(_v) {
        this.#___private_content_4 = _v;
    }
    #___private_loading_5 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_5;
    }
    set loading(_v) {
        this.#___private_loading_5 = _v;
    }
    #___private_maxHeight_6 = (()=>{
        const _value = _init_maxHeight(this, DEFAULT_MAX_HEIGHT);
        _init_extra__init_maxHeight(this);
        return _value;
    })();
    get maxHeight() {
        return this.#___private_maxHeight_6;
    }
    set maxHeight(_v) {
        this.#___private_maxHeight_6 = _v;
    }
    #___private_seeMoreLabel_7 = (()=>{
        const _value = _init_seeMoreLabel(this, '查看更多');
        _init_extra__init_seeMoreLabel(this);
        return _value;
    })();
    get seeMoreLabel() {
        return this.#___private_seeMoreLabel_7;
    }
    set seeMoreLabel(_v) {
        this.#___private_seeMoreLabel_7 = _v;
    }
    #___private_seeLessLabel_8 = (()=>{
        const _value = _init_seeLessLabel(this, '收起');
        _init_extra__init_seeLessLabel(this);
        return _value;
    })();
    get seeLessLabel() {
        return this.#___private_seeLessLabel_8;
    }
    set seeLessLabel(_v) {
        this.#___private_seeLessLabel_8 = _v;
    }
    #___private_hasQuote_9 = (()=>{
        const _value = _init_hasQuote(this, false);
        _init_extra__init_hasQuote(this);
        return _value;
    })();
    get hasQuote() {
        return this.#___private_hasQuote_9;
    }
    set hasQuote(_v) {
        this.#___private_hasQuote_9 = _v;
    }
    #___private_quoteLabel_10 = (()=>{
        const _value = _init_quoteLabel(this, '');
        _init_extra__init_quoteLabel(this);
        return _value;
    })();
    get quoteLabel() {
        return this.#___private_quoteLabel_10;
    }
    set quoteLabel(_v) {
        this.#___private_quoteLabel_10 = _v;
    }
    #___private_editable_11 = (()=>{
        const _value = _init_editable(this, false);
        _init_extra__init_editable(this);
        return _value;
    })();
    get editable() {
        return this.#___private_editable_11;
    }
    set editable(_v) {
        this.#___private_editable_11 = _v;
    }
    #___private_editConfirmLabel_12 = (()=>{
        const _value = _init_editConfirmLabel(this, '发送');
        _init_extra__init_editConfirmLabel(this);
        return _value;
    })();
    get editConfirmLabel() {
        return this.#___private_editConfirmLabel_12;
    }
    set editConfirmLabel(_v) {
        this.#___private_editConfirmLabel_12 = _v;
    }
    #___private_editCancelLabel_13 = (()=>{
        const _value = _init_editCancelLabel(this, '取消');
        _init_extra__init_editCancelLabel(this);
        return _value;
    })();
    get editCancelLabel() {
        return this.#___private_editCancelLabel_13;
    }
    set editCancelLabel(_v) {
        this.#___private_editCancelLabel_13 = _v;
    }
    #___private__editValue_14 = (()=>{
        const _value = _init__editValue(this, '');
        _init_extra__init__editValue(this);
        return _value;
    })();
    get _editValue() {
        return this.#___private__editValue_14;
    }
    set _editValue(_v) {
        this.#___private__editValue_14 = _v;
    }
    #___private__editMinWidth_15 = (()=>{
        const _value = _init__editMinWidth(this, null);
        _init_extra__init__editMinWidth(this);
        return _value;
    })();
    get _editMinWidth() {
        return this.#___private__editMinWidth_15;
    }
    set _editMinWidth(_v) {
        this.#___private__editMinWidth_15 = _v;
    }
    #___private__expanded_16 = (()=>{
        const _value = _init__expanded(this, false);
        _init_extra__init__expanded(this);
        return _value;
    })();
    get _expanded() {
        return this.#___private__expanded_16;
    }
    set _expanded(_v) {
        this.#___private__expanded_16 = _v;
    }
    #___private__overflowing_17 = (()=>{
        const _value = _init__overflowing(this, false);
        _init_extra__init__overflowing(this);
        return _value;
    })();
    get _overflowing() {
        return this.#___private__overflowing_17;
    }
    set _overflowing(_v) {
        this.#___private__overflowing_17 = _v;
    }
    #___private_contentEl_18 = (()=>{
        const _value = _init_contentEl(this);
        _init_extra__init_contentEl(this);
        return _value;
    })();
    get contentEl() {
        return this.#___private_contentEl_18;
    }
    set contentEl(_v) {
        this.#___private_contentEl_18 = _v;
    }
    #___private_contentInnerEl_19 = (()=>{
        const _value = _init_contentInnerEl(this);
        _init_extra__init_contentInnerEl(this);
        return _value;
    })();
    get contentInnerEl() {
        return this.#___private_contentInnerEl_19;
    }
    set contentInnerEl(_v) {
        this.#___private_contentInnerEl_19 = _v;
    }
    #___private_editInputEl_20 = (()=>{
        const _value = _init_editInputEl(this);
        _init_extra__init_editInputEl(this);
        return _value;
    })();
    get editInputEl() {
        return this.#___private_editInputEl_20;
    }
    set editInputEl(_v) {
        this.#___private_editInputEl_20 = _v;
    }
    #___private_avatarSlot_21 = (()=>{
        const _value = _init_avatarSlot(this);
        _init_extra__init_avatarSlot(this);
        return _value;
    })();
    get avatarSlot() {
        return this.#___private_avatarSlot_21;
    }
    set avatarSlot(_v) {
        this.#___private_avatarSlot_21 = _v;
    }
    #___private_headerSlot_22 = (()=>{
        const _value = _init_headerSlot(this);
        _init_extra__init_headerSlot(this);
        return _value;
    })();
    get headerSlot() {
        return this.#___private_headerSlot_22;
    }
    set headerSlot(_v) {
        this.#___private_headerSlot_22 = _v;
    }
    #___private_footerSlot_23 = (()=>{
        const _value = _init_footerSlot(this);
        _init_extra__init_footerSlot(this);
        return _value;
    })();
    get footerSlot() {
        return this.#___private_footerSlot_23;
    }
    set footerSlot(_v) {
        this.#___private_footerSlot_23 = _v;
    }
    #___private_defaultSlot_24 = (()=>{
        const _value = _init_defaultSlot(this);
        _init_extra__init_defaultSlot(this);
        return _value;
    })();
    get defaultSlot() {
        return this.#___private_defaultSlot_24;
    }
    set defaultSlot(_v) {
        this.#___private_defaultSlot_24 = _v;
    }
    #___private_loadingSlot_25 = (()=>{
        const _value = _init_loadingSlot(this);
        _init_extra__init_loadingSlot(this);
        return _value;
    })();
    get loadingSlot() {
        return this.#___private_loadingSlot_25;
    }
    set loadingSlot(_v) {
        this.#___private_loadingSlot_25 = _v;
    }
    _resizeObserver = null;
    _observedContentEl = null;
    #___private_hasAvatar_26 = (()=>{
        const _value = _init_hasAvatar(this, false);
        _init_extra__init_hasAvatar(this);
        return _value;
    })();
    get hasAvatar() {
        return this.#___private_hasAvatar_26;
    }
    set hasAvatar(_v) {
        this.#___private_hasAvatar_26 = _v;
    }
    #___private_hasHeader_27 = (()=>{
        const _value = _init_hasHeader(this, false);
        _init_extra__init_hasHeader(this);
        return _value;
    })();
    get hasHeader() {
        return this.#___private_hasHeader_27;
    }
    set hasHeader(_v) {
        this.#___private_hasHeader_27 = _v;
    }
    #___private_hasFooter_28 = (()=>{
        const _value = _init_hasFooter(this, false);
        _init_extra__init_hasFooter(this);
        return _value;
    })();
    get hasFooter() {
        return this.#___private_hasFooter_28;
    }
    set hasFooter(_v) {
        this.#___private_hasFooter_28 = _v;
    }
    #___private_hasDefaultSlotContent_29 = (()=>{
        const _value = _init_hasDefaultSlotContent(this, false);
        _init_extra__init_hasDefaultSlotContent(this);
        return _value;
    })();
    get hasDefaultSlotContent() {
        return this.#___private_hasDefaultSlotContent_29;
    }
    set hasDefaultSlotContent(_v) {
        this.#___private_hasDefaultSlotContent_29 = _v;
    }
    #___private_hasLoadingSlotContent_30 = (()=>{
        const _value = _init_hasLoadingSlotContent(this, false);
        _init_extra__init_hasLoadingSlotContent(this);
        return _value;
    })();
    get hasLoadingSlotContent() {
        return this.#___private_hasLoadingSlotContent_30;
    }
    set hasLoadingSlotContent(_v) {
        this.#___private_hasLoadingSlotContent_30 = _v;
    }
    static slotHasContent(slot) {
        if (!slot) return false;
        const nodes = slot.assignedNodes();
        return nodes.some((node)=>{
            if (node.nodeType === Node.ELEMENT_NODE) return true;
            if (node.nodeType === Node.TEXT_NODE) return (node.textContent ?? '').trim().length > 0;
            return false;
        });
    }
    updateSlotState = ()=>{
        this.hasAvatar = VeBubbleElement.slotHasContent(this.avatarSlot);
        this.hasHeader = VeBubbleElement.slotHasContent(this.headerSlot);
        this.hasFooter = VeBubbleElement.slotHasContent(this.footerSlot);
        this.hasDefaultSlotContent = VeBubbleElement.slotHasContent(this.defaultSlot);
        this.hasLoadingSlotContent = VeBubbleElement.slotHasContent(this.loadingSlot);
    };
    handleDefaultSlotChange = ()=>{
        this.updateSlotState();
        this.requestOverflowMeasure();
    };
    firstUpdated() {
        this.updateSlotState();
        this.updateContentObserver();
        this.requestOverflowMeasure();
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this._resizeObserver?.disconnect();
        this._resizeObserver = null;
        this._observedContentEl = null;
    }
    requestOverflowMeasure() {
        this.updateComplete.then(()=>{
            this.updateContentObserver();
            this.measureOverflow();
        });
    }
    updateContentObserver() {
        const inner = this.contentInnerEl;
        if (!inner || this._observedContentEl === inner) return;
        this._resizeObserver?.disconnect();
        this._resizeObserver = new ResizeObserver(()=>this.measureOverflow());
        this._resizeObserver.observe(inner);
        this._observedContentEl = inner;
    }
    resolveMaxHeight(content) {
        const probe = document.createElement('div');
        probe.style.position = 'absolute';
        probe.style.visibility = 'hidden';
        probe.style.pointerEvents = 'none';
        probe.style.blockSize = this.resolvedMaxHeight;
        content.append(probe);
        const height = probe.getBoundingClientRect().height;
        probe.remove();
        return height;
    }
    get resolvedMaxHeight() {
        const maxHeight = 'string' == typeof this.maxHeight ? this.maxHeight.trim() : '';
        return maxHeight || DEFAULT_MAX_HEIGHT;
    }
    measureOverflow() {
        const inner = this.contentInnerEl;
        if (!inner || this.loading || this.editable) {
            this._overflowing = false;
            this._expanded = false;
            return;
        }
        const content = this.contentEl ?? inner;
        const maxHeight = this.resolveMaxHeight(content);
        if (maxHeight <= 0) {
            this._overflowing = false;
            this._expanded = false;
            return;
        }
        const contentStyle = getComputedStyle(content);
        const paddingBlock = Number.parseFloat(contentStyle.paddingBlockStart) + Number.parseFloat(contentStyle.paddingBlockEnd);
        const contentHeight = inner.scrollHeight + paddingBlock;
        const overflowing = contentHeight > maxHeight;
        this._overflowing = overflowing;
        if (!overflowing) this._expanded = false;
    }
    renderContent() {
        const variant = this.variant;
        const shape = this.shape;
        const contentClasses = classMap({
            've-bubble-content': true,
            [`ve-bubble-content-${variant}`]: true,
            [`ve-bubble-content-${shape}`]: 'text' !== variant,
            've-bubble-content-overflowing': this._overflowing,
            've-bubble-content-collapsed': this._overflowing && !this._expanded,
            've-bubble-content-expanded': this._overflowing && this._expanded
        });
        const contentStyle = styleMap({
            '--ve-bubble-current-max-height': this.resolvedMaxHeight
        });
        return html`<div class=${contentClasses} part="content" style=${contentStyle}>
      <div class="ve-bubble-content-inner">
        ${this.renderQuote()}
        <slot @slotchange=${this.handleDefaultSlotChange}
          >${this.hasDefaultSlotContent ? nothing : this.content ?? nothing}</slot
        >
      </div>
      ${this.renderToggle()}
    </div>`;
    }
    renderQuote() {
        if (!this.hasQuote) return nothing;
        return html`
      <div class="ve-bubble-quote" part="quote">
        <span class="ve-bubble-quote-bar" aria-hidden="true"></span>
        <span class="ve-bubble-quote-inner">
          <slot name="quote">${this.quoteLabel}</slot>
        </span>
      </div>
    `;
    }
    renderToggle() {
        if (!this._overflowing) return nothing;
        if (!this._expanded) return html`
        <button
          class="ve-bubble-toggle ve-bubble-toggle-overlay"
          part="toggle overlay"
          aria-expanded="false"
          @click=${this.handleToggle}
        >${this.seeMoreLabel}</button>
      `;
        return html`
      <button
        class="ve-bubble-toggle"
        part="toggle"
        aria-expanded="true"
        @click=${this.handleToggle}
      >${this.seeLessLabel}</button>
    `;
    }
    handleToggle() {
        this._expanded = !this._expanded;
    }
    renderLoading() {
        return html`<div class="ve-bubble-loading" part="loading">
      <slot name="loading" @slotchange=${this.updateSlotState}></slot>
      ${this.hasLoadingSlotContent ? nothing : html`<span class="ve-bubble-loading-dots" aria-hidden="true">
          <span class="ve-bubble-loading-dot"></span>
          <span class="ve-bubble-loading-dot"></span>
          <span class="ve-bubble-loading-dot"></span>
        </span>`}
    </div>`;
    }
    renderEditable() {
        const editStyle = null != this._editMinWidth ? `min-width: ${this._editMinWidth}px;` : '';
        return html`
      <div class="ve-bubble-edit" part="edit" style=${editStyle}>
        <ve-input
          class="ve-bubble-edit-textarea"
          type="textarea"
          .value=${this._editValue}
          auto-size
          @ve-input=${this._handleEditInput}
        ></ve-input>
        <div class="ve-bubble-edit-actions" part="edit-actions">
          <ve-button
            type="outline"
            size="small"
            @click=${this._handleEditCancel}
          >${this.editCancelLabel}</ve-button>
          <ve-button
            type="primary"
            size="small"
            @click=${this._handleEditConfirm}
          >${this.editConfirmLabel}</ve-button>
        </div>
      </div>
    `;
    }
    _handleEditInput(e) {
        this._editValue = e.detail.value;
    }
    _handleEditConfirm() {
        this.dispatchEvent(new CustomEvent('ve-bubble-edit-confirm', {
            detail: {
                content: this._editValue
            },
            bubbles: true,
            composed: true
        }));
    }
    _handleEditCancel() {
        this.dispatchEvent(new CustomEvent('ve-bubble-edit-cancel', {
            bubbles: true,
            composed: true
        }));
    }
    willUpdate(changed) {
        if (changed.has('editable')) if (this.editable) {
            const prevEditable = changed.get('editable');
            if (true !== prevEditable) {
                const measured = this.contentEl?.getBoundingClientRect().width ?? 0;
                this._editMinWidth = measured > 0 ? Math.ceil(measured) : null;
            }
            this._editValue = this.content ?? '';
            import("./ve-input.js");
            import("./ve-button.js");
        } else this._editMinWidth = null;
    }
    updated(changed) {
        if (changed.has('content') || changed.has('maxHeight') || changed.has('loading') || changed.has('editable')) this.requestOverflowMeasure();
        if (changed.has('editable') && this.editable) this.updateComplete.then(()=>{
            const input = this.editInputEl;
            if (!input) return;
            const focusable = input;
            const doFocus = ()=>focusable.focus({
                    preventScroll: true
                });
            if (focusable.updateComplete) focusable.updateComplete.then(doFocus);
            else doFocus();
        });
    }
    render() {
        if (this.editable) return html`<div class="ve-bubble" part="root">
        <div
          class=${classMap({
            've-bubble-avatar': true,
            've-bubble-avatar-empty': !this.hasAvatar
        })}
          part="avatar"
        >
          <slot name="avatar" @slotchange=${this.updateSlotState}></slot>
        </div>
        <div class="ve-bubble-body" part="body">
          ${this.renderEditable()}
        </div>
      </div>`;
        return html`<div class="ve-bubble" part="root">
      <div
        class=${classMap({
            've-bubble-avatar': true,
            've-bubble-avatar-empty': !this.hasAvatar
        })}
        part="avatar"
      >
        <slot name="avatar" @slotchange=${this.updateSlotState}></slot>
      </div>
      <div class="ve-bubble-body" part="body">
        <div
          class=${classMap({
            've-bubble-header': true,
            've-bubble-header-empty': !this.hasHeader
        })}
          part="header"
        >
          <slot name="header" @slotchange=${this.updateSlotState}></slot>
        </div>
        ${this.loading ? this.renderLoading() : this.renderContent()}
        <div
          class=${classMap({
            've-bubble-footer': true,
            've-bubble-footer-empty': !this.hasFooter
        })}
          part="footer"
        >
          <slot name="footer" @slotchange=${this.updateSlotState}></slot>
        </div>
      </div>
    </div>`;
    }
}
const ve_bubble_list = [
    componentReset,
    token,
    css`
    :host {
      display: flex;
      flex-direction: column;
      align-items: flex-start;
      gap: var(--ve-bubble-list-gap);
      box-sizing: border-box;
      inline-size: 100%;
      position: relative;
    }

    ::slotted(*) {
      max-inline-size: 100%;
    }

    /* placement=end 的子项靠右对齐 */
    ::slotted([placement='end']) {
      align-self: flex-end;
    }

    /* ── 指示器通用样式（与 ve-message-divider preload pill 一致） ── */
    .ve-bubble-list__indicator {
      position: relative;
      display: flex;
      align-items: center;
      justify-content: center;
      align-self: center;
      inline-size: 48px;
      block-size: 32px;
      border: var(--stroke-0_5, 0.5px) solid var(--color-border-default, #d2d1e0);
      border-radius: var(--radius-full, 99px);
      background: color-mix(in srgb, var(--color-bg-overlay, #ffffff) 72%, transparent);
      color: var(--color-text-primary, #08080c);
      backdrop-filter: blur(6px);
      overflow: hidden;
      transition: opacity 0.2s ease;
      cursor: default;
    }

    /* SVG 图标尺寸 */
    .ve-bubble-list__indicator-icon {
      display: block;
      inline-size: 16px;
      block-size: 16px;
      fill: none;
      stroke: currentColor;
      stroke-width: 1.35px;
      stroke-linecap: round;
      stroke-linejoin: round;
    }

    /* ── Loading dots 动画（与 ve-message-divider 预加载中一致） ── */
    .ve-bubble-list__indicator--loading {
      cursor: default;
      position: sticky;
      inset-block-end: 20px;
    }

    .ve-bubble-list__loading-dot {
      fill: currentColor;
      stroke: none;
      transform-box: fill-box;
      transform-origin: center;
      animation: ve-bubble-list-dot-bounce 900ms ease-in-out infinite;
    }

    .ve-bubble-list__loading-dot:nth-child(2) {
      animation-delay: 120ms;
    }

    .ve-bubble-list__loading-dot:nth-child(3) {
      animation-delay: 240ms;
    }

    @keyframes ve-bubble-list-dot-bounce {
      0%,
      72%,
      100% {
        opacity: 1;
        transform: translateY(0);
      }
      28% {
        opacity: 1;
        transform: translateY(-2px);
      }
      48% {
        opacity: 0.56;
        transform: translateY(1px);
      }
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-bubble-list__loading-dot {
        animation: none;
      }
    }

    /* ── Scroll-to-bottom pill（与 ve-message-divider 预加载完成一致） ── */
    .ve-bubble-list__indicator--scroll {
      cursor: pointer;
      position: sticky;
      inset-block-end: 20px;
    }

    .ve-bubble-list__indicator--scroll:hover {
      background: color-mix(in srgb, var(--color-bg-overlay, #ffffff) 88%, transparent);
      border-color: var(--color-border-hover, #b8b7c8);
    }

    .ve-bubble-list__indicator--scroll:active {
      transform: scale(0.95);
    }

    .ve-bubble-list__indicator--scroll:focus-visible {
      outline: 2px solid var(--color-primary, #165dff);
      outline-offset: 2px;
    }
  `
];
function ve_bubble_list_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_bubble_list_element_toPropertyKey(arg) {
    var key = ve_bubble_list_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_bubble_list_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_bubble_list_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_bubble_list_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_bubble_list_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_bubble_list_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_bubble_list_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_bubble_list_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_bubble_list_element_dec, ve_bubble_list_element_dec1, ve_bubble_list_element_dec2, ve_bubble_list_element_dec3, _init_gap, _init_extra__init_gap, ve_bubble_list_element_init_loading, ve_bubble_list_element_init_extra_init_loading, _init_scrollIndicator, _init_extra__init_scrollIndicator, _init__showScrollBottom, _init_extra__init__showScrollBottom;
const VE_BUBBLE_LIST_TAG = 've-bubble-list';
ve_bubble_list_element_dec = property({
    type: String
}), ve_bubble_list_element_dec1 = property({
    type: Boolean,
    reflect: true
}), ve_bubble_list_element_dec2 = property({
    type: Boolean,
    attribute: 'scroll-indicator',
    reflect: true
}), ve_bubble_list_element_dec3 = state();
class VeBubbleListElement extends VeElement {
    static{
        ({ e: [_init_gap, _init_extra__init_gap, ve_bubble_list_element_init_loading, ve_bubble_list_element_init_extra_init_loading, _init_scrollIndicator, _init_extra__init_scrollIndicator, _init__showScrollBottom, _init_extra__init__showScrollBottom] } = ve_bubble_list_element_apply_decs_2311(this, [], [
            [
                ve_bubble_list_element_dec,
                1,
                "gap"
            ],
            [
                ve_bubble_list_element_dec1,
                1,
                "loading"
            ],
            [
                ve_bubble_list_element_dec2,
                1,
                "scrollIndicator"
            ],
            [
                ve_bubble_list_element_dec3,
                1,
                "_showScrollBottom"
            ]
        ]));
    }
    static styles = ve_bubble_list;
    #___private_gap_1 = (()=>{
        const _value = _init_gap(this, null);
        _init_extra__init_gap(this);
        return _value;
    })();
    get gap() {
        return this.#___private_gap_1;
    }
    set gap(_v) {
        this.#___private_gap_1 = _v;
    }
    #___private_loading_2 = (()=>{
        const _value = ve_bubble_list_element_init_loading(this, false);
        ve_bubble_list_element_init_extra_init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_2;
    }
    set loading(_v) {
        this.#___private_loading_2 = _v;
    }
    #___private_scrollIndicator_3 = (()=>{
        const _value = _init_scrollIndicator(this, false);
        _init_extra__init_scrollIndicator(this);
        return _value;
    })();
    get scrollIndicator() {
        return this.#___private_scrollIndicator_3;
    }
    set scrollIndicator(_v) {
        this.#___private_scrollIndicator_3 = _v;
    }
    #___private__showScrollBottom_4 = (()=>{
        const _value = _init__showScrollBottom(this, false);
        _init_extra__init__showScrollBottom(this);
        return _value;
    })();
    get _showScrollBottom() {
        return this.#___private__showScrollBottom_4;
    }
    set _showScrollBottom(_v) {
        this.#___private__showScrollBottom_4 = _v;
    }
    _scrollContainer = null;
    _resizeObserver = null;
    _rafId = null;
    connectedCallback() {
        super.connectedCallback();
        requestAnimationFrame(()=>this._findScrollContainer());
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this._detachScrollListener();
    }
    updated(changed) {
        super.updated(changed);
        if (changed.has('gap')) if (null == this.gap || '' === this.gap.trim()) this.style.removeProperty('--ve-bubble-list-gap');
        else this.style.setProperty('--ve-bubble-list-gap', this.gap);
        if (changed.has('scrollIndicator')) if (this.scrollIndicator) this._findScrollContainer();
        else {
            this._detachScrollListener();
            this._showScrollBottom = false;
        }
    }
    _findScrollContainer() {
        if (!this.scrollIndicator) return;
        let el = this.parentElement;
        while(el){
            const style = getComputedStyle(el);
            const overflowY = style.overflowY;
            if ('auto' === overflowY || 'scroll' === overflowY) return void this._attachScrollListener(el);
            el = el.parentElement;
        }
    }
    _attachScrollListener(container) {
        this._detachScrollListener();
        this._scrollContainer = container;
        container.addEventListener('scroll', this._handleScroll, {
            passive: true
        });
        this._resizeObserver = new ResizeObserver(()=>{
            if (null !== this._rafId) return;
            this._rafId = requestAnimationFrame(()=>{
                this._rafId = null;
                this._checkScrollPosition();
            });
        });
        this._resizeObserver.observe(container);
        this._checkScrollPosition();
    }
    _detachScrollListener() {
        if (this._scrollContainer) {
            this._scrollContainer.removeEventListener('scroll', this._handleScroll);
            this._scrollContainer = null;
        }
        if (this._resizeObserver) {
            this._resizeObserver.disconnect();
            this._resizeObserver = null;
        }
        if (null !== this._rafId) {
            cancelAnimationFrame(this._rafId);
            this._rafId = null;
        }
    }
    _handleScroll = ()=>{
        this._checkScrollPosition();
    };
    _checkScrollPosition() {
        if (!this._scrollContainer || !this.scrollIndicator) return;
        const el = this._scrollContainer;
        const threshold = 50;
        const isAtBottom = el.scrollHeight - el.scrollTop - el.clientHeight < threshold;
        this._showScrollBottom = !isAtBottom;
    }
    _handleScrollBottomClick() {
        if (this._scrollContainer) this._scrollContainer.scrollTo({
            top: this._scrollContainer.scrollHeight,
            behavior: 'smooth'
        });
        this.dispatchEvent(new CustomEvent('ve-bubble-list-scroll-bottom', {
            bubbles: true,
            composed: true
        }));
    }
    _renderLoadingIndicator() {
        return html`
      <div class="ve-bubble-list__indicator ve-bubble-list__indicator--loading" part="indicator">
        <svg class="ve-bubble-list__indicator-icon" viewBox="0 0 16 16" aria-hidden="true" focusable="false">
          <circle class="ve-bubble-list__loading-dot" cx="3.5" cy="8" r="1"></circle>
          <circle class="ve-bubble-list__loading-dot" cx="8" cy="8" r="1"></circle>
          <circle class="ve-bubble-list__loading-dot" cx="12.5" cy="8" r="1"></circle>
        </svg>
      </div>
    `;
    }
    _renderScrollBottomIndicator() {
        return html`
      <div
        class="ve-bubble-list__indicator ve-bubble-list__indicator--scroll"
        part="indicator"
        role="button"
        tabindex="0"
        aria-label="滚动到底部"
        @click=${this._handleScrollBottomClick}
        @keydown=${(e)=>{
            if ('Enter' === e.key || ' ' === e.key) {
                e.preventDefault();
                this._handleScrollBottomClick();
            }
        }}
      >
        <svg class="ve-bubble-list__indicator-icon" viewBox="0 0 16 16" aria-hidden="true" focusable="false">
          <path d="M8 3.25v8.25"/>
          <path d="M4.75 8.75 8 12 11.25 8.75"/>
        </svg>
      </div>
    `;
    }
    render() {
        return html`
      <slot role="log"></slot>
      ${this.loading ? this._renderLoadingIndicator() : nothing}
      ${this.scrollIndicator && this._showScrollBottom ? this._renderScrollBottomIndicator() : nothing}
    `;
    }
}
defineCustomElement(VE_BUBBLE_TAG, VeBubbleElement);
defineCustomElement(VE_BUBBLE_LIST_TAG, VeBubbleListElement);
export { VE_BUBBLE_LIST_TAG, VE_BUBBLE_TAG, VeBubbleElement, VeBubbleListElement };
