// @ve-design/web/dist/ve-badge.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { styleMap } from "lit/directives/style-map.js";
import { property, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
const veBadgeTokenDefaults = css`
  :host {
    --ve-badge-dot-size: 8px;
    --ve-badge-min-size: 18px;
    --ve-badge-padding-inline: var(--space-xxxs);
    --ve-badge-radius: var(--radius-full);
    --ve-badge-offset-x: 0px;
    --ve-badge-offset-y: 0px;
    --ve-badge-count-offset-x: calc(var(--space-xxxs) * -1);
    --ve-badge-count-offset-y: var(--space-xxxs);
    --ve-badge-dot-offset-block: 1px;
    --ve-badge-dot-offset-inline: -1px;
    --ve-badge-font-family: var(--font-en);
    --ve-badge-font-size: var(--text-caption);
    --ve-badge-font-weight: var(--font-weight-normal);
    --ve-badge-line-height: var(--line-height-body);

    --ve-badge-surface-dark: var(--color-bg-primary);
    --ve-badge-text-dark: var(--color-text-foreground);
    --ve-badge-border-dark: transparent;

    --ve-badge-surface-ash: var(--color-border-strong);
    --ve-badge-text-ash: var(--color-text-foreground);
    --ve-badge-border-ash: transparent;

    --ve-badge-surface-blue: var(--color-bg-info);
    --ve-badge-text-blue: var(--color-text-foreground);
    --ve-badge-border-blue: transparent;

    --ve-badge-surface-light: var(--color-bg-base);
    --ve-badge-text-light: var(--color-text-secondary);
    --ve-badge-border-light: var(--color-border-default);

    --ve-badge-background: var(--color-bg-primary);
    --ve-badge-text-color: var(--ve-badge-text-dark);
    --ve-badge-border-color: var(--ve-badge-border-dark);
    --ve-badge-border-width: 0px;

    --ve-badge-current-surface: var(--ve-badge-background);
    --ve-badge-current-text: var(--ve-badge-text-color);
    --ve-badge-current-border: var(--ve-badge-border-color);
    --ve-badge-current-border-width: var(--ve-badge-border-width);
  }

  :host([color='dark']) {
    --ve-badge-current-surface: var(--ve-badge-surface-dark);
    --ve-badge-current-text: var(--ve-badge-text-dark);
    --ve-badge-current-border: var(--ve-badge-border-dark);
    --ve-badge-current-border-width: 0px;
  }

  :host([color='ash']) {
    --ve-badge-current-surface: var(--ve-badge-surface-ash);
    --ve-badge-current-text: var(--ve-badge-text-ash);
    --ve-badge-current-border: var(--ve-badge-border-ash);
    --ve-badge-current-border-width: 0px;
  }

  :host([color='blue']) {
    --ve-badge-current-surface: var(--ve-badge-surface-blue);
    --ve-badge-current-text: var(--ve-badge-text-blue);
    --ve-badge-current-border: var(--ve-badge-border-blue);
    --ve-badge-current-border-width: 0px;
  }

  :host([color='light']) {
    --ve-badge-current-surface: var(--ve-badge-surface-light);
    --ve-badge-current-text: var(--ve-badge-text-light);
    --ve-badge-current-border: var(--ve-badge-border-light);
    --ve-badge-current-border-width: var(--stroke-weight-base);
  }
`;
const token = veBadgeTokenDefaults;
const ve_badge = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-badge {
      position: relative;
      display: inline-flex;
      align-items: center;
      inline-size: max-content;
      max-inline-size: 100%;
      vertical-align: middle;
    }

    .ve-badge-slot {
      display: inline-flex;
      align-items: center;
      max-inline-size: 100%;
    }

    .ve-badge-indicator {
      position: static;
      inset: auto;
      flex: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      border-style: solid;
      border-width: var(--ve-badge-current-border-width);
      border-color: var(--ve-badge-current-border);
      border-radius: var(--ve-badge-radius);
      background: var(--ve-badge-current-surface);
      color: var(--ve-badge-current-text);
      font-family: var(--ve-badge-font-family);
      font-size: var(--ve-badge-font-size);
      font-weight: var(--ve-badge-font-weight);
      line-height: var(--ve-badge-line-height);
      vertical-align: middle;
      white-space: nowrap;
      user-select: none;
      pointer-events: none;
    }

    :host(:not([dot])) .ve-badge-indicator {
      min-inline-size: var(--ve-badge-min-size);
      min-block-size: var(--ve-badge-min-size);
      padding-inline: var(--ve-badge-padding-inline);
    }

    .ve-badge-indicator--dot {
      inline-size: var(--ve-badge-dot-size);
      block-size: var(--ve-badge-dot-size);
      padding: 0;
    }

    .ve-badge-indicator--anchored {
      position: absolute;
      inset-block-start: 0;
      inset-inline-end: 0;
      z-index: 1;
      transform: translate(
        calc(50% + var(--ve-badge-offset-x)),
        calc(-50% + var(--ve-badge-offset-y))
      );
      transform-origin: top right;
    }

    .ve-badge-indicator--anchored:not(.ve-badge-indicator--dot) {
      transform: translate(
        calc(50% + var(--ve-badge-count-offset-x) + var(--ve-badge-offset-x)),
        calc(-50% + var(--ve-badge-count-offset-y) + var(--ve-badge-offset-y))
      );
    }

    .ve-badge-indicator--dot.ve-badge-indicator--anchored {
      inset-block-start: calc(
        var(--ve-badge-dot-offset-block) + var(--ve-badge-offset-y)
      );
      inset-inline-end: calc(
        var(--ve-badge-dot-offset-inline) - var(--ve-badge-offset-x)
      );
      transform: none;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _init_dot, _init_extra__init_dot, _init_status, _init_extra__init_status, _init_text, _init_extra__init_text, _init_color, _init_extra__init_color, _init_offset, _init_extra__init_offset, _init_hasWrappedContent, _init_extra__init_hasWrappedContent;
const VE_BADGE_TAG = 've-badge';
const PRESET_COLORS = new Set([
    'dark',
    'ash',
    'blue',
    'light'
]);
const STATUS_DOT_COLOR = {
    default: 'var(--color-text-tertiary)',
    success: 'var(--color-bg-success)',
    warning: 'var(--color-bg-warning)',
    danger: 'var(--color-bg-danger)',
    processing: 'var(--color-bg-info)'
};
const OFFSET_ATTRIBUTE_CONVERTER = {
    fromAttribute (value) {
        if (!value) return [
            0,
            0
        ];
        const values = value.split(',').map((item)=>Number.parseFloat(item.trim())).filter((item)=>Number.isFinite(item));
        return [
            values[0] ?? 0,
            values[1] ?? 0
        ];
    },
    toAttribute (value) {
        if (!value) return null;
        return `${value[0] ?? 0},${value[1] ?? 0}`;
    }
};
_dec = property({
    type: Boolean,
    reflect: true
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    type: String
}), _dec3 = property({
    type: String,
    reflect: true
}), _dec4 = property({
    converter: OFFSET_ATTRIBUTE_CONVERTER
}), _dec5 = state();
class VeBadgeElement extends VeElement {
    static{
        ({ e: [_init_dot, _init_extra__init_dot, _init_status, _init_extra__init_status, _init_text, _init_extra__init_text, _init_color, _init_extra__init_color, _init_offset, _init_extra__init_offset, _init_hasWrappedContent, _init_extra__init_hasWrappedContent] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "dot"
            ],
            [
                _dec1,
                1,
                "status"
            ],
            [
                _dec2,
                1,
                "text"
            ],
            [
                _dec3,
                1,
                "color"
            ],
            [
                _dec4,
                1,
                "offset"
            ],
            [
                _dec5,
                1,
                "hasWrappedContent"
            ]
        ]));
    }
    static styles = ve_badge;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                'style'
            ])
        ];
    }
    #___private_dot_1 = (()=>{
        const _value = _init_dot(this, false);
        _init_extra__init_dot(this);
        return _value;
    })();
    get dot() {
        return this.#___private_dot_1;
    }
    set dot(_v) {
        this.#___private_dot_1 = _v;
    }
    #___private_status_2 = (()=>{
        const _value = _init_status(this, '');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_2;
    }
    set status(_v) {
        this.#___private_status_2 = _v;
    }
    #___private_text_3 = (()=>{
        const _value = _init_text(this, '');
        _init_extra__init_text(this);
        return _value;
    })();
    get text() {
        return this.#___private_text_3;
    }
    set text(_v) {
        this.#___private_text_3 = _v;
    }
    #___private_color_4 = (()=>{
        const _value = _init_color(this, '');
        _init_extra__init_color(this);
        return _value;
    })();
    get color() {
        return this.#___private_color_4;
    }
    set color(_v) {
        this.#___private_color_4 = _v;
    }
    #___private_offset_5 = (()=>{
        const _value = _init_offset(this, [
            0,
            0
        ]);
        _init_extra__init_offset(this);
        return _value;
    })();
    get offset() {
        return this.#___private_offset_5;
    }
    set offset(_v) {
        this.#___private_offset_5 = _v;
    }
    #___private_hasWrappedContent_6 = (()=>{
        const _value = _init_hasWrappedContent(this, false);
        _init_extra__init_hasWrappedContent(this);
        return _value;
    })();
    get hasWrappedContent() {
        return this.#___private_hasWrappedContent_6;
    }
    set hasWrappedContent(_v) {
        this.#___private_hasWrappedContent_6 = _v;
    }
    connectedCallback() {
        super.connectedCallback();
        this.syncOffsetTokens();
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('offset')) this.syncOffsetTokens();
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.syncWrappedContentState();
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
        this.syncOffsetTokens();
        this.syncWrappedContentState();
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if ('style' === name && oldValue !== newValue) this.syncOffsetTokens();
    }
    syncWrappedContentState() {
        this.hasWrappedContent = [
            ...this.childNodes
        ].some((node)=>{
            if (node.nodeType === Node.TEXT_NODE) return (node.textContent?.trim().length ?? 0) > 0;
            return node.nodeType === Node.ELEMENT_NODE;
        });
    }
    get normalizedColor() {
        return (this.color ?? '').trim();
    }
    syncOffsetTokens() {
        const [offsetX, offsetY] = this.offset;
        this.style.setProperty('--ve-badge-offset-x', `${offsetX}px`);
        this.style.setProperty('--ve-badge-offset-y', `${offsetY}px`);
    }
    get displayText() {
        return (this.text ?? '').trim();
    }
    get shouldRenderIndicator() {
        return this.dot || this.displayText.length > 0;
    }
    get indicatorStyle() {
        const style = {};
        const hasPresetColor = PRESET_COLORS.has(this.normalizedColor);
        if (this.normalizedColor.length > 0 && !hasPresetColor) {
            style['--ve-badge-current-surface'] = this.normalizedColor;
            style['--ve-badge-current-text'] = 'var(--color-text-foreground)';
            style['--ve-badge-current-border'] = 'transparent';
            style['--ve-badge-current-border-width'] = '0px';
        } else if (this.dot && this.status in STATUS_DOT_COLOR) {
            style['--ve-badge-current-surface'] = STATUS_DOT_COLOR[this.status];
            style['--ve-badge-current-text'] = 'var(--color-text-foreground)';
            style['--ve-badge-current-border'] = 'transparent';
            style['--ve-badge-current-border-width'] = '0px';
        }
        return style;
    }
    renderIndicator() {
        if (!this.shouldRenderIndicator) return nothing;
        const isDot = this.dot;
        const indicatorClass = classMap({
            've-badge-indicator': true,
            've-badge-indicator--dot': isDot,
            've-badge-indicator--anchored': this.hasWrappedContent
        });
        const label = isDot ? '有新消息' : this.displayText;
        return html`
      <sup
        class=${indicatorClass}
        part="indicator"
        style=${styleMap(this.indicatorStyle)}
        role="status"
        aria-label=${label}
      >
        ${isDot ? nothing : this.displayText}
      </sup>
    `;
    }
    render() {
        return html`
      <span class="ve-badge">
        <span class="ve-badge-slot">
          <slot @slotchange=${this.syncWrappedContentState}></slot>
        </span>
        ${this.renderIndicator()}
      </span>
    `;
    }
}
defineCustomElement(VE_BADGE_TAG, VeBadgeElement);
export { VE_BADGE_TAG, VeBadgeElement };
