// @ve-design/web/dist/ve-clarify.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { registerIcons } from "./icons/registry.js";
import { downUpIcon } from "./icons/down-up.js";
import { cornerDownLeftIcon } from "./icons/corner-down-left.js";
import { closeIcon } from "./icons/close.js";
import { chevronLeftMdIcon } from "./icons/chevron-left-md.js";
import { chevronRightMdIcon } from "./icons/chevron-right-md.js";
import { arrowRightMdIcon } from "./icons/arrow-right-md.js";
import { arrowUpMdIcon } from "./icons/arrow-up-md.js";
import { checkIcon } from "./76270.js";
import { editIcon } from "./icons/edit.js";
import { dragIcon } from "./icons/drag.js";
const veClarifyTokenDefaults = css`
  :host {
    /* Hairline border — one step lighter than --color-border-default; adapts to dark mode */
    --ve-clarify-border-subtle: var(--color-bg-muted, #ebebf5);

    /* ── 尺寸 ── */
    --ve-clarify-width: 100%;
    --ve-clarify-radius: var(--radius-2xl, 20px);
    --ve-clarify-border-width: var(--stroke-weight-base, 0.5px);

    /* ── 卡片颜色 ── */
    --ve-clarify-bg: var(--color-bg-base, #ffffff);
    --ve-clarify-border: var(--color-border-default, #d2d1e0);
    --ve-clarify-shadow: var(--shadow-md,
      0px 4px 13px rgba(16, 16, 19, 0.05),
      0px 2px 2px rgba(16, 16, 19, 0.05));

    /* ── 问题标题 ── */
    --ve-clarify-question-color: var(--color-text-primary, #101013);
    --ve-clarify-question-size: var(--text-body-lg, 16px);
    --ve-clarify-question-lh: var(--line-height-body, 26px);

    /* ── 步骤导航 ── */
    --ve-clarify-step-color: var(--color-text-tertiary, #7a7880);
    --ve-clarify-step-size: var(--text-body, 14px);
    --ve-clarify-step-lh: 22px;

    /* ── 序号徽章 ── */
    --ve-clarify-badge-size: 24px;
    --ve-clarify-badge-radius: var(--radius-sm, 6px);
    --ve-clarify-badge-bg: var(--color-bg-base, #ffffff);
    --ve-clarify-badge-border: var(--ve-clarify-border-subtle);
    --ve-clarify-badge-color: var(--color-text-primary, #101013);
    --ve-clarify-badge-font-family: var(--font-en, 'SF Pro', sans-serif);
    --ve-clarify-badge-font-size: var(--text-body, 14px);

    /* ── 选项行 ── */
    --ve-clarify-option-height: 48px;
    --ve-clarify-option-radius: var(--radius-lg, 12px);
    --ve-clarify-option-padding: var(--space-xs, 12px);
    --ve-clarify-option-gap: var(--space-xxs, 8px);
    --ve-clarify-option-bg: transparent;
    --ve-clarify-option-bg-selected: var(--color-bg-secondary, #f6f6fc);
    --ve-clarify-option-text: var(--color-text-secondary, #383740);
    --ve-clarify-option-text-selected: var(--color-text-primary, #101013);
    --ve-clarify-option-text-disabled: var(--color-text-disable, #b8b7c3);
    --ve-clarify-option-divider: var(--ve-clarify-border-subtle);
    --ve-clarify-option-divider-inset: var(--space-xxxs, 4px);

    /* ── 多选复选框 ── */
    --ve-clarify-checkbox-size: 16px;
    --ve-clarify-checkbox-radius: var(--radius-xs, 4px);
    --ve-clarify-checkbox-border: var(--color-border-strong, #b8b7c3);
    --ve-clarify-checkbox-bg: var(--color-bg-base, #ffffff);
    --ve-clarify-checkbox-checked-bg: var(--color-bg-primary, #101013);
    --ve-clarify-checkbox-icon-color: var(--color-text-foreground, #ffffff);

    /* ── 拖拽指示器 ── */
    --ve-clarify-drop-indicator-color: var(--color-bg-primary, #101013);
    --ve-clarify-drop-indicator-size: var(--stroke-2, 2px);
    --ve-clarify-drop-indicator-radius: var(--radius-full, 999px);

    /* ── 补充信息输入 ── */
    --ve-clarify-supplement-placeholder: var(--color-text-disable, #b8b7c3);

    /* ── 拖拽图标 ── */
    --ve-clarify-drag-color: var(--color-text-tertiary, #7a7880);
    --ve-clarify-drag-color-active: var(--color-text-secondary, #383740);

    /* ── 底部操作区 ── */
    --ve-clarify-footer-label-color: var(--color-text-secondary, #383740);
    --ve-clarify-footer-count-color: var(--color-text-primary, #101013);
    --ve-clarify-footer-hint-color: var(--color-text-tertiary, #7a7880);

    /* ── 跳过按钮 ── */
    --ve-clarify-btn-skip-height: 32px;
    --ve-clarify-btn-skip-radius: var(--radius-md, 8px);
    --ve-clarify-btn-skip-border: var(--color-border-strong, #b8b7c3);
    --ve-clarify-btn-skip-color: var(--color-text-primary, #101013);

    /* ── 提交按钮 ── */
    --ve-clarify-btn-submit-bg: var(--color-bg-primary, #101013);
    --ve-clarify-btn-submit-color: var(--color-text-foreground, #ffffff);
    --ve-clarify-btn-submit-radius: var(--radius-md, 8px);
    --ve-clarify-btn-submit-size: 32px;

    /* ── 快捷键栏 ── */
    --ve-clarify-shortcut-color: var(--color-text-tertiary, #7a7880);
    --ve-clarify-shortcut-size: var(--text-caption, 12px);
    --ve-clarify-shortcut-opacity: 0.7;

    /* ── 交互态 / 动画 ── */
    --ve-clarify-hover-ratio: var(--state-hover-subtle-ratio, 8%);
    --ve-clarify-hover-mix: var(--state-hover-subtle-mix, currentColor);
    --ve-clarify-disabled-ratio: var(--state-disabled-ratio, 0.4);
    --ve-clarify-transition-duration: var(--motion-fast, 160ms);
    --ve-clarify-transition-ease: var(--ease-standard, ease);
  }
`;
const token = veClarifyTokenDefaults;
const ve_clarify = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      width: var(--ve-clarify-width);
      font-family: var(--font-sans, 'PingFang SC', sans-serif);
    }

    /* ===========================================================
     * Card
     * =========================================================== */
    .ve-clarify {
      display: flex;
      flex-direction: column;
      align-items: center;
      position: relative;
      width: 100%;
      background: var(--ve-clarify-bg);
      border: var(--ve-clarify-border-width) solid var(--ve-clarify-border);
      border-radius: var(--ve-clarify-radius);
      box-shadow: var(--ve-clarify-shadow);
      overflow: hidden;
      outline: none;
    }

    /* ===========================================================
     * Header
     * =========================================================== */
    .ve-clarify-header {
      display: flex;
      align-items: center;
      justify-content: space-between;
      width: 100%;
      padding: var(--space-m, 20px) var(--space-l, 24px) var(--space-s, 16px);
      box-sizing: border-box;
    }

    .ve-clarify-header-title {
      flex: 1 0 0;
      min-width: 0;
      margin: 0;
      font-size: var(--ve-clarify-question-size);
      line-height: var(--ve-clarify-question-lh);
      color: var(--ve-clarify-question-color);
      white-space: normal;
      overflow-wrap: anywhere;
      word-break: break-word;
    }

    .ve-clarify-header-actions {
      display: flex;
      align-items: center;
      gap: var(--space-s, 16px);
      flex-shrink: 0;
    }

    /* ── 步骤导航 ── */
    .ve-clarify-pagination {
      display: flex;
      align-items: center;
      gap: var(--space-xxxs, 4px);
    }

    .ve-clarify-pagination-btn {
      background: none;
      border: none;
      padding: 0;
      cursor: pointer;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      width: 16px;
      height: 16px;
      color: var(--ve-clarify-step-color);
      transition: opacity 160ms ease;
    }

    .ve-clarify-pagination-btn:hover:not(:disabled) {
      opacity: 0.7;
    }

    .ve-clarify-pagination-btn:disabled {
      opacity: 0.35;
      cursor: default;
    }

    .ve-clarify-pagination-btn svg {
      width: 16px;
      height: 16px;
      display: block;
    }

    .ve-clarify-pagination-text {
      font-size: var(--ve-clarify-step-size);
      line-height: var(--ve-clarify-step-lh);
      color: var(--ve-clarify-step-color);
      white-space: nowrap;
    }

    /* ── 关闭按钮 ── */
    .ve-clarify-close {
      background: transparent;
      border: none;
      padding: 0;
      cursor: pointer;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      width: 24px;
      height: 24px;
      border-radius: var(--radius-sm, 6px);
      color: var(--color-text-secondary, #383740);
      transition: background-color var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease),
        color var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease);
    }

    .ve-clarify-close:hover {
      background: var(--color-bg-secondary, #f6f6fc);
      color: var(--color-text-primary, #101013);
    }

    .ve-clarify-close:active {
      background: color-mix(
        in srgb,
        var(--color-text-primary, #101013)
          var(--state-pressed-subtle-ratio, 12%),
        transparent
      );
    }

    .ve-clarify-close:focus-visible {
      outline: var(--stroke-2, 2px) solid var(--color-border-interactive,
        #165dff);
      outline-offset: 2px;
    }

    .ve-clarify-close svg {
      width: 16px;
      height: 16px;
      display: block;
    }

    /* ===========================================================
     * Body / Options
     * =========================================================== */
    .ve-clarify-body {
      display: flex;
      flex-direction: column;
      align-items: center;
      width: 100%;
      padding: 0 var(--space-xs, 12px) var(--space-xxs, 8px);
      box-sizing: border-box;
    }

    .ve-clarify-options {
      display: flex;
      flex-direction: column;
      align-items: center;
      width: 100%;
    }

    /* ── 单个选项容器 ── */
    .ve-clarify-option {
      display: flex;
      flex-direction: column;
      align-items: flex-start;
      width: 100%;
      border-radius: var(--ve-clarify-option-radius);
      background: var(--ve-clarify-option-bg);
      cursor: pointer;
      transition: background-color var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease),
        color var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease);
      position: relative;
      outline: none;
      overflow: hidden;
    }

    /* hover：bg-secondary（disabled 与 supplement 除外） */
    .ve-clarify-option:not(.ve-clarify-option-disabled):not(
        .ve-clarify-option-supplement
      ):hover {
      background: var(--ve-clarify-option-bg-selected);
    }

    /* 键盘 highlight 态（与 hover 视觉一致） */
    .ve-clarify-option-highlighted:not(.ve-clarify-option-disabled) {
      background: var(--ve-clarify-option-bg-selected);
    }

    /* 高亮态相邻分隔线隐藏 */
    .ve-clarify-option-highlighted ~ .ve-clarify-divider,
    .ve-clarify-option-highlighted .ve-clarify-divider {
      visibility: hidden;
    }

    /* 补充信息行位于列表底部，不展示分割线；其余选项（含多选/排序的最后一项）均保留 */
    .ve-clarify-options
      > .ve-clarify-option-supplement
      .ve-clarify-divider {
      display: none;
    }

    /* 非排序模式下若末项是真实选项（无补充信息行），同样隐藏底部分割线 */
    :host(:not([data-mode='sort']))
      .ve-clarify-options
      > .ve-clarify-option:last-child:not(.ve-clarify-option-supplement)
      .ve-clarify-divider {
      display: none;
    }

    /* 单选选中态：持久 bg-secondary */
    :host([data-mode='single']) .ve-clarify-option-selected {
      background: var(--ve-clarify-option-bg-selected);
    }

    /* 多选已勾选：默认不影响背景 */
    :host([data-mode='multi']) .ve-clarify-option-selected {
      background: var(--ve-clarify-option-bg);
    }

    /* 多选已勾选 + hover → bg-secondary */
    :host([data-mode='multi'])
      .ve-clarify-option-selected:not(.ve-clarify-option-disabled):hover {
      background: var(--ve-clarify-option-bg-selected);
    }

    .ve-clarify-option-disabled {
      cursor: not-allowed;
    }

    /* 禁用态徽章 */
    .ve-clarify-option-disabled .ve-clarify-badge {
      background: var(--ve-clarify-option-bg-selected);
      border-color: transparent;
      color: var(--ve-clarify-option-text-disabled);
    }

    .ve-clarify-option-disabled .ve-clarify-checkbox {
      background: var(--ve-clarify-option-bg-selected);
      border-color: var(--ve-clarify-checkbox-border);
      opacity: 0.5;
    }

    /* 排序模式下的拖拽幽灵元素也保留圆角 */
    .ve-clarify-option-sortable {
      border-radius: var(--ve-clarify-option-radius);
    }

    .ve-clarify-option-sortable .ve-clarify-option-row {
      border-radius: var(--ve-clarify-option-radius);
    }

    .ve-clarify-option-dragging {
      opacity: 0;
    }

    .ve-clarify-option-supplement {
      cursor: default;
    }

    /* 预留行内按钮高度，避免输入文字时「跳过 / 提交」按钮出现导致行高跳动 */
    :host([data-mode='single'])
      .ve-clarify-option-supplement
      .ve-clarify-option-row {
      min-height: var(--ve-clarify-option-height);
      padding-block: calc(
        (
            var(--ve-clarify-option-height) -
              var(--ve-clarify-btn-submit-size)
          ) / 2
      );
    }

    .ve-clarify-option-supplement .ve-clarify-badge-checkbox {
      cursor: pointer;
    }

    /* ── 选项内容行 ── */
    .ve-clarify-option-row {
      display: flex;
      align-items: center;
      gap: var(--ve-clarify-option-gap);
      width: 100%;
      min-height: var(--ve-clarify-option-height);
      padding: var(--ve-clarify-option-padding);
      box-sizing: border-box;
      border-radius: var(--ve-clarify-option-radius);
    }

    /* ── 分隔线 ── */
    .ve-clarify-divider {
      width: 100%;
      padding-inline: var(--ve-clarify-option-divider-inset);
      box-sizing: border-box;
    }

    .ve-clarify-divider-line {
      border-top: var(--ve-clarify-border-width) solid
        var(--ve-clarify-option-divider);
      width: 100%;
      height: 0;
    }

    /* 选中/hover 的 bg-secondary 项无 divider */
    :host([data-mode='single'])
      .ve-clarify-option-selected
      .ve-clarify-divider,
    .ve-clarify-option:not(.ve-clarify-option-disabled):not(
        .ve-clarify-option-supplement
      ):hover
      .ve-clarify-divider {
      visibility: hidden;
    }

    /* ── 序号徽章 ── */
    .ve-clarify-badge {
      flex-shrink: 0;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      width: var(--ve-clarify-badge-size);
      height: var(--ve-clarify-badge-size);
      border-radius: var(--ve-clarify-badge-radius);
      background: var(--ve-clarify-badge-bg);
      border: var(--ve-clarify-border-width) solid
        var(--ve-clarify-badge-border);
      font-family: var(--ve-clarify-badge-font-family);
      font-size: var(--ve-clarify-badge-font-size);
      line-height: 22px;
      color: var(--ve-clarify-badge-color);
      overflow: clip;
      box-sizing: border-box;
    }

    /* 多选徽章：透明背景，内嵌复选框 */
    .ve-clarify-badge-checkbox {
      background: transparent;
      border: none;
      padding: 0;
    }

    .ve-clarify-checkbox {
      width: var(--ve-clarify-checkbox-size);
      height: var(--ve-clarify-checkbox-size);
      border-radius: var(--ve-clarify-checkbox-radius);
      border: var(--ve-clarify-border-width) solid
        var(--ve-clarify-checkbox-border);
      background: var(--ve-clarify-checkbox-bg);
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex-shrink: 0;
      transition: background var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease),
        border-color var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease);
    }

    .ve-clarify-checkbox-checked {
      background: var(--ve-clarify-checkbox-checked-bg);
      border-color: transparent;
      color: var(--ve-clarify-checkbox-icon-color);
    }

    .ve-clarify-checkbox svg {
      width: 12px;
      height: 12px;
      display: block;
    }

    /* 补充信息徽章（铅笔） */
    .ve-clarify-badge-supplement {
      background: var(--ve-clarify-badge-bg);
      border: var(--ve-clarify-border-width) solid
        var(--ve-clarify-badge-border);
      color: var(--color-text-tertiary, #7a7880);
      padding: 0;
    }

    .ve-clarify-badge-supplement svg {
      width: 16px;
      height: 16px;
      display: block;
    }

    /* ── 选项文本 ── */
    .ve-clarify-option-content {
      flex: 1 0 0;
      min-width: 0;
      overflow: hidden;
    }

    .ve-clarify-option-label {
      display: block;
      font-size: 14px;
      line-height: 22px;
      color: var(--ve-clarify-option-text);
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    /* 选中态文字加深 */
    .ve-clarify-option-selected .ve-clarify-option-label {
      color: var(--ve-clarify-option-text-selected);
    }

    /* hover 态文字保持 text-secondary（不加深） */
    .ve-clarify-option:not(.ve-clarify-option-disabled):hover
      .ve-clarify-option-label {
      color: var(--ve-clarify-option-text);
    }

    /* 多选 checked+hover 保留 text-primary */
    :host([data-mode='multi'])
      .ve-clarify-option-selected:hover
      .ve-clarify-option-label {
      color: var(--ve-clarify-option-text-selected);
    }

    .ve-clarify-option-disabled .ve-clarify-option-label {
      color: var(--ve-clarify-option-text-disabled);
    }

    /* ── 单选右箭头（仅 selected 显示） ── */
    .ve-clarify-arrow {
      flex-shrink: 0;
      width: 16px;
      height: 16px;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      color: var(--color-text-secondary, #383740);
    }

    .ve-clarify-arrow svg {
      width: 16px;
      height: 16px;
      display: block;
    }

    /* ── 拖拽图标 ── */
    .ve-clarify-drag-handle {
      flex-shrink: 0;
      width: 16px;
      height: 16px;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      cursor: grab;
      color: var(--ve-clarify-drag-color);
    }

    .ve-clarify-drag-handle:active {
      cursor: grabbing;
    }

    .ve-clarify-option-selected .ve-clarify-drag-handle,
    .ve-clarify-option:hover .ve-clarify-drag-handle {
      color: var(--ve-clarify-drag-color-active);
    }

    .ve-clarify-drag-handle svg {
      width: 16px;
      height: 16px;
      display: block;
      fill: currentColor;
    }

    /* ── 排序模式 ── */
    .ve-clarify-option-sortable {
      cursor: grab;
    }

    .ve-clarify-option-sortable:active {
      cursor: grabbing;
    }

    .ve-clarify-option-sortable.ve-clarify-option-disabled {
      cursor: not-allowed;
    }

    /* 拖拽中的源项透明度参考 ac-clarify-wizard */
    .ve-clarify-option-sortable.ve-clarify-option-dragging {
      opacity: var(--ve-clarify-disabled-ratio);
      cursor: grabbing;
    }

    /* ── 补充信息输入框 ── */
    .ve-clarify-input-field {
      flex: 1 0 0;
      min-width: 0;
      background: none;
      border: none;
      outline: none;
      font-family: inherit;
      font-size: 14px;
      line-height: 22px;
      color: var(--color-text-primary, #101013);
      width: 100%;
    }

    .ve-clarify-input-field::placeholder {
      color: var(--ve-clarify-supplement-placeholder);
    }

    /* ===========================================================
     * Footer
     * =========================================================== */
    .ve-clarify-footer {
      display: flex;
      align-items: center;
      justify-content: space-between;
      gap: var(--space-xxs, 8px);
      width: 100%;
      padding: 10px var(--space-xs, 12px) var(--space-xs, 12px);
      box-sizing: border-box;
    }

    .ve-clarify-footer-info {
      flex: 1 0 0;
      min-width: 0;
      display: flex;
      align-items: center;
      gap: var(--space-xxs, 8px);
      padding-left: var(--space-xs, 12px);
      font-size: 14px;
      line-height: 22px;
    }

    .ve-clarify-footer-count {
      color: var(--ve-clarify-footer-count-color);
      text-align: center;
      min-width: 24px;
    }

    .ve-clarify-footer-label {
      color: var(--ve-clarify-footer-label-color);
      white-space: nowrap;
    }

    .ve-clarify-footer-hint {
      color: var(--ve-clarify-footer-hint-color);
      white-space: nowrap;
    }

    .ve-clarify-footer-actions {
      display: flex;
      align-items: center;
      gap: var(--space-xxs, 8px);
      flex-shrink: 0;
    }

    /* ── 跳过按钮 ── */
    .ve-clarify-btn-skip {
      height: var(--ve-clarify-btn-skip-height);
      padding: 5px 12px;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      background: none;
      border: var(--ve-clarify-border-width) solid
        var(--ve-clarify-btn-skip-border);
      border-radius: var(--ve-clarify-btn-skip-radius);
      font-family: inherit;
      font-size: 14px;
      line-height: 22px;
      color: var(--ve-clarify-btn-skip-color);
      cursor: pointer;
      white-space: nowrap;
      transition: background var(--ve-clarify-transition-duration)
        var(--ve-clarify-transition-ease);
    }

    .ve-clarify-btn-skip:hover {
      background: var(--color-bg-secondary, #f6f6fc);
    }

    /* ── 提交按钮（→） ── */
    .ve-clarify-btn-submit {
      width: var(--ve-clarify-btn-submit-size);
      height: var(--ve-clarify-btn-submit-size);
      display: inline-flex;
      align-items: center;
      justify-content: center;
      background: var(--ve-clarify-btn-submit-bg);
      border: none;
      border-radius: var(--ve-clarify-btn-submit-radius);
      cursor: pointer;
      color: var(--ve-clarify-btn-submit-color);
      transition: background var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease),
        opacity var(--ve-clarify-transition-duration)
          var(--ve-clarify-transition-ease);
    }

    .ve-clarify-btn-submit:disabled {
      background: var(--ve-clarify-badge-bg);
      border: var(--ve-clarify-border-width) solid
        var(--ve-clarify-btn-skip-border);
      color: var(--ve-clarify-option-text-disabled);
      cursor: not-allowed;
    }

    .ve-clarify-btn-submit:hover:not(:disabled) {
      opacity: 0.85;
    }

    .ve-clarify-btn-submit svg {
      width: 16px;
      height: 16px;
      display: block;
    }

    /* ===========================================================
     * Keyboard hint (shortcuts bar)
     * =========================================================== */
    .ve-clarify-keyboard-hint {
      display: flex;
      gap: var(--space-m, 20px);
      align-items: center;
      justify-content: center;
      padding-block: var(--space-xxs, 8px);
      width: 100%;
    }

    .ve-clarify-keyboard-hint-item {
      display: flex;
      gap: var(--space-xxxs, 4px);
      align-items: center;
      font-size: var(--ve-clarify-shortcut-size);
      line-height: 1.5;
      color: var(--ve-clarify-shortcut-color);
      opacity: var(--ve-clarify-shortcut-opacity);
    }

    .ve-clarify-keyboard-hint-icons {
      display: flex;
      align-items: center;
    }

    .ve-clarify-keyboard-hint-icons svg + svg {
      margin-left: -4px;
    }

    .ve-clarify-keyboard-hint-key {
      font-family: 'SF Pro', var(--font-sans), sans-serif;
      white-space: nowrap;
    }

    .ve-clarify-keyboard-hint-text {
      white-space: nowrap;
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-clarify-option,
      .ve-clarify-checkbox,
      .ve-clarify-btn-skip,
      .ve-clarify-btn-submit,
      .ve-clarify-close,
      .ve-clarify-pagination-btn {
        transition: none;
        animation: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _init_questions, _init_extra__init_questions, _init_activeIndex, _init_extra__init_activeIndex, _init_defaultActiveIndex, _init_extra__init_defaultActiveIndex, _init_disabled, _init_extra__init_disabled, _init_closable, _init_extra__init_closable, _init_showKeyboardHint, _init_extra__init_showKeyboardHint, _init_keyboardNavigation, _init_extra__init_keyboardNavigation, _init_internalActiveIndex, _init_extra__init_internalActiveIndex, _init_highlightedIndex, _init_extra__init_highlightedIndex, _init_selections, _init_extra__init_selections, _init_inputs, _init_extra__init_inputs, _init_inputSelected, _init_extra__init_inputSelected, _init_sortOrders, _init_extra__init_sortOrders, _init_draggingValue, _init_extra__init_draggingValue, _init_results, _init_extra__init_results;
const VE_CLARIFY_TAG = 've-clarify';
const VE_CLARIFY_ACTIVE_INDEX_CHANGE_EVENT = 've-active-index-change';
const VE_CLARIFY_SUBMIT_EVENT = 've-submit';
const VE_CLARIFY_SKIP_EVENT = 've-skip';
const VE_CLARIFY_COMPLETE_EVENT = 've-complete';
const VE_CLARIFY_CLOSE_EVENT = 've-close';
_dec = property({
    attribute: false
}), _dec1 = property({
    type: Number,
    attribute: 'active-index',
    reflect: true
}), _dec2 = property({
    type: Number,
    attribute: 'default-active-index'
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = property({
    type: Boolean,
    attribute: 'show-keyboard-hint',
    reflect: true
}), _dec6 = property({
    type: Boolean,
    attribute: 'keyboard-navigation',
    reflect: true
}), _dec7 = state(), _dec8 = state(), _dec9 = state(), _dec10 = state(), _dec11 = state(), _dec12 = state(), _dec13 = state(), _dec14 = state();
class VeClarifyElement extends VeElement {
    static{
        ({ e: [_init_questions, _init_extra__init_questions, _init_activeIndex, _init_extra__init_activeIndex, _init_defaultActiveIndex, _init_extra__init_defaultActiveIndex, _init_disabled, _init_extra__init_disabled, _init_closable, _init_extra__init_closable, _init_showKeyboardHint, _init_extra__init_showKeyboardHint, _init_keyboardNavigation, _init_extra__init_keyboardNavigation, _init_internalActiveIndex, _init_extra__init_internalActiveIndex, _init_highlightedIndex, _init_extra__init_highlightedIndex, _init_selections, _init_extra__init_selections, _init_inputs, _init_extra__init_inputs, _init_inputSelected, _init_extra__init_inputSelected, _init_sortOrders, _init_extra__init_sortOrders, _init_draggingValue, _init_extra__init_draggingValue, _init_results, _init_extra__init_results] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "questions"
            ],
            [
                _dec1,
                1,
                "activeIndex"
            ],
            [
                _dec2,
                1,
                "defaultActiveIndex"
            ],
            [
                _dec3,
                1,
                "disabled"
            ],
            [
                _dec4,
                1,
                "closable"
            ],
            [
                _dec5,
                1,
                "showKeyboardHint"
            ],
            [
                _dec6,
                1,
                "keyboardNavigation"
            ],
            [
                _dec7,
                1,
                "internalActiveIndex"
            ],
            [
                _dec8,
                1,
                "highlightedIndex"
            ],
            [
                _dec9,
                1,
                "selections"
            ],
            [
                _dec10,
                1,
                "inputs"
            ],
            [
                _dec11,
                1,
                "inputSelected"
            ],
            [
                _dec12,
                1,
                "sortOrders"
            ],
            [
                _dec13,
                1,
                "draggingValue"
            ],
            [
                _dec14,
                1,
                "results"
            ]
        ]));
    }
    static styles = ve_clarify;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    static{
        registerIcons([
            downUpIcon,
            cornerDownLeftIcon,
            closeIcon,
            chevronLeftMdIcon,
            chevronRightMdIcon,
            arrowRightMdIcon,
            arrowUpMdIcon,
            checkIcon,
            editIcon,
            dragIcon
        ]);
    }
    #___private_questions_1 = (()=>{
        const _value = _init_questions(this, []);
        _init_extra__init_questions(this);
        return _value;
    })();
    get questions() {
        return this.#___private_questions_1;
    }
    set questions(_v) {
        this.#___private_questions_1 = _v;
    }
    #___private_activeIndex_2 = (()=>{
        const _value = _init_activeIndex(this, null);
        _init_extra__init_activeIndex(this);
        return _value;
    })();
    get activeIndex() {
        return this.#___private_activeIndex_2;
    }
    set activeIndex(_v) {
        this.#___private_activeIndex_2 = _v;
    }
    #___private_defaultActiveIndex_3 = (()=>{
        const _value = _init_defaultActiveIndex(this, 0);
        _init_extra__init_defaultActiveIndex(this);
        return _value;
    })();
    get defaultActiveIndex() {
        return this.#___private_defaultActiveIndex_3;
    }
    set defaultActiveIndex(_v) {
        this.#___private_defaultActiveIndex_3 = _v;
    }
    #___private_disabled_4 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_4;
    }
    set disabled(_v) {
        this.#___private_disabled_4 = _v;
    }
    #___private_closable_5 = (()=>{
        const _value = _init_closable(this, false);
        _init_extra__init_closable(this);
        return _value;
    })();
    get closable() {
        return this.#___private_closable_5;
    }
    set closable(_v) {
        this.#___private_closable_5 = _v;
    }
    #___private_showKeyboardHint_6 = (()=>{
        const _value = _init_showKeyboardHint(this, true);
        _init_extra__init_showKeyboardHint(this);
        return _value;
    })();
    get showKeyboardHint() {
        return this.#___private_showKeyboardHint_6;
    }
    set showKeyboardHint(_v) {
        this.#___private_showKeyboardHint_6 = _v;
    }
    #___private_keyboardNavigation_7 = (()=>{
        const _value = _init_keyboardNavigation(this, true);
        _init_extra__init_keyboardNavigation(this);
        return _value;
    })();
    get keyboardNavigation() {
        return this.#___private_keyboardNavigation_7;
    }
    set keyboardNavigation(_v) {
        this.#___private_keyboardNavigation_7 = _v;
    }
    #___private_internalActiveIndex_8 = (()=>{
        const _value = _init_internalActiveIndex(this, 0);
        _init_extra__init_internalActiveIndex(this);
        return _value;
    })();
    get internalActiveIndex() {
        return this.#___private_internalActiveIndex_8;
    }
    set internalActiveIndex(_v) {
        this.#___private_internalActiveIndex_8 = _v;
    }
    #___private_highlightedIndex_9 = (()=>{
        const _value = _init_highlightedIndex(this, -1);
        _init_extra__init_highlightedIndex(this);
        return _value;
    })();
    get highlightedIndex() {
        return this.#___private_highlightedIndex_9;
    }
    set highlightedIndex(_v) {
        this.#___private_highlightedIndex_9 = _v;
    }
    #___private_selections_10 = (()=>{
        const _value = _init_selections(this, {});
        _init_extra__init_selections(this);
        return _value;
    })();
    get selections() {
        return this.#___private_selections_10;
    }
    set selections(_v) {
        this.#___private_selections_10 = _v;
    }
    #___private_inputs_11 = (()=>{
        const _value = _init_inputs(this, {});
        _init_extra__init_inputs(this);
        return _value;
    })();
    get inputs() {
        return this.#___private_inputs_11;
    }
    set inputs(_v) {
        this.#___private_inputs_11 = _v;
    }
    #___private_inputSelected_12 = (()=>{
        const _value = _init_inputSelected(this, false);
        _init_extra__init_inputSelected(this);
        return _value;
    })();
    get inputSelected() {
        return this.#___private_inputSelected_12;
    }
    set inputSelected(_v) {
        this.#___private_inputSelected_12 = _v;
    }
    #___private_sortOrders_13 = (()=>{
        const _value = _init_sortOrders(this, {});
        _init_extra__init_sortOrders(this);
        return _value;
    })();
    get sortOrders() {
        return this.#___private_sortOrders_13;
    }
    set sortOrders(_v) {
        this.#___private_sortOrders_13 = _v;
    }
    #___private_draggingValue_14 = (()=>{
        const _value = _init_draggingValue(this, null);
        _init_extra__init_draggingValue(this);
        return _value;
    })();
    get draggingValue() {
        return this.#___private_draggingValue_14;
    }
    set draggingValue(_v) {
        this.#___private_draggingValue_14 = _v;
    }
    #___private_results_15 = (()=>{
        const _value = _init_results(this, []);
        _init_extra__init_results(this);
        return _value;
    })();
    get results() {
        return this.#___private_results_15;
    }
    set results(_v) {
        this.#___private_results_15 = _v;
    }
    submittedIds = new Set();
    completeFired = false;
    lastQuestionsRef = null;
    connectedCallback() {
        super.connectedCallback();
        this.internalActiveIndex = this.clampIndex(this.defaultActiveIndex);
        this.tabIndex = this.tabIndex < 0 ? 0 : this.tabIndex;
        this.addEventListener('keydown', this.handleHostKeyDown);
        this.addEventListener('mouseleave', this.handleHostMouseLeave);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('keydown', this.handleHostKeyDown);
        this.removeEventListener('mouseleave', this.handleHostMouseLeave);
    }
    willUpdate(changed) {
        super.willUpdate(changed);
        if (this.lastQuestionsRef !== this.questions) {
            this.lastQuestionsRef = this.questions;
            this.submittedIds = new Set();
            this.completeFired = false;
            this.internalActiveIndex = this.clampIndex(this.internalActiveIndex);
        }
        const has = (key)=>changed.has(key);
        const indexChanged = has('activeIndex') || has('internalActiveIndex') || has('questions');
        if (indexChanged) this.highlightedIndex = -1;
        const question = this.currentQuestion;
        const mode = question?.sortable ? 'sort' : question?.multiple ? 'multi' : 'single';
        if (this.dataset.mode !== mode) this.dataset.mode = mode;
        if (indexChanged && question?.sortable && !this.sortOrders[question.id]) this.sortOrders = {
            ...this.sortOrders,
            [question.id]: question.options.map((opt)=>opt.value)
        };
    }
    updated(changed) {
        super.updated(changed);
        if (changed.has('results') && this.questions.length > 0 && this.results.length === this.questions.length && !this.completeFired) {
            this.completeFired = true;
            this.dispatchEvent(new CustomEvent(VE_CLARIFY_COMPLETE_EVENT, {
                detail: {
                    results: [
                        ...this.results
                    ]
                },
                bubbles: true,
                composed: true
            }));
        }
    }
    get currentIndex() {
        if ('number' == typeof this.activeIndex && !Number.isNaN(this.activeIndex)) return this.clampIndex(this.activeIndex);
        return this.clampIndex(this.internalActiveIndex);
    }
    get currentQuestion() {
        return this.questions[this.currentIndex];
    }
    clampIndex(index) {
        if (0 === this.questions.length) return 0;
        if (index < 0) return 0;
        if (index >= this.questions.length) return this.questions.length - 1;
        return index;
    }
    getSelected(questionId) {
        return this.selections[questionId] ?? new Set();
    }
    getInputText(questionId) {
        return this.inputs[questionId] ?? '';
    }
    isOptionSelected(questionId, value) {
        return this.getSelected(questionId).has(value);
    }
    getSortOrder(question) {
        const cached = this.sortOrders[question.id];
        if (!cached) return question.options.map((opt)=>opt.value);
        const valueSet = new Set(question.options.map((opt)=>opt.value));
        const valid = cached.length === question.options.length && cached.every((v)=>valueSet.has(v));
        if (!valid) return question.options.map((opt)=>opt.value);
        return cached;
    }
    getSortedOptions(question) {
        const order = this.getSortOrder(question);
        const map = new Map(question.options.map((opt)=>[
                opt.value,
                opt
            ]));
        const result = [];
        order.forEach((value)=>{
            const opt = map.get(value);
            if (opt) result.push(opt);
        });
        return result;
    }
    upsertResult(result) {
        const next = this.results.filter((r)=>r.questionId !== result.questionId);
        next.push(result);
        this.results = next;
    }
    setActiveIndex(nextIndex) {
        const clamped = this.clampIndex(nextIndex);
        if (clamped === this.currentIndex) return;
        this.inputSelected = false;
        if (null === this.activeIndex) this.internalActiveIndex = clamped;
        this.dispatchEvent(new CustomEvent(VE_CLARIFY_ACTIVE_INDEX_CHANGE_EVENT, {
            detail: {
                index: clamped
            },
            bubbles: true,
            composed: true
        }));
    }
    goToNextQuestion() {
        const idx = this.currentIndex;
        if (idx < this.questions.length - 1) {
            this.setActiveIndex(idx + 1);
            return true;
        }
        return false;
    }
    handleOptionClick = (option)=>{
        const question = this.currentQuestion;
        if (!question) return;
        if (this.disabled || option.disabled) return;
        if (question.sortable) return;
        const qid = question.id;
        const selected = new Set(this.getSelected(qid));
        const wasSelected = selected.has(option.value);
        const hasManualInput = this.getInputText(qid).trim().length > 0;
        if (question.multiple) {
            if (wasSelected) selected.delete(option.value);
            else selected.add(option.value);
            this.selections = {
                ...this.selections,
                [qid]: selected
            };
            this.highlightedIndex = -1;
            return;
        }
        if (wasSelected) selected.clear();
        else {
            selected.clear();
            selected.add(option.value);
        }
        this.selections = {
            ...this.selections,
            [qid]: selected
        };
        if (!question.multiple && !wasSelected && !hasManualInput) {
            const result = {
                questionId: qid,
                selectedValues: [
                    option.value
                ],
                skipped: false
            };
            this.dispatchSubmit(result);
            this.recordResult(result);
            this.goToNextQuestion();
        }
    };
    handleInputChange = (event)=>{
        const question = this.currentQuestion;
        if (!question) return;
        const target = event.target;
        this.inputs = {
            ...this.inputs,
            [question.id]: target.value
        };
        if (question.multiple && target.value.trim().length > 0) this.inputSelected = true;
    };
    handleInputCheckboxClick = ()=>{
        this.inputSelected = !this.inputSelected;
    };
    handleSubmit = ()=>{
        const question = this.currentQuestion;
        if (!question) return;
        const qid = question.id;
        const inputText = this.getInputText(qid);
        const selectedValues = question.sortable ? this.getSortOrder(question) : Array.from(this.getSelected(qid));
        const result = {
            questionId: qid,
            selectedValues,
            ...inputText ? {
                inputText
            } : {},
            skipped: false
        };
        this.dispatchSubmit(result);
        this.recordResult(result);
        this.goToNextQuestion();
    };
    handleSkip = ()=>{
        const question = this.currentQuestion;
        if (!question) return;
        const qid = question.id;
        this.dispatchEvent(new CustomEvent(VE_CLARIFY_SKIP_EVENT, {
            detail: {
                questionId: qid
            },
            bubbles: true,
            composed: true
        }));
        this.recordResult({
            questionId: qid,
            selectedValues: [],
            skipped: true
        });
        this.goToNextQuestion();
    };
    handlePrev = ()=>{
        const idx = this.currentIndex;
        if (idx > 0) this.setActiveIndex(idx - 1);
    };
    handleNext = ()=>{
        const idx = this.currentIndex;
        if (idx < this.questions.length - 1) this.setActiveIndex(idx + 1);
    };
    handleClose = ()=>{
        this.dispatchEvent(new CustomEvent(VE_CLARIFY_CLOSE_EVENT, {
            bubbles: true,
            composed: true
        }));
    };
    handleDragStart = (event, option)=>{
        if (this.disabled || option.disabled) return void event.preventDefault();
        if (event.dataTransfer) {
            event.dataTransfer.effectAllowed = 'move';
            try {
                event.dataTransfer.setData('text/plain', option.value);
            } catch  {}
            this.applyRoundedDragImage(event);
        }
        requestAnimationFrame(()=>{
            this.draggingValue = option.value;
        });
    };
    applyRoundedDragImage(event) {
        const source = event.currentTarget;
        if (!source || !event.dataTransfer) return;
        const rect = source.getBoundingClientRect();
        const ghost = source.cloneNode(true);
        ghost.removeAttribute('draggable');
        ghost.style.position = 'absolute';
        ghost.style.top = '-10000px';
        ghost.style.left = '-10000px';
        ghost.style.width = `${rect.width}px`;
        ghost.style.minHeight = `${rect.height}px`;
        ghost.style.pointerEvents = 'none';
        ghost.style.margin = '0';
        ghost.style.zIndex = '-1';
        const host = this.renderRoot;
        host.appendChild(ghost);
        const offsetX = event.clientX - rect.left;
        const offsetY = event.clientY - rect.top;
        event.dataTransfer.setDragImage(ghost, offsetX, offsetY);
        requestAnimationFrame(()=>{
            ghost.remove();
        });
    }
    reorderTo(target) {
        const question = this.currentQuestion;
        if (!question || !question.sortable) return;
        if (!this.draggingValue || this.draggingValue === target.value) return;
        const order = this.getSortOrder(question).slice();
        const fromIndex = order.indexOf(this.draggingValue);
        const toIndex = order.indexOf(target.value);
        if (-1 === fromIndex || -1 === toIndex) return;
        if (fromIndex === toIndex) return;
        order.splice(fromIndex, 1);
        order.splice(toIndex, 0, this.draggingValue);
        this.sortOrders = {
            ...this.sortOrders,
            [question.id]: order
        };
    }
    handleDragEnter = (event, target)=>{
        if (!this.draggingValue) return;
        event.preventDefault();
        this.reorderTo(target);
    };
    handleDragOver = (event)=>{
        if (!this.draggingValue) return;
        event.preventDefault();
        if (event.dataTransfer) event.dataTransfer.dropEffect = 'move';
    };
    handleDragEnd = ()=>{
        this.draggingValue = null;
    };
    handleDrop = (event)=>{
        if (!this.draggingValue) return;
        event.preventDefault();
        this.draggingValue = null;
    };
    handleHostKeyDown = (event)=>{
        if (!this.keyboardNavigation || this.disabled) return;
        const question = this.currentQuestion;
        if (!question) return;
        const target = event.target;
        const inInput = target && ('INPUT' === target.tagName || 'TEXTAREA' === target.tagName || target.isContentEditable);
        const total = this.getNavigableCount(question);
        if (0 === total) return;
        switch(event.key){
            case 'ArrowDown':
                if (inInput) return;
                event.preventDefault();
                this.highlightedIndex = this.highlightedIndex < 0 ? 0 : (this.highlightedIndex + 1) % total;
                break;
            case 'ArrowUp':
                if (inInput) return;
                event.preventDefault();
                this.highlightedIndex = this.highlightedIndex < 0 ? total - 1 : (this.highlightedIndex - 1 + total) % total;
                break;
            case 'Enter':
                if (inInput) return;
                event.preventDefault();
                this.activateHighlighted();
                break;
            case 'Escape':
            case 'Esc':
                event.preventDefault();
                if (false !== question.showSkip) this.handleSkip();
                break;
        }
    };
    syncHighlight = (index)=>{
        if (!this.keyboardNavigation) return;
        if (this.highlightedIndex !== index) this.highlightedIndex = index;
    };
    handleHostMouseLeave = ()=>{
        if (!this.keyboardNavigation) return;
        if (-1 !== this.highlightedIndex) this.highlightedIndex = -1;
    };
    getNavigableCount(question) {
        if (question.sortable) return question.options.length;
        return question.options.length + (false !== question.showInput ? 1 : 0);
    }
    activateHighlighted() {
        const question = this.currentQuestion;
        if (!question) return;
        const optCount = question.options.length;
        const idx = this.highlightedIndex;
        if (question.sortable) return void this.handleSubmit();
        if (idx < optCount) {
            const option = question.options[idx];
            if (option) this.handleOptionClick(option);
            return;
        }
        if (question.multiple) {
            const selectedCount = this.getSelected(question.id).size;
            const inputText = this.getInputText(question.id);
            if (selectedCount > 0 || inputText.length > 0) this.handleSubmit();
        } else {
            const inputText = this.getInputText(question.id);
            if (inputText.trim().length > 0) this.handleSubmit();
            else if (false !== question.showSkip) this.handleSkip();
        }
    }
    dispatchSubmit(result) {
        this.dispatchEvent(new CustomEvent(VE_CLARIFY_SUBMIT_EVENT, {
            detail: result,
            bubbles: true,
            composed: true
        }));
    }
    recordResult(result) {
        if (this.submittedIds.has(result.questionId)) return void this.upsertResult(result);
        this.submittedIds.add(result.questionId);
        this.upsertResult(result);
    }
    renderPagination() {
        const total = this.questions.length;
        if (total <= 1) return nothing;
        if (this.currentQuestion?.sortable) return nothing;
        const idx = this.currentIndex;
        const isFirst = 0 === idx;
        const isLast = idx === total - 1;
        return html`<div
      class="ve-clarify-pagination"
      aria-label="第 ${idx + 1} / ${total} 题"
    >
      <button
        type="button"
        class="ve-clarify-pagination-btn"
        aria-label="上一题"
        ?disabled=${isFirst || this.disabled}
        @click=${this.handlePrev}
      >
        <ve-icon .definition=${chevronLeftMdIcon}></ve-icon>
      </button>
      <span class="ve-clarify-pagination-text">${idx + 1}/${total}</span>
      <button
        type="button"
        class="ve-clarify-pagination-btn"
        aria-label="下一题"
        ?disabled=${isLast || this.disabled}
        @click=${this.handleNext}
      >
        <ve-icon .definition=${chevronRightMdIcon}></ve-icon>
      </button>
    </div>`;
    }
    renderHeader(question) {
        const closeBtn = this.closable ? html`<button
          type="button"
          class="ve-clarify-close"
          aria-label="关闭"
          ?disabled=${this.disabled}
          @click=${this.handleClose}
        >
          <ve-icon .definition=${closeIcon}></ve-icon>
        </button>` : nothing;
        return html`<div class="ve-clarify-header" part="header">
      <div class="ve-clarify-header-title" part="title">${question.title}</div>
      <div class="ve-clarify-header-actions">
        ${this.renderPagination()}${closeBtn}
      </div>
    </div>`;
    }
    renderBadge(question, displayIndex, selected) {
        if (question.multiple) return html`<span
        class="ve-clarify-badge ve-clarify-badge-checkbox"
        aria-hidden="true"
      >
        <span
          class=${classMap({
            've-clarify-checkbox': true,
            've-clarify-checkbox-checked': selected
        })}
        >
          ${selected ? html`<ve-icon .definition=${checkIcon}></ve-icon>` : nothing}
        </span>
      </span>`;
        return html`<span class="ve-clarify-badge" aria-hidden="true">
      ${displayIndex + 1}
    </span>`;
    }
    renderOptionItem(question, option, index) {
        const selected = this.isOptionSelected(question.id, option.value);
        const optionDisabled = this.disabled || true === option.disabled;
        const multiple = true === question.multiple;
        const highlighted = this.keyboardNavigation && this.highlightedIndex === index;
        const classes = classMap({
            've-clarify-option': true,
            've-clarify-option-selected': selected,
            've-clarify-option-disabled': optionDisabled,
            've-clarify-option-highlighted': highlighted
        });
        return html`<div
      class=${classes}
      part="option"
      role=${multiple ? 'checkbox' : 'option'}
      aria-selected=${selected ? 'true' : 'false'}
      aria-disabled=${optionDisabled ? 'true' : 'false'}
      tabindex=${optionDisabled ? '-1' : '0'}
      data-option-key=${option.key}
      @click=${()=>this.handleOptionClick(option)}
      @mouseenter=${()=>this.syncHighlight(index)}
      @focus=${()=>this.syncHighlight(index)}
    >
      <div class="ve-clarify-option-row">
        ${this.renderBadge(question, index, selected)}
        <div class="ve-clarify-option-content">
          <span class="ve-clarify-option-label">${option.label}</span>
        </div>
        ${!multiple && selected ? html`<span class="ve-clarify-arrow"
              ><ve-icon
                name="${this.currentIndex >= this.questions.length - 1 ? 'arrow-up-md' : 'arrow-right-md'}"
              ></ve-icon
            ></span>` : nothing}
      </div>
      <div class="ve-clarify-divider">
        <div class="ve-clarify-divider-line"></div>
      </div>
    </div>`;
    }
    renderSupplementRow(question, supplementIndex) {
        const isSingle = !question.multiple;
        const value = this.getInputText(question.id);
        const placeholder = question.inputPlaceholder ?? (isSingle ? '补充信息' : '补充更多，用顿号隔开');
        const highlighted = this.keyboardNavigation && this.highlightedIndex === supplementIndex;
        const badge = isSingle ? html`<span
          class="ve-clarify-badge ve-clarify-badge-supplement"
          aria-hidden="true"
        >
          <ve-icon .definition=${editIcon}></ve-icon>
        </span>` : html`<span
          class="ve-clarify-badge ve-clarify-badge-checkbox"
          aria-hidden="true"
          @click=${this.handleInputCheckboxClick}
        >
          <span
            class=${classMap({
            've-clarify-checkbox': true,
            've-clarify-checkbox-checked': this.inputSelected
        })}
          >
            ${this.inputSelected ? html`<ve-icon name="check"></ve-icon>` : nothing}
          </span>
        </span>`;
        const showInlineSubmit = isSingle && value.trim().length > 0;
        const showInlineSkip = isSingle && false !== question.showSkip && !showInlineSubmit;
        const submitIcon = this.currentIndex >= this.questions.length - 1 ? 'arrow-up-md' : 'arrow-right-md';
        const classes = classMap({
            've-clarify-option': true,
            've-clarify-option-supplement': true,
            've-clarify-option-highlighted': highlighted
        });
        return html`<div
      class=${classes}
      @mouseenter=${()=>this.syncHighlight(supplementIndex)}
    >
      <div class="ve-clarify-option-row">
        ${badge}
        <input
          class="ve-clarify-input-field"
          type="text"
          .value=${value}
          placeholder=${placeholder}
          ?disabled=${this.disabled}
          @input=${this.handleInputChange}
        />
        ${showInlineSkip ? html`<button
              type="button"
              class="ve-clarify-btn-skip"
              ?disabled=${this.disabled}
              @click=${this.handleSkip}
            >
              跳过
            </button>` : nothing}
        ${showInlineSubmit ? html`<button
              type="button"
              class="ve-clarify-btn-submit"
              aria-label="提交"
              ?disabled=${this.disabled}
              @click=${this.handleSubmit}
            >
              <ve-icon name=${submitIcon}></ve-icon>
            </button>` : nothing}
      </div>
      <div class="ve-clarify-divider">
        <div class="ve-clarify-divider-line"></div>
      </div>
    </div>`;
    }
    renderOptionsList(question) {
        if (question.sortable) return this.renderSortableList(question);
        const showInput = false !== question.showInput;
        return html`<div class="ve-clarify-options" part="options">
      ${question.options.map((option, index)=>this.renderOptionItem(question, option, index))}
      ${showInput ? this.renderSupplementRow(question, question.options.length) : nothing}
    </div>`;
    }
    renderSortableList(question) {
        const sortedOptions = this.getSortedOptions(question);
        return html`<div class="ve-clarify-options" role="list" part="options">
      ${sortedOptions.map((option, index)=>this.renderSortableItem(option, index))}
    </div>`;
    }
    renderSortableItem(option, index) {
        const optionDisabled = this.disabled || true === option.disabled;
        const highlighted = this.keyboardNavigation && this.highlightedIndex === index;
        const isDragging = this.draggingValue === option.value;
        const classes = classMap({
            've-clarify-option': true,
            've-clarify-option-sortable': true,
            've-clarify-option-disabled': optionDisabled,
            've-clarify-option-highlighted': highlighted,
            've-clarify-option-dragging': isDragging
        });
        return html`<div
      class=${classes}
      part="option"
      role="listitem"
      aria-disabled=${optionDisabled ? 'true' : 'false'}
      aria-grabbed=${isDragging ? 'true' : 'false'}
      data-option-key=${option.key}
      data-option-value=${option.value}
      draggable=${optionDisabled ? 'false' : 'true'}
      @mouseenter=${()=>this.syncHighlight(index)}
      @dragstart=${(e)=>this.handleDragStart(e, option)}
      @dragenter=${(e)=>this.handleDragEnter(e, option)}
      @dragover=${this.handleDragOver}
      @dragend=${this.handleDragEnd}
      @drop=${this.handleDrop}
    >
      <div class="ve-clarify-option-row">
        <span class="ve-clarify-badge" aria-hidden="true">${index + 1}</span>
        <div class="ve-clarify-option-content">
          <span class="ve-clarify-option-label">${option.label}</span>
        </div>
        <span
          class="ve-clarify-drag-handle"
          aria-label="拖拽排序"
          aria-hidden="true"
          ><ve-icon name="drag"></ve-icon
        ></span>
      </div>
      <div class="ve-clarify-divider">
        <div class="ve-clarify-divider-line"></div>
      </div>
    </div>`;
    }
    renderFooter(question) {
        if (question.sortable) return html`<div class="ve-clarify-footer" part="footer">
        <div class="ve-clarify-footer-info">
          <span class="ve-clarify-footer-hint">拖拽排序</span>
        </div>
        <div class="ve-clarify-footer-actions">
          ${false !== question.showSkip ? html`<button
                type="button"
                class="ve-clarify-btn-skip"
                ?disabled=${this.disabled}
                @click=${this.handleSkip}
              >
                跳过
              </button>` : nothing}
          <button
            type="button"
            class="ve-clarify-btn-submit"
            aria-label="提交"
            ?disabled=${this.disabled}
            @click=${this.handleSubmit}
          >
            <ve-icon
              name="${this.currentIndex >= this.questions.length - 1 ? 'arrow-up-md' : 'arrow-right-md'}"
            ></ve-icon>
          </button>
        </div>
      </div>`;
        if (!question.multiple) return nothing;
        const selectedCount = this.getSelected(question.id).size;
        const inputText = this.getInputText(question.id);
        const canSubmit = selectedCount > 0 || this.inputSelected && inputText.length > 0;
        return html`<div class="ve-clarify-footer" part="footer">
      <div class="ve-clarify-footer-info">
        <span class="ve-clarify-footer-count">${selectedCount}</span>
        <span class="ve-clarify-footer-label">已选</span>
      </div>
      <div class="ve-clarify-footer-actions">
        ${false !== question.showSkip ? html`<button
              type="button"
              class="ve-clarify-btn-skip"
              ?disabled=${this.disabled}
              @click=${this.handleSkip}
            >
              跳过
            </button>` : nothing}
        <button
          type="button"
          class="ve-clarify-btn-submit"
          aria-label="提交"
          ?disabled=${this.disabled || !canSubmit}
          @click=${this.handleSubmit}
        >
          <ve-icon
            name="${this.currentIndex >= this.questions.length - 1 ? 'arrow-up-md' : 'arrow-right-md'}"
          ></ve-icon>
        </button>
      </div>
    </div>`;
    }
    renderKeyboardHint() {
        if (!this.showKeyboardHint) return nothing;
        const question = this.currentQuestion;
        const switchLabel = question?.sortable ? '拖拽' : '切换';
        const confirmLabel = question?.multiple || question?.sortable ? '确认' : '选择';
        return html`<div
      class="ve-clarify-keyboard-hint"
      part="keyboard-hint"
      aria-hidden="true"
    >
      <span class="ve-clarify-keyboard-hint-item">
        <span class="ve-clarify-keyboard-hint-icons">
          <ve-icon .definition=${downUpIcon}></ve-icon>
        </span>
        <span class="ve-clarify-keyboard-hint-text">${switchLabel}</span>
      </span>
      <span class="ve-clarify-keyboard-hint-item">
        <ve-icon .definition=${cornerDownLeftIcon}></ve-icon>
        <span class="ve-clarify-keyboard-hint-text">${confirmLabel}</span>
      </span>
      <span class="ve-clarify-keyboard-hint-item">
        <span class="ve-clarify-keyboard-hint-key">esc</span>
        <span class="ve-clarify-keyboard-hint-text">跳过</span>
      </span>
    </div>`;
    }
    render() {
        const question = this.currentQuestion;
        if (!question) return null;
        return html`<div class="ve-clarify" part="panel">
        ${this.renderHeader(question)}
        <div class="ve-clarify-body">${this.renderOptionsList(question)}</div>
        ${this.renderFooter(question)}
      </div>
      ${this.renderKeyboardHint()}`;
    }
}
defineCustomElement(VE_CLARIFY_TAG, VeClarifyElement);
export { VE_CLARIFY_ACTIVE_INDEX_CHANGE_EVENT, VE_CLARIFY_CLOSE_EVENT, VE_CLARIFY_COMPLETE_EVENT, VE_CLARIFY_SKIP_EVENT, VE_CLARIFY_SUBMIT_EVENT, VE_CLARIFY_TAG, VeClarifyElement };
