// @ve-design/web/dist/ve-citation.js
import { css, html, nothing } from "lit";
import { property, state } from "lit/decorators.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { VeDropdownElement, VE_DROPDOWN_TAG } from "./97605.js";
import { globeIcon } from "./icons/globe.js";
import { documentIcon } from "./icons/document.js";
import { chevronDownSmIcon } from "./67641.js";
const acCitationTokenDefaults = css`
  :host {
    /* === Surface === */
    --ve-citation-background: var(--color-bg-base, #ffffff);
    --ve-citation-surface-background: var(--color-bg-surface, #f6f6fc);
    --ve-citation-hover-background: var(--color-bg-secondary, #f7f8fa);
    --ve-citation-border-color: var(--color-border-default, #e5e6eb);
    --ve-citation-border-width: var(--stroke-weight-base, 0.5px);
    --ve-citation-radius: var(--radius-lg, 12px);

    /* === Header === */
    --ve-citation-header-padding-block: var(--space-xs, 8px);
    --ve-citation-header-padding-inline: var(--space-s, 12px);
    --ve-citation-header-gap: var(--space-xxs, 6px);
    --ve-citation-title-color: var(--color-text-primary, #1d2129);
    --ve-citation-title-font-size: var(--text-body, 14px);
    --ve-citation-title-font-weight: var(--font-weight-medium, 500);
    --ve-citation-title-line-height: 22px;
    --ve-citation-icon-color: var(--color-icon-secondary, #86909c);
    --ve-citation-icon-size: 16px;
    --ve-citation-chevron-rotation-collapsed: 0deg;
    --ve-citation-chevron-rotation-expanded: 90deg;

    /* === Item (card form, aligned with reference .ve-citation__source) === */
    --ve-citation-list-padding-block: var(--space-xxs, 6px);
    --ve-citation-list-padding-inline: var(--space-xs, 8px);
    --ve-citation-list-gap: 0;
    --ve-citation-item-padding-block: var(--space-xs, 8px);
    --ve-citation-item-padding-inline: var(--space-s, 12px);
    --ve-citation-item-radius: var(--radius-md, 8px);
    --ve-citation-item-gap: var(--space-xxs, 6px);

    --ve-citation-item-title-color: var(--color-text-primary, #1d2129);
    --ve-citation-item-title-font-size: var(--text-ui-strong, 14px);
    --ve-citation-item-title-font-weight: var(--font-weight-medium, 500);
    --ve-citation-item-title-line-height: 28px;

    --ve-citation-item-description-color: var(
      --color-text-tertiary,
      #86909c
    );
    --ve-citation-item-description-font-size: var(--text-body-sm, 14px);
    --ve-citation-item-description-line-height: var(--line-height-body, 22px);

    --ve-citation-item-meta-color: var(--color-text-tertiary, #86909c);
    --ve-citation-item-meta-font-size: var(--text-caption, 12px);
    --ve-citation-item-meta-line-height: var(--line-height-body, 22px);

    --ve-citation-item-icon-size: 16px;
    --ve-citation-item-icon-radius: var(--radius-full, 99px);
    --ve-citation-item-icon-background: var(--color-bg-base, #ffffff);
    --ve-citation-item-icon-color: var(--color-icon-secondary, #86909c);
    --ve-citation-item-divider-color: var(--color-border-default, #e5e6eb);
    --ve-citation-item-divider-size: var(--stroke-weight-base, 0.5px);

    --ve-citation-item-active-background: var(
      --color-bg-secondary,
      #eaf3ff
    );

    /* === Inline / popover === */
    --ve-citation-inline-gap: var(--space-xxxs, 4px);
    --ve-citation-inline-color: var(--color-text-tertiary, #86909c);
    --ve-citation-inline-trigger-height: 28px;
    --ve-citation-inline-trigger-padding-inline: var(--space-xxxs, 4px);
    --ve-citation-inline-trigger-radius: var(--radius-sm, 4px);
    --ve-citation-inline-trigger-font-size: var(--text-body-sm, 13px);
    --ve-citation-inline-trigger-hover-background: var(
      --color-bg-secondary,
      #f7f8fa
    );
    --ve-citation-inline-count-size: 18px;
    --ve-citation-inline-count-padding-inline: var(--space-xxxs, 4px);
    --ve-citation-inline-count-background: var(
      --color-bg-secondary,
      #f7f8fa
    );
    --ve-citation-inline-count-color: var(--color-text-tertiary, #86909c);
    --ve-citation-inline-count-font-size: var(--text-caption, 12px);
    --ve-citation-popover-overlay-width: 320px;
    --ve-citation-popover-padding: var(--space-xs, 8px);
    --ve-citation-popover-radius: var(--radius-lg, 12px);
    --ve-citation-popover-shadow: var(--shadow-md, 0 4px 12px rgba(0, 0, 0, 0.08));
    --ve-citation-popover-background: var(--color-bg-overlay, #ffffff);

    /* === Link list (mode='link') === */
    --ve-citation-link-list-width: 100%;
    --ve-citation-link-list-padding: var(--space-xs, 8px);
    --ve-citation-link-list-radius: var(--radius-md, 8px);
    --ve-citation-link-list-border-color: var(
      --color-border-interactive,
      #c9cdd4
    );
    --ve-citation-link-list-border-style: dashed;
    --ve-citation-link-list-border-width: var(--stroke-weight-base, 0.5px);
    --ve-citation-link-list-heading-color: var(--color-text-primary, #1d2129);
    --ve-citation-link-list-heading-font-size: var(--text-ui-strong, 14px);
    --ve-citation-link-list-heading-line-height: 28px;
    --ve-citation-link-row-gap: var(--space-xxs, 6px);
    --ve-citation-link-row-min-height: 34px;
    --ve-citation-link-color: var(--color-text-info, #165dff);
    --ve-citation-link-hover-background: color-mix(
      in srgb,
      var(--color-text-info, #165dff) 8%,
      transparent
    );
    --ve-citation-link-index-size: 18px;
    --ve-citation-link-index-background: var(--color-bg-secondary, #f7f8fa);
    --ve-citation-link-index-color: var(--color-text-tertiary, #86909c);
    --ve-citation-link-index-font-size: var(--text-caption, 12px);
    --ve-citation-more-color: var(--color-text-tertiary, #86909c);
    --ve-citation-more-hover-background: var(--color-bg-secondary, #f7f8fa);
    --ve-citation-more-radius: var(--radius-sm, 4px);
    --ve-citation-more-height: 28px;
    --ve-citation-more-padding-block: 0;
    --ve-citation-more-padding-inline: var(--space-xxxs, 4px);
    --ve-citation-more-font-size: var(--text-body, 14px);
    --ve-citation-more-arrow-size: 16px;

    /* === Animation === */
    --ve-citation-transition: background-color 150ms ease,
      transform 150ms ease;
  }
`;
const token = acCitationTokenDefaults;
const ve_citation = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      inline-size: 100%;
      color: var(--color-text-primary, #1d2129);
      font-family: var(--font-sans);
    }

    :host([inline]) {
      display: inline-flex;
      align-items: center;
      inline-size: auto;
      gap: var(--ve-citation-inline-gap);
    }

    /* ===========================================================
     * Block (folding) form
     * =========================================================== */
    .ve-citation-content {
      display: flex;
      flex-direction: column;
      gap: var(--ve-citation-list-gap);
      inline-size: 100%;
    }

    :host([mode='link']) .ve-citation-content {
      gap: 0;
    }

    /* ===========================================================
     * Card item (default form, aligned with reference .ve-citation__source)
     * =========================================================== */
    .ve-citation-item {
      display: flex;
      flex-direction: column;
      gap: var(--ve-citation-item-gap);
      box-sizing: border-box;
      inline-size: 100%;
      padding-block: var(--ve-citation-item-padding-block);
      padding-inline: var(--ve-citation-item-padding-inline);
      border: 0;
      border-radius: var(--ve-citation-item-radius);
      background: var(--ve-citation-background);
      color: inherit;
      text-align: start;
      text-decoration: none;
      cursor: pointer;
      transition: var(--ve-citation-transition);
      outline: none;
    }

    .ve-citation-item:hover,
    .ve-citation-item:focus-visible {
      background: var(--ve-citation-hover-background);
    }

    .ve-citation-item[aria-current='true'] {
      background: var(--ve-citation-item-active-background);
    }

    .ve-citation-item-header {
      display: flex;
      align-items: center;
      gap: var(--ve-citation-item-gap);
      inline-size: 100%;
      min-block-size: 18px;
      min-inline-size: 0;
    }

    .ve-citation-item-source {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs, 4px);
      color: var(--ve-citation-item-meta-color);
      font-size: var(--ve-citation-item-meta-font-size);
      line-height: var(--ve-citation-item-meta-line-height);
      white-space: nowrap;
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ve-citation-item-source-name {
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ve-citation-item-icon {
      display: inline-flex;
      flex: 0 0 auto;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-citation-item-icon-size);
      block-size: var(--ve-citation-item-icon-size);
      border-radius: var(--ve-citation-item-icon-radius);
      background: var(--ve-citation-item-icon-background);
      color: var(--ve-citation-item-icon-color);
      overflow: hidden;
    }

    .ve-citation-item-icon img {
      inline-size: 100%;
      block-size: 100%;
      object-fit: cover;
      display: block;
    }

    .ve-citation-item-icon ve-icon {
      --ve-icon-size: var(--ve-citation-item-icon-size);
    }

    .ve-citation-item-divider {
      flex: 0 0 auto;
      inline-size: var(--ve-citation-item-divider-size);
      block-size: 12px;
      background: var(--ve-citation-item-divider-color);
    }

    .ve-citation-item-date {
      color: var(--ve-citation-item-meta-color);
      font-size: var(--ve-citation-item-meta-font-size);
      line-height: var(--ve-citation-item-meta-line-height);
      white-space: nowrap;
    }

    .ve-citation-item-index {
      display: inline-flex;
      flex: 0 0 auto;
      align-items: center;
      justify-content: center;
      inline-size: 18px;
      block-size: 18px;
      margin-inline-start: auto;
      border-radius: var(--radius-full, 99px);
      background: var(--ve-citation-link-index-background);
      color: var(--ve-citation-link-index-color);
      font-size: var(--ve-citation-link-index-font-size);
      line-height: 1;
    }

    .ve-citation-item-body {
      display: flex;
      flex-direction: column;
      gap: var(--space-xxxs, 2px);
      inline-size: 100%;
      min-inline-size: 0;
    }

    .ve-citation-item-title {
      margin: 0;
      color: var(--ve-citation-item-title-color);
      font-size: var(--ve-citation-item-title-font-size);
      font-weight: var(--ve-citation-item-title-font-weight);
      line-height: var(--ve-citation-item-title-line-height);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-citation-item-description {
      margin: 0;
      color: var(--ve-citation-item-description-color);
      font-size: var(--ve-citation-item-description-font-size);
      line-height: var(--ve-citation-item-description-line-height);
      display: -webkit-box;
      -webkit-line-clamp: 2;
      -webkit-box-orient: vertical;
      overflow: hidden;
      word-break: break-word;
    }

    .ve-citation-item-url {
      margin: 0;
      color: var(--ve-citation-item-meta-color);
      font-size: var(--ve-citation-item-meta-font-size);
      line-height: var(--ve-citation-item-meta-line-height);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    /* ===========================================================
     * Link mode (collapsible link list)
     * =========================================================== */
    .ve-citation-link-row {
      display: flex;
      align-items: center;
      gap: var(--ve-citation-link-row-gap);
      inline-size: 100%;
      min-block-size: var(--ve-citation-link-row-min-height);
      min-inline-size: 0;
      padding-block: var(--space-xxxs, 2px);
      color: var(--color-text-primary, #1d2129);
      font-size: var(--text-body, 14px);
      line-height: 22px;
    }

    .ve-citation-link-index {
      display: inline-flex;
      flex: 0 0 auto;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-citation-link-index-size);
      block-size: var(--ve-citation-link-index-size);
      border-radius: var(--radius-full, 99px);
      background: var(--ve-citation-link-index-background);
      color: var(--ve-citation-link-index-color);
      font-size: var(--ve-citation-link-index-font-size);
      line-height: 1;
    }

    .ve-citation-link {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs, 4px);
      flex: 0 1 auto;
      min-inline-size: 0;
      padding-block: 2px;
      padding-inline: var(--space-xxxs, 4px);
      border-radius: var(--radius-md, 6px);
      color: var(--ve-citation-link-color);
      text-decoration: none;
      white-space: nowrap;
      overflow: hidden;
      transition: background-color 150ms ease;
    }

    .ve-citation-link:hover,
    .ve-citation-link:focus-visible {
      background: var(--ve-citation-link-hover-background);
      outline: none;
    }

    .ve-citation-link-icon {
      display: inline-flex;
      flex: 0 0 auto;
      align-items: center;
      justify-content: center;
      inline-size: 14px;
      block-size: 14px;
      color: inherit;
    }

    .ve-citation-link-icon img {
      inline-size: 100%;
      block-size: 100%;
      object-fit: cover;
      display: block;
    }

    .ve-citation-link-icon ve-icon {
      --ve-icon-size: 14px;
    }

    .ve-citation-link-label {
      flex: 0 1 auto;
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
      color: inherit;
    }

    .ve-citation-link-date {
      flex: 0 0 auto;
      color: var(--ve-citation-item-meta-color);
      font-size: var(--ve-citation-item-meta-font-size);
      line-height: var(--ve-citation-item-meta-line-height);
      white-space: nowrap;
    }

    .ve-citation-more {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      align-self: flex-start;
      gap: var(--space-xxxs, 4px);
      box-sizing: border-box;
      block-size: var(--ve-citation-more-height);
      min-inline-size: 0;
      max-inline-size: max-content;
      margin-block-start: var(--space-xxs, 8px);
      padding-block: var(--ve-citation-more-padding-block);
      padding-inline: var(--ve-citation-more-padding-inline);
      border: 0;
      border-radius: var(--ve-citation-more-radius);
      background: transparent;
      color: var(--ve-citation-more-color);
      font: inherit;
      font-size: var(--ve-citation-more-font-size);
      line-height: 22px;
      cursor: pointer;
      transition: background-color 150ms ease;
    }

    .ve-citation-more:hover,
    .ve-citation-more:focus-visible {
      background: var(--ve-citation-more-hover-background);
      outline: none;
    }

    .ve-citation-more[disabled] {
      cursor: default;
      opacity: 0.5;
    }

    .ve-citation-more-arrow {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-citation-more-arrow-size);
      block-size: var(--ve-citation-more-arrow-size);
      color: inherit;
      transform: rotate(0deg);
      transition: transform 200ms ease;
    }

    .ve-citation-more-arrow ve-icon {
      inline-size: 100%;
      block-size: 100%;
    }

    .ve-citation-more[aria-expanded='true'] .ve-citation-more-arrow {
      transform: rotate(180deg);
    }

    /* ===========================================================
     * Inline (popover) form
     * =========================================================== */
    .ve-citation-inline-trigger {
      position: relative;
      display: inline-flex;
      align-items: center;
      gap: var(--ve-citation-inline-gap);
      box-sizing: border-box;
      block-size: var(--ve-citation-inline-trigger-height);
      padding-block: 0;
      padding-inline: var(--ve-citation-inline-trigger-padding-inline);
      border: 0;
      margin: 0;
      border-radius: var(--ve-citation-inline-trigger-radius);
      background: transparent;
      color: var(--ve-citation-inline-color);
      font: inherit;
      font-size: var(--ve-citation-inline-trigger-font-size);
      line-height: 1.5;
      cursor: pointer;
      outline: none;
      transition: background-color 150ms ease;
    }

    .ve-citation-inline-trigger:hover,
    .ve-citation-inline-trigger:focus-visible {
      background: var(--ve-citation-inline-trigger-hover-background);
    }

    .ve-citation-inline-trigger[disabled] {
      cursor: default;
      opacity: 0.5;
    }

    .ve-citation-inline-count {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      min-inline-size: var(--ve-citation-inline-count-size);
      block-size: var(--ve-citation-inline-count-size);
      padding-inline: var(--ve-citation-inline-count-padding-inline);
      border-radius: var(--radius-full, 99px);
      background: var(--ve-citation-inline-count-background);
      color: var(--ve-citation-inline-count-color);
      font-family: var(--font-en, inherit);
      font-size: var(--ve-citation-inline-count-font-size);
      line-height: 1.5;
    }

    .ve-citation-popover {
      box-sizing: border-box;
      inline-size: var(--ve-citation-popover-overlay-width);
      max-inline-size: 90vw;
      display: flex;
      flex-direction: column;
      gap: var(--ve-citation-list-gap);
    }

    /* 复用 ve-dropdown 浮层定位与白底圆角阴影；自定义滚动区域内边距留给 dropdown-menu。 */
    ve-dropdown.ve-citation-dropdown {
      --ve-dropdown-min-width: 0;
      --ve-dropdown-padding-block: var(--ve-citation-popover-padding);
      --ve-dropdown-padding-inline: var(--ve-citation-popover-padding);
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _init_items, _init_extra__init_items, _init_mode, _init_extra__init_mode, _init_triggerLabel, _init_extra__init_triggerLabel, _init_showCount, _init_extra__init_showCount, _init_showIndex, _init_extra__init_showIndex, _init_inline, _init_extra__init_inline, _init_expanded, _init_extra__init_expanded, _init_defaultExpanded, _init_extra__init_defaultExpanded, _init_collapseCount, _init_extra__init_collapseCount, _init_activeKey, _init_extra__init_activeKey, _init_popoverOverlayWidth, _init_extra__init_popoverOverlayWidth, _init_disabled, _init_extra__init_disabled, _init_internalExpanded, _init_extra__init_internalExpanded;
const VE_CITATION_TAG = 've-citation';
_dec = property({
    attribute: false
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    type: String,
    attribute: 'trigger-label'
}), _dec3 = property({
    type: Boolean,
    attribute: 'show-count'
}), _dec4 = property({
    type: Boolean,
    attribute: 'show-index'
}), _dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = property({
    type: Boolean,
    reflect: true
}), _dec7 = property({
    type: Boolean,
    attribute: 'default-expanded'
}), _dec8 = property({
    type: Number,
    attribute: 'collapse-count'
}), _dec9 = property({
    attribute: 'active-key'
}), _dec10 = property({
    attribute: 'popover-overlay-width'
}), _dec11 = property({
    type: Boolean,
    reflect: true
}), _dec12 = state();
class VeCitationElement extends VeElement {
    static{
        ({ e: [_init_items, _init_extra__init_items, _init_mode, _init_extra__init_mode, _init_triggerLabel, _init_extra__init_triggerLabel, _init_showCount, _init_extra__init_showCount, _init_showIndex, _init_extra__init_showIndex, _init_inline, _init_extra__init_inline, _init_expanded, _init_extra__init_expanded, _init_defaultExpanded, _init_extra__init_defaultExpanded, _init_collapseCount, _init_extra__init_collapseCount, _init_activeKey, _init_extra__init_activeKey, _init_popoverOverlayWidth, _init_extra__init_popoverOverlayWidth, _init_disabled, _init_extra__init_disabled, _init_internalExpanded, _init_extra__init_internalExpanded] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "items"
            ],
            [
                _dec1,
                1,
                "mode"
            ],
            [
                _dec2,
                1,
                "triggerLabel"
            ],
            [
                _dec3,
                1,
                "showCount"
            ],
            [
                _dec4,
                1,
                "showIndex"
            ],
            [
                _dec5,
                1,
                "inline"
            ],
            [
                _dec6,
                1,
                "expanded"
            ],
            [
                _dec7,
                1,
                "defaultExpanded"
            ],
            [
                _dec8,
                1,
                "collapseCount"
            ],
            [
                _dec9,
                1,
                "activeKey"
            ],
            [
                _dec10,
                1,
                "popoverOverlayWidth"
            ],
            [
                _dec11,
                1,
                "disabled"
            ],
            [
                _dec12,
                1,
                "internalExpanded"
            ]
        ]));
    }
    static styles = ve_citation;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement,
        [VE_DROPDOWN_TAG]: VeDropdownElement
    };
    #___private_items_1 = (()=>{
        const _value = _init_items(this, []);
        _init_extra__init_items(this);
        return _value;
    })();
    get items() {
        return this.#___private_items_1;
    }
    set items(_v) {
        this.#___private_items_1 = _v;
    }
    #___private_mode_2 = (()=>{
        const _value = _init_mode(this, 'card');
        _init_extra__init_mode(this);
        return _value;
    })();
    get mode() {
        return this.#___private_mode_2;
    }
    set mode(_v) {
        this.#___private_mode_2 = _v;
    }
    #___private_triggerLabel_3 = (()=>{
        const _value = _init_triggerLabel(this, void 0);
        _init_extra__init_triggerLabel(this);
        return _value;
    })();
    get triggerLabel() {
        return this.#___private_triggerLabel_3;
    }
    set triggerLabel(_v) {
        this.#___private_triggerLabel_3 = _v;
    }
    #___private_showCount_4 = (()=>{
        const _value = _init_showCount(this, true);
        _init_extra__init_showCount(this);
        return _value;
    })();
    get showCount() {
        return this.#___private_showCount_4;
    }
    set showCount(_v) {
        this.#___private_showCount_4 = _v;
    }
    #___private_showIndex_5 = (()=>{
        const _value = _init_showIndex(this, true);
        _init_extra__init_showIndex(this);
        return _value;
    })();
    get showIndex() {
        return this.#___private_showIndex_5;
    }
    set showIndex(_v) {
        this.#___private_showIndex_5 = _v;
    }
    #___private_inline_6 = (()=>{
        const _value = _init_inline(this, false);
        _init_extra__init_inline(this);
        return _value;
    })();
    get inline() {
        return this.#___private_inline_6;
    }
    set inline(_v) {
        this.#___private_inline_6 = _v;
    }
    #___private_expanded_7 = (()=>{
        const _value = _init_expanded(this, void 0);
        _init_extra__init_expanded(this);
        return _value;
    })();
    get expanded() {
        return this.#___private_expanded_7;
    }
    set expanded(_v) {
        this.#___private_expanded_7 = _v;
    }
    #___private_defaultExpanded_8 = (()=>{
        const _value = _init_defaultExpanded(this, void 0);
        _init_extra__init_defaultExpanded(this);
        return _value;
    })();
    get defaultExpanded() {
        return this.#___private_defaultExpanded_8;
    }
    set defaultExpanded(_v) {
        this.#___private_defaultExpanded_8 = _v;
    }
    #___private_collapseCount_9 = (()=>{
        const _value = _init_collapseCount(this, 5);
        _init_extra__init_collapseCount(this);
        return _value;
    })();
    get collapseCount() {
        return this.#___private_collapseCount_9;
    }
    set collapseCount(_v) {
        this.#___private_collapseCount_9 = _v;
    }
    #___private_activeKey_10 = (()=>{
        const _value = _init_activeKey(this, void 0);
        _init_extra__init_activeKey(this);
        return _value;
    })();
    get activeKey() {
        return this.#___private_activeKey_10;
    }
    set activeKey(_v) {
        this.#___private_activeKey_10 = _v;
    }
    #___private_popoverOverlayWidth_11 = (()=>{
        const _value = _init_popoverOverlayWidth(this, void 0);
        _init_extra__init_popoverOverlayWidth(this);
        return _value;
    })();
    get popoverOverlayWidth() {
        return this.#___private_popoverOverlayWidth_11;
    }
    set popoverOverlayWidth(_v) {
        this.#___private_popoverOverlayWidth_11 = _v;
    }
    #___private_disabled_12 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_12;
    }
    set disabled(_v) {
        this.#___private_disabled_12 = _v;
    }
    #___private_internalExpanded_13 = (()=>{
        const _value = _init_internalExpanded(this, true);
        _init_extra__init_internalExpanded(this);
        return _value;
    })();
    get internalExpanded() {
        return this.#___private_internalExpanded_13;
    }
    set internalExpanded(_v) {
        this.#___private_internalExpanded_13 = _v;
    }
    seededInitialState = false;
    get actualExpanded() {
        if ('boolean' == typeof this.expanded) return this.expanded;
        return this.internalExpanded;
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (!this.seededInitialState) {
            if ('boolean' == typeof this.defaultExpanded) this.internalExpanded = this.defaultExpanded;
            else this.internalExpanded = 'link' !== this.mode;
            this.seededInitialState = true;
        }
        if (changedProperties.has('expanded') && 'boolean' == typeof this.expanded) this.internalExpanded = this.expanded;
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (this.actualExpanded) this.dataset['expanded'] = '';
        else delete this.dataset['expanded'];
        if (changedProperties.has('popoverOverlayWidth')) this.applyPopoverWidthVar();
    }
    applyPopoverWidthVar() {
        if (void 0 === this.popoverOverlayWidth || '' === this.popoverOverlayWidth) return void this.style.removeProperty('--ve-citation-popover-overlay-width');
        const value = String(this.popoverOverlayWidth);
        this.style.setProperty('--ve-citation-popover-overlay-width', /^\d+(?:\.\d+)?$/.test(value) ? `${value}px` : value);
    }
    toggleExpanded() {
        if (this.disabled) return;
        const next = !this.actualExpanded;
        if (void 0 === this.expanded) this.internalExpanded = next;
        this.dispatchEvent(new CustomEvent('ve-citation-expand', {
            detail: {
                expanded: next
            },
            bubbles: true,
            composed: true
        }));
    }
    handleItemClick(event, item, index) {
        if (this.disabled) return void event.preventDefault();
        this.dispatchEvent(new CustomEvent('ve-citation-click', {
            detail: {
                item,
                index
            },
            bubbles: true,
            composed: true
        }));
    }
    getItemKey(item, index) {
        return item.key ?? index;
    }
    renderCardItem(item, index) {
        const key = this.getItemKey(item, index);
        const isActive = void 0 !== this.activeKey && this.activeKey === key;
        const hasUrl = Boolean(item.url);
        const hostname = hasUrl ? this.tryGetHostname(item.url ?? '') : '';
        const iconNode = this.renderItemIcon(item.icon, globeIcon);
        const sourceLabel = item.source || hostname;
        const header = html`
      <header class="ve-citation-item-header">
        <span class="ve-citation-item-source" part="item-source">
          <span class="ve-citation-item-icon" part="item-icon"
            >${iconNode}</span
          >
          ${sourceLabel ? html`<span class="ve-citation-item-source-name"
                >${sourceLabel}</span
              >` : nothing}
        </span>
        ${item.date ? html`
              <span class="ve-citation-item-divider" aria-hidden="true"></span>
              <time
                class="ve-citation-item-date"
                part="item-date"
                datetime=${item.date}
                >${item.date}</time
              >
            ` : nothing}
        ${this.showIndex ? html`<span class="ve-citation-item-index" part="item-index"
              >${index + 1}</span
            >` : nothing}
      </header>
    `;
        const body = html`
      <div class="ve-citation-item-body">
        <h3 class="ve-citation-item-title" part="item-title">${item.title}</h3>
        ${item.description ? html`<p
              class="ve-citation-item-description"
              part="item-description"
            >
              ${item.description}
            </p>` : nothing}
      </div>
    `;
        const partAttr = `item${isActive ? ' item-active' : ''}`;
        const ariaCurrent = isActive ? 'true' : void 0;
        if (hasUrl) return html`<a
        class="ve-citation-item"
        part=${partAttr}
        href=${item.url ?? ''}
        target="_blank"
        rel="noopener noreferrer"
        aria-current=${ifDefined(ariaCurrent)}
        @click=${(event)=>this.handleItemClick(event, item, index)}
        >${header}${body}</a
      >`;
        return html`<button
      type="button"
      class="ve-citation-item"
      part=${partAttr}
      aria-current=${ifDefined(ariaCurrent)}
      ?disabled=${this.disabled}
      @click=${(event)=>this.handleItemClick(event, item, index)}
    >
      ${header}${body}
    </button>`;
    }
    tryGetHostname(url) {
        try {
            return new URL(url).hostname;
        } catch  {
            return url;
        }
    }
    renderItemIcon(icon, fallback) {
        if (icon && 'object' == typeof icon) return html`<ve-icon .definition=${icon}></ve-icon>`;
        if ('string' == typeof icon && icon) {
            if (this.isIconName(icon)) return html`<ve-icon name=${icon}></ve-icon>`;
            return html`<img src=${icon} alt="" />`;
        }
        return html`<ve-icon .definition=${fallback}></ve-icon>`;
    }
    isIconName(value) {
        return /^[a-z0-9-]+$/.test(value);
    }
    renderCardList(items) {
        return html`${items.map((item, index)=>this.renderCardItem(item, index))}`;
    }
    renderLinkRow(item, index) {
        const iconNode = this.renderItemIcon(item.icon, documentIcon);
        const linkContent = html`
      <span class="ve-citation-link-icon" aria-hidden="true">${iconNode}</span>
      <span class="ve-citation-link-label">${item.title}</span>
    `;
        const link = item.url ? html`<a
          class="ve-citation-link"
          part="link"
          href=${item.url}
          target="_blank"
          rel="noopener noreferrer"
          @click=${(event)=>this.handleItemClick(event, item, index)}
          >${linkContent}</a
        >` : html`<button
          type="button"
          class="ve-citation-link"
          part="link"
          ?disabled=${this.disabled}
          @click=${(event)=>this.handleItemClick(event, item, index)}
        >
          ${linkContent}
        </button>`;
        return html`
      <div class="ve-citation-link-row" part="link-row">
        ${this.showIndex ? html`<span class="ve-citation-link-index" part="link-index"
              >${index + 1}</span
            >` : nothing}
        ${link}
        ${item.date ? html`<time
              class="ve-citation-link-date"
              datetime=${item.date}
              part="link-date"
              >${item.date}</time
            >` : nothing}
      </div>
    `;
    }
    renderLinkList(items) {
        const expanded = this.actualExpanded;
        const limit = Math.max(0, 0 | this.collapseCount);
        const total = items.length;
        const visible = !expanded && limit > 0 && total > limit ? items.slice(0, limit) : items;
        const hasMore = total > limit && limit > 0;
        return html`
      ${visible.map((item, index)=>this.renderLinkRow(item, index))}
      ${hasMore ? html`
            <button
              type="button"
              class="ve-citation-more"
              part="more"
              aria-expanded=${expanded ? 'true' : 'false'}
              ?disabled=${this.disabled}
              @click=${this.toggleExpanded}
            >
              <span>${expanded ? '折叠' : '查看全部'}</span>
              <span class="ve-citation-more-arrow" aria-hidden="true">
                <ve-icon .definition=${chevronDownSmIcon}></ve-icon>
              </span>
            </button>
          ` : nothing}
    `;
    }
    renderListContent() {
        if ('link' === this.mode) return this.renderLinkList(this.items);
        return this.renderCardList(this.items);
    }
    renderBlock() {
        return html`
      <div class="ve-citation-content" part="content">
        <slot>${this.renderListContent()}</slot>
      </div>
    `;
    }
    renderInline() {
        const count = this.items.length;
        const hasCustomLabel = void 0 !== this.triggerLabel;
        const label = this.triggerLabel ?? '参考来源';
        const triggerContent = hasCustomLabel ? html`<span class="ve-citation-inline-count" part="count"
          >${label}</span
        >` : html`${label}${this.showCount && count > 0 ? html`<span class="ve-citation-inline-count" part="count"
              >${count}</span
            >` : nothing}`;
        return html`
      <ve-dropdown
        class="ve-citation-dropdown"
        trigger="hover"
        position="bottom-start"
        ?disabled=${this.disabled || 0 === count}
      >
        <button
          slot="trigger"
          type="button"
          class="ve-citation-inline-trigger"
          part="trigger"
          ?disabled=${this.disabled}
        >
          <slot name="trigger">${triggerContent}</slot>
        </button>
        <div class="ve-citation-popover" part="popover">
          <slot>${this.renderListContent()}</slot>
        </div>
      </ve-dropdown>
    `;
    }
    render() {
        return this.inline ? this.renderInline() : this.renderBlock();
    }
}
defineCustomElement(VE_CITATION_TAG, VeCitationElement);
export { VE_CITATION_TAG, VeCitationElement };
