// @ve-design/web/dist/ve-digital-input.js
import { css, html } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { addIcon } from "./icons/add.js";
import { minusIcon } from "./icons/minus.js";
const veDigitalInputTokenDefaults = css`
  :host {
    --ve-digital-input-width: 160px;
    --ve-digital-input-description-width: 240px;

    --ve-digital-input-height-small: var(--height-small);
    --ve-digital-input-height-default: var(--height-default);
    --ve-digital-input-height-large: var(--height-large);

    --ve-digital-input-radius: var(--radius-md);
    --ve-digital-input-border-width: var(--stroke-weight-base);

    --ve-digital-input-font-family: var(--font-en);
    --ve-digital-input-font-size: var(--text-body);
    --ve-digital-input-font-weight: var(--font-weight-normal);
    --ve-digital-input-line-height: 22px;

    --ve-digital-input-button-size-small: var(--height-small);
    --ve-digital-input-button-size-default: var(--height-default);
    --ve-digital-input-button-size-large: var(--height-large);
    --ve-digital-input-icon-size: var(--text-caption);

    --ve-digital-input-bg: var(--color-bg-base);
    --ve-digital-input-bg-hover: var(--color-bg-base);
    --ve-digital-input-border: var(--color-border-strong);
    --ve-digital-input-border-hover: var(--color-border-interactive);
    --ve-digital-input-border-active: var(--color-border-interactive);
    --ve-digital-input-text: var(--color-text-primary);
    --ve-digital-input-icon: var(--color-icon-secondary);

    --ve-digital-input-error-bg: var(--color-bg-danger-secondary);
    --ve-digital-input-error-bg-hover: color-mix(
      var(--ve-digital-input-error-bg) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-digital-input-error-border: var(--color-border-danger);
    --ve-digital-input-error-border-hover: var(--color-text-danger);
    --ve-digital-input-error-border-active: var(--color-text-danger);
    --ve-digital-input-error-description: var(--color-text-danger);
    --ve-digital-input-error-active-shadow:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        rgb(from var(--color-text-danger) r g b / 25%),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        rgb(from var(--color-text-danger) r g b / 10%);

    --ve-digital-input-disabled-bg: var(--color-bg-control);
    --ve-digital-input-disabled-border: var(--color-border-default);
    --ve-digital-input-disabled-text: var(--color-text-disable);

    --ve-digital-input-description-gap: var(--space-xxxs);
    --ve-digital-input-description-font-family: var(--font-cn);
    --ve-digital-input-description-font-size: var(--text-caption);
    --ve-digital-input-description-line-height: var(--line-height-body);
    --ve-digital-input-description-color: var(--color-text-tertiary);

    --ve-digital-input-transition-duration: var(--motion-fast);
    --ve-digital-input-transition-timing: var(--ease-standard);
    --ve-digital-input-active-shadow:
      0 0 var(--radius-xs) calc(var(--stroke-weight-strong) * -1)
        rgb(from var(--color-border-interactive-active) r g b / 25%),
      0 0 var(--space-l) calc(var(--stroke-weight-strong) * -1)
        rgb(from var(--color-border-interactive-active) r g b / 10%);

    --ve-digital-input-current-height: var(--ve-digital-input-height-default);
    --ve-digital-input-current-button-size: var(
      --ve-digital-input-button-size-default
    );
  }

  :host([size='small']) {
    --ve-digital-input-current-height: var(--ve-digital-input-height-small);
    --ve-digital-input-current-button-size: var(
      --ve-digital-input-button-size-small
    );
  }

  :host([size='large']) {
    --ve-digital-input-current-height: var(--ve-digital-input-height-large);
    --ve-digital-input-current-button-size: var(
      --ve-digital-input-button-size-large
    );
  }
`;
const token = veDigitalInputTokenDefaults;
const ve_digital_input = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
      box-sizing: border-box;
    }

    :host([hidden]) {
      display: none;
    }

    :host([disabled]) {
      cursor: not-allowed;
    }

    .ve-digital-input-root {
      display: inline-flex;
      flex-direction: column;
      align-items: flex-start;
      gap: var(--ve-digital-input-description-gap);
      inline-size: var(--ve-digital-input-width);
      max-inline-size: 100%;
      box-sizing: border-box;
    }

    .ve-digital-input-root-has-description {
      inline-size: var(--ve-digital-input-description-width);
    }

    .ve-digital-input-control {
      display: inline-flex;
      align-items: stretch;
      inline-size: var(--ve-digital-input-width);
      min-inline-size: 0;
      block-size: var(--ve-digital-input-current-height);
      box-sizing: border-box;
      color: var(--ve-digital-input-text);
      font-family: var(--ve-digital-input-font-family);
      font-size: var(--ve-digital-input-font-size);
      font-weight: var(--ve-digital-input-font-weight);
      line-height: var(--ve-digital-input-line-height);
    }

    .ve-digital-input-step,
    .ve-digital-input-field-wrap {
      border: var(--ve-digital-input-border-width) solid
        var(--ve-digital-input-border);
      background: var(--ve-digital-input-bg);
      transition:
        background-color var(--ve-digital-input-transition-duration)
          var(--ve-digital-input-transition-timing),
        border-color var(--ve-digital-input-transition-duration)
          var(--ve-digital-input-transition-timing),
        box-shadow var(--ve-digital-input-transition-duration)
          var(--ve-digital-input-transition-timing),
        color var(--ve-digital-input-transition-duration)
          var(--ve-digital-input-transition-timing),
        opacity var(--ve-digital-input-transition-duration)
          var(--ve-digital-input-transition-timing);
    }

    .ve-digital-input-step {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: 0 0 var(--ve-digital-input-current-button-size);
      inline-size: var(--ve-digital-input-current-button-size);
      block-size: var(--ve-digital-input-current-height);
      box-sizing: border-box;
      margin: 0;
      padding: 0;
      appearance: none;
      color: var(--ve-digital-input-icon);
      cursor: pointer;
      user-select: none;
    }

    .ve-digital-input-step-increase {
      border-start-start-radius: var(--ve-digital-input-radius);
      border-end-start-radius: var(--ve-digital-input-radius);
    }

    .ve-digital-input-step-decrease {
      border-start-end-radius: var(--ve-digital-input-radius);
      border-end-end-radius: var(--ve-digital-input-radius);
    }

    .ve-digital-input-step ve-icon {
      --ve-icon-size: var(--ve-digital-input-icon-size);

      display: inline-flex;
      flex: none;
    }

    .ve-digital-input-field-wrap {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: 1 1 auto;
      min-inline-size: 0;
      block-size: var(--ve-digital-input-current-height);
      box-sizing: border-box;
      margin-inline: calc(var(--ve-digital-input-border-width) * -1);
      padding-inline: var(--space-xxs);
    }

    .ve-digital-input-field {
      inline-size: 100%;
      min-inline-size: 0;
      border: none;
      outline: none;
      appearance: textfield;
      box-sizing: border-box;
      margin: 0;
      padding: 0;
      background: transparent;
      color: inherit;
      font: inherit;
      line-height: inherit;
      text-align: center;
      font-variant-numeric: tabular-nums;
      -webkit-tap-highlight-color: transparent;
    }

    .ve-digital-input-field::placeholder {
      color: var(--color-text-disable);
    }

    .ve-digital-input-field::-webkit-outer-spin-button,
    .ve-digital-input-field::-webkit-inner-spin-button {
      margin: 0;
      appearance: none;
    }

    .ve-digital-input-control:not(.ve-digital-input-control-disabled)
      .ve-digital-input-step:hover,
    .ve-digital-input-control:not(.ve-digital-input-control-disabled)
      .ve-digital-input-field-wrap:hover {
      border-color: var(--ve-digital-input-border-hover);
      background: var(--ve-digital-input-bg-hover);
      z-index: 1;
    }

    .ve-digital-input-control:not(.ve-digital-input-control-disabled)
      .ve-digital-input-field:focus {
      outline: none;
    }

    .ve-digital-input-control:not(.ve-digital-input-control-disabled)
      .ve-digital-input-field-wrap:focus-within {
      border-color: var(--ve-digital-input-border-active);
      background: var(--ve-digital-input-bg);
      box-shadow: var(--ve-digital-input-active-shadow);
      z-index: 2;
    }

    .ve-digital-input-control-error .ve-digital-input-field-wrap {
      background: var(--ve-digital-input-error-bg);
      border-color: var(--ve-digital-input-error-border);
    }

    .ve-digital-input-control-error:not(.ve-digital-input-control-disabled)
      .ve-digital-input-field-wrap:hover {
      background: var(--ve-digital-input-error-bg-hover);
      border-color: var(--ve-digital-input-error-border-hover);
    }

    .ve-digital-input-control-error:not(.ve-digital-input-control-disabled)
      .ve-digital-input-field-wrap:focus-within {
      border-color: var(--ve-digital-input-error-border-active);
      background: var(--ve-digital-input-bg);
      box-shadow: var(--ve-digital-input-error-active-shadow);
    }

    .ve-digital-input-control-disabled .ve-digital-input-step,
    .ve-digital-input-control-disabled .ve-digital-input-step:hover,
    .ve-digital-input-control-disabled .ve-digital-input-field-wrap,
    .ve-digital-input-control-disabled .ve-digital-input-field-wrap:hover,
    .ve-digital-input-control-disabled .ve-digital-input-field-wrap:focus-within {
      background: var(--ve-digital-input-disabled-bg);
      border-color: var(--ve-digital-input-disabled-border);
      color: var(--ve-digital-input-disabled-text);
      cursor: not-allowed;
      box-shadow: none;
    }

    .ve-digital-input-step:disabled {
      color: var(--ve-digital-input-disabled-text);
      cursor: not-allowed;
    }

    .ve-digital-input-field:disabled {
      color: var(--ve-digital-input-disabled-text);
      -webkit-text-fill-color: var(--ve-digital-input-disabled-text);
      cursor: not-allowed;
    }

    .ve-digital-input-control-disabled .ve-digital-input-field::placeholder {
      color: var(--ve-digital-input-disabled-text);
    }

    .ve-digital-input-description {
      inline-size: 100%;
      min-block-size: 18px;
      box-sizing: border-box;
      overflow: hidden;
      color: var(--ve-digital-input-description-color);
      font-family: var(--ve-digital-input-description-font-family);
      font-size: var(--ve-digital-input-description-font-size);
      line-height: var(--ve-digital-input-description-line-height);
      white-space: nowrap;
      text-overflow: ellipsis;
    }

    .ve-digital-input-description-error {
      color: var(--ve-digital-input-error-description);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-digital-input-step,
      .ve-digital-input-field-wrap {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_min, _init_extra__init_min, _init_max, _init_extra__init_max, _init_step, _init_extra__init_step, _init_precision, _init_extra__init_precision, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_placeholder, _init_extra__init_placeholder, _init_description, _init_extra__init_description, _init_controlElement, _init_extra__init_controlElement;
const VE_DIGITAL_INPUT_TAG = 've-digital-input';
const INCREASE_LABEL = 'Increase';
const DECREASE_LABEL = 'Decrease';
const INPUT_LABEL = 'Value';
function isFiniteNumber(value) {
    return 'number' == typeof value && Number.isFinite(value);
}
function normalizePrecision(value) {
    if (!isFiniteNumber(value) || value < 0) return;
    return Math.floor(value);
}
function toValidStep(value) {
    return isFiniteNumber(value) && value > 0 ? value : 1;
}
function roundToPrecision(value, precision) {
    if (void 0 === precision) return value;
    const factor = 10 ** precision;
    return Math.round((value + Number.EPSILON) * factor) / factor;
}
_dec = property({
    type: Number
}), _dec1 = property({
    type: Number,
    attribute: 'default-value'
}), _dec2 = property({
    type: Number
}), _dec3 = property({
    type: Number
}), _dec4 = property({
    type: Number
}), _dec5 = property({
    type: Number
}), _dec6 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec8 = property({
    type: Boolean,
    reflect: true
}), _dec9 = property({
    type: String
}), _dec10 = property({
    type: String
}), _dec11 = query('.ve-digital-input-field');
class VeDigitalInputElement extends VeElement {
    static{
        ({ e: [_init_value, _init_extra__init_value, _init_defaultValue, _init_extra__init_defaultValue, _init_min, _init_extra__init_min, _init_max, _init_extra__init_max, _init_step, _init_extra__init_step, _init_precision, _init_extra__init_precision, _init_size, _init_extra__init_size, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_placeholder, _init_extra__init_placeholder, _init_description, _init_extra__init_description, _init_controlElement, _init_extra__init_controlElement] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "value"
            ],
            [
                _dec1,
                1,
                "defaultValue"
            ],
            [
                _dec2,
                1,
                "min"
            ],
            [
                _dec3,
                1,
                "max"
            ],
            [
                _dec4,
                1,
                "step"
            ],
            [
                _dec5,
                1,
                "precision"
            ],
            [
                _dec6,
                1,
                "size"
            ],
            [
                _dec7,
                1,
                "status"
            ],
            [
                _dec8,
                1,
                "disabled"
            ],
            [
                _dec9,
                1,
                "placeholder"
            ],
            [
                _dec10,
                1,
                "description"
            ],
            [
                _dec11,
                1,
                "controlElement"
            ]
        ]));
    }
    static styles = ve_digital_input;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_value_1 = (()=>{
        const _value = _init_value(this, 0);
        _init_extra__init_value(this);
        return _value;
    })();
    get value() {
        return this.#___private_value_1;
    }
    set value(_v) {
        this.#___private_value_1 = _v;
    }
    #___private_defaultValue_2 = (()=>{
        const _value = _init_defaultValue(this, 0);
        _init_extra__init_defaultValue(this);
        return _value;
    })();
    get defaultValue() {
        return this.#___private_defaultValue_2;
    }
    set defaultValue(_v) {
        this.#___private_defaultValue_2 = _v;
    }
    #___private_min_3 = (()=>{
        const _value = _init_min(this, void 0);
        _init_extra__init_min(this);
        return _value;
    })();
    get min() {
        return this.#___private_min_3;
    }
    set min(_v) {
        this.#___private_min_3 = _v;
    }
    #___private_max_4 = (()=>{
        const _value = _init_max(this, void 0);
        _init_extra__init_max(this);
        return _value;
    })();
    get max() {
        return this.#___private_max_4;
    }
    set max(_v) {
        this.#___private_max_4 = _v;
    }
    #___private_step_5 = (()=>{
        const _value = _init_step(this, 1);
        _init_extra__init_step(this);
        return _value;
    })();
    get step() {
        return this.#___private_step_5;
    }
    set step(_v) {
        this.#___private_step_5 = _v;
    }
    #___private_precision_6 = (()=>{
        const _value = _init_precision(this, void 0);
        _init_extra__init_precision(this);
        return _value;
    })();
    get precision() {
        return this.#___private_precision_6;
    }
    set precision(_v) {
        this.#___private_precision_6 = _v;
    }
    #___private_size_7 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_7;
    }
    set size(_v) {
        this.#___private_size_7 = _v;
    }
    #___private_status_8 = (()=>{
        const _value = _init_status(this, 'default');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_8;
    }
    set status(_v) {
        this.#___private_status_8 = _v;
    }
    #___private_disabled_9 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_9;
    }
    set disabled(_v) {
        this.#___private_disabled_9 = _v;
    }
    #___private_placeholder_10 = (()=>{
        const _value = _init_placeholder(this, '');
        _init_extra__init_placeholder(this);
        return _value;
    })();
    get placeholder() {
        return this.#___private_placeholder_10;
    }
    set placeholder(_v) {
        this.#___private_placeholder_10 = _v;
    }
    #___private_description_11 = (()=>{
        const _value = _init_description(this, '');
        _init_extra__init_description(this);
        return _value;
    })();
    get description() {
        return this.#___private_description_11;
    }
    set description(_v) {
        this.#___private_description_11 = _v;
    }
    #___private_controlElement_12 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_12;
    }
    set controlElement(_v) {
        this.#___private_controlElement_12 = _v;
    }
    initializedDefaultValue = false;
    lastCommittedValue = 0;
    pendingInternalValueChange = false;
    connectedCallback() {
        super.connectedCallback();
        if (!this.initializedDefaultValue) {
            this.initializedDefaultValue = true;
            if (!this.hasAttribute('value') && this.hasAttribute('default-value')) this.value = this.normalizeValue(this.defaultValue);
            else this.value = this.normalizeValue(this.value);
            this.lastCommittedValue = this.value;
        }
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('value') || changedProperties.has('min') || changedProperties.has('max') || changedProperties.has('precision')) {
            const normalizedValue = this.normalizeValue(this.value);
            if (!Object.is(this.value, normalizedValue)) this.value = normalizedValue;
        }
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('value')) {
            if (!this.pendingInternalValueChange) this.lastCommittedValue = this.value;
            this.pendingInternalValueChange = false;
        }
    }
    addEventListener(type, listener, options) {
        super.addEventListener(type, listener, options);
    }
    removeEventListener(type, listener, options) {
        super.removeEventListener(type, listener, options);
    }
    focus(options) {
        this.controlElement?.focus(options);
    }
    blur() {
        this.controlElement?.blur();
    }
    get normalizedSize() {
        return 'small' === this.size || 'large' === this.size ? this.size : 'default';
    }
    get normalizedStatus() {
        return 'error' === this.status ? 'error' : 'default';
    }
    get normalizedStep() {
        return toValidStep(this.step);
    }
    get normalizedPrecision() {
        return normalizePrecision(this.precision);
    }
    get hasMin() {
        return isFiniteNumber(this.min);
    }
    get hasMax() {
        return isFiniteNumber(this.max);
    }
    get canIncrease() {
        return !this.disabled && (!this.hasMax || this.value < this.max);
    }
    get canDecrease() {
        return !this.disabled && (!this.hasMin || this.value > this.min);
    }
    get displayValue() {
        const precision = this.normalizedPrecision;
        if (void 0 === precision) return String(this.value);
        return this.value.toFixed(precision);
    }
    normalizeValue(value) {
        let nextValue = isFiniteNumber(value) ? value : 0;
        if (this.hasMin) nextValue = Math.max(nextValue, this.min);
        if (this.hasMax) nextValue = Math.min(nextValue, this.max);
        return roundToPrecision(nextValue, this.normalizedPrecision);
    }
    parseInputValue(value) {
        const trimmed = value.trim();
        if (!trimmed) return this.normalizeValue(0);
        const parsed = Number(trimmed);
        return this.normalizeValue(parsed);
    }
    setValueFromInteraction(nextValue, source, commit) {
        const oldValue = commit ? this.value : this.lastCommittedValue;
        const previousValue = this.value;
        const normalizedValue = this.normalizeValue(nextValue);
        if (Object.is(previousValue, normalizedValue) && !commit) return;
        if (!Object.is(previousValue, normalizedValue)) {
            this.pendingInternalValueChange = true;
            this.value = normalizedValue;
            this.emitValueEvent('ve-input', {
                value: normalizedValue,
                oldValue: previousValue,
                source
            });
        }
        if (commit && !Object.is(this.lastCommittedValue, normalizedValue)) {
            this.emitValueEvent('ve-change', {
                value: normalizedValue,
                oldValue,
                source
            });
            this.lastCommittedValue = normalizedValue;
        }
    }
    commitValue(source) {
        if (Object.is(this.lastCommittedValue, this.value)) return;
        this.emitValueEvent('ve-change', {
            value: this.value,
            oldValue: this.lastCommittedValue,
            source
        });
        this.lastCommittedValue = this.value;
    }
    emitValueEvent(type, detail) {
        this.dispatchEvent(new CustomEvent(type, {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    handleInput(event) {
        const input = event.currentTarget;
        this.setValueFromInteraction(this.parseInputValue(input.value), 'input', false);
    }
    handleChange() {
        this.commitValue('input');
    }
    handleBlur() {
        this.commitValue('blur');
    }
    handleKeydown(event) {
        if ('Enter' !== event.key) return;
        this.commitValue('enter');
    }
    handleIncrease() {
        if (!this.canIncrease) return;
        this.setValueFromInteraction(this.value + this.normalizedStep, 'step', true);
    }
    handleDecrease() {
        if (!this.canDecrease) return;
        this.setValueFromInteraction(this.value - this.normalizedStep, 'step', true);
    }
    renderStepButton(direction) {
        const isIncrease = 'increase' === direction;
        const disabled = isIncrease ? !this.canIncrease : !this.canDecrease;
        return html`
      <button
        class=${classMap({
            've-digital-input-step': true,
            've-digital-input-step-increase': isIncrease,
            've-digital-input-step-decrease': !isIncrease
        })}
        part=${isIncrease ? 'increase' : 'decrease'}
        type="button"
        ?disabled=${disabled}
        aria-label=${isIncrease ? INCREASE_LABEL : DECREASE_LABEL}
        @click=${isIncrease ? this.handleIncrease : this.handleDecrease}
      >
        <ve-icon .definition=${isIncrease ? addIcon : minusIcon}></ve-icon>
      </button>
    `;
    }
    render() {
        const hasDescription = (this.description ?? '').trim().length > 0;
        return html`
      <div
        class=${classMap({
            've-digital-input-root': true,
            "ve-digital-input-root-has-description": hasDescription
        })}
      >
        <div
          class=${classMap({
            've-digital-input-control': true,
            [`ve-digital-input-control-${this.normalizedSize}`]: true,
            've-digital-input-control-error': 'error' === this.normalizedStatus,
            've-digital-input-control-disabled': this.disabled
        })}
          part="control"
        >
          ${this.renderStepButton('increase')}
          <div class="ve-digital-input-field-wrap">
            <input
              class="ve-digital-input-field"
              part="input"
              type="text"
              inputmode="decimal"
              role="spinbutton"
              ?disabled=${this.disabled}
              .value=${this.displayValue}
              placeholder=${ifDefined(this.placeholder || void 0)}
              aria-label=${INPUT_LABEL}
              aria-disabled=${ifDefined(this.disabled ? 'true' : void 0)}
              aria-invalid=${ifDefined('error' === this.normalizedStatus ? 'true' : void 0)}
              aria-valuemin=${ifDefined(this.hasMin ? this.min : void 0)}
              aria-valuemax=${ifDefined(this.hasMax ? this.max : void 0)}
              aria-valuenow=${this.value}
              @input=${this.handleInput}
              @change=${this.handleChange}
              @blur=${this.handleBlur}
              @keydown=${this.handleKeydown}
            />
          </div>
          ${this.renderStepButton('decrease')}
        </div>
        ${hasDescription ? html`
              <div
                class=${classMap({
            "ve-digital-input-description": true,
            "ve-digital-input-description-error": 'error' === this.normalizedStatus
        })}
                part="description"
              >
                ${this.description}
              </div>
            ` : null}
      </div>
    `;
    }
}
defineCustomElement(VE_DIGITAL_INPUT_TAG, VeDigitalInputElement);
export { VE_DIGITAL_INPUT_TAG, VeDigitalInputElement };
