// @ve-design/web/dist/ve-link.js
import { css, html } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
const veLinkTokenDefaults = css`
  :host {
    --ve-link-font-family: var(--font-sans);
    --ve-link-font-size: var(--text-body);
    --ve-link-font-weight: var(--font-weight-normal);
    --ve-link-line-height: 1;

    --ve-link-padding-block: var(--space-xxxs);
    --ve-link-padding-inline: var(--space-xxxs);
    --ve-link-radius: var(--radius-xs);
    --ve-link-underline-thickness: var(--stroke-weight-strong);
    --ve-link-underline-offset: var(--space-xxxs);

    --ve-link-text-default: var(--color-blue-600, #1664FF);
    --ve-link-text-default-hover: var(--color-blue-500, #2F79FF);
    --ve-link-text-default-active: var(--color-blue-700, #0D4BDB);
    --ve-link-text-default-disabled: var(--color-blue-400, #5D9BFF);
    --ve-link-text-primary: var(--color-text-accent);
    --ve-link-text-success: var(--color-text-success);
    --ve-link-text-warning: var(--color-text-warning);
    --ve-link-text-error: var(--color-text-danger);

    --ve-link-disabled-opacity: var(--state-disabled-ratio);

    --ve-link-current-text: var(--ve-link-text-default);
    --ve-link-current-text-hover: var(--ve-link-text-default-hover);
    --ve-link-current-text-active: var(--ve-link-text-default-active);
    --ve-link-current-text-disabled: var(--ve-link-text-default-disabled);
    --ve-link-current-disabled-opacity: 1;
  }

  :host([status='primary']) {
    --ve-link-current-text: var(--ve-link-text-primary);
    --ve-link-current-text-hover: var(--ve-link-text-primary);
    --ve-link-current-text-active: var(--ve-link-text-primary);
    --ve-link-current-text-disabled: var(--ve-link-text-primary);
    --ve-link-current-disabled-opacity: var(--ve-link-disabled-opacity);
  }

  :host([status='success']) {
    --ve-link-current-text: var(--ve-link-text-success);
    --ve-link-current-text-hover: var(--ve-link-text-success);
    --ve-link-current-text-active: var(--ve-link-text-success);
    --ve-link-current-text-disabled: var(--ve-link-text-success);
    --ve-link-current-disabled-opacity: var(--ve-link-disabled-opacity);
  }

  :host([status='warning']) {
    --ve-link-current-text: var(--ve-link-text-warning);
    --ve-link-current-text-hover: var(--ve-link-text-warning);
    --ve-link-current-text-active: var(--ve-link-text-warning);
    --ve-link-current-text-disabled: var(--ve-link-text-warning);
    --ve-link-current-disabled-opacity: var(--ve-link-disabled-opacity);
  }

  :host([status='error']) {
    --ve-link-current-text: var(--ve-link-text-error);
    --ve-link-current-text-hover: var(--ve-link-text-error);
    --ve-link-current-text-active: var(--ve-link-text-error);
    --ve-link-current-text-disabled: var(--ve-link-text-error);
    --ve-link-current-disabled-opacity: var(--ve-link-disabled-opacity);
  }
`;
const token = veLinkTokenDefaults;
const ve_link = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    :host([disabled]) {
      cursor: not-allowed;
    }

    .ve-link {
      position: relative;
      display: inline-flex;
      align-items: center;
      box-sizing: border-box;
      margin: 0;
      padding-block: var(--ve-link-padding-block);
      padding-inline: var(--ve-link-padding-inline);
      border: none;
      border-radius: var(--ve-link-radius);
      background: transparent;
      color: var(--ve-link-current-text);
      font-family: var(--ve-link-font-family);
      font-size: var(--ve-link-font-size);
      font-weight: var(--ve-link-font-weight);
      line-height: var(--ve-link-line-height);
      text-decoration: none;
      cursor: default;
      transition:
        color 120ms ease,
        opacity 120ms ease;
    }

    a.ve-link {
      cursor: pointer;
    }

    a.ve-link:not([href]) {
      cursor: default;
    }

    .ve-link:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--stroke-2) color-mix(currentColor 24%, transparent);
    }

    :host(:not([disabled])) a.ve-link[href]:hover {
      color: var(--ve-link-current-text-hover);
    }

    :host(:not([disabled])) a.ve-link[href]:active {
      color: var(--ve-link-current-text-active);
    }

    :host([underline]) .ve-link {
      text-decoration-line: underline;
      text-decoration-style: dotted;
      text-decoration-color: currentColor;
      text-decoration-thickness: var(--ve-link-underline-thickness);
      text-underline-offset: var(--ve-link-underline-offset);
      text-decoration-skip-ink: none;
    }

    :host([disabled]) .ve-link {
      background: none;
      color: var(--ve-link-current-text-disabled);
      cursor: not-allowed;
      opacity: var(--ve-link-current-disabled-opacity);
    }

    .ve-link-content {
      display: inline-flex;
      align-items: center;
      line-height: inherit;
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-link {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _init_href, _init_extra__init_href, _init_target, _init_extra__init_target, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_underline, _init_extra__init_underline, _init_controlElement, _init_extra__init_controlElement;
const VE_LINK_TAG = 've-link';
const CONTROL_ACCESSIBILITY_ATTRIBUTES = [
    'aria-label',
    'aria-labelledby',
    'aria-describedby'
];
_dec = property({
    type: String
}), _dec1 = property({
    type: String
}), _dec2 = property({
    type: String,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = query('.ve-link');
class VeLinkElement extends VeElement {
    static{
        ({ e: [_init_href, _init_extra__init_href, _init_target, _init_extra__init_target, _init_status, _init_extra__init_status, _init_disabled, _init_extra__init_disabled, _init_underline, _init_extra__init_underline, _init_controlElement, _init_extra__init_controlElement] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "href"
            ],
            [
                _dec1,
                1,
                "target"
            ],
            [
                _dec2,
                1,
                "status"
            ],
            [
                _dec3,
                1,
                "disabled"
            ],
            [
                _dec4,
                1,
                "underline"
            ],
            [
                _dec5,
                1,
                "controlElement"
            ]
        ]));
    }
    static styles = ve_link;
    static get observedAttributes() {
        return [
            ...new Set([
                ...super.observedAttributes,
                ...CONTROL_ACCESSIBILITY_ATTRIBUTES
            ])
        ];
    }
    #___private_href_1 = (()=>{
        const _value = _init_href(this, '');
        _init_extra__init_href(this);
        return _value;
    })();
    get href() {
        return this.#___private_href_1;
    }
    set href(_v) {
        this.#___private_href_1 = _v;
    }
    #___private_target_2 = (()=>{
        const _value = _init_target(this, '_self');
        _init_extra__init_target(this);
        return _value;
    })();
    get target() {
        return this.#___private_target_2;
    }
    set target(_v) {
        this.#___private_target_2 = _v;
    }
    #___private_status_3 = (()=>{
        const _value = _init_status(this, void 0);
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_3;
    }
    set status(_v) {
        this.#___private_status_3 = _v;
    }
    #___private_disabled_4 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_4;
    }
    set disabled(_v) {
        this.#___private_disabled_4 = _v;
    }
    #___private_underline_5 = (()=>{
        const _value = _init_underline(this, false);
        _init_extra__init_underline(this);
        return _value;
    })();
    get underline() {
        return this.#___private_underline_5;
    }
    set underline(_v) {
        this.#___private_underline_5 = _v;
    }
    #___private_controlElement_6 = (()=>{
        const _value = _init_controlElement(this, null);
        _init_extra__init_controlElement(this);
        return _value;
    })();
    get controlElement() {
        return this.#___private_controlElement_6;
    }
    set controlElement(_v) {
        this.#___private_controlElement_6 = _v;
    }
    get controlAriaLabel() {
        return this.getAttribute('aria-label') ?? void 0;
    }
    get controlAriaLabelledby() {
        return this.getAttribute('aria-labelledby') ?? void 0;
    }
    get controlAriaDescribedby() {
        return this.getAttribute('aria-describedby') ?? void 0;
    }
    focus(options) {
        this.controlElement?.focus(options);
    }
    attributeChangedCallback(name, oldValue, newValue) {
        super.attributeChangedCallback(name, oldValue, newValue);
        if (oldValue !== newValue && CONTROL_ACCESSIBILITY_ATTRIBUTES.includes(name)) this.requestUpdate();
    }
    handleClick(event) {
        if (!this.disabled) return;
        event.preventDefault();
        event.stopPropagation();
    }
    renderContent() {
        return html`
      <span class="ve-link-content" part="content">
        <slot></slot>
      </span>
    `;
    }
    renderAnchor() {
        return html`
      <a
        class="ve-link"
        part="control"
        href=${ifDefined(this.disabled || 0 === (this.href ?? '').trim().length ? void 0 : this.href)}
        target=${ifDefined(this.disabled ? void 0 : this.target)}
        rel=${ifDefined(!this.disabled && '_blank' === this.target ? 'noreferrer noopener' : void 0)}
        aria-label=${ifDefined(this.controlAriaLabel)}
        aria-labelledby=${ifDefined(this.controlAriaLabelledby)}
        aria-describedby=${ifDefined(this.controlAriaDescribedby)}
        aria-disabled=${ifDefined(this.disabled ? 'true' : void 0)}
        tabindex=${ifDefined(this.disabled ? '-1' : void 0)}
        @click=${this.handleClick}
      >
        ${this.renderContent()}
      </a>
    `;
    }
    render() {
        return this.renderAnchor();
    }
}
defineCustomElement(VE_LINK_TAG, VeLinkElement);
export { VE_LINK_TAG, VeLinkElement };
