// @ve-design/web/dist/ve-modal.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state as decorators_js_state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeButtonElement, VE_BUTTON_TAG } from "./68533.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
const veModalTokenDefaults = css`
  :host {
    --ve-modal-z-index: 1000;
    --ve-modal-width: 480px;
    --ve-modal-max-width: calc(100vw - var(--space-xl));

    --ve-modal-overlay-background: rgba(16, 16, 19, 0.45);
    --ve-modal-panel-background: var(--color-bg-overlay);
    --ve-modal-panel-border-width: var(--stroke-weight-base);
    --ve-modal-panel-border-color: var(--color-border-default);
    --ve-modal-panel-radius: var(--radius-xl);
    --ve-modal-panel-shadow: var(--shadow-lg);

    --ve-modal-header-padding-block: var(--space-m);
    --ve-modal-header-padding-inline-start: var(--space-l);
    --ve-modal-header-padding-inline-end: 18px;
    --ve-modal-title-gap: var(--space-xxs);
    --ve-modal-title-color: var(--color-text-primary);
    --ve-modal-title-font-family: var(--font-sans);
    --ve-modal-title-font-size: var(--text-subsection-title);
    --ve-modal-title-font-weight: var(--font-weight-semibold);
    --ve-modal-title-line-height: var(--line-height-title);

    --ve-modal-close-button-size: 28px;
    --ve-modal-close-icon-size: var(--text-ui-strong);
    --ve-modal-close-button-radius: var(--radius-xs);
    --ve-modal-close-color: var(--color-text-primary);
    --ve-modal-close-color-hover: var(--color-icon-primary);
    --ve-modal-close-background-hover: color-mix(
      var(--color-bg-secondary) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-modal-close-background-pressed: color-mix(
      var(--color-bg-secondary) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );

    --ve-modal-body-padding-inline: var(--space-l);
    --ve-modal-body-color: var(--color-text-primary);
    --ve-modal-body-font-family: var(--font-sans);
    --ve-modal-body-font-size: var(--text-body);
    --ve-modal-body-line-height: var(--line-height-body);

    --ve-modal-footer-padding-block: var(--space-m);
    --ve-modal-footer-padding-inline: var(--space-l);
    --ve-modal-footer-gap: var(--space-xs);

    --ve-modal-button-height: var(--height-default);
    --ve-modal-button-min-width: 60px;
    --ve-modal-button-padding-inline: var(--space-xs);
    --ve-modal-button-radius: var(--radius-md);
    --ve-modal-button-font-family: var(--font-sans);
    --ve-modal-button-font-size: var(--text-body);
    --ve-modal-button-font-weight: var(--font-weight-normal);
    --ve-modal-button-line-height: 22px;
    --ve-modal-button-disabled-opacity: var(--state-disabled-ratio);

    --ve-modal-cancel-background: var(--color-bg-secondary);
    --ve-modal-cancel-color: var(--color-text-primary);
    --ve-modal-cancel-background-hover: color-mix(
      var(--ve-modal-cancel-background) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-modal-cancel-background-pressed: color-mix(
      var(--ve-modal-cancel-background) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );

    --ve-modal-ok-background: var(--color-bg-primary);
    --ve-modal-ok-color: var(--color-text-foreground);
    --ve-modal-ok-background-hover: color-mix(
      var(--ve-modal-ok-background) var(--state-hover-solid-ratio),
      var(--state-hover-solid-mix)
    );
    --ve-modal-ok-background-pressed: color-mix(
      var(--ve-modal-ok-background) var(--state-pressed-solid-ratio),
      var(--state-pressed-solid-mix)
    );

    --ve-modal-motion-duration: var(--motion-base);
    --ve-modal-motion-easing: var(--ease-standard);
  }
`;
const token = veModalTokenDefaults;
const ve_modal = [
    componentReset,
    token,
    css`
    :host {
      display: contents;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-modal-overlay {
      position: fixed;
      inset: 0;
      z-index: var(--ve-modal-current-z-index, var(--ve-modal-z-index));
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: center;
      padding: var(--space-xl) var(--space-s);
      background: var(--ve-modal-overlay-background);
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
      overscroll-behavior: contain;
      transition:
        opacity var(--ve-modal-motion-duration) var(--ve-modal-motion-easing),
        visibility 0s linear var(--ve-modal-motion-duration);
    }

    .ve-modal-overlay-open {
      opacity: 1;
      visibility: visible;
      pointer-events: auto;
      transition-delay: 0s;
    }

    .ve-modal-overlay-no-mask {
      background: transparent;
    }

    .ve-modal-panel {
      box-sizing: border-box;
      display: flex;
      flex-direction: column;
      width: var(--ve-modal-current-width, var(--ve-modal-width));
      max-width: var(--ve-modal-max-width);
      max-height: calc(100vh - var(--space-xl) * 2);
      overflow: hidden;
      border: var(--ve-modal-panel-border-width) solid
        var(--ve-modal-panel-border-color);
      border-radius: var(--ve-modal-panel-radius);
      background: var(--ve-modal-panel-background);
      box-shadow: var(--ve-modal-panel-shadow);
      color: var(--ve-modal-body-color);
      font-family: var(--ve-modal-body-font-family);
      transform: translate3d(0, 8px, 0) scale(0.98);
      transition: transform var(--ve-modal-motion-duration)
        var(--ve-modal-motion-easing);
    }

    .ve-modal-overlay-open .ve-modal-panel {
      transform: translate3d(0, 0, 0) scale(1);
    }

    .ve-modal-header {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: space-between;
      gap: var(--space-s);
      width: 100%;
      padding-block: var(--ve-modal-header-padding-block);
      padding-inline: var(--ve-modal-header-padding-inline-start)
        var(--ve-modal-header-padding-inline-end);
    }

    .ve-modal-title {
      display: inline-flex;
      align-items: center;
      min-width: 0;
      gap: var(--ve-modal-title-gap);
      color: var(--ve-modal-title-color);
      font-family: var(--ve-modal-title-font-family);
      font-size: var(--ve-modal-title-font-size);
      font-weight: var(--ve-modal-title-font-weight);
      line-height: var(--ve-modal-title-line-height);
      overflow-wrap: anywhere;
    }

    .ve-modal-close-button {
      --ve-button-height-small: var(--ve-modal-close-button-size);
      --ve-button-padding-inline-small: 0px;
      --ve-button-radius-default: var(--ve-modal-close-button-radius);
      --ve-button-background-secondary: transparent;
      --ve-button-text-secondary: var(--ve-modal-close-color);

      flex: none;
      inline-size: var(--ve-modal-close-button-size);
      color: var(--ve-modal-close-color);
    }

    .ve-modal-close-button:hover {
      --ve-button-background-secondary: var(--ve-modal-close-background-hover);
      --ve-button-text-secondary: var(--ve-modal-close-color-hover);
    }

    .ve-modal-close-button:active {
      --ve-button-background-secondary: var(--ve-modal-close-background-pressed);
    }

    .ve-modal-action:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--stroke-weight-bold) var(--color-border-interactive);
    }

    .ve-modal-close-button svg,
    .ve-modal-close-button ::slotted(*) {
      width: var(--ve-modal-close-icon-size);
      height: var(--ve-modal-close-icon-size);
      flex: none;
    }

    .ve-modal-body {
      box-sizing: border-box;
      flex: 1 1 auto;
      min-height: 0;
      overflow: auto;
      overscroll-behavior: contain;
      padding-inline: var(--ve-modal-body-padding-inline);
      color: var(--ve-modal-body-color);
      font-size: var(--ve-modal-body-font-size);
      line-height: var(--ve-modal-body-line-height);
    }

    .ve-modal-footer {
      box-sizing: border-box;
      display: flex;
      align-items: center;
      justify-content: flex-end;
      gap: var(--ve-modal-footer-gap);
      width: 100%;
      padding-block: var(--ve-modal-footer-padding-block);
      padding-inline: var(--ve-modal-footer-padding-inline);
    }

    .ve-modal-footer-slot {
      display: contents;
    }

    .ve-modal-action {
      appearance: none;
      box-sizing: border-box;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      min-width: var(--ve-modal-button-min-width);
      height: var(--ve-modal-button-height);
      padding-block: 0;
      padding-inline: var(--ve-modal-button-padding-inline);
      border: 0;
      border-radius: var(--ve-modal-button-radius);
      font-family: var(--ve-modal-button-font-family);
      font-size: var(--ve-modal-button-font-size);
      font-weight: var(--ve-modal-button-font-weight);
      line-height: var(--ve-modal-button-line-height);
      white-space: nowrap;
      cursor: pointer;
      transition:
        background-color var(--ve-modal-motion-duration) var(--ease-standard),
        color var(--ve-modal-motion-duration) var(--ease-standard),
        opacity var(--ve-modal-motion-duration) var(--ease-standard);
    }

    .ve-modal-action:disabled {
      opacity: var(--ve-modal-button-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-modal-action-cancel {
      background: var(--ve-modal-cancel-background);
      color: var(--ve-modal-cancel-color);
    }

    .ve-modal-action-cancel:hover:not(:disabled) {
      background: var(--ve-modal-cancel-background-hover);
    }

    .ve-modal-action-cancel:active:not(:disabled) {
      background: var(--ve-modal-cancel-background-pressed);
    }

    .ve-modal-action-ok {
      background: var(--ve-modal-ok-background);
      color: var(--ve-modal-ok-color);
    }

    .ve-modal-action-ok:hover:not(:disabled) {
      background: var(--ve-modal-ok-background-hover);
    }

    .ve-modal-action-ok:active:not(:disabled) {
      background: var(--ve-modal-ok-background-pressed);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-modal-overlay,
      .ve-modal-panel,
      .ve-modal-action,
      .ve-modal-close-button {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _init_visible, _init_extra__init_visible, _init_width, _init_extra__init_width, _init_closable, _init_extra__init_closable, _init_mask, _init_extra__init_mask, _init_maskClosable, _init_extra__init_maskClosable, _init_okText, _init_extra__init_okText, _init_cancelText, _init_extra__init_cancelText, _init_okDisabled, _init_extra__init_okDisabled, _init_cancelDisabled, _init_extra__init_cancelDisabled, _init_hideFooter, _init_extra__init_hideFooter, _init_zIndex, _init_extra__init_zIndex, _init_titleSlot, _init_extra__init_titleSlot, _init_footerSlot, _init_extra__init_footerSlot, _init_closeIconSlot, _init_extra__init_closeIconSlot, _init_overlayElement, _init_extra__init_overlayElement, _init_panelElement, _init_extra__init_panelElement, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasFooterSlot, _init_extra__init_hasFooterSlot, _init_hasCloseIconSlot, _init_extra__init_hasCloseIconSlot;
const VE_MODAL_TAG = 've-modal';
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
function normalizeWidth(value) {
    const trimmed = value.trim();
    return trimmed.length > 0 ? trimmed : '480px';
}
function shouldReduceMotion() {
    return "u" > typeof window && 'function' == typeof window.matchMedia && window.matchMedia('(prefers-reduced-motion: reduce)').matches;
}
function parseTimeValue(value) {
    const trimmed = value.trim();
    if (trimmed.endsWith('ms')) return Number.parseFloat(trimmed);
    if (trimmed.endsWith('s')) return 1000 * Number.parseFloat(trimmed);
    return Number.parseFloat(trimmed) || 0;
}
function parseTimeList(value) {
    return value.split(',').map(parseTimeValue);
}
function getLongestTransitionMs(element) {
    if ("u" < typeof window) return 0;
    const styles = window.getComputedStyle(element);
    const durations = parseTimeList(styles.transitionDuration);
    const delays = parseTimeList(styles.transitionDelay);
    return durations.reduce((longest, duration, index)=>{
        const delay = delays[index] ?? delays[0] ?? 0;
        return Math.max(longest, duration + delay);
    }, 0);
}
const SCROLLABLE_OVERFLOW_PATTERN = /(auto|scroll|overlay)/;
const scrollLockStates = new Map();
function canElementScroll(element) {
    return element.scrollHeight > element.clientHeight || element.scrollWidth > element.clientWidth;
}
function isScrollableElement(element) {
    if (scrollLockStates.has(element)) return true;
    if (!canElementScroll(element)) return false;
    const styles = window.getComputedStyle(element);
    return SCROLLABLE_OVERFLOW_PATTERN.test(styles.overflow) || SCROLLABLE_OVERFLOW_PATTERN.test(styles.overflowX) || SCROLLABLE_OVERFLOW_PATTERN.test(styles.overflowY);
}
function getScrollLockTarget(element) {
    let parent = element.parentElement;
    while(parent && parent !== document.body && parent !== document.documentElement){
        if (isScrollableElement(parent)) return parent;
        parent = parent.parentElement;
    }
    return document.body;
}
function canConsumeScroll(element, deltaX, deltaY) {
    const styles = window.getComputedStyle(element);
    const scrollableY = SCROLLABLE_OVERFLOW_PATTERN.test(styles.overflowY) && element.scrollHeight > element.clientHeight;
    const scrollableX = SCROLLABLE_OVERFLOW_PATTERN.test(styles.overflowX) && element.scrollWidth > element.clientWidth;
    if (0 !== deltaY && scrollableY) {
        const maxScrollTop = element.scrollHeight - element.clientHeight;
        if (deltaY < 0 && element.scrollTop > 0) return true;
        if (deltaY > 0 && element.scrollTop < maxScrollTop) return true;
    }
    if (0 !== deltaX && scrollableX) {
        const maxScrollLeft = element.scrollWidth - element.clientWidth;
        if (deltaX < 0 && element.scrollLeft > 0) return true;
        if (deltaX > 0 && element.scrollLeft < maxScrollLeft) return true;
    }
    return false;
}
function getScrollbarWidth(element) {
    if (element === document.body) return window.innerWidth - document.documentElement.clientWidth;
    return element.offsetWidth - element.clientWidth;
}
function lockScrollTarget(target) {
    const existingState = scrollLockStates.get(target);
    if (existingState) {
        existingState.count += 1;
        return;
    }
    const { style } = target;
    const state = {
        count: 1,
        overflow: style.overflow,
        paddingRight: style.paddingRight
    };
    const scrollbarWidth = getScrollbarWidth(target);
    scrollLockStates.set(target, state);
    style.overflow = 'hidden';
    if (scrollbarWidth <= 0) return;
    const currentPaddingRight = parseFloat(window.getComputedStyle(target).paddingRight) || 0;
    style.paddingRight = `${currentPaddingRight + scrollbarWidth}px`;
}
function unlockScrollTarget(target) {
    const state = scrollLockStates.get(target);
    if (!state) return;
    state.count -= 1;
    if (state.count > 0) return;
    target.style.overflow = state.overflow;
    target.style.paddingRight = state.paddingRight;
    scrollLockStates.delete(target);
}
_dec = property({
    type: Boolean,
    reflect: true
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec3 = property({
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec4 = property({
    attribute: 'mask-closable',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec5 = property({
    type: String,
    attribute: 'ok-text',
    reflect: true,
    useDefault: true
}), _dec6 = property({
    type: String,
    attribute: 'cancel-text',
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: Boolean,
    attribute: 'ok-disabled',
    reflect: true
}), _dec8 = property({
    type: Boolean,
    attribute: 'cancel-disabled',
    reflect: true
}), _dec9 = property({
    type: Boolean,
    attribute: 'hide-footer',
    reflect: true
}), _dec10 = property({
    type: Number,
    attribute: 'z-index',
    reflect: true,
    useDefault: true
}), _dec11 = query('slot[name="title"]'), _dec12 = query('slot[name="footer"]'), _dec13 = query('slot[name="close-icon"]'), _dec14 = query('.ve-modal-overlay'), _dec15 = query('.ve-modal-panel'), _dec16 = decorators_js_state(), _dec17 = decorators_js_state(), _dec18 = decorators_js_state();
class VeModalElement extends VeElement {
    static{
        ({ e: [_init_visible, _init_extra__init_visible, _init_width, _init_extra__init_width, _init_closable, _init_extra__init_closable, _init_mask, _init_extra__init_mask, _init_maskClosable, _init_extra__init_maskClosable, _init_okText, _init_extra__init_okText, _init_cancelText, _init_extra__init_cancelText, _init_okDisabled, _init_extra__init_okDisabled, _init_cancelDisabled, _init_extra__init_cancelDisabled, _init_hideFooter, _init_extra__init_hideFooter, _init_zIndex, _init_extra__init_zIndex, _init_titleSlot, _init_extra__init_titleSlot, _init_footerSlot, _init_extra__init_footerSlot, _init_closeIconSlot, _init_extra__init_closeIconSlot, _init_overlayElement, _init_extra__init_overlayElement, _init_panelElement, _init_extra__init_panelElement, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasFooterSlot, _init_extra__init_hasFooterSlot, _init_hasCloseIconSlot, _init_extra__init_hasCloseIconSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "visible"
            ],
            [
                _dec1,
                1,
                "width"
            ],
            [
                _dec2,
                1,
                "closable"
            ],
            [
                _dec3,
                1,
                "mask"
            ],
            [
                _dec4,
                1,
                "maskClosable"
            ],
            [
                _dec5,
                1,
                "okText"
            ],
            [
                _dec6,
                1,
                "cancelText"
            ],
            [
                _dec7,
                1,
                "okDisabled"
            ],
            [
                _dec8,
                1,
                "cancelDisabled"
            ],
            [
                _dec9,
                1,
                "hideFooter"
            ],
            [
                _dec10,
                1,
                "zIndex"
            ],
            [
                _dec11,
                1,
                "titleSlot"
            ],
            [
                _dec12,
                1,
                "footerSlot"
            ],
            [
                _dec13,
                1,
                "closeIconSlot"
            ],
            [
                _dec14,
                1,
                "overlayElement"
            ],
            [
                _dec15,
                1,
                "panelElement"
            ],
            [
                _dec16,
                1,
                "hasTitleSlot"
            ],
            [
                _dec17,
                1,
                "hasFooterSlot"
            ],
            [
                _dec18,
                1,
                "hasCloseIconSlot"
            ]
        ]));
    }
    static styles = ve_modal;
    static dependencies = {
        [VE_BUTTON_TAG]: VeButtonElement
    };
    #___private_visible_1 = (()=>{
        const _value = _init_visible(this, false);
        _init_extra__init_visible(this);
        return _value;
    })();
    get visible() {
        return this.#___private_visible_1;
    }
    set visible(_v) {
        this.#___private_visible_1 = _v;
    }
    #___private_width_2 = (()=>{
        const _value = _init_width(this, '480px');
        _init_extra__init_width(this);
        return _value;
    })();
    get width() {
        return this.#___private_width_2;
    }
    set width(_v) {
        this.#___private_width_2 = _v;
    }
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_closable_3 = (()=>{
        const _value = _init_closable(this, true);
        _init_extra__init_closable(this);
        return _value;
    })();
    get closable() {
        return this.#___private_closable_3;
    }
    set closable(_v) {
        this.#___private_closable_3 = _v;
    }
    #___private_mask_4 = (()=>{
        const _value = _init_mask(this, true);
        _init_extra__init_mask(this);
        return _value;
    })();
    get mask() {
        return this.#___private_mask_4;
    }
    set mask(_v) {
        this.#___private_mask_4 = _v;
    }
    #___private_maskClosable_5 = (()=>{
        const _value = _init_maskClosable(this, true);
        _init_extra__init_maskClosable(this);
        return _value;
    })();
    get maskClosable() {
        return this.#___private_maskClosable_5;
    }
    set maskClosable(_v) {
        this.#___private_maskClosable_5 = _v;
    }
    #___private_okText_6 = (()=>{
        const _value = _init_okText(this, '确定');
        _init_extra__init_okText(this);
        return _value;
    })();
    get okText() {
        return this.#___private_okText_6;
    }
    set okText(_v) {
        this.#___private_okText_6 = _v;
    }
    #___private_cancelText_7 = (()=>{
        const _value = _init_cancelText(this, '取消');
        _init_extra__init_cancelText(this);
        return _value;
    })();
    get cancelText() {
        return this.#___private_cancelText_7;
    }
    set cancelText(_v) {
        this.#___private_cancelText_7 = _v;
    }
    #___private_okDisabled_8 = (()=>{
        const _value = _init_okDisabled(this, false);
        _init_extra__init_okDisabled(this);
        return _value;
    })();
    get okDisabled() {
        return this.#___private_okDisabled_8;
    }
    set okDisabled(_v) {
        this.#___private_okDisabled_8 = _v;
    }
    #___private_cancelDisabled_9 = (()=>{
        const _value = _init_cancelDisabled(this, false);
        _init_extra__init_cancelDisabled(this);
        return _value;
    })();
    get cancelDisabled() {
        return this.#___private_cancelDisabled_9;
    }
    set cancelDisabled(_v) {
        this.#___private_cancelDisabled_9 = _v;
    }
    #___private_hideFooter_10 = (()=>{
        const _value = _init_hideFooter(this, false);
        _init_extra__init_hideFooter(this);
        return _value;
    })();
    get hideFooter() {
        return this.#___private_hideFooter_10;
    }
    set hideFooter(_v) {
        this.#___private_hideFooter_10 = _v;
    }
    #___private_zIndex_11 = (()=>{
        const _value = _init_zIndex(this, 1000);
        _init_extra__init_zIndex(this);
        return _value;
    })();
    get zIndex() {
        return this.#___private_zIndex_11;
    }
    set zIndex(_v) {
        this.#___private_zIndex_11 = _v;
    }
    #___private_titleSlot_12 = (()=>{
        const _value = _init_titleSlot(this, null);
        _init_extra__init_titleSlot(this);
        return _value;
    })();
    get titleSlot() {
        return this.#___private_titleSlot_12;
    }
    set titleSlot(_v) {
        this.#___private_titleSlot_12 = _v;
    }
    #___private_footerSlot_13 = (()=>{
        const _value = _init_footerSlot(this, null);
        _init_extra__init_footerSlot(this);
        return _value;
    })();
    get footerSlot() {
        return this.#___private_footerSlot_13;
    }
    set footerSlot(_v) {
        this.#___private_footerSlot_13 = _v;
    }
    #___private_closeIconSlot_14 = (()=>{
        const _value = _init_closeIconSlot(this, null);
        _init_extra__init_closeIconSlot(this);
        return _value;
    })();
    get closeIconSlot() {
        return this.#___private_closeIconSlot_14;
    }
    set closeIconSlot(_v) {
        this.#___private_closeIconSlot_14 = _v;
    }
    #___private_overlayElement_15 = (()=>{
        const _value = _init_overlayElement(this, null);
        _init_extra__init_overlayElement(this);
        return _value;
    })();
    get overlayElement() {
        return this.#___private_overlayElement_15;
    }
    set overlayElement(_v) {
        this.#___private_overlayElement_15 = _v;
    }
    #___private_panelElement_16 = (()=>{
        const _value = _init_panelElement(this, null);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_16;
    }
    set panelElement(_v) {
        this.#___private_panelElement_16 = _v;
    }
    #___private_hasTitleSlot_17 = (()=>{
        const _value = _init_hasTitleSlot(this, false);
        _init_extra__init_hasTitleSlot(this);
        return _value;
    })();
    get hasTitleSlot() {
        return this.#___private_hasTitleSlot_17;
    }
    set hasTitleSlot(_v) {
        this.#___private_hasTitleSlot_17 = _v;
    }
    #___private_hasFooterSlot_18 = (()=>{
        const _value = _init_hasFooterSlot(this, false);
        _init_extra__init_hasFooterSlot(this);
        return _value;
    })();
    get hasFooterSlot() {
        return this.#___private_hasFooterSlot_18;
    }
    set hasFooterSlot(_v) {
        this.#___private_hasFooterSlot_18 = _v;
    }
    #___private_hasCloseIconSlot_19 = (()=>{
        const _value = _init_hasCloseIconSlot(this, false);
        _init_extra__init_hasCloseIconSlot(this);
        return _value;
    })();
    get hasCloseIconSlot() {
        return this.#___private_hasCloseIconSlot_19;
    }
    set hasCloseIconSlot(_v) {
        this.#___private_hasCloseIconSlot_19 = _v;
    }
    lastCloseSource = 'api';
    closeFallbackTimer = null;
    closeTransitionTarget = null;
    scrollLocked = false;
    scrollLockTarget = null;
    scrollGuardTarget = null;
    touchStartX = 0;
    touchStartY = 0;
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this.syncDocumentKeydownListener();
    }
    firstUpdated() {
        this.syncSlotState();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('visible')) {
            this.syncDocumentKeydownListener();
            if (this.visible) {
                this.lockBackgroundScroll();
                this.cancelAfterCloseWait();
            } else if (true === changedProperties.get('visible')) this.waitForCloseTransition();
        }
    }
    disconnectedCallback() {
        this.cancelAfterCloseWait();
        this.removeDocumentKeydownListener();
        this.unlockBackgroundScroll();
        super.disconnectedCallback();
    }
    show() {
        this.requestVisible(true, 'api');
    }
    hide(source = 'api', sourceEvent) {
        this.requestVisible(false, source, sourceEvent);
    }
    syncSlotState() {
        this.hasTitleSlot = hasAssignedSlotContent(this.titleSlot) || hasNamedSlottedContent(this, 'title');
        this.hasFooterSlot = hasAssignedSlotContent(this.footerSlot) || hasNamedSlottedContent(this, 'footer');
        this.hasCloseIconSlot = hasAssignedSlotContent(this.closeIconSlot) || hasNamedSlottedContent(this, 'close-icon');
    }
    handleSlotChange = ()=>{
        this.syncSlotState();
    };
    dispatchVisibleChange(visible, source, sourceEvent) {
        return this.dispatchEvent(new CustomEvent('ve-visible-change', {
            detail: sourceEvent ? {
                visible,
                source,
                sourceEvent
            } : {
                visible,
                source
            },
            bubbles: true,
            composed: true,
            cancelable: true
        }));
    }
    requestVisible(visible, source, sourceEvent) {
        if (this.visible === visible) return;
        if (!this.dispatchVisibleChange(visible, source, sourceEvent)) return;
        if (!visible) this.lastCloseSource = source;
        this.visible = visible;
    }
    requestClose(source, sourceEvent) {
        const closeEvent = new CustomEvent('ve-close', {
            detail: sourceEvent ? {
                source,
                sourceEvent
            } : {
                source
            },
            bubbles: true,
            composed: true,
            cancelable: true
        });
        this.dispatchEvent(closeEvent);
        if (!closeEvent.defaultPrevented) this.requestVisible(false, source, sourceEvent);
    }
    dispatchAfterClose() {
        this.unlockBackgroundScroll();
        this.dispatchEvent(new CustomEvent('ve-after-close', {
            detail: {
                source: this.lastCloseSource
            },
            bubbles: true,
            composed: true
        }));
    }
    waitForCloseTransition() {
        this.cancelAfterCloseWait();
        if (shouldReduceMotion()) return void this.dispatchAfterClose();
        const targets = [
            this.overlayElement,
            this.panelElement
        ].filter((target)=>null !== target);
        const transitionDuration = targets.reduce((longest, target)=>Math.max(longest, getLongestTransitionMs(target)), 0);
        if (transitionDuration <= 0) return void this.dispatchAfterClose();
        const transitionTarget = targets.find((target)=>getLongestTransitionMs(target) === transitionDuration) ?? targets[0];
        if (!transitionTarget) return void this.dispatchAfterClose();
        this.closeTransitionTarget = transitionTarget;
        transitionTarget.addEventListener('transitionend', this.handleCloseTransitionEnd);
        this.closeFallbackTimer = window.setTimeout(()=>{
            this.closeFallbackTimer = null;
            this.finishAfterCloseWait();
        }, transitionDuration + 50);
    }
    finishAfterCloseWait() {
        if (this.visible) return void this.cancelAfterCloseWait();
        this.cancelAfterCloseWait();
        this.dispatchAfterClose();
    }
    cancelAfterCloseWait() {
        if (null !== this.closeFallbackTimer) {
            window.clearTimeout(this.closeFallbackTimer);
            this.closeFallbackTimer = null;
        }
        this.closeTransitionTarget?.removeEventListener('transitionend', this.handleCloseTransitionEnd);
        this.closeTransitionTarget = null;
    }
    syncDocumentKeydownListener() {
        if (this.visible) return void document.addEventListener('keydown', this.handleDocumentKeydown);
        this.removeDocumentKeydownListener();
    }
    removeDocumentKeydownListener() {
        document.removeEventListener('keydown', this.handleDocumentKeydown);
    }
    lockBackgroundScroll() {
        if (this.scrollLocked) return;
        const target = getScrollLockTarget(this);
        lockScrollTarget(target);
        this.scrollLockTarget = target;
        this.scrollLocked = true;
        this.addScrollGuard();
    }
    unlockBackgroundScroll() {
        if (!this.scrollLocked || !this.scrollLockTarget) return;
        unlockScrollTarget(this.scrollLockTarget);
        this.scrollLockTarget = null;
        this.scrollLocked = false;
        this.removeScrollGuard();
    }
    addScrollGuard() {
        const overlay = this.overlayElement;
        if (!overlay || this.scrollGuardTarget) return;
        overlay.addEventListener('wheel', this.handleOverlayWheel, {
            passive: false
        });
        overlay.addEventListener('touchstart', this.handleOverlayTouchStart, {
            passive: true
        });
        overlay.addEventListener('touchmove', this.handleOverlayTouchMove, {
            passive: false
        });
        this.scrollGuardTarget = overlay;
    }
    removeScrollGuard() {
        const overlay = this.scrollGuardTarget;
        if (!overlay) return;
        overlay.removeEventListener('wheel', this.handleOverlayWheel);
        overlay.removeEventListener('touchstart', this.handleOverlayTouchStart);
        overlay.removeEventListener('touchmove', this.handleOverlayTouchMove);
        this.scrollGuardTarget = null;
    }
    handleOverlayClick = (event)=>{
        if (event.target !== event.currentTarget || !this.maskClosable) return;
        this.requestClose('mask', event);
    };
    isScrollConsumedInside(event, deltaX, deltaY) {
        const overlay = this.overlayElement;
        if (!overlay) return false;
        const path = event.composedPath();
        for (const node of path){
            if (node === overlay) break;
            if (node instanceof HTMLElement && canConsumeScroll(node, deltaX, deltaY)) return true;
        }
        return false;
    }
    handleOverlayWheel = (event)=>{
        if (this.isScrollConsumedInside(event, event.deltaX, event.deltaY)) return;
        if (event.cancelable) event.preventDefault();
    };
    handleOverlayTouchStart = (event)=>{
        const touch = event.touches[0];
        if (touch) {
            this.touchStartX = touch.clientX;
            this.touchStartY = touch.clientY;
        }
    };
    handleOverlayTouchMove = (event)=>{
        const touch = event.touches[0];
        if (!touch) return;
        const deltaX = this.touchStartX - touch.clientX;
        const deltaY = this.touchStartY - touch.clientY;
        if (this.isScrollConsumedInside(event, deltaX, deltaY)) return;
        if (event.cancelable) event.preventDefault();
    };
    handleDocumentKeydown = (event)=>{
        if (!this.visible || 'Escape' !== event.key) return;
        event.stopPropagation();
        this.requestClose('esc', event);
    };
    handleCloseTransitionEnd = (event)=>{
        if (event.target !== this.closeTransitionTarget) return;
        this.finishAfterCloseWait();
    };
    handleCloseClick = (event)=>{
        this.requestClose('close', event);
    };
    handleCancelClick = (event)=>{
        const cancelEvent = new CustomEvent('ve-cancel', {
            detail: {
                source: 'cancel',
                sourceEvent: event
            },
            bubbles: true,
            composed: true,
            cancelable: true
        });
        this.dispatchEvent(cancelEvent);
        if (!cancelEvent.defaultPrevented) this.requestVisible(false, 'cancel', event);
    };
    handleOkClick = (event)=>{
        const okEvent = new CustomEvent('ve-ok', {
            detail: {
                source: 'ok',
                sourceEvent: event
            },
            bubbles: true,
            composed: true,
            cancelable: true
        });
        this.dispatchEvent(okEvent);
        if (!okEvent.defaultPrevented) this.requestVisible(false, 'ok', event);
    };
    renderCloseIcon() {
        return html`
      <slot name="close-icon" @slotchange=${this.handleSlotChange}></slot>
      ${this.hasCloseIconSlot ? nothing : html`
            <svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
              <path
                d="M18 6 6 18M6 6l12 12"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="1.5"
              ></path>
            </svg>
          `}
    `;
    }
    renderCloseButton() {
        if (!this.closable) return nothing;
        return html`
      <ve-button
        class="ve-modal-close-button"
        part="close-button"
        type="secondary"
        size="small"
        shape="default"
        aria-label="Close"
        @click=${this.handleCloseClick}
      >
        ${this.renderCloseIcon()}
      </ve-button>
    `;
    }
    renderTitle() {
        return html`
      <div class="ve-modal-title" part="title">
        <slot name="title" @slotchange=${this.handleSlotChange}></slot>
        ${this.hasTitleSlot ? nothing : this.title}
      </div>
    `;
    }
    renderFooter() {
        if (this.hideFooter) return nothing;
        return html`
      <div class="ve-modal-footer" part="footer">
        <slot
          class="ve-modal-footer-slot"
          name="footer"
          @slotchange=${this.handleSlotChange}
        ></slot>
        ${this.hasFooterSlot ? nothing : html`
              <button
                class="ve-modal-action ve-modal-action-cancel"
                part="cancel-button"
                type="button"
                ?disabled=${this.cancelDisabled}
                @click=${this.handleCancelClick}
              >
                <slot name="cancel">${this.cancelText}</slot>
              </button>
              <button
                class="ve-modal-action ve-modal-action-ok"
                part="ok-button"
                type="button"
                ?disabled=${this.okDisabled}
                @click=${this.handleOkClick}
              >
                <slot name="ok">${this.okText}</slot>
              </button>
            `}
      </div>
    `;
    }
    render() {
        const overlayClass = classMap({
            've-modal-overlay': true,
            've-modal-overlay-open': this.visible,
            've-modal-overlay-no-mask': !this.mask
        });
        const style = `--ve-modal-current-width:${normalizeWidth(this.width)};--ve-modal-current-z-index:${this.zIndex};`;
        return html`
      <div
        class=${overlayClass}
        part="overlay"
        style=${style}
        role="presentation"
        @click=${this.handleOverlayClick}
      >
        <section
          class="ve-modal-panel"
          part="panel"
          role="dialog"
          aria-modal="true"
          aria-hidden=${this.visible ? 'false' : 'true'}
          tabindex="-1"
        >
          <header class="ve-modal-header" part="header">
            ${this.renderTitle()} ${this.renderCloseButton()}
          </header>
          <div class="ve-modal-body" part="body">
            <slot></slot>
          </div>
          ${this.renderFooter()}
        </section>
      </div>
    `;
    }
}
defineCustomElement(VE_MODAL_TAG, VeModalElement);
export { VE_MODAL_TAG, VeModalElement };
