// @ve-design/web/dist/ve-notification.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { infoIcon } from "./icons/info.js";
import { checkCircleIcon } from "./icons/check-circle.js";
import { warningIcon } from "./icons/warning.js";
import { errorIcon } from "./icons/error.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
import { closeIcon } from "./icons/close.js";
const veNotificationTokenDefaults = css`
  :host {
    --ve-notification-wrapper-offset-block: var(--space-xxl);
    --ve-notification-wrapper-offset-inline: var(--space-l);
    --ve-notification-wrapper-gap: var(--space-s);
    --ve-notification-z-index: var(--z-index-notification, 1004);

    --ve-notification-width: 400px;
    --ve-notification-max-width: calc(100vw - var(--space-l) * 2);
    --ve-notification-padding: var(--space-s);
    --ve-notification-gap: var(--space-xxs);
    --ve-notification-content-gap: var(--space-xxxs);
    --ve-notification-action-gap: var(--space-xxxs);
    --ve-notification-border-width: var(--stroke-0_5);
    --ve-notification-border-style: solid;
    --ve-notification-border-radius: var(--radius-xl);
    --ve-notification-box-shadow: var(--shadow-lg);
    --ve-notification-background: var(--color-bg-base);
    --ve-notification-border-color: var(--color-border-default);
    --ve-notification-font-family: var(--font-cn);

    --ve-notification-title-color: var(--color-text-primary);
    --ve-notification-title-font-size: var(--text-body);
    --ve-notification-title-font-weight: var(--font-weight-medium);
    --ve-notification-title-line-height: 22px;

    --ve-notification-content-color: var(--color-text-tertiary);
    --ve-notification-content-font-size: var(--text-body-sm);
    --ve-notification-content-font-weight: var(--font-weight-normal);
    --ve-notification-content-line-height: var(--line-height-body);
    --ve-notification-content-link-color: var(--color-text-secondary);

    --ve-notification-icon-size: 18px;
    --ve-notification-icon-line-height: 22px;
    --ve-notification-info-color-icon: var(--color-bg-info);
    --ve-notification-success-color-icon: var(--color-bg-success);
    --ve-notification-warning-color-icon: var(--color-bg-warning);
    --ve-notification-error-color-icon: var(--color-bg-danger);
    --ve-notification-current-color-icon: var(
      --ve-notification-info-color-icon
    );

    --ve-notification-close-button-size: 22px;
    --ve-notification-close-button-radius: var(--radius-xs);
    --ve-notification-close-icon-size: var(--text-ui-strong);
    --ve-notification-close-icon-color: var(--color-icon-secondary);
    --ve-notification-close-icon-color-hover: var(--color-icon-primary);
    --ve-notification-close-button-hover-background: color-mix(
      var(--color-bg-base) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );

    --ve-notification-enter-duration: 100ms;
    --ve-notification-fade-duration: 100ms;
    --ve-notification-exit-duration: 300ms;
    --ve-notification-icon-scale-duration: 400ms;
    --ve-notification-linear-easing: cubic-bezier(0, 0, 1, 1);
    --ve-notification-standard-easing: cubic-bezier(0.34, 0.69, 0.1, 1);
    --ve-notification-overshoot-easing: cubic-bezier(0.3, 1.3, 0.3, 1);
  }
`;
const token = veNotificationTokenDefaults;
const ve_notification = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      box-sizing: border-box;
      width: var(--ve-notification-width);
      max-width: var(--ve-notification-max-width);
      pointer-events: none;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-notification {
      position: relative;
      display: flex;
      flex-direction: column;
      box-sizing: border-box;
      width: 100%;
      min-width: 0;
      gap: var(--ve-notification-gap);
      padding: var(--ve-notification-padding);
      color: var(--ve-notification-content-color);
      font-family: var(--ve-notification-font-family);
      letter-spacing: var(--letter-spacing-normal);
      text-align: start;
      background: var(--ve-notification-background);
      border: var(--ve-notification-border-width)
        var(--ve-notification-border-style) var(--ve-notification-border-color);
      border-radius: var(--ve-notification-border-radius);
      box-shadow: var(--ve-notification-box-shadow);
      opacity: 0;
      overflow: hidden;
      pointer-events: auto;
      transition: opacity var(--ve-notification-enter-duration)
        var(--ve-notification-linear-easing);
    }

    .ve-notification-info {
      --ve-notification-current-color-icon: var(
        --ve-notification-info-color-icon
      );
    }

    .ve-notification-success {
      --ve-notification-current-color-icon: var(
        --ve-notification-success-color-icon
      );
    }

    .ve-notification-warning {
      --ve-notification-current-color-icon: var(
        --ve-notification-warning-color-icon
      );
    }

    .ve-notification-error {
      --ve-notification-current-color-icon: var(
        --ve-notification-error-color-icon
      );
    }

    .ve-notification-entered {
      opacity: 1;
    }

    .ve-notification-leaving {
      opacity: 0;
      transition: opacity var(--ve-notification-fade-duration)
        var(--ve-notification-linear-easing);
    }

    .ve-notification-body {
      display: flex;
      align-items: flex-start;
      box-sizing: border-box;
      min-width: 0;
      width: 100%;
      gap: var(--ve-notification-gap);
      padding-inline-end: calc(
        var(--ve-notification-close-button-size) + var(--ve-notification-gap)
      );
    }

    .ve-notification:not(.ve-notification-closable) .ve-notification-body {
      padding-inline-end: 0;
    }

    .ve-notification-icon {
      display: inline-flex;
      flex: none;
      align-items: flex-start;
      justify-content: center;
      width: var(--ve-notification-icon-size);
      height: var(--ve-notification-icon-line-height);
      padding-top: 2px;
      color: var(--ve-notification-current-color-icon);
      font-size: var(--ve-notification-icon-size);
      line-height: 1;
      animation:
        ve-notification-fade var(--ve-notification-fade-duration)
          var(--ve-notification-linear-easing),
        ve-notification-scale var(--ve-notification-icon-scale-duration)
          var(--ve-notification-overshoot-easing);
    }

    .ve-notification-icon ve-icon,
    .ve-notification-icon svg,
    .ve-notification-icon ::slotted(svg) {
      display: inline-flex;
      width: 1em;
      height: 1em;
      color: inherit;
      flex: none;
    }

    .ve-notification-icon ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-width: 1em;
      max-height: 1em;
      color: inherit;
      line-height: 1;
    }

    .ve-notification-main {
      display: flex;
      flex: 1 1 auto;
      flex-direction: column;
      min-width: 0;
      gap: var(--ve-notification-content-gap);
    }

    .ve-notification-title {
      display: block;
      min-width: 0;
      color: var(--ve-notification-title-color);
      font-size: var(--ve-notification-title-font-size);
      font-weight: var(--ve-notification-title-font-weight);
      line-height: var(--ve-notification-title-line-height);
      word-break: break-word;
    }

    .ve-notification-content {
      display: block;
      min-width: 0;
      color: var(--ve-notification-content-color);
      font-size: var(--ve-notification-content-font-size);
      font-weight: var(--ve-notification-content-font-weight);
      line-height: var(--ve-notification-content-line-height);
      word-break: break-word;
    }

    .ve-notification-content ::slotted(a) {
      color: var(--ve-notification-content-link-color);
      text-decoration-line: underline;
      text-decoration-style: dotted;
      text-decoration-thickness: 6%;
      text-underline-offset: 2px;
    }

    .ve-notification-actions {
      display: flex;
      align-items: center;
      justify-content: flex-end;
      box-sizing: border-box;
      width: 100%;
      min-width: 0;
      gap: var(--ve-notification-action-gap);
    }

    .ve-notification-actions ::slotted(*) {
      flex: none;
    }

    .ve-notification-close-btn {
      position: absolute;
      inset-block-start: var(--ve-notification-padding);
      inset-inline-end: var(--ve-notification-padding);
      display: inline-flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      width: var(--ve-notification-close-button-size);
      height: var(--ve-notification-close-button-size);
      padding: 0;
      color: var(--ve-notification-close-icon-color);
      font: inherit;
      font-size: var(--ve-notification-close-icon-size);
      line-height: 1;
      background: transparent;
      border: 0;
      border-radius: var(--ve-notification-close-button-radius);
      cursor: pointer;
      transition:
        background-color var(--ve-notification-fade-duration)
          var(--ve-notification-linear-easing),
        color var(--ve-notification-fade-duration)
          var(--ve-notification-linear-easing);
    }

    .ve-notification-close-btn:hover,
    .ve-notification-close-btn:focus-visible {
      color: var(--ve-notification-close-icon-color-hover);
      background: var(--ve-notification-close-button-hover-background);
    }

    .ve-notification-close-btn slot,
    .ve-notification-close-btn ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
      width: 100%;
      height: 100%;
      color: inherit;
      line-height: 1;
    }

    .ve-notification-close-btn ve-icon,
    .ve-notification-close-btn svg,
    .ve-notification-close-btn ::slotted(svg) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      width: 1em;
      height: 1em;
      color: inherit;
      line-height: 0;
    }

    @keyframes ve-notification-fade {
      from {
        opacity: 0;
      }

      to {
        opacity: 1;
      }
    }

    @keyframes ve-notification-scale {
      from {
        transform: scale(0);
      }

      to {
        transform: scale(1);
      }
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-notification,
      .ve-notification-leaving,
      .ve-notification-close-btn {
        transition: none;
      }

      .ve-notification-icon {
        animation: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _init_type, _init_extra__init_type, _init_content, _init_extra__init_content, _init_notificationId, _init_extra__init_notificationId, _init_position, _init_extra__init_position, _init_showIcon, _init_extra__init_showIcon, _init_closable, _init_extra__init_closable, _init_duration, _init_extra__init_duration, _init_enterDuration, _init_extra__init_enterDuration, _init_exitDuration, _init_extra__init_exitDuration, _init_titleSlot, _init_extra__init_titleSlot, _init_iconSlot, _init_extra__init_iconSlot, _init_actionsSlot, _init_extra__init_actionsSlot, _init_closeIconSlot, _init_extra__init_closeIconSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasActionsSlot, _init_extra__init_hasActionsSlot, _init_hasCloseIconSlot, _init_extra__init_hasCloseIconSlot, _init_phase, _init_extra__init_phase;
const VE_NOTIFICATION_TAG = 've-notification';
const VE_NOTIFICATION_CLOSE_EVENT = 've-close';
const VE_NOTIFICATION_AFTER_CLOSE_EVENT = 've-after-close';
const NOTIFICATION_TYPES = [
    'info',
    'success',
    'warning',
    'error'
];
const NOTIFICATION_POSITIONS = [
    'top-right',
    'top-left',
    'bottom-right',
    'bottom-left'
];
const DEFAULT_ICON_BY_TYPE = {
    info: infoIcon,
    success: checkCircleIcon,
    warning: warningIcon,
    error: errorIcon
};
const ICON_SLOT_SIZE = 'var(--ve-notification-icon-size)';
const CLOSE_ICON_SLOT_SIZE = 'var(--ve-notification-close-icon-size)';
const NOTIFICATION_DEFAULT_ENTER_DURATION = 100;
const NOTIFICATION_DEFAULT_EXIT_DURATION = 300;
const NOTIFICATION_DEFAULT_DURATION = 3000;
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
function normalizeDuration(value, fallback) {
    return Number.isFinite(value) && value >= 0 ? value : fallback;
}
function shouldReduceMotion() {
    return "u" > typeof window && 'function' == typeof window.matchMedia && window.matchMedia('(prefers-reduced-motion: reduce)').matches;
}
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: String,
    reflect: true
}), _dec2 = property({
    type: String,
    attribute: 'notification-id',
    reflect: true
}), _dec3 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec4 = property({
    attribute: 'show-icon',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = property({
    type: Number,
    reflect: true,
    useDefault: true
}), _dec7 = property({
    type: Number,
    attribute: 'enter-duration',
    reflect: true,
    useDefault: true
}), _dec8 = property({
    type: Number,
    attribute: 'exit-duration',
    reflect: true,
    useDefault: true
}), _dec9 = query('slot[name="title"]'), _dec10 = query('slot[name="icon"]'), _dec11 = query('slot[name="actions"]'), _dec12 = query('slot[name="close-icon"]'), _dec13 = state(), _dec14 = state(), _dec15 = state(), _dec16 = state(), _dec17 = state();
class VeNotificationElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_content, _init_extra__init_content, _init_notificationId, _init_extra__init_notificationId, _init_position, _init_extra__init_position, _init_showIcon, _init_extra__init_showIcon, _init_closable, _init_extra__init_closable, _init_duration, _init_extra__init_duration, _init_enterDuration, _init_extra__init_enterDuration, _init_exitDuration, _init_extra__init_exitDuration, _init_titleSlot, _init_extra__init_titleSlot, _init_iconSlot, _init_extra__init_iconSlot, _init_actionsSlot, _init_extra__init_actionsSlot, _init_closeIconSlot, _init_extra__init_closeIconSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasActionsSlot, _init_extra__init_hasActionsSlot, _init_hasCloseIconSlot, _init_extra__init_hasCloseIconSlot, _init_phase, _init_extra__init_phase] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "type"
            ],
            [
                _dec1,
                1,
                "content"
            ],
            [
                _dec2,
                1,
                "notificationId"
            ],
            [
                _dec3,
                1,
                "position"
            ],
            [
                _dec4,
                1,
                "showIcon"
            ],
            [
                _dec5,
                1,
                "closable"
            ],
            [
                _dec6,
                1,
                "duration"
            ],
            [
                _dec7,
                1,
                "enterDuration"
            ],
            [
                _dec8,
                1,
                "exitDuration"
            ],
            [
                _dec9,
                1,
                "titleSlot"
            ],
            [
                _dec10,
                1,
                "iconSlot"
            ],
            [
                _dec11,
                1,
                "actionsSlot"
            ],
            [
                _dec12,
                1,
                "closeIconSlot"
            ],
            [
                _dec13,
                1,
                "hasTitleSlot"
            ],
            [
                _dec14,
                1,
                "hasIconSlot"
            ],
            [
                _dec15,
                1,
                "hasActionsSlot"
            ],
            [
                _dec16,
                1,
                "hasCloseIconSlot"
            ],
            [
                _dec17,
                1,
                "phase"
            ]
        ]));
    }
    static styles = ve_notification;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'info');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_content_2 = (()=>{
        const _value = _init_content(this, '');
        _init_extra__init_content(this);
        return _value;
    })();
    get content() {
        return this.#___private_content_2;
    }
    set content(_v) {
        this.#___private_content_2 = _v;
    }
    #___private_notificationId_3 = (()=>{
        const _value = _init_notificationId(this, '');
        _init_extra__init_notificationId(this);
        return _value;
    })();
    get notificationId() {
        return this.#___private_notificationId_3;
    }
    set notificationId(_v) {
        this.#___private_notificationId_3 = _v;
    }
    #___private_position_4 = (()=>{
        const _value = _init_position(this, 'top-right');
        _init_extra__init_position(this);
        return _value;
    })();
    get position() {
        return this.#___private_position_4;
    }
    set position(_v) {
        this.#___private_position_4 = _v;
    }
    #___private_showIcon_5 = (()=>{
        const _value = _init_showIcon(this, true);
        _init_extra__init_showIcon(this);
        return _value;
    })();
    get showIcon() {
        return this.#___private_showIcon_5;
    }
    set showIcon(_v) {
        this.#___private_showIcon_5 = _v;
    }
    #___private_closable_6 = (()=>{
        const _value = _init_closable(this, false);
        _init_extra__init_closable(this);
        return _value;
    })();
    get closable() {
        return this.#___private_closable_6;
    }
    set closable(_v) {
        this.#___private_closable_6 = _v;
    }
    #___private_duration_7 = (()=>{
        const _value = _init_duration(this, NOTIFICATION_DEFAULT_DURATION);
        _init_extra__init_duration(this);
        return _value;
    })();
    get duration() {
        return this.#___private_duration_7;
    }
    set duration(_v) {
        this.#___private_duration_7 = _v;
    }
    #___private_enterDuration_8 = (()=>{
        const _value = _init_enterDuration(this, NOTIFICATION_DEFAULT_ENTER_DURATION);
        _init_extra__init_enterDuration(this);
        return _value;
    })();
    get enterDuration() {
        return this.#___private_enterDuration_8;
    }
    set enterDuration(_v) {
        this.#___private_enterDuration_8 = _v;
    }
    #___private_exitDuration_9 = (()=>{
        const _value = _init_exitDuration(this, NOTIFICATION_DEFAULT_EXIT_DURATION);
        _init_extra__init_exitDuration(this);
        return _value;
    })();
    get exitDuration() {
        return this.#___private_exitDuration_9;
    }
    set exitDuration(_v) {
        this.#___private_exitDuration_9 = _v;
    }
    #___private_titleSlot_10 = (()=>{
        const _value = _init_titleSlot(this, null);
        _init_extra__init_titleSlot(this);
        return _value;
    })();
    get titleSlot() {
        return this.#___private_titleSlot_10;
    }
    set titleSlot(_v) {
        this.#___private_titleSlot_10 = _v;
    }
    #___private_iconSlot_11 = (()=>{
        const _value = _init_iconSlot(this, null);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_11;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_11 = _v;
    }
    #___private_actionsSlot_12 = (()=>{
        const _value = _init_actionsSlot(this, null);
        _init_extra__init_actionsSlot(this);
        return _value;
    })();
    get actionsSlot() {
        return this.#___private_actionsSlot_12;
    }
    set actionsSlot(_v) {
        this.#___private_actionsSlot_12 = _v;
    }
    #___private_closeIconSlot_13 = (()=>{
        const _value = _init_closeIconSlot(this, null);
        _init_extra__init_closeIconSlot(this);
        return _value;
    })();
    get closeIconSlot() {
        return this.#___private_closeIconSlot_13;
    }
    set closeIconSlot(_v) {
        this.#___private_closeIconSlot_13 = _v;
    }
    #___private_hasTitleSlot_14 = (()=>{
        const _value = _init_hasTitleSlot(this, false);
        _init_extra__init_hasTitleSlot(this);
        return _value;
    })();
    get hasTitleSlot() {
        return this.#___private_hasTitleSlot_14;
    }
    set hasTitleSlot(_v) {
        this.#___private_hasTitleSlot_14 = _v;
    }
    #___private_hasIconSlot_15 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_15;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_15 = _v;
    }
    #___private_hasActionsSlot_16 = (()=>{
        const _value = _init_hasActionsSlot(this, false);
        _init_extra__init_hasActionsSlot(this);
        return _value;
    })();
    get hasActionsSlot() {
        return this.#___private_hasActionsSlot_16;
    }
    set hasActionsSlot(_v) {
        this.#___private_hasActionsSlot_16 = _v;
    }
    #___private_hasCloseIconSlot_17 = (()=>{
        const _value = _init_hasCloseIconSlot(this, false);
        _init_extra__init_hasCloseIconSlot(this);
        return _value;
    })();
    get hasCloseIconSlot() {
        return this.#___private_hasCloseIconSlot_17;
    }
    set hasCloseIconSlot(_v) {
        this.#___private_hasCloseIconSlot_17 = _v;
    }
    #___private_phase_18 = (()=>{
        const _value = _init_phase(this, 'entering');
        _init_extra__init_phase(this);
        return _value;
    })();
    get phase() {
        return this.#___private_phase_18;
    }
    set phase(_v) {
        this.#___private_phase_18 = _v;
    }
    autoCloseTimer = null;
    enterFrame = null;
    closeOpacityTimer = null;
    closeCompleteTimer = null;
    hovering = false;
    focusWithin = false;
    lightDomSlotObserver;
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    get normalizedType() {
        return NOTIFICATION_TYPES.includes(this.type) ? this.type : 'info';
    }
    get normalizedPosition() {
        return NOTIFICATION_POSITIONS.includes(this.position) ? this.position : 'top-right';
    }
    get normalizedEnterDuration() {
        return normalizeDuration(this.enterDuration, NOTIFICATION_DEFAULT_ENTER_DURATION);
    }
    get normalizedExitDuration() {
        return normalizeDuration(this.exitDuration, NOTIFICATION_DEFAULT_EXIT_DURATION);
    }
    get normalizedDuration() {
        return normalizeDuration(this.duration, NOTIFICATION_DEFAULT_DURATION);
    }
    get notificationRole() {
        return 'error' === this.normalizedType || 'warning' === this.normalizedType ? 'alert' : 'status';
    }
    get notificationAriaLive() {
        return 'status' === this.notificationRole ? 'polite' : 'assertive';
    }
    get hasTitleContent() {
        return this.hasTitleSlot || hasNamedSlottedContent(this, 'title') || (this.title ?? '').trim().length > 0;
    }
    get hasIconContent() {
        return this.hasIconSlot || hasNamedSlottedContent(this, 'icon');
    }
    get hasActionsContent() {
        return this.hasActionsSlot || hasNamedSlottedContent(this, 'actions');
    }
    get hasCloseIconContent() {
        return this.hasCloseIconSlot || hasNamedSlottedContent(this, 'close-icon');
    }
    connectedCallback() {
        super.connectedCallback();
        this.observeLightDomSlots();
        this.syncSlotState();
        this.applyTransitionVariables();
    }
    firstUpdated() {
        this.syncSlotState();
        this.scheduleEnter();
        this.startTimer();
    }
    updated(changedProperties) {
        if (changedProperties.has('duration') && 'entered' === this.phase) this.restartTimer();
        if (changedProperties.has('enterDuration') || changedProperties.has('exitDuration')) this.applyTransitionVariables();
    }
    disconnectedCallback() {
        this.lightDomSlotObserver?.disconnect();
        this.lightDomSlotObserver = void 0;
        this.clearTimers();
        super.disconnectedCallback();
    }
    restartTimer() {
        this.clearAutoCloseTimer();
        this.syncAutoCloseTimer();
    }
    close(sourceEvent) {
        if ('leaving' === this.phase || 'closed' === this.phase) return;
        this.clearAutoCloseTimer();
        this.clearCloseTimers();
        if (!this.dispatchNotificationEvent(VE_NOTIFICATION_CLOSE_EVENT, sourceEvent)) return void this.syncAutoCloseTimer();
        this.phase = 'leaving';
        if (shouldReduceMotion()) return void this.finishClose(sourceEvent);
        const hostHeight = this.scrollHeight || this.getBoundingClientRect().height;
        this.style.height = `${hostHeight}px`;
        this.style.overflow = 'hidden';
        this.style.transition = 'none';
        this.getBoundingClientRect();
        const fadeDuration = this.normalizedEnterDuration;
        this.closeOpacityTimer = window.setTimeout(()=>{
            this.closeOpacityTimer = null;
            this.style.transition = `height ${this.normalizedExitDuration}ms var(--ve-notification-standard-easing)`;
            this.style.height = '0px';
        }, fadeDuration);
        this.closeCompleteTimer = window.setTimeout(()=>{
            this.closeCompleteTimer = null;
            this.finishClose(sourceEvent);
        }, fadeDuration + this.normalizedExitDuration);
    }
    applyTransitionVariables() {
        this.style.setProperty('--ve-notification-enter-duration', `${this.normalizedEnterDuration}ms`);
        this.style.setProperty('--ve-notification-fade-duration', `${this.normalizedEnterDuration}ms`);
        this.style.setProperty('--ve-notification-exit-duration', `${this.normalizedExitDuration}ms`);
    }
    scheduleEnter() {
        this.cancelEnterFrame();
        this.phase = 'entering';
        this.enterFrame = window.requestAnimationFrame(()=>{
            this.enterFrame = null;
            if ('entering' === this.phase) this.phase = 'entered';
        });
    }
    startTimer() {
        if (0 === this.duration || 'leaving' === this.phase || 'closed' === this.phase) return;
        if (null !== this.autoCloseTimer) return;
        this.autoCloseTimer = window.setTimeout(()=>{
            this.autoCloseTimer = null;
            this.close();
        }, this.normalizedDuration);
    }
    clearAutoCloseTimer() {
        if (null === this.autoCloseTimer) return;
        window.clearTimeout(this.autoCloseTimer);
        this.autoCloseTimer = null;
    }
    clearCloseTimers() {
        if (null !== this.closeOpacityTimer) {
            window.clearTimeout(this.closeOpacityTimer);
            this.closeOpacityTimer = null;
        }
        if (null !== this.closeCompleteTimer) {
            window.clearTimeout(this.closeCompleteTimer);
            this.closeCompleteTimer = null;
        }
    }
    cancelEnterFrame() {
        if (null === this.enterFrame) return;
        window.cancelAnimationFrame(this.enterFrame);
        this.enterFrame = null;
    }
    clearTimers() {
        this.clearAutoCloseTimer();
        this.clearCloseTimers();
        this.cancelEnterFrame();
    }
    syncAutoCloseTimer() {
        if (this.hovering || this.focusWithin) return void this.clearAutoCloseTimer();
        this.startTimer();
    }
    isNotificationNode(target) {
        return target instanceof Node && (this.contains(target) || this.renderRoot.contains(target));
    }
    syncSlotState() {
        this.hasTitleSlot = hasAssignedSlotContent(this.titleSlot) || hasNamedSlottedContent(this, 'title');
        this.hasIconSlot = hasAssignedSlotContent(this.iconSlot) || hasNamedSlottedContent(this, 'icon');
        this.hasActionsSlot = hasAssignedSlotContent(this.actionsSlot) || hasNamedSlottedContent(this, 'actions');
        this.hasCloseIconSlot = hasAssignedSlotContent(this.closeIconSlot) || hasNamedSlottedContent(this, 'close-icon');
        this.syncSlottedIconSize(this.iconSlot, ICON_SLOT_SIZE);
        this.syncSlottedIconSize(this.closeIconSlot, CLOSE_ICON_SLOT_SIZE);
    }
    syncSlottedIconSize(slot, fallbackSize) {
        const icons = slot?.assignedElements({
            flatten: true
        }).filter((element)=>element.localName === VE_ICON_TAG);
        icons?.forEach((icon)=>{
            if ('' === `${icon.getAttribute('size') ?? ''}`.trim() && '' === `${icon.size ?? ''}`.trim()) icon.size = fallbackSize;
        });
    }
    observeLightDomSlots() {
        if ("u" < typeof MutationObserver || this.lightDomSlotObserver) return;
        this.lightDomSlotObserver = new MutationObserver(()=>{
            this.syncSlotState();
        });
        this.lightDomSlotObserver.observe(this, {
            childList: true,
            subtree: false,
            attributes: true,
            attributeFilter: [
                'slot'
            ]
        });
    }
    handleSlotChange() {
        this.syncSlotState();
    }
    handleMouseEnter() {
        this.hovering = true;
        this.syncAutoCloseTimer();
    }
    handleMouseLeave() {
        this.hovering = false;
        this.syncAutoCloseTimer();
    }
    handleFocusIn() {
        this.focusWithin = true;
        this.syncAutoCloseTimer();
    }
    handleFocusOut(event) {
        if (this.isNotificationNode(event.relatedTarget)) return;
        this.focusWithin = false;
        this.syncAutoCloseTimer();
    }
    handleClose(event) {
        this.close(event);
    }
    createCloseDetail(sourceEvent) {
        const detail = {
            type: this.normalizedType
        };
        if (this.notificationId) detail.id = this.notificationId;
        if (sourceEvent) detail.sourceEvent = sourceEvent;
        return detail;
    }
    dispatchNotificationEvent(type, sourceEvent) {
        return this.dispatchEvent(new CustomEvent(type, {
            detail: this.createCloseDetail(sourceEvent),
            bubbles: true,
            composed: true,
            cancelable: type === VE_NOTIFICATION_CLOSE_EVENT
        }));
    }
    finishClose(sourceEvent) {
        this.phase = 'closed';
        this.style.removeProperty('height');
        this.style.removeProperty('overflow');
        this.style.removeProperty('transition');
        this.dispatchNotificationEvent(VE_NOTIFICATION_AFTER_CLOSE_EVENT, sourceEvent);
    }
    renderDefaultIcon() {
        return html`<ve-icon
      .definition=${DEFAULT_ICON_BY_TYPE[this.normalizedType]}
      size="var(--ve-notification-icon-size)"
    ></ve-icon>`;
    }
    renderIcon() {
        if (!this.showIcon) return nothing;
        return html`
      <span class="ve-notification-icon" part="icon" aria-hidden="true">
        <slot name="icon" @slotchange=${this.handleSlotChange}></slot>
        ${this.hasIconContent ? nothing : this.renderDefaultIcon()}
      </span>
    `;
    }
    renderTitle() {
        if (!this.hasTitleContent) return nothing;
        return html`
      <div class="ve-notification-title" part="title">
        <slot name="title" @slotchange=${this.handleSlotChange}>
          ${this.title}
        </slot>
      </div>
    `;
    }
    renderCloseIcon() {
        if (this.hasCloseIconContent) return html`<slot
        name="close-icon"
        @slotchange=${this.handleSlotChange}
      ></slot>`;
        return html`<ve-icon
      .definition=${closeIcon}
      size="var(--ve-notification-close-icon-size)"
    ></ve-icon>`;
    }
    render() {
        if ('closed' === this.phase) return nothing;
        const notificationClass = classMap({
            've-notification': true,
            [`ve-notification-${this.normalizedType}`]: true,
            [`ve-notification-${this.normalizedPosition}`]: true,
            've-notification-with-actions': this.hasActionsContent,
            've-notification-closable': this.closable,
            've-notification-entered': 'entered' === this.phase,
            've-notification-leaving': 'leaving' === this.phase
        });
        return html`<section
      class=${notificationClass}
      part="notification"
      role=${this.notificationRole}
      aria-live=${this.notificationAriaLive}
      @mouseenter=${this.handleMouseEnter}
      @mouseleave=${this.handleMouseLeave}
      @focusin=${this.handleFocusIn}
      @focusout=${this.handleFocusOut}
    >
      <div class="ve-notification-body">
        ${this.renderIcon()}
        <div class="ve-notification-main">
          ${this.renderTitle()}
          <div class="ve-notification-content" part="content">
            <slot @slotchange=${this.handleSlotChange}>${this.content}</slot>
          </div>
        </div>
      </div>
      ${this.hasActionsContent ? html`<div class="ve-notification-actions" part="actions">
            <slot name="actions" @slotchange=${this.handleSlotChange}></slot>
          </div>` : nothing}
      ${this.closable ? html`<button
            type="button"
            class="ve-notification-close-btn"
            part="close-button"
            aria-label="Close notification"
            @click=${this.handleClose}
          >
            ${this.renderCloseIcon()}
          </button>` : nothing}
    </section>`;
    }
}
defineCustomElement(VE_NOTIFICATION_TAG, VeNotificationElement);
let holders = {};
const notificationOnCloseCallbacks = new WeakMap();
const notificationContentCleanups = new WeakMap();
const notificationRecordByElement = new WeakMap();
function isBrowser() {
    return "u" > typeof document;
}
function isNode(value) {
    return "u" > typeof Node && value instanceof Node;
}
function isContentRenderer(value) {
    return 'function' == typeof value;
}
function isNotificationContent(value) {
    return 'string' == typeof value || 'function' == typeof value || isNode(value);
}
function normalizeInput(type, input) {
    const options = isNotificationContent(input) ? {
        content: input
    } : input;
    return {
        ...options,
        type
    };
}
function generateNotificationId() {
    if ("u" > typeof crypto && 'function' == typeof crypto.randomUUID) return `vedesign_notification_id_${crypto.randomUUID()}`;
    return `vedesign_notification_id_${Math.random().toFixed(10).slice(2)}`;
}
function getContainerElement() {
    if (!isBrowser()) return null;
    return document.body;
}
function applyHolderStyle(holder, position) {
    const isTop = position.startsWith('top');
    const isRight = position.endsWith('right');
    holder.className = `ve-notification-wrapper ve-notification-wrapper-${position}`;
    holder.dataset.veNotificationPosition = position;
    holder.style.position = 'fixed';
    holder.style.zIndex = 'var(--ve-notification-z-index, var(--z-index-notification, 1004))';
    holder.style.display = 'flex';
    holder.style.flexDirection = 'column';
    holder.style.gap = 'var(--ve-notification-wrapper-gap, 16px)';
    holder.style.pointerEvents = 'none';
    holder.style.boxSizing = 'border-box';
    holder.style.maxWidth = 'var(--ve-notification-max-width, calc(100vw - 48px))';
    holder.style.top = isTop ? 'var(--ve-notification-wrapper-offset-block, 40px)' : '';
    holder.style.bottom = isTop ? '' : 'var(--ve-notification-wrapper-offset-block, 40px)';
    holder.style.right = isRight ? 'var(--ve-notification-wrapper-offset-inline, 24px)' : '';
    holder.style.left = isRight ? '' : 'var(--ve-notification-wrapper-offset-inline, 24px)';
    holder.style.alignItems = isRight ? 'flex-end' : 'flex-start';
}
function getHolder(position) {
    const container = getContainerElement();
    if (!container) return null;
    const existing = holders[position];
    if (existing && existing.container === container && existing.holder.isConnected) return existing;
    const holder = document.createElement('div');
    applyHolderStyle(holder, position);
    container.appendChild(holder);
    const record = {
        holder,
        notifications: [],
        container,
        position
    };
    holders[position] = record;
    return record;
}
function getNodeSlotName(node) {
    if (node instanceof Element) return node.getAttribute('slot') ?? '';
    return '';
}
function removeSlottedChildren(element, slotName) {
    for (const child of Array.from(element.childNodes))if (getNodeSlotName(child) === slotName) child.remove();
}
function toSlottedNode(value, slotName) {
    if (!isNode(value)) {
        const span = document.createElement('span');
        span.textContent = value;
        if (slotName) span.slot = slotName;
        return span;
    }
    if (!slotName) {
        if (value instanceof Element) value.removeAttribute('slot');
        return value;
    }
    if (value instanceof Element) {
        value.setAttribute('slot', slotName);
        return value;
    }
    const span = document.createElement('span');
    span.slot = slotName;
    span.appendChild(value);
    return span;
}
function setSlottedContent(element, slotName, value) {
    removeSlottedChildren(element, slotName);
    if (void 0 === value) return;
    element.appendChild(toSlottedNode(value, slotName));
}
function cleanupRenderedContent(element, slotName) {
    const cleanups = notificationContentCleanups.get(element);
    if (!cleanups) return;
    if (void 0 === slotName) {
        for (const cleanup of cleanups.values())cleanup();
        notificationContentCleanups.delete(element);
        return;
    }
    cleanups.get(slotName)?.();
    cleanups.delete(slotName);
    if (0 === cleanups.size) notificationContentCleanups.delete(element);
}
function setRenderedContentCleanup(element, slotName, cleanup) {
    const cleanups = notificationContentCleanups.get(element) ?? new Map();
    cleanups.set(slotName, cleanup);
    notificationContentCleanups.set(element, cleanups);
}
function applyContentToElement(element, slotName, content) {
    cleanupRenderedContent(element, slotName);
    removeSlottedChildren(element, slotName);
    if (void 0 === content) {
        if ('' === slotName) element.content = '';
        else if ('title' === slotName) element.title = '';
        return;
    }
    if ('string' == typeof content) {
        if ('' === slotName) {
            element.content = content;
            return;
        }
        if ('title' === slotName) {
            element.title = content;
            return;
        }
    }
    if ('' === slotName) element.content = '';
    else if ('title' === slotName) element.title = '';
    if (!isContentRenderer(content)) return void element.appendChild(toSlottedNode(content, slotName));
    const container = document.createElement('div');
    if (slotName) container.slot = slotName;
    element.appendChild(container);
    const result = content(container, {
        element,
        close: ()=>element.close()
    });
    if ('function' == typeof result) return void setRenderedContentCleanup(element, slotName, result);
    if ('string' == typeof result) {
        container.textContent = result;
        return;
    }
    if (isNode(result)) container.appendChild(result);
}
function applyOptionsToElement(element, options, fallbackId) {
    element.type = options.type ?? 'info';
    element.position = options.position ?? 'top-right';
    element.notificationId = options.id ?? fallbackId;
    element.duration = options.duration ?? NOTIFICATION_DEFAULT_DURATION;
    element.showIcon = options.showIcon ?? true;
    element.closable = options.closable ?? false;
    element.enterDuration = NOTIFICATION_DEFAULT_ENTER_DURATION;
    element.exitDuration = NOTIFICATION_DEFAULT_EXIT_DURATION;
    applyContentToElement(element, 'title', options.title);
    applyContentToElement(element, '', options.content);
    applyContentToElement(element, 'actions', options.actions);
    setSlottedContent(element, 'icon', options.icon);
    setSlottedContent(element, 'close-icon', options.closeIcon);
}
function cleanupHolder(record) {
    if (record.holder.childElementCount > 0) return;
    record.holder.remove();
    if (holders[record.position] === record) delete holders[record.position];
}
function detachNotificationFromRecord(record, element) {
    record.notifications = record.notifications.filter((notification)=>notification !== element);
}
function removeNotificationFromRecord(record, element) {
    detachNotificationFromRecord(record, element);
    notificationRecordByElement.delete(element);
    cleanupRenderedContent(element);
    element.remove();
    cleanupHolder(record);
}
function findNotificationById(id) {
    for (const record of Object.values(holders)){
        const notification = record.notifications.find((item)=>item.notificationId === id);
        if (notification) return {
            record,
            notification
        };
    }
    return null;
}
function moveNotificationToRecord(element, currentRecord, nextRecord) {
    if (currentRecord === nextRecord) return;
    detachNotificationFromRecord(currentRecord, element);
    nextRecord.notifications.push(element);
    notificationRecordByElement.set(element, nextRecord);
    nextRecord.holder.appendChild(element);
    cleanupHolder(currentRecord);
}
function addInstance(options) {
    const position = options.position ?? 'top-right';
    const record = getHolder(position);
    if (!record) return ()=>void 0;
    const existing = options.id ? findNotificationById(options.id) : null;
    if (existing) {
        moveNotificationToRecord(existing.notification, existing.record, record);
        applyOptionsToElement(existing.notification, options, existing.notification.notificationId);
        notificationOnCloseCallbacks.set(existing.notification, options.onClose);
        existing.notification.restartTimer();
        return ()=>existing.notification.close();
    }
    const element = document.createElement(VE_NOTIFICATION_TAG);
    const id = options.id ?? generateNotificationId();
    applyOptionsToElement(element, options, id);
    notificationOnCloseCallbacks.set(element, options.onClose);
    element.addEventListener(VE_NOTIFICATION_AFTER_CLOSE_EVENT, ()=>{
        const currentRecord = notificationRecordByElement.get(element);
        if (currentRecord) removeNotificationFromRecord(currentRecord, element);
        else element.remove();
        const onClose = notificationOnCloseCallbacks.get(element);
        notificationOnCloseCallbacks.delete(element);
        onClose?.();
    }, {
        once: true
    });
    record.notifications.push(element);
    notificationRecordByElement.set(element, record);
    record.holder.appendChild(element);
    return ()=>element.close();
}
function clearNotifications() {
    for (const record of Object.values(holders))for (const notification of [
        ...record.notifications
    ])notification.close();
}
const Notification = {
    info (config) {
        return addInstance(normalizeInput('info', config));
    },
    success (config) {
        return addInstance(normalizeInput('success', config));
    },
    warning (config) {
        return addInstance(normalizeInput('warning', config));
    },
    error (config) {
        return addInstance(normalizeInput('error', config));
    },
    addInstance (config) {
        return addInstance(config);
    },
    clear () {
        clearNotifications();
    }
};
const ve_notification_ve_notification = Notification;
export default ve_notification_ve_notification;
export { Notification, VE_NOTIFICATION_AFTER_CLOSE_EVENT, VE_NOTIFICATION_CLOSE_EVENT, VE_NOTIFICATION_TAG, VeNotificationElement };
