// @ve-design/web/dist/ve-pagination.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { VE_SELECT_TAG, VE_SELECT_ITEM_TAG, VeSelectItemElement, VeSelectElement } from "./30129.js";
import { moreHorizontalIcon } from "./icons/more-horizontal.js";
import { chevronLeftMdIcon } from "./icons/chevron-left-md.js";
import { chevronRightMdIcon } from "./icons/chevron-right-md.js";
const vePaginationTokenDefaults = css`
  :host {
    --ve-pagination-gap: var(--space-xxs);
    --ve-pagination-page-gap: var(--space-xxxs);
    --ve-pagination-item-size: var(--height-small);
    --ve-pagination-item-min-inline-size: var(--height-small);
    --ve-pagination-item-padding-inline: calc(var(--space-xxs) + 2px);
    --ve-pagination-item-radius: var(--radius-sm);
    --ve-pagination-font-family: var(--font-sans);
    --ve-pagination-font-size: var(--text-body);
    --ve-pagination-font-weight: var(--font-weight-normal);
    --ve-pagination-line-height: 22px;
    --ve-pagination-total-color: var(--color-text-secondary);
    --ve-pagination-text: var(--color-text-secondary);
    --ve-pagination-text-current: var(--color-text-primary);
    --ve-pagination-text-disabled: var(--color-text-disable);
    --ve-pagination-icon-color: var(--color-icon-primary);
    --ve-pagination-background: transparent;
    --ve-pagination-background-hover: var(--color-bg-surface);
    --ve-pagination-background-pressed: color-mix(
      var(--ve-pagination-background-hover) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );
    --ve-pagination-background-current: var(--color-bg-muted);
    --ve-pagination-select-arrow-color: var(--color-icon-disable);
    --ve-pagination-focus-ring: var(--color-border-interactive);
    --ve-pagination-disabled-opacity: var(--state-disabled-ratio);

    --ve-pagination-size-width: fit-content;
    --ve-pagination-size-min-width: 0px;
    --ve-pagination-size-height: var(--height-small);
    --ve-pagination-transition-duration: var(--motion-fast);
    --ve-pagination-transition-timing: var(--ease-standard);
  }
`;
const token = vePaginationTokenDefaults;
const ve_pagination = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      max-inline-size: 100%;
      vertical-align: middle;
      box-sizing: border-box;
    }

    :host([hidden]) {
      display: none;
    }

    :host([disabled]) {
      cursor: not-allowed;
      color: var(--ve-pagination-text-disabled);
    }

    .ve-pagination {
      display: inline-flex;
      align-items: center;
      max-inline-size: 100%;
      gap: var(--ve-pagination-gap);
      box-sizing: border-box;
      color: var(--ve-pagination-text);
      font-family: var(--ve-pagination-font-family);
      font-size: var(--ve-pagination-font-size);
      font-weight: var(--ve-pagination-font-weight);
      line-height: var(--ve-pagination-line-height);
    }

    .ve-pagination-total {
      flex: none;
      color: var(--ve-pagination-total-color);
      white-space: nowrap;
    }

    :host([disabled]) .ve-pagination-total {
      color: var(--ve-pagination-text-disabled);
      opacity: var(--ve-pagination-disabled-opacity);
    }

    .ve-pagination-pages {
      display: inline-flex;
      align-items: center;
      flex: none;
      gap: var(--ve-pagination-page-gap);
    }

    .ve-pagination-item {
      appearance: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: 0 0 auto;
      min-inline-size: var(--ve-pagination-item-min-inline-size);
      block-size: var(--ve-pagination-item-size);
      box-sizing: border-box;
      margin: 0;
      border: none;
      border-radius: var(--ve-pagination-item-radius);
      padding: 0 var(--ve-pagination-item-padding-inline);
      background: var(--ve-pagination-background);
      color: var(--ve-pagination-text);
      font: inherit;
      font-variant-numeric: tabular-nums;
      line-height: var(--ve-pagination-line-height);
      text-align: center;
      white-space: nowrap;
      cursor: pointer;
      user-select: none;
      -webkit-tap-highlight-color: transparent;
      transition:
        background-color var(--ve-pagination-transition-duration)
          var(--ve-pagination-transition-timing),
        color var(--ve-pagination-transition-duration)
          var(--ve-pagination-transition-timing),
        opacity var(--ve-pagination-transition-duration)
          var(--ve-pagination-transition-timing),
        box-shadow var(--ve-pagination-transition-duration)
          var(--ve-pagination-transition-timing);
    }

    .ve-pagination-item-icon,
    .ve-pagination-item-ellipsis {
      inline-size: var(--ve-pagination-item-size);
      min-inline-size: var(--ve-pagination-item-size);
      padding: 0;
      color: var(--ve-pagination-icon-color);
    }

    .ve-pagination-item ve-icon {
      display: inline-flex;
      flex: none;
    }

    .ve-pagination-item:not(:disabled):hover {
      background: var(--ve-pagination-background-hover);
      color: var(--ve-pagination-text-current);
    }

    .ve-pagination-item:not(:disabled):active {
      background: var(--ve-pagination-background-pressed);
    }

    .ve-pagination-item:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--stroke-weight-bold)
        var(--ve-pagination-focus-ring);
    }

    .ve-pagination-item-current {
      background: var(--ve-pagination-background-current);
      color: var(--ve-pagination-text-current);
    }

    .ve-pagination-item:disabled {
      cursor: not-allowed;
      opacity: var(--ve-pagination-disabled-opacity);
      box-shadow: none;
    }

    .ve-pagination-item-ellipsis:disabled {
      opacity: 1;
      cursor: default;
    }

    :host([disabled]) .ve-pagination-item-current,
    :host([disabled]) .ve-pagination-item-ellipsis {
      color: var(--ve-pagination-text-disabled);
    }

    :host([disabled]) .ve-pagination-item-ellipsis:disabled {
      opacity: var(--ve-pagination-disabled-opacity);
    }

    .ve-pagination-size {
      display: inline-flex;
      align-items: center;
      flex: none;
      inline-size: var(--ve-pagination-size-width);
      min-inline-size: var(--ve-pagination-size-min-width);
      block-size: var(--ve-pagination-size-height);
      box-sizing: border-box;
    }

    .ve-pagination-size-select {
      inline-size: fit-content;
      --ve-select-arrow-color: var(--ve-pagination-select-arrow-color);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-pagination-item {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _init_current, _init_extra__init_current, _init_defaultCurrent, _init_extra__init_defaultCurrent, _init_pageSize, _init_extra__init_pageSize, _init_defaultPageSize, _init_extra__init_defaultPageSize, _init_total, _init_extra__init_total, _init_pageSizeOptions, _init_extra__init_pageSizeOptions, _init_showTotal, _init_extra__init_showTotal, _init_showPageSize, _init_extra__init_showPageSize, _init_totalTemplate, _init_extra__init_totalTemplate, _init_bufferSize, _init_extra__init_bufferSize, _init_disabled, _init_extra__init_disabled, _init_currentButtonElement, _init_extra__init_currentButtonElement, _init_firstEnabledButtonElement, _init_extra__init_firstEnabledButtonElement, _init_innerCurrent, _init_extra__init_innerCurrent, _init_innerPageSize, _init_extra__init_innerPageSize;
const VE_PAGINATION_TAG = 've-pagination';
const DEFAULT_CURRENT = 1;
const DEFAULT_PAGE_SIZE = 10;
const DEFAULT_PAGE_SIZE_OPTIONS = [
    10,
    20,
    50,
    100
];
const DEFAULT_BUFFER_SIZE = 2;
const DEFAULT_TOTAL_TEMPLATE = '共 {total} 项';
const OPTIONAL_NUMBER_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return;
        const numeric = Number(value);
        return Number.isFinite(numeric) ? numeric : void 0;
    },
    toAttribute (value) {
        return void 0 === value ? null : String(value);
    }
};
const NUMBER_ARRAY_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return [
            ...DEFAULT_PAGE_SIZE_OPTIONS
        ];
        const trimmed = value.trim();
        if (!trimmed) return [];
        if (trimmed.startsWith('[')) try {
            const parsed = JSON.parse(trimmed);
            if (Array.isArray(parsed)) return parsed.filter(isPositiveNumber).map((item)=>Math.floor(item));
        } catch  {
            return [];
        }
        return trimmed.split(',').map((item)=>Number(item.trim())).filter(isPositiveNumber).map((item)=>Math.floor(item));
    },
    toAttribute (value) {
        return value.length > 0 ? value.join(',') : null;
    }
};
function isPositiveNumber(value) {
    return 'number' == typeof value && Number.isFinite(value) && value > 0;
}
function toPositiveInteger(value, fallback) {
    return isPositiveNumber(value) ? Math.floor(value) : fallback;
}
function clamp(value, min, max) {
    return Math.min(Math.max(value, min), max);
}
function uniquePositiveIntegers(values) {
    const seen = new Set();
    const result = [];
    for (const value of values){
        if (!isPositiveNumber(value)) continue;
        const integer = Math.floor(value);
        if (!seen.has(integer)) {
            seen.add(integer);
            result.push(integer);
        }
    }
    return result;
}
function createPaginationItems(current, totalPages, bufferSize) {
    if (totalPages <= 1) return [
        {
            type: 'page',
            value: 1
        }
    ];
    const normalizedBuffer = Math.max(0, Math.floor(bufferSize));
    const allVisibleLimit = 2 * normalizedBuffer + 5;
    if (totalPages <= allVisibleLimit) return Array.from({
        length: totalPages
    }, (_, index)=>({
            type: 'page',
            value: index + 1
        }));
    let left = Math.max(2, current - normalizedBuffer);
    let right = Math.min(totalPages - 1, current + normalizedBuffer);
    if (current <= normalizedBuffer + 2) {
        left = 2;
        right = Math.min(totalPages - 1, 2 * normalizedBuffer + 1);
    }
    if (current >= totalPages - normalizedBuffer - 1) {
        left = Math.max(2, totalPages - 2 * normalizedBuffer);
        right = totalPages - 1;
    }
    const items = [
        {
            type: 'page',
            value: 1
        }
    ];
    if (left > 2) items.push({
        type: 'ellipsis',
        key: 'start-ellipsis'
    });
    for(let page = left; page <= right; page += 1)items.push({
        type: 'page',
        value: page
    });
    if (right < totalPages - 1) items.push({
        type: 'ellipsis',
        key: 'end-ellipsis'
    });
    items.push({
        type: 'page',
        value: totalPages
    });
    return items;
}
_dec = property({
    converter: OPTIONAL_NUMBER_CONVERTER
}), _dec1 = property({
    type: Number,
    attribute: 'default-current'
}), _dec2 = property({
    converter: OPTIONAL_NUMBER_CONVERTER,
    attribute: 'page-size'
}), _dec3 = property({
    type: Number,
    attribute: 'default-page-size'
}), _dec4 = property({
    type: Number
}), _dec5 = property({
    converter: NUMBER_ARRAY_CONVERTER,
    attribute: 'page-size-options'
}), _dec6 = property({
    type: Boolean,
    attribute: 'show-total',
    reflect: true
}), _dec7 = property({
    type: Boolean,
    attribute: 'show-page-size',
    reflect: true
}), _dec8 = property({
    type: String,
    attribute: 'total-template'
}), _dec9 = property({
    type: Number,
    attribute: 'buffer-size'
}), _dec10 = property({
    type: Boolean,
    reflect: true
}), _dec11 = query('.ve-pagination-item-current'), _dec12 = query('button:not(:disabled)'), _dec13 = state(), _dec14 = state();
class VePaginationElement extends VeElement {
    static{
        ({ e: [_init_current, _init_extra__init_current, _init_defaultCurrent, _init_extra__init_defaultCurrent, _init_pageSize, _init_extra__init_pageSize, _init_defaultPageSize, _init_extra__init_defaultPageSize, _init_total, _init_extra__init_total, _init_pageSizeOptions, _init_extra__init_pageSizeOptions, _init_showTotal, _init_extra__init_showTotal, _init_showPageSize, _init_extra__init_showPageSize, _init_totalTemplate, _init_extra__init_totalTemplate, _init_bufferSize, _init_extra__init_bufferSize, _init_disabled, _init_extra__init_disabled, _init_currentButtonElement, _init_extra__init_currentButtonElement, _init_firstEnabledButtonElement, _init_extra__init_firstEnabledButtonElement, _init_innerCurrent, _init_extra__init_innerCurrent, _init_innerPageSize, _init_extra__init_innerPageSize] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "current"
            ],
            [
                _dec1,
                1,
                "defaultCurrent"
            ],
            [
                _dec2,
                1,
                "pageSize"
            ],
            [
                _dec3,
                1,
                "defaultPageSize"
            ],
            [
                _dec4,
                1,
                "total"
            ],
            [
                _dec5,
                1,
                "pageSizeOptions"
            ],
            [
                _dec6,
                1,
                "showTotal"
            ],
            [
                _dec7,
                1,
                "showPageSize"
            ],
            [
                _dec8,
                1,
                "totalTemplate"
            ],
            [
                _dec9,
                1,
                "bufferSize"
            ],
            [
                _dec10,
                1,
                "disabled"
            ],
            [
                _dec11,
                1,
                "currentButtonElement"
            ],
            [
                _dec12,
                1,
                "firstEnabledButtonElement"
            ],
            [
                _dec13,
                1,
                "innerCurrent"
            ],
            [
                _dec14,
                1,
                "innerPageSize"
            ]
        ]));
    }
    static styles = ve_pagination;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement,
        [VE_SELECT_TAG]: VeSelectElement,
        [VE_SELECT_ITEM_TAG]: VeSelectItemElement
    };
    #___private_current_1 = (()=>{
        const _value = _init_current(this, void 0);
        _init_extra__init_current(this);
        return _value;
    })();
    get current() {
        return this.#___private_current_1;
    }
    set current(_v) {
        this.#___private_current_1 = _v;
    }
    #___private_defaultCurrent_2 = (()=>{
        const _value = _init_defaultCurrent(this, DEFAULT_CURRENT);
        _init_extra__init_defaultCurrent(this);
        return _value;
    })();
    get defaultCurrent() {
        return this.#___private_defaultCurrent_2;
    }
    set defaultCurrent(_v) {
        this.#___private_defaultCurrent_2 = _v;
    }
    #___private_pageSize_3 = (()=>{
        const _value = _init_pageSize(this, void 0);
        _init_extra__init_pageSize(this);
        return _value;
    })();
    get pageSize() {
        return this.#___private_pageSize_3;
    }
    set pageSize(_v) {
        this.#___private_pageSize_3 = _v;
    }
    #___private_defaultPageSize_4 = (()=>{
        const _value = _init_defaultPageSize(this, DEFAULT_PAGE_SIZE);
        _init_extra__init_defaultPageSize(this);
        return _value;
    })();
    get defaultPageSize() {
        return this.#___private_defaultPageSize_4;
    }
    set defaultPageSize(_v) {
        this.#___private_defaultPageSize_4 = _v;
    }
    #___private_total_5 = (()=>{
        const _value = _init_total(this, 0);
        _init_extra__init_total(this);
        return _value;
    })();
    get total() {
        return this.#___private_total_5;
    }
    set total(_v) {
        this.#___private_total_5 = _v;
    }
    #___private_pageSizeOptions_6 = (()=>{
        const _value = _init_pageSizeOptions(this, [
            ...DEFAULT_PAGE_SIZE_OPTIONS
        ]);
        _init_extra__init_pageSizeOptions(this);
        return _value;
    })();
    get pageSizeOptions() {
        return this.#___private_pageSizeOptions_6;
    }
    set pageSizeOptions(_v) {
        this.#___private_pageSizeOptions_6 = _v;
    }
    #___private_showTotal_7 = (()=>{
        const _value = _init_showTotal(this, false);
        _init_extra__init_showTotal(this);
        return _value;
    })();
    get showTotal() {
        return this.#___private_showTotal_7;
    }
    set showTotal(_v) {
        this.#___private_showTotal_7 = _v;
    }
    #___private_showPageSize_8 = (()=>{
        const _value = _init_showPageSize(this, false);
        _init_extra__init_showPageSize(this);
        return _value;
    })();
    get showPageSize() {
        return this.#___private_showPageSize_8;
    }
    set showPageSize(_v) {
        this.#___private_showPageSize_8 = _v;
    }
    #___private_totalTemplate_9 = (()=>{
        const _value = _init_totalTemplate(this, DEFAULT_TOTAL_TEMPLATE);
        _init_extra__init_totalTemplate(this);
        return _value;
    })();
    get totalTemplate() {
        return this.#___private_totalTemplate_9;
    }
    set totalTemplate(_v) {
        this.#___private_totalTemplate_9 = _v;
    }
    #___private_bufferSize_10 = (()=>{
        const _value = _init_bufferSize(this, DEFAULT_BUFFER_SIZE);
        _init_extra__init_bufferSize(this);
        return _value;
    })();
    get bufferSize() {
        return this.#___private_bufferSize_10;
    }
    set bufferSize(_v) {
        this.#___private_bufferSize_10 = _v;
    }
    #___private_disabled_11 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_11;
    }
    set disabled(_v) {
        this.#___private_disabled_11 = _v;
    }
    #___private_currentButtonElement_12 = (()=>{
        const _value = _init_currentButtonElement(this, null);
        _init_extra__init_currentButtonElement(this);
        return _value;
    })();
    get currentButtonElement() {
        return this.#___private_currentButtonElement_12;
    }
    set currentButtonElement(_v) {
        this.#___private_currentButtonElement_12 = _v;
    }
    #___private_firstEnabledButtonElement_13 = (()=>{
        const _value = _init_firstEnabledButtonElement(this, null);
        _init_extra__init_firstEnabledButtonElement(this);
        return _value;
    })();
    get firstEnabledButtonElement() {
        return this.#___private_firstEnabledButtonElement_13;
    }
    set firstEnabledButtonElement(_v) {
        this.#___private_firstEnabledButtonElement_13 = _v;
    }
    #___private_innerCurrent_14 = (()=>{
        const _value = _init_innerCurrent(this, DEFAULT_CURRENT);
        _init_extra__init_innerCurrent(this);
        return _value;
    })();
    get innerCurrent() {
        return this.#___private_innerCurrent_14;
    }
    set innerCurrent(_v) {
        this.#___private_innerCurrent_14 = _v;
    }
    #___private_innerPageSize_15 = (()=>{
        const _value = _init_innerPageSize(this, DEFAULT_PAGE_SIZE);
        _init_extra__init_innerPageSize(this);
        return _value;
    })();
    get innerPageSize() {
        return this.#___private_innerPageSize_15;
    }
    set innerPageSize(_v) {
        this.#___private_innerPageSize_15 = _v;
    }
    hasInitializedDefaults = false;
    get isCurrentControlled() {
        return void 0 !== this.current;
    }
    get isPageSizeControlled() {
        return void 0 !== this.pageSize;
    }
    get normalizedTotal() {
        return Math.max(0, Math.floor(Number.isFinite(this.total) ? this.total : 0));
    }
    get resolvedPageSize() {
        const source = this.isPageSizeControlled ? this.pageSize : this.innerPageSize;
        return toPositiveInteger(source ?? DEFAULT_PAGE_SIZE, DEFAULT_PAGE_SIZE);
    }
    get totalPages() {
        return Math.max(1, Math.ceil(this.normalizedTotal / this.resolvedPageSize));
    }
    get resolvedCurrent() {
        const source = this.isCurrentControlled ? this.current : this.innerCurrent;
        return clamp(toPositiveInteger(source ?? DEFAULT_CURRENT, DEFAULT_CURRENT), 1, this.totalPages);
    }
    get normalizedPageSizeOptions() {
        const options = uniquePositiveIntegers(this.pageSizeOptions);
        const currentPageSize = this.resolvedPageSize;
        return options.includes(currentPageSize) ? options : [
            currentPageSize,
            ...options
        ];
    }
    get totalText() {
        return this.totalTemplate.replace('{total}', String(this.normalizedTotal));
    }
    focus(options) {
        (this.currentButtonElement ?? this.firstEnabledButtonElement)?.focus(options);
    }
    connectedCallback() {
        super.connectedCallback();
        if (!this.hasInitializedDefaults) {
            this.hasInitializedDefaults = true;
            this.innerCurrent = toPositiveInteger(this.defaultCurrent, DEFAULT_CURRENT);
            this.innerPageSize = toPositiveInteger(this.defaultPageSize, DEFAULT_PAGE_SIZE);
        }
    }
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (!this.isCurrentControlled && (changedProperties.has('total') || changedProperties.has('pageSize') || changedProperties.has('innerPageSize'))) this.innerCurrent = clamp(this.innerCurrent, 1, this.totalPages);
    }
    addEventListener(type, listener, options) {
        super.addEventListener(type, listener, options);
    }
    removeEventListener(type, listener, options) {
        super.removeEventListener(type, listener, options);
    }
    emitPageChange(nextPage, source) {
        if (this.disabled) return;
        const previousCurrent = this.resolvedCurrent;
        const current = clamp(nextPage, 1, this.totalPages);
        if (current === previousCurrent) return;
        if (!this.isCurrentControlled) this.innerCurrent = current;
        this.dispatchEvent(new CustomEvent('ve-change', {
            bubbles: true,
            composed: true,
            detail: {
                current,
                previousCurrent,
                pageSize: this.resolvedPageSize,
                total: this.normalizedTotal,
                totalPages: this.totalPages,
                source
            }
        }));
    }
    handlePageClick(page) {
        this.emitPageChange(page, 'page');
    }
    handlePreviousClick() {
        this.emitPageChange(this.resolvedCurrent - 1, 'prev');
    }
    handleNextClick() {
        this.emitPageChange(this.resolvedCurrent + 1, 'next');
    }
    handlePageSizeChange(event) {
        event.stopPropagation();
        if (this.disabled) return;
        const value = event.detail.value;
        if (Array.isArray(value) || '' === value) return;
        const nextPageSize = toPositiveInteger(Number(value), this.resolvedPageSize);
        const previousPageSize = this.resolvedPageSize;
        if (nextPageSize === previousPageSize) return;
        const previousCurrent = this.resolvedCurrent;
        const totalPages = Math.max(1, Math.ceil(this.normalizedTotal / nextPageSize));
        const current = clamp(previousCurrent, 1, totalPages);
        if (!this.isPageSizeControlled) this.innerPageSize = nextPageSize;
        if (!this.isCurrentControlled) this.innerCurrent = current;
        this.dispatchEvent(new CustomEvent('ve-page-size-change', {
            bubbles: true,
            composed: true,
            detail: {
                pageSize: nextPageSize,
                previousPageSize,
                current,
                previousCurrent,
                total: this.normalizedTotal,
                totalPages
            }
        }));
    }
    renderTotal() {
        if (!this.showTotal) return nothing;
        return html`
      <span class="ve-pagination-total" part="total">${this.totalText}</span>
    `;
    }
    renderIconButton(action, icon, disabled, label, onClick) {
        return html`
      <button
        class="ve-pagination-item ve-pagination-item-icon"
        part="item"
        type="button"
        data-pagination-action=${action}
        aria-label=${label}
        ?disabled=${this.disabled || disabled}
        @click=${onClick}
      >
        <ve-icon .definition=${icon} size="16"></ve-icon>
      </button>
    `;
    }
    renderPageItem(item) {
        if ('ellipsis' === item.type) return html`
        <button
          class="ve-pagination-item ve-pagination-item-ellipsis"
          part="item"
          type="button"
          tabindex="-1"
          aria-label="More pages"
          disabled
        >
          <ve-icon .definition=${moreHorizontalIcon} size="16"></ve-icon>
        </button>
      `;
        const page = item.value;
        const active = page === this.resolvedCurrent;
        const classes = {
            've-pagination-item': true,
            've-pagination-item-page': true,
            've-pagination-item-current': active
        };
        return html`
      <button
        class=${classMap(classes)}
        part="item"
        type="button"
        aria-label=${`Page ${page}`}
        aria-current=${ifDefined(active ? 'page' : void 0)}
        ?disabled=${this.disabled}
        @click=${()=>this.handlePageClick(page)}
      >
        ${page}
      </button>
    `;
    }
    renderPages() {
        const items = createPaginationItems(this.resolvedCurrent, this.totalPages, this.bufferSize);
        return html`
      <div class="ve-pagination-pages" part="pages">
        ${this.renderIconButton('prev', chevronLeftMdIcon, this.resolvedCurrent <= 1, 'Previous page', ()=>this.handlePreviousClick())}
        ${items.map((item)=>this.renderPageItem(item))}
        ${this.renderIconButton('next', chevronRightMdIcon, this.resolvedCurrent >= this.totalPages, 'Next page', ()=>this.handleNextClick())}
      </div>
    `;
    }
    renderPageSizeSelect() {
        if (!this.showPageSize) return nothing;
        return html`
      <span class="ve-pagination-size" part="select">
        <ve-select
          class="ve-pagination-size-select"
          aria-label="Page size"
          size="small"
          popup-match-select-width="false"
          .value=${String(this.resolvedPageSize)}
          ?disabled=${this.disabled}
          @ve-change=${this.handlePageSizeChange}
        >
          ${this.normalizedPageSizeOptions.map((size)=>html`
              <ve-select-item .value=${String(size)}>
                ${size}条/页
              </ve-select-item>
            `)}
        </ve-select>
      </span>
    `;
    }
    render() {
        return html`
      <nav class="ve-pagination" part="root" aria-label="Pagination">
        ${this.renderTotal()} ${this.renderPages()}
        ${this.renderPageSizeSelect()}
      </nav>
    `;
    }
}
defineCustomElement(VE_PAGINATION_TAG, VePaginationElement);
export { VE_PAGINATION_TAG, VePaginationElement };
