// @ve-design/web/dist/ve-prompt-item.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
import { arrowRightMdIcon } from "./icons/arrow-right-md.js";
const vePromptItemTokenDefaults = css`
  :host {
    /* === Surface === */
    --ve-prompt-item-background: var(--color-bg-base, #ffffff);
    --ve-prompt-item-hover-background: var(--color-bg-surface, #f6f6fc);
    --ve-prompt-item-muted-background: var(--color-bg-muted, #ebebf5);
    --ve-prompt-item-border-color: var(--color-border-default, #d2d1e0);
    --ve-prompt-item-border-width: var(--stroke-weight-base, 0.5px);

    /* === Text === */
    --ve-prompt-item-title-color: var(--color-text-primary, #1f1f23);
    --ve-prompt-item-title-secondary-color: var(--color-text-secondary, #383740);
    --ve-prompt-item-description-color: var(--color-text-tertiary, #7a7880);
    --ve-prompt-item-icon-color: var(--color-bg-icon, #383740);

    /* === Card-grid (type=card) === */
    --ve-prompt-item-card-min-height: 4.625rem; /* 74px */
    --ve-prompt-item-card-padding: var(--space-s, 8px);
    --ve-prompt-item-card-radius: var(--radius-lg, 12px);
    --ve-prompt-item-card-gap: var(--space-xxs, 4px);

    /* === Line (text/line)：合并原 list / compact，宽度由外部控制 === */
    --ve-prompt-item-line-min-height: 2.375rem; /* 38px */
    --ve-prompt-item-line-padding-block: var(--space-xs, 6px);
    --ve-prompt-item-line-padding-inline: var(--space-m, 12px);
    --ve-prompt-item-line-radius: var(--radius-lg, 12px);
    --ve-prompt-item-line-gap: var(--space-xxs, 4px);

    /* === Media card (card/media)：默认 3:2，1:1 等比例由外部 CSS 变量覆盖 === */
    --ve-prompt-item-media-width: 15.625rem; /* 250px */
    --ve-prompt-item-media-height: 10.375rem; /* 166px */
    --ve-prompt-item-media-radius: var(--radius-2xl, 20px);
    --ve-prompt-item-media-mask-radius: var(--radius-2xl, 20px);
    --ve-prompt-item-media-mask-padding: var(--space-xs, 6px);
    --ve-prompt-item-media-foreground-color: var(--color-text-foreground, #ffffff);

    /* === Title typography === */
    --ve-prompt-item-title-font-size: 0.875rem; /* 14px */
    --ve-prompt-item-title-line-height: 1.375rem; /* 22px */
    --ve-prompt-item-title-weight: var(--font-weight-normal, 400);

    --ve-prompt-item-description-font-size: 0.8125rem; /* 13px */
    --ve-prompt-item-description-line-height: 1.5;

    --ve-prompt-item-mask-copy-font-size: 0.75rem; /* 12px */
    --ve-prompt-item-mask-copy-line-height: 1.25rem; /* 20px */
    --ve-prompt-item-mask-button-font-size: 0.75rem; /* 12px */
    --ve-prompt-item-mask-button-line-height: 1.25rem; /* 20px */
    --ve-prompt-item-mask-button-min-height: 1.75rem; /* 28px */
    --ve-prompt-item-mask-button-radius: var(--radius-md, 8px);
    --ve-prompt-item-mask-button-background: rgb(255 255 255 / 20%);

    /* === Disabled === */
    --ve-prompt-item-disabled-opacity: var(--state-disabled-ratio, 0.4);
  }
`;
const token = vePromptItemTokenDefaults;
const ve_prompt_item = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: top;
      box-sizing: border-box;
    }

    :host([disabled]) {
      cursor: not-allowed;
    }

    :host([disabled]) .ve-prompt-item {
      opacity: var(--ve-prompt-item-disabled-opacity);
      pointer-events: none;
    }

    /* ===========================================================
     * Root container — 默认为可点击容器
     * =========================================================== */
    .ve-prompt-item {
      display: inline-flex;
      box-sizing: border-box;
      max-inline-size: 100%;
      margin: 0;
      padding: 0;
      border: 0;
      background: transparent;
      color: inherit;
      font: inherit;
      text-align: start;
      cursor: pointer;
      outline: none;
      transition: background-color 150ms ease;
    }

    .ve-prompt-item:focus-visible {
      outline: 2px solid var(--color-border-interactive, #2f74ff);
      outline-offset: 2px;
    }

    /* ===========================================================
     * Type = card / line —— 文字形态共享基础
     * =========================================================== */
    :host([type='card']) .ve-prompt-item,
    :host([type='line']) .ve-prompt-item {
      align-items: center;
      gap: var(--space-xxs, 4px);
      background: var(--ve-prompt-item-background);
    }

    /* type = card → 弹性宽度，title + description + arrow */
    :host([type='card']) .ve-prompt-item {
      align-items: flex-end;
      justify-content: flex-end;
      inline-size: 100%;
      min-block-size: var(--ve-prompt-item-card-min-height);
      padding: var(--ve-prompt-item-card-padding);
      border: var(--ve-prompt-item-border-width) solid
        var(--ve-prompt-item-border-color);
      border-radius: var(--ve-prompt-item-card-radius);
      gap: var(--ve-prompt-item-card-gap);
    }

    :host([type='card']) .ve-prompt-item:hover:not([aria-disabled='true']) {
      background: var(--ve-prompt-item-hover-background);
    }

    /* type = line → 弹性宽度，单行 title */
    :host([type='line']) .ve-prompt-item {
      inline-size: 100%;
      min-block-size: var(--ve-prompt-item-line-min-height);
      padding-block: var(--ve-prompt-item-line-padding-block);
      padding-inline: var(--ve-prompt-item-line-padding-inline);
      border-radius: var(--ve-prompt-item-line-radius);
      gap: var(--ve-prompt-item-line-gap);
    }

    :host([type='line']) .ve-prompt-item:hover:not([aria-disabled='true']) {
      background: var(--ve-prompt-item-hover-background);
    }

    /* ===========================================================
     * Text body — copy 区（title + description）
     * =========================================================== */
    .ve-prompt-item-copy {
      display: flex;
      flex: 1 1 0;
      flex-direction: column;
      min-inline-size: 0;
      gap: var(--space-xxxs, 2px);
      overflow: hidden;
    }

    .ve-prompt-item-title-row {
      display: flex;
      align-items: center;
      inline-size: 100%;
      min-inline-size: 0;
      gap: var(--space-xxxs, 2px);
    }

    .ve-prompt-item-title {
      flex: 1 1 0;
      min-inline-size: 0;
      margin: 0;
      color: var(--ve-prompt-item-title-color);
      font-size: var(--ve-prompt-item-title-font-size);
      font-weight: var(--ve-prompt-item-title-weight);
      line-height: var(--ve-prompt-item-title-line-height);
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    /* line 模式下标题色与 secondary 行保持一致 */
    :host([type='line']) .ve-prompt-item-title {
      color: var(--ve-prompt-item-title-secondary-color);
    }

    .ve-prompt-item-description {
      margin: 0;
      inline-size: 100%;
      color: var(--ve-prompt-item-description-color);
      font-size: var(--ve-prompt-item-description-font-size);
      line-height: var(--ve-prompt-item-description-line-height);
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    /* ===========================================================
     * Slot — prefix / suffix / arrow
     * =========================================================== */
    .ve-prompt-item-prefix,
    .ve-prompt-item-suffix {
      display: inline-flex;
      flex: 0 0 auto;
      align-items: center;
      justify-content: center;
      color: var(--ve-prompt-item-icon-color);
    }

    .ve-prompt-item-arrow {
      flex: 0 0 auto;
      inline-size: 1rem;
      block-size: 1rem;
      color: currentColor;
    }

    /* line 下箭头略小，匹配紧凑节奏 */
    :host([type='line']) .ve-prompt-item-arrow {
      inline-size: 0.875rem;
      block-size: 0.875rem;
    }

    /* ===========================================================
     * Type = media —— 媒体卡片
     * =========================================================== */
    :host([type='media']) .ve-prompt-item {
      position: relative;
      align-items: flex-end;
      justify-content: center;
      overflow: hidden;
      inline-size: var(--ve-prompt-item-media-width);
      block-size: var(--ve-prompt-item-media-height);
      border-radius: var(--ve-prompt-item-media-radius);
      background: var(--ve-prompt-item-hover-background);
    }

    .ve-prompt-item-media {
      position: absolute;
      inset: 0;
      display: block;
      inline-size: 100%;
      block-size: 100%;
      object-fit: cover;
    }

    /* ===========================================================
     * Mask —— hover 时浮现，不 hover 时隐藏
     * =========================================================== */
    .ve-prompt-item-mask {
      position: absolute;
      right: 0;
      bottom: 0;
      left: 0;
      z-index: 1;
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      gap: var(--space-xxs, 4px);
      inline-size: 100%;
      padding: var(--ve-prompt-item-media-mask-padding);
      border-radius: var(--radius-xs, 2px) var(--radius-xs, 2px)
        var(--ve-prompt-item-media-mask-radius)
        var(--ve-prompt-item-media-mask-radius);
      background: linear-gradient(
        180deg,
        rgba(0, 0, 0, 0) 0%,
        rgba(0, 0, 0, 0.5) 100%
      );
      color: var(--ve-prompt-item-media-foreground-color);
      text-shadow: 0 4px 10px rgb(0 0 0 / 10%);
      backdrop-filter: blur(28px);
      -webkit-backdrop-filter: blur(28px);
      opacity: 0;
      pointer-events: none;
      overflow: hidden;
      transition: opacity 200ms ease;
    }

    :host([type='media']) .ve-prompt-item:hover .ve-prompt-item-mask,
    :host([type='media'][force-hover]) .ve-prompt-item-mask {
      opacity: 1;
    }

    .ve-prompt-item-mask-title {
      display: -webkit-box;
      inline-size: 100%;
      margin: 0;
      color: var(--ve-prompt-item-media-foreground-color);
      font-family: var(--font-mono, monospace);
      font-size: var(--ve-prompt-item-mask-copy-font-size);
      line-height: var(--ve-prompt-item-mask-copy-line-height);
      white-space: normal;
      word-break: break-word;
      overflow: hidden;
      text-overflow: ellipsis;
      -webkit-box-orient: vertical;
      -webkit-line-clamp: 2;
    }

    .ve-prompt-item-mask-button {
      display: flex;
      align-items: center;
      justify-content: center;
      inline-size: 100%;
      min-block-size: var(--ve-prompt-item-mask-button-min-height);
      padding-block: var(--space-xxxs, 2px);
      padding-inline: var(--space-xs, 6px);
      border-radius: var(--ve-prompt-item-mask-button-radius);
      background: var(--ve-prompt-item-mask-button-background);
      color: var(--ve-prompt-item-media-foreground-color);
      font-size: var(--ve-prompt-item-mask-button-font-size);
      font-weight: var(--font-weight-medium, 500);
      line-height: var(--ve-prompt-item-mask-button-line-height);
      white-space: nowrap;
      overflow: hidden;
      text-overflow: ellipsis;
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _init_type, _init_extra__init_type, _init_disabled, _init_extra__init_disabled, _init_arrow, _init_extra__init_arrow, _init_forceHover, _init_extra__init_forceHover, _init_description, _init_extra__init_description, _init_image, _init_extra__init_image, _init_alt, _init_extra__init_alt, _init_titleSlot, _init_extra__init_titleSlot, _init_descriptionSlot, _init_extra__init_descriptionSlot, _init_mediaSlot, _init_extra__init_mediaSlot, _init_maskSlot, _init_extra__init_maskSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasDescriptionSlot, _init_extra__init_hasDescriptionSlot, _init_hasMediaSlot, _init_extra__init_hasMediaSlot, _init_hasMaskSlot, _init_extra__init_hasMaskSlot;
const VE_PROMPT_ITEM_TAG = 've-prompt-item';
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true,
    attribute: 'force-hover'
}), _dec4 = property({
    type: String
}), _dec5 = property({
    type: String
}), _dec6 = property({
    type: String
}), _dec7 = query('slot[name="title"]'), _dec8 = query('slot[name="description"]'), _dec9 = query('slot[name="media"]'), _dec10 = query('slot[name="mask"]'), _dec11 = state(), _dec12 = state(), _dec13 = state(), _dec14 = state();
class VePromptItemElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_disabled, _init_extra__init_disabled, _init_arrow, _init_extra__init_arrow, _init_forceHover, _init_extra__init_forceHover, _init_description, _init_extra__init_description, _init_image, _init_extra__init_image, _init_alt, _init_extra__init_alt, _init_titleSlot, _init_extra__init_titleSlot, _init_descriptionSlot, _init_extra__init_descriptionSlot, _init_mediaSlot, _init_extra__init_mediaSlot, _init_maskSlot, _init_extra__init_maskSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasDescriptionSlot, _init_extra__init_hasDescriptionSlot, _init_hasMediaSlot, _init_extra__init_hasMediaSlot, _init_hasMaskSlot, _init_extra__init_hasMaskSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "type"
            ],
            [
                _dec1,
                1,
                "disabled"
            ],
            [
                _dec2,
                1,
                "arrow"
            ],
            [
                _dec3,
                1,
                "forceHover"
            ],
            [
                _dec4,
                1,
                "description"
            ],
            [
                _dec5,
                1,
                "image"
            ],
            [
                _dec6,
                1,
                "alt"
            ],
            [
                _dec7,
                1,
                "titleSlot"
            ],
            [
                _dec8,
                1,
                "descriptionSlot"
            ],
            [
                _dec9,
                1,
                "mediaSlot"
            ],
            [
                _dec10,
                1,
                "maskSlot"
            ],
            [
                _dec11,
                1,
                "hasTitleSlot"
            ],
            [
                _dec12,
                1,
                "hasDescriptionSlot"
            ],
            [
                _dec13,
                1,
                "hasMediaSlot"
            ],
            [
                _dec14,
                1,
                "hasMaskSlot"
            ]
        ]));
    }
    static styles = ve_prompt_item;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'card');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_disabled_2 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_2;
    }
    set disabled(_v) {
        this.#___private_disabled_2 = _v;
    }
    #___private_arrow_3 = (()=>{
        const _value = _init_arrow(this, void 0);
        _init_extra__init_arrow(this);
        return _value;
    })();
    get arrow() {
        return this.#___private_arrow_3;
    }
    set arrow(_v) {
        this.#___private_arrow_3 = _v;
    }
    #___private_forceHover_4 = (()=>{
        const _value = _init_forceHover(this, false);
        _init_extra__init_forceHover(this);
        return _value;
    })();
    get forceHover() {
        return this.#___private_forceHover_4;
    }
    set forceHover(_v) {
        this.#___private_forceHover_4 = _v;
    }
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_description_5 = (()=>{
        const _value = _init_description(this, '');
        _init_extra__init_description(this);
        return _value;
    })();
    get description() {
        return this.#___private_description_5;
    }
    set description(_v) {
        this.#___private_description_5 = _v;
    }
    #___private_image_6 = (()=>{
        const _value = _init_image(this, '');
        _init_extra__init_image(this);
        return _value;
    })();
    get image() {
        return this.#___private_image_6;
    }
    set image(_v) {
        this.#___private_image_6 = _v;
    }
    #___private_alt_7 = (()=>{
        const _value = _init_alt(this, '');
        _init_extra__init_alt(this);
        return _value;
    })();
    get alt() {
        return this.#___private_alt_7;
    }
    set alt(_v) {
        this.#___private_alt_7 = _v;
    }
    #___private_titleSlot_8 = (()=>{
        const _value = _init_titleSlot(this, null);
        _init_extra__init_titleSlot(this);
        return _value;
    })();
    get titleSlot() {
        return this.#___private_titleSlot_8;
    }
    set titleSlot(_v) {
        this.#___private_titleSlot_8 = _v;
    }
    #___private_descriptionSlot_9 = (()=>{
        const _value = _init_descriptionSlot(this, null);
        _init_extra__init_descriptionSlot(this);
        return _value;
    })();
    get descriptionSlot() {
        return this.#___private_descriptionSlot_9;
    }
    set descriptionSlot(_v) {
        this.#___private_descriptionSlot_9 = _v;
    }
    #___private_mediaSlot_10 = (()=>{
        const _value = _init_mediaSlot(this, null);
        _init_extra__init_mediaSlot(this);
        return _value;
    })();
    get mediaSlot() {
        return this.#___private_mediaSlot_10;
    }
    set mediaSlot(_v) {
        this.#___private_mediaSlot_10 = _v;
    }
    #___private_maskSlot_11 = (()=>{
        const _value = _init_maskSlot(this, null);
        _init_extra__init_maskSlot(this);
        return _value;
    })();
    get maskSlot() {
        return this.#___private_maskSlot_11;
    }
    set maskSlot(_v) {
        this.#___private_maskSlot_11 = _v;
    }
    #___private_hasTitleSlot_12 = (()=>{
        const _value = _init_hasTitleSlot(this, false);
        _init_extra__init_hasTitleSlot(this);
        return _value;
    })();
    get hasTitleSlot() {
        return this.#___private_hasTitleSlot_12;
    }
    set hasTitleSlot(_v) {
        this.#___private_hasTitleSlot_12 = _v;
    }
    #___private_hasDescriptionSlot_13 = (()=>{
        const _value = _init_hasDescriptionSlot(this, false);
        _init_extra__init_hasDescriptionSlot(this);
        return _value;
    })();
    get hasDescriptionSlot() {
        return this.#___private_hasDescriptionSlot_13;
    }
    set hasDescriptionSlot(_v) {
        this.#___private_hasDescriptionSlot_13 = _v;
    }
    #___private_hasMediaSlot_14 = (()=>{
        const _value = _init_hasMediaSlot(this, false);
        _init_extra__init_hasMediaSlot(this);
        return _value;
    })();
    get hasMediaSlot() {
        return this.#___private_hasMediaSlot_14;
    }
    set hasMediaSlot(_v) {
        this.#___private_hasMediaSlot_14 = _v;
    }
    #___private_hasMaskSlot_15 = (()=>{
        const _value = _init_hasMaskSlot(this, false);
        _init_extra__init_hasMaskSlot(this);
        return _value;
    })();
    get hasMaskSlot() {
        return this.#___private_hasMaskSlot_15;
    }
    set hasMaskSlot(_v) {
        this.#___private_hasMaskSlot_15 = _v;
    }
    lightDomSlotObserver;
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    get showArrow() {
        if (void 0 !== this.arrow) return this.arrow;
        return 'media' !== this.type;
    }
    connectedCallback() {
        super.connectedCallback();
        this.observeLightDomSlots();
        this.syncSlotState();
    }
    firstUpdated() {
        this.syncSlotState();
    }
    disconnectedCallback() {
        this.lightDomSlotObserver?.disconnect();
        this.lightDomSlotObserver = void 0;
        super.disconnectedCallback();
    }
    observeLightDomSlots() {
        if ("u" < typeof MutationObserver || this.lightDomSlotObserver) return;
        this.lightDomSlotObserver = new MutationObserver(()=>{
            this.syncSlotState();
        });
        this.lightDomSlotObserver.observe(this, {
            childList: true,
            subtree: false,
            attributes: true,
            attributeFilter: [
                'slot'
            ]
        });
    }
    syncSlotState() {
        this.hasTitleSlot = hasAssignedSlotContent(this.titleSlot) || hasNamedSlottedContent(this, 'title');
        this.hasDescriptionSlot = hasAssignedSlotContent(this.descriptionSlot) || hasNamedSlottedContent(this, "description");
        this.hasMediaSlot = hasAssignedSlotContent(this.mediaSlot) || hasNamedSlottedContent(this, 'media');
        this.hasMaskSlot = hasAssignedSlotContent(this.maskSlot) || hasNamedSlottedContent(this, 'mask');
    }
    handleSlotChange() {
        this.syncSlotState();
    }
    emitClick() {
        if (this.disabled) return;
        this.dispatchEvent(new CustomEvent('ve-prompt-item-click', {
            detail: {
                type: this.type
            },
            bubbles: true,
            composed: true
        }));
    }
    handleClick(event) {
        if (this.disabled) {
            event.preventDefault();
            event.stopImmediatePropagation();
            return;
        }
        this.emitClick();
    }
    handleKeydown(event) {
        if (this.disabled) return;
        if ('Enter' !== event.key) return;
        event.preventDefault();
        this.emitClick();
    }
    renderTitle() {
        return html`
      <p class="ve-prompt-item-title" part="title">
        <slot name="title" @slotchange=${this.handleSlotChange}></slot>
        ${this.hasTitleSlot ? nothing : this.title}
      </p>
    `;
    }
    renderDescription() {
        return html`
      <p class="ve-prompt-item-description" part="description">
        <slot name="description" @slotchange=${this.handleSlotChange}></slot>
        ${this.hasDescriptionSlot ? nothing : this.description}
      </p>
    `;
    }
    renderArrow() {
        if (!this.showArrow) return nothing;
        return html`<ve-icon
      class="ve-prompt-item-arrow"
      part="arrow"
      .definition=${arrowRightMdIcon}
      aria-hidden="true"
    ></ve-icon>`;
    }
    renderPrefix() {
        return html`<span class="ve-prompt-item-prefix"
      ><slot name="prefix"></slot
    ></span>`;
    }
    renderSuffix() {
        return html`<span class="ve-prompt-item-suffix"
      ><slot name="suffix"></slot
    ></span>`;
    }
    renderCardBody() {
        return html`
      ${this.renderPrefix()}
      <div class="ve-prompt-item-copy">
        <div class="ve-prompt-item-title-row">${this.renderTitle()}</div>
        ${this.renderDescription()}
      </div>
      ${this.renderSuffix()} ${this.renderArrow()}
    `;
    }
    renderLineBody() {
        return html`
      ${this.renderPrefix()} ${this.renderTitle()} ${this.renderSuffix()}
      ${this.renderArrow()}
    `;
    }
    renderMedia() {
        const fallback = this.image ? html`<img
          class="ve-prompt-item-media"
          part="media"
          src=${this.image}
          alt=${this.alt}
        />` : nothing;
        return html`<slot name="media" @slotchange=${this.handleSlotChange}></slot>
      ${this.hasMediaSlot ? nothing : fallback}`;
    }
    renderMaskFallback() {
        return html`
      ${this.description || this.hasDescriptionSlot ? html`<p class="ve-prompt-item-mask-title" part="mask-title">
            <slot
              name="description"
              @slotchange=${this.handleSlotChange}
            ></slot>
            ${this.hasDescriptionSlot ? nothing : this.description}
          </p>` : nothing}
      ${this.title || this.hasTitleSlot ? html`<span class="ve-prompt-item-mask-button" part="mask-button">
            <slot name="title" @slotchange=${this.handleSlotChange}></slot>
            ${this.hasTitleSlot ? nothing : this.title}
          </span>` : nothing}
    `;
    }
    renderMask() {
        const hasMaskCopy = Boolean(this.description || this.title || this.hasDescriptionSlot || this.hasTitleSlot);
        if (!hasMaskCopy && !this.hasMaskSlot) return nothing;
        return html`
      <div class="ve-prompt-item-mask" part="mask">
        <slot name="mask" @slotchange=${this.handleSlotChange}></slot>
        ${this.hasMaskSlot ? nothing : this.renderMaskFallback()}
      </div>
    `;
    }
    renderMediaBody() {
        return html`${this.renderMedia()} ${this.renderMask()}`;
    }
    renderBody() {
        switch(this.type){
            case 'media':
                return this.renderMediaBody();
            case 'line':
                return this.renderLineBody();
            case 'card':
            default:
                return this.renderCardBody();
        }
    }
    render() {
        const classes = classMap({
            've-prompt-item': true,
            've-prompt-item-disabled': this.disabled
        });
        return html`
      <button
        class=${classes}
        part="root"
        type="button"
        ?disabled=${this.disabled}
        aria-disabled=${ifDefined(this.disabled ? 'true' : void 0)}
        @click=${this.handleClick}
        @keydown=${this.handleKeydown}
      >
        ${this.renderBody()}
      </button>
    `;
    }
}
defineCustomElement(VE_PROMPT_ITEM_TAG, VePromptItemElement);
export { VE_PROMPT_ITEM_TAG, VePromptItemElement };
