// @ve-design/web/dist/ve-popconfirm.js
import { css, html, nothing } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
import { createFloatingAutoUpdate, getFloatingDistance, createRafScheduler, computeFloatingPanelPosition } from "./42896.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
import { infoIcon } from "./icons/info.js";
const vePopconfirmTokenDefaults = css`
  :host {
    --ve-popconfirm-z-index: 1000;
    --ve-popconfirm-padding-block: var(--space-s);
    --ve-popconfirm-padding-inline: var(--space-m);
    --ve-popconfirm-gap: var(--space-xxs);
    --ve-popconfirm-content-gap: var(--space-s);
    --ve-popconfirm-text-gap: var(--space-xxxs);
    --ve-popconfirm-actions-gap: var(--space-xxxs);
    --ve-popconfirm-border-radius: var(--radius-xl);
    --ve-popconfirm-border-width: var(--stroke-weight-base);
    --ve-popconfirm-border-color: var(--color-border-default);
    --ve-popconfirm-background: var(--color-bg-overlay);
    --ve-popconfirm-shadow: var(--shadow-lg);

    --ve-popconfirm-icon-size: 18px;
    --ve-popconfirm-icon-padding-block-start: 1px;
    --ve-popconfirm-icon-color: var(--color-text-info);

    --ve-popconfirm-title-color: var(--color-text-primary);
    --ve-popconfirm-title-font-size: var(--text-body);
    --ve-popconfirm-title-font-weight: var(--font-weight-medium);
    --ve-popconfirm-title-line-height: 22px;
    --ve-popconfirm-content-color: var(--color-text-tertiary);
    --ve-popconfirm-content-font-size: var(--text-body-sm);
    --ve-popconfirm-content-font-weight: var(--font-weight-normal);
    --ve-popconfirm-content-line-height: var(--line-height-body);

    --ve-popconfirm-action-height: var(--height-small);
    --ve-popconfirm-action-padding-inline: var(--space-xs);
    --ve-popconfirm-action-radius: var(--radius-md);
    --ve-popconfirm-action-gap: var(--space-xxxs);
    --ve-popconfirm-action-font-size: var(--text-body);
    --ve-popconfirm-action-font-weight: var(--font-weight-normal);
    --ve-popconfirm-action-line-height: 22px;
    --ve-popconfirm-action-text-color: var(--color-text-primary);
    --ve-popconfirm-action-hover-background: color-mix(
      var(--color-bg-secondary) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-popconfirm-action-pressed-background: color-mix(
      var(--color-bg-secondary) var(--state-pressed-subtle-ratio),
      var(--state-pressed-subtle-mix)
    );
    --ve-popconfirm-confirm-text-color: var(--color-text-primary);
    --ve-popconfirm-action-disabled-opacity: var(--state-disabled-ratio);

    --ve-popconfirm-trigger-focus-radius: var(--radius-sm);
    --ve-popconfirm-trigger-focus-ring-width: var(--stroke-2);
    --ve-popconfirm-trigger-focus-ring-color: color-mix(
      currentColor 24%,
      transparent
    );

    --ve-popconfirm-offset: var(--space-xxs);
    --ve-popconfirm-arrow-size: var(--space-xs);
    --ve-popconfirm-current-arrow-size: 0px;
    --ve-popconfirm-arrow-offset: calc(
      var(--ve-popconfirm-current-arrow-size) * -0.5
    );
    --ve-popconfirm-arrow-background: var(--ve-popconfirm-background);
    --ve-popconfirm-arrow-border-color: var(--ve-popconfirm-border-color);

    --ve-popconfirm-transition-duration: var(--motion-fast);
    --ve-popconfirm-transition-easing: var(--ease-enter);
    --ve-popconfirm-enter-transform-y: var(--space-xxs);
    --ve-popconfirm-enter-transform-x: var(--space-xxs);

    --ve-popconfirm-current-enter-transform: translate3d(
      0,
      var(--ve-popconfirm-enter-transform-y),
      0
    );
  }

  :host([arrow]) {
    --ve-popconfirm-current-arrow-size: var(--ve-popconfirm-arrow-size);
  }

  :host([position^='top']) {
    --ve-popconfirm-current-enter-transform: translate3d(
      0,
      var(--ve-popconfirm-enter-transform-y),
      0
    );
  }

  :host([position^='bottom']) {
    --ve-popconfirm-current-enter-transform: translate3d(
      0,
      calc(var(--ve-popconfirm-enter-transform-y) * -1),
      0
    );
  }

  :host([position^='left']) {
    --ve-popconfirm-current-enter-transform: translate3d(
      var(--ve-popconfirm-enter-transform-x),
      0,
      0
    );
  }

  :host([position^='right']) {
    --ve-popconfirm-current-enter-transform: translate3d(
      calc(var(--ve-popconfirm-enter-transform-x) * -1),
      0,
      0
    );
  }
`;
const token = vePopconfirmTokenDefaults;
const ve_popconfirm = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-popconfirm {
      position: relative;
      display: inline-flex;
      inline-size: fit-content;
      max-inline-size: 100%;
    }

    .ve-popconfirm-trigger {
      display: inline-flex;
      align-items: inherit;
      justify-content: inherit;
      inline-size: fit-content;
      max-inline-size: 100%;
      outline: none;
    }

    .ve-popconfirm-trigger:focus-visible {
      border-radius: var(--ve-popconfirm-trigger-focus-radius);
      box-shadow: 0 0 0 var(--ve-popconfirm-trigger-focus-ring-width)
        var(--ve-popconfirm-trigger-focus-ring-color);
    }

    .ve-popconfirm-disabled .ve-popconfirm-trigger {
      cursor: not-allowed;
    }

    .ve-popconfirm-panel {
      position: fixed;
      inset: auto;
      margin: 0;
      z-index: var(--ve-popconfirm-z-index);
      box-sizing: border-box;
      display: flex;
      align-items: stretch;
      gap: var(--ve-popconfirm-gap);
      padding-block: var(--ve-popconfirm-padding-block);
      padding-inline: var(--ve-popconfirm-padding-inline);
      border: var(--ve-popconfirm-border-width) solid
        var(--ve-popconfirm-border-color);
      border-radius: var(--ve-popconfirm-border-radius);
      background: var(--ve-popconfirm-background);
      color: var(--ve-popconfirm-title-color);
      box-shadow: var(--ve-popconfirm-shadow);
      font-family: var(--font-sans);
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
      transform: var(--ve-popconfirm-current-enter-transform);
      transform-origin: center;
      transition:
        opacity var(--ve-popconfirm-transition-duration)
          var(--ve-popconfirm-transition-easing),
        transform var(--ve-popconfirm-transition-duration)
          var(--ve-popconfirm-transition-easing),
        visibility 0s linear var(--ve-popconfirm-transition-duration);
    }

    .ve-popconfirm-panel.is-open {
      opacity: 1;
      visibility: visible;
      pointer-events: auto;
      transform: translate3d(0, 0, 0);
      transition-delay: 0s;
    }

    .ve-popconfirm-panel.is-closed {
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
    }

    .ve-popconfirm-arrow {
      position: absolute;
      box-sizing: border-box;
      inline-size: var(--ve-popconfirm-current-arrow-size);
      block-size: var(--ve-popconfirm-current-arrow-size);
      border: 0 solid var(--ve-popconfirm-arrow-border-color);
      background: var(--ve-popconfirm-arrow-background);
      transform: rotate(45deg);
      pointer-events: none;
    }

    .ve-popconfirm-placement-top .ve-popconfirm-arrow,
    .ve-popconfirm-placement-top-start .ve-popconfirm-arrow,
    .ve-popconfirm-placement-top-end .ve-popconfirm-arrow {
      inset-block-end: var(--ve-popconfirm-arrow-offset);
      border-inline-end-width: var(--ve-popconfirm-border-width);
      border-block-end-width: var(--ve-popconfirm-border-width);
    }

    .ve-popconfirm-placement-bottom .ve-popconfirm-arrow,
    .ve-popconfirm-placement-bottom-start .ve-popconfirm-arrow,
    .ve-popconfirm-placement-bottom-end .ve-popconfirm-arrow {
      inset-block-start: var(--ve-popconfirm-arrow-offset);
      border-inline-start-width: var(--ve-popconfirm-border-width);
      border-block-start-width: var(--ve-popconfirm-border-width);
    }

    .ve-popconfirm-placement-left .ve-popconfirm-arrow,
    .ve-popconfirm-placement-left-start .ve-popconfirm-arrow,
    .ve-popconfirm-placement-left-end .ve-popconfirm-arrow {
      inset-inline-end: var(--ve-popconfirm-arrow-offset);
      border-block-start-width: var(--ve-popconfirm-border-width);
      border-inline-end-width: var(--ve-popconfirm-border-width);
    }

    .ve-popconfirm-placement-right .ve-popconfirm-arrow,
    .ve-popconfirm-placement-right-start .ve-popconfirm-arrow,
    .ve-popconfirm-placement-right-end .ve-popconfirm-arrow {
      inset-inline-start: var(--ve-popconfirm-arrow-offset);
      border-inline-start-width: var(--ve-popconfirm-border-width);
      border-block-end-width: var(--ve-popconfirm-border-width);
    }

    .ve-popconfirm-icon {
      display: inline-flex;
      align-items: flex-start;
      justify-content: center;
      flex: none;
      padding-block-start: var(--ve-popconfirm-icon-padding-block-start);
      color: var(--ve-popconfirm-icon-color);
      line-height: 1;
    }

    .ve-popconfirm-icon ve-icon,
    .ve-popconfirm-icon ::slotted(*) {
      inline-size: var(--ve-popconfirm-icon-size);
      block-size: var(--ve-popconfirm-icon-size);
      flex: none;
    }

    .ve-popconfirm-main {
      display: flex;
      flex: 1 1 auto;
      flex-direction: column;
      gap: var(--ve-popconfirm-content-gap);
      align-items: stretch;
      min-inline-size: 0;
    }

    .ve-popconfirm-text {
      display: flex;
      flex-direction: column;
      gap: var(--ve-popconfirm-text-gap);
      min-inline-size: 0;
    }

    .ve-popconfirm-title {
      min-inline-size: 0;
      color: var(--ve-popconfirm-title-color);
      font-size: var(--ve-popconfirm-title-font-size);
      font-weight: var(--ve-popconfirm-title-font-weight);
      line-height: var(--ve-popconfirm-title-line-height);
      overflow-wrap: break-word;
    }

    .ve-popconfirm-content {
      min-inline-size: 0;
      color: var(--ve-popconfirm-content-color);
      font-size: var(--ve-popconfirm-content-font-size);
      font-weight: var(--ve-popconfirm-content-font-weight);
      line-height: var(--ve-popconfirm-content-line-height);
      overflow-wrap: break-word;
    }

    .ve-popconfirm-title ::slotted(*),
    .ve-popconfirm-content ::slotted(*) {
      margin: 0;
    }

    .ve-popconfirm-actions {
      display: flex;
      align-items: center;
      justify-content: flex-end;
      gap: var(--ve-popconfirm-actions-gap);
      inline-size: 100%;
    }

    .ve-popconfirm-actions-slot {
      display: contents;
    }

    .ve-popconfirm-actions ::slotted(*) {
      margin: 0;
    }

    .ve-popconfirm-action {
      appearance: none;
      box-sizing: border-box;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: var(--ve-popconfirm-action-gap);
      min-block-size: var(--ve-popconfirm-action-height);
      block-size: var(--ve-popconfirm-action-height);
      padding-block: 0;
      padding-inline: var(--ve-popconfirm-action-padding-inline);
      border: 0;
      border-radius: var(--ve-popconfirm-action-radius);
      background: transparent;
      color: var(--ve-popconfirm-action-text-color);
      font: inherit;
      font-size: var(--ve-popconfirm-action-font-size);
      font-weight: var(--ve-popconfirm-action-font-weight);
      line-height: var(--ve-popconfirm-action-line-height);
      white-space: nowrap;
      cursor: pointer;
      transition:
        background-color var(--ve-popconfirm-transition-duration)
          var(--ease-standard),
        color var(--ve-popconfirm-transition-duration) var(--ease-standard);
    }

    .ve-popconfirm-action:hover:not(:disabled) {
      background: var(--ve-popconfirm-action-hover-background);
    }

    .ve-popconfirm-action:active:not(:disabled) {
      background: var(--ve-popconfirm-action-pressed-background);
    }

    .ve-popconfirm-action:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--ve-popconfirm-trigger-focus-ring-width)
        var(--ve-popconfirm-trigger-focus-ring-color);
    }

    .ve-popconfirm-action:disabled {
      opacity: var(--ve-popconfirm-action-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-popconfirm-action-confirm {
      color: var(--ve-popconfirm-confirm-text-color);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-popconfirm-panel,
      .ve-popconfirm-action {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _init_position, _init_extra__init_position, _init_content, _init_extra__init_content, _init_okText, _init_extra__init_okText, _init_cancelText, _init_extra__init_cancelText, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_disabled, _init_extra__init_disabled, _init_okDisabled, _init_extra__init_okDisabled, _init_cancelDisabled, _init_extra__init_cancelDisabled, _init_showIcon, _init_extra__init_showIcon, _init_arrow, _init_extra__init_arrow, _init_innerOpen, _init_extra__init_innerOpen, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasContentSlot, _init_extra__init_hasContentSlot, _init_hasActionsSlot, _init_extra__init_hasActionsSlot, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_arrowElement, _init_extra__init_arrowElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_iconSlot, _init_extra__init_iconSlot, _init_titleSlot, _init_extra__init_titleSlot, _init_contentSlot, _init_extra__init_contentSlot;
const VE_POPCONFIRM_TAG = 've-popconfirm';
const VE_POPCONFIRM_OPEN_CHANGE_EVENT = 've-open-change';
const VE_POPCONFIRM_CONFIRM_EVENT = 've-popconfirm-confirm';
const VE_POPCONFIRM_CANCEL_EVENT = 've-popconfirm-cancel';
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
const VIEWPORT_PADDING = 8;
const FALLBACK_OFFSET = 8;
const FALLBACK_ARROW_SIZE = 8;
const ARROW_GAP = 2;
const ICON_SLOT_SIZE = 'var(--ve-popconfirm-icon-size)';
let popconfirmPanelSeed = 0;
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    type: String,
    attribute: 'ok-text',
    reflect: true,
    useDefault: true
}), _dec3 = property({
    type: String,
    attribute: 'cancel-text',
    reflect: true,
    useDefault: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = property({
    type: Boolean,
    attribute: 'default-open'
}), _dec6 = property({
    type: Boolean,
    reflect: true
}), _dec7 = property({
    type: Boolean,
    attribute: 'ok-disabled',
    reflect: true
}), _dec8 = property({
    type: Boolean,
    attribute: 'cancel-disabled',
    reflect: true
}), _dec9 = property({
    attribute: 'show-icon',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec10 = property({
    type: Boolean,
    reflect: true
}), _dec11 = state(), _dec12 = state(), _dec13 = state(), _dec14 = state(), _dec15 = state(), _dec16 = state(), _dec17 = query('.ve-popconfirm-trigger'), _dec18 = query('.ve-popconfirm-panel'), _dec19 = query('.ve-popconfirm-arrow'), _dec20 = query('slot:not([name])'), _dec21 = query('slot[name="icon"]'), _dec22 = query('slot[name="title"]'), _dec23 = query('slot[name="content"]');
class VePopconfirmElement extends VeElement {
    static{
        ({ e: [_init_position, _init_extra__init_position, _init_content, _init_extra__init_content, _init_okText, _init_extra__init_okText, _init_cancelText, _init_extra__init_cancelText, _init_open, _init_extra__init_open, _init_defaultOpen, _init_extra__init_defaultOpen, _init_disabled, _init_extra__init_disabled, _init_okDisabled, _init_extra__init_okDisabled, _init_cancelDisabled, _init_extra__init_cancelDisabled, _init_showIcon, _init_extra__init_showIcon, _init_arrow, _init_extra__init_arrow, _init_innerOpen, _init_extra__init_innerOpen, _init_resolvedPlacement, _init_extra__init_resolvedPlacement, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_hasTitleSlot, _init_extra__init_hasTitleSlot, _init_hasContentSlot, _init_extra__init_hasContentSlot, _init_hasActionsSlot, _init_extra__init_hasActionsSlot, _init_triggerElement, _init_extra__init_triggerElement, _init_panelElement, _init_extra__init_panelElement, _init_arrowElement, _init_extra__init_arrowElement, _init_defaultSlot, _init_extra__init_defaultSlot, _init_iconSlot, _init_extra__init_iconSlot, _init_titleSlot, _init_extra__init_titleSlot, _init_contentSlot, _init_extra__init_contentSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "position"
            ],
            [
                _dec1,
                1,
                "content"
            ],
            [
                _dec2,
                1,
                "okText"
            ],
            [
                _dec3,
                1,
                "cancelText"
            ],
            [
                _dec4,
                1,
                "open"
            ],
            [
                _dec5,
                1,
                "defaultOpen"
            ],
            [
                _dec6,
                1,
                "disabled"
            ],
            [
                _dec7,
                1,
                "okDisabled"
            ],
            [
                _dec8,
                1,
                "cancelDisabled"
            ],
            [
                _dec9,
                1,
                "showIcon"
            ],
            [
                _dec10,
                1,
                "arrow"
            ],
            [
                _dec11,
                1,
                "innerOpen"
            ],
            [
                _dec12,
                1,
                "resolvedPlacement"
            ],
            [
                _dec13,
                1,
                "hasIconSlot"
            ],
            [
                _dec14,
                1,
                "hasTitleSlot"
            ],
            [
                _dec15,
                1,
                "hasContentSlot"
            ],
            [
                _dec16,
                1,
                "hasActionsSlot"
            ],
            [
                _dec17,
                1,
                "triggerElement"
            ],
            [
                _dec18,
                1,
                "panelElement"
            ],
            [
                _dec19,
                1,
                "arrowElement"
            ],
            [
                _dec20,
                1,
                "defaultSlot"
            ],
            [
                _dec21,
                1,
                "iconSlot"
            ],
            [
                _dec22,
                1,
                "titleSlot"
            ],
            [
                _dec23,
                1,
                "contentSlot"
            ]
        ]));
    }
    static styles = ve_popconfirm;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_position_1 = (()=>{
        const _value = _init_position(this, 'top');
        _init_extra__init_position(this);
        return _value;
    })();
    get position() {
        return this.#___private_position_1;
    }
    set position(_v) {
        this.#___private_position_1 = _v;
    }
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_content_2 = (()=>{
        const _value = _init_content(this, '这里是一条文本描述内容。');
        _init_extra__init_content(this);
        return _value;
    })();
    get content() {
        return this.#___private_content_2;
    }
    set content(_v) {
        this.#___private_content_2 = _v;
    }
    #___private_okText_3 = (()=>{
        const _value = _init_okText(this, '确定');
        _init_extra__init_okText(this);
        return _value;
    })();
    get okText() {
        return this.#___private_okText_3;
    }
    set okText(_v) {
        this.#___private_okText_3 = _v;
    }
    #___private_cancelText_4 = (()=>{
        const _value = _init_cancelText(this, '取消');
        _init_extra__init_cancelText(this);
        return _value;
    })();
    get cancelText() {
        return this.#___private_cancelText_4;
    }
    set cancelText(_v) {
        this.#___private_cancelText_4 = _v;
    }
    #___private_open_5 = (()=>{
        const _value = _init_open(this, void 0);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_5;
    }
    set open(_v) {
        this.#___private_open_5 = _v;
    }
    #___private_defaultOpen_6 = (()=>{
        const _value = _init_defaultOpen(this, false);
        _init_extra__init_defaultOpen(this);
        return _value;
    })();
    get defaultOpen() {
        return this.#___private_defaultOpen_6;
    }
    set defaultOpen(_v) {
        this.#___private_defaultOpen_6 = _v;
    }
    #___private_disabled_7 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_7;
    }
    set disabled(_v) {
        this.#___private_disabled_7 = _v;
    }
    #___private_okDisabled_8 = (()=>{
        const _value = _init_okDisabled(this, false);
        _init_extra__init_okDisabled(this);
        return _value;
    })();
    get okDisabled() {
        return this.#___private_okDisabled_8;
    }
    set okDisabled(_v) {
        this.#___private_okDisabled_8 = _v;
    }
    #___private_cancelDisabled_9 = (()=>{
        const _value = _init_cancelDisabled(this, false);
        _init_extra__init_cancelDisabled(this);
        return _value;
    })();
    get cancelDisabled() {
        return this.#___private_cancelDisabled_9;
    }
    set cancelDisabled(_v) {
        this.#___private_cancelDisabled_9 = _v;
    }
    #___private_showIcon_10 = (()=>{
        const _value = _init_showIcon(this, true);
        _init_extra__init_showIcon(this);
        return _value;
    })();
    get showIcon() {
        return this.#___private_showIcon_10;
    }
    set showIcon(_v) {
        this.#___private_showIcon_10 = _v;
    }
    #___private_arrow_11 = (()=>{
        const _value = _init_arrow(this, false);
        _init_extra__init_arrow(this);
        return _value;
    })();
    get arrow() {
        return this.#___private_arrow_11;
    }
    set arrow(_v) {
        this.#___private_arrow_11 = _v;
    }
    #___private_innerOpen_12 = (()=>{
        const _value = _init_innerOpen(this, false);
        _init_extra__init_innerOpen(this);
        return _value;
    })();
    get innerOpen() {
        return this.#___private_innerOpen_12;
    }
    set innerOpen(_v) {
        this.#___private_innerOpen_12 = _v;
    }
    #___private_resolvedPlacement_13 = (()=>{
        const _value = _init_resolvedPlacement(this, this.position);
        _init_extra__init_resolvedPlacement(this);
        return _value;
    })();
    get resolvedPlacement() {
        return this.#___private_resolvedPlacement_13;
    }
    set resolvedPlacement(_v) {
        this.#___private_resolvedPlacement_13 = _v;
    }
    #___private_hasIconSlot_14 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_14;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_14 = _v;
    }
    #___private_hasTitleSlot_15 = (()=>{
        const _value = _init_hasTitleSlot(this, false);
        _init_extra__init_hasTitleSlot(this);
        return _value;
    })();
    get hasTitleSlot() {
        return this.#___private_hasTitleSlot_15;
    }
    set hasTitleSlot(_v) {
        this.#___private_hasTitleSlot_15 = _v;
    }
    #___private_hasContentSlot_16 = (()=>{
        const _value = _init_hasContentSlot(this, false);
        _init_extra__init_hasContentSlot(this);
        return _value;
    })();
    get hasContentSlot() {
        return this.#___private_hasContentSlot_16;
    }
    set hasContentSlot(_v) {
        this.#___private_hasContentSlot_16 = _v;
    }
    #___private_hasActionsSlot_17 = (()=>{
        const _value = _init_hasActionsSlot(this, false);
        _init_extra__init_hasActionsSlot(this);
        return _value;
    })();
    get hasActionsSlot() {
        return this.#___private_hasActionsSlot_17;
    }
    set hasActionsSlot(_v) {
        this.#___private_hasActionsSlot_17 = _v;
    }
    #___private_triggerElement_18 = (()=>{
        const _value = _init_triggerElement(this);
        _init_extra__init_triggerElement(this);
        return _value;
    })();
    get triggerElement() {
        return this.#___private_triggerElement_18;
    }
    set triggerElement(_v) {
        this.#___private_triggerElement_18 = _v;
    }
    #___private_panelElement_19 = (()=>{
        const _value = _init_panelElement(this);
        _init_extra__init_panelElement(this);
        return _value;
    })();
    get panelElement() {
        return this.#___private_panelElement_19;
    }
    set panelElement(_v) {
        this.#___private_panelElement_19 = _v;
    }
    #___private_arrowElement_20 = (()=>{
        const _value = _init_arrowElement(this);
        _init_extra__init_arrowElement(this);
        return _value;
    })();
    get arrowElement() {
        return this.#___private_arrowElement_20;
    }
    set arrowElement(_v) {
        this.#___private_arrowElement_20 = _v;
    }
    #___private_defaultSlot_21 = (()=>{
        const _value = _init_defaultSlot(this);
        _init_extra__init_defaultSlot(this);
        return _value;
    })();
    get defaultSlot() {
        return this.#___private_defaultSlot_21;
    }
    set defaultSlot(_v) {
        this.#___private_defaultSlot_21 = _v;
    }
    #___private_iconSlot_22 = (()=>{
        const _value = _init_iconSlot(this);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_22;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_22 = _v;
    }
    #___private_titleSlot_23 = (()=>{
        const _value = _init_titleSlot(this);
        _init_extra__init_titleSlot(this);
        return _value;
    })();
    get titleSlot() {
        return this.#___private_titleSlot_23;
    }
    set titleSlot(_v) {
        this.#___private_titleSlot_23 = _v;
    }
    #___private_contentSlot_24 = (()=>{
        const _value = _init_contentSlot(this);
        _init_extra__init_contentSlot(this);
        return _value;
    })();
    get contentSlot() {
        return this.#___private_contentSlot_24;
    }
    set contentSlot(_v) {
        this.#___private_contentSlot_24 = _v;
    }
    panelId = `ve-popconfirm-panel-${++popconfirmPanelSeed}`;
    titleId = `${this.panelId}-title`;
    contentId = `${this.panelId}-content`;
    documentBound = false;
    positionRequestId = 0;
    cleanupAutoUpdate = null;
    autoUpdateReference = null;
    repositionScheduler = createRafScheduler();
    slottedTriggerElement = null;
    slottedTriggerHadDisabledAttribute = false;
    titleText = '标题';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    get isOpen() {
        return 'boolean' == typeof this.open ? this.open : this.innerOpen;
    }
    get isControlled() {
        return 'boolean' == typeof this.open;
    }
    get canShowPanel() {
        return this.isOpen && !this.disabled;
    }
    get referenceElement() {
        return this.getSlottedTriggerElement() ?? this.triggerElement;
    }
    get hasTitle() {
        return this.hasTitleSlot || (this.title ?? '').trim().length > 0;
    }
    get hasContent() {
        return this.hasContentSlot || (this.content ?? '').trim().length > 0;
    }
    show() {
        this.setOpen(true, 'api');
    }
    hide() {
        this.setOpen(false, 'api');
    }
    toggle(open = !this.isOpen) {
        this.setOpen(open, 'api');
    }
    focus(options) {
        this.referenceElement?.focus(options);
    }
    connectedCallback() {
        super.connectedCallback();
        this.innerOpen = this.open ?? this.defaultOpen;
    }
    firstUpdated() {
        this.syncSlotState();
        this.syncSlottedTriggerDisabledState();
        this.syncOverlayState();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('open') && 'boolean' == typeof this.open) this.innerOpen = this.open;
        if (changedProperties.has('defaultOpen') && 1 === changedProperties.size && 'boolean' != typeof this.open) this.innerOpen = this.defaultOpen;
        if (changedProperties.has('disabled') && this.disabled && !this.isControlled) this.innerOpen = false;
        if (changedProperties.has('disabled')) this.syncSlottedTriggerDisabledState();
        if (changedProperties.has('open') || changedProperties.has('innerOpen') || changedProperties.has('disabled') || changedProperties.has('position') || changedProperties.has('arrow')) this.syncOverlayState();
    }
    disconnectedCallback() {
        this.repositionScheduler.cancel();
        this.positionRequestId += 1;
        this.teardownAutoUpdate();
        this.removeDocumentListener();
        this.restoreSlottedTriggerDisabledState();
        super.disconnectedCallback();
    }
    dispatchOpenChange(nextOpen, source) {
        this.dispatchEvent(new CustomEvent(VE_POPCONFIRM_OPEN_CHANGE_EVENT, {
            detail: {
                open: nextOpen,
                source
            },
            bubbles: true,
            composed: true
        }));
    }
    setOpen(nextOpen, source) {
        if (nextOpen && this.disabled) return;
        if (this.isOpen === nextOpen) return;
        if (!this.isControlled) this.innerOpen = nextOpen;
        this.dispatchOpenChange(nextOpen, source);
        if (!nextOpen) {
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            this.teardownAutoUpdate();
            this.syncDocumentListener();
            return;
        }
        this.updateComplete.then(()=>{
            if (!this.isConnected || !this.canShowPanel) return;
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
            this.syncDocumentListener();
        });
    }
    dispatchActionEvent(eventName, sourceEvent) {
        return this.dispatchEvent(new CustomEvent(eventName, {
            detail: {
                sourceEvent
            },
            bubbles: true,
            composed: true,
            cancelable: true
        }));
    }
    getSlottedTriggerElement() {
        const assignedElements = this.defaultSlot?.assignedElements({
            flatten: true
        }) ?? [];
        return assignedElements.find((element)=>element instanceof HTMLElement) ?? null;
    }
    syncSlotState() {
        this.hasIconSlot = hasAssignedSlotContent(this.iconSlot) || hasNamedSlottedContent(this, 'icon');
        this.hasTitleSlot = hasAssignedSlotContent(this.titleSlot) || hasNamedSlottedContent(this, 'title');
        this.hasContentSlot = hasAssignedSlotContent(this.contentSlot) || hasNamedSlottedContent(this, 'content');
        this.hasActionsSlot = hasNamedSlottedContent(this, 'actions');
        this.syncSlottedIconSize();
        if (this.canShowPanel) this.scheduleUpdatePosition();
    }
    syncSlottedIconSize() {
        const icons = this.iconSlot?.assignedElements({
            flatten: true
        }).filter((element)=>element.localName === VE_ICON_TAG);
        icons?.forEach((icon)=>{
            if ('' === `${icon.getAttribute('size') ?? ''}`.trim() && '' === `${icon.size ?? ''}`.trim()) icon.size = ICON_SLOT_SIZE;
        });
    }
    syncSlottedTriggerDisabledState() {
        const triggerElement = this.getSlottedTriggerElement();
        if (this.slottedTriggerElement && this.slottedTriggerElement !== triggerElement) {
            if (!this.slottedTriggerHadDisabledAttribute) {
                this.slottedTriggerElement.removeAttribute('disabled');
                Reflect.set(this.slottedTriggerElement, 'disabled', false);
            }
            this.slottedTriggerElement = null;
        }
        if (!triggerElement) return;
        if (this.slottedTriggerElement !== triggerElement) {
            this.slottedTriggerElement = triggerElement;
            this.slottedTriggerHadDisabledAttribute = triggerElement.hasAttribute('disabled');
        }
        if (this.disabled) {
            triggerElement.setAttribute('disabled', '');
            Reflect.set(triggerElement, 'disabled', true);
            return;
        }
        if (!this.slottedTriggerHadDisabledAttribute) {
            triggerElement.removeAttribute('disabled');
            Reflect.set(triggerElement, 'disabled', false);
        }
    }
    restoreSlottedTriggerDisabledState() {
        if (!this.slottedTriggerElement) return;
        if (!this.slottedTriggerHadDisabledAttribute) {
            this.slottedTriggerElement.removeAttribute('disabled');
            Reflect.set(this.slottedTriggerElement, 'disabled', false);
        }
        this.slottedTriggerElement = null;
    }
    syncDocumentListener() {
        if (!this.isConnected) return void this.removeDocumentListener();
        if (this.canShowPanel && !this.documentBound) {
            document.addEventListener('pointerdown', this.handleDocumentPointerDown, true);
            this.documentBound = true;
            return;
        }
        if (!this.canShowPanel) this.removeDocumentListener();
    }
    removeDocumentListener() {
        if (!this.documentBound) return;
        document.removeEventListener('pointerdown', this.handleDocumentPointerDown, true);
        this.documentBound = false;
    }
    syncOverlayState() {
        this.syncDocumentListener();
        if (!this.canShowPanel) {
            hidePanelPopover(this.panelElement);
            this.repositionScheduler.cancel();
            this.positionRequestId += 1;
            this.teardownAutoUpdate();
            return;
        }
        this.updateComplete.then(()=>{
            if (!this.canShowPanel || !this.isConnected) return;
            showPanelPopover(this.panelElement);
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
            this.syncDocumentListener();
        });
    }
    setupAutoUpdate() {
        const referenceElement = this.referenceElement;
        if (!this.isConnected || !this.canShowPanel || !referenceElement || !this.panelElement) return;
        if (this.cleanupAutoUpdate && this.autoUpdateReference === referenceElement) return;
        this.teardownAutoUpdate();
        this.autoUpdateReference = referenceElement;
        this.cleanupAutoUpdate = createFloatingAutoUpdate(referenceElement, this.panelElement, ()=>{
            this.scheduleUpdatePosition();
        });
    }
    teardownAutoUpdate() {
        this.cleanupAutoUpdate?.();
        this.cleanupAutoUpdate = null;
        this.autoUpdateReference = null;
    }
    scheduleUpdatePosition() {
        this.repositionScheduler.schedule(()=>{
            this.updatePosition();
        });
    }
    async updatePosition() {
        const referenceElement = this.referenceElement;
        if (!this.canShowPanel || !this.isConnected || !referenceElement || !this.panelElement) return;
        const requestId = ++this.positionRequestId;
        const distance = getFloatingDistance(this, {
            offsetVarName: '--ve-popconfirm-offset',
            arrowVarName: '--ve-popconfirm-current-arrow-size',
            fallbackOffset: FALLBACK_OFFSET,
            fallbackArrowSize: this.arrow ? FALLBACK_ARROW_SIZE : 0,
            arrowGap: this.arrow ? ARROW_GAP : 0
        });
        this.panelElement.style.position = 'fixed';
        this.panelElement.style.top = '0px';
        this.panelElement.style.left = '0px';
        const position = await computeFloatingPanelPosition({
            reference: referenceElement,
            floating: this.panelElement,
            arrowElement: this.arrow ? this.arrowElement : null,
            placement: this.position,
            distance,
            viewportPadding: VIEWPORT_PADDING
        });
        if (requestId !== this.positionRequestId || !this.canShowPanel || !this.isConnected) return;
        this.resolvedPlacement = position.placement;
        this.panelElement.style.left = `${position.x}px`;
        this.panelElement.style.top = `${position.y}px`;
        if (this.arrowElement) {
            this.arrowElement.style.removeProperty('left');
            this.arrowElement.style.removeProperty('top');
            if (null !== position.arrowX) this.arrowElement.style.left = `${position.arrowX}px`;
            if (null !== position.arrowY) this.arrowElement.style.top = `${position.arrowY}px`;
        }
        this.panelElement.style.visibility = position.hidden ? 'hidden' : 'visible';
    }
    handleDocumentPointerDown = (event)=>{
        if (!this.canShowPanel) return;
        const referenceElement = this.referenceElement;
        const panelElement = this.panelElement;
        const path = event.composedPath();
        const clickedInside = path.includes(this) || (referenceElement ? path.includes(referenceElement) : false) || (panelElement ? path.includes(panelElement) : false);
        if (!clickedInside) this.setOpen(false, 'outside');
    };
    handleTriggerClick = (event)=>{
        if (this.disabled) return;
        event.stopPropagation();
        this.setOpen(!this.isOpen, 'trigger');
    };
    handleTriggerKeydown = (event)=>{
        if (this.disabled || event.target !== this.triggerElement) return;
        if ('Enter' !== event.key && ' ' !== event.key) return;
        event.preventDefault();
        this.setOpen(!this.isOpen, 'trigger');
    };
    handleKeydown = (event)=>{
        if ('Escape' !== event.key || !this.canShowPanel) return;
        event.stopPropagation();
        this.setOpen(false, 'escape');
        this.focus();
    };
    handleConfirmClick = (event)=>{
        if (this.disabled || this.okDisabled) return;
        if (this.dispatchActionEvent(VE_POPCONFIRM_CONFIRM_EVENT, event)) {
            this.setOpen(false, 'confirm');
            if (!this.isControlled) this.focus();
        }
    };
    handleCancelClick = (event)=>{
        if (this.disabled || this.cancelDisabled) return;
        if (this.dispatchActionEvent(VE_POPCONFIRM_CANCEL_EVENT, event)) {
            this.setOpen(false, 'cancel');
            if (!this.isControlled) this.focus();
        }
    };
    handleDefaultSlotChange = ()=>{
        this.syncSlottedTriggerDisabledState();
        if (this.canShowPanel) {
            this.teardownAutoUpdate();
            this.setupAutoUpdate();
            this.scheduleUpdatePosition();
        }
    };
    handleNamedSlotChange = ()=>{
        this.syncSlotState();
    };
    renderIcon() {
        if (!this.showIcon) return nothing;
        return html`
      <div class="ve-popconfirm-icon" part="icon" aria-hidden="true">
        <slot name="icon" @slotchange=${this.handleNamedSlotChange}></slot>
        ${this.hasIconSlot ? nothing : html`<ve-icon
              .definition=${infoIcon}
              size="var(--ve-popconfirm-icon-size)"
            ></ve-icon>`}
      </div>
    `;
    }
    renderTitle() {
        if (!this.hasTitle) return nothing;
        return html`
      <div class="ve-popconfirm-title" part="title" id=${this.titleId}>
        <slot name="title" @slotchange=${this.handleNamedSlotChange}>
          ${this.title}
        </slot>
      </div>
    `;
    }
    renderContent() {
        if (!this.hasContent) return nothing;
        return html`
      <div class="ve-popconfirm-content" part="content" id=${this.contentId}>
        <slot name="content" @slotchange=${this.handleNamedSlotChange}>
          ${this.content}
        </slot>
      </div>
    `;
    }
    renderActions() {
        return html`
      <div class="ve-popconfirm-actions" part="actions">
        <slot
          class="ve-popconfirm-actions-slot"
          name="actions"
          @slotchange=${this.handleNamedSlotChange}
        ></slot>
        ${this.hasActionsSlot ? nothing : html`
              <button
                class="ve-popconfirm-action ve-popconfirm-action-cancel"
                part="cancel-button"
                type="button"
                ?disabled=${this.disabled || this.cancelDisabled}
                @click=${this.handleCancelClick}
              >
                <slot name="cancel" @slotchange=${this.handleNamedSlotChange}>
                  ${this.cancelText}
                </slot>
              </button>
              <button
                class="ve-popconfirm-action ve-popconfirm-action-confirm"
                part="confirm-button"
                type="button"
                ?disabled=${this.disabled || this.okDisabled}
                @click=${this.handleConfirmClick}
              >
                <slot name="confirm" @slotchange=${this.handleNamedSlotChange}>
                  ${this.okText}
                </slot>
              </button>
            `}
      </div>
    `;
    }
    render() {
        const hostClass = classMap({
            've-popconfirm': true,
            've-popconfirm-disabled': this.disabled
        });
        const panelClass = classMap({
            've-popconfirm-panel': true,
            [`ve-popconfirm-placement-${this.resolvedPlacement}`]: true,
            'is-open': this.canShowPanel,
            'is-closed': !this.canShowPanel
        });
        return html`
      <span class=${hostClass} @keydown=${this.handleKeydown}>
        <span
          class="ve-popconfirm-trigger"
          part="trigger"
          tabindex=${this.disabled ? -1 : 0}
          aria-haspopup="dialog"
          aria-expanded=${this.canShowPanel ? 'true' : 'false'}
          aria-disabled=${this.disabled ? 'true' : 'false'}
          aria-controls=${this.canShowPanel ? this.panelId : nothing}
          @click=${this.handleTriggerClick}
          @keydown=${this.handleTriggerKeydown}
        >
          <slot @slotchange=${this.handleDefaultSlotChange}></slot>
        </span>

        <div
          id=${this.panelId}
          class=${panelClass}
          part="panel"
          popover=${"manual"}
          role="dialog"
          aria-modal="false"
          aria-hidden=${this.canShowPanel ? 'false' : 'true'}
          aria-labelledby=${ifDefined(this.hasTitle ? this.titleId : void 0)}
          aria-describedby=${ifDefined(this.hasContent ? this.contentId : void 0)}
        >
          ${this.arrow ? html`<div class="ve-popconfirm-arrow" part="arrow"></div>` : nothing}
          ${this.renderIcon()}
          <div class="ve-popconfirm-main">
            <div class="ve-popconfirm-text">
              ${this.renderTitle()} ${this.renderContent()}
            </div>
            ${this.renderActions()}
          </div>
        </div>
      </span>
    `;
    }
}
defineCustomElement(VE_POPCONFIRM_TAG, VePopconfirmElement);
export { VE_POPCONFIRM_TAG, VePopconfirmElement };
