// @ve-design/web/dist/ve-spin.js
import { css, html, nothing } from "lit";
import { ifDefined } from "lit/directives/if-defined.js";
import { classMap } from "lit/directives/class-map.js";
import { property } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { hasNamedSlottedContent } from "./89163.js";
import { loadingIcon } from "./42938.js";
const veSpinTokenDefaults = css`
  :host {
    --ve-spin-gap-horizontal: var(--space-xxxs);
    --ve-spin-gap-vertical: var(--space-xxs);

    --ve-spin-icon-color: var(--color-icon-secondary);

    --ve-spin-small-icon-size: 14px;
    --ve-spin-default-icon-size: 16px;
    --ve-spin-large-icon-size: 20px;

    --ve-spin-small-font-size: var(--text-caption);
    --ve-spin-default-font-size: var(--text-body-sm);
    --ve-spin-large-font-size: var(--text-body);

    --ve-spin-small-line-height: var(--line-height-body);
    --ve-spin-default-line-height: var(--line-height-body);
    --ve-spin-large-line-height: var(--line-height-body);

    --ve-spin-text-color: var(--color-text-tertiary);
    --ve-spin-font-family: var(--font-sans);
    --ve-spin-font-weight: var(--font-weight-normal);
    --ve-spin-animation-duration: 0.8s;

    --ve-spin-current-icon-size: var(--ve-spin-default-icon-size);
    --ve-spin-current-font-size: var(--ve-spin-default-font-size);
    --ve-spin-current-line-height: var(--ve-spin-default-line-height);
  }

  :host([size='small']) {
    --ve-spin-current-icon-size: var(--ve-spin-small-icon-size);
    --ve-spin-current-font-size: var(--ve-spin-small-font-size);
    --ve-spin-current-line-height: var(--ve-spin-small-line-height);
  }

  :host([size='large']) {
    --ve-spin-current-icon-size: var(--ve-spin-large-icon-size);
    --ve-spin-current-font-size: var(--ve-spin-large-font-size);
    --ve-spin-current-line-height: var(--ve-spin-large-line-height);
  }
`;
const token = veSpinTokenDefaults;
const ve_spin = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      box-sizing: border-box;
      max-inline-size: 100%;
      color: var(--ve-spin-text-color);
      font-family: var(--ve-spin-font-family);
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-spin {
      box-sizing: border-box;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-inline-size: 100%;
      min-inline-size: 0;
      color: var(--ve-spin-text-color);
    }

    .ve-spin-horizontal {
      flex-direction: row;
      gap: var(--ve-spin-gap-horizontal);
    }

    .ve-spin-vertical {
      flex-direction: column;
      gap: var(--ve-spin-gap-vertical);
      text-align: center;
    }

    .ve-spin-icon-only {
      gap: 0;
    }

    .ve-spin-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: var(--ve-spin-current-icon-size);
      block-size: var(--ve-spin-current-icon-size);
      color: var(--ve-spin-icon-color);
      font-size: var(--ve-spin-current-icon-size);
      line-height: 0;
    }

    .ve-spin-icon-graphic {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      transform-origin: center;
    }

    .ve-spin-icon-rotating .ve-spin-icon-graphic {
      animation: ve-spin-rotate var(--ve-spin-animation-duration) linear
        infinite;
    }

    .ve-spin-icon ::slotted(*) {
      display: block;
      flex: none;
      color: inherit;
      line-height: 1;
    }

    .ve-spin-icon ::slotted(ve-icon) {
      inline-size: var(--ve-spin-current-icon-size);
      block-size: var(--ve-spin-current-icon-size);
      --ve-icon-size: var(--ve-spin-current-icon-size);
    }

    .ve-spin-default-icon {
      display: inline-flex;
      inline-size: var(--ve-spin-current-icon-size);
      block-size: var(--ve-spin-current-icon-size);
    }

    .ve-spin-tip {
      min-inline-size: 0;
      color: var(--ve-spin-text-color);
      font-size: var(--ve-spin-current-font-size);
      font-weight: var(--ve-spin-font-weight);
      line-height: var(--ve-spin-current-line-height);
      letter-spacing: var(--letter-spacing-normal);
      overflow-wrap: anywhere;
      word-break: break-word;
    }

    @keyframes ve-spin-rotate {
      to {
        transform: rotate(360deg);
      }
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-spin-icon-rotating .ve-spin-icon-graphic {
        animation-duration: 1.6s;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _init_size, _init_extra__init_size, _init_direction, _init_extra__init_direction, _init_autoRotate, _init_extra__init_autoRotate, _init_tip, _init_extra__init_tip;
const VE_SPIN_TAG = 've-spin';
const SPIN_SIZES = new Set([
    'small',
    'default',
    'large'
]);
const SPIN_DIRECTIONS = new Set([
    'horizontal',
    'vertical'
]);
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
_dec = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    attribute: 'auto-rotate',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec3 = property({
    type: String
});
class VeSpinElement extends VeElement {
    static{
        ({ e: [_init_size, _init_extra__init_size, _init_direction, _init_extra__init_direction, _init_autoRotate, _init_extra__init_autoRotate, _init_tip, _init_extra__init_tip] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "size"
            ],
            [
                _dec1,
                1,
                "direction"
            ],
            [
                _dec2,
                1,
                "autoRotate"
            ],
            [
                _dec3,
                1,
                "tip"
            ]
        ]));
    }
    static styles = ve_spin;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_size_1 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_1;
    }
    set size(_v) {
        this.#___private_size_1 = _v;
    }
    #___private_direction_2 = (()=>{
        const _value = _init_direction(this, 'horizontal');
        _init_extra__init_direction(this);
        return _value;
    })();
    get direction() {
        return this.#___private_direction_2;
    }
    set direction(_v) {
        this.#___private_direction_2 = _v;
    }
    #___private_autoRotate_3 = (()=>{
        const _value = _init_autoRotate(this, true);
        _init_extra__init_autoRotate(this);
        return _value;
    })();
    get autoRotate() {
        return this.#___private_autoRotate_3;
    }
    set autoRotate(_v) {
        this.#___private_autoRotate_3 = _v;
    }
    #___private_tip_4 = (()=>{
        const _value = _init_tip(this, '');
        _init_extra__init_tip(this);
        return _value;
    })();
    get tip() {
        return this.#___private_tip_4;
    }
    set tip(_v) {
        this.#___private_tip_4 = _v;
    }
    get normalizedSize() {
        return SPIN_SIZES.has(this.size) ? this.size : 'default';
    }
    get normalizedDirection() {
        return SPIN_DIRECTIONS.has(this.direction) ? this.direction : 'horizontal';
    }
    get hasCustomIcon() {
        return hasNamedSlottedContent(this, 'icon');
    }
    get hasTip() {
        return (this.tip ?? '').trim().length > 0;
    }
    get statusAriaLabel() {
        if (this.hasTip) return;
        return this.getAttribute('aria-label') ?? 'Loading';
    }
    handleSlotChange() {
        this.requestUpdate();
    }
    renderDefaultIcon() {
        return html`<ve-icon
      class="ve-spin-default-icon"
      .definition=${loadingIcon}
    ></ve-icon>`;
    }
    render() {
        const spinClass = classMap({
            've-spin': true,
            [`ve-spin-${this.normalizedSize}`]: true,
            [`ve-spin-${this.normalizedDirection}`]: true,
            've-spin-icon-only': !this.hasTip
        });
        const iconClass = classMap({
            've-spin-icon': true,
            've-spin-icon-rotating': this.autoRotate
        });
        return html`
      <div
        class=${spinClass}
        part="base"
        role="status"
        aria-live="polite"
        aria-busy="true"
        aria-label=${ifDefined(this.statusAriaLabel)}
      >
        <span class=${iconClass} part="icon">
          <span class="ve-spin-icon-graphic" part="icon-graphic">
            <slot name="icon" @slotchange=${this.handleSlotChange}></slot>
            ${this.hasCustomIcon ? nothing : this.renderDefaultIcon()}
          </span>
        </span>

        ${this.hasTip ? html`<span class="ve-spin-tip" part="tip">${this.tip}</span>` : nothing}
      </div>
    `;
    }
}
defineCustomElement(VE_SPIN_TAG, VeSpinElement);
export { VE_SPIN_TAG, VeSpinElement };
