// @ve-design/web/dist/ve-sidebar.js
import { css, html, nothing } from "lit";
import { property, query, state } from "lit/decorators.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { chevronDownSmIcon } from "./67641.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
import { loadingIcon } from "./42938.js";
import { sidebarSimpleLeftSquare1Icon } from "./icons/sidebar-simple-left-square-1.js";
const veSidebarTokenDefaults = css`
  :host {
    /* ========================================
     * ve-sidebar — root layout
     * ======================================== */
    --ve-sidebar-width-expanded: 220px;
    --ve-sidebar-width-collapsed: 54px;
    --ve-sidebar-min-height: 100%;

    --ve-sidebar-background: var(--color-bg-surface);
    --ve-sidebar-color: var(--color-text-primary);
    --ve-sidebar-font-family: var(--font-sans);

    --ve-sidebar-padding-block: var(--space-s);
    --ve-sidebar-padding-inline-expanded: var(--space-xs);
    --ve-sidebar-padding-inline-collapsed: var(--space-xxs);
    --ve-sidebar-section-gap: var(--space-s);
    --ve-sidebar-group-gap: var(--space-xs);

    --ve-sidebar-current-width: var(--ve-sidebar-width-expanded);
    --ve-sidebar-current-padding-inline: var(--ve-sidebar-padding-inline-expanded);

    /* Pinned 区滚动时的分割线（Figma: border/subtle，对应 gray-100） */
    --ve-sidebar-divider-color: var(--color-bg-muted);
    --ve-sidebar-divider-width: var(--stroke-0_5);

    /* ========================================
     * ve-sidebar — header (logo + toggle)
     * ======================================== */
    --ve-sidebar-header-gap: var(--space-xxxs);
    --ve-sidebar-header-min-height: 36px;

    --ve-sidebar-logo-gap: 6px;
    --ve-sidebar-logo-padding: 6px;
    --ve-sidebar-logo-radius: var(--radius-lg);
    --ve-sidebar-logo-icon-size: 28px;
    --ve-sidebar-logo-text-font-family: var(--font-en);
    --ve-sidebar-logo-text-font-size: var(--text-ui-strong);
    --ve-sidebar-logo-text-font-weight: var(--font-weight-medium);
    --ve-sidebar-logo-text-line-height: var(--line-height-body);
    --ve-sidebar-logo-text-color: var(--color-text-primary);

    --ve-sidebar-toggle-padding: 9px;
    --ve-sidebar-toggle-radius: var(--radius-lg);
    --ve-sidebar-toggle-icon-size: 18px;
    --ve-sidebar-toggle-color: var(--color-text-primary);
    --ve-sidebar-toggle-hover-ratio: var(--state-hover-subtle-ratio);
    --ve-sidebar-toggle-hover-mix: var(--state-hover-subtle-mix);

    --ve-sidebar-logo-toggle-size: 36px;

    --ve-sidebar-focus-ring-color: var(--color-border-interactive);
    --ve-sidebar-focus-ring-width: var(--stroke-2);
    --ve-sidebar-focus-ring-offset: 1px;
    --ve-sidebar-focus-ring-offset-inset: -2px;

    /* ========================================
     * ve-sidebar — footer (user)
     * ======================================== */
    --ve-sidebar-footer-background: var(--color-bg-surface);
    --ve-sidebar-footer-padding: var(--space-xxs);
    --ve-sidebar-footer-min-height: 54px;

    /* ========================================
     * ve-sidebar-group
     * ======================================== */
    --ve-sidebar-group-item-gap: 2px;
    --ve-sidebar-group-radius: var(--radius-lg);

    --ve-sidebar-group-label-padding-block-start: var(--space-xxs);
    --ve-sidebar-group-label-padding-block-end: var(--space-xxxs);
    --ve-sidebar-group-label-padding-inline: var(--space-xxs);
    --ve-sidebar-group-label-gap: var(--space-xxxs);
    --ve-sidebar-group-label-color: var(--color-text-tertiary);
    --ve-sidebar-group-label-font-size: var(--text-caption);
    --ve-sidebar-group-label-font-weight: var(--font-weight-normal);
    --ve-sidebar-group-label-line-height: var(--line-height-body);

    --ve-sidebar-group-chevron-size: 14px;
    --ve-sidebar-group-chevron-color: var(--color-text-tertiary);

    /* ========================================
     * ve-sidebar-item
     * ======================================== */
    --ve-sidebar-item-height: 38px;
    --ve-sidebar-item-padding-inline: var(--space-xxs);
    --ve-sidebar-item-padding-block: var(--space-xxs);
    --ve-sidebar-item-radius: var(--radius-lg);
    --ve-sidebar-item-gap: var(--space-xxs);

    --ve-sidebar-item-icon-size: 20px;
    --ve-sidebar-item-icon-color: var(--color-text-secondary);

    --ve-sidebar-item-text-color: var(--color-text-primary);
    --ve-sidebar-item-text-font-size: var(--text-body);
    --ve-sidebar-item-text-font-weight: var(--font-weight-normal);
    --ve-sidebar-item-text-line-height: 22px;

    /* Hover、selected 共用同一个「选中底」(Figma: bg-muted-surface 选中)；
     * pressed 在此基础上用 subtle recipe 再压暗一档。 */
    --ve-sidebar-item-background: transparent;
    --ve-sidebar-item-current-background: var(--ve-sidebar-item-background);
    --ve-sidebar-item-hover-surface: var(--color-bg-muted);
    --ve-sidebar-item-selected-background: var(--color-bg-muted);
    --ve-sidebar-item-selected-text-color: var(--color-text-primary);
    --ve-sidebar-item-selected-font-weight: var(--font-weight-medium);

    --ve-sidebar-item-pressed-ratio: var(--state-pressed-subtle-ratio);
    --ve-sidebar-item-pressed-mix: var(--state-pressed-subtle-mix);
    --ve-sidebar-item-disabled-opacity: var(--state-disabled-ratio);

    --ve-sidebar-item-suffix-color: var(--color-text-tertiary);
    --ve-sidebar-item-suffix-icon-size: 16px;
    --ve-sidebar-item-suffix-gap: var(--space-xxs);

    --ve-sidebar-item-shortcut-color: var(--color-text-disable);
    --ve-sidebar-item-shortcut-font-size: 11px;
    --ve-sidebar-item-shortcut-line-height: 16px;
    --ve-sidebar-item-shortcut-letter-spacing: 0.2px;

    --ve-sidebar-item-indicator-size: 6px;
    --ve-sidebar-item-indicator-color: var(--color-bg-info);

    /* ========================================
     * ve-sidebar-item — tooltip (collapsed hover)
     * ======================================== */
    --ve-sidebar-tooltip-background: var(--color-bg-inverse);
    --ve-sidebar-tooltip-color: var(--color-text-foreground);
    --ve-sidebar-tooltip-padding-block: var(--space-xxxs);
    --ve-sidebar-tooltip-padding-inline: var(--space-xxs);
    --ve-sidebar-tooltip-radius: var(--radius-sm);
    --ve-sidebar-tooltip-font-size: var(--text-caption);
    --ve-sidebar-tooltip-line-height: var(--line-height-title);
    --ve-sidebar-tooltip-gap: var(--space-xxxs);
    --ve-sidebar-tooltip-offset: var(--space-xxxs);
    --ve-sidebar-tooltip-shortcut-radius: var(--radius-xs);
    --ve-sidebar-tooltip-shortcut-font-size: 11px;
    --ve-sidebar-tooltip-shortcut-padding-inline: var(--space-xxxs);
    --ve-sidebar-tooltip-shortcut-mix-ratio: 12%;
    --ve-sidebar-tooltip-z-index: 1000;

    /* ========================================
     * ve-sidebar — scrollbar
     * ========================================
     * Scrollbar styling is inherited from componentReset
     * (foundation/style/reset.ts). Override the shared
     * --ve-scrollbar-* variables here if a sidebar-specific
     * look is ever required. */
  }

  :host([collapsed]) {
    --ve-sidebar-current-width: var(--ve-sidebar-width-collapsed);
    --ve-sidebar-current-padding-inline: var(--ve-sidebar-padding-inline-collapsed);
  }

  :host([selected]) {
    --ve-sidebar-item-current-background: var(--ve-sidebar-item-selected-background);
    --ve-sidebar-item-text-color: var(--ve-sidebar-item-selected-text-color);
    --ve-sidebar-item-text-font-weight: var(--ve-sidebar-item-selected-font-weight);
  }

  :host([active]) {
    --ve-sidebar-item-current-background: var(--ve-sidebar-item-selected-background);
    --ve-sidebar-item-suffix-color: var(--color-text-primary);
  }
`;
const token = veSidebarTokenDefaults;
const groupLayout = css`
  :host {
    display: flex;
    flex-direction: column;
    gap: var(--ve-sidebar-group-item-gap);
    inline-size: 100%;
    border-radius: var(--ve-sidebar-group-radius);
    box-sizing: border-box;
  }

  .ve-sidebar-group__header {
    display: flex;
    align-items: center;
    gap: var(--ve-sidebar-group-label-gap);
    inline-size: 100%;
    padding-block-start: var(--ve-sidebar-group-label-padding-block-start);
    padding-block-end: var(--ve-sidebar-group-label-padding-block-end);
    padding-inline-start: var(--ve-sidebar-group-label-padding-inline);
    padding-inline-end: 0;
    color: var(--ve-sidebar-group-label-color);
    font-size: var(--ve-sidebar-group-label-font-size);
    font-weight: var(--ve-sidebar-group-label-font-weight);
    line-height: var(--ve-sidebar-group-label-line-height);
    box-sizing: border-box;
    background: transparent;
    border: 0;
    text-align: start;
    font-family: inherit;
  }

  .ve-sidebar-group__header[role='button'] {
    cursor: pointer;
  }

  .ve-sidebar-group__header[role='button']:focus-visible {
    outline: var(--ve-sidebar-focus-ring-width) solid
      var(--ve-sidebar-focus-ring-color);
    outline-offset: var(--ve-sidebar-focus-ring-offset);
    border-radius: var(--ve-sidebar-group-radius);
  }

  :host(:not([label]):not([with-label])) .ve-sidebar-group__header,
  :host([collapsed-parent]) .ve-sidebar-group__header {
    display: none;
  }

  .ve-sidebar-group__label {
    flex: 1 1 auto;
    min-inline-size: 0;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .ve-sidebar-group__header-suffix {
    display: inline-flex;
    align-items: center;
    flex: none;
    color: var(--ve-sidebar-group-label-color);
  }

  .ve-sidebar-group__header-suffix:empty {
    display: none;
  }

  .ve-sidebar-group__chevron {
    inline-size: var(--ve-sidebar-group-chevron-size);
    block-size: var(--ve-sidebar-group-chevron-size);
    color: var(--ve-sidebar-group-chevron-color);
    display: inline-flex;
    align-items: center;
    justify-content: center;
    transition: transform var(--motion-fast) var(--ease-standard);
    flex: none;
  }

  :host([collapsed]) .ve-sidebar-group__chevron {
    transform: rotate(-90deg);
  }

  .ve-sidebar-group__items {
    display: flex;
    flex-direction: column;
    gap: var(--ve-sidebar-group-item-gap);
    inline-size: 100%;
  }

  :host([collapsed]) .ve-sidebar-group__items {
    display: none;
  }
`;
const ve_sidebar_group = [
    token,
    groupLayout
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _init_label, _init_extra__init_label, _init_collapsible, _init_extra__init_collapsible, _init_collapsed, _init_extra__init_collapsed, _init_withLabel, _init_extra__init_withLabel, _init_labelSlot, _init_extra__init_labelSlot, _init_hasLabelSlot, _init_extra__init_hasLabelSlot;
const VE_SIDEBAR_GROUP_TAG = 've-sidebar-group';
const CHEVRON_DOWN_ICON = html`<ve-icon
  .definition=${chevronDownSmIcon}
></ve-icon>`;
_dec = property({
    type: String,
    reflect: true
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    type: Boolean,
    reflect: true
}), _dec3 = property({
    type: Boolean,
    reflect: true,
    attribute: 'with-label'
}), _dec4 = query('slot[name="label"]'), _dec5 = state();
class VeSidebarGroupElement extends VeElement {
    static{
        ({ e: [_init_label, _init_extra__init_label, _init_collapsible, _init_extra__init_collapsible, _init_collapsed, _init_extra__init_collapsed, _init_withLabel, _init_extra__init_withLabel, _init_labelSlot, _init_extra__init_labelSlot, _init_hasLabelSlot, _init_extra__init_hasLabelSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "label"
            ],
            [
                _dec1,
                1,
                "collapsible"
            ],
            [
                _dec2,
                1,
                "collapsed"
            ],
            [
                _dec3,
                1,
                "withLabel"
            ],
            [
                _dec4,
                1,
                "labelSlot"
            ],
            [
                _dec5,
                1,
                "hasLabelSlot"
            ]
        ]));
    }
    static styles = ve_sidebar_group;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_label_1 = (()=>{
        const _value = _init_label(this, '');
        _init_extra__init_label(this);
        return _value;
    })();
    get label() {
        return this.#___private_label_1;
    }
    set label(_v) {
        this.#___private_label_1 = _v;
    }
    #___private_collapsible_2 = (()=>{
        const _value = _init_collapsible(this, false);
        _init_extra__init_collapsible(this);
        return _value;
    })();
    get collapsible() {
        return this.#___private_collapsible_2;
    }
    set collapsible(_v) {
        this.#___private_collapsible_2 = _v;
    }
    #___private_collapsed_3 = (()=>{
        const _value = _init_collapsed(this, false);
        _init_extra__init_collapsed(this);
        return _value;
    })();
    get collapsed() {
        return this.#___private_collapsed_3;
    }
    set collapsed(_v) {
        this.#___private_collapsed_3 = _v;
    }
    #___private_withLabel_4 = (()=>{
        const _value = _init_withLabel(this, false);
        _init_extra__init_withLabel(this);
        return _value;
    })();
    get withLabel() {
        return this.#___private_withLabel_4;
    }
    set withLabel(_v) {
        this.#___private_withLabel_4 = _v;
    }
    #___private_labelSlot_5 = (()=>{
        const _value = _init_labelSlot(this, null);
        _init_extra__init_labelSlot(this);
        return _value;
    })();
    get labelSlot() {
        return this.#___private_labelSlot_5;
    }
    set labelSlot(_v) {
        this.#___private_labelSlot_5 = _v;
    }
    #___private_hasLabelSlot_6 = (()=>{
        const _value = _init_hasLabelSlot(this, false);
        _init_extra__init_hasLabelSlot(this);
        return _value;
    })();
    get hasLabelSlot() {
        return this.#___private_hasLabelSlot_6;
    }
    set hasLabelSlot(_v) {
        this.#___private_hasLabelSlot_6 = _v;
    }
    lightDomSlotObserver;
    connectedCallback() {
        super.connectedCallback();
        this.observeLightDomSlots();
        this.syncSlotState();
    }
    firstUpdated() {
        this.syncSlotState();
    }
    disconnectedCallback() {
        this.lightDomSlotObserver?.disconnect();
        this.lightDomSlotObserver = void 0;
        super.disconnectedCallback();
    }
    toggle() {
        if (!this.collapsible) return;
        this.collapsed = !this.collapsed;
        this.dispatchEvent(new CustomEvent('ve-sidebar-group-toggle', {
            detail: {
                collapsed: this.collapsed,
                label: this.label
            },
            bubbles: true,
            composed: true
        }));
    }
    handleHeaderClick = (event)=>{
        if (!this.collapsible) return;
        event.preventDefault();
        this.toggle();
    };
    handleHeaderKeydown = (event)=>{
        if (!this.collapsible) return;
        if ('Enter' === event.key || ' ' === event.key) {
            event.preventDefault();
            this.toggle();
        }
    };
    observeLightDomSlots() {
        if ("u" < typeof MutationObserver || this.lightDomSlotObserver) return;
        this.lightDomSlotObserver = new MutationObserver(()=>{
            this.syncSlotState();
        });
        this.lightDomSlotObserver.observe(this, {
            childList: true,
            subtree: false,
            attributes: true,
            attributeFilter: [
                'slot'
            ]
        });
    }
    syncSlotState() {
        this.hasLabelSlot = hasAssignedSlotContent(this.labelSlot) || hasNamedSlottedContent(this, 'label');
    }
    handleSlotChange() {
        this.syncSlotState();
    }
    get shouldRenderHeader() {
        return Boolean(this.label && this.label.length > 0 || this.hasLabelSlot || this.collapsible || this.withLabel);
    }
    renderHeader() {
        const interactive = this.collapsible;
        return html`
      <div
        class="ve-sidebar-group__header"
        part="header"
        role=${ifDefined(interactive ? 'button' : void 0)}
        tabindex=${ifDefined(interactive ? '0' : void 0)}
        aria-expanded=${ifDefined(interactive ? this.collapsed ? 'false' : 'true' : void 0)}
        @click=${this.handleHeaderClick}
        @keydown=${this.handleHeaderKeydown}
      >
        <span class="ve-sidebar-group__label" part="label">
          <slot name="label" @slotchange=${this.handleSlotChange}></slot>
          ${this.hasLabelSlot ? nothing : this.label}
        </span>
        <span class="ve-sidebar-group__header-suffix" part="header-suffix">
          <slot name="header-suffix"></slot>
        </span>
        ${this.collapsible ? html`<span class="ve-sidebar-group__chevron" aria-hidden="true">
              ${CHEVRON_DOWN_ICON}
            </span>` : nothing}
      </div>
    `;
    }
    render() {
        return html`
      ${this.shouldRenderHeader ? this.renderHeader() : nothing}
      <div class="ve-sidebar-group__items" part="items">
        <slot></slot>
      </div>
    `;
    }
}
const itemLayout = css`
  :host {
    display: block;
    inline-size: 100%;
    box-sizing: border-box;
  }

  .ve-sidebar-item {
    position: relative;
    box-sizing: border-box;
    display: flex;
    align-items: center;
    gap: var(--ve-sidebar-item-gap);
    inline-size: 100%;
    min-block-size: var(--ve-sidebar-item-height);
    padding-block: var(--ve-sidebar-item-padding-block);
    padding-inline: var(--ve-sidebar-item-padding-inline);
    border-radius: var(--ve-sidebar-item-radius);
    background: var(--ve-sidebar-item-current-background);
    color: var(--ve-sidebar-item-text-color);
    font-family: inherit;
    font-size: var(--ve-sidebar-item-text-font-size);
    font-weight: var(--ve-sidebar-item-text-font-weight);
    line-height: var(--ve-sidebar-item-text-line-height);
    text-decoration: none;
    appearance: none;
    border: 0;
    margin: 0;
    cursor: pointer;
    text-align: start;
    overflow: hidden;
    transition:
      background var(--motion-fast) var(--ease-standard),
      color var(--motion-fast) var(--ease-standard);
  }

  /* hover 与 selected 同色（选中底），保持视觉一致 */
  :host(:not([disabled])) .ve-sidebar-item:hover {
    background: var(--ve-sidebar-item-hover-surface);
  }

  /* pressed 在选中底上再压暗一档 */
  :host(:not([disabled])) .ve-sidebar-item:active {
    background: color-mix(
      var(--ve-sidebar-item-hover-surface)
        var(--ve-sidebar-item-pressed-ratio),
      var(--ve-sidebar-item-pressed-mix)
    );
  }

  :host([disabled]) .ve-sidebar-item {
    opacity: var(--ve-sidebar-item-disabled-opacity);
    cursor: not-allowed;
    pointer-events: none;
  }

  .ve-sidebar-item:focus-visible {
    outline: var(--ve-sidebar-focus-ring-width) solid
      var(--ve-sidebar-focus-ring-color);
    outline-offset: var(--ve-sidebar-focus-ring-offset-inset);
  }

  ::slotted([slot='prefix']) {
    flex: none;
    inline-size: var(--ve-sidebar-item-icon-size);
    block-size: var(--ve-sidebar-item-icon-size);
    color: var(--ve-sidebar-item-icon-color);
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }

  .ve-sidebar-item__label {
    flex: 1 1 auto;
    min-inline-size: 0;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    display: inline-block;
  }

  .ve-sidebar-item__suffix {
    display: inline-flex;
    align-items: center;
    gap: var(--ve-sidebar-item-suffix-gap);
    color: var(--ve-sidebar-item-suffix-color);
    flex: none;
  }

  ::slotted([slot='suffix']) {
    flex: none;
    color: var(--ve-sidebar-item-suffix-color);
    inline-size: var(--ve-sidebar-item-suffix-icon-size);
    block-size: var(--ve-sidebar-item-suffix-icon-size);
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }

  .ve-sidebar-item__shortcut {
    color: var(--ve-sidebar-item-shortcut-color);
    font-size: var(--ve-sidebar-item-shortcut-font-size);
    line-height: var(--ve-sidebar-item-shortcut-line-height);
    letter-spacing: var(--ve-sidebar-item-shortcut-letter-spacing);
    padding-block: 1px;
    flex: none;
  }

  .ve-sidebar-item__indicator {
    inline-size: var(--ve-sidebar-item-suffix-icon-size);
    block-size: var(--ve-sidebar-item-suffix-icon-size);
    flex: none;
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }

  .ve-sidebar-item__indicator-dot {
    inline-size: var(--ve-sidebar-item-indicator-size);
    block-size: var(--ve-sidebar-item-indicator-size);
    border-radius: var(--radius-full);
    background: var(--ve-sidebar-item-indicator-color);
    display: inline-block;
  }

  .ve-sidebar-item__spinner {
    --ve-icon-size: var(--ve-sidebar-item-suffix-icon-size);
    animation: ve-sidebar-spin 0.8s linear infinite;
    flex: none;
  }

  :host([collapsed-parent]) .ve-sidebar-item {
    justify-content: center;
    padding-inline: 0;
  }

  :host([collapsed-parent]) .ve-sidebar-item__label,
  :host([collapsed-parent]) .ve-sidebar-item__suffix {
    display: none;
  }

  /* When collapsed without a prefix icon, hide the empty rounded box. */
  :host([collapsed-parent][data-no-icon]) {
    display: none;
  }

  /* Tooltip — shown on hover only when collapsed-parent + tooltip attr set.
   * Uses position: fixed (JS-positioned) so it escapes ancestor overflow.
   */
  .ve-sidebar-item__tooltip {
    position: fixed;
    inset-block-start: 0;
    inset-inline-start: 0;
    transform: translateY(-50%);
    z-index: var(--ve-sidebar-tooltip-z-index);
    display: none;
    align-items: center;
    gap: var(--ve-sidebar-tooltip-gap);
    background: var(--ve-sidebar-tooltip-background);
    color: var(--ve-sidebar-tooltip-color);
    padding-block: var(--ve-sidebar-tooltip-padding-block);
    padding-inline: var(--ve-sidebar-tooltip-padding-inline);
    border-radius: var(--ve-sidebar-tooltip-radius);
    font-family: inherit;
    font-size: var(--ve-sidebar-tooltip-font-size);
    line-height: var(--ve-sidebar-tooltip-line-height);
    white-space: nowrap;
    pointer-events: none;
    margin: 0;
    border: 0;
  }
  .ve-sidebar-item__tooltip[data-visible] {
    display: inline-flex;
  }
  .ve-sidebar-item__tooltip-shortcut {
    background: color-mix(
      in srgb,
      var(--ve-sidebar-tooltip-color)
        var(--ve-sidebar-tooltip-shortcut-mix-ratio),
      transparent
    );
    padding-inline: var(--ve-sidebar-tooltip-shortcut-padding-inline);
    border-radius: var(--ve-sidebar-tooltip-shortcut-radius);
    font-size: var(--ve-sidebar-tooltip-shortcut-font-size);
    line-height: 1.4;
    font-family: inherit;
    color: inherit;
  }

  @keyframes ve-sidebar-spin {
    to {
      transform: rotate(360deg);
    }
  }

  @media (prefers-reduced-motion: reduce) {
    .ve-sidebar-item {
      transition: none;
    }
    .ve-sidebar-item__spinner {
      animation: none;
    }
  }
`;
const ve_sidebar_item = [
    token,
    itemLayout
];
function ve_sidebar_item_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_sidebar_item_element_toPropertyKey(arg) {
    var key = ve_sidebar_item_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_sidebar_item_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_sidebar_item_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_sidebar_item_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_sidebar_item_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_sidebar_item_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_sidebar_item_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_sidebar_item_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_sidebar_item_element_dec, ve_sidebar_item_element_dec1, ve_sidebar_item_element_dec2, ve_sidebar_item_element_dec3, ve_sidebar_item_element_dec4, ve_sidebar_item_element_dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _init_type, _init_extra__init_type, _init_href, _init_extra__init_href, _init_target, _init_extra__init_target, _init_selected, _init_extra__init_selected, _init_active, _init_extra__init_active, _init_disabled, _init_extra__init_disabled, _init_loading, _init_extra__init_loading, _init_unread, _init_extra__init_unread, _init_shortcut, _init_extra__init_shortcut, _init_tooltip, _init_extra__init_tooltip, _init_collapsedParent, _init_extra__init_collapsedParent, _init__tooltipLabel, _init_extra__init__tooltipLabel, _init__hasPrefixIcon, _init_extra__init__hasPrefixIcon;
const VE_SIDEBAR_ITEM_TAG = 've-sidebar-item';
const ITEM_TYPES = new Set([
    'primary',
    'history'
]);
ve_sidebar_item_element_dec = property({
    type: String,
    reflect: true
}), ve_sidebar_item_element_dec1 = property({
    type: String
}), ve_sidebar_item_element_dec2 = property({
    type: String
}), ve_sidebar_item_element_dec3 = property({
    type: Boolean,
    reflect: true
}), ve_sidebar_item_element_dec4 = property({
    type: Boolean,
    reflect: true
}), ve_sidebar_item_element_dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = property({
    type: Boolean,
    reflect: true
}), _dec7 = property({
    type: Boolean,
    reflect: true
}), _dec8 = property({
    type: String
}), _dec9 = property({
    type: String
}), _dec10 = property({
    type: Boolean,
    reflect: true,
    attribute: 'collapsed-parent'
}), _dec11 = state(), _dec12 = state();
class VeSidebarItemElement extends VeElement {
    static{
        ({ e: [_init_type, _init_extra__init_type, _init_href, _init_extra__init_href, _init_target, _init_extra__init_target, _init_selected, _init_extra__init_selected, _init_active, _init_extra__init_active, _init_disabled, _init_extra__init_disabled, _init_loading, _init_extra__init_loading, _init_unread, _init_extra__init_unread, _init_shortcut, _init_extra__init_shortcut, _init_tooltip, _init_extra__init_tooltip, _init_collapsedParent, _init_extra__init_collapsedParent, _init__tooltipLabel, _init_extra__init__tooltipLabel, _init__hasPrefixIcon, _init_extra__init__hasPrefixIcon] } = ve_sidebar_item_element_apply_decs_2311(this, [], [
            [
                ve_sidebar_item_element_dec,
                1,
                "type"
            ],
            [
                ve_sidebar_item_element_dec1,
                1,
                "href"
            ],
            [
                ve_sidebar_item_element_dec2,
                1,
                "target"
            ],
            [
                ve_sidebar_item_element_dec3,
                1,
                "selected"
            ],
            [
                ve_sidebar_item_element_dec4,
                1,
                "active"
            ],
            [
                ve_sidebar_item_element_dec5,
                1,
                "disabled"
            ],
            [
                _dec6,
                1,
                "loading"
            ],
            [
                _dec7,
                1,
                "unread"
            ],
            [
                _dec8,
                1,
                "shortcut"
            ],
            [
                _dec9,
                1,
                "tooltip"
            ],
            [
                _dec10,
                1,
                "collapsedParent"
            ],
            [
                _dec11,
                1,
                "_tooltipLabel"
            ],
            [
                _dec12,
                1,
                "_hasPrefixIcon"
            ]
        ]));
    }
    static styles = ve_sidebar_item;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_type_1 = (()=>{
        const _value = _init_type(this, 'primary');
        _init_extra__init_type(this);
        return _value;
    })();
    get type() {
        return this.#___private_type_1;
    }
    set type(_v) {
        this.#___private_type_1 = _v;
    }
    #___private_href_2 = (()=>{
        const _value = _init_href(this, '');
        _init_extra__init_href(this);
        return _value;
    })();
    get href() {
        return this.#___private_href_2;
    }
    set href(_v) {
        this.#___private_href_2 = _v;
    }
    #___private_target_3 = (()=>{
        const _value = _init_target(this, '_self');
        _init_extra__init_target(this);
        return _value;
    })();
    get target() {
        return this.#___private_target_3;
    }
    set target(_v) {
        this.#___private_target_3 = _v;
    }
    #___private_selected_4 = (()=>{
        const _value = _init_selected(this, false);
        _init_extra__init_selected(this);
        return _value;
    })();
    get selected() {
        return this.#___private_selected_4;
    }
    set selected(_v) {
        this.#___private_selected_4 = _v;
    }
    #___private_active_5 = (()=>{
        const _value = _init_active(this, false);
        _init_extra__init_active(this);
        return _value;
    })();
    get active() {
        return this.#___private_active_5;
    }
    set active(_v) {
        this.#___private_active_5 = _v;
    }
    #___private_disabled_6 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_6;
    }
    set disabled(_v) {
        this.#___private_disabled_6 = _v;
    }
    #___private_loading_7 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_7;
    }
    set loading(_v) {
        this.#___private_loading_7 = _v;
    }
    #___private_unread_8 = (()=>{
        const _value = _init_unread(this, false);
        _init_extra__init_unread(this);
        return _value;
    })();
    get unread() {
        return this.#___private_unread_8;
    }
    set unread(_v) {
        this.#___private_unread_8 = _v;
    }
    #___private_shortcut_9 = (()=>{
        const _value = _init_shortcut(this, '');
        _init_extra__init_shortcut(this);
        return _value;
    })();
    get shortcut() {
        return this.#___private_shortcut_9;
    }
    set shortcut(_v) {
        this.#___private_shortcut_9 = _v;
    }
    #___private_tooltip_10 = (()=>{
        const _value = _init_tooltip(this, '');
        _init_extra__init_tooltip(this);
        return _value;
    })();
    get tooltip() {
        return this.#___private_tooltip_10;
    }
    set tooltip(_v) {
        this.#___private_tooltip_10 = _v;
    }
    #___private_collapsedParent_11 = (()=>{
        const _value = _init_collapsedParent(this, false);
        _init_extra__init_collapsedParent(this);
        return _value;
    })();
    get collapsedParent() {
        return this.#___private_collapsedParent_11;
    }
    set collapsedParent(_v) {
        this.#___private_collapsedParent_11 = _v;
    }
    #___private__tooltipLabel_12 = (()=>{
        const _value = _init__tooltipLabel(this, '');
        _init_extra__init__tooltipLabel(this);
        return _value;
    })();
    get _tooltipLabel() {
        return this.#___private__tooltipLabel_12;
    }
    set _tooltipLabel(_v) {
        this.#___private__tooltipLabel_12 = _v;
    }
    #___private__hasPrefixIcon_13 = (()=>{
        const _value = _init__hasPrefixIcon(this, false);
        _init_extra__init__hasPrefixIcon(this);
        return _value;
    })();
    get _hasPrefixIcon() {
        return this.#___private__hasPrefixIcon_13;
    }
    set _hasPrefixIcon(_v) {
        this.#___private__hasPrefixIcon_13 = _v;
    }
    tooltipEl = null;
    connectedCallback() {
        super.connectedCallback();
        this.addEventListener('mouseenter', this.handleMouseEnter);
        this.addEventListener('mouseleave', this.handleMouseLeave);
        this.addEventListener('focus', this.handleMouseEnter, true);
        this.addEventListener('blur', this.handleMouseLeave, true);
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.removeEventListener('mouseenter', this.handleMouseEnter);
        this.removeEventListener('mouseleave', this.handleMouseLeave);
        this.removeEventListener('focus', this.handleMouseEnter, true);
        this.removeEventListener('blur', this.handleMouseLeave, true);
        this.unbindReposition();
    }
    firstUpdated() {
        this.tooltipEl = this.renderRoot?.querySelector('.ve-sidebar-item__tooltip') ?? null;
        this.handleLabelSlotChange();
        this.syncPrefixState();
    }
    handlePrefixSlotChange = ()=>{
        this.syncPrefixState();
    };
    syncPrefixState() {
        const slot = this.renderRoot?.querySelector('slot[name="prefix"]');
        this._hasPrefixIcon = hasAssignedSlotContent(slot);
        if (this._hasPrefixIcon) this.removeAttribute('data-no-icon');
        else this.setAttribute('data-no-icon', '');
    }
    handleLabelSlotChange = ()=>{
        const slot = this.renderRoot?.querySelector('slot:not([name])');
        if (!slot) return;
        const text = slot.assignedNodes({
            flatten: true
        }).map((node)=>node.textContent || '').join('').trim();
        this._tooltipLabel = text;
    };
    get effectiveTooltip() {
        return this.tooltip ?? '';
    }
    get itemType() {
        return ITEM_TYPES.has(this.type) ? this.type : 'primary';
    }
    get isLink() {
        return 'string' == typeof this.href && this.href.trim().length > 0;
    }
    shouldShowTooltip() {
        if (this.disabled) return false;
        if (!this.collapsedParent) return false;
        return Boolean(this.tooltip);
    }
    positionTooltip() {
        if (!this.tooltipEl) return;
        const rect = this.getBoundingClientRect();
        const offsetVar = getComputedStyle(this).getPropertyValue('--ve-sidebar-tooltip-offset') || '8px';
        const offset = parseFloat(offsetVar) || 8;
        this.tooltipEl.style.insetInlineStart = `${rect.right + offset}px`;
        this.tooltipEl.style.insetBlockStart = `${rect.top + rect.height / 2}px`;
    }
    bindReposition() {
        window.addEventListener('scroll', this.handleScrollOrResize, true);
        window.addEventListener('resize', this.handleScrollOrResize);
    }
    unbindReposition() {
        window.removeEventListener('scroll', this.handleScrollOrResize, true);
        window.removeEventListener('resize', this.handleScrollOrResize);
    }
    handleScrollOrResize = ()=>{
        if (this.tooltipEl?.hasAttribute('data-visible')) this.positionTooltip();
    };
    handleMouseEnter = ()=>{
        if (!this.shouldShowTooltip()) return;
        if (!this.tooltipEl) return;
        this.positionTooltip();
        this.tooltipEl.setAttribute('data-visible', '');
        this.bindReposition();
    };
    handleMouseLeave = ()=>{
        if (!this.tooltipEl) return;
        this.tooltipEl.removeAttribute('data-visible');
        this.unbindReposition();
    };
    handleClick = (event)=>{
        if (this.disabled) {
            event.preventDefault();
            event.stopPropagation();
            return;
        }
        this.dispatchEvent(new CustomEvent('ve-sidebar-item-click', {
            detail: {
                type: this.itemType,
                selected: this.selected
            },
            bubbles: true,
            composed: true
        }));
    };
    renderIndicator() {
        if (this.loading) return html`<span
        class="ve-sidebar-item__indicator"
        part="indicator"
        aria-hidden="true"
      >
        <ve-icon
          class="ve-sidebar-item__spinner"
          .definition=${loadingIcon}
        ></ve-icon>
      </span>`;
        if (this.unread) return html`<span
        class="ve-sidebar-item__indicator"
        part="indicator"
        aria-label="未读"
      >
        <span class="ve-sidebar-item__indicator-dot"></span>
      </span>`;
        return nothing;
    }
    renderSuffix() {
        return html`
      <span class="ve-sidebar-item__suffix" part="suffix">
        ${this.renderIndicator()}
        <slot name="suffix"></slot>
        ${this.shortcut ? html`<kbd class="ve-sidebar-item__shortcut" part="shortcut"
              >${this.shortcut}</kbd
            >` : nothing}
      </span>
    `;
    }
    renderContent() {
        return html`
      <slot name="prefix" @slotchange=${this.handlePrefixSlotChange}></slot>
      <span class="ve-sidebar-item__label" part="label">
        <slot @slotchange=${this.handleLabelSlotChange}></slot>
      </span>
      ${this.renderSuffix()}
    `;
    }
    renderTooltip() {
        return html`
      <span
        class="ve-sidebar-item__tooltip"
        part="tooltip"
        role="tooltip"
        aria-hidden="true"
      >
        <span class="ve-sidebar-item__tooltip-label"
          >${this.effectiveTooltip}</span
        >
        ${this.shortcut ? html`<kbd class="ve-sidebar-item__tooltip-shortcut"
              >${this.shortcut}</kbd
            >` : nothing}
      </span>
    `;
    }
    render() {
        if (this.isLink) return html`
        <a
          class="ve-sidebar-item"
          part="control"
          href=${ifDefined(this.disabled ? void 0 : this.href)}
          target=${ifDefined(this.disabled ? void 0 : this.target)}
          rel=${ifDefined(!this.disabled && '_blank' === this.target ? 'noreferrer noopener' : void 0)}
          aria-current=${ifDefined(this.selected ? 'page' : void 0)}
          aria-disabled=${ifDefined(this.disabled ? 'true' : void 0)}
          tabindex=${ifDefined(this.disabled ? '-1' : void 0)}
          @click=${this.handleClick}
        >
          ${this.renderContent()}
        </a>
        ${this.renderTooltip()}
      `;
        return html`
      <button
        class="ve-sidebar-item"
        part="control"
        type="button"
        ?disabled=${this.disabled}
        aria-pressed=${ifDefined(this.selected ? 'true' : void 0)}
        @click=${this.handleClick}
      >
        ${this.renderContent()}
      </button>
      ${this.renderTooltip()}
    `;
    }
}
const sidebarLayout = css`
  :host {
    display: flex;
    flex-direction: column;
    box-sizing: border-box;
    inline-size: var(--ve-sidebar-current-width);
    min-block-size: var(--ve-sidebar-min-height);
    background: var(--ve-sidebar-background);
    color: var(--ve-sidebar-color);
    font-family: var(--ve-sidebar-font-family);
    overflow: hidden;
    transition:
      inline-size var(--motion-base) var(--ease-standard),
      padding var(--motion-base) var(--ease-standard);
  }

  .ve-sidebar {
    display: flex;
    flex-direction: column;
    box-sizing: border-box;
    block-size: 100%;
    inline-size: 100%;
  }

  .ve-sidebar__content {
    display: flex;
    flex-direction: column;
    flex: 1 1 auto;
    min-block-size: 0;
    gap: var(--ve-sidebar-section-gap);
  }

  .ve-sidebar__header {
    display: flex;
    align-items: center;
    justify-content: space-between;
    inline-size: 100%;
    min-block-size: var(--ve-sidebar-header-min-height);
    flex: 0 0 auto;
    padding-block-start: var(--ve-sidebar-padding-block);
    padding-inline: var(--ve-sidebar-current-padding-inline);
    box-sizing: border-box;
  }

  :host([collapsed]) .ve-sidebar__header {
    justify-content: center;
  }

  .ve-sidebar__logo {
    display: inline-flex;
    align-items: center;
    gap: var(--ve-sidebar-logo-gap);
    padding: var(--ve-sidebar-logo-padding);
    border-radius: var(--ve-sidebar-logo-radius);
    color: var(--ve-sidebar-logo-text-color);
    font-family: var(--ve-sidebar-logo-text-font-family);
    font-size: var(--ve-sidebar-logo-text-font-size);
    font-weight: var(--ve-sidebar-logo-text-font-weight);
    line-height: var(--ve-sidebar-logo-text-line-height);
    box-sizing: border-box;
  }

  .ve-sidebar__logo-icon,
  ::slotted([slot='logo']) {
    inline-size: var(--ve-sidebar-logo-icon-size);
    block-size: var(--ve-sidebar-logo-icon-size);
    flex: none;
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }

  .ve-sidebar__brand {
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }

  :host([collapsed]) .ve-sidebar__brand {
    display: none;
  }

  .ve-sidebar__toggle {
    appearance: none;
    background: transparent;
    border: 0;
    margin: 0;
    padding: var(--ve-sidebar-toggle-padding);
    border-radius: var(--ve-sidebar-toggle-radius);
    color: var(--ve-sidebar-toggle-color);
    cursor: pointer;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    line-height: 0;
    transition: background var(--motion-fast) var(--ease-standard);
    box-sizing: border-box;
  }

  .ve-sidebar__toggle:hover {
    background: color-mix(
      var(--ve-sidebar-background) var(--ve-sidebar-toggle-hover-ratio),
      var(--ve-sidebar-toggle-hover-mix)
    );
  }

  .ve-sidebar__toggle:focus-visible {
    outline: var(--ve-sidebar-focus-ring-width) solid
      var(--ve-sidebar-focus-ring-color);
    outline-offset: var(--ve-sidebar-focus-ring-offset);
  }

  .ve-sidebar__toggle-icon {
    inline-size: var(--ve-sidebar-toggle-icon-size);
    block-size: var(--ve-sidebar-toggle-icon-size);
    display: inline-flex;
    align-items: center;
    justify-content: center;
  }

  .ve-sidebar__toggle-icon ve-icon {
    --ve-icon-size: var(--ve-sidebar-toggle-icon-size);
  }

  /* Collapsed mode: logo becomes the expand trigger */
  .ve-sidebar__logo-toggle {
    appearance: none;
    background: transparent;
    border: 0;
    margin: 0 auto;
    padding: var(--ve-sidebar-logo-padding);
    border-radius: var(--ve-sidebar-logo-radius);
    color: var(--ve-sidebar-toggle-color);
    cursor: pointer;
    position: relative;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    inline-size: var(--ve-sidebar-logo-toggle-size);
    block-size: var(--ve-sidebar-logo-toggle-size);
    box-sizing: border-box;
    font: inherit;
    transition: background var(--motion-fast) var(--ease-standard);
  }

  .ve-sidebar__logo-toggle:hover {
    background: color-mix(
      var(--ve-sidebar-background) var(--ve-sidebar-toggle-hover-ratio),
      var(--ve-sidebar-toggle-hover-mix)
    );
  }

  .ve-sidebar__logo-toggle:focus-visible {
    outline: var(--ve-sidebar-focus-ring-width) solid
      var(--ve-sidebar-focus-ring-color);
    outline-offset: var(--ve-sidebar-focus-ring-offset);
  }

  .ve-sidebar__logo-base,
  .ve-sidebar__logo-hover {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    color: var(--ve-sidebar-toggle-color);
    transition: opacity var(--motion-fast) var(--ease-standard);
  }

  .ve-sidebar__logo-base {
    inline-size: var(--ve-sidebar-logo-icon-size);
    block-size: var(--ve-sidebar-logo-icon-size);
  }

  .ve-sidebar__logo-hover {
    position: absolute;
    inset: 0;
    display: inline-flex;
    align-items: center;
    justify-content: center;
    opacity: 0;
    pointer-events: none;
  }

  .ve-sidebar__logo-hover ve-icon {
    --ve-icon-size: var(--ve-sidebar-toggle-icon-size);
  }

  .ve-sidebar__logo-toggle:hover .ve-sidebar__logo-base,
  .ve-sidebar__logo-toggle:focus-visible .ve-sidebar__logo-base {
    opacity: 0;
  }

  .ve-sidebar__logo-toggle:hover .ve-sidebar__logo-hover,
  .ve-sidebar__logo-toggle:focus-visible .ve-sidebar__logo-hover {
    opacity: 1;
  }

  .ve-sidebar__body {
    display: flex;
    flex-direction: column;
    gap: var(--ve-sidebar-group-gap);
    flex: 1 1 auto;
    min-block-size: 0;
    inline-size: 100%;
    overflow-y: auto;
    overflow-x: hidden;
    padding-inline: var(--ve-sidebar-current-padding-inline);
    padding-block-end: var(--ve-sidebar-padding-block);
    box-sizing: border-box;
  }

  :host([hide-scrollbar]) .ve-sidebar__body {
    scrollbar-width: none;
  }

  :host([hide-scrollbar]) .ve-sidebar__body::-webkit-scrollbar {
    display: none;
  }

  /* Pinned region — sticky function module; the rest of the body scrolls
   * beneath it. A divider appears under it once the body is scrolled. */
  .ve-sidebar__pinned {
    position: sticky;
    inset-block-start: 0;
    z-index: 2;
    display: flex;
    flex-direction: column;
    gap: var(--ve-sidebar-group-gap);
    inline-size: 100%;
    background: var(--ve-sidebar-background);
    border-block-end: var(--ve-sidebar-divider-width) solid transparent;
    transition: border-color var(--motion-fast) var(--ease-standard);
  }

  .ve-sidebar__pinned[data-empty] {
    display: none;
  }

  :host([data-scrolled]) .ve-sidebar__pinned {
    border-block-end-color: var(--ve-sidebar-divider-color);
  }

  .ve-sidebar__footer {
    display: flex;
    align-items: center;
    justify-content: center;
    inline-size: 100%;
    min-block-size: var(--ve-sidebar-footer-min-height);
    padding: var(--ve-sidebar-footer-padding);
    background: var(--ve-sidebar-footer-background);
    box-sizing: border-box;
    flex: 0 0 auto;
  }

  :host(:not([with-footer])) .ve-sidebar__footer {
    display: none;
  }

  /* Logo tooltip — shown on hover of collapsed logo-toggle */
  .ve-sidebar__logo-tooltip {
    position: fixed;
    inset-block-start: 0;
    inset-inline-start: 0;
    transform: translateY(-50%);
    z-index: var(--ve-sidebar-tooltip-z-index);
    display: none;
    align-items: center;
    gap: var(--ve-sidebar-tooltip-gap);
    background: var(--ve-sidebar-tooltip-background);
    color: var(--ve-sidebar-tooltip-color);
    padding-block: var(--ve-sidebar-tooltip-padding-block);
    padding-inline: var(--ve-sidebar-tooltip-padding-inline);
    border-radius: var(--ve-sidebar-tooltip-radius);
    font-family: inherit;
    font-size: var(--ve-sidebar-tooltip-font-size);
    line-height: var(--ve-sidebar-tooltip-line-height);
    white-space: nowrap;
    pointer-events: none;
    margin: 0;
  }
  .ve-sidebar__logo-tooltip[data-visible] {
    display: inline-flex;
  }

  @media (prefers-reduced-motion: reduce) {
    :host {
      transition: none;
    }
  }
`;
const ve_sidebar = [
    componentReset,
    token,
    sidebarLayout
];
function ve_sidebar_element_toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function ve_sidebar_element_toPropertyKey(arg) {
    var key = ve_sidebar_element_toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function ve_sidebar_element_checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function ve_sidebar_element_setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function ve_sidebar_element_apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: ve_sidebar_element_setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) ve_sidebar_element_setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : ve_sidebar_element_toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return ve_sidebar_element_checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let ve_sidebar_element_dec, ve_sidebar_element_dec1, ve_sidebar_element_dec2, ve_sidebar_element_dec3, ve_sidebar_element_dec4, ve_sidebar_element_dec5, ve_sidebar_element_init_collapsed, ve_sidebar_element_init_extra_init_collapsed, _init_brand, _init_extra__init_brand, _init_withFooter, _init_extra__init_withFooter, _init_expandTooltip, _init_extra__init_expandTooltip, _init_hideScrollbar, _init_extra__init_hideScrollbar, _init__hasPinned, _init_extra__init__hasPinned;
const VE_SIDEBAR_TAG = 've-sidebar';
const SIDEBAR_TOGGLE_ICON = html`
  <ve-icon .definition=${sidebarSimpleLeftSquare1Icon}></ve-icon>
`;
ve_sidebar_element_dec = property({
    type: Boolean,
    reflect: true
}), ve_sidebar_element_dec1 = property({
    type: String
}), ve_sidebar_element_dec2 = property({
    type: Boolean,
    reflect: true,
    attribute: 'with-footer'
}), ve_sidebar_element_dec3 = property({
    type: String,
    attribute: 'expand-tooltip'
}), ve_sidebar_element_dec4 = property({
    type: Boolean,
    reflect: true,
    attribute: 'hide-scrollbar'
}), ve_sidebar_element_dec5 = state();
class VeSidebarElement extends VeElement {
    static{
        ({ e: [ve_sidebar_element_init_collapsed, ve_sidebar_element_init_extra_init_collapsed, _init_brand, _init_extra__init_brand, _init_withFooter, _init_extra__init_withFooter, _init_expandTooltip, _init_extra__init_expandTooltip, _init_hideScrollbar, _init_extra__init_hideScrollbar, _init__hasPinned, _init_extra__init__hasPinned] } = ve_sidebar_element_apply_decs_2311(this, [], [
            [
                ve_sidebar_element_dec,
                1,
                "collapsed"
            ],
            [
                ve_sidebar_element_dec1,
                1,
                "brand"
            ],
            [
                ve_sidebar_element_dec2,
                1,
                "withFooter"
            ],
            [
                ve_sidebar_element_dec3,
                1,
                "expandTooltip"
            ],
            [
                ve_sidebar_element_dec4,
                1,
                "hideScrollbar"
            ],
            [
                ve_sidebar_element_dec5,
                1,
                "_hasPinned"
            ]
        ]));
    }
    static styles = ve_sidebar;
    static dependencies = {
        [VE_SIDEBAR_GROUP_TAG]: VeSidebarGroupElement,
        [VE_SIDEBAR_ITEM_TAG]: VeSidebarItemElement,
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_collapsed_1 = (()=>{
        const _value = ve_sidebar_element_init_collapsed(this, false);
        ve_sidebar_element_init_extra_init_collapsed(this);
        return _value;
    })();
    get collapsed() {
        return this.#___private_collapsed_1;
    }
    set collapsed(_v) {
        this.#___private_collapsed_1 = _v;
    }
    #___private_brand_2 = (()=>{
        const _value = _init_brand(this, '');
        _init_extra__init_brand(this);
        return _value;
    })();
    get brand() {
        return this.#___private_brand_2;
    }
    set brand(_v) {
        this.#___private_brand_2 = _v;
    }
    #___private_withFooter_3 = (()=>{
        const _value = _init_withFooter(this, false);
        _init_extra__init_withFooter(this);
        return _value;
    })();
    get withFooter() {
        return this.#___private_withFooter_3;
    }
    set withFooter(_v) {
        this.#___private_withFooter_3 = _v;
    }
    #___private_expandTooltip_4 = (()=>{
        const _value = _init_expandTooltip(this, '展开');
        _init_extra__init_expandTooltip(this);
        return _value;
    })();
    get expandTooltip() {
        return this.#___private_expandTooltip_4;
    }
    set expandTooltip(_v) {
        this.#___private_expandTooltip_4 = _v;
    }
    #___private_hideScrollbar_5 = (()=>{
        const _value = _init_hideScrollbar(this, false);
        _init_extra__init_hideScrollbar(this);
        return _value;
    })();
    get hideScrollbar() {
        return this.#___private_hideScrollbar_5;
    }
    set hideScrollbar(_v) {
        this.#___private_hideScrollbar_5 = _v;
    }
    #___private__hasPinned_6 = (()=>{
        const _value = _init__hasPinned(this, false);
        _init_extra__init__hasPinned(this);
        return _value;
    })();
    get _hasPinned() {
        return this.#___private__hasPinned_6;
    }
    set _hasPinned(_v) {
        this.#___private__hasPinned_6 = _v;
    }
    logoTooltipEl = null;
    boundLogoTrigger = null;
    slotObserver;
    connectedCallback() {
        super.connectedCallback();
        if (!this.slotObserver) {
            this.slotObserver = new MutationObserver(()=>{
                this.propagateCollapsedState();
                this.bindLogoTooltip();
            });
            this.slotObserver.observe(this, {
                childList: true,
                subtree: true
            });
        }
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        if (this.slotObserver) {
            this.slotObserver.disconnect();
            this.slotObserver = void 0;
        }
        const body = this.renderRoot?.querySelector('.ve-sidebar__body');
        body?.removeEventListener('scroll', this.handleBodyScroll);
        this.unbindLogoTrigger();
        this.removeLogoListeners();
    }
    firstUpdated() {
        this.bindLogoTooltip();
        this.propagateCollapsedState();
        const body = this.renderRoot?.querySelector('.ve-sidebar__body');
        body?.addEventListener('scroll', this.handleBodyScroll, {
            passive: true
        });
    }
    updated(changed) {
        if (changed.has('collapsed')) {
            this.propagateCollapsedState();
            this.bindLogoTooltip();
        }
    }
    toggle() {
        this.collapsed = !this.collapsed;
        this.dispatchEvent(new CustomEvent('ve-sidebar-toggle', {
            detail: {
                collapsed: this.collapsed
            },
            bubbles: true,
            composed: true
        }));
    }
    handleToggleClick = (event)=>{
        event.preventDefault();
        this.toggle();
    };
    handlePinnedSlotChange = (event)=>{
        const slot = event.target;
        this._hasPinned = slot.assignedElements().length > 0;
    };
    handleBodyScroll = ()=>{
        const body = this.renderRoot?.querySelector('.ve-sidebar__body');
        if (!body) return;
        if (body.scrollTop > 0) this.setAttribute('data-scrolled', '');
        else this.removeAttribute('data-scrolled');
    };
    propagateCollapsedState() {
        const items = this.querySelectorAll(VE_SIDEBAR_ITEM_TAG);
        items.forEach((node)=>{
            if (this.collapsed) node.setAttribute('collapsed-parent', '');
            else node.removeAttribute('collapsed-parent');
        });
        const groups = this.querySelectorAll(VE_SIDEBAR_GROUP_TAG);
        groups.forEach((node)=>{
            if (this.collapsed) node.setAttribute('collapsed-parent', '');
            else node.removeAttribute('collapsed-parent');
        });
    }
    unbindLogoTrigger() {
        if (this.boundLogoTrigger) {
            this.boundLogoTrigger.removeEventListener('mouseenter', this.handleLogoMouseEnter);
            this.boundLogoTrigger.removeEventListener('mouseleave', this.handleLogoMouseLeave);
            this.boundLogoTrigger = null;
        }
    }
    bindLogoTooltip() {
        this.logoTooltipEl = this.renderRoot?.querySelector('.ve-sidebar__logo-tooltip') ?? null;
        const trigger = this.renderRoot?.querySelector('.ve-sidebar__logo-toggle');
        if (this.boundLogoTrigger && this.boundLogoTrigger !== trigger) this.unbindLogoTrigger();
        if (trigger && this.boundLogoTrigger !== trigger) {
            trigger.addEventListener('mouseenter', this.handleLogoMouseEnter);
            trigger.addEventListener('mouseleave', this.handleLogoMouseLeave);
            this.boundLogoTrigger = trigger;
        }
    }
    positionLogoTooltip() {
        if (!this.logoTooltipEl || !this.boundLogoTrigger) return;
        const rect = this.boundLogoTrigger.getBoundingClientRect();
        const offsetVar = getComputedStyle(this).getPropertyValue('--ve-sidebar-tooltip-offset') || '8px';
        const offset = parseFloat(offsetVar) || 8;
        this.logoTooltipEl.style.insetInlineStart = `${rect.right + offset}px`;
        this.logoTooltipEl.style.insetBlockStart = `${rect.top + rect.height / 2}px`;
    }
    handleLogoMouseEnter = ()=>{
        if (!this.collapsed || !this.expandTooltip) return;
        if (!this.logoTooltipEl) return;
        this.positionLogoTooltip();
        this.logoTooltipEl.setAttribute('data-visible', '');
        window.addEventListener('scroll', this.handleLogoScrollOrResize, true);
        window.addEventListener('resize', this.handleLogoScrollOrResize);
    };
    handleLogoMouseLeave = ()=>{
        if (!this.logoTooltipEl) return;
        this.logoTooltipEl.removeAttribute('data-visible');
        this.removeLogoListeners();
    };
    removeLogoListeners() {
        window.removeEventListener('scroll', this.handleLogoScrollOrResize, true);
        window.removeEventListener('resize', this.handleLogoScrollOrResize);
    }
    handleLogoScrollOrResize = ()=>{
        if (this.logoTooltipEl?.hasAttribute('data-visible')) this.positionLogoTooltip();
    };
    renderCollapsedLogo() {
        return html`
      <button
        class="ve-sidebar__logo-toggle"
        part="logo"
        type="button"
        aria-label="Expand sidebar"
        aria-expanded="false"
        @click=${this.handleToggleClick}
      >
        <span class="ve-sidebar__logo-base" part="logo-base">
          <slot name="logo"></slot>
        </span>
        <span class="ve-sidebar__logo-hover" part="logo-hover" aria-hidden="true">
          ${SIDEBAR_TOGGLE_ICON}
        </span>
      </button>
      ${this.expandTooltip ? html`<span
            class="ve-sidebar__logo-tooltip"
            part="logo-tooltip"
            role="tooltip"
            aria-hidden="true"
            >${this.expandTooltip}</span
          >` : nothing}
    `;
    }
    renderExpandedLogo() {
        return html`
      <div class="ve-sidebar__logo" part="logo">
        <slot name="logo"></slot>
        <span class="ve-sidebar__brand" part="brand">
          <slot name="brand">${this.brand ? this.brand : nothing}</slot>
        </span>
      </div>
    `;
    }
    renderToggle() {
        return html`
      <button
        class="ve-sidebar__toggle"
        part="toggle"
        type="button"
        aria-label="Collapse sidebar"
        aria-expanded="true"
        @click=${this.handleToggleClick}
      >
        <span class="ve-sidebar__toggle-icon" aria-hidden="true">
          <slot name="toggle-icon">${SIDEBAR_TOGGLE_ICON}</slot>
        </span>
      </button>
    `;
    }
    render() {
        return html`
      <div class="ve-sidebar" part="container">
        <div class="ve-sidebar__content" part="content">
          <header class="ve-sidebar__header" part="header">
            ${this.collapsed ? this.renderCollapsedLogo() : html`${this.renderExpandedLogo()} ${this.renderToggle()}`}
          </header>
          <div class="ve-sidebar__body" part="body">
            <div
              class="ve-sidebar__pinned"
              part="pinned"
              ?data-empty=${!this._hasPinned}
            >
              <slot
                name="pinned"
                @slotchange=${this.handlePinnedSlotChange}
              ></slot>
            </div>
            <slot></slot>
          </div>
        </div>
        <footer class="ve-sidebar__footer" part="footer">
          <slot name="footer"></slot>
        </footer>
      </div>
    `;
    }
}
defineCustomElement(VE_SIDEBAR_TAG, VeSidebarElement);
defineCustomElement(VE_SIDEBAR_GROUP_TAG, VeSidebarGroupElement);
defineCustomElement(VE_SIDEBAR_ITEM_TAG, VeSidebarItemElement);
export { VE_SIDEBAR_GROUP_TAG, VE_SIDEBAR_ITEM_TAG, VE_SIDEBAR_TAG, VeSidebarElement, VeSidebarGroupElement, VeSidebarItemElement };
