// @ve-design/web/dist/ve-text-selection.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { showPanelPopover, hidePanelPopover } from "./7332.js";
import { hasNamedSlottedContent } from "./89163.js";
const veTextSelectionTokenDefaults = css`
  :host {
    /* Surface tokens */
    --ve-text-selection-surface: var(--color-bg-base);
    --ve-text-selection-border: var(--color-border-default);
    --ve-text-selection-text: var(--color-text-primary);
    --ve-text-selection-active-text: var(--color-text-secondary);
    --ve-text-selection-action-hover-surface: var(--color-bg-surface);

    /* Layout tokens */
    --ve-text-selection-padding-block: var(--space-xxxs);
    --ve-text-selection-padding-inline: var(--space-xxxs);
    --ve-text-selection-gap: 2px;
    --ve-text-selection-max-inline-size: 620.8px;

    /* Action tokens */
    --ve-text-selection-action-gap: 4px;
    --ve-text-selection-action-height: 32px;
    --ve-text-selection-action-padding-block: 5px;
    --ve-text-selection-action-padding-inline: var(--space-xs, 8px);
    --ve-text-selection-icon-size: 16px;

    /* Visual tokens */
    --ve-text-selection-radius: var(--radius-lg);
    --ve-text-selection-action-radius: var(--radius-md);
    --ve-text-selection-border-width: var(--stroke-weight-base, 0.5px);
    --ve-text-selection-shadow: var(--shadow-lg);

    /* State tokens */
    --ve-text-selection-disabled-opacity: var(--state-disabled-ratio, 0.4);
    --ve-text-selection-press-transform: translateY(1px);

    /* Typography tokens */
    --ve-text-selection-font-family: var(--font-sans);
    --ve-text-selection-font-size: var(--text-body);
    --ve-text-selection-font-weight: var(--font-weight-normal);
    --ve-text-selection-line-height: 22px;

    /* Floating popup tokens */
    --ve-text-selection-z-index: 1000;
    --ve-text-selection-offset: var(--space-xxs, 8px);
    --ve-text-selection-transition-duration: var(--motion-fast, 150ms);
    --ve-text-selection-transition-easing: ease;
    --ve-text-selection-enter-transform-y: var(--space-xxs, 4px);
  }
`;
const token = veTextSelectionTokenDefaults;
const ve_text_selection = [
    componentReset,
    token,
    css`
    :host {
      display: inline;
      color: inherit;
      font: inherit;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-text-selection-content {
      display: inline;
    }

    .ve-text-selection-popup {
      position: fixed;
      inset: auto;
      margin: 0;
      z-index: var(--ve-text-selection-z-index);
      display: flex;
      align-items: center;
      justify-content: center;
      gap: var(--ve-text-selection-gap);
      max-inline-size: var(--ve-text-selection-max-inline-size);
      padding-block: var(--ve-text-selection-padding-block);
      padding-inline: var(--ve-text-selection-padding-inline);
      border: var(--ve-text-selection-border-width) solid
        var(--ve-text-selection-border);
      border-radius: var(--ve-text-selection-radius);
      background: var(--ve-text-selection-surface);
      box-shadow: var(--ve-text-selection-shadow);
      color: var(--ve-text-selection-text);
      font-family: var(--ve-text-selection-font-family);
      font-size: var(--ve-text-selection-font-size);
      font-weight: var(--ve-text-selection-font-weight);
      line-height: var(--ve-text-selection-line-height);
      white-space: nowrap;
      opacity: 0;
      visibility: hidden;
      pointer-events: none;
      transform: translate3d(0, var(--ve-text-selection-enter-transform-y), 0);
      transition:
        opacity var(--ve-text-selection-transition-duration)
          var(--ve-text-selection-transition-easing),
        transform var(--ve-text-selection-transition-duration)
          var(--ve-text-selection-transition-easing),
        visibility 0s linear var(--ve-text-selection-transition-duration);
    }

    .ve-text-selection-popup.is-open {
      opacity: 1;
      visibility: visible;
      pointer-events: auto;
      transform: translate3d(0, 0, 0);
      transition-delay: 0s;
    }

    .ve-text-selection-action {
      appearance: none;
      position: relative;
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      gap: var(--ve-text-selection-action-gap);
      min-block-size: var(--ve-text-selection-action-height);
      block-size: var(--ve-text-selection-action-height);
      padding-block: var(--ve-text-selection-action-padding-block);
      padding-inline: var(--ve-text-selection-action-padding-inline);
      border: 0;
      border-radius: var(--ve-text-selection-action-radius);
      background: transparent;
      color: var(--ve-text-selection-text);
      font: inherit;
      letter-spacing: 0;
      text-align: center;
      text-decoration: none;
      white-space: nowrap;
      cursor: pointer;
      user-select: none;
      transition:
        background-color 160ms ease,
        color 160ms ease,
        opacity 160ms ease,
        transform 90ms ease-out;
    }

    .ve-text-selection-action:hover {
      background: var(--ve-text-selection-action-hover-surface);
      color: var(--ve-text-selection-active-text);
    }

    .ve-text-selection-action:active {
      transform: var(--ve-text-selection-press-transform);
    }

    .ve-text-selection-action:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px color-mix(in srgb, currentColor 24%, transparent);
    }

    .ve-text-selection-action[disabled] {
      cursor: not-allowed;
      opacity: var(--ve-text-selection-disabled-opacity);
    }

    .ve-text-selection-icon {
      display: inline-flex;
      flex: none;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-text-selection-icon-size);
      block-size: var(--ve-text-selection-icon-size);
      color: currentColor;
    }

    .ve-text-selection-icon ve-icon {
      inline-size: 100%;
      block-size: 100%;
    }

    .ve-text-selection-label {
      display: inline-flex;
      align-items: center;
      min-inline-size: 0;
      color: currentColor;
      white-space: nowrap;
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-text-selection-popup,
      .ve-text-selection-action {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _init_actions, _init_extra__init_actions, _init_disabled, _init_extra__init_disabled, _init_forceOpen, _init_extra__init_forceOpen, _init_open, _init_extra__init_open, _init_selectedText, _init_extra__init_selectedText, _init_popupElement, _init_extra__init_popupElement, _init_contentElement, _init_extra__init_contentElement;
const VE_TEXT_SELECTION_TAG = 've-text-selection';
const VIEWPORT_PADDING = 8;
const POPUP_VERTICAL_GAP = 8;
_dec = property({
    attribute: false
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    type: Boolean,
    reflect: true,
    attribute: 'force-open'
}), _dec3 = state(), _dec4 = state(), _dec5 = query('.ve-text-selection-popup'), _dec6 = query('.ve-text-selection-content');
class VeTextSelectionElement extends VeElement {
    static{
        ({ e: [_init_actions, _init_extra__init_actions, _init_disabled, _init_extra__init_disabled, _init_forceOpen, _init_extra__init_forceOpen, _init_open, _init_extra__init_open, _init_selectedText, _init_extra__init_selectedText, _init_popupElement, _init_extra__init_popupElement, _init_contentElement, _init_extra__init_contentElement] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "actions"
            ],
            [
                _dec1,
                1,
                "disabled"
            ],
            [
                _dec2,
                1,
                "forceOpen"
            ],
            [
                _dec3,
                1,
                "open"
            ],
            [
                _dec4,
                1,
                "selectedText"
            ],
            [
                _dec5,
                1,
                "popupElement"
            ],
            [
                _dec6,
                1,
                "contentElement"
            ]
        ]));
    }
    static styles = ve_text_selection;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_actions_1 = (()=>{
        const _value = _init_actions(this, []);
        _init_extra__init_actions(this);
        return _value;
    })();
    get actions() {
        return this.#___private_actions_1;
    }
    set actions(_v) {
        this.#___private_actions_1 = _v;
    }
    #___private_disabled_2 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_2;
    }
    set disabled(_v) {
        this.#___private_disabled_2 = _v;
    }
    #___private_forceOpen_3 = (()=>{
        const _value = _init_forceOpen(this, false);
        _init_extra__init_forceOpen(this);
        return _value;
    })();
    get forceOpen() {
        return this.#___private_forceOpen_3;
    }
    set forceOpen(_v) {
        this.#___private_forceOpen_3 = _v;
    }
    #___private_open_4 = (()=>{
        const _value = _init_open(this, false);
        _init_extra__init_open(this);
        return _value;
    })();
    get open() {
        return this.#___private_open_4;
    }
    set open(_v) {
        this.#___private_open_4 = _v;
    }
    #___private_selectedText_5 = (()=>{
        const _value = _init_selectedText(this, '');
        _init_extra__init_selectedText(this);
        return _value;
    })();
    get selectedText() {
        return this.#___private_selectedText_5;
    }
    set selectedText(_v) {
        this.#___private_selectedText_5 = _v;
    }
    #___private_popupElement_6 = (()=>{
        const _value = _init_popupElement(this);
        _init_extra__init_popupElement(this);
        return _value;
    })();
    get popupElement() {
        return this.#___private_popupElement_6;
    }
    set popupElement(_v) {
        this.#___private_popupElement_6 = _v;
    }
    #___private_contentElement_7 = (()=>{
        const _value = _init_contentElement(this);
        _init_extra__init_contentElement(this);
        return _value;
    })();
    get contentElement() {
        return this.#___private_contentElement_7;
    }
    set contentElement(_v) {
        this.#___private_contentElement_7 = _v;
    }
    lightDomSlotObserver;
    documentBound = false;
    rafId = 0;
    lastRect = null;
    connectedCallback() {
        super.connectedCallback();
        this.observeLightDomSlots();
        this.bindDocument();
    }
    disconnectedCallback() {
        this.lightDomSlotObserver?.disconnect();
        this.lightDomSlotObserver = void 0;
        this.unbindDocument();
        hidePanelPopover(this.popupElement);
        if (this.rafId) {
            cancelAnimationFrame(this.rafId);
            this.rafId = 0;
        }
        super.disconnectedCallback();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('forceOpen') || changedProperties.has('actions')) if (this.forceOpen && this.actions.length > 0) {
            this.open = true;
            this.scheduleReposition();
        } else if (this.forceOpen || this.hasActiveSelectionInside()) {
            if (0 === this.actions.length) this.open = false;
        } else this.open = false;
        if ((changedProperties.has('open') || changedProperties.has('actions')) && this.open && this.lastRect) this.applyPopupPosition(this.lastRect);
        if (changedProperties.has('open') && !this.open) hidePanelPopover(this.popupElement);
    }
    observeLightDomSlots() {
        if ("u" < typeof MutationObserver || this.lightDomSlotObserver) return;
        this.lightDomSlotObserver = new MutationObserver(()=>{
            this.requestUpdate();
        });
        this.lightDomSlotObserver.observe(this, {
            childList: true,
            subtree: false,
            attributes: true,
            attributeFilter: [
                'slot'
            ]
        });
    }
    bindDocument() {
        if (this.documentBound || "u" < typeof document) return;
        document.addEventListener('selectionchange', this.handleSelectionChange);
        window.addEventListener('scroll', this.handleViewportChange, true);
        window.addEventListener('resize', this.handleViewportChange);
        this.documentBound = true;
    }
    unbindDocument() {
        if (!this.documentBound || "u" < typeof document) return;
        document.removeEventListener('selectionchange', this.handleSelectionChange);
        window.removeEventListener('scroll', this.handleViewportChange, true);
        window.removeEventListener('resize', this.handleViewportChange);
        this.documentBound = false;
    }
    handleViewportChange = ()=>{
        if (!this.open) return;
        this.scheduleReposition();
    };
    handleSelectionChange = ()=>{
        if (this.disabled || 0 === this.actions.length || this.forceOpen) {
            if (this.forceOpen && this.actions.length > 0 && !this.open) {
                this.open = true;
                this.scheduleReposition();
            }
            return;
        }
        const detail = this.computeSelectionRect();
        if (!detail) {
            if (this.open) {
                this.open = false;
                this.selectedText = '';
                this.dispatchEvent(new CustomEvent('ve-text-selection-change', {
                    detail: {
                        selectedText: ''
                    },
                    bubbles: true,
                    composed: true
                }));
            }
            return;
        }
        this.selectedText = detail.text;
        this.lastRect = detail.rect;
        this.open = true;
        this.scheduleReposition();
        this.dispatchEvent(new CustomEvent('ve-text-selection-change', {
            detail: {
                selectedText: detail.text
            },
            bubbles: true,
            composed: true
        }));
    };
    computeSelectionRect() {
        const selection = "u" > typeof window ? window.getSelection() : null;
        if (!selection || selection.isCollapsed || 0 === selection.rangeCount) return null;
        const range = selection.getRangeAt(0);
        if (!this.contentElement) return null;
        const slot = this.contentElement.querySelector('slot');
        if (!slot) return null;
        const assigned = slot.assignedNodes({
            flatten: true
        });
        if (0 === assigned.length) return null;
        const containedAll = this.rangeContainedByNodes(range, assigned);
        if (!containedAll) return null;
        const text = selection.toString().trim();
        if (!text) return null;
        const rect = range.getBoundingClientRect();
        if (0 === rect.width && 0 === rect.height) return null;
        return {
            rect,
            text
        };
    }
    rangeContainedByNodes(range, nodes) {
        return nodes.some((node)=>{
            if (node.nodeType === Node.TEXT_NODE) return range.startContainer === node && range.endContainer === node;
            if (node.nodeType === Node.ELEMENT_NODE) {
                const element = node;
                return element.contains(range.startContainer) && element.contains(range.endContainer);
            }
            return false;
        });
    }
    hasActiveSelectionInside() {
        return null !== this.computeSelectionRect();
    }
    scheduleReposition() {
        if (this.rafId) cancelAnimationFrame(this.rafId);
        this.rafId = requestAnimationFrame(()=>{
            this.rafId = 0;
            if (!this.open) return;
            let rect = this.lastRect;
            if (this.forceOpen) {
                if (!rect) rect = this.contentElement?.getBoundingClientRect() ?? null;
            } else {
                const detail = this.computeSelectionRect();
                if (!detail) return;
                rect = detail.rect;
                this.lastRect = rect;
            }
            if (rect) this.applyPopupPosition(rect);
        });
    }
    applyPopupPosition(rect) {
        const popup = this.popupElement;
        if (!popup) return;
        showPanelPopover(popup);
        const previousVisibility = popup.style.visibility;
        popup.style.visibility = 'hidden';
        popup.style.left = '0px';
        popup.style.top = '0px';
        const popupRect = popup.getBoundingClientRect();
        popup.style.visibility = previousVisibility;
        const viewportWidth = "u" > typeof window ? window.innerWidth : 0;
        const viewportHeight = "u" > typeof window ? window.innerHeight : 0;
        let left = rect.left + (rect.width - popupRect.width) / 2;
        left = Math.min(Math.max(VIEWPORT_PADDING, left), Math.max(VIEWPORT_PADDING, viewportWidth - popupRect.width - VIEWPORT_PADDING));
        let top = rect.top - popupRect.height - POPUP_VERTICAL_GAP;
        if (top < VIEWPORT_PADDING) top = rect.bottom + POPUP_VERTICAL_GAP;
        top = Math.min(Math.max(VIEWPORT_PADDING, top), Math.max(VIEWPORT_PADDING, viewportHeight - popupRect.height - VIEWPORT_PADDING));
        popup.style.left = `${left}px`;
        popup.style.top = `${top}px`;
    }
    handleActionClick(action, event) {
        if (this.disabled || action.disabled) {
            event.preventDefault();
            event.stopPropagation();
            return;
        }
        const detail = {
            action: action.key,
            label: action.label,
            selectedText: this.selectedText
        };
        action.onClick?.(detail);
        this.dispatchEvent(new CustomEvent('ve-text-selection-action', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    handlePopupMouseDown = (event)=>{
        event.preventDefault();
    };
    renderAction(action) {
        const iconSlotName = `${action.key}-icon`;
        const labelSlotName = `${action.key}-label`;
        const hasIconSlot = hasNamedSlottedContent(this, iconSlotName);
        const hasLabelSlot = hasNamedSlottedContent(this, labelSlotName);
        return html`
      <button
        class="ve-text-selection-action"
        part="action"
        type="button"
        data-action=${action.key}
        ?disabled=${action.disabled ?? false}
        @click=${(event)=>this.handleActionClick(action, event)}
      >
        ${action.icon || hasIconSlot ? html`<span class="ve-text-selection-icon" part="icon">
              <slot name=${iconSlotName}></slot>
              ${hasIconSlot ? nothing : html`<ve-icon name=${action.icon}></ve-icon>`}
            </span>` : nothing}
        <span class="ve-text-selection-label" part="label">
          <slot name=${labelSlotName}></slot>
          ${hasLabelSlot ? nothing : action.label}
        </span>
      </button>
    `;
    }
    render() {
        const isVisible = this.open && !this.disabled && this.actions.length > 0;
        const popupClass = classMap({
            've-text-selection-popup': true,
            'is-open': isVisible
        });
        return html`
      <span class="ve-text-selection-content"><slot></slot></span>
      <div
        class=${popupClass}
        part="popup"
        popover=${"manual"}
        role="toolbar"
        aria-label="文本操作"
        aria-hidden=${isVisible ? 'false' : 'true'}
        @mousedown=${this.handlePopupMouseDown}
      >
        ${this.actions.map((action)=>this.renderAction(action))}
      </div>
    `;
    }
}
defineCustomElement(VE_TEXT_SELECTION_TAG, VeTextSelectionElement);
export { VE_TEXT_SELECTION_TAG, VeTextSelectionElement };
