// @ve-design/web/dist/ve-tag.js
import { css, html, nothing } from "lit";
import { styleMap } from "lit/directives/style-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { successIcon } from "./icons/success.js";
import { warningIcon } from "./icons/warning.js";
import { errorIcon } from "./icons/error.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
const veTagTokenDefaults = css`
  :host {
    --ve-tag-height: 22px;
    --ve-tag-padding-inline: calc(var(--space-xxs) - var(--stroke-weight-bold));
    --ve-tag-gap: var(--space-xxxs);
    --ve-tag-radius: var(--radius-sm);
    --ve-tag-font-family: var(--font-sans);
    --ve-tag-font-size: var(--text-caption);
    --ve-tag-font-weight: var(--font-weight-normal);
    --ve-tag-line-height: var(--line-height-body);
    --ve-tag-background: var(--color-bg-surface);
    --ve-tag-text-color: var(--color-text-secondary);
    --ve-tag-border-color: var(--color-border-default);
    --ve-tag-border-width: var(--stroke-weight-base);
    --ve-tag-icon-size: 14px;
    --ve-tag-close-size: 12px;
    --ve-tag-close-icon-size: 12px;
    --ve-tag-focus-ring-width: var(--stroke-weight-bold);
    --ve-tag-focus-ring-color: color-mix(
      var(--ve-tag-current-text-color) 24%,
      transparent
    );

    --ve-tag-status-success-background: var(--color-bg-success-secondary);
    --ve-tag-status-success-text-color: var(--color-text-success);
    --ve-tag-status-success-border-color: var(--color-border-success);
    --ve-tag-status-warning-background: var(--color-bg-warning-secondary);
    --ve-tag-status-warning-text-color: var(--color-text-warning);
    --ve-tag-status-warning-border-color: var(--color-border-warning);
    --ve-tag-status-error-background: var(--color-bg-danger-secondary);
    --ve-tag-status-error-text-color: var(--color-text-danger);
    --ve-tag-status-error-border-color: var(--color-border-danger);

    --ve-tag-current-background: var(--ve-tag-background);
    --ve-tag-current-text-color: var(--ve-tag-text-color);
    --ve-tag-current-border-color: transparent;
    --ve-tag-current-close-color: var(--ve-tag-current-text-color);
  }

  :host([size='small']) {
    --ve-tag-height: 16px;
    --ve-tag-padding-inline: calc(var(--space-xxs) - var(--stroke-weight-bold));
    --ve-tag-radius: var(--radius-xs);
    --ve-tag-font-size: calc(var(--text-caption) - var(--stroke-weight-bold));
    --ve-tag-icon-size: 12px;
    --ve-tag-close-size: 10px;
    --ve-tag-close-icon-size: 10px;
  }

  :host([size='medium']) {
    --ve-tag-height: 18px;
    --ve-tag-padding-inline: calc(var(--space-xxs) - var(--stroke-weight-bold));
    --ve-tag-radius: var(--radius-xs);
    --ve-tag-font-size: calc(var(--text-caption) - var(--stroke-weight-bold));
    --ve-tag-icon-size: 12px;
    --ve-tag-close-size: 10px;
    --ve-tag-close-icon-size: 10px;
  }

  :host([size='large']) {
    --ve-tag-height: 24px;
    --ve-tag-padding-inline: var(--space-xxs);
    --ve-tag-radius: var(--radius-sm);
    --ve-tag-font-size: var(--text-body-sm);
    --ve-tag-icon-size: 14px;
    --ve-tag-close-size: 12px;
    --ve-tag-close-icon-size: 12px;
  }

  :host([bordered]) {
    --ve-tag-current-border-color: var(--ve-tag-border-color);
  }

  :host([status='success']) {
    --ve-tag-current-background: var(--ve-tag-status-success-background);
    --ve-tag-current-text-color: var(--ve-tag-status-success-text-color);
  }

  :host([status='warning']) {
    --ve-tag-current-background: var(--ve-tag-status-warning-background);
    --ve-tag-current-text-color: var(--ve-tag-status-warning-text-color);
  }

  :host([status='error']) {
    --ve-tag-current-background: var(--ve-tag-status-error-background);
    --ve-tag-current-text-color: var(--ve-tag-status-error-text-color);
  }

  :host([status='success'][bordered]) {
    --ve-tag-current-border-color: var(--ve-tag-status-success-border-color);
  }

  :host([status='warning'][bordered]) {
    --ve-tag-current-border-color: var(--ve-tag-status-warning-border-color);
  }

  :host([status='error'][bordered]) {
    --ve-tag-current-border-color: var(--ve-tag-status-error-border-color);
  }
`;
const token = veTagTokenDefaults;
const ve_tag = [
    componentReset,
    token,
    css`
    :host {
      display: inline-flex;
      vertical-align: middle;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-tag {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-tag-gap);
      box-sizing: border-box;
      block-size: var(--ve-tag-height);
      max-inline-size: 100%;
      padding-inline: var(--ve-tag-padding-inline);
      border: var(--ve-tag-border-width) solid
        var(--ve-tag-current-border-color);
      border-radius: var(--ve-tag-radius);
      background: var(--ve-tag-current-background);
      color: var(--ve-tag-current-text-color);
      font-family: var(--ve-tag-font-family);
      font-size: var(--ve-tag-font-size);
      font-weight: var(--ve-tag-font-weight);
      line-height: var(--ve-tag-line-height);
      white-space: nowrap;
      overflow: hidden;
      transition:
        background-color var(--motion-fast) ease,
        border-color var(--motion-fast) ease,
        color var(--motion-fast) ease,
        box-shadow var(--motion-fast) ease;
      outline: none;
    }

    .ve-tag-control {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-tag-gap);
      min-inline-size: 0;
      color: inherit;
      outline: none;
    }

    .ve-tag-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: var(--ve-tag-icon-size);
      block-size: var(--ve-tag-icon-size);
      color: inherit;
      line-height: 0;
      --ve-icon-size: var(--ve-tag-icon-size);
    }

    .ve-tag-icon ::slotted(*) {
      flex: none;
      inline-size: var(--ve-tag-icon-size);
      block-size: var(--ve-tag-icon-size);
      color: inherit;
      line-height: 0;
    }

    .ve-tag-close:focus-visible {
      box-shadow: 0 0 0 var(--ve-tag-focus-ring-width)
        var(--ve-tag-focus-ring-color);
    }

    .ve-tag-label {
      display: block;
      min-inline-size: 0;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ve-tag-close {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: var(--ve-tag-close-size);
      block-size: var(--ve-tag-close-size);
      padding: 0;
      border: none;
      border-radius: var(--radius-full);
      background: transparent;
      color: var(--ve-tag-current-close-color);
      cursor: pointer;
      outline: none;
      transition:
        background-color var(--motion-fast) ease,
        box-shadow var(--motion-fast) ease;
    }

    .ve-tag-close svg {
      inline-size: var(--ve-tag-close-icon-size);
      block-size: var(--ve-tag-close-icon-size);
    }

    .ve-tag-close:hover {
      background: color-mix(
        var(--ve-tag-current-background) var(--state-hover-subtle-ratio),
        var(--state-hover-subtle-mix)
      );
    }

    .ve-tag-close:active {
      background: color-mix(
        var(--ve-tag-current-background) var(--state-pressed-subtle-ratio),
        var(--state-pressed-subtle-mix)
      );
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _init_color, _init_extra__init_color, _init_bgColor, _init_extra__init_bgColor, _init_status, _init_extra__init_status, _init_size, _init_extra__init_size, _init_closable, _init_extra__init_closable, _init_bordered, _init_extra__init_bordered, _init_iconSlot, _init_extra__init_iconSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot;
const VE_TAG_TAG = 've-tag';
const TAG_STATUS_ICONS = {
    success: successIcon,
    warning: warningIcon,
    error: errorIcon
};
_dec = property({
    type: String,
    reflect: true
}), _dec1 = property({
    type: String,
    attribute: 'bg-color',
    reflect: true
}), _dec2 = property({
    type: String,
    reflect: true
}), _dec3 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = property({
    type: Boolean,
    reflect: true
}), _dec6 = query('slot[name="icon"]'), _dec7 = state();
class VeTagElement extends VeElement {
    static{
        ({ e: [_init_color, _init_extra__init_color, _init_bgColor, _init_extra__init_bgColor, _init_status, _init_extra__init_status, _init_size, _init_extra__init_size, _init_closable, _init_extra__init_closable, _init_bordered, _init_extra__init_bordered, _init_iconSlot, _init_extra__init_iconSlot, _init_hasIconSlot, _init_extra__init_hasIconSlot] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "color"
            ],
            [
                _dec1,
                1,
                "bgColor"
            ],
            [
                _dec2,
                1,
                "status"
            ],
            [
                _dec3,
                1,
                "size"
            ],
            [
                _dec4,
                1,
                "closable"
            ],
            [
                _dec5,
                1,
                "bordered"
            ],
            [
                _dec6,
                1,
                "iconSlot"
            ],
            [
                _dec7,
                1,
                "hasIconSlot"
            ]
        ]));
    }
    static styles = ve_tag;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_color_1 = (()=>{
        const _value = _init_color(this, '');
        _init_extra__init_color(this);
        return _value;
    })();
    get color() {
        return this.#___private_color_1;
    }
    set color(_v) {
        this.#___private_color_1 = _v;
    }
    #___private_bgColor_2 = (()=>{
        const _value = _init_bgColor(this, '');
        _init_extra__init_bgColor(this);
        return _value;
    })();
    get bgColor() {
        return this.#___private_bgColor_2;
    }
    set bgColor(_v) {
        this.#___private_bgColor_2 = _v;
    }
    #___private_status_3 = (()=>{
        const _value = _init_status(this, '');
        _init_extra__init_status(this);
        return _value;
    })();
    get status() {
        return this.#___private_status_3;
    }
    set status(_v) {
        this.#___private_status_3 = _v;
    }
    #___private_size_4 = (()=>{
        const _value = _init_size(this, 'default');
        _init_extra__init_size(this);
        return _value;
    })();
    get size() {
        return this.#___private_size_4;
    }
    set size(_v) {
        this.#___private_size_4 = _v;
    }
    #___private_closable_5 = (()=>{
        const _value = _init_closable(this, false);
        _init_extra__init_closable(this);
        return _value;
    })();
    get closable() {
        return this.#___private_closable_5;
    }
    set closable(_v) {
        this.#___private_closable_5 = _v;
    }
    #___private_bordered_6 = (()=>{
        const _value = _init_bordered(this, false);
        _init_extra__init_bordered(this);
        return _value;
    })();
    get bordered() {
        return this.#___private_bordered_6;
    }
    set bordered(_v) {
        this.#___private_bordered_6 = _v;
    }
    #___private_iconSlot_7 = (()=>{
        const _value = _init_iconSlot(this, null);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_7;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_7 = _v;
    }
    #___private_hasIconSlot_8 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_8;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_8 = _v;
    }
    firstUpdated() {
        this.syncSlotState();
    }
    get normalizedColor() {
        return (this.color ?? '').trim();
    }
    get normalizedBgColor() {
        return (this.bgColor ?? '').trim();
    }
    get customColorStyle() {
        const styles = {};
        if (this.normalizedColor) {
            styles['--ve-tag-text-color'] = this.normalizedColor;
            styles['--ve-tag-border-color'] = this.normalizedColor;
            styles['--ve-tag-current-text-color'] = this.normalizedColor;
            styles['--ve-tag-current-close-color'] = this.normalizedColor;
            if (this.bordered) styles['--ve-tag-current-border-color'] = this.normalizedColor;
        }
        if (this.normalizedBgColor) {
            styles['--ve-tag-background'] = this.normalizedBgColor;
            styles['--ve-tag-current-background'] = this.normalizedBgColor;
        }
        return styles;
    }
    get normalizedStatus() {
        return 'success' === this.status || 'warning' === this.status || 'error' === this.status ? this.status : '';
    }
    get statusIcon() {
        const status = this.normalizedStatus;
        return status ? TAG_STATUS_ICONS[status] : void 0;
    }
    get shouldRenderIcon() {
        return Boolean(this.statusIcon) || this.hasIconSlot || hasNamedSlottedContent(this, 'icon');
    }
    syncSlotState() {
        this.hasIconSlot = hasAssignedSlotContent(this.iconSlot) || hasNamedSlottedContent(this, 'icon');
    }
    handleIconSlotChange() {
        this.syncSlotState();
    }
    emitClose(sourceEvent) {
        return this.dispatchEvent(new CustomEvent('ve-close', {
            detail: {
                sourceEvent
            },
            bubbles: true,
            composed: true,
            cancelable: true
        }));
    }
    handleCloseClick(event) {
        event.preventDefault();
        event.stopPropagation();
        this.emitClose(event);
    }
    renderLabelContent() {
        return html`<span class="ve-tag-label" part="label"><slot></slot></span>`;
    }
    renderIcon() {
        if (!this.shouldRenderIcon) return nothing;
        const icon = this.statusIcon;
        return html`
      <span class="ve-tag-icon" part="icon" aria-hidden="true">
        <slot name="icon" @slotchange=${this.handleIconSlotChange}></slot>
        ${this.hasIconSlot || !icon ? nothing : html`<ve-icon .definition=${icon}></ve-icon>`}
      </span>
    `;
    }
    renderCloseButton() {
        if (!this.closable) return nothing;
        return html`
      <button
        class="ve-tag-close"
        part="close"
        type="button"
        aria-label="Close tag"
        @click=${this.handleCloseClick}
      >
        <svg viewBox="0 0 16 16" aria-hidden="true" focusable="false">
          <path
            d="M4.47 4.47a.75.75 0 0 1 1.06 0L8 6.94l2.47-2.47a.75.75 0 1 1 1.06 1.06L9.06 8l2.47 2.47a.75.75 0 1 1-1.06 1.06L8 9.06l-2.47 2.47a.75.75 0 1 1-1.06-1.06L6.94 8 4.47 5.53a.75.75 0 0 1 0-1.06Z"
            fill="currentColor"
          ></path>
        </svg>
      </button>
    `;
    }
    renderControl() {
        return html`
      <span class="ve-tag-control" part="control">
        ${this.renderIcon()}
        ${this.renderLabelContent()}
      </span>
    `;
    }
    render() {
        return html`
      <span
        class="ve-tag"
        part="container"
        style=${styleMap(this.customColorStyle)}
      >
        ${this.renderControl()}
        ${this.renderCloseButton()}
      </span>
    `;
    }
}
defineCustomElement(VE_TAG_TAG, VeTagElement);
export { VE_TAG_TAG, VeTagElement };
