// @ve-design/web/dist/ve-thinking.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { createNativeTitleAttributeController, setComponentTitle } from "./24968.js";
import { hasAssignedSlotContent, hasNamedSlottedContent } from "./89163.js";
const veThinkingTokenDefaults = css`
  :host {
    --ve-thinking-font-family: var(--font-sans);
    --ve-thinking-font-size: var(--text-body);
    --ve-thinking-font-weight: var(--font-weight-normal);
    --ve-thinking-line-height: var(--line-height-body);

    --ve-thinking-text-color: var(--color-text-tertiary);
    --ve-thinking-label-color: var(--color-text-tertiary);
    --ve-thinking-line-color: var(--color-border-default);
    --ve-thinking-focus-color: var(--color-border-strong);
    --ve-thinking-scrollbar-color: var(--color-border-default);
    --ve-thinking-indicator-color: var(--color-text-tertiary);
    --ve-thinking-indicator-secondary-color: var(--color-bg-info);
    --ve-thinking-completed-color: var(--color-text-tertiary);
    --ve-thinking-shimmer-base-color: var(--color-text-tertiary);
    --ve-thinking-shimmer-highlight-color: var(--color-bg-muted);
    --ve-thinking-hover-label-color: color-mix(
      var(--ve-thinking-label-color) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );

    --ve-thinking-block-gap: var(--space-xxs);
    --ve-thinking-header-gap: var(--space-xxxs);
    --ve-thinking-body-gap: var(--space-xxxs);
    --ve-thinking-content-gap: 0;
    --ve-thinking-scrollbar-gutter: var(--space-xs);
    --ve-thinking-icon-size: 20px;
    --ve-thinking-caret-size: 16px;
    --ve-thinking-caret-stroke-width: var(--stroke-weight-strong);
    --ve-thinking-step-width: var(--ve-thinking-icon-size);
    --ve-thinking-step-line-width: var(--stroke-weight-base);
    --ve-thinking-focus-width: var(--stroke-weight-bold);
    --ve-thinking-fade-size: 51px;
    --ve-thinking-scrollbar-size: var(--space-xxxs);
    --ve-thinking-radius: var(--radius-xs);

    --ve-thinking-motion-duration: var(--motion-base);
    --ve-thinking-motion-duration-fast: var(--motion-fast);
    --ve-thinking-shimmer-duration: 1800ms;
    --ve-thinking-collapse-max-height: 80vh;
    --ve-thinking-current-label-color: var(--ve-thinking-label-color);
    --ve-thinking-current-max-height: 489px;
  }

  :host([loading]) {
    --ve-thinking-current-label-color: var(--ve-thinking-shimmer-base-color);
  }
`;
const token = veThinkingTokenDefaults;
const ve_thinking = [
    componentReset,
    token,
    css`
    :host {
      display: flex;
      flex-direction: column;
      align-items: stretch;
      gap: var(--ve-thinking-block-gap);
      width: 100%;
      box-sizing: border-box;
      color: var(--ve-thinking-text-color);
      font-family: var(--ve-thinking-font-family);
      font-size: var(--ve-thinking-font-size);
      font-weight: var(--ve-thinking-font-weight);
      line-height: var(--ve-thinking-line-height);
      transition: gap var(--ve-thinking-motion-duration) ease;
    }

    :host(:not([expanded])),
    :host([loading]:not([title]):empty) {
      gap: 0;
    }

    :host([hidden]) {
      display: none;
    }

    .ve-thinking-header-row {
      display: flex;
      align-items: center;
      width: 100%;
      gap: var(--ve-thinking-header-gap);
    }

    .ve-thinking-header-extra {
      display: inline-flex;
      align-items: center;
      gap: var(--ve-thinking-header-gap);
      margin-left: auto;
    }

    .ve-thinking-header {
      appearance: none;
      display: inline-flex;
      align-items: center;
      align-self: flex-start;
      gap: var(--ve-thinking-header-gap);
      margin: 0;
      padding: 0;
      border: 0;
      background: transparent;
      color: inherit;
      font: inherit;
      user-select: none;
    }

    .ve-thinking-header-clickable {
      cursor: pointer;
    }

    .ve-thinking-header-clickable:hover .ve-thinking-label {
      color: var(--ve-thinking-hover-label-color);
    }

    .ve-thinking-header-clickable:focus-visible {
      outline: var(--ve-thinking-focus-width) solid
        var(--ve-thinking-focus-color);
      outline-offset: var(--space-xxxs);
      border-radius: var(--ve-thinking-radius);
    }

    .ve-thinking-indicator,
    .ve-thinking-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      width: var(--ve-thinking-icon-size);
      height: var(--ve-thinking-icon-size);
      color: var(--ve-thinking-indicator-color);
    }

    .ve-thinking-blob,
    .ve-thinking-icon {
      display: block;
      width: 100%;
      height: 100%;
    }

    .ve-thinking-icon {
      object-fit: contain;
    }

    .ve-thinking-indicator ::slotted(*) {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      max-width: 100%;
      max-height: 100%;
      line-height: 1;
      flex: none;
    }

    .ve-thinking-label {
      color: var(--ve-thinking-current-label-color);
      font-size: var(--ve-thinking-font-size);
      font-weight: var(--ve-thinking-font-weight);
      line-height: var(--ve-thinking-line-height);
      white-space: nowrap;
      transition: color var(--ve-thinking-motion-duration-fast) ease;
    }

    :host([loading]) .ve-thinking-label {
      background-image: linear-gradient(
        90deg,
        var(--ve-thinking-shimmer-base-color) 6%,
        var(--ve-thinking-shimmer-highlight-color) 39%,
        var(--ve-thinking-shimmer-base-color) 72%
      );
      background-size: 220% 100%;
      background-clip: text;
      -webkit-background-clip: text;
      -webkit-text-fill-color: transparent;
      color: transparent;
      animation: ve-thinking-shimmer var(--ve-thinking-shimmer-duration) linear
        infinite;
    }

    .ve-thinking-caret {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      width: var(--ve-thinking-caret-size);
      height: var(--ve-thinking-line-height);
      color: var(--ve-thinking-text-color);
      transition: transform var(--ve-thinking-motion-duration-fast) ease;
    }

    .ve-thinking-caret svg,
    .ve-thinking-caret ve-icon {
      width: var(--ve-thinking-caret-size);
      height: var(--ve-thinking-caret-size);
      --ve-icon-size: var(--ve-thinking-caret-size);
    }

    :host(:not([expanded])) .ve-thinking-caret {
      transform: rotate(-90deg);
    }

    :host([loading]:not([title]):empty) .ve-thinking-caret,
    :host([loading]:not([title]):empty) .ve-thinking-body {
      display: none;
    }

    .ve-thinking-body {
      position: relative;
      display: flex;
      align-items: stretch;
      gap: var(--ve-thinking-body-gap);
      width: 100%;
      min-width: 0;
      max-height: var(--ve-thinking-collapse-max-height);
      overflow: hidden;
      opacity: 1;
      transition:
        max-height var(--ve-thinking-motion-duration) ease,
        opacity var(--ve-thinking-motion-duration-fast) ease;
    }

    :host(:not([expanded])) .ve-thinking-body {
      max-height: 0;
      opacity: 0;
    }

    .ve-thinking-steps {
      display: flex;
      align-items: stretch;
      justify-content: center;
      flex: 0 0 var(--ve-thinking-step-width);
    }

    .ve-thinking-step-line {
      display: block;
      align-self: stretch;
      width: var(--ve-thinking-step-line-width);
      min-height: var(--space-s);
      border-radius: var(--radius-full);
      background: var(--ve-thinking-line-color);
    }

    .ve-thinking-content {
      flex: 1 1 0;
      min-width: 0;
      color: var(--ve-thinking-text-color);
      word-break: break-word;
    }

    ::slotted(*) {
      margin: 0;
    }

    ::slotted(*:not(:first-child)) {
      margin-top: var(--ve-thinking-content-gap);
    }

    .ve-thinking-body-scrollable .ve-thinking-content {
      max-height: var(--ve-thinking-current-max-height);
      overflow-y: auto;
      overflow-x: hidden;
      padding-right: var(--ve-thinking-scrollbar-gutter);
      scrollbar-width: thin;
      scrollbar-color: transparent transparent;
    }

    .ve-thinking-body-scrollable .ve-thinking-content::-webkit-scrollbar {
      width: var(--ve-thinking-scrollbar-size);
    }

    .ve-thinking-body-scrollable .ve-thinking-content::-webkit-scrollbar-track {
      background: transparent;
    }

    .ve-thinking-body-scrollable .ve-thinking-content::-webkit-scrollbar-thumb {
      border-radius: var(--radius-full);
      background: transparent;
    }

    .ve-thinking-body-scrollable:hover .ve-thinking-content,
    .ve-thinking-body-scrollable:focus-within .ve-thinking-content {
      scrollbar-color: var(--ve-thinking-scrollbar-color) transparent;
    }

    .ve-thinking-body-scrollable:hover
      .ve-thinking-content::-webkit-scrollbar-thumb,
    .ve-thinking-body-scrollable:focus-within
      .ve-thinking-content::-webkit-scrollbar-thumb {
      background: var(--ve-thinking-scrollbar-color);
    }

    .ve-thinking-body-scrollable .ve-thinking-content {
      --ve-thinking-scroll-mask-start: black;
      --ve-thinking-scroll-mask-end: black;
      -webkit-mask-image:
        linear-gradient(
          to bottom,
          var(--ve-thinking-scroll-mask-start) 0,
          black var(--ve-thinking-fade-size),
          black calc(100% - var(--ve-thinking-fade-size)),
          var(--ve-thinking-scroll-mask-end) 100%
        ),
        linear-gradient(
          to right,
          transparent calc(100% - var(--ve-thinking-scrollbar-gutter)),
          black calc(100% - var(--ve-thinking-scrollbar-gutter))
        );
      -webkit-mask-composite: source-over;
      mask-image:
        linear-gradient(
          to bottom,
          var(--ve-thinking-scroll-mask-start) 0,
          black var(--ve-thinking-fade-size),
          black calc(100% - var(--ve-thinking-fade-size)),
          var(--ve-thinking-scroll-mask-end) 100%
        ),
        linear-gradient(
          to right,
          transparent calc(100% - var(--ve-thinking-scrollbar-gutter)),
          black calc(100% - var(--ve-thinking-scrollbar-gutter))
        );
      mask-composite: add;
    }

    .ve-thinking-body-scrollable[data-scroll-start='true'] .ve-thinking-content {
      --ve-thinking-scroll-mask-start: transparent;
    }

    .ve-thinking-body-scrollable[data-scroll-end='true'] .ve-thinking-content {
      --ve-thinking-scroll-mask-end: transparent;
    }

    @keyframes ve-thinking-shimmer {
      0% {
        background-position: 100% 0;
      }
      100% {
        background-position: -100% 0;
      }
    }

    @media (prefers-reduced-motion: reduce) {
      :host {
        transition: none;
      }

      :host([loading]) .ve-thinking-indicator,
      :host([loading]) .ve-thinking-label,
      :host([loading]) .ve-thinking-label {
        background: none;
        -webkit-text-fill-color: var(--ve-thinking-current-label-color);
        color: var(--ve-thinking-current-label-color);
      }

      .ve-thinking-caret,
      .ve-thinking-body,
      .ve-thinking-label {
        transition: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _init_loading, _init_extra__init_loading, _init_expanded, _init_extra__init_expanded, _init_maxHeight, _init_extra__init_maxHeight, _init_autoScroll, _init_extra__init_autoScroll, _init_scrollContainer, _init_extra__init_scrollContainer, _init_iconSlot, _init_extra__init_iconSlot, _init_loadingCanvas, _init_extra__init_loadingCanvas, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_canScrollStart, _init_extra__init_canScrollStart, _init_canScrollEnd, _init_extra__init_canScrollEnd;
const VE_THINKING_TAG = 've-thinking';
const VE_THINKING_TOGGLE_EVENT = 've-thinking-toggle';
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
const THINKING_BLOB_PATHS = [
    'M-55.363 -172.052 C-55.363 -172.052 -187.257 35.800 -187.257 35.800 C-210.788 72.884 -222.636 94.054 -219.327 110.707 C-215.783 128.545 -194.850 141.200 -152.260 162.678 C-152.260 162.678 -88.770 194.695 -88.770 194.695 C-52.787 212.841 -21.75 228.250 0 228.593 C22.297 228.754 43.982 217.302 88.812 194.695 C88.812 194.695 152.301 162.678 152.301 162.678 C193.026 142.141 214.103 129.850 218.836 112.979 C223.568 96.108 211.955 74.657 187.298 35.800 C187.298 35.800 55.404 -172.052 55.404 -172.052 C39.424 -197.236 26.805 -213.190 15.906 -221.516 C4.162 -230.486 -5.644 -230.534 -15.335 -223.896 C-27.340 -215.673 -39.426 -197.167 -55.363 -172.052 Z',
    'M-45.863 -166.739 C-57.999 -139.75 -112.480 -5.800 -121.757 27.800 C-129.750 56.750 -138.905 81.915 -135.077 101.707 C-131.250 121.499 -109 141.000 -92.010 156.928 C-77.945 170.113 -55.421 191.468 -44.770 201.445 C-29.500 215.750 -17.000 228.686 0 228.593 C17.000 228.500 32.934 218.956 52.812 208.695 C74.500 197.500 116.991 169.305 130.301 159.928 C155.750 142.000 163.250 129.000 166.086 119.479 C173.191 95.627 166.347 84.100 158.548 59.300 C150.749 34.499 57.000 -154.000 48.904 -170.739 C40.984 -187.118 26.836 -213.378 15.937 -221.703 C4.194 -230.673 -5.613 -230.722 -15.303 -224.084 C-27.309 -215.861 -34.638 -191.701 -45.863 -166.739 Z',
    'M-20.295 -115.546 C-14.655 -79.764 -13.489 -50.044 -1.792 -3.903 C7.500 32.750 5.750 29.999 24.071 75.053 C33.437 98.086 37.000 108.500 37.460 131.081 C37.777 146.585 20.000 188.250 16.011 202.240 C11.738 217.230 0.822 232.764 -8.832 234.225 C-22.946 236.360 -44.445 215.440 -54.790 200.223 C-70.750 176.750 -81.424 162.789 -76.786 137.611 C-73.981 122.386 -48.398 56.705 -43.330 43.433 C-23.500 -8.500 -17.427 -21.868 -11.031 -49.446 C-7.250 -65.750 -6.601 -73.671 3.029 -118.359 C11.250 -156.500 15.260 -192.000 6.645 -216.221 C0.684 -232.978 -15.023 -233.937 -21.397 -209.733 C-26.989 -188.500 -26.750 -156.500 -20.295 -115.546 Z',
    'M22.625 -102.380 C43.975 -62.331 112.646 7.407 139.092 33.852 C166.295 61.056 202.846 80.380 202.846 111.153 C202.846 131.030 187.975 140.564 159.116 156.944 C134.604 170.857 90.208 195.042 65.030 207.006 C50.951 213.695 31.901 226.294 2.453 226.294 C-26.993 226.294 -50.846 215.656 -81.177 203.877 C-95.673 198.247 -142.074 176.475 -166.074 162.690 C-195.205 145.957 -222.085 126.800 -222.085 109.524 C-222.085 87.263 -195.154 64.272 -186.012 53.398 C-145.374 5.067 -84.122 -57.325 -35.533 -114.159 C-7.602 -146.830 4.147 -159.975 4.147 -208.134 C4.147 -229.779 -19.484 -239.276 -23.680 -207.251 C-27.877 -175.227 1.276 -142.429 22.625 -102.380 Z'
];
_dec = property({
    type: Boolean,
    reflect: true
}), _dec1 = property({
    type: Boolean,
    reflect: true
}), _dec2 = property({
    attribute: 'max-height',
    reflect: true
}), _dec3 = property({
    attribute: 'auto-scroll',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec4 = query('.ve-thinking-content'), _dec5 = query('slot[name="icon"]'), _dec6 = query('.ve-thinking-blob'), _dec7 = state(), _dec8 = state(), _dec9 = state();
class VeThinkingElement extends VeElement {
    static{
        ({ e: [_init_loading, _init_extra__init_loading, _init_expanded, _init_extra__init_expanded, _init_maxHeight, _init_extra__init_maxHeight, _init_autoScroll, _init_extra__init_autoScroll, _init_scrollContainer, _init_extra__init_scrollContainer, _init_iconSlot, _init_extra__init_iconSlot, _init_loadingCanvas, _init_extra__init_loadingCanvas, _init_hasIconSlot, _init_extra__init_hasIconSlot, _init_canScrollStart, _init_extra__init_canScrollStart, _init_canScrollEnd, _init_extra__init_canScrollEnd] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "loading"
            ],
            [
                _dec1,
                1,
                "expanded"
            ],
            [
                _dec2,
                1,
                "maxHeight"
            ],
            [
                _dec3,
                1,
                "autoScroll"
            ],
            [
                _dec4,
                1,
                "scrollContainer"
            ],
            [
                _dec5,
                1,
                "iconSlot"
            ],
            [
                _dec6,
                1,
                "loadingCanvas"
            ],
            [
                _dec7,
                1,
                "hasIconSlot"
            ],
            [
                _dec8,
                1,
                "canScrollStart"
            ],
            [
                _dec9,
                1,
                "canScrollEnd"
            ]
        ]));
    }
    static styles = ve_thinking;
    get title() {
        return this.titleText;
    }
    set title(value) {
        setComponentTitle(this, this.titleText, value, (title)=>{
            this.titleText = title;
        });
    }
    #___private_loading_1 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_1;
    }
    set loading(_v) {
        this.#___private_loading_1 = _v;
    }
    #___private_expanded_2 = (()=>{
        const _value = _init_expanded(this, true);
        _init_extra__init_expanded(this);
        return _value;
    })();
    get expanded() {
        return this.#___private_expanded_2;
    }
    set expanded(_v) {
        this.#___private_expanded_2 = _v;
    }
    #___private_maxHeight_3 = (()=>{
        const _value = _init_maxHeight(this, '');
        _init_extra__init_maxHeight(this);
        return _value;
    })();
    get maxHeight() {
        return this.#___private_maxHeight_3;
    }
    set maxHeight(_v) {
        this.#___private_maxHeight_3 = _v;
    }
    #___private_autoScroll_4 = (()=>{
        const _value = _init_autoScroll(this, true);
        _init_extra__init_autoScroll(this);
        return _value;
    })();
    get autoScroll() {
        return this.#___private_autoScroll_4;
    }
    set autoScroll(_v) {
        this.#___private_autoScroll_4 = _v;
    }
    #___private_scrollContainer_5 = (()=>{
        const _value = _init_scrollContainer(this, null);
        _init_extra__init_scrollContainer(this);
        return _value;
    })();
    get scrollContainer() {
        return this.#___private_scrollContainer_5;
    }
    set scrollContainer(_v) {
        this.#___private_scrollContainer_5 = _v;
    }
    #___private_iconSlot_6 = (()=>{
        const _value = _init_iconSlot(this, null);
        _init_extra__init_iconSlot(this);
        return _value;
    })();
    get iconSlot() {
        return this.#___private_iconSlot_6;
    }
    set iconSlot(_v) {
        this.#___private_iconSlot_6 = _v;
    }
    #___private_loadingCanvas_7 = (()=>{
        const _value = _init_loadingCanvas(this, null);
        _init_extra__init_loadingCanvas(this);
        return _value;
    })();
    get loadingCanvas() {
        return this.#___private_loadingCanvas_7;
    }
    set loadingCanvas(_v) {
        this.#___private_loadingCanvas_7 = _v;
    }
    #___private_hasIconSlot_8 = (()=>{
        const _value = _init_hasIconSlot(this, false);
        _init_extra__init_hasIconSlot(this);
        return _value;
    })();
    get hasIconSlot() {
        return this.#___private_hasIconSlot_8;
    }
    set hasIconSlot(_v) {
        this.#___private_hasIconSlot_8 = _v;
    }
    #___private_canScrollStart_9 = (()=>{
        const _value = _init_canScrollStart(this, false);
        _init_extra__init_canScrollStart(this);
        return _value;
    })();
    get canScrollStart() {
        return this.#___private_canScrollStart_9;
    }
    set canScrollStart(_v) {
        this.#___private_canScrollStart_9 = _v;
    }
    #___private_canScrollEnd_10 = (()=>{
        const _value = _init_canScrollEnd(this, false);
        _init_extra__init_canScrollEnd(this);
        return _value;
    })();
    get canScrollEnd() {
        return this.#___private_canScrollEnd_10;
    }
    set canScrollEnd(_v) {
        this.#___private_canScrollEnd_10 = _v;
    }
    scrollResizeObserver;
    contentMutationObserver;
    scrollFrame = 0;
    blobFrame = 0;
    blobLastTime = 0;
    blobAnimationFrame = 0;
    shouldStickToBottom = true;
    titleText = '';
    constructor(){
        super();
        createNativeTitleAttributeController(this);
    }
    connectedCallback() {
        super.connectedCallback();
        this.applyMaxHeightToken();
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.syncSlotState();
        this.observeScrollableContent();
        this.scheduleScrollStateUpdate();
    }
    updated(_changedProperties) {
        super.updated(_changedProperties);
        if (_changedProperties.has('maxHeight')) this.applyMaxHeightToken();
        if (_changedProperties.has('maxHeight') || _changedProperties.has('expanded') || _changedProperties.has('loading')) {
            this.observeScrollableContent();
            this.scheduleScrollStateUpdate();
        }
        this.syncLoadingBlobAnimation();
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        this.scrollResizeObserver?.disconnect();
        this.contentMutationObserver?.disconnect();
        window.cancelAnimationFrame(this.scrollFrame);
        this.stopLoadingBlobAnimation();
    }
    toggle() {
        this.expanded = !this.expanded;
        this.dispatchEvent(new CustomEvent(VE_THINKING_TOGGLE_EVENT, {
            detail: {
                expanded: this.expanded
            },
            bubbles: true,
            composed: true
        }));
    }
    applyMaxHeightToken() {
        if (this.maxHeight) return void this.style.setProperty('--ve-thinking-current-max-height', this.maxHeight);
        this.style.removeProperty('--ve-thinking-current-max-height');
    }
    isNearBottom(scrollContainer) {
        const tolerance = 8;
        return scrollContainer.scrollTop + scrollContainer.clientHeight >= scrollContainer.scrollHeight - tolerance;
    }
    observeScrollableContent() {
        this.scrollResizeObserver?.disconnect();
        this.contentMutationObserver?.disconnect();
        const scrollContainer = this.scrollContainer;
        if (!scrollContainer || !this.maxHeight || !this.expanded) return;
        this.scrollResizeObserver = new ResizeObserver(()=>{
            this.handleContentChange();
        });
        this.scrollResizeObserver.observe(scrollContainer);
        Array.from(scrollContainer.children).forEach((child)=>{
            this.scrollResizeObserver?.observe(child);
        });
        this.contentMutationObserver = new MutationObserver(()=>{
            this.handleContentChange();
        });
        this.contentMutationObserver.observe(this, {
            childList: true,
            subtree: true,
            characterData: true
        });
    }
    handleContentChange() {
        const scrollContainer = this.scrollContainer;
        if (scrollContainer && this.autoScroll && this.shouldStickToBottom) scrollContainer.scrollTop = scrollContainer.scrollHeight;
        this.scheduleScrollStateUpdate();
    }
    scheduleScrollStateUpdate() {
        window.cancelAnimationFrame(this.scrollFrame);
        this.scrollFrame = window.requestAnimationFrame(()=>{
            this.updateScrollState();
        });
    }
    updateScrollState() {
        const scrollContainer = this.scrollContainer;
        if (!scrollContainer || !this.maxHeight || !this.expanded) {
            this.canScrollStart = false;
            this.canScrollEnd = false;
            return;
        }
        const tolerance = 1;
        const maxScrollTop = scrollContainer.scrollHeight - scrollContainer.clientHeight;
        this.canScrollStart = scrollContainer.scrollTop > tolerance;
        this.canScrollEnd = maxScrollTop - scrollContainer.scrollTop > tolerance;
        this.shouldStickToBottom = this.isNearBottom(scrollContainer);
    }
    handleScroll() {
        this.updateScrollState();
    }
    handleHeaderClick() {
        this.toggle();
    }
    handleHeaderKeydown(event) {
        if ('Enter' === event.key || ' ' === event.key) {
            event.preventDefault();
            this.toggle();
        }
    }
    syncSlotState() {
        this.hasIconSlot = hasAssignedSlotContent(this.iconSlot) || hasNamedSlottedContent(this, 'icon');
    }
    handleIconSlotChange() {
        this.syncSlotState();
        this.syncLoadingBlobAnimation();
    }
    syncLoadingBlobAnimation() {
        if (this.hasIconSlot) return void this.stopLoadingBlobAnimation();
        const canvas = this.loadingCanvas;
        if (!canvas) return void this.stopLoadingBlobAnimation();
        if (this.loading) return void this.startLoadingBlobAnimation();
        this.stopLoadingBlobAnimation();
        this.blobFrame = 0;
        this.drawLoadingBlobFrame(canvas);
    }
    startLoadingBlobAnimation() {
        if (this.blobAnimationFrame) return;
        const canvas = this.loadingCanvas;
        if (!canvas) return;
        this.drawLoadingBlobFrame(canvas);
        this.blobAnimationFrame = window.requestAnimationFrame((timestamp)=>{
            this.tickLoadingBlob(timestamp);
        });
    }
    stopLoadingBlobAnimation() {
        window.cancelAnimationFrame(this.blobAnimationFrame);
        this.blobAnimationFrame = 0;
        this.blobLastTime = 0;
    }
    tickLoadingBlob(timestamp) {
        const canvas = this.loadingCanvas;
        if (!canvas || !this.loading || this.hasIconSlot) return void this.stopLoadingBlobAnimation();
        if (!this.blobLastTime || timestamp - this.blobLastTime >= 160) {
            this.drawLoadingBlobFrame(canvas);
            this.blobFrame = (this.blobFrame + 1) % THINKING_BLOB_PATHS.length;
            this.blobLastTime = timestamp;
        }
        this.blobAnimationFrame = window.requestAnimationFrame((nextTimestamp)=>{
            this.tickLoadingBlob(nextTimestamp);
        });
    }
    createLoadingBlobGradient(context) {
        const gradient = context.createConicGradient(105 * Math.PI / 180, 10, 10);
        gradient.addColorStop(0, '#006aff');
        gradient.addColorStop(0.29, '#786aff');
        gradient.addColorStop(0.46, '#ffcc6b');
        gradient.addColorStop(0.5, '#ffcc6b');
        gradient.addColorStop(0.8, '#006aff');
        gradient.addColorStop(1, '#006aff');
        return gradient;
    }
    drawLoadingBlobFrame(canvas) {
        const context = canvas.getContext('2d');
        if (!context) return;
        const size = 20;
        const dpr = window.devicePixelRatio || 1;
        const width = size * dpr;
        if (canvas.width !== width || canvas.height !== width) {
            canvas.width = width;
            canvas.height = width;
        }
        const scale = 0.0015 * size;
        context.setTransform(dpr, 0, 0, dpr, 0, 0);
        context.clearRect(0, 0, size, size);
        context.fillStyle = this.createLoadingBlobGradient(context);
        context.fillRect(0, 0, size, size);
        context.globalCompositeOperation = 'destination-in';
        context.translate(size / 2, size / 2);
        context.scale(scale, scale);
        context.lineWidth = 2 / scale;
        context.lineCap = 'round';
        context.lineJoin = 'round';
        context.strokeStyle = '#fff';
        context.stroke(new Path2D(THINKING_BLOB_PATHS[this.blobFrame]));
        context.globalCompositeOperation = 'source-over';
        context.setTransform(1, 0, 0, 1, 0, 0);
    }
    renderThinkingIcon(part) {
        const hasIconSlot = this.hasIconSlot || hasNamedSlottedContent(this, 'icon');
        return html`
      <span class="ve-thinking-indicator" part=${part} aria-hidden="true">
        <slot name="icon" @slotchange=${this.handleIconSlotChange}></slot>
        ${hasIconSlot ? nothing : html`<canvas
              class="ve-thinking-blob"
              width="20"
              height="20"
              aria-hidden="true"
            ></canvas>`}
      </span>
    `;
    }
    renderCaret() {
        return html`
      <span class="ve-thinking-caret" part="caret" aria-hidden="true">
        <svg viewBox="0 0 24 24" fill="none" aria-hidden="true">
          <path
            d="m19 9-7 7-7-7"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            style="stroke-width: var(--ve-thinking-caret-stroke-width);"
          ></path>
        </svg>
      </span>
    `;
    }
    renderHeaderLabel() {
        if (!this.title) return nothing;
        return html`<span class="ve-thinking-label" part="label"
      >${this.title}</span
    >`;
    }
    renderHeader() {
        return html`
      <div class="ve-thinking-header-row" part="header-row">
        <button
          class="ve-thinking-header ve-thinking-header-clickable"
          part="header"
          type="button"
          aria-expanded=${this.expanded ? 'true' : 'false'}
          @click=${this.handleHeaderClick}
          @keydown=${this.handleHeaderKeydown}
        >
          ${this.renderThinkingIcon(this.loading ? 'indicator' : 'icon')}
          ${this.renderHeaderLabel()} ${this.renderCaret()}
        </button>
        <span class="ve-thinking-header-extra" part="header-extra">
          <slot name="header-extra"></slot>
        </span>
      </div>
    `;
    }
    renderBody() {
        const bodyClass = classMap({
            've-thinking-body': true,
            've-thinking-body-scrollable': Boolean(this.maxHeight)
        });
        return html`
      <div
        class=${bodyClass}
        part="body"
        data-scroll-start=${String(this.canScrollStart)}
        data-scroll-end=${String(this.canScrollEnd)}
      >
        <div class="ve-thinking-steps" aria-hidden="true">
          <span class="ve-thinking-step-line" part="step-line"></span>
        </div>
        <div
          class="ve-thinking-content"
          part="content"
          @scroll=${this.handleScroll}
        >
          <slot @slotchange=${this.handleContentChange}></slot>
        </div>
      </div>
    `;
    }
    render() {
        return html`${this.renderHeader()}${this.renderBody()}`;
    }
}
defineCustomElement(VE_THINKING_TAG, VeThinkingElement);
export { VE_THINKING_TAG, VE_THINKING_TOGGLE_EVENT, VeThinkingElement };
