// @ve-design/web/dist/ve-table.js
import { createTable, flexRender, getCoreRowModel, getSortedRowModel } from "@tanstack/lit-table";
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { VeEmptyElement, VE_EMPTY_TAG } from "./77133.js";
import { arrowUpSmIcon } from "./icons/arrow-up-sm.js";
import { arrowDownSmIcon } from "./icons/arrow-down-sm.js";
import { downUpIcon } from "./icons/down-up.js";
const veTableTokenDefaults = css`
  :host {
    --ve-table-font-family: var(--font-sans);
    --ve-table-font-size: var(--text-body);
    --ve-table-font-weight-header: var(--font-weight-medium);
    --ve-table-line-height: var(--line-height-body);

    --ve-table-row-size: 48px;
    --ve-table-padding-inline: var(--space-xs);
    --ve-table-padding-block: 0px;

    --ve-table-radius: var(--radius-xl);
    --ve-table-border-width: var(--stroke-weight-base);
    --ve-table-border-color: var(--color-border-default);
    --ve-table-header-border-color: var(--color-border-default);
    --ve-table-row-border-color: var(--color-border-default);
    --ve-table-shadow: none;

    --ve-table-background: var(--color-bg-base);
    --ve-table-header-background: var(--color-bg-secondary);
    --ve-table-header-text: var(--color-text-primary);
    --ve-table-cell-text: var(--color-text-primary);
    --ve-table-cell-text-secondary: var(--color-text-secondary);
    --ve-table-empty-text: var(--color-text-tertiary);
    --ve-table-hover-background: color-mix(
      var(--ve-table-background) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-table-focus-ring-width: var(--stroke-2);
    --ve-table-focus-ring-color: color-mix(
      var(--color-text-accent) 24%,
      transparent
    );
    --ve-table-disabled-opacity: var(--state-disabled-ratio);
    --ve-table-transition-duration: var(--motion-fast);

    --ve-table-checkbox-size: 16px;
    --ve-table-checkbox-border-width: var(--stroke-weight-strong);
    --ve-table-checkbox-radius: var(--radius-xs);
    --ve-table-checkbox-border: var(--color-border-strong);
    --ve-table-checkbox-border-hover: var(--color-border-interactive);
    --ve-table-checkbox-selected-border: transparent;
    --ve-table-checkbox-background: transparent;
    --ve-table-checkbox-selected-background: var(--color-bg-primary);
    --ve-table-checkbox-selected-color: var(--color-text-foreground);
    --ve-table-checkbox-selected-hover-background: color-mix(
      var(--ve-table-checkbox-selected-background)
        var(--state-hover-solid-ratio),
      var(--state-hover-solid-mix)
    );
    --ve-table-checkbox-icon-size: 12px;
    --ve-table-checkbox-icon-stroke-width: var(--stroke-weight-bold);
    --ve-table-sort-icon-size: 14px;
    --ve-table-control-column-size: 40px;
    --ve-table-skeleton-size: 10px;
    --ve-table-skeleton-radius: var(--radius-md);
    --ve-table-skeleton-background: linear-gradient(
      90deg,
      var(--color-bg-surface) 0%,
      color-mix(var(--color-bg-surface) 55%, var(--color-bg-base)) 45%,
      var(--color-bg-surface) 100%
    );
    --ve-table-pin-shadow: var(--shadow-sm);
    --ve-table-pin-left-shadow:
      var(--ve-table-pin-shadow),
      8px 0 12px -12px rgb(from var(--color-text-primary) r g b / 24%);
    --ve-table-pin-right-shadow:
      var(--ve-table-pin-shadow),
      -8px 0 12px -12px rgb(from var(--color-text-primary) r g b / 24%);

    /* Runtime custom properties (internal) */
    --ve-table-pin-left-offset: 0px;
    --ve-table-pin-right-offset: 0px;
    --ve-table-scroll-x: auto;
    --ve-table-scroll-y: auto;
  }
`;
const token = veTableTokenDefaults;
const ve_table = [
    componentReset,
    token,
    css`
    :host {
      display: block;
      color: var(--ve-table-cell-text);
      font-family: var(--ve-table-font-family);
      font-size: var(--ve-table-font-size);
      line-height: var(--ve-table-line-height);
    }

    :host([hidden]) {
      display: none;
    }

    .ve-table {
      position: relative;
      display: flex;
      flex-direction: column;
      gap: var(--space-s);
      min-inline-size: 0;
    }

    .ve-table-scroll {
      position: relative;
      overflow: auto;
      border-radius: var(--ve-table-radius);
      border: var(--ve-table-border-width) solid transparent;
      background: var(--ve-table-background);
      box-shadow: var(--ve-table-shadow);
    }

    :host([bordered]) .ve-table-scroll {
      border-color: var(--ve-table-border-color);
    }

    .ve-table-native {
      inline-size: 100%;
      min-inline-size: 100%;
      border-collapse: separate;
      border-spacing: 0;
      table-layout: auto;
      background: var(--ve-table-background);
    }

    .ve-table-head-cell,
    .ve-table-cell,
    .ve-table-select-cell {
      box-sizing: border-box;
      block-size: var(--ve-table-row-size);
      padding: var(--ve-table-padding-block) var(--ve-table-padding-inline);
      border-block-end: var(--ve-table-border-width) solid
        var(--ve-table-row-border-color);
      text-align: left;
      vertical-align: middle;
      background: var(--ve-table-background);
      color: var(--ve-table-cell-text);
    }

    .ve-table-head-cell {
      position: sticky;
      inset-block-start: 0;
      z-index: 2;
      background: var(--ve-table-header-background);
      color: var(--ve-table-header-text);
      font-weight: var(--ve-table-font-weight-header);
      border-block-end-color: var(--ve-table-header-border-color);
    }

    .ve-table-select-cell {
      inline-size: var(--ve-table-control-column-size);
      min-inline-size: var(--ve-table-control-column-size);
      max-inline-size: var(--ve-table-control-column-size);
      padding-inline: 0;
      text-align: center;
      display: table-cell;
      vertical-align: middle;
      line-height: 1;
    }

    .ve-table-cell-align-center,
    .ve-table-head-cell-align-center {
      text-align: center;
    }

    .ve-table-cell-align-right,
    .ve-table-head-cell-align-right {
      text-align: right;
    }

    :host([vertical-line]) .ve-table-head-cell,
    :host([vertical-line]) .ve-table-cell,
    :host([vertical-line]) .ve-table-select-cell {
      border-inline-end: var(--ve-table-border-width) solid
        var(--ve-table-border-color);
    }

    :host([vertical-line]) .ve-table-head-row > :last-child,
    :host([vertical-line]) .ve-table-row > :last-child {
      border-inline-end: none;
    }

    tbody tr:last-child > .ve-table-cell,
    tbody tr:last-child > .ve-table-select-cell {
      border-block-end: none;
    }

    .ve-table-row {
      transition: background-color var(--ve-table-transition-duration) ease;
    }

    :host([hover]) .ve-table-row:hover .ve-table-cell,
    :host([hover]) .ve-table-row:hover .ve-table-select-cell {
      background: var(--ve-table-hover-background);
    }

    .ve-table-selection,
    .ve-table-sort-button {
      font: inherit;
    }

    .ve-table-selection,
    .ve-table-sort-button {
      border: none;
      background: transparent;
      color: inherit;
    }

    .ve-table-sort-button {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: var(--space-xxxs);
      border-radius: var(--radius-sm);
      cursor: pointer;
      transition:
        background-color var(--ve-table-transition-duration) ease,
        color var(--ve-table-transition-duration) ease,
        opacity var(--ve-table-transition-duration) ease;
    }

    .ve-table-selection {
      margin-inline: auto;
      vertical-align: middle;
    }

    .ve-table-selection {
      position: relative;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-table-checkbox-size);
      block-size: var(--ve-table-checkbox-size);
      padding: 0;
      border: var(--ve-table-checkbox-border-width) solid
        var(--ve-table-checkbox-border);
      border-radius: var(--ve-table-checkbox-radius);
      background: var(--ve-table-checkbox-background);
      color: var(--ve-table-checkbox-selected-color);
      cursor: pointer;
      transition:
        background-color var(--ve-table-transition-duration) ease,
        border-color var(--ve-table-transition-duration) ease,
        box-shadow var(--ve-table-transition-duration) ease,
        opacity var(--ve-table-transition-duration) ease;
    }

    .ve-table-selection-checked,
    .ve-table-selection-indeterminate {
      border-color: var(--ve-table-checkbox-selected-border);
      background: var(--ve-table-checkbox-selected-background);
    }

    .ve-table-selection:hover:not(.ve-table-selection-checked):not(
        .ve-table-selection-indeterminate
      ) {
      border-color: var(--ve-table-checkbox-border-hover);
    }

    .ve-table-selection-checked:hover,
    .ve-table-selection-indeterminate:hover {
      background: var(--ve-table-checkbox-selected-hover-background);
    }

    .ve-table-selection-icon {
      position: absolute;
      inset-block-start: 50%;
      inset-inline-start: 50%;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-table-checkbox-icon-size);
      block-size: var(--ve-table-checkbox-icon-size);
      --ve-icon-size: var(--ve-table-checkbox-icon-size);
      opacity: 0;
      transform: translate(-50%, -50%);
      pointer-events: none;
    }

    .ve-table-selection-checked .ve-table-selection-icon-check,
    .ve-table-selection-indeterminate .ve-table-selection-icon-minus {
      opacity: 1;
    }

    .ve-table-sort-button {
      inline-size: 100%;
      justify-content: inherit;
      padding: 0;
      text-align: inherit;
    }

    .ve-table-sort-button:hover {
      background: color-mix(
        var(--ve-table-background) var(--state-hover-subtle-ratio),
        var(--state-hover-subtle-mix)
      );
    }

    .ve-table-sort-button:focus-visible,
    .ve-table-selection:focus-visible {
      outline: none;
      box-shadow: 0 0 0 var(--ve-table-focus-ring-width)
        var(--ve-table-focus-ring-color);
    }

    .ve-table-selection[disabled] {
      opacity: var(--ve-table-disabled-opacity);
      cursor: not-allowed;
    }

    .ve-table-header-content,
    .ve-table-cell-content,
    .ve-table-empty {
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      min-inline-size: 0;
    }

    .ve-table-header-content {
      justify-content: inherit;
      inline-size: 100%;
    }

    .ve-table-cell-content {
      color: inherit;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-table-cell-content-wrap {
      white-space: normal;
    }

    .ve-table-sort-indicator {
      display: inline-flex;
      inline-size: var(--ve-table-sort-icon-size);
      block-size: var(--ve-table-sort-icon-size);
      color: var(--color-text-tertiary);
      font-size: var(--ve-table-sort-icon-size);
      line-height: 1;
      flex: none;
    }

    .ve-table-sort-indicator ve-icon {
      --ve-icon-size: var(--ve-table-sort-icon-size);
      inline-size: var(--ve-table-sort-icon-size);
      block-size: var(--ve-table-sort-icon-size);
    }

    .ve-table-sort-indicator-active {
      color: var(--color-text-accent);
    }

    .ve-table-pin-left {
      position: sticky;
      inset-inline-start: var(--ve-table-pin-left-offset, 0px);
      z-index: 1;
    }

    .ve-table-pin-right {
      position: sticky;
      inset-inline-end: var(--ve-table-pin-right-offset, 0px);
      z-index: 1;
    }

    .ve-table-scroll-can-left .ve-table-pin-left {
      box-shadow: var(--ve-table-pin-left-shadow);
    }

    .ve-table-scroll-can-right .ve-table-pin-right {
      box-shadow: var(--ve-table-pin-right-shadow);
    }

    .ve-table-head-cell.ve-table-pin-left,
    .ve-table-head-cell.ve-table-pin-right {
      z-index: 3;
    }

    .ve-table-empty {
      display: flex;
      justify-content: center;
      align-items: center;
      inline-size: 100%;
      max-inline-size: 100%;
      min-inline-size: 0;
      padding: calc(var(--space-l) * 2) var(--space-m);
      color: var(--ve-table-empty-text);
      box-sizing: border-box;
    }

    .ve-table-skeleton-cell,
    .ve-table-skeleton-selection {
      display: inline-block;
      block-size: var(--ve-table-skeleton-size);
      border-radius: var(--ve-table-skeleton-radius);
      background: var(--ve-table-skeleton-background);
      background-size: 200% 100%;
      animation: ve-table-skeleton 1.4s ease-in-out infinite;
    }

    .ve-table-skeleton-cell {
      inline-size: var(--ve-table-skeleton-width, 64%);
      max-inline-size: 100%;
    }

    .ve-table-skeleton-selection {
      inline-size: var(--ve-table-checkbox-size);
      block-size: var(--ve-table-checkbox-size);
      border-radius: var(--radius-xs);
    }

    @keyframes ve-table-skeleton {
      0% {
        background-position: 100% 0;
      }

      to {
        background-position: -100% 0;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _init_columns, _init_extra__init_columns, _init_data, _init_extra__init_data, _init_rowKey, _init_extra__init_rowKey, _init_loading, _init_extra__init_loading, _init_bordered, _init_extra__init_bordered, _init_verticalLine, _init_extra__init_verticalLine, _init_hover, _init_extra__init_hover, _init_selectable, _init_extra__init_selectable, _init_scrollConfig, _init_extra__init_scrollConfig, _init_sortingState, _init_extra__init_sortingState, _init_rowSelectionState, _init_extra__init_rowSelectionState, _init_canScrollLeft, _init_extra__init_canScrollLeft, _init_canScrollRight, _init_extra__init_canScrollRight, _init_scrollContainer, _init_extra__init_scrollContainer;
const VE_TABLE_TAG = 've-table';
const DEFAULT_COLUMN_WIDTH = 160;
const MIN_COLUMN_WIDTH = 80;
const MAX_COLUMN_WIDTH = 1200;
const CONTROL_COLUMN_WIDTH = 40;
const TABLE_SKELETON_WIDTHS = [
    64,
    42,
    72,
    54,
    36
];
_dec = property({
    attribute: false
}), _dec1 = property({
    attribute: false
}), _dec2 = property({
    type: String,
    attribute: 'row-key'
}), _dec3 = property({
    type: Boolean,
    reflect: true
}), _dec4 = property({
    type: Boolean,
    reflect: true
}), _dec5 = property({
    type: Boolean,
    attribute: 'vertical-line',
    reflect: true
}), _dec6 = property({
    type: Boolean,
    reflect: true
}), _dec7 = property({
    type: Boolean,
    reflect: true
}), _dec8 = property({
    attribute: false
}), _dec9 = state(), _dec10 = state(), _dec11 = state(), _dec12 = state(), _dec13 = query('.ve-table-scroll');
class VeTableElement extends VeElement {
    static{
        ({ e: [_init_columns, _init_extra__init_columns, _init_data, _init_extra__init_data, _init_rowKey, _init_extra__init_rowKey, _init_loading, _init_extra__init_loading, _init_bordered, _init_extra__init_bordered, _init_verticalLine, _init_extra__init_verticalLine, _init_hover, _init_extra__init_hover, _init_selectable, _init_extra__init_selectable, _init_scrollConfig, _init_extra__init_scrollConfig, _init_sortingState, _init_extra__init_sortingState, _init_rowSelectionState, _init_extra__init_rowSelectionState, _init_canScrollLeft, _init_extra__init_canScrollLeft, _init_canScrollRight, _init_extra__init_canScrollRight, _init_scrollContainer, _init_extra__init_scrollContainer] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "columns"
            ],
            [
                _dec1,
                1,
                "data"
            ],
            [
                _dec2,
                1,
                "rowKey"
            ],
            [
                _dec3,
                1,
                "loading"
            ],
            [
                _dec4,
                1,
                "bordered"
            ],
            [
                _dec5,
                1,
                "verticalLine"
            ],
            [
                _dec6,
                1,
                "hover"
            ],
            [
                _dec7,
                1,
                "selectable"
            ],
            [
                _dec8,
                1,
                "scrollConfig"
            ],
            [
                _dec9,
                1,
                "sortingState"
            ],
            [
                _dec10,
                1,
                "rowSelectionState"
            ],
            [
                _dec11,
                1,
                "canScrollLeft"
            ],
            [
                _dec12,
                1,
                "canScrollRight"
            ],
            [
                _dec13,
                1,
                "scrollContainer"
            ]
        ]));
    }
    static styles = ve_table;
    static dependencies = {
        [VE_ICON_TAG]: VeIconElement,
        [VE_EMPTY_TAG]: VeEmptyElement
    };
    #___private_columns_1 = (()=>{
        const _value = _init_columns(this, []);
        _init_extra__init_columns(this);
        return _value;
    })();
    get columns() {
        return this.#___private_columns_1;
    }
    set columns(_v) {
        this.#___private_columns_1 = _v;
    }
    #___private_data_2 = (()=>{
        const _value = _init_data(this, []);
        _init_extra__init_data(this);
        return _value;
    })();
    get data() {
        return this.#___private_data_2;
    }
    set data(_v) {
        this.#___private_data_2 = _v;
    }
    #___private_rowKey_3 = (()=>{
        const _value = _init_rowKey(this, 'id');
        _init_extra__init_rowKey(this);
        return _value;
    })();
    get rowKey() {
        return this.#___private_rowKey_3;
    }
    set rowKey(_v) {
        this.#___private_rowKey_3 = _v;
    }
    #___private_loading_4 = (()=>{
        const _value = _init_loading(this, false);
        _init_extra__init_loading(this);
        return _value;
    })();
    get loading() {
        return this.#___private_loading_4;
    }
    set loading(_v) {
        this.#___private_loading_4 = _v;
    }
    #___private_bordered_5 = (()=>{
        const _value = _init_bordered(this, false);
        _init_extra__init_bordered(this);
        return _value;
    })();
    get bordered() {
        return this.#___private_bordered_5;
    }
    set bordered(_v) {
        this.#___private_bordered_5 = _v;
    }
    #___private_verticalLine_6 = (()=>{
        const _value = _init_verticalLine(this, false);
        _init_extra__init_verticalLine(this);
        return _value;
    })();
    get verticalLine() {
        return this.#___private_verticalLine_6;
    }
    set verticalLine(_v) {
        this.#___private_verticalLine_6 = _v;
    }
    #___private_hover_7 = (()=>{
        const _value = _init_hover(this, true);
        _init_extra__init_hover(this);
        return _value;
    })();
    get hover() {
        return this.#___private_hover_7;
    }
    set hover(_v) {
        this.#___private_hover_7 = _v;
    }
    #___private_selectable_8 = (()=>{
        const _value = _init_selectable(this, false);
        _init_extra__init_selectable(this);
        return _value;
    })();
    get selectable() {
        return this.#___private_selectable_8;
    }
    set selectable(_v) {
        this.#___private_selectable_8 = _v;
    }
    #___private_scrollConfig_9 = (()=>{
        const _value = _init_scrollConfig(this, void 0);
        _init_extra__init_scrollConfig(this);
        return _value;
    })();
    get scrollConfig() {
        return this.#___private_scrollConfig_9;
    }
    set scrollConfig(_v) {
        this.#___private_scrollConfig_9 = _v;
    }
    #___private_sortingState_10 = (()=>{
        const _value = _init_sortingState(this, []);
        _init_extra__init_sortingState(this);
        return _value;
    })();
    get sortingState() {
        return this.#___private_sortingState_10;
    }
    set sortingState(_v) {
        this.#___private_sortingState_10 = _v;
    }
    #___private_rowSelectionState_11 = (()=>{
        const _value = _init_rowSelectionState(this, {});
        _init_extra__init_rowSelectionState(this);
        return _value;
    })();
    get rowSelectionState() {
        return this.#___private_rowSelectionState_11;
    }
    set rowSelectionState(_v) {
        this.#___private_rowSelectionState_11 = _v;
    }
    #___private_canScrollLeft_12 = (()=>{
        const _value = _init_canScrollLeft(this, false);
        _init_extra__init_canScrollLeft(this);
        return _value;
    })();
    get canScrollLeft() {
        return this.#___private_canScrollLeft_12;
    }
    set canScrollLeft(_v) {
        this.#___private_canScrollLeft_12 = _v;
    }
    #___private_canScrollRight_13 = (()=>{
        const _value = _init_canScrollRight(this, false);
        _init_extra__init_canScrollRight(this);
        return _value;
    })();
    get canScrollRight() {
        return this.#___private_canScrollRight_13;
    }
    set canScrollRight(_v) {
        this.#___private_canScrollRight_13 = _v;
    }
    #___private_scrollContainer_14 = (()=>{
        const _value = _init_scrollContainer(this, null);
        _init_extra__init_scrollContainer(this);
        return _value;
    })();
    get scrollContainer() {
        return this.#___private_scrollContainer_14;
    }
    set scrollContainer(_v) {
        this.#___private_scrollContainer_14 = _v;
    }
    cachedColumns = null;
    cachedColumnDefs = [];
    resizeObserver = null;
    willUpdate(changedProperties) {
        super.willUpdate(changedProperties);
        if (changedProperties.has('data')) this.pruneRowStates();
    }
    firstUpdated(changedProperties) {
        super.firstUpdated(changedProperties);
        this.observeScrollContainer();
        this.updateScrollShadowState();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('columns') || changedProperties.has('data') || changedProperties.has('loading') || changedProperties.has('scrollConfig') || changedProperties.has('selectable')) this.updateScrollShadowState();
    }
    disconnectedCallback() {
        this.resizeObserver?.disconnect();
        this.resizeObserver = null;
        super.disconnectedCallback();
    }
    observeScrollContainer() {
        if ("u" < typeof ResizeObserver || !this.scrollContainer || this.resizeObserver) return;
        this.resizeObserver = new ResizeObserver(()=>{
            this.updateScrollShadowState();
        });
        this.resizeObserver.observe(this.scrollContainer);
    }
    updateScrollShadowState() {
        const container = this.scrollContainer;
        if (!container) {
            this.canScrollLeft = false;
            this.canScrollRight = false;
            return;
        }
        const maxScrollLeft = Math.max(0, container.scrollWidth - container.clientWidth);
        const scrollLeft = Math.min(Math.max(container.scrollLeft, 0), maxScrollLeft);
        const nextCanScrollLeft = scrollLeft > 1;
        const nextCanScrollRight = maxScrollLeft - scrollLeft > 1;
        if (this.canScrollLeft !== nextCanScrollLeft) this.canScrollLeft = nextCanScrollLeft;
        if (this.canScrollRight !== nextCanScrollRight) this.canScrollRight = nextCanScrollRight;
    }
    handleScroll = ()=>{
        this.updateScrollShadowState();
    };
    createTableInstance() {
        let table;
        table = createTable({
            data: this.data,
            columns: this.buildColumnDefs(),
            state: {
                sorting: this.sortingState,
                rowSelection: this.rowSelectionState,
                columnPinning: this.buildColumnPinningState(),
                columnSizing: {},
                columnSizingInfo: {
                    startOffset: null,
                    startSize: null,
                    deltaOffset: null,
                    deltaPercentage: null,
                    isResizingColumn: false,
                    columnSizingStart: []
                }
            },
            getRowId: (row, index)=>this.resolveRowId(row, index),
            enableRowSelection: this.selectable,
            onSortingChange: (updater)=>{
                this.sortingState = this.resolveUpdater(this.sortingState, updater);
                this.emitSortChange();
            },
            onRowSelectionChange: (updater)=>{
                this.rowSelectionState = this.resolveUpdater(this.rowSelectionState, updater);
                this.emitSelectionChange(table);
            },
            getCoreRowModel: getCoreRowModel(),
            getSortedRowModel: getSortedRowModel(),
            onStateChange: ()=>void 0,
            renderFallbackValue: null
        });
        return table;
    }
    createRenderContext(table = this.createTableInstance()) {
        const leafColumns = table.getVisibleLeafColumns();
        const columnWidths = new Map();
        const columnStyles = new Map();
        const leftOffsets = new Map();
        const rightOffsets = new Map();
        let leftOffset = this.selectable ? CONTROL_COLUMN_WIDTH : 0;
        let rightOffset = 0;
        for (const column of leafColumns)columnWidths.set(column.id, this.getColumnPixelWidth(column));
        for (const column of leafColumns)if (this.getColumnMeta(column)?.fixed === 'left') {
            leftOffsets.set(column.id, leftOffset);
            leftOffset += columnWidths.get(column.id) ?? DEFAULT_COLUMN_WIDTH;
        }
        for(let index = leafColumns.length - 1; index >= 0; index -= 1){
            const column = leafColumns[index];
            if (column && this.getColumnMeta(column)?.fixed === 'right') {
                rightOffsets.set(column.id, rightOffset);
                rightOffset += columnWidths.get(column.id) ?? DEFAULT_COLUMN_WIDTH;
            }
        }
        const context = {
            table,
            leafColumns,
            columnWidths,
            columnStyles,
            leftOffsets,
            rightOffsets,
            renderedColumnCount: leafColumns.length + (this.selectable ? 1 : 0)
        };
        for (const column of leafColumns)columnStyles.set(column.id, this.getColumnCellStyle(column, context));
        return context;
    }
    resolveUpdater(previous, updater) {
        return 'function' == typeof updater ? updater(previous) : updater;
    }
    resolveRowId(row, index) {
        const candidate = row[this.rowKey];
        if ('string' == typeof candidate || 'number' == typeof candidate) return String(candidate);
        return String(index);
    }
    getSelectedRowKeys() {
        return Object.entries(this.rowSelectionState).filter(([, selected])=>selected).map(([key])=>key);
    }
    getCurrentSortDetail() {
        const first = this.sortingState[0];
        if (!first) return null;
        return {
            key: first.id,
            direction: first.desc ? 'desc' : 'asc'
        };
    }
    emitTableEvent(type, detail) {
        this.dispatchEvent(new CustomEvent(type, {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    emitAggregateChange() {
        this.emitTableEvent('ve-change', {
            sortState: this.getCurrentSortDetail(),
            selectedRowKeys: this.getSelectedRowKeys()
        });
    }
    emitSortChange() {
        const sortDetail = this.getCurrentSortDetail();
        this.emitTableEvent('ve-sort-change', sortDetail);
        this.emitAggregateChange();
    }
    emitSelectionChange(table) {
        this.emitTableEvent('ve-selection-change', {
            selectedRowKeys: this.getSelectedRowKeys(),
            selectedRows: table.getSelectedRowModel().rows.map((row)=>row.original)
        });
        this.emitAggregateChange();
    }
    pruneRowStates() {
        const validKeys = new Set(this.data.map((row, index)=>this.resolveRowId(row, index)));
        this.rowSelectionState = Object.fromEntries(Object.entries(this.rowSelectionState).filter(([key])=>validKeys.has(key)));
    }
    buildColumnPinningState() {
        const left = [];
        const right = [];
        for (const column of this.columns){
            if ('left' === column.fixed) left.push(column.key);
            if ('right' === column.fixed) right.push(column.key);
        }
        return {
            left,
            right
        };
    }
    buildColumnDefs() {
        if (this.cachedColumns === this.columns) return this.cachedColumnDefs;
        this.cachedColumns = this.columns;
        this.cachedColumnDefs = this.columns.map((column)=>{
            const meta = {
                tableColumn: column
            };
            if (column.align) meta.align = column.align;
            if (column.fixed) meta.fixed = column.fixed;
            if (column.wrap) meta.wrap = column.wrap;
            const shared = {
                id: column.key,
                header: ()=>this.renderHeaderContent(column),
                cell: (context)=>this.renderCellContent(column, context.getValue(), context.row.original, context.row.index),
                enableSorting: Boolean(column.sortable),
                meta
            };
            if ('number' == typeof column.width) Object.assign(shared, {
                size: column.width
            });
            if ('number' == typeof column.minWidth) Object.assign(shared, {
                minSize: column.minWidth
            });
            if ('number' == typeof column.maxWidth) Object.assign(shared, {
                maxSize: column.maxWidth
            });
            return {
                ...shared,
                accessorFn: (row)=>this.getCellValue(row, column)
            };
        });
        return this.cachedColumnDefs;
    }
    getCellValue(row, column) {
        const dataIndex = column.dataIndex ?? column.key;
        return row[dataIndex];
    }
    renderHeaderContent(column) {
        const content = 'function' == typeof column.title ? column.title(column) : column.title;
        return html`<span class="ve-table-header-content">${content}</span>`;
    }
    renderCellContent(column, value, row, rowIndex) {
        if (column.render) return this.wrapCellContent(column, column.render(value, row, rowIndex));
        return this.wrapCellContent(column, this.stringifyCellValue(value));
    }
    wrapCellContent(column, content) {
        return html`
      <span
        class=${classMap({
            've-table-cell-content': true,
            've-table-cell-content-wrap': Boolean(column.wrap)
        })}
      >
        ${content ?? nothing}
      </span>
    `;
    }
    stringifyCellValue(value) {
        if (null == value || '' === value) return '--';
        return String(value);
    }
    getColumnMeta(column) {
        return column.columnDef.meta;
    }
    getColumnPixelWidth(column) {
        const tableColumn = this.getColumnMeta(column)?.tableColumn;
        const minWidth = 'number' == typeof tableColumn?.minWidth && Number.isFinite(tableColumn.minWidth) ? Math.max(MIN_COLUMN_WIDTH, tableColumn.minWidth) : MIN_COLUMN_WIDTH;
        const maxWidth = 'number' == typeof tableColumn?.maxWidth && Number.isFinite(tableColumn.maxWidth) ? Math.max(minWidth, tableColumn.maxWidth) : MAX_COLUMN_WIDTH;
        const baseWidth = 'number' == typeof tableColumn?.width && Number.isFinite(tableColumn.width) ? tableColumn.width : DEFAULT_COLUMN_WIDTH;
        return Math.min(Math.max(baseWidth, minWidth), maxWidth);
    }
    shouldLockColumnWidth(column) {
        const tableColumn = this.getColumnMeta(column)?.tableColumn;
        return Boolean(tableColumn?.fixed || 'number' == typeof tableColumn?.width || 'number' == typeof tableColumn?.minWidth || 'number' == typeof tableColumn?.maxWidth);
    }
    getColumnCellStyle(column, context) {
        const fixed = this.getColumnMeta(column)?.fixed;
        const shouldLockWidth = this.shouldLockColumnWidth(column);
        const columnWidth = context.columnWidths.get(column.id) ?? DEFAULT_COLUMN_WIDTH;
        const widthStyle = shouldLockWidth ? `width:${columnWidth}px; min-width:${columnWidth}px;` : '';
        if ('left' === fixed) return `--ve-table-pin-left-offset:${context.leftOffsets.get(column.id) ?? 0}px; ${widthStyle}`;
        if ('right' === fixed) return `--ve-table-pin-right-offset:${context.rightOffsets.get(column.id) ?? 0}px; ${widthStyle}`;
        return shouldLockWidth ? widthStyle : void 0;
    }
    renderSelectionHeader(context) {
        const table = context.table;
        const checked = table.getIsAllRowsSelected();
        const indeterminate = table.getIsSomeRowsSelected() && !checked;
        return html`
      <th class="ve-table-head-cell ve-table-select-cell" scope="col">
        <button
          class=${classMap({
            've-table-selection': true,
            've-table-selection-checked': checked,
            've-table-selection-indeterminate': indeterminate
        })}
          type="button"
          role="checkbox"
          aria-label="Select all rows"
          aria-checked=${indeterminate ? 'mixed' : checked ? 'true' : 'false'}
          @click=${()=>table.toggleAllRowsSelected(!checked)}
        >
          ${this.renderSelectionIcons()}
        </button>
      </th>
    `;
    }
    renderSelectionIcons() {
        return html`
      <svg
        class="ve-table-selection-icon ve-table-selection-icon-check"
        aria-hidden="true"
        focusable="false"
        viewBox="0 0 16 16"
        fill="none"
      >
        <path
          d="M3.25 8.1 6.35 11.2 12.75 4.8"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="var(--ve-table-checkbox-icon-stroke-width)"
        ></path>
      </svg>
      <svg
        class="ve-table-selection-icon ve-table-selection-icon-minus"
        aria-hidden="true"
        focusable="false"
        viewBox="0 0 16 16"
        fill="none"
      >
        <path
          d="M4 8h8"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-width="var(--ve-table-checkbox-icon-stroke-width)"
        ></path>
      </svg>
    `;
    }
    renderHead(context) {
        const headerGroups = context.table.getHeaderGroups();
        return html`
      <thead part="head">
        ${headerGroups.map((group)=>html`
            <tr class="ve-table-head-row">
              ${this.selectable ? this.renderSelectionHeader(context) : nothing}
              ${group.headers.map((header)=>{
                const meta = this.getColumnMeta(header.column);
                const align = meta?.align ?? 'left';
                const fixed = meta?.fixed;
                const style = context.columnStyles.get(header.column.id);
                const canSort = header.column.getCanSort();
                const sortState = header.column.getIsSorted();
                const ariaSort = 'asc' === sortState ? 'ascending' : 'desc' === sortState ? 'descending' : 'none';
                const sortIcon = 'asc' === sortState ? arrowUpSmIcon : 'desc' === sortState ? arrowDownSmIcon : downUpIcon;
                return html`
                  <th
                    class=${classMap({
                    've-table-head-cell': true,
                    've-table-head-cell-align-center': 'center' === align,
                    've-table-head-cell-align-right': 'right' === align,
                    've-table-pin-left': 'left' === fixed,
                    've-table-pin-right': 'right' === fixed
                })}
                    scope="col"
                    aria-sort=${ariaSort}
                    style=${ifDefined(style)}
                  >
                    ${header.isPlaceholder ? nothing : canSort ? html`
                            <button
                              class="ve-table-sort-button"
                              type="button"
                              @click=${header.column.getToggleSortingHandler()}
                            >
                              ${flexRender(header.column.columnDef.header, header.getContext())}
                              <span
                                class=${classMap({
                    've-table-sort-indicator': true,
                    've-table-sort-indicator-active': Boolean(sortState)
                })}
                                aria-hidden="true"
                              >
                                <ve-icon .definition=${sortIcon}></ve-icon>
                              </span>
                            </button>
                          ` : flexRender(header.column.columnDef.header, header.getContext())}
                  </th>
                `;
            })}
            </tr>
          `)}
      </thead>
    `;
    }
    renderRowSelectionCell(row) {
        const checked = row.getIsSelected();
        const indeterminate = row.getIsSomeSelected() && !checked;
        return html`
      <td class="ve-table-select-cell">
        <button
          class=${classMap({
            've-table-selection': true,
            've-table-selection-checked': checked,
            've-table-selection-indeterminate': indeterminate
        })}
          type="button"
          role="checkbox"
          aria-label=${ifDefined(`Select row ${row.id}`)}
          aria-checked=${indeterminate ? 'mixed' : checked ? 'true' : 'false'}
          ?disabled=${!row.getCanSelect()}
          @click=${()=>row.toggleSelected(!checked)}
        >
          ${this.renderSelectionIcons()}
        </button>
      </td>
    `;
    }
    renderCell(cell, context) {
        const meta = this.getColumnMeta(cell.column);
        const align = meta?.align ?? 'left';
        const fixed = meta?.fixed;
        const style = context.columnStyles.get(cell.column.id);
        return html`
      <td
        class=${classMap({
            've-table-cell': true,
            've-table-cell-align-center': 'center' === align,
            've-table-cell-align-right': 'right' === align,
            've-table-pin-left': 'left' === fixed,
            've-table-pin-right': 'right' === fixed
        })}
        style=${ifDefined(style)}
      >
        ${flexRender(cell.column.columnDef.cell, cell.getContext())}
      </td>
    `;
    }
    renderSkeletonCell(column, context, columnIndex) {
        const meta = this.getColumnMeta(column);
        const align = meta?.align ?? 'left';
        const fixed = meta?.fixed;
        const skeletonWidth = TABLE_SKELETON_WIDTHS[columnIndex % TABLE_SKELETON_WIDTHS.length] ?? 64;
        const style = [
            context.columnStyles.get(column.id),
            `--ve-table-skeleton-width:${skeletonWidth}%;`
        ].filter(Boolean).join(' ');
        return html`
      <td
        class=${classMap({
            've-table-cell': true,
            've-table-cell-align-center': 'center' === align,
            've-table-cell-align-right': 'right' === align,
            've-table-pin-left': 'left' === fixed,
            've-table-pin-right': 'right' === fixed
        })}
        style=${style}
      >
        <span class="ve-table-skeleton-cell" aria-hidden="true"></span>
      </td>
    `;
    }
    renderSkeletonRows(context) {
        if (!this.loading) return nothing;
        return Array.from({
            length: 2
        }, (_, rowIndex)=>html`
        <tr
          class="ve-table-row ve-table-skeleton-row"
          part="row"
          aria-hidden="true"
        >
          ${this.selectable ? html`
                <td class="ve-table-select-cell">
                  <span class="ve-table-skeleton-selection"></span>
                </td>
              ` : nothing}
          ${context.leafColumns.map((column, columnIndex)=>this.renderSkeletonCell(column, context, columnIndex + rowIndex))}
        </tr>
      `);
    }
    renderBody(context) {
        const rows = context.table.getRowModel().rows;
        const emptyContent = html`<slot name="empty-content"
      ><ve-empty></ve-empty
    ></slot>`;
        if (0 === rows.length && !this.loading) return html`
        <tbody part="body">
          <tr>
            <td class="ve-table-cell" colspan=${context.renderedColumnCount}>
              <div class="ve-table-empty" part="empty">${emptyContent}</div>
            </td>
          </tr>
        </tbody>
      `;
        return html`
      <tbody part="body">
        ${rows.map((row)=>html`
            <tr
              class=${classMap({
                've-table-row': true
            })}
              part="row"
            >
              ${this.selectable ? this.renderRowSelectionCell(row) : nothing}
              ${row.getVisibleCells().map((cell)=>this.renderCell(cell, context))}
            </tr>
          `)}
        ${this.renderSkeletonRows(context)}
      </tbody>
    `;
    }
    normalizeScrollSize(value) {
        if ('number' == typeof value) return Number.isFinite(value) && value > 0 ? `${value}px` : void 0;
        const trimmed = value?.trim();
        return trimmed || void 0;
    }
    getScrollContainerStyle() {
        const scrollY = this.normalizeScrollSize(this.scrollConfig?.y);
        return scrollY ? `--ve-table-scroll-y:${scrollY}; max-block-size:var(--ve-table-scroll-y);` : void 0;
    }
    getNativeTableStyle() {
        const scrollX = this.normalizeScrollSize(this.scrollConfig?.x);
        return scrollX ? `--ve-table-scroll-x:${scrollX}; inline-size:max(100%, var(--ve-table-scroll-x)); min-inline-size:max(100%, var(--ve-table-scroll-x));` : void 0;
    }
    render() {
        const context = this.createRenderContext();
        return html`
      <div class="ve-table">
        <div
          class=${classMap({
            've-table-scroll': true,
            've-table-scroll-can-left': this.canScrollLeft,
            've-table-scroll-can-right': this.canScrollRight
        })}
          style=${ifDefined(this.getScrollContainerStyle())}
          @scroll=${this.handleScroll}
        >
          <table
            class="ve-table-native"
            part="table"
            style=${ifDefined(this.getNativeTableStyle())}
          >
            ${this.renderHead(context)} ${this.renderBody(context)}
          </table>
        </div>
      </div>
    `;
    }
}
defineCustomElement(VE_TABLE_TAG, VeTableElement);
export { VE_TABLE_TAG, VeTableElement };
