// @ve-design/web/dist/ve-upload.js
import { css, html, nothing } from "lit";
import { classMap } from "lit/directives/class-map.js";
import { ifDefined } from "lit/directives/if-defined.js";
import { property, query, state } from "lit/decorators.js";
import { defineCustomElement, componentReset, VeElement } from "./78928.js";
import { uploadIcon } from "./icons/upload.js";
import { addIcon } from "./icons/add.js";
import { fileIcon as file_fileIcon } from "./icons/file.js";
import { typeImageStateDefaultIcon } from "./icons/type-image-state-default.js";
import { typePdfStateDefaultIcon } from "./icons/type-pdf-state-default.js";
import { typeVideoStateDefaultIcon } from "./icons/type-video-state-default.js";
import { typeAudioStateDefaultIcon } from "./icons/type-audio-state-default.js";
import { warningIcon } from "./icons/warning.js";
import { imageIcon } from "./icons/image.js";
import { closeIcon } from "./icons/close.js";
import { successIcon as success_successIcon } from "./icons/success.js";
import { trash03Icon } from "./icons/trash-03.js";
import { previewIcon } from "./icons/preview.js";
import { playSquareIcon } from "./icons/play-square.js";
import { pauseIcon } from "./icons/pause.js";
import { VeButtonElement, VE_BUTTON_TAG } from "./68533.js";
import { VeIconElement, VE_ICON_TAG } from "./29989.js";
import { typeWordStateDefaultIcon } from "./icons/type-word-state-default.js";
const veUploadTokenDefaults = css`
  :host {
    --ve-upload-tip-color-text: var(--color-text-tertiary);
    --ve-upload-tip-margin-top: 4px;
    --ve-upload-tip-font-size: 12px;
    --ve-upload-list-margin-top: var(--space-xxs);
    --ve-upload-list-gap: var(--space-xxs);

    --ve-upload-attachment-color-bg: var(--color-bg-secondary);
    --ve-upload-attachment-color-bg-hover: color-mix(
      var(--ve-upload-attachment-color-bg) var(--state-hover-subtle-ratio),
      var(--state-hover-subtle-mix)
    );
    --ve-upload-attachment-color-bg-error: color-mix(
      var(--color-bg-danger) 12%,
      var(--color-bg-base)
    );
    --ve-upload-attachment-color-text: var(--color-text-primary);
    --ve-upload-attachment-color-text-secondary: var(--color-text-tertiary);
    --ve-upload-attachment-color-text-danger: var(--color-text-danger);
    --ve-upload-attachment-color-icon: var(--color-icon-accent);
    --ve-upload-attachment-color-icon-foreground: var(--color-icon-foreground);
    --ve-upload-attachment-color-icon-bg: transparent;
    --ve-upload-attachment-color-gradient-start: rgb(246 246 252 / 0%);
    --ve-upload-attachment-color-gradient-end: rgb(16 16 19 / 34%);
    --ve-upload-attachment-color-skeleton: var(--color-bg-muted);
    --ve-upload-attachment-progress-bg: var(--color-bg-muted);
    --ve-upload-attachment-progress-opacity: 0.6;
    --ve-upload-attachment-loading-shimmer-color: color-mix(
      var(--color-bg-base) 72%,
      transparent
    );
    --ve-upload-attachment-loading-shimmer-duration: 1200ms;
    --ve-upload-attachment-radius: var(--radius-lg);
    --ve-upload-attachment-radius-inner: var(--radius-md);
    --ve-upload-attachment-radius-badge: var(--radius-xs);
    --ve-upload-attachment-card-size: 120px;
    --ve-upload-attachment-card-padding: var(--space-xs);
    --ve-upload-attachment-card-title-line-height: 22px;
    --ve-upload-attachment-card-meta-inline: var(--space-xxxs);
    --ve-upload-attachment-row-width: 200px;
    --ve-upload-attachment-row-height: 56px;
    --ve-upload-attachment-row-padding-block: var(--space-xxs);
    --ve-upload-attachment-row-padding-inline: var(--space-xs);
    --ve-upload-attachment-row-icon-size: 32px;
    --ve-upload-attachment-tag-height: 38px;
    --ve-upload-attachment-tag-max-width: 200px;
    --ve-upload-attachment-tag-padding: var(--space-xxs);
    --ve-upload-attachment-tag-icon-size: 20px;
    --ve-upload-attachment-title-font-size: var(--text-body);
    --ve-upload-attachment-meta-font-size: var(--text-caption);
    --ve-upload-attachment-remove-size: 20px;
    --ve-upload-attachment-remove-icon-size: 12px;
    --ve-upload-attachment-remove-bg: var(--color-bg-inverse);
    --ve-upload-attachment-remove-color: var(--color-text-foreground);
    --ve-upload-attachment-remove-opacity: 0.82;
    --ve-upload-attachment-video-play-size: 22px;
    --ve-upload-attachment-video-play-bg: rgb(16 16 19 / 56%);

    --ve-upload-picture-item-width: 80px;
    --ve-upload-picture-color-bg: var(--color-bg-secondary);
    --ve-upload-picture-border-radius: 2px;
    --ve-upload-picture-border-width: var(--stroke-1);
    --ve-upload-picture-border-style: dashed;
    --ve-upload-picture-color-border: var(--color-border-default);
    --ve-upload-picture-color-border-disabled: var(--color-border-strong);
    --ve-upload-picture-color-border-hover: var(--color-border-strong);
    --ve-upload-picture-color-bg-hover: var(--color-bg-muted);
    --ve-upload-picture-color-bg-disabled: var(--color-bg-secondary);
    --ve-upload-picture-color-text: var(--color-text-secondary);
    --ve-upload-picture-color-text-hover: var(--color-text-secondary);
    --ve-upload-picture-color-text-disabled: var(--color-text-disable);

    --ve-upload-drag-font-size: 14px;
    --ve-upload-drag-border-radius: 2px;
    --ve-upload-drag-color-text: var(--color-text-primary);
    --ve-upload-drag-border-style: dashed;
    --ve-upload-drag-border-width: var(--stroke-1);
    --ve-upload-drag-padding-vertical: var(--space-xxl);
    --ve-upload-drag-padding-inline: var(--space-xl);
    --ve-upload-drag-min-height: 180px;
    --ve-upload-drag-content-gap: var(--space-s);
    --ve-upload-drag-color-bg: var(--color-bg-secondary);
    --ve-upload-drag-color-bg-hover: var(--color-bg-muted);
    --ve-upload-drag-color-bg-active: color-mix(
      var(--color-bg-primary) 10%,
      var(--color-bg-base)
    );
    --ve-upload-drag-color-bg-disabled: var(--color-bg-secondary);
    --ve-upload-drag-color-border: var(--color-border-default);
    --ve-upload-drag-color-border-active: var(--color-text-accent);
    --ve-upload-drag-color-border-hover: var(--color-border-strong);
    --ve-upload-drag-color-border-disabled: var(--color-text-disable);
    --ve-upload-drag-color-icon: var(--color-text-secondary);
    --ve-upload-drag-color-icon-hover: var(--color-text-secondary);
    --ve-upload-drag-color-icon-active: var(--color-text-accent);
    --ve-upload-drag-color-text-hover: var(--color-text-primary);
    --ve-upload-drag-color-text-active: var(--color-text-primary);
    --ve-upload-drag-color-text-disabled: var(--color-text-disable);

    --ve-upload-text-item-size-operation-icon: 12px;
    --ve-upload-text-item-margin-top: 12px;
    --ve-upload-text-item-font-size: 14px;
    --ve-upload-text-item-color-text: var(--color-text-primary);
    --ve-upload-text-item-padding-left: 12px;
    --ve-upload-text-item-color-bg: var(--color-bg-secondary);
    --ve-upload-text-item-padding-vertical: 8px;
    --ve-upload-text-item-margin-remove-icon-left: 12px;
    --ve-upload-text-item-color-remove-icon: var(--color-text-secondary);
    --ve-upload-text-item-color-status-icon: var(--color-text-foreground);
    --ve-upload-text-item-color-file-icon-success: var(--color-text-accent);
    --ve-upload-text-item-color-progress-bg-hover: color-mix(
      var(--color-text-primary) 20%,
      transparent
    );
    --ve-upload-text-item-color-progress-bg-hover-active: var(--color-blue-700);
    --ve-upload-text-item-size-file-icon: 16px;
    --ve-upload-text-item-margin-file-icon-right: 12px;
    --ve-upload-text-item-color-file-icon: var(--color-text-accent);
    --ve-upload-text-item-padding-right: 10px;
    --ve-upload-text-item-color-link: var(--color-text-accent);
    --ve-upload-text-item-color-reupload-icon: var(--color-text-accent);
    --ve-upload-text-item-color-reupload-icon-hover: var(--color-blue-700);
    --ve-upload-text-item-size-status-icon: 12px;
    --ve-upload-text-item-color-error-icon: var(--color-text-danger);
    --ve-upload-text-item-color-success-icon: var(--color-text-success);
    --ve-upload-text-item-border-radius: 2px;
    --ve-upload-text-item-margin-error-icon-left: 4px;
    --ve-upload-text-item-margin-status-left: 10px;
    --ve-upload-text-item-thumbnail-width: 40px;
    --ve-upload-text-item-margin-thumbnail-right: 12px;
    --ve-upload-text-item-color-text-error: var(--color-text-danger);
    --ve-upload-text-item-color-text-success: var(
      --ve-upload-text-item-color-text
    );
    --ve-upload-text-item-color-text-uploading: var(
      --ve-upload-text-item-color-text
    );

    --ve-upload-picture-item-margin-preview-icon-right: 20px;
    --ve-upload-picture-item-size-width: 80px;
    --ve-upload-picture-item-border-radius: 2px;
    --ve-upload-picture-item-margin-right: 8px;
    --ve-upload-picture-item-margin-bottom: 8px;
    --ve-upload-picture-item-color-bg: var(--color-bg-secondary);
    --ve-upload-picture-item-color-operation-bg: rgb(0 0 0 / 50%);
    --ve-upload-picture-item-color-operation-icon: var(--color-text-foreground);
    --ve-upload-picture-item-color-error-icon: var(--color-text-foreground);
    --ve-upload-picture-item-color-progress-bg-hover: rgb(255 255 255 / 34%);
    --ve-upload-picture-item-color-progress-bg-hover-active: var(
      --color-text-accent
    );
    --ve-upload-picture-text-item-color-bg-error: color-mix(
      var(--color-bg-danger) 12%,
      var(--color-bg-base)
    );
    --ve-upload-picture-text-item-color-text-error: var(--color-text-danger);
    --ve-upload-picture-text-item-color-text-success: var(
      --ve-upload-text-item-color-text
    );
    --ve-upload-picture-text-item-color-text-uploading: var(
      --ve-upload-text-item-color-text
    );
    --ve-upload-picture-text-item-padding-vertical: 8px;

    --ve-upload-drag-size-icon: 28px;
    --ve-upload-picture-item-size-mask-icon: 16px;
    --ve-upload-picture-item-size-error-icon: 26px;
    --ve-upload-text-item-size-reupload-icon: 14px;
    --ve-upload-text-item-size-success-icon: 14px;
    --ve-upload-text-item-size-error-icon: 14px;
    --ve-upload-picture-item-size-operation-icon: 14px;

    --ve-upload-color-icon-box-shadow-default: color-mix(
      var(--color-text-accent) 28%,
      transparent
    );
    --ve-upload-color-icon-box-shadow-active: var(--color-text-accent);

    --ve-upload-transition-duration-1: 100ms;
    --ve-upload-transition-duration-2: 200ms;
    --ve-upload-transition-duration-3: 300ms;
    --ve-upload-transition-timing-linear: cubic-bezier(0, 0, 1, 1);
    --ve-upload-transition-timing-standard: cubic-bezier(0.34, 0.69, 0.1, 1);

    --ve-upload-preview-mask-bg: rgb(0 0 0 / 65%);
    --ve-upload-preview-action-bg: rgb(0 0 0 / 40%);
    --ve-upload-preview-action-color: var(--color-text-foreground);
  }
`;
const token = veUploadTokenDefaults;
const ve_upload = [
    componentReset,
    token,
    css`
    :host {
      display: inline-block;
      max-inline-size: 100%;
      vertical-align: top;
    }

    :host([hidden]) {
      display: none;
    }

    :host([drag]) {
      display: block;
      inline-size: 100%;
    }

    .ve-upload-root {
      display: inline-block;
      inline-size: 100%;
      max-inline-size: 100%;
    }

    .ve-upload {
      display: inline-block;
      inline-size: 100%;
      max-inline-size: 100%;
      vertical-align: top;
    }

    .ve-upload-type-picture-card {
      inline-size: auto;
      vertical-align: top;
    }

    .ve-upload-drag {
      display: block;
      inline-size: 100%;
    }

    .ve-upload-hide {
      display: none;
    }

    .ve-upload-disabled .ve-upload-trigger-picture,
    .ve-upload-disabled .ve-upload-trigger-picture:hover {
      cursor: not-allowed;
      border-color: var(--ve-upload-picture-color-border-disabled);
      background-color: var(--ve-upload-picture-color-bg-disabled);
      color: var(--ve-upload-picture-color-text-disabled);
    }

    .ve-upload-disabled .ve-upload-trigger-drag,
    .ve-upload-disabled .ve-upload-trigger-drag:hover {
      cursor: not-allowed;
      border-color: var(--ve-upload-drag-color-border-disabled);
      background-color: var(--ve-upload-drag-color-bg-disabled);
      color: var(--ve-upload-drag-color-text-disabled);
    }

    .ve-upload-disabled .ve-upload-trigger-drag-icon,
    .ve-upload-disabled .ve-upload-trigger-drag-text,
    .ve-upload-disabled .ve-upload-trigger-tip {
      color: var(--ve-upload-drag-color-text-disabled);
    }

    .ve-upload-trigger {
      cursor: pointer;
      display: inline-block;
      vertical-align: top;
      inline-size: 100%;
      box-sizing: border-box;
    }

    .ve-upload-trigger:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
      border-radius: var(--radius-xs);
    }

    .ve-upload-trigger-tip {
      color: var(--ve-upload-tip-color-text);
      margin-top: var(--ve-upload-tip-margin-top);
      font-size: var(--ve-upload-tip-font-size);
      line-height: 1.5;
      white-space: nowrap;
      text-overflow: ellipsis;
      overflow: hidden;
    }

    .ve-upload-trigger-with-icon {
      vertical-align: top;
    }

    .ve-upload-trigger-picture-wrapper {
      display: inline-block;
      vertical-align: top;
    }

    .ve-upload-trigger-picture {
      min-inline-size: var(--ve-upload-picture-item-width);
      block-size: var(--ve-upload-picture-item-width);
      color: var(--ve-upload-picture-color-text);
      text-align: center;
      margin-bottom: 0;
      background: var(--ve-upload-picture-color-bg);
      border-radius: var(--ve-upload-picture-border-radius);
      border: var(--ve-upload-picture-border-width)
        var(--ve-upload-picture-border-style)
        var(--ve-upload-picture-color-border);
      transition: all var(--ve-upload-transition-duration-1)
        var(--ve-upload-transition-timing-linear);
      display: inline-flex;
      align-items: center;
      justify-content: center;
      box-sizing: border-box;
    }

    .ve-upload-trigger-picture:hover {
      border-color: var(--ve-upload-picture-color-border-hover);
      background-color: var(--ve-upload-picture-color-bg-hover);
      color: var(--ve-upload-picture-color-text-hover);
    }

    .ve-upload-trigger-picture:focus-visible {
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
      outline: none;
    }

    .ve-upload-trigger-picture-text {
      line-height: 1;
      font-size: 20px;
    }

    .ve-upload-trigger-drag {
      display: flex;
      flex-direction: column;
      align-items: center;
      justify-content: center;
      gap: var(--ve-upload-drag-content-gap);
      inline-size: 100%;
      min-block-size: var(--ve-upload-drag-min-height);
      border-radius: var(--ve-upload-drag-border-radius);
      border: var(--ve-upload-drag-border-width)
        var(--ve-upload-drag-border-style) var(--ve-upload-drag-color-border);
      background-color: var(--ve-upload-drag-color-bg);
      text-align: center;
      transition: all var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-standard);
      color: var(--ve-upload-drag-color-text);
      padding: var(--ve-upload-drag-padding-vertical)
        var(--ve-upload-drag-padding-inline);
      box-sizing: border-box;
    }

    .ve-upload-trigger-drag-icon {
      font-size: var(--ve-upload-drag-size-icon);
      color: var(--ve-upload-drag-color-icon);
      line-height: 1;
      display: inline-flex;
      align-items: center;
      justify-content: center;
    }

    .ve-upload-trigger-drag:hover {
      border-color: var(--ve-upload-drag-color-border-hover);
      background-color: var(--ve-upload-drag-color-bg-hover);
    }

    .ve-upload-trigger-drag:hover .ve-upload-trigger-drag-text {
      color: var(--ve-upload-drag-color-text-hover);
    }

    .ve-upload-trigger-drag:hover .ve-upload-trigger-drag-icon {
      color: var(--ve-upload-drag-color-icon-hover);
    }

    .ve-upload-trigger-drag:focus-visible {
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
      outline: none;
    }

    .ve-upload-trigger-drag-active {
      border-color: var(--ve-upload-drag-color-border-active);
      color: var(--ve-upload-drag-color-text-active);
      background-color: var(--ve-upload-drag-color-bg-active);
    }

    .ve-upload-trigger-drag-active .ve-upload-trigger-drag-text {
      color: var(--ve-upload-drag-color-text-active);
    }

    .ve-upload-trigger-drag-active .ve-upload-trigger-drag-icon {
      color: var(--ve-upload-drag-color-icon-active);
    }

    .ve-upload-trigger-drag .ve-upload-trigger-tip {
      margin-top: 0;
    }

    .ve-upload-trigger-drag-text {
      color: var(--ve-upload-drag-color-text);
      line-height: 1.5;
      font-size: var(--ve-upload-drag-font-size);
      margin: 0;
    }

    .ve-upload-trigger-custom-active {
      box-shadow: inset 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
      border-radius: var(--radius-sm);
      transition: box-shadow var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-standard);
    }

    .ve-upload-trigger-input {
      display: none;
    }

    .ve-upload-icon {
      display: inline-flex;
      flex: none;
      color: inherit;
      line-height: 0;
    }

    .ve-upload-hide + .ve-upload-list .ve-upload-list-item:first-of-type {
      margin-top: 0;
    }

    .ve-upload-list {
      inline-size: 100%;
      box-sizing: border-box;
    }

    .ve-upload-list-type-text .ve-upload-list-item:first-of-type,
    .ve-upload-list-type-picture-list .ve-upload-list-item:first-of-type {
      margin-top: var(--ve-upload-list-margin-top);
    }

    .ve-upload-list-file-icon {
      line-height: var(--ve-upload-text-item-size-file-icon);
      font-size: var(--ve-upload-text-item-size-file-icon);
      color: var(--ve-upload-text-item-color-file-icon);
      margin-right: var(--ve-upload-text-item-margin-file-icon-right);
      flex: none;
    }

    .ve-upload-list-preview-icon {
      cursor: pointer;
    }

    .ve-upload-list-error-icon {
      cursor: pointer;
      color: var(--ve-upload-text-item-color-error-icon);
      font-size: var(--ve-upload-text-item-size-error-icon);
      margin-left: var(--ve-upload-text-item-margin-error-icon-left);
      line-height: 1;
      display: inline-flex;
      align-items: center;
    }

    .ve-upload-list-success-icon {
      cursor: pointer;
      color: var(--ve-upload-text-item-color-success-icon);
      font-size: var(--ve-upload-text-item-size-success-icon);
      line-height: var(--ve-upload-text-item-size-success-icon);
      display: inline-flex;
      align-items: center;
      justify-content: center;
    }

    .ve-upload-list-remove-icon {
      position: relative;
      cursor: pointer;
      font-size: var(--ve-upload-picture-item-size-operation-icon);
      line-height: 1;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      border-radius: var(--radius-xs);
      opacity: 0;
      pointer-events: none;
      transition: opacity var(--ve-upload-transition-duration-1)
        var(--ve-upload-transition-timing-linear);
    }

    .ve-upload-list-item:hover .ve-upload-list-remove-icon,
    .ve-upload-list-item:focus-within .ve-upload-list-remove-icon,
    .ve-upload-list-remove-icon:focus-visible {
      opacity: 1;
      pointer-events: auto;
    }

    .ve-upload-list-remove-icon:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
    }

    .ve-upload-list-start-icon,
    .ve-upload-list-cancel-icon {
      position: absolute;
      color: var(--ve-upload-text-item-color-status-icon);
      transform: translateX(-50%) translateY(-50%);
      top: 50%;
      left: 50%;
      font-size: var(--ve-upload-text-item-size-status-icon);
      line-height: 1;
      border-radius: var(--radius-xs);
      padding: 0;
      cursor: pointer;
      z-index: 1;
    }

    .ve-upload-list-start-icon:focus-visible,
    .ve-upload-list-cancel-icon:focus-visible {
      color: var(--ve-upload-color-icon-box-shadow-active);
      outline: none;
    }

    .ve-upload-list-reupload-icon {
      cursor: pointer;
      color: var(--ve-upload-text-item-color-reupload-icon);
      font-size: var(--ve-upload-text-item-size-reupload-icon);
      line-height: 1;
      transition: all var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-standard);
      border-radius: var(--radius-xs);
      display: inline-flex;
      align-items: center;
      justify-content: center;
    }

    .ve-upload-list-reupload-icon:hover,
    .ve-upload-list-reupload-icon:active {
      color: var(--ve-upload-text-item-color-reupload-icon-hover);
    }

    .ve-upload-list-reupload-icon:focus-visible {
      box-shadow: inset 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
      outline: none;
    }

    .ve-upload-list-status {
      position: relative;
      cursor: pointer;
      inline-size: 14px;
      block-size: 14px;
      line-height: var(--ve-upload-text-item-size-status-icon);
      display: inline-flex;
      align-items: center;
      justify-content: center;
      margin-left: var(--ve-upload-text-item-margin-status-left);
      flex: none;
    }

    .ve-upload-list-status-ring {
      inline-size: 14px;
      block-size: 14px;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      position: absolute;
      inset: 0;
    }

    .ve-upload-list-status-ring-svg {
      inline-size: 100%;
      block-size: 100%;
      transform: rotate(-90deg);
      overflow: visible;
    }

    .ve-upload-list-status-ring-track,
    .ve-upload-list-status-ring-path {
      fill: none;
      stroke-width: 2px;
      stroke-linecap: round;
    }

    .ve-upload-list-status-ring-track {
      stroke: var(--ve-upload-text-item-color-progress-bg-hover);
    }

    .ve-upload-list-status-ring-path {
      stroke: var(--ve-upload-text-item-color-progress-bg-hover-active);
      stroke-dasharray: 100;
      stroke-dashoffset: calc(100 - var(--ve-upload-progress-percent, 0));
      transition: stroke-dashoffset var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-standard);
    }

    .ve-upload-list-status::after {
      content: '';
      inline-size: 9px;
      block-size: 9px;
      border-radius: 50%;
      background: var(--ve-upload-text-item-color-bg);
      position: absolute;
      inset: 0;
      margin: auto;
      z-index: 0;
    }

    .ve-upload-list-status-label {
      display: none;
    }

    .ve-upload-list-status-line {
      display: block;
      inline-size: 100%;
      block-size: auto;
    }

    .ve-upload-list-status-line .ve-upload-list-status-ring,
    .ve-upload-list-status-line::after {
      display: none;
    }

    .ve-upload-list-progress {
      display: flex;
      align-items: center;
      gap: 8px;
      inline-size: 100%;
    }

    .ve-upload-list-progress-line {
      flex: 1;
      min-inline-size: 0;
      block-size: 4px;
      border-radius: 999px;
      background: var(--ve-upload-text-item-color-progress-bg-hover);
      overflow: hidden;
    }

    .ve-upload-list-progress-line-fill {
      display: block;
      block-size: 100%;
      inline-size: calc(var(--ve-upload-progress-percent, 0) * 1%);
      border-radius: 999px;
      background: var(--ve-upload-text-item-color-progress-bg-hover-active);
      transition: inline-size var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-standard);
    }

    .ve-upload-list-progress-text {
      flex: none;
      font-size: 12px;
      color: var(--ve-upload-tip-color-text);
      line-height: 1;
      white-space: nowrap;
    }

    .ve-upload-list-item-done .ve-upload-list-file-icon {
      color: var(--ve-upload-text-item-color-file-icon-success);
    }

    .ve-upload-list-item {
      box-sizing: border-box;
      padding-right: calc(
        var(--ve-upload-text-item-size-operation-icon) +
          var(--ve-upload-text-item-margin-remove-icon-left)
      );
      margin-top: var(--ve-upload-text-item-margin-top);
      position: relative;
    }

    .ve-upload-list-item-operation {
      font-size: var(--ve-upload-text-item-size-operation-icon);
      position: absolute;
      right: 0;
      top: 50%;
      transform: translateY(-50%);
      color: var(--ve-upload-text-item-color-remove-icon);
      line-height: 1;
    }

    .ve-upload-list-item-text {
      font-size: var(--ve-upload-text-item-font-size);
      display: flex;
      align-items: center;
      border-radius: var(--ve-upload-text-item-border-radius);
      inline-size: 100%;
      box-sizing: border-box;
      background-color: var(--ve-upload-text-item-color-bg);
      padding: var(--ve-upload-text-item-padding-vertical) 0;
      padding-left: var(--ve-upload-text-item-padding-left);
      padding-right: var(--ve-upload-text-item-padding-right);
      flex-wrap: nowrap;
    }

    .ve-upload-list-item-text-content {
      flex: 1;
      display: flex;
      align-items: center;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
      justify-content: space-between;
      flex-wrap: nowrap;
      transition: background-color var(--ve-upload-transition-duration-1)
        var(--ve-upload-transition-timing-linear);
      min-inline-size: 0;
    }

    .ve-upload-list-item-text-thumbnail {
      block-size: var(--ve-upload-text-item-thumbnail-width);
      inline-size: var(--ve-upload-text-item-thumbnail-width);
      margin-right: var(--ve-upload-text-item-margin-thumbnail-right);
      flex-shrink: 0;
      border-radius: var(--radius-xs);
      overflow: hidden;
      background: var(--ve-upload-picture-item-color-bg);
    }

    .ve-upload-list-item-text-thumbnail img {
      inline-size: 100%;
      block-size: 100%;
      object-fit: cover;
    }

    .ve-upload-list-item-text-name {
      white-space: nowrap;
      display: flex;
      overflow: hidden;
      flex-shrink: 1;
      flex-grow: 1;
      align-items: center;
      color: var(--ve-upload-text-item-color-text);
      font-size: var(--ve-upload-text-item-font-size);
      text-overflow: ellipsis;
      line-height: 1.4286;
      margin-right: var(--ve-upload-text-item-margin-status-left);
      min-inline-size: 0;
    }

    .ve-upload-list-item-text-name-link {
      cursor: pointer;
      text-decoration: none;
      overflow: hidden;
      color: var(--ve-upload-text-item-color-link);
      text-overflow: ellipsis;
      white-space: nowrap;
      transition: opacity var(--ve-upload-transition-duration-1)
        var(--ve-upload-transition-timing-linear);
    }

    .ve-upload-list-item-text-name-link:hover {
      opacity: 0.85;
    }

    .ve-upload-list-item-text-name-text {
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-upload-list-item-error .ve-upload-list-status,
    .ve-upload-list-item-done .ve-upload-list-status {
      display: none;
    }

    .ve-upload-list-type-text
      .ve-upload-list-item-error
      .ve-upload-list-item-text-name-link,
    .ve-upload-list-type-text
      .ve-upload-list-item-error
      .ve-upload-list-item-text-name {
      color: var(--ve-upload-text-item-color-text-error);
    }

    .ve-upload-list-type-text
      .ve-upload-list-item-done
      .ve-upload-list-item-text-name {
      color: var(--ve-upload-text-item-color-text-success);
    }

    .ve-upload-list-type-text
      .ve-upload-list-item-uploading
      .ve-upload-list-item-text-name {
      color: var(--ve-upload-text-item-color-text-uploading);
    }

    .ve-upload-list-type-picture-card {
      display: inline;
      vertical-align: top;
    }

    .ve-upload-list-type-picture-card .ve-upload-list-status {
      top: 50%;
      transform: translateY(-50%);
      margin-left: 0;
      inline-size: 24px;
      block-size: 24px;
    }

    .ve-upload-list-type-picture-card .ve-upload-list-status-ring {
      inline-size: 24px;
      block-size: 24px;
    }

    .ve-upload-list-type-picture-card .ve-upload-list-status-ring-track {
      stroke: var(--ve-upload-picture-item-color-progress-bg-hover);
    }

    .ve-upload-list-type-picture-card .ve-upload-list-status-ring-path {
      stroke: var(--ve-upload-picture-item-color-progress-bg-hover-active);
    }

    .ve-upload-list-type-picture-card .ve-upload-list-status::after {
      inline-size: 16px;
      block-size: 16px;
      background: rgb(0 0 0 / 50%);
    }

    .ve-upload-list-type-picture-card .ve-upload-list-status-label {
      display: block;
      font-size: 0;
    }

    .ve-upload-list-type-picture-card .ve-upload-list-item {
      display: inline-block;
      vertical-align: top;
      margin-top: 0;
      padding-right: 0;
      margin-right: var(--ve-upload-picture-item-margin-right);
      margin-bottom: var(--ve-upload-picture-item-margin-bottom);
      overflow: hidden;
      transition: all var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-standard);
    }

    .ve-upload-list-type-picture-card
      .ve-upload-list-item-error
      .ve-upload-list-item-picture-mask {
      opacity: 1;
    }

    .ve-upload-list-item-picture {
      inline-size: var(--ve-upload-picture-item-size-width);
      block-size: var(--ve-upload-picture-item-size-width);
      position: relative;
      overflow: hidden;
      border-radius: var(--ve-upload-picture-item-border-radius);
      box-sizing: border-box;
      text-align: center;
      vertical-align: top;
      background-color: var(--ve-upload-picture-item-color-bg);
      display: inline-flex;
      align-items: center;
      justify-content: center;
    }

    .ve-upload-list-item-picture img {
      inline-size: 100%;
      block-size: 100%;
      object-fit: cover;
    }

    .ve-upload-list-item-picture-progress {
      inline-size: 100%;
      block-size: 100%;
      background-color: var(--ve-upload-picture-item-color-operation-bg);
      display: flex;
      align-items: center;
      justify-content: center;
    }

    .ve-upload-list-item-picture-mask {
      cursor: pointer;
      position: absolute;
      text-align: center;
      inset: 0;
      background: var(--ve-upload-picture-item-color-operation-bg);
      color: var(--ve-upload-picture-item-color-operation-icon);
      font-size: var(--ve-upload-picture-item-size-mask-icon);
      line-height: var(--ve-upload-picture-item-size-width);
      opacity: 0;
      transition: opacity var(--ve-upload-transition-duration-1)
        var(--ve-upload-transition-timing-linear);
    }

    .ve-upload-list-item-picture-operation {
      font-size: var(--ve-upload-picture-item-size-operation-icon);
      line-height: 1;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      gap: var(--ve-upload-picture-item-margin-preview-icon-right);
      position: absolute;
      inset: 0;
      margin: auto;
      block-size: fit-content;
      inline-size: fit-content;
    }

    .ve-upload-list-item-picture-operation .ve-upload-list-reupload-icon {
      color: var(--ve-upload-picture-item-color-operation-icon);
    }

    .ve-upload-list-item-picture-operation .ve-upload-list-reupload-icon,
    .ve-upload-list-item-picture-operation .ve-upload-list-preview-icon,
    .ve-upload-list-item-picture-operation .ve-upload-list-remove-icon {
      color: var(--ve-upload-picture-item-color-operation-icon);
      border-radius: 2px;
    }

    .ve-upload-list-item-picture-operation
      .ve-upload-list-reupload-icon:focus-visible,
    .ve-upload-list-item-picture-operation
      .ve-upload-list-preview-icon:focus-visible,
    .ve-upload-list-item-picture-operation
      .ve-upload-list-remove-icon:focus-visible {
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
      outline: none;
    }

    .ve-upload-list-item-picture-error-tip {
      position: absolute;
      inset: 0;
      margin: auto;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: fit-content;
      block-size: fit-content;
    }

    .ve-upload-list-item-picture-error-tip
      .ve-upload-list-item-picture-operation {
      opacity: 0;
      inline-size: 0;
      block-size: 0;
    }

    .ve-upload-list-item-picture-error-tip .ve-upload-list-error-icon {
      font-size: var(--ve-upload-picture-item-size-error-icon);
      color: var(--ve-upload-picture-item-color-error-icon);
      margin-left: 0;
    }

    .ve-upload-list-item-picture-mask:hover,
    .ve-upload-list-item-picture-mask:focus-within,
    .ve-upload-list-item-picture-mask:hover
      .ve-upload-list-item-picture-operation,
    .ve-upload-list-item-picture-mask:focus-within
      .ve-upload-list-item-picture-operation {
      opacity: 1;
      display: block;
    }

    .ve-upload-list-item-picture-mask:hover
      .ve-upload-list-item-picture-error-tip,
    .ve-upload-list-item-picture-mask:focus-within
      .ve-upload-list-item-picture-error-tip {
      display: none;
    }

    .ve-upload-list-type-picture-list .ve-upload-list-item-text {
      padding-top: var(--ve-upload-picture-text-item-padding-vertical);
      padding-bottom: var(--ve-upload-picture-text-item-padding-vertical);
    }

    .ve-upload-list-type-picture-list
      .ve-upload-list-item-error
      .ve-upload-list-item-text {
      background-color: var(--ve-upload-picture-text-item-color-bg-error);
    }

    .ve-upload-list-type-picture-list
      .ve-upload-list-item-error
      .ve-upload-list-item-text-name-link,
    .ve-upload-list-type-picture-list
      .ve-upload-list-item-error
      .ve-upload-list-item-text-name {
      color: var(--ve-upload-picture-text-item-color-text-error);
    }

    .ve-upload-list-type-picture-list
      .ve-upload-list-item-done
      .ve-upload-list-item-text-name {
      color: var(--ve-upload-picture-text-item-color-text-success);
    }

    .ve-upload-list-type-picture-list
      .ve-upload-list-item-uploading
      .ve-upload-list-item-text-name {
      color: var(--ve-upload-picture-text-item-color-text-uploading);
    }

    .ve-upload-list-type-card,
    .ve-upload-list-type-row,
    .ve-upload-list-type-tag {
      display: flex;
      flex-wrap: wrap;
      align-items: center;
      gap: var(--ve-upload-list-gap);
      inline-size: auto;
    }

    .ve-upload-root
      > .ve-upload:not(.ve-upload-hide)
      + .ve-upload-list-type-row,
    .ve-upload-root
      > .ve-upload:not(.ve-upload-hide)
      + .ve-upload-list-type-tag {
      margin-top: var(--ve-upload-list-margin-top);
    }

    .ve-upload-list-type-card {
      display: inline-flex;
      vertical-align: top;
    }

    .ve-upload-card-stack {
      display: inline-flex;
      flex-wrap: wrap;
      align-items: center;
      gap: var(--ve-upload-list-gap);
      max-inline-size: 100%;
      vertical-align: top;
    }

    .ve-upload-card-stack > .ve-upload-list-type-card {
      display: contents;
    }

    .ve-upload-type-card {
      inline-size: auto;
      vertical-align: top;
    }

    .ve-upload-type-card .ve-upload-trigger-picture {
      inline-size: var(--ve-upload-attachment-card-size);
      min-inline-size: 0;
      block-size: var(--ve-upload-attachment-card-size);
      border-radius: var(--ve-upload-attachment-radius);
      background: var(--ve-upload-attachment-color-bg);
      border-width: var(--stroke-weight-base);
      border-color: var(--color-border-default);
    }

    .ve-upload-type-card .ve-upload-trigger-picture:hover {
      border-color: var(--color-border-strong);
      background: var(--ve-upload-attachment-color-bg-hover);
    }

    .ve-upload-list-type-card .ve-upload-list-item,
    .ve-upload-list-type-row .ve-upload-list-item,
    .ve-upload-list-type-tag .ve-upload-list-item {
      margin: 0;
      padding: 0;
    }

    .ve-upload-attachment-card,
    .ve-upload-attachment-row,
    .ve-upload-attachment-tag {
      position: relative;
      cursor: pointer;
      background: var(--ve-upload-attachment-color-bg);
      color: var(--ve-upload-attachment-color-text);
      transition:
        background-color var(--ve-upload-transition-duration-1)
          var(--ve-upload-transition-timing-linear),
        opacity var(--ve-upload-transition-duration-1)
          var(--ve-upload-transition-timing-linear);
    }

    .ve-upload-attachment-card:hover,
    .ve-upload-attachment-row:hover,
    .ve-upload-attachment-tag:hover {
      background: var(--ve-upload-attachment-color-bg-hover);
    }

    .ve-upload-attachment-card:focus-visible,
    .ve-upload-attachment-row:focus-visible,
    .ve-upload-attachment-tag:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
    }

    :host([disabled]) .ve-upload-attachment-card,
    :host([disabled]) .ve-upload-attachment-row,
    :host([disabled]) .ve-upload-attachment-tag {
      opacity: var(--state-disabled-ratio);
    }

    .ve-upload-attachment-card {
      inline-size: var(--ve-upload-attachment-card-size);
      block-size: var(--ve-upload-attachment-card-size);
      border-radius: var(--ve-upload-attachment-radius);
      box-sizing: border-box;
      display: flex;
      flex-direction: column;
      overflow: visible;
    }

    .ve-upload-list-item.ve-upload-attachment-card-file,
    .ve-upload-list-item.ve-upload-attachment-card.ve-upload-attachment-loading {
      justify-content: space-between;
      padding: var(--ve-upload-attachment-card-padding);
    }

    .ve-upload-attachment-card-file.ve-upload-list-item-error {
      background: var(--ve-upload-attachment-color-bg);
    }

    .ve-upload-attachment-card-title {
      display: -webkit-box;
      -webkit-box-orient: vertical;
      -webkit-line-clamp: 2;
      overflow: hidden;
      min-inline-size: 0;
      max-block-size: 54px;
      font-size: var(--ve-upload-attachment-title-font-size);
      line-height: var(--ve-upload-attachment-card-title-line-height);
      color: var(--ve-upload-attachment-color-text);
      word-break: break-word;
    }

    .ve-upload-attachment-title-link,
    .ve-upload-attachment-title-text {
      color: inherit;
      text-decoration: none;
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ve-upload-attachment-title-link:hover {
      color: var(--color-text-accent);
    }

    .ve-upload-attachment-card-footer {
      min-inline-size: 0;
    }

    .ve-upload-attachment-card-extension {
      display: flex;
      align-items: center;
      min-block-size: 18px;
      border-radius: var(--ve-upload-attachment-radius-badge);
      padding-inline: var(--ve-upload-attachment-card-meta-inline);
      color: var(--ve-upload-attachment-color-text-secondary);
      font-size: var(--ve-upload-attachment-meta-font-size);
      line-height: 1.5;
      text-transform: uppercase;
    }

    .ve-upload-attachment-card-media {
      padding: 0;
    }

    .ve-upload-attachment-media {
      position: relative;
      display: flex;
      align-items: center;
      justify-content: center;
      inline-size: 100%;
      block-size: 100%;
      overflow: hidden;
      border-radius: var(--ve-upload-attachment-radius);
      cursor: pointer;
      color: var(--ve-upload-attachment-color-icon-foreground);
    }

    .ve-upload-attachment-media:focus-visible {
      outline: none;
      box-shadow: inset 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
    }

    .ve-upload-attachment-media img,
    .ve-upload-attachment-thumb img {
      inline-size: 100%;
      block-size: 100%;
      object-fit: cover;
    }

    .ve-upload-list-item-error .ve-upload-attachment-media img,
    .ve-upload-list-item-error .ve-upload-attachment-thumb img {
      filter: grayscale(1);
    }

    .ve-upload-attachment-media-gradient {
      position: absolute;
      inset-inline: 0;
      bottom: 0;
      block-size: 64px;
      background: linear-gradient(
        180deg,
        var(--ve-upload-attachment-color-gradient-start) 0%,
        var(--ve-upload-attachment-color-gradient-end) 100%
      );
      pointer-events: none;
    }

    .ve-upload-attachment-card-media .ve-upload-attachment-card-extension {
      position: absolute;
      inset-inline: var(--space-xxs);
      bottom: var(--space-xxs);
      z-index: 1;
      color: var(--ve-upload-attachment-color-icon-foreground);
    }

    .ve-upload-attachment-card-error {
      position: absolute;
      inset-inline: var(--space-xxs);
      bottom: var(--space-xxs);
      z-index: 1;
    }

    .ve-upload-list-item.ve-upload-attachment-row {
      inline-size: var(--ve-upload-attachment-row-width);
      min-block-size: var(--ve-upload-attachment-row-height);
      border-radius: var(--ve-upload-attachment-radius);
      display: flex;
      align-items: center;
      gap: var(--space-xxs);
      padding: var(--ve-upload-attachment-row-padding-block)
        var(--ve-upload-attachment-row-padding-inline);
      box-sizing: border-box;
    }

    .ve-upload-attachment-row-content {
      position: relative;
      z-index: 1;
      min-inline-size: 0;
      flex: 1;
      display: flex;
      flex-direction: column;
      overflow: hidden;
    }

    .ve-upload-attachment-row-title,
    .ve-upload-attachment-tag-title {
      min-inline-size: 0;
      color: var(--ve-upload-attachment-color-text);
      font-size: var(--ve-upload-attachment-title-font-size);
      line-height: var(--ve-upload-attachment-card-title-line-height);
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-upload-attachment-row-title {
      inline-size: 100%;
    }

    .ve-upload-attachment-row-title .ve-upload-attachment-title-link,
    .ve-upload-attachment-row-title .ve-upload-attachment-title-text,
    .ve-upload-attachment-tag-title .ve-upload-attachment-title-link,
    .ve-upload-attachment-tag-title .ve-upload-attachment-title-text {
      display: block;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    .ve-upload-list-item.ve-upload-attachment-tag {
      min-block-size: var(--ve-upload-attachment-tag-height);
      max-inline-size: var(--ve-upload-attachment-tag-max-width);
      border-radius: var(--ve-upload-attachment-radius);
      display: inline-flex;
      align-items: center;
      gap: var(--space-xxxs);
      padding: var(--ve-upload-attachment-tag-padding);
      box-sizing: border-box;
    }

    .ve-upload-attachment-tag-title {
      position: relative;
      z-index: 1;
      flex: 1;
    }

    .ve-upload-attachment-thumb {
      position: relative;
      z-index: 1;
      flex: none;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      overflow: hidden;
      border-radius: var(--ve-upload-attachment-radius-inner);
      color: var(--ve-upload-attachment-color-icon);
      background: var(--color-bg-control);
    }

    .ve-upload-attachment-thumb-file {
      background: var(--ve-upload-attachment-color-icon-bg);
    }

    .ve-upload-attachment-thumb-media {
      background: var(--color-bg-control);
    }

    .ve-upload-attachment-thumb-row {
      inline-size: var(--ve-upload-attachment-row-icon-size);
      block-size: var(--ve-upload-attachment-row-icon-size);
    }

    .ve-upload-attachment-thumb-tag {
      inline-size: var(--ve-upload-attachment-tag-icon-size);
      block-size: var(--ve-upload-attachment-tag-icon-size);
      border-radius: var(--ve-upload-attachment-radius-badge);
    }

    .ve-upload-attachment-file-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: 100%;
      block-size: 100%;
      font-size: 18px;
      color: var(--ve-upload-attachment-color-icon);
    }

    .ve-upload-attachment-thumb-tag .ve-upload-attachment-file-icon {
      font-size: 14px;
    }

    .ve-upload-attachment-play {
      position: absolute;
      inset: 0;
      margin: auto;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-upload-attachment-video-play-size);
      block-size: var(--ve-upload-attachment-video-play-size);
      border-radius: var(--radius-full);
      background: var(--ve-upload-attachment-video-play-bg);
      color: var(--ve-upload-attachment-color-icon-foreground);
      font-size: 10px;
    }

    .ve-upload-attachment-meta {
      display: flex;
      align-items: center;
      gap: var(--space-xxs);
      min-inline-size: 0;
      color: var(--ve-upload-attachment-color-text-secondary);
      font-size: var(--ve-upload-attachment-meta-font-size);
      line-height: 1.5;
      white-space: nowrap;
    }

    .ve-upload-attachment-meta span {
      overflow: hidden;
      text-overflow: ellipsis;
    }

    .ve-upload-attachment-meta-error {
      gap: 2px;
      color: var(--ve-upload-attachment-color-text-danger);
    }

    .ve-upload-attachment-card-file .ve-upload-attachment-meta-error {
      gap: var(--space-xxxs);
      font-size: var(--text-body);
      line-height: 1.5;
    }

    .ve-upload-attachment-error-icon {
      display: inline-flex;
      align-items: center;
      justify-content: center;
      flex: none;
      inline-size: 14px;
      block-size: 14px;
      font-size: 14px;
      color: var(--ve-upload-attachment-color-text-danger);
    }

    .ve-upload-attachment-card-file .ve-upload-attachment-error-icon {
      inline-size: 18px;
      block-size: 18px;
      font-size: 18px;
    }

    .ve-upload-attachment-tag .ve-upload-attachment-meta-error {
      flex: none;
      inline-size: 14px;
      overflow: hidden;
    }

    .ve-upload-attachment-tag .ve-upload-attachment-meta-error span:last-child {
      display: none;
    }

    .ve-upload-attachment-remove {
      position: absolute;
      z-index: 2;
      top: -6px;
      right: -6px;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      inline-size: var(--ve-upload-attachment-remove-size);
      block-size: var(--ve-upload-attachment-remove-size);
      border-radius: var(--radius-full);
      color: var(--ve-upload-attachment-remove-color);
      background: var(--ve-upload-attachment-remove-bg);
      font-size: var(--ve-upload-attachment-remove-icon-size);
      line-height: 1;
      cursor: pointer;
      opacity: 0;
      pointer-events: none;
      transform: scale(0.92);
      transition:
        opacity var(--ve-upload-transition-duration-1)
          var(--ve-upload-transition-timing-linear),
        transform var(--ve-upload-transition-duration-1)
          var(--ve-upload-transition-timing-linear);
    }

    .ve-upload-attachment-card:hover .ve-upload-attachment-remove,
    .ve-upload-attachment-card:focus-within .ve-upload-attachment-remove,
    .ve-upload-attachment-row:hover .ve-upload-attachment-remove,
    .ve-upload-attachment-row:focus-within .ve-upload-attachment-remove,
    .ve-upload-attachment-tag:hover .ve-upload-attachment-remove,
    .ve-upload-attachment-tag:focus-within .ve-upload-attachment-remove,
    .ve-upload-attachment-remove:focus-visible {
      opacity: var(--ve-upload-attachment-remove-opacity);
      pointer-events: auto;
      transform: scale(1);
    }

    .ve-upload-attachment-card .ve-upload-attachment-remove {
      top: -6px;
      right: -6px;
    }

    .ve-upload-attachment-remove:hover {
      opacity: 1;
      transform: scale(1.04);
    }

    .ve-upload-attachment-tag .ve-upload-attachment-remove {
      top: 50%;
      right: var(--space-xxs);
      z-index: 3;
      color: var(--ve-upload-attachment-color-text);
      background: var(--ve-upload-attachment-color-bg-hover);
      border-radius: var(--radius-xs);
      transform: translateY(-50%) scale(0.92);
      box-shadow: calc(var(--space-s) * -1) 0 var(--space-s)
        var(--ve-upload-attachment-color-bg-hover);
    }

    .ve-upload-attachment-tag:hover .ve-upload-attachment-remove,
    .ve-upload-attachment-tag:focus-within .ve-upload-attachment-remove,
    .ve-upload-attachment-tag .ve-upload-attachment-remove:focus-visible {
      opacity: 1;
      transform: translateY(-50%) scale(1);
    }

    .ve-upload-attachment-tag .ve-upload-attachment-remove:hover {
      transform: translateY(-50%) scale(1.04);
    }

    .ve-upload-attachment-remove:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
    }

    .ve-upload-remove-svg {
      display: block;
      inline-size: var(--ve-upload-attachment-remove-icon-size);
      block-size: var(--ve-upload-attachment-remove-icon-size);
    }

    .ve-upload-attachment-card.ve-upload-attachment-loading {
      pointer-events: none;
    }

    .ve-upload-attachment-progress-fill {
      position: absolute;
      z-index: 0;
      inset-block: 0;
      inset-inline-start: 0;
      inline-size: var(--ve-upload-attachment-progress-size, 0%);
      border-radius: var(--ve-upload-attachment-radius) 0 0
        var(--ve-upload-attachment-radius);
      background: var(--ve-upload-attachment-progress-bg);
      opacity: var(--ve-upload-attachment-progress-opacity);
      pointer-events: none;
      overflow: hidden;
    }

    .ve-upload-attachment-skeleton-lines {
      display: flex;
      flex-direction: column;
      gap: var(--space-xxxs);
      inline-size: 100%;
    }

    .ve-upload-attachment-skeleton-lines span,
    .ve-upload-attachment-skeleton-meta,
    .ve-upload-attachment-skeleton-title,
    .ve-upload-attachment-skeleton-subtitle {
      display: block;
      border-radius: var(--ve-upload-attachment-radius-badge);
      background: var(--ve-upload-attachment-color-skeleton);
      overflow: hidden;
      position: relative;
    }

    .ve-upload-attachment-progress-fill::after,
    .ve-upload-attachment-skeleton-lines span::after,
    .ve-upload-attachment-skeleton-meta::after,
    .ve-upload-attachment-skeleton-title::after,
    .ve-upload-attachment-skeleton-subtitle::after {
      content: '';
      position: absolute;
      inset: 0;
      background: linear-gradient(
        90deg,
        transparent 0%,
        var(--ve-upload-attachment-loading-shimmer-color) 50%,
        transparent 100%
      );
      transform: translateX(-100%);
      animation: ve-upload-loading-shimmer
        var(--ve-upload-attachment-loading-shimmer-duration) ease-in-out
        infinite;
    }

    .ve-upload-attachment-skeleton-lines span {
      block-size: 13px;
    }

    .ve-upload-attachment-skeleton-lines span:first-child {
      inline-size: 100%;
    }

    .ve-upload-attachment-skeleton-lines span:last-child {
      inline-size: 58px;
    }

    .ve-upload-attachment-skeleton-meta {
      inline-size: 58px;
      block-size: 13px;
    }

    .ve-upload-attachment-skeleton-title {
      inline-size: 100%;
      block-size: 13px;
      margin-bottom: var(--space-xxxs);
    }

    .ve-upload-attachment-skeleton-subtitle {
      inline-size: 58px;
      block-size: 13px;
    }

    @keyframes ve-upload-loading-shimmer {
      100% {
        transform: translateX(100%);
      }
    }

    .ve-upload-slide-up-enter {
      opacity: 0;
    }

    .ve-upload-slide-up-enter-active {
      opacity: 1;
      transition: opacity var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-standard);
    }

    .ve-upload-slide-up-exit {
      opacity: 1;
    }

    .ve-upload-slide-up-exit-active {
      opacity: 0;
      overflow: hidden;
      margin: 0;
      transition:
        opacity var(--ve-upload-transition-duration-1)
          var(--ve-upload-transition-timing-linear),
        block-size var(--ve-upload-transition-duration-3)
          var(--ve-upload-transition-timing-standard)
          var(--ve-upload-transition-duration-1),
        margin var(--ve-upload-transition-duration-3)
          var(--ve-upload-transition-timing-standard)
          var(--ve-upload-transition-duration-1);
    }

    .ve-upload-list-item.ve-upload-slide-inline-enter {
      opacity: 0;
    }

    .ve-upload-list-item.ve-upload-slide-inline-enter-active {
      opacity: 1;
      transition: opacity var(--ve-upload-transition-duration-2)
        var(--ve-upload-transition-timing-linear);
    }

    .ve-upload-list-item.ve-upload-slide-inline-exit {
      opacity: 1;
    }

    .ve-upload-list-item.ve-upload-slide-inline-exit-active {
      opacity: 0;
      overflow: hidden;
      margin: 0;
      transition:
        opacity var(--ve-upload-transition-duration-1)
          var(--ve-upload-transition-timing-linear),
        inline-size var(--ve-upload-transition-duration-3)
          var(--ve-upload-transition-timing-standard)
          var(--ve-upload-transition-duration-1),
        margin var(--ve-upload-transition-duration-3)
          var(--ve-upload-transition-timing-standard)
          var(--ve-upload-transition-duration-1);
    }

    .ve-upload-rtl {
      direction: rtl;
    }

    .ve-upload-list-rtl {
      direction: rtl;
    }

    .ve-upload-list-rtl .ve-upload-list-file-icon {
      margin-right: 0;
      margin-left: var(--ve-upload-text-item-margin-file-icon-right);
    }

    .ve-upload-list-rtl .ve-upload-list-error-icon {
      margin-left: 0;
      margin-right: var(--ve-upload-text-item-margin-error-icon-left);
    }

    .ve-upload-list-rtl .ve-upload-list-item {
      padding-right: 0;
      padding-left: calc(
        var(--ve-upload-text-item-size-operation-icon) +
          var(--ve-upload-text-item-margin-remove-icon-left)
      );
    }

    .ve-upload-list-rtl .ve-upload-list-item-operation {
      right: auto;
      left: 0;
    }

    .ve-upload-list-rtl .ve-upload-list-item-text {
      padding-right: var(--ve-upload-text-item-padding-left);
      padding-left: var(--ve-upload-text-item-padding-right);
    }

    .ve-upload-list-rtl .ve-upload-list-item-text-thumbnail {
      margin-right: 0;
      margin-left: var(--ve-upload-text-item-margin-thumbnail-right);
    }

    .ve-upload-list-rtl .ve-upload-list-item-text-name {
      margin-right: 0;
      margin-left: var(--ve-upload-text-item-margin-status-left);
    }

    .ve-upload-list-rtl
      .ve-upload-list-item-picture-operation
      .ve-upload-list-reupload-icon
      + .ve-upload-list-remove-icon,
    .ve-upload-list-rtl
      .ve-upload-list-item-picture-operation
      .ve-upload-list-preview-icon
      + .ve-upload-list-remove-icon {
      margin-left: 0;
      margin-right: var(--ve-upload-picture-item-margin-preview-icon-right);
    }

    .ve-upload-list-rtl.ve-upload-list-type-picture-card
      .ve-upload-list-status {
      margin-left: auto;
      margin-right: 0;
    }

    .ve-upload-list-rtl.ve-upload-list-type-picture-card .ve-upload-list-item {
      margin-right: 0;
      margin-left: var(--ve-upload-picture-item-margin-right);
    }

    .ve-upload-preview {
      position: fixed;
      inset: 0;
      background: var(--ve-upload-preview-mask-bg);
      z-index: 1000;
      display: flex;
      align-items: center;
      justify-content: center;
      padding: var(--space-s);
      box-sizing: border-box;
    }

    .ve-upload-preview img {
      max-inline-size: min(90vw, 980px);
      max-block-size: 90vh;
      border-radius: var(--radius-sm);
      object-fit: contain;
      box-shadow: 0 12px 36px rgb(0 0 0 / 28%);
      background: color-mix(var(--color-bg-base) 95%, transparent);
    }

    .ve-upload-preview-close,
    .ve-upload-preview-prev,
    .ve-upload-preview-next {
      position: fixed;
      border: none;
      background: var(--ve-upload-preview-action-bg);
      color: var(--ve-upload-preview-action-color);
      cursor: pointer;
      inline-size: 36px;
      block-size: 36px;
      border-radius: 999px;
      line-height: 1;
      font-size: 20px;
      display: inline-flex;
      align-items: center;
      justify-content: center;
      transition: background-color var(--ve-upload-transition-duration-1)
        var(--ve-upload-transition-timing-linear);
    }

    .ve-upload-preview-close:hover,
    .ve-upload-preview-prev:hover,
    .ve-upload-preview-next:hover {
      background: color-mix(var(--ve-upload-preview-action-bg) 88%, white);
    }

    .ve-upload-preview-close:focus-visible,
    .ve-upload-preview-prev:focus-visible,
    .ve-upload-preview-next:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px var(--ve-upload-color-icon-box-shadow-default);
    }

    .ve-upload-preview-close {
      top: var(--space-s);
      right: var(--space-s);
    }

    .ve-upload-preview-prev {
      left: var(--space-s);
      top: 50%;
      transform: translateY(-50%);
    }

    .ve-upload-preview-next {
      right: var(--space-s);
      top: 50%;
      transform: translateY(-50%);
    }

    @media (prefers-reduced-motion: reduce) {
      .ve-upload-trigger-picture,
      .ve-upload-trigger-drag,
      .ve-upload-list-reupload-icon,
      .ve-upload-preview-close,
      .ve-upload-preview-prev,
      .ve-upload-preview-next {
        transition: none;
      }

      .ve-upload-attachment-progress-fill::after,
      .ve-upload-attachment-skeleton-lines span::after,
      .ve-upload-attachment-skeleton-meta::after,
      .ve-upload-attachment-skeleton-title::after,
      .ve-upload-attachment-skeleton-subtitle::after {
        animation: none;
      }
    }
  `
];
function toPrimitive(input, hint) {
    if ("object" != typeof input || null === input) return input;
    var prim = input[Symbol.toPrimitive];
    if (void 0 !== prim) {
        var res = prim.call(input, hint || "default");
        if ("object" != typeof res) return res;
        throw new TypeError("@@toPrimitive must return a primitive value.");
    }
    return ("string" === hint ? String : Number)(input);
}
function toPropertyKey(arg) {
    var key = toPrimitive(arg, "string");
    return "symbol" == typeof key ? key : String(key);
}
function checkInRHS(value) {
    if (Object(value) !== value) throw TypeError("right-hand side of 'in' should be an object, got " + (null !== value ? typeof value : "null"));
    return value;
}
function setFunctionName(fn, name, prefix) {
    if ("symbol" == typeof name) {
        name = name.description;
        name = name ? "[" + name + "]" : "";
    }
    try {
        Object.defineProperty(fn, "name", {
            configurable: true,
            value: prefix ? prefix + " " + name : name
        });
    } catch (_) {}
    return fn;
}
function _apply_decs_2311(targetClass, classDecs, memberDecs, classDecsHaveThis, instanceBrand, parentClass) {
    var symbolMetadata = Symbol.metadata || Symbol.for("Symbol.metadata");
    var defineProperty = Object.defineProperty;
    var create = Object.create;
    var metadata;
    var existingNonFields = [
        create(null),
        create(null)
    ];
    var hasClassDecs = classDecs.length;
    var _;
    function createRunInitializers(initializers, useStaticThis, hasValue) {
        return function(thisArg, value) {
            if (useStaticThis) {
                value = thisArg;
                thisArg = targetClass;
            }
            for(var i = 0; i < initializers.length; i++)value = initializers[i].apply(thisArg, hasValue ? [
                value
            ] : []);
            return hasValue ? value : thisArg;
        };
    }
    function assertCallable(fn, hint1, hint2, throwUndefined) {
        if ("function" != typeof fn) {
            if (throwUndefined || void 0 !== fn) throw new TypeError(hint1 + " must " + (hint2 || "be") + " a function" + (throwUndefined ? "" : " or undefined"));
        }
        return fn;
    }
    function applyDec(Class, decInfo, decoratorsHaveThis, name, kind, initializers, ret, isStatic, isPrivate, isField, hasPrivateBrand) {
        function assertInstanceIfPrivate(target) {
            if (!hasPrivateBrand(target)) throw new TypeError("Attempted to access private element on non-instance");
        }
        var decs = [].concat(decInfo[0]);
        var decVal = decInfo[3];
        var isClass = !ret;
        var isAccessor = 1 === kind;
        var isGetter = 3 === kind;
        var isSetter = 4 === kind;
        var isMethod = 2 === kind;
        function bindPropCall(name, useStaticThis, before) {
            return function(_this, value) {
                if (useStaticThis) {
                    value = _this;
                    _this = Class;
                }
                if (before) before(_this);
                return desc[name].call(_this, value);
            };
        }
        var desc1 = {};
        var init = [];
        var key = isGetter ? "get" : isSetter || isAccessor ? "set" : "value";
        if (!isClass) {
            if (isPrivate) {
                if (isField || isAccessor) desc1 = {
                    get: setFunctionName(function() {
                        return decVal(this);
                    }, name, "get"),
                    set: function(value) {
                        decInfo[4](this, value);
                    }
                };
                else desc1[key] = decVal;
                if (!isField) setFunctionName(desc1[key], name, isMethod ? "" : key);
            } else if (!isField) desc1 = Object.getOwnPropertyDescriptor(Class, name);
            if (!isField && !isPrivate) {
                _ = existingNonFields[+isStatic][name];
                if (_ && (_ ^ kind) !== 7) throw new Error("Decorating two elements with the same name (" + desc1[key].name + ") is not supported yet");
                existingNonFields[+isStatic][name] = kind < 3 ? 1 : kind;
            }
        }
        var newValue = Class;
        for(var i = decs.length - 1; i >= 0; i -= decoratorsHaveThis ? 2 : 1){
            var dec = assertCallable(decs[i], "A decorator", "be", true);
            var decThis = decoratorsHaveThis ? decs[i - 1] : void 0;
            var decoratorFinishedRef = {};
            var ctx = {
                kind: [
                    "field",
                    "accessor",
                    "method",
                    "getter",
                    "setter",
                    "class"
                ][kind],
                name: name,
                metadata: metadata,
                addInitializer: (function(decoratorFinishedRef, initializer) {
                    if (decoratorFinishedRef.v) throw new TypeError("attempted to call addInitializer after decoration was finished");
                    assertCallable(initializer, "An initializer", "be", true);
                    initializers.push(initializer);
                }).bind(null, decoratorFinishedRef)
            };
            if (isClass) {
                _ = dec.call(decThis, newValue, ctx);
                decoratorFinishedRef.v = 1;
                if (assertCallable(_, "class decorators", "return")) newValue = _;
                continue;
            }
            ctx.static = isStatic;
            ctx.private = isPrivate;
            _ = ctx.access = {
                has: isPrivate ? hasPrivateBrand.bind() : function(target) {
                    return name in target;
                }
            };
            if (!isSetter) _.get = isPrivate ? isMethod ? function(_this) {
                assertInstanceIfPrivate(_this);
                return desc1.value;
            } : bindPropCall("get", 0, assertInstanceIfPrivate) : function(target) {
                return target[name];
            };
            if (!isMethod && !isGetter) _.set = isPrivate ? bindPropCall("set", 0, assertInstanceIfPrivate) : function(target, value) {
                target[name] = value;
            };
            newValue = dec.call(decThis, isAccessor ? {
                get: desc1.get,
                set: desc1.set
            } : desc1[key], ctx);
            decoratorFinishedRef.v = 1;
            if (isAccessor) {
                if ("object" == typeof newValue && newValue) {
                    if (_ = assertCallable(newValue.get, "accessor.get")) desc1.get = _;
                    if (_ = assertCallable(newValue.set, "accessor.set")) desc1.set = _;
                    if (_ = assertCallable(newValue.init, "accessor.init")) init.unshift(_);
                } else if (void 0 !== newValue) throw new TypeError("accessor decorators must return an object with get, set, or init properties or undefined");
            } else if (assertCallable(newValue, (isField ? "field" : "method") + " decorators", "return")) if (isField) init.unshift(newValue);
            else desc1[key] = newValue;
        }
        if (kind < 2) ret.push(createRunInitializers(init, isStatic, 1), createRunInitializers(initializers, isStatic, 0));
        if (!isField && !isClass) if (isPrivate) if (isAccessor) ret.splice(-1, 0, bindPropCall("get", isStatic), bindPropCall("set", isStatic));
        else ret.push(isMethod ? desc1[key] : assertCallable.call.bind(desc1[key]));
        else defineProperty(Class, name, desc1);
        return newValue;
    }
    function applyMemberDecs() {
        var ret = [];
        var protoInitializers;
        var staticInitializers;
        var pushInitializers = function(initializers) {
            if (initializers) ret.push(createRunInitializers(initializers));
        };
        var applyMemberDecsOfKind = function(isStatic, isField) {
            for(var i = 0; i < memberDecs.length; i++){
                var decInfo = memberDecs[i];
                var kind = decInfo[1];
                var kindOnly = 7 & kind;
                if ((8 & kind) == isStatic && !kindOnly == isField) {
                    var name = decInfo[2];
                    var isPrivate = !!decInfo[3];
                    var decoratorsHaveThis = 16 & kind;
                    applyDec(isStatic ? targetClass : targetClass.prototype, decInfo, decoratorsHaveThis, isPrivate ? "#" + name : toPropertyKey(name), kindOnly, kindOnly < 2 ? [] : isStatic ? staticInitializers = staticInitializers || [] : protoInitializers = protoInitializers || [], ret, !!isStatic, isPrivate, isField, isStatic && isPrivate ? function(_) {
                        return checkInRHS(_) === targetClass;
                    } : instanceBrand);
                }
            }
        };
        applyMemberDecsOfKind(8, 0);
        applyMemberDecsOfKind(0, 0);
        applyMemberDecsOfKind(8, 1);
        applyMemberDecsOfKind(0, 1);
        pushInitializers(protoInitializers);
        pushInitializers(staticInitializers);
        return ret;
    }
    function defineMetadata(Class) {
        return defineProperty(Class, symbolMetadata, {
            configurable: true,
            enumerable: true,
            value: metadata
        });
    }
    if (void 0 !== parentClass) metadata = parentClass[symbolMetadata];
    metadata = create(null == metadata ? null : metadata);
    _ = applyMemberDecs();
    if (!hasClassDecs) defineMetadata(targetClass);
    return {
        e: _,
        get c () {
            var initializers = [];
            return hasClassDecs && [
                defineMetadata(targetClass = applyDec(targetClass, [
                    classDecs
                ], classDecsHaveThis, targetClass.name, 5, initializers)),
                createRunInitializers(initializers, 1)
            ];
        }
    };
}
let _dec, _dec1, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _dec14, _dec15, _dec16, _dec17, _dec18, _dec19, _dec20, _dec21, _dec22, _dec23, _dec24, _dec25, _dec26, _dec27, _dec28, _dec29, _dec30, _dec31, _init_action, _init_extra__init_action, _init_method, _init_extra__init_method, _init_name, _init_extra__init_name, _init_accept, _init_extra__init_accept, _init_listType, _init_extra__init_listType, _init_imagePreview, _init_extra__init_imagePreview, _init_multiple, _init_extra__init_multiple, _init_disabled, _init_extra__init_disabled, _init_drag, _init_extra__init_drag, _init_directory, _init_extra__init_directory, _init_withCredentials, _init_extra__init_withCredentials, _init_autoUpload, _init_extra__init_autoUpload, _init_showUploadList, _init_extra__init_showUploadList, _init_limit, _init_extra__init_limit, _init_tip, _init_extra__init_tip, _init_headers, _init_extra__init_headers, _init_data, _init_extra__init_data, _init_customRequest, _init_extra__init_customRequest, _init_beforeUpload, _init_extra__init_beforeUpload, _init_progressProps, _init_extra__init_progressProps, _init_renderUploadItem, _init_extra__init_renderUploadItem, _init_renderUploadList, _init_extra__init_renderUploadList, _init_defaultFileList, _init_extra__init_defaultFileList, _init_fileList, _init_extra__init_fileList, _init_inputElement, _init_extra__init_inputElement, _init_innerFileList, _init_extra__init_innerFileList, _init_hasTriggerSlot, _init_extra__init_hasTriggerSlot, _init_hasTipSlot, _init_extra__init_hasTipSlot, _init_dragActive, _init_extra__init_dragActive, _init_dragEnterCount, _init_extra__init_dragEnterCount, _init_previewVisible, _init_extra__init_previewVisible, _init_previewCurrent, _init_extra__init_previewCurrent;
const VE_UPLOAD_TAG = 've-upload';
const STATUS = {
    init: 'init',
    uploading: 'uploading',
    success: 'done',
    fail: 'error'
};
const TRUE_BY_DEFAULT_BOOLEAN_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return true;
        return 'false' !== value;
    },
    toAttribute (value) {
        return value ? null : 'false';
    }
};
const LIMIT_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return;
        const trimmed = value.trim();
        if (0 === trimmed.length) return;
        try {
            const parsed = JSON.parse(trimmed);
            if ('number' == typeof parsed && Number.isFinite(parsed)) return Math.max(0, parsed);
            if (isObject(parsed)) {
                const maxCount = Number(parsed.maxCount);
                if (!Number.isFinite(maxCount)) return;
                return {
                    maxCount: Math.max(0, maxCount),
                    ...'boolean' == typeof parsed.hideOnExceedLimit ? {
                        hideOnExceedLimit: parsed.hideOnExceedLimit
                    } : {}
                };
            }
        } catch  {}
        const parsed = Number(trimmed);
        return Number.isFinite(parsed) ? Math.max(0, parsed) : void 0;
    },
    toAttribute (value) {
        if ('number' == typeof value) return String(value);
        if (isObject(value)) try {
            return JSON.stringify(value);
        } catch  {}
        return null;
    }
};
const LIST_TYPE_VALUES = [
    'card',
    'row',
    'tag'
];
function isObject(value) {
    return 'object' == typeof value && null !== value;
}
const ACCEPT_CONVERTER = {
    fromAttribute (value) {
        if (null === value) return '';
        const trimmed = value.trim();
        if (0 === trimmed.length) return '';
        try {
            const parsed = JSON.parse(trimmed);
            if (isObject(parsed) && 'string' == typeof parsed.type && parsed.type.trim().length > 0) return {
                type: parsed.type,
                ...'boolean' == typeof parsed.strict ? {
                    strict: parsed.strict
                } : {}
            };
        } catch  {}
        return value;
    },
    toAttribute (value) {
        if ('string' == typeof value) return value;
        try {
            return JSON.stringify(value);
        } catch  {
            return null;
        }
    }
};
const UPLOAD_ICON_MAP = {
    upload: uploadIcon,
    plus: addIcon,
    file: file_fileIcon,
    'file-image': typeImageStateDefaultIcon,
    'file-pdf': typePdfStateDefaultIcon,
    'file-video': typeVideoStateDefaultIcon,
    'file-audio': typeAudioStateDefaultIcon,
    'exclamation-circle-fill': warningIcon,
    'image-close': imageIcon,
    close: closeIcon,
    check: success_successIcon,
    delete: trash03Icon,
    eye: previewIcon,
    'play-arrow-fill': playSquareIcon,
    pause: pauseIcon
};
function renderUploadIcon(definition, size) {
    const iconSize = void 0 === size ? void 0 : String(size);
    return html`
    <ve-icon
      class="ve-upload-icon"
      .definition=${definition}
      size=${ifDefined(iconSize)}
    ></ve-icon>
  `;
}
function renderMappedUploadIcon(name, size) {
    return renderUploadIcon(UPLOAD_ICON_MAP[name], size);
}
function renderUploadRemoveIcon() {
    return html`
    <svg
      class="ve-upload-remove-svg"
      viewBox="0 0 24 24"
      fill="none"
      aria-hidden="true"
      focusable="false"
    >
      <path
        d="M18 6 6 18M6 6l12 12"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
      ></path>
    </svg>
  `;
}
const isFunction = (value)=>'function' == typeof value;
const isFile = (value)=>value instanceof File;
const parseXhrResponse = (xhr)=>{
    const text = xhr.responseText || xhr.response;
    if (!text) return text;
    if ('string' != typeof text) return text;
    try {
        return JSON.parse(text);
    } catch  {
        return text;
    }
};
const appendFormDataValue = (formData, key, value)=>{
    if (null == value) return;
    if (value instanceof Blob) return void formData.append(key, value);
    formData.append(key, String(value));
};
const createDefaultUploadRequest = (options)=>{
    const { onProgress, onSuccess, onError, action, method, headers, name, file, data, withCredentials } = options;
    const getValue = (value)=>{
        if (isFunction(value)) return value(file);
        return value;
    };
    const requestFieldName = getValue(name ?? 'file') || 'file';
    const requestData = getValue(data ?? {});
    const xhr = new XMLHttpRequest();
    if (xhr.upload) xhr.upload.onprogress = (event)=>{
        let percent = 0;
        if (event.total > 0) percent = Math.floor(event.loaded / event.total * 100);
        onProgress(percent, event);
    };
    xhr.onerror = (event)=>{
        onError(event);
    };
    xhr.onload = ()=>{
        if (xhr.status < 200 || xhr.status >= 300) return void onError(parseXhrResponse(xhr));
        onSuccess(parseXhrResponse(xhr));
    };
    const formData = new FormData();
    if (isObject(requestData)) for (const [key, value] of Object.entries(requestData))appendFormDataValue(formData, key, value);
    formData.append(requestFieldName, file);
    xhr.open((method || 'post').toUpperCase(), action || '', true);
    if (withCredentials && 'withCredentials' in xhr) xhr.withCredentials = true;
    if (headers) {
        for (const [headerName, headerValue] of Object.entries(headers))if (null != headerValue) xhr.setRequestHeader(headerName, String(headerValue));
    }
    xhr.send(formData);
    return {
        abort () {
            xhr.abort();
        }
    };
};
_dec = property({
    type: String
}), _dec1 = property({
    type: String,
    reflect: true,
    useDefault: true
}), _dec2 = property({
    attribute: 'name'
}), _dec3 = property({
    attribute: 'accept',
    converter: ACCEPT_CONVERTER
}), _dec4 = property({
    type: String,
    attribute: 'list-type',
    reflect: true,
    useDefault: true
}), _dec5 = property({
    type: Boolean,
    attribute: 'image-preview',
    reflect: true
}), _dec6 = property({
    type: Boolean,
    reflect: true
}), _dec7 = property({
    type: Boolean,
    reflect: true
}), _dec8 = property({
    type: Boolean,
    reflect: true
}), _dec9 = property({
    type: Boolean,
    reflect: true
}), _dec10 = property({
    type: Boolean,
    attribute: 'with-credentials',
    reflect: true
}), _dec11 = property({
    attribute: 'auto-upload',
    reflect: true,
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec12 = property({
    attribute: 'show-upload-list',
    converter: TRUE_BY_DEFAULT_BOOLEAN_CONVERTER,
    useDefault: true
}), _dec13 = property({
    attribute: 'limit',
    converter: LIMIT_CONVERTER
}), _dec14 = property({
    type: String
}), _dec15 = property({
    attribute: false
}), _dec16 = property({
    attribute: false
}), _dec17 = property({
    attribute: false
}), _dec18 = property({
    attribute: false
}), _dec19 = property({
    attribute: false
}), _dec20 = property({
    attribute: false
}), _dec21 = property({
    attribute: false
}), _dec22 = property({
    attribute: false
}), _dec23 = property({
    attribute: false
}), _dec24 = query('.ve-upload-trigger-input'), _dec25 = state(), _dec26 = state(), _dec27 = state(), _dec28 = state(), _dec29 = state(), _dec30 = state(), _dec31 = state();
class VeUploadElement extends VeElement {
    static{
        ({ e: [_init_action, _init_extra__init_action, _init_method, _init_extra__init_method, _init_name, _init_extra__init_name, _init_accept, _init_extra__init_accept, _init_listType, _init_extra__init_listType, _init_imagePreview, _init_extra__init_imagePreview, _init_multiple, _init_extra__init_multiple, _init_disabled, _init_extra__init_disabled, _init_drag, _init_extra__init_drag, _init_directory, _init_extra__init_directory, _init_withCredentials, _init_extra__init_withCredentials, _init_autoUpload, _init_extra__init_autoUpload, _init_showUploadList, _init_extra__init_showUploadList, _init_limit, _init_extra__init_limit, _init_tip, _init_extra__init_tip, _init_headers, _init_extra__init_headers, _init_data, _init_extra__init_data, _init_customRequest, _init_extra__init_customRequest, _init_beforeUpload, _init_extra__init_beforeUpload, _init_progressProps, _init_extra__init_progressProps, _init_renderUploadItem, _init_extra__init_renderUploadItem, _init_renderUploadList, _init_extra__init_renderUploadList, _init_defaultFileList, _init_extra__init_defaultFileList, _init_fileList, _init_extra__init_fileList, _init_inputElement, _init_extra__init_inputElement, _init_innerFileList, _init_extra__init_innerFileList, _init_hasTriggerSlot, _init_extra__init_hasTriggerSlot, _init_hasTipSlot, _init_extra__init_hasTipSlot, _init_dragActive, _init_extra__init_dragActive, _init_dragEnterCount, _init_extra__init_dragEnterCount, _init_previewVisible, _init_extra__init_previewVisible, _init_previewCurrent, _init_extra__init_previewCurrent] } = _apply_decs_2311(this, [], [
            [
                _dec,
                1,
                "action"
            ],
            [
                _dec1,
                1,
                "method"
            ],
            [
                _dec2,
                1,
                "name"
            ],
            [
                _dec3,
                1,
                "accept"
            ],
            [
                _dec4,
                1,
                "listType"
            ],
            [
                _dec5,
                1,
                "imagePreview"
            ],
            [
                _dec6,
                1,
                "multiple"
            ],
            [
                _dec7,
                1,
                "disabled"
            ],
            [
                _dec8,
                1,
                "drag"
            ],
            [
                _dec9,
                1,
                "directory"
            ],
            [
                _dec10,
                1,
                "withCredentials"
            ],
            [
                _dec11,
                1,
                "autoUpload"
            ],
            [
                _dec12,
                1,
                "showUploadList"
            ],
            [
                _dec13,
                1,
                "limit"
            ],
            [
                _dec14,
                1,
                "tip"
            ],
            [
                _dec15,
                1,
                "headers"
            ],
            [
                _dec16,
                1,
                "data"
            ],
            [
                _dec17,
                1,
                "customRequest"
            ],
            [
                _dec18,
                1,
                "beforeUpload"
            ],
            [
                _dec19,
                1,
                "progressProps"
            ],
            [
                _dec20,
                1,
                "renderUploadItem"
            ],
            [
                _dec21,
                1,
                "renderUploadList"
            ],
            [
                _dec22,
                1,
                "defaultFileList"
            ],
            [
                _dec23,
                1,
                "fileList"
            ],
            [
                _dec24,
                1,
                "inputElement"
            ],
            [
                _dec25,
                1,
                "innerFileList"
            ],
            [
                _dec26,
                1,
                "hasTriggerSlot"
            ],
            [
                _dec27,
                1,
                "hasTipSlot"
            ],
            [
                _dec28,
                1,
                "dragActive"
            ],
            [
                _dec29,
                1,
                "dragEnterCount"
            ],
            [
                _dec30,
                1,
                "previewVisible"
            ],
            [
                _dec31,
                1,
                "previewCurrent"
            ]
        ]));
    }
    static styles = ve_upload;
    static dependencies = {
        [VE_BUTTON_TAG]: VeButtonElement,
        [VE_ICON_TAG]: VeIconElement
    };
    #___private_action_1 = (()=>{
        const _value = _init_action(this, '');
        _init_extra__init_action(this);
        return _value;
    })();
    get action() {
        return this.#___private_action_1;
    }
    set action(_v) {
        this.#___private_action_1 = _v;
    }
    #___private_method_2 = (()=>{
        const _value = _init_method(this, 'post');
        _init_extra__init_method(this);
        return _value;
    })();
    get method() {
        return this.#___private_method_2;
    }
    set method(_v) {
        this.#___private_method_2 = _v;
    }
    #___private_name_3 = (()=>{
        const _value = _init_name(this, 'file');
        _init_extra__init_name(this);
        return _value;
    })();
    get name() {
        return this.#___private_name_3;
    }
    set name(_v) {
        this.#___private_name_3 = _v;
    }
    #___private_accept_4 = (()=>{
        const _value = _init_accept(this, '');
        _init_extra__init_accept(this);
        return _value;
    })();
    get accept() {
        return this.#___private_accept_4;
    }
    set accept(_v) {
        this.#___private_accept_4 = _v;
    }
    #___private_listType_5 = (()=>{
        const _value = _init_listType(this, 'row');
        _init_extra__init_listType(this);
        return _value;
    })();
    get listType() {
        return this.#___private_listType_5;
    }
    set listType(_v) {
        this.#___private_listType_5 = _v;
    }
    #___private_imagePreview_6 = (()=>{
        const _value = _init_imagePreview(this, false);
        _init_extra__init_imagePreview(this);
        return _value;
    })();
    get imagePreview() {
        return this.#___private_imagePreview_6;
    }
    set imagePreview(_v) {
        this.#___private_imagePreview_6 = _v;
    }
    #___private_multiple_7 = (()=>{
        const _value = _init_multiple(this, false);
        _init_extra__init_multiple(this);
        return _value;
    })();
    get multiple() {
        return this.#___private_multiple_7;
    }
    set multiple(_v) {
        this.#___private_multiple_7 = _v;
    }
    #___private_disabled_8 = (()=>{
        const _value = _init_disabled(this, false);
        _init_extra__init_disabled(this);
        return _value;
    })();
    get disabled() {
        return this.#___private_disabled_8;
    }
    set disabled(_v) {
        this.#___private_disabled_8 = _v;
    }
    #___private_drag_9 = (()=>{
        const _value = _init_drag(this, false);
        _init_extra__init_drag(this);
        return _value;
    })();
    get drag() {
        return this.#___private_drag_9;
    }
    set drag(_v) {
        this.#___private_drag_9 = _v;
    }
    #___private_directory_10 = (()=>{
        const _value = _init_directory(this, false);
        _init_extra__init_directory(this);
        return _value;
    })();
    get directory() {
        return this.#___private_directory_10;
    }
    set directory(_v) {
        this.#___private_directory_10 = _v;
    }
    #___private_withCredentials_11 = (()=>{
        const _value = _init_withCredentials(this, false);
        _init_extra__init_withCredentials(this);
        return _value;
    })();
    get withCredentials() {
        return this.#___private_withCredentials_11;
    }
    set withCredentials(_v) {
        this.#___private_withCredentials_11 = _v;
    }
    #___private_autoUpload_12 = (()=>{
        const _value = _init_autoUpload(this, true);
        _init_extra__init_autoUpload(this);
        return _value;
    })();
    get autoUpload() {
        return this.#___private_autoUpload_12;
    }
    set autoUpload(_v) {
        this.#___private_autoUpload_12 = _v;
    }
    #___private_showUploadList_13 = (()=>{
        const _value = _init_showUploadList(this, true);
        _init_extra__init_showUploadList(this);
        return _value;
    })();
    get showUploadList() {
        return this.#___private_showUploadList_13;
    }
    set showUploadList(_v) {
        this.#___private_showUploadList_13 = _v;
    }
    #___private_limit_14 = (()=>{
        const _value = _init_limit(this);
        _init_extra__init_limit(this);
        return _value;
    })();
    get limit() {
        return this.#___private_limit_14;
    }
    set limit(_v) {
        this.#___private_limit_14 = _v;
    }
    #___private_tip_15 = (()=>{
        const _value = _init_tip(this, '');
        _init_extra__init_tip(this);
        return _value;
    })();
    get tip() {
        return this.#___private_tip_15;
    }
    set tip(_v) {
        this.#___private_tip_15 = _v;
    }
    #___private_headers_16 = (()=>{
        const _value = _init_headers(this, {});
        _init_extra__init_headers(this);
        return _value;
    })();
    get headers() {
        return this.#___private_headers_16;
    }
    set headers(_v) {
        this.#___private_headers_16 = _v;
    }
    #___private_data_17 = (()=>{
        const _value = _init_data(this, {});
        _init_extra__init_data(this);
        return _value;
    })();
    get data() {
        return this.#___private_data_17;
    }
    set data(_v) {
        this.#___private_data_17 = _v;
    }
    #___private_customRequest_18 = (()=>{
        const _value = _init_customRequest(this);
        _init_extra__init_customRequest(this);
        return _value;
    })();
    get customRequest() {
        return this.#___private_customRequest_18;
    }
    set customRequest(_v) {
        this.#___private_customRequest_18 = _v;
    }
    #___private_beforeUpload_19 = (()=>{
        const _value = _init_beforeUpload(this, ()=>true);
        _init_extra__init_beforeUpload(this);
        return _value;
    })();
    get beforeUpload() {
        return this.#___private_beforeUpload_19;
    }
    set beforeUpload(_v) {
        this.#___private_beforeUpload_19 = _v;
    }
    #___private_progressProps_20 = (()=>{
        const _value = _init_progressProps(this);
        _init_extra__init_progressProps(this);
        return _value;
    })();
    get progressProps() {
        return this.#___private_progressProps_20;
    }
    set progressProps(_v) {
        this.#___private_progressProps_20 = _v;
    }
    #___private_renderUploadItem_21 = (()=>{
        const _value = _init_renderUploadItem(this);
        _init_extra__init_renderUploadItem(this);
        return _value;
    })();
    get renderUploadItem() {
        return this.#___private_renderUploadItem_21;
    }
    set renderUploadItem(_v) {
        this.#___private_renderUploadItem_21 = _v;
    }
    #___private_renderUploadList_22 = (()=>{
        const _value = _init_renderUploadList(this);
        _init_extra__init_renderUploadList(this);
        return _value;
    })();
    get renderUploadList() {
        return this.#___private_renderUploadList_22;
    }
    set renderUploadList(_v) {
        this.#___private_renderUploadList_22 = _v;
    }
    #___private_defaultFileList_23 = (()=>{
        const _value = _init_defaultFileList(this, []);
        _init_extra__init_defaultFileList(this);
        return _value;
    })();
    get defaultFileList() {
        return this.#___private_defaultFileList_23;
    }
    set defaultFileList(_v) {
        this.#___private_defaultFileList_23 = _v;
    }
    #___private_fileList_24 = (()=>{
        const _value = _init_fileList(this);
        _init_extra__init_fileList(this);
        return _value;
    })();
    get fileList() {
        return this.#___private_fileList_24;
    }
    set fileList(_v) {
        this.#___private_fileList_24 = _v;
    }
    #___private_inputElement_25 = (()=>{
        const _value = _init_inputElement(this, null);
        _init_extra__init_inputElement(this);
        return _value;
    })();
    get inputElement() {
        return this.#___private_inputElement_25;
    }
    set inputElement(_v) {
        this.#___private_inputElement_25 = _v;
    }
    #___private_innerFileList_26 = (()=>{
        const _value = _init_innerFileList(this, []);
        _init_extra__init_innerFileList(this);
        return _value;
    })();
    get innerFileList() {
        return this.#___private_innerFileList_26;
    }
    set innerFileList(_v) {
        this.#___private_innerFileList_26 = _v;
    }
    #___private_hasTriggerSlot_27 = (()=>{
        const _value = _init_hasTriggerSlot(this, false);
        _init_extra__init_hasTriggerSlot(this);
        return _value;
    })();
    get hasTriggerSlot() {
        return this.#___private_hasTriggerSlot_27;
    }
    set hasTriggerSlot(_v) {
        this.#___private_hasTriggerSlot_27 = _v;
    }
    #___private_hasTipSlot_28 = (()=>{
        const _value = _init_hasTipSlot(this, false);
        _init_extra__init_hasTipSlot(this);
        return _value;
    })();
    get hasTipSlot() {
        return this.#___private_hasTipSlot_28;
    }
    set hasTipSlot(_v) {
        this.#___private_hasTipSlot_28 = _v;
    }
    #___private_dragActive_29 = (()=>{
        const _value = _init_dragActive(this, false);
        _init_extra__init_dragActive(this);
        return _value;
    })();
    get dragActive() {
        return this.#___private_dragActive_29;
    }
    set dragActive(_v) {
        this.#___private_dragActive_29 = _v;
    }
    #___private_dragEnterCount_30 = (()=>{
        const _value = _init_dragEnterCount(this, 0);
        _init_extra__init_dragEnterCount(this);
        return _value;
    })();
    get dragEnterCount() {
        return this.#___private_dragEnterCount_30;
    }
    set dragEnterCount(_v) {
        this.#___private_dragEnterCount_30 = _v;
    }
    #___private_previewVisible_31 = (()=>{
        const _value = _init_previewVisible(this, false);
        _init_extra__init_previewVisible(this);
        return _value;
    })();
    get previewVisible() {
        return this.#___private_previewVisible_31;
    }
    set previewVisible(_v) {
        this.#___private_previewVisible_31 = _v;
    }
    #___private_previewCurrent_32 = (()=>{
        const _value = _init_previewCurrent(this, 0);
        _init_extra__init_previewCurrent(this);
        return _value;
    })();
    get previewCurrent() {
        return this.#___private_previewCurrent_32;
    }
    set previewCurrent(_v) {
        this.#___private_previewCurrent_32 = _v;
    }
    uidSeed = 0;
    objectUrlStore = new Map();
    fileUidStore = new WeakMap();
    uploadRequests = new Map();
    initializedDefaultFileList = false;
    choose() {
        if (this.getDerivedDisabledState()) return;
        this.inputElement?.click();
    }
    submit(file) {
        if (this.getDerivedDisabledState()) return;
        if (file) return void this.uploadByUid(file.uid);
        const files = this.getMergedFileList().filter((item)=>item.status === STATUS.init);
        for (const currentFile of files)this.uploadByUid(currentFile.uid);
    }
    abort(file) {
        const request = this.uploadRequests.get(file.uid);
        if (!request?.abort) return;
        request.abort();
        this.uploadRequests.delete(file.uid);
        const updated = this.updateFile(file.uid, (current)=>({
                ...current,
                status: STATUS.fail
            }));
        if (updated) this.emitAbortEvent(updated);
    }
    reupload(file) {
        const updated = this.updateFile(file.uid, (current)=>({
                ...current,
                percent: 0,
                status: STATUS.uploading
            }));
        if (!updated) return;
        this.emitReuploadEvent(updated);
        this.uploadByUid(updated.uid);
    }
    clear() {
        for (const request of this.uploadRequests.values())request.abort?.();
        this.uploadRequests.clear();
        this.revokeAllObjectUrls();
        this.applyFileList([], void 0);
    }
    addFiles(files) {
        this.handleFiles(this.normalizeFiles(files), 'api');
    }
    firstUpdated(_changedProperties) {
        super.firstUpdated(_changedProperties);
        this.syncSlotState();
        this.initializeDefaultFileListIfNeeded();
    }
    updated(changedProperties) {
        super.updated(changedProperties);
        if (changedProperties.has('listType') && !LIST_TYPE_VALUES.includes(this.listType)) this.listType = 'row';
        if (changedProperties.has('fileList')) this.ensureObjectUrlsForFiles(this.getMergedFileList());
        if (changedProperties.has('defaultFileList')) this.initializeDefaultFileListIfNeeded(true);
        this.cleanupRemovedObjectUrls();
    }
    disconnectedCallback() {
        super.disconnectedCallback();
        for (const request of this.uploadRequests.values())request.abort?.();
        this.uploadRequests.clear();
        this.revokeAllObjectUrls();
    }
    createUid() {
        this.uidSeed += 1;
        return `${Date.now()}-${this.uidSeed}`;
    }
    getStableFileUid(file, index, fileOccurrenceMap) {
        const occurrence = fileOccurrenceMap?.get(file) ?? 0;
        fileOccurrenceMap?.set(file, occurrence + 1);
        const existingUid = this.fileUidStore.get(file);
        if (existingUid) return 0 === occurrence ? existingUid : `${existingUid}-${occurrence}`;
        const uid = `${Date.now()}-${index}-${this.uidSeed + 1}`;
        this.uidSeed += 1;
        this.fileUidStore.set(file, uid);
        return 0 === occurrence ? uid : `${uid}-${occurrence}`;
    }
    normalizeStatus(status) {
        if (status === STATUS.init || status === STATUS.uploading || status === STATUS.success || status === STATUS.fail) return status;
        return STATUS.success;
    }
    normalizeExternalFileList(fileList) {
        if (!Array.isArray(fileList) || 0 === fileList.length) return [];
        const uniqueByUid = new Map();
        const fileOccurrenceMap = new WeakMap();
        fileList.forEach((rawItem, index)=>{
            const uid = 'string' == typeof rawItem.uid && rawItem.uid.length > 0 ? rawItem.uid : isFile(rawItem.originFile) ? this.getStableFileUid(rawItem.originFile, index, fileOccurrenceMap) : `${Date.now()}-${index}`;
            const status = this.normalizeStatus(rawItem.status);
            const percent = rawItem.percent ?? (status === STATUS.success ? 100 : status === STATUS.uploading ? 0 : 0);
            const normalizedName = rawItem.name ?? rawItem.originFile?.name;
            const normalized = {
                ...rawItem,
                uid,
                status,
                percent,
                ...normalizedName ? {
                    name: normalizedName
                } : {}
            };
            uniqueByUid.set(uid, normalized);
            if (isFile(normalized.originFile)) this.ensureObjectUrl(normalized);
        });
        return [
            ...uniqueByUid.values()
        ];
    }
    initializeDefaultFileListIfNeeded(force = false) {
        if (this.isControlled()) {
            this.initializedDefaultFileList = true;
            return;
        }
        if (!force && this.initializedDefaultFileList) return;
        this.innerFileList = this.normalizeExternalFileList(this.defaultFileList);
        this.initializedDefaultFileList = true;
    }
    isControlled() {
        return Array.isArray(this.fileList);
    }
    getMergedFileList() {
        if (this.isControlled()) return this.normalizeExternalFileList(this.fileList);
        return this.innerFileList;
    }
    cloneFileItem(file) {
        return {
            ...file
        };
    }
    cloneFileList(fileList) {
        return fileList.map((file)=>this.cloneFileItem(file));
    }
    applyFileList(nextFileList, changedFile) {
        const normalized = this.normalizeExternalFileList(nextFileList);
        if (!this.isControlled()) this.innerFileList = normalized;
        const fileListClone = this.cloneFileList(normalized);
        if (changedFile) return void this.emitChangeEvent(fileListClone, changedFile);
        this.emitChangeEvent(fileListClone, void 0);
    }
    updateFile(uid, updater) {
        const sourceFileList = this.getMergedFileList();
        let changedFile;
        const nextFileList = sourceFileList.map((file)=>{
            if (file.uid !== uid) return file;
            const updated = updater(file);
            changedFile = updated;
            return updated;
        });
        if (!changedFile) return;
        this.applyFileList(nextFileList, changedFile);
        return changedFile;
    }
    removeFile(uid) {
        const sourceFileList = this.getMergedFileList();
        const targetFile = sourceFileList.find((file)=>file.uid === uid);
        if (!targetFile) return;
        const nextFileList = sourceFileList.filter((file)=>file.uid !== uid);
        const request = this.uploadRequests.get(uid);
        request?.abort?.();
        this.uploadRequests.delete(uid);
        this.revokeObjectUrl(uid);
        this.applyFileList(nextFileList, targetFile);
        this.emitRemoveEvent(targetFile, nextFileList);
    }
    emitSelectEvent(files) {
        this.dispatchEvent(new CustomEvent('ve-select', {
            detail: {
                files: this.cloneFileList(files)
            },
            bubbles: true,
            composed: true
        }));
    }
    emitChangeEvent(fileList, file) {
        const detail = {
            fileList: this.cloneFileList(fileList)
        };
        if (file) detail.file = this.cloneFileItem(file);
        this.dispatchEvent(new CustomEvent('ve-change', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    emitSuccessEvent(file, response) {
        this.dispatchEvent(new CustomEvent('ve-success', {
            detail: {
                file: this.cloneFileItem(file),
                response
            },
            bubbles: true,
            composed: true
        }));
    }
    emitErrorEvent(file, response) {
        this.dispatchEvent(new CustomEvent('ve-error', {
            detail: {
                file: this.cloneFileItem(file),
                response
            },
            bubbles: true,
            composed: true
        }));
    }
    emitRemoveEvent(file, fileList) {
        this.dispatchEvent(new CustomEvent('ve-remove', {
            detail: {
                file: this.cloneFileItem(file),
                fileList: this.cloneFileList(fileList)
            },
            bubbles: true,
            composed: true
        }));
    }
    emitExceedLimitEvent(files, fileList, maxCount) {
        const normalizedFiles = files.map((file, index)=>({
                uid: `${Date.now()}-exceed-${index}`,
                status: STATUS.init,
                originFile: file,
                name: file.name,
                percent: 0
            }));
        this.dispatchEvent(new CustomEvent('ve-exceed-limit', {
            detail: {
                files: this.cloneFileList(normalizedFiles),
                fileList: this.cloneFileList(fileList),
                ...'number' == typeof maxCount ? {
                    maxCount
                } : {}
            },
            bubbles: true,
            composed: true
        }));
    }
    emitDropEvent(files) {
        this.dispatchEvent(new CustomEvent('ve-drop', {
            detail: {
                files: this.cloneFileList(files)
            },
            bubbles: true,
            composed: true
        }));
    }
    emitProgressEvent(file, event) {
        const detail = {
            file: this.cloneFileItem(file)
        };
        if (event) detail.event = event;
        this.dispatchEvent(new CustomEvent('ve-progress', {
            detail,
            bubbles: true,
            composed: true
        }));
    }
    emitPreviewEvent(file) {
        this.dispatchEvent(new CustomEvent('ve-preview', {
            detail: {
                file: this.cloneFileItem(file)
            },
            bubbles: true,
            composed: true
        }));
    }
    emitItemClickEvent(file, fileList) {
        this.dispatchEvent(new CustomEvent('ve-item-click', {
            detail: {
                file: this.cloneFileItem(file),
                fileList: this.cloneFileList(fileList)
            },
            bubbles: true,
            composed: true
        }));
    }
    emitReuploadEvent(file) {
        this.dispatchEvent(new CustomEvent('ve-reupload', {
            detail: {
                file: this.cloneFileItem(file)
            },
            bubbles: true,
            composed: true
        }));
    }
    emitAbortEvent(file) {
        this.dispatchEvent(new CustomEvent('ve-abort', {
            detail: {
                file: this.cloneFileItem(file)
            },
            bubbles: true,
            composed: true
        }));
    }
    resolveLimitConfig() {
        if ('number' == typeof this.limit) return {
            maxCount: this.limit,
            hideOnExceedLimit: true
        };
        if (isObject(this.limit)) {
            const maxCount = Number(this.limit.maxCount);
            const resolvedHideOnExceedLimit = 'boolean' == typeof this.limit.hideOnExceedLimit ? this.limit.hideOnExceedLimit : true;
            if (!Number.isFinite(maxCount)) return {
                hideOnExceedLimit: resolvedHideOnExceedLimit
            };
            return {
                maxCount,
                hideOnExceedLimit: resolvedHideOnExceedLimit
            };
        }
        return {
            hideOnExceedLimit: true
        };
    }
    getDerivedDisabledState() {
        const fileList = this.getMergedFileList();
        const { maxCount, hideOnExceedLimit } = this.resolveLimitConfig();
        const exceedLimit = 'number' == typeof maxCount ? maxCount <= fileList.length : false;
        return this.disabled || !hideOnExceedLimit && exceedLimit;
    }
    shouldHideTrigger() {
        const fileList = this.getMergedFileList();
        const { maxCount, hideOnExceedLimit } = this.resolveLimitConfig();
        if (!hideOnExceedLimit || 'number' != typeof maxCount) return false;
        return maxCount <= fileList.length;
    }
    getShowUploadList() {
        return false !== this.showUploadList;
    }
    getCustomIconConfig() {
        return isObject(this.showUploadList) ? this.showUploadList : {};
    }
    resolveIcon(iconName, fallback) {
        const config = this.getCustomIconConfig();
        const icon = config[iconName];
        if (null === icon) return null;
        return icon ?? fallback;
    }
    resolveDefaultFileIconName(file) {
        const fileType = (file.type || file.originFile?.type || '').toLowerCase();
        const fileName = this.resolveDefaultFileName(file).toLowerCase();
        const extension = fileName.includes('.') ? fileName.split('.').pop() : '';
        if (fileType.includes('pdf') || 'pdf' === extension) return typePdfStateDefaultIcon;
        return typeWordStateDefaultIcon;
    }
    resolveDefaultFileIcon(file, size) {
        return renderUploadIcon(this.resolveDefaultFileIconName(file), size);
    }
    resolveDefaultFileName(file) {
        return file.name ?? file.originFile?.name ?? '';
    }
    resolveFileName(file) {
        const config = this.getCustomIconConfig();
        if (isFunction(config.fileName)) return config.fileName(file);
        return this.resolveDefaultFileName(file);
    }
    hasCustomFileNameRenderer() {
        return isFunction(this.getCustomIconConfig().fileName);
    }
    resolveAcceptValue() {
        if ('string' == typeof this.accept) return this.accept;
        if (isObject(this.accept) && 'string' == typeof this.accept.type) return this.accept.type;
        return '';
    }
    normalizeFiles(files) {
        return Array.from(files).filter((file)=>file instanceof File);
    }
    isAcceptFile(file) {
        const accept = this.accept;
        if (!accept) return true;
        const acceptText = 'string' == typeof accept ? accept : accept.type;
        const strict = !(isObject(accept) && false === accept.strict);
        if (!strict) return true;
        if (!acceptText || 0 === acceptText.trim().length) return true;
        const accepts = acceptText.split(',').map((item)=>item.trim().toLowerCase()).filter((item)=>item.length > 0);
        const fileType = (file.type || '').toLowerCase();
        const baseFileType = fileType.split('/')[0];
        const extension = (file.name.includes('.') ? `.${file.name.split('.').pop()}` : '').toLowerCase();
        if (0 === accepts.length) return true;
        return accepts.some((type)=>{
            if (type === fileType || `${baseFileType}${extension.replace('.', '/')}` === type) return true;
            if (/^\*(\/\*)?$/.test(type)) return true;
            if (/\/\*/.test(type)) return fileType.replace(/\/.*$/, '') === type.replace(/\/.*$/, '');
            if (/\..*/.test(type)) {
                if ('.jpg' === type || '.jpeg' === type) return '.jpg' === extension || '.jpeg' === extension;
                return extension === type;
            }
            return false;
        });
    }
    resolveProgressType(listType) {
        if ('card' === listType) return 'circle';
        return this.progressProps?.type === 'line' ? 'line' : 'circle';
    }
    shouldShowProgressText(listType) {
        if ('card' === listType) return false;
        return this.progressProps?.showText === true;
    }
    resolveProgressStyle(percent) {
        const width = this.progressProps?.width;
        const widthStyle = void 0 === width ? '' : `inline-size:${'number' == typeof width ? `${width}px` : width};`;
        return `--ve-upload-progress-percent:${percent};${widthStyle}`;
    }
    getProgressText(file, percent) {
        const formatText = this.progressProps?.formatText;
        if (isFunction(formatText)) return formatText(percent, this.cloneFileItem(file));
        const status = file.status ?? STATUS.init;
        if (status === STATUS.uploading) return `${percent}%`;
        if (status === STATUS.init) return 'Ready';
        if (status === STATUS.success) return 'Done';
        return 'Error';
    }
    formatProgressDisplay(file, originNode) {
        const config = this.getCustomIconConfig();
        if (isFunction(config.progressRender)) return config.progressRender(this.cloneFileItem(file), originNode);
        return originNode;
    }
    isRtl() {
        if ("u" < typeof window) return false;
        return 'rtl' === window.getComputedStyle(this).direction;
    }
    ensureObjectUrl(file) {
        if (file.url || !file.originFile || this.objectUrlStore.has(file.uid)) return;
        if (!isFunction(URL.createObjectURL)) return;
        const objectUrl = URL.createObjectURL(file.originFile);
        this.objectUrlStore.set(file.uid, objectUrl);
    }
    ensureObjectUrlsForFiles(fileList) {
        fileList.forEach((file)=>{
            this.ensureObjectUrl(file);
        });
    }
    revokeObjectUrl(uid) {
        const targetUrl = this.objectUrlStore.get(uid);
        if (!targetUrl) return;
        URL.revokeObjectURL(targetUrl);
        this.objectUrlStore.delete(uid);
    }
    revokeAllObjectUrls() {
        for (const url of this.objectUrlStore.values())URL.revokeObjectURL(url);
        this.objectUrlStore.clear();
    }
    cleanupRemovedObjectUrls() {
        const currentUidSet = new Set(this.getMergedFileList().map((file)=>file.uid));
        for (const [uid, url] of this.objectUrlStore.entries())if (!currentUidSet.has(uid)) {
            URL.revokeObjectURL(url);
            this.objectUrlStore.delete(uid);
        }
    }
    getFileUrl(file) {
        if ('string' == typeof file.url && file.url.length > 0) return file.url;
        this.ensureObjectUrl(file);
        return this.objectUrlStore.get(file.uid);
    }
    createUploadItem(file) {
        const uid = this.createUid();
        const uploadItem = {
            uid,
            originFile: file,
            percent: 0,
            status: STATUS.init,
            name: file.name,
            size: file.size,
            ...file.type ? {
                type: file.type
            } : {}
        };
        this.ensureObjectUrl(uploadItem);
        return uploadItem;
    }
    async runBeforeUpload(file, files) {
        if (!isFunction(this.beforeUpload)) return file;
        try {
            const result = await this.beforeUpload(file, files);
            if (false === result) return null;
            if (isFile(result)) return result;
            return file;
        } catch (error) {
            console.error(error);
            return null;
        }
    }
    async handleFiles(files, source) {
        if (0 === files.length || this.getDerivedDisabledState()) return;
        const sourceFileList = this.getMergedFileList();
        const { maxCount } = this.resolveLimitConfig();
        if ('number' == typeof maxCount && maxCount < sourceFileList.length + files.length) return void this.emitExceedLimitEvent(files, sourceFileList, maxCount);
        const acceptedFiles = files.filter((file)=>this.isAcceptFile(file));
        const beforeUploadPassedFiles = [];
        for (const acceptedFile of acceptedFiles){
            const checkedFile = await this.runBeforeUpload(acceptedFile, acceptedFiles);
            if (checkedFile) beforeUploadPassedFiles.push(checkedFile);
        }
        if (0 === beforeUploadPassedFiles.length) return;
        const newItems = beforeUploadPassedFiles.map((file)=>this.createUploadItem(file));
        const nextFileList = [
            ...sourceFileList,
            ...newItems
        ];
        this.applyFileList(nextFileList, newItems[newItems.length - 1]);
        this.emitSelectEvent(newItems);
        if ('drop' === source) this.emitDropEvent(newItems);
        if (!this.autoUpload) return;
        for (const file of newItems)this.uploadByUid(file.uid);
    }
    async uploadByUid(uid) {
        const targetFile = this.getMergedFileList().find((file)=>file.uid === uid);
        if (!targetFile || !targetFile.originFile) return;
        await this.doUpload(targetFile);
    }
    async doUpload(file) {
        const rawFile = file.originFile;
        if (!rawFile) return;
        const uploadCandidate = this.updateFile(file.uid, (current)=>({
                ...current,
                status: STATUS.uploading,
                percent: 0
            }));
        if (!uploadCandidate) return;
        let completed = false;
        const onProgress = (percent, event)=>{
            const progressed = this.updateFile(file.uid, (current)=>({
                    ...current,
                    status: STATUS.uploading,
                    percent
                }));
            if (!progressed) return;
            this.emitProgressEvent(progressed, event);
        };
        const onSuccess = (response)=>{
            completed = true;
            const succeeded = this.updateFile(file.uid, (current)=>({
                    ...current,
                    status: STATUS.success,
                    percent: 100,
                    response
                }));
            if (!succeeded) return;
            this.uploadRequests.delete(file.uid);
            this.emitSuccessEvent(succeeded, response);
        };
        const onError = (response)=>{
            completed = true;
            const failed = this.updateFile(file.uid, (current)=>({
                    ...current,
                    status: STATUS.fail,
                    response
                }));
            if (!failed) return;
            this.uploadRequests.delete(file.uid);
            this.emitErrorEvent(failed, response);
        };
        const options = {
            onProgress,
            onSuccess,
            onError,
            headers: this.headers,
            name: this.name,
            file: rawFile,
            data: this.data,
            withCredentials: this.withCredentials,
            action: this.action,
            method: this.method
        };
        if (!this.action && !this.customRequest) return void onError(new Error('`action` or `customRequest` is required to upload.'));
        let request;
        try {
            if (this.action) request = createDefaultUploadRequest(options);
            else if (this.customRequest) request = await this.customRequest(options);
        } catch (error) {
            if (!completed) onError(error);
            return;
        }
        if (completed) return;
        const currentFile = this.getMergedFileList().find((item)=>item.uid === file.uid);
        if (!this.isConnected || !currentFile || currentFile.status !== STATUS.uploading) return void request?.abort?.();
        this.uploadRequests.set(file.uid, request ?? {});
    }
    handleInputChange(event) {
        const target = event.target;
        if (!(target instanceof HTMLInputElement)) return;
        const files = target.files ? this.normalizeFiles(target.files) : [];
        target.value = '';
        this.handleFiles(files, 'select');
    }
    handleTriggerClick(event) {
        if (this.getDerivedDisabledState()) {
            event.preventDefault();
            event.stopPropagation();
            return;
        }
        this.choose();
    }
    handleKeyboardActivate(event, callback) {
        if ('Enter' !== event.key && ' ' !== event.key) return;
        event.preventDefault();
        callback();
    }
    stopListItemActionEvent(event) {
        event.stopPropagation();
    }
    handleTriggerKeyDown(event) {
        if (this.getDerivedDisabledState()) return;
        this.handleKeyboardActivate(event, ()=>{
            this.choose();
        });
    }
    handleTriggerSlotChange() {
        this.syncSlotState();
    }
    handleTipSlotChange() {
        this.syncSlotState();
    }
    syncSlotState() {
        this.hasTriggerSlot = null !== this.querySelector('[slot="trigger"]');
        this.hasTipSlot = null !== this.querySelector('[slot="tip"]');
    }
    handleRemoveAction(file) {
        this.removeFile(file.uid);
    }
    handleItemClick(file) {
        this.emitItemClickEvent(file, this.getMergedFileList());
    }
    handleItemKeyDown(event, file) {
        this.handleKeyboardActivate(event, ()=>{
            this.handleItemClick(file);
        });
    }
    handlePreviewAction(file) {
        this.emitPreviewEvent(file);
        if (!this.imagePreview || 'card' !== this.listType) return;
        const previewFileList = this.getPreviewFileList(this.getMergedFileList());
        if (0 === previewFileList.length) return;
        const index = previewFileList.findIndex((item)=>item.uid === file.uid);
        this.previewCurrent = index >= 0 ? index : 0;
        this.previewVisible = true;
    }
    getPreviewFileList(fileList) {
        return fileList.filter((file)=>{
            const url = this.getFileUrl(file);
            return 'string' == typeof url && url.length > 0;
        });
    }
    closePreview() {
        this.previewVisible = false;
        this.previewCurrent = 0;
    }
    showPreviewByOffset(offset) {
        const previewFileList = this.getPreviewFileList(this.getMergedFileList());
        if (0 === previewFileList.length) return void this.closePreview();
        const nextIndex = (this.previewCurrent + offset + previewFileList.length) % previewFileList.length;
        this.previewCurrent = nextIndex;
    }
    handleDropDragEnter(event) {
        if (!this.drag || this.getDerivedDisabledState()) return;
        event.preventDefault();
        this.dragEnterCount += 1;
    }
    handleDropDragLeave(event) {
        if (!this.drag || this.getDerivedDisabledState()) return;
        event.preventDefault();
        this.dragEnterCount = Math.max(this.dragEnterCount - 1, 0);
        if (0 === this.dragEnterCount) this.dragActive = false;
    }
    handleDropDragOver(event) {
        if (!this.drag || this.getDerivedDisabledState()) return;
        event.preventDefault();
        if (event.dataTransfer) event.dataTransfer.dropEffect = 'copy';
        if (!this.dragActive) this.dragActive = true;
    }
    getDropEntries(event) {
        return Array.from(event.dataTransfer?.items ?? []).map((item)=>item.webkitGetAsEntry?.());
    }
    async readDirectoryEntry(entry, files) {
        if (entry.isFile) {
            const fileEntry = entry;
            const file = await new Promise((resolve)=>{
                fileEntry.file((targetFile)=>resolve(targetFile), ()=>resolve(null));
            });
            if (!file || !this.isAcceptFile(file)) return;
            try {
                Object.defineProperty(file, 'webkitRelativePath', {
                    value: entry.fullPath.replace(/^\//, ''),
                    configurable: true
                });
            } catch  {}
            files.push(file);
            return;
        }
        if (!entry.isDirectory) return;
        const directoryEntry = entry;
        const reader = directoryEntry.createReader();
        while(true){
            const readEntries = await new Promise((resolve)=>{
                reader.readEntries((entries)=>resolve(entries), ()=>resolve([]));
            });
            if (0 === readEntries.length) break;
            await Promise.all(readEntries.map((currentEntry)=>this.readDirectoryEntry(currentEntry, files)));
        }
    }
    async getDroppedFiles(event) {
        if (!this.directory) {
            const entries = this.getDropEntries(event);
            const directoryIndices = entries.reduce((result, entry, index)=>{
                if (entry?.isDirectory) return [
                    ...result,
                    index
                ];
                return result;
            }, []);
            const droppedFiles = Array.from(event.dataTransfer?.files ?? []).filter((_, index)=>!directoryIndices.includes(index));
            return droppedFiles.filter((file)=>this.isAcceptFile(file));
        }
        const entries = this.getDropEntries(event).filter((entry)=>Boolean(entry));
        const files = [];
        await Promise.all(entries.map((entry)=>this.readDirectoryEntry(entry, files)));
        return this.multiple ? files : files.slice(0, 1);
    }
    handleDrop = async (event)=>{
        if (!this.drag || this.getDerivedDisabledState()) return;
        event.preventDefault();
        this.dragEnterCount = 0;
        this.dragActive = false;
        const droppedFiles = await this.getDroppedFiles(event);
        if (droppedFiles.length > 0) await this.handleFiles(droppedFiles, 'drop');
    };
    renderUploadProgress(file, listType) {
        const status = file.status ?? STATUS.init;
        let originNode = nothing;
        const reuploadIcon = this.resolveIcon('reuploadIcon', 'card' === listType ? renderMappedUploadIcon('upload') : 'Reupload');
        const reuploadNode = status !== STATUS.fail || null === reuploadIcon ? nothing : html`
            <span
              class="ve-upload-list-reupload-icon"
              role="button"
              tabindex="0"
              aria-label="reupload"
              @click=${(event)=>{
            this.stopListItemActionEvent(event);
            this.reupload(file);
        }}
              @keydown=${(event)=>{
            this.stopListItemActionEvent(event);
            this.handleKeyboardActivate(event, ()=>{
                this.reupload(file);
            });
        }}
            >
              ${reuploadIcon}
            </span>
          `;
        if (status === STATUS.success) {
            const successIcon = this.resolveIcon('successIcon', renderMappedUploadIcon('check'));
            if (null !== successIcon) originNode = html`
          <span class="ve-upload-list-success-icon"> ${successIcon} </span>
        `;
            return this.formatProgressDisplay(file, originNode);
        }
        const currentPercent = Math.max(0, Math.min(100, Math.round(file.percent ?? 0)));
        const progressText = this.getProgressText({
            ...file,
            percent: currentPercent
        }, currentPercent);
        const progressType = this.resolveProgressType(listType);
        const showProgressText = this.shouldShowProgressText(listType);
        const startIcon = this.resolveIcon('startIcon', renderMappedUploadIcon('play-arrow-fill'));
        const cancelIcon = this.resolveIcon('cancelIcon', renderMappedUploadIcon('pause'));
        originNode = html`
      ${reuploadNode}
      <span
        class=${classMap({
            've-upload-list-status': true,
            've-upload-list-status-line': 'line' === progressType
        })}
        style=${this.resolveProgressStyle(currentPercent)}
      >
        ${'line' === progressType ? html`
              <span class="ve-upload-list-progress">
                <span class="ve-upload-list-progress-line">
                  <span class="ve-upload-list-progress-line-fill"></span>
                </span>
                ${showProgressText ? html`<span class="ve-upload-list-progress-text"
                      >${progressText}</span
                    >` : nothing}
              </span>
            ` : html`
              <span class="ve-upload-list-status-ring" aria-hidden="true">
                <svg class="ve-upload-list-status-ring-svg" viewBox="0 0 20 20">
                  <circle
                    class="ve-upload-list-status-ring-track"
                    cx="10"
                    cy="10"
                    r="8"
                  ></circle>
                  <circle
                    class="ve-upload-list-status-ring-path"
                    cx="10"
                    cy="10"
                    r="8"
                    pathLength="100"
                  ></circle>
                </svg>
              </span>
              <span class="ve-upload-list-status-label">${progressText}</span>
            `}
        ${status === STATUS.init ? null === startIcon ? nothing : html`
                <span
                  class="ve-upload-list-start-icon"
                  role="button"
                  tabindex="0"
                  aria-label="start"
                  @click=${(event)=>{
            this.stopListItemActionEvent(event);
            this.submit(file);
        }}
                  @keydown=${(event)=>{
            this.stopListItemActionEvent(event);
            this.handleKeyboardActivate(event, ()=>{
                this.submit(file);
            });
        }}
                >
                  ${startIcon}
                </span>
              ` : status === STATUS.uploading ? null === cancelIcon ? nothing : html`
                  <span
                    class="ve-upload-list-cancel-icon"
                    role="button"
                    tabindex="0"
                    aria-label="cancel"
                    @click=${(event)=>{
            this.stopListItemActionEvent(event);
            this.abort(file);
        }}
                    @keydown=${(event)=>{
            this.stopListItemActionEvent(event);
            this.handleKeyboardActivate(event, ()=>{
                this.abort(file);
            });
        }}
                  >
                    ${cancelIcon}
                  </span>
                ` : nothing}
      </span>
    `;
        return this.formatProgressDisplay(file, originNode);
    }
    resolveFileExtension(file) {
        const fileName = this.resolveDefaultFileName(file);
        const extension = fileName.includes('.') ? fileName.split('.').pop() : '';
        if (extension && extension.length > 0) return extension.toUpperCase();
        const fileType = (file.type || file.originFile?.type || '').trim();
        const mimeSubtype = fileType.includes('/') ? fileType.split('/').pop() : '';
        return mimeSubtype && mimeSubtype.length > 0 ? mimeSubtype.toUpperCase() : 'FILE';
    }
    resolveAttachmentKind(file) {
        const fileType = (file.type || file.originFile?.type || '').toLowerCase();
        const extension = this.resolveFileExtension(file).toLowerCase();
        if (fileType.includes('image') || [
            'png',
            'jpg',
            'jpeg',
            'bmp',
            'gif',
            'webp'
        ].includes(extension)) return 'image';
        if (fileType.includes('video') || [
            'mp4',
            'mov',
            'm2v',
            'mkv',
            'webm'
        ].includes(extension)) return 'video';
        return 'file';
    }
    resolveFileSizeText(file) {
        const size = file.size ?? file.originFile?.size;
        if ('number' != typeof size || !Number.isFinite(size) || size < 0) return '';
        if (size < 1024) return `${size} B`;
        const units = [
            'KB',
            'MB',
            'GB',
            'TB'
        ];
        let value = size / 1024;
        let unitIndex = 0;
        while(value >= 1024 && unitIndex < units.length - 1){
            value /= 1024;
            unitIndex += 1;
        }
        return `${value.toFixed(2)} ${units[unitIndex]}`;
    }
    renderAttachmentRemove(file) {
        const removeIcon = this.resolveIcon('removeIcon', renderUploadRemoveIcon());
        if (this.getDerivedDisabledState() || null === removeIcon) return nothing;
        return html`
      <span
        class="ve-upload-attachment-remove"
        role="button"
        tabindex="0"
        aria-label="remove"
        @click=${(event)=>{
            this.stopListItemActionEvent(event);
            this.handleRemoveAction(file);
        }}
        @keydown=${(event)=>{
            this.stopListItemActionEvent(event);
            this.handleKeyboardActivate(event, ()=>{
                this.handleRemoveAction(file);
            });
        }}
      >
        ${removeIcon}
      </span>
    `;
    }
    renderAttachmentFileIcon(file, size) {
        const fileIcon = this.resolveIcon('fileIcon', this.resolveDefaultFileIcon(file, 'row' === size ? 'var(--ve-upload-attachment-row-icon-size)' : 'var(--ve-upload-attachment-tag-icon-size)'));
        if (null === fileIcon) return nothing;
        return html`<span class="ve-upload-attachment-file-icon">${fileIcon}</span>`;
    }
    renderAttachmentThumb(file, size) {
        const iconClass = 'row' === size ? 've-upload-attachment-thumb ve-upload-attachment-thumb-row' : 've-upload-attachment-thumb ve-upload-attachment-thumb-tag';
        return html`
      <span class="${iconClass} ve-upload-attachment-thumb-file">
        ${this.renderAttachmentFileIcon(file, size)}
      </span>
    `;
    }
    renderAttachmentMeta(file) {
        const status = file.status ?? STATUS.init;
        const errorIcon = this.resolveIcon('errorIcon', renderMappedUploadIcon('exclamation-circle-fill'));
        if (status === STATUS.fail) return html`
        <span class="ve-upload-attachment-meta ve-upload-attachment-meta-error">
          ${null === errorIcon ? nothing : html`<span class="ve-upload-attachment-error-icon"
                >${errorIcon}</span
              >`}
          <span>上传失败</span>
        </span>
      `;
        if (status === STATUS.uploading) return html`
        <span class="ve-upload-attachment-meta">
          <span>上传中...</span>
        </span>
      `;
        const extension = this.resolveFileExtension(file).toLowerCase();
        const fileSizeText = this.resolveFileSizeText(file);
        return html`
      <span class="ve-upload-attachment-meta">
        <span>${extension}</span>
        ${fileSizeText ? html`<span>${fileSizeText}</span>` : nothing}
      </span>
    `;
    }
    resolveAttachmentProgressPercent(file) {
        const percent = Number(file.percent ?? 0);
        if (!Number.isFinite(percent)) return 0;
        return Math.max(0, Math.min(100, percent));
    }
    renderAttachmentProgressFill(file) {
        const percent = this.resolveAttachmentProgressPercent(file);
        return html`
      <span
        class="ve-upload-attachment-progress-fill"
        style="--ve-upload-attachment-progress-size: ${percent}%;"
      ></span>
    `;
    }
    renderAttachmentTitle(file) {
        const fileName = this.resolveFileName(file);
        const defaultFileName = this.resolveDefaultFileName(file);
        if (this.hasCustomFileNameRenderer()) return html`<span class="ve-upload-attachment-title-text">${fileName}</span>`;
        return html`
      <span class="ve-upload-attachment-title-text">${defaultFileName}</span>
    `;
    }
    renderAttachmentCardSkeleton() {
        return html`
      <div
        class="ve-upload-list-item ve-upload-attachment-card ve-upload-attachment-loading"
      >
        <div class="ve-upload-attachment-skeleton-lines">
          <span></span>
          <span></span>
        </div>
        <span class="ve-upload-attachment-skeleton-meta"></span>
      </div>
    `;
    }
    renderAttachmentCardItem(file) {
        const status = file.status ?? STATUS.init;
        const kind = this.resolveAttachmentKind(file);
        const fileUrl = this.getFileUrl(file);
        const defaultFileName = this.resolveDefaultFileName(file);
        const extension = this.resolveFileExtension(file);
        const iconConfig = this.getCustomIconConfig();
        if (status === STATUS.uploading) return this.renderAttachmentCardSkeleton();
        if (('image' === kind || 'video' === kind) && fileUrl) return html`
        <div
          class=${classMap({
            've-upload-list-item': true,
            've-upload-attachment-card': true,
            've-upload-attachment-card-media': true,
            [`ve-upload-list-item-${status}`]: true
        })}
          role="button"
          tabindex="0"
          @click=${()=>{
            this.handleItemClick(file);
        }}
          @keydown=${(event)=>{
            this.handleItemKeyDown(event, file);
        }}
        >
          <div class="ve-upload-attachment-media">
            ${isFunction(iconConfig.imageRender) ? iconConfig.imageRender(file) : html`<img
                  src=${fileUrl}
                  alt=${ifDefined(defaultFileName || void 0)}
                />`}
          </div>
          ${status === STATUS.fail ? html`<div class="ve-upload-attachment-card-error">
                ${this.renderAttachmentMeta(file)}
              </div>` : nothing}
          ${this.renderAttachmentRemove(file)}
        </div>
      `;
        return html`
      <div
        class=${classMap({
            've-upload-list-item': true,
            've-upload-attachment-card': true,
            've-upload-attachment-card-file': true,
            [`ve-upload-list-item-${status}`]: true
        })}
        role="button"
        tabindex="0"
        @click=${()=>{
            this.handleItemClick(file);
        }}
        @keydown=${(event)=>{
            this.handleItemKeyDown(event, file);
        }}
      >
        <div class="ve-upload-attachment-card-title">
          ${this.renderAttachmentTitle(file)}
        </div>
        <div class="ve-upload-attachment-card-footer">
          ${status === STATUS.fail ? this.renderAttachmentMeta(file) : html`<span class="ve-upload-attachment-card-extension"
                >${extension}</span
              >`}
        </div>
        ${this.renderAttachmentRemove(file)}
      </div>
    `;
    }
    renderAttachmentRowItem(file) {
        const status = file.status ?? STATUS.init;
        return html`
      <div
        class=${classMap({
            've-upload-list-item': true,
            've-upload-attachment-row': true,
            've-upload-attachment-loading': status === STATUS.uploading,
            [`ve-upload-list-item-${status}`]: true
        })}
        role="button"
        tabindex="0"
        @click=${()=>{
            this.handleItemClick(file);
        }}
        @keydown=${(event)=>{
            this.handleItemKeyDown(event, file);
        }}
      >
        ${status === STATUS.uploading ? this.renderAttachmentProgressFill(file) : nothing}
        ${this.renderAttachmentThumb(file, 'row')}
        <div class="ve-upload-attachment-row-content">
          <div class="ve-upload-attachment-row-title">
            ${this.renderAttachmentTitle(file)}
          </div>
          ${this.renderAttachmentMeta(file)}
        </div>
        ${this.renderAttachmentRemove(file)}
      </div>
    `;
    }
    renderAttachmentTagItem(file) {
        const status = file.status ?? STATUS.init;
        return html`
      <div
        class=${classMap({
            've-upload-list-item': true,
            've-upload-attachment-tag': true,
            've-upload-attachment-loading': status === STATUS.uploading,
            [`ve-upload-list-item-${status}`]: true
        })}
        role="button"
        tabindex="0"
        @click=${()=>{
            this.handleItemClick(file);
        }}
        @keydown=${(event)=>{
            this.handleItemKeyDown(event, file);
        }}
      >
        ${status === STATUS.uploading ? this.renderAttachmentProgressFill(file) : nothing}
        ${this.renderAttachmentThumb(file, 'tag')}
        <span class="ve-upload-attachment-tag-title">
          ${this.renderAttachmentTitle(file)}
        </span>
        ${status === STATUS.fail ? this.renderAttachmentMeta(file) : nothing}
        ${status === STATUS.uploading ? nothing : this.renderAttachmentRemove(file)}
      </div>
    `;
    }
    renderAttachmentItem(file) {
        if ('card' === this.listType) return this.renderAttachmentCardItem(file);
        if ('tag' === this.listType) return this.renderAttachmentTagItem(file);
        return this.renderAttachmentRowItem(file);
    }
    resolveFileByUid(uid) {
        return this.getMergedFileList().find((file)=>file.uid === uid);
    }
    getUploadListRenderProps(fileList) {
        const resolveCurrentFile = (file)=>this.resolveFileByUid(file.uid) ?? file;
        return {
            imagePreview: this.imagePreview,
            listType: this.listType,
            fileList: this.cloneFileList(fileList),
            showUploadList: this.showUploadList,
            ...this.progressProps ? {
                progressProps: this.progressProps
            } : {},
            onUpload: (file)=>{
                this.submit(resolveCurrentFile(file));
            },
            onAbort: (file)=>{
                this.abort(resolveCurrentFile(file));
            },
            onRemove: (file)=>{
                this.handleRemoveAction(resolveCurrentFile(file));
            },
            onReupload: (file)=>{
                this.reupload(resolveCurrentFile(file));
            },
            onPreview: (file)=>{
                this.handlePreviewAction(resolveCurrentFile(file));
            },
            disabled: this.getDerivedDisabledState()
        };
    }
    renderUploadListContent(fileList) {
        if (!this.getShowUploadList()) return nothing;
        if (0 === fileList.length) return nothing;
        const clonedFileList = this.cloneFileList(fileList);
        if (isFunction(this.renderUploadList)) {
            const customUploadList = this.renderUploadList(clonedFileList, this.getUploadListRenderProps(fileList));
            return html`
        <div
          class=${classMap({
                've-upload-list': true,
                've-upload-list-rtl': this.isRtl()
            })}
          part="file-list"
        >
          ${customUploadList}
        </div>
      `;
        }
        const renderedItems = fileList.map((file)=>{
            const originNode = this.renderAttachmentItem(file);
            if (!isFunction(this.renderUploadItem)) return originNode;
            const customUploadItem = this.renderUploadItem(originNode, this.cloneFileItem(file), clonedFileList);
            return customUploadItem ?? originNode;
        });
        return html`
      <div
        class=${classMap({
            've-upload-list': true,
            [`ve-upload-list-type-${this.listType}`]: true,
            've-upload-list-rtl': this.isRtl()
        })}
        part="file-list"
      >
        ${renderedItems}
      </div>
    `;
    }
    renderTipSlot() {
        if (this.hasTipSlot) return html`<slot
        name="tip"
        @slotchange=${this.handleTipSlotChange}
      ></slot>`;
        if ((this.tip ?? '').trim().length > 0) return html`${this.tip}`;
        return nothing;
    }
    renderTriggerContent() {
        const disabledUploadDom = this.getDerivedDisabledState();
        if (this.hasTriggerSlot) return html`
        <div
          class=${classMap({
            've-upload-trigger-custom-active': this.dragActive
        })}
        >
          <slot
            name="trigger"
            @slotchange=${this.handleTriggerSlotChange}
          ></slot>
        </div>
      `;
        if ('card' === this.listType) return html`
        <div class="ve-upload-trigger-picture-wrapper">
          <div
            class="ve-upload-trigger-picture"
            aria-label="upload"
            tabindex=${ifDefined(disabledUploadDom ? '-1' : '0')}
          >
            <div class="ve-upload-trigger-picture-text">
              ${renderMappedUploadIcon('plus')}
            </div>
          </div>
        </div>
      `;
        if (this.drag) return html`
        <div
          class=${classMap({
            've-upload-trigger-drag': true,
            've-upload-trigger-drag-active': this.dragActive
        })}
          aria-label="drag"
          tabindex=${ifDefined(disabledUploadDom ? '-1' : '0')}
        >
          <span class="ve-upload-trigger-drag-icon">
            ${renderMappedUploadIcon('upload', 'var(--ve-upload-drag-size-icon)')}
          </span>
          <p class="ve-upload-trigger-drag-text">
            ${this.dragActive ? 'Drop files here' : 'Click or drag file to this area to upload'}
          </p>
          ${(this.tip ?? '').trim().length > 0 ? html`<div class="ve-upload-trigger-tip">${this.tip}</div>` : nothing}
        </div>
      `;
        return html`
      <ve-button
        class="ve-upload-trigger-with-icon"
        type="primary"
        ?disabled=${disabledUploadDom}
      >
        ${renderMappedUploadIcon('upload')}
        Upload
      </ve-button>
    `;
    }
    renderUploadTrigger() {
        const disabledUploadDom = this.getDerivedDisabledState();
        const hideTrigger = this.shouldHideTrigger();
        return html`
      <div
        class=${classMap({
            've-upload': true,
            [`ve-upload-type-${this.listType}`]: true,
            've-upload-drag': this.drag,
            've-upload-disabled': disabledUploadDom,
            've-upload-hide': hideTrigger,
            've-upload-rtl': this.isRtl()
        })}
      >
        <div
          class="ve-upload-trigger"
          role="button"
          tabindex=${ifDefined(disabledUploadDom ? '-1' : '0')}
          aria-disabled=${disabledUploadDom ? 'true' : 'false'}
          @click=${this.handleTriggerClick}
          @keydown=${this.handleTriggerKeyDown}
          @dragenter=${this.handleDropDragEnter}
          @dragleave=${this.handleDropDragLeave}
          @dragover=${this.handleDropDragOver}
          @drop=${this.handleDrop}
        >
          ${this.renderTriggerContent()}
        </div>

        ${(this.tip ?? '').trim().length > 0 && !this.drag && 'card' !== this.listType ? html`
              <div class="ve-upload-trigger-tip" part="tip">
                ${this.renderTipSlot()}
              </div>
            ` : this.hasTipSlot && !this.drag && 'card' !== this.listType ? html`
                <div class="ve-upload-trigger-tip" part="tip">
                  ${this.renderTipSlot()}
                </div>
              ` : nothing}
      </div>
    `;
    }
    renderPictureCardTip() {
        if ('card' !== this.listType) return nothing;
        if (!this.hasTipSlot && 0 === (this.tip ?? '').trim().length) return nothing;
        return html`
      <div class="ve-upload-trigger-tip" part="tip">
        ${this.renderTipSlot()}
      </div>
    `;
    }
    renderPreview(fileList) {
        if (!this.imagePreview || 'card' !== this.listType || !this.previewVisible) return nothing;
        const previewFileList = this.getPreviewFileList(fileList);
        if (0 === previewFileList.length) return nothing;
        const currentFile = previewFileList[this.previewCurrent] ?? previewFileList[previewFileList.length - 1];
        if (!currentFile) return nothing;
        const currentImageUrl = this.getFileUrl(currentFile);
        if (!currentImageUrl) return nothing;
        return html`
      <div
        class="ve-upload-preview"
        @click=${(event)=>{
            const target = event.target;
            if (target instanceof HTMLElement && target.classList.contains('ve-upload-preview')) this.closePreview();
        }}
      >
        <button
          class="ve-upload-preview-close"
          type="button"
          @click=${(event)=>{
            event.stopPropagation();
            this.closePreview();
        }}
        >
          ${renderMappedUploadIcon('close')}
        </button>
        ${previewFileList.length > 1 ? html`
              <button
                class="ve-upload-preview-prev"
                type="button"
                @click=${(event)=>{
            event.stopPropagation();
            this.showPreviewByOffset(-1);
        }}
              >
                ‹
              </button>
              <button
                class="ve-upload-preview-next"
                type="button"
                @click=${(event)=>{
            event.stopPropagation();
            this.showPreviewByOffset(1);
        }}
              >
                ›
              </button>
            ` : nothing}
        <img
          src=${currentImageUrl}
          alt=${ifDefined(this.resolveDefaultFileName(currentFile))}
        />
      </div>
    `;
    }
    render() {
        const fileList = this.getMergedFileList();
        return html`
      <div class="ve-upload-root">
        ${'card' === this.listType ? html`
              <div class="ve-upload-card-stack">
                ${this.renderUploadListContent(fileList)}
                ${this.renderUploadTrigger()}
              </div>
            ` : html`
              ${this.renderUploadTrigger()}
              ${this.renderUploadListContent(fileList)}
            `}
        ${this.renderPictureCardTip()}

        <input
          class="ve-upload-trigger-input"
          type="file"
          accept=${ifDefined(this.resolveAcceptValue() || void 0)}
          ?disabled=${this.getDerivedDisabledState()}
          ?multiple=${this.multiple}
          ?webkitdirectory=${this.directory}
          ?directory=${this.directory}
          @change=${this.handleInputChange}
        />

        ${this.renderPreview(fileList)}
      </div>
    `;
    }
}
defineCustomElement(VE_UPLOAD_TAG, VeUploadElement);
export { VE_UPLOAD_TAG, VeUploadElement };
