// @ve-design/web/web/dist/adapter/garfish.js
const GARFISH_PROXY_NODE = Symbol.for('garfish.proxyNode');
const GARFISH_PATCHED_CREATE_TREE_WALKER = Symbol.for('garfish.patchedCreateTreeWalker');
const GARFISH_PATCHED_CURRENT_NODE = Symbol.for('garfish.patchedTreeWalkerCurrentNode');
const PATCHED_CREATE_TREE_WALKER = Symbol.for('@ve-design/web/adapter/garfish/createTreeWalker');
const PATCHED_CURRENT_NODE = Symbol.for('@ve-design/web/adapter/garfish/currentNode');
const SHOW_ALL = 0xffffffff;
const SHOW_ELEMENT = 0x1;
const SHOW_ATTRIBUTE = 0x2;
const SHOW_TEXT = 0x4;
const SHOW_CDATA_SECTION = 0x8;
const SHOW_PROCESSING_INSTRUCTION = 0x40;
const SHOW_COMMENT = 0x80;
const SHOW_DOCUMENT = 0x100;
const SHOW_DOCUMENT_TYPE = 0x200;
const SHOW_DOCUMENT_FRAGMENT = 0x400;
const LIT_TEMPLATE_WALKER_MASK = SHOW_ELEMENT | SHOW_COMMENT;
const FILTER_ACCEPT = 1;
const ELEMENT_NODE = 1;
const ATTRIBUTE_NODE = 2;
const TEXT_NODE = 3;
const CDATA_SECTION_NODE = 4;
const PROCESSING_INSTRUCTION_NODE = 7;
const COMMENT_NODE = 8;
const DOCUMENT_NODE = 9;
const DOCUMENT_TYPE_NODE = 10;
const DOCUMENT_FRAGMENT_NODE = 11;
function isObject(value) {
    return ('object' == typeof value || 'function' == typeof value) && !!value;
}
function readProperty(value, key) {
    try {
        return isObject(value) ? value[key] : void 0;
    } catch  {
        return;
    }
}
function isMarked(value, marker) {
    return true === readProperty(value, marker);
}
function mark(value, marker) {
    try {
        Object.defineProperty(value, marker, {
            configurable: false,
            enumerable: false,
            value: true
        });
    } catch  {}
}
function hasGarfishProxyTarget(value) {
    return isObject(value) && void 0 !== readProperty(value, GARFISH_PROXY_NODE);
}
function hasGarfishRuntime() {
    return true === readProperty(globalThis, '__GARFISH__') || void 0 !== readProperty(globalThis, 'Garfish');
}
function unwrapGarfishProxy(value) {
    let current = value;
    const seen = [];
    while(current && 'object' == typeof current && seen.indexOf(current) < 0){
        seen.push(current);
        const next = readProperty(current, GARFISH_PROXY_NODE);
        if (!next || 'object' != typeof next) return current;
        current = next;
    }
    return value;
}
function getShowBit(node1) {
    switch(node1.nodeType){
        case ELEMENT_NODE:
            return SHOW_ELEMENT;
        case ATTRIBUTE_NODE:
            return SHOW_ATTRIBUTE;
        case TEXT_NODE:
            return SHOW_TEXT;
        case CDATA_SECTION_NODE:
            return SHOW_CDATA_SECTION;
        case PROCESSING_INSTRUCTION_NODE:
            return SHOW_PROCESSING_INSTRUCTION;
        case COMMENT_NODE:
            return SHOW_COMMENT;
        case DOCUMENT_NODE:
            return SHOW_DOCUMENT;
        case DOCUMENT_TYPE_NODE:
            return SHOW_DOCUMENT_TYPE;
        case DOCUMENT_FRAGMENT_NODE:
            return SHOW_DOCUMENT_FRAGMENT;
        default:
            return 0;
    }
}
function filterAccepts(node1, filter) {
    if (!filter) return true;
    const result = 'function' == typeof filter ? filter(node1) : filter.acceptNode(node1);
    return result === FILTER_ACCEPT;
}
function createSafeTreeWalker(root, whatToShow = SHOW_ALL, filter = null) {
    let traversalRoot = root;
    let currentNode = root;
    const nextInDocumentOrder = (node1)=>{
        if (node1.firstChild) return node1.firstChild;
        let cursor = node1;
        while(cursor && cursor !== traversalRoot){
            if (cursor.nextSibling) return cursor.nextSibling;
            cursor = cursor.parentNode;
        }
        return null;
    };
    return {
        get root () {
            return traversalRoot;
        },
        get whatToShow () {
            return whatToShow;
        },
        get filter () {
            return filter;
        },
        get currentNode () {
            return currentNode;
        },
        set currentNode (node){
            const realNode = unwrapGarfishProxy(node);
            traversalRoot = realNode;
            currentNode = realNode;
        },
        nextNode () {
            let next = nextInDocumentOrder(currentNode);
            while(next){
                currentNode = next;
                if ((whatToShow & getShowBit(next)) !== 0 && filterAccepts(next, filter)) return next;
                next = nextInDocumentOrder(next);
            }
            return null;
        }
    };
}
function isLitTemplateWalkerMask(whatToShow) {
    return (whatToShow & LIT_TEMPLATE_WALKER_MASK) === LIT_TEMPLATE_WALKER_MASK;
}
function patchTreeWalkerCurrentNode(nativeCreateTreeWalker, realDocument) {
    const sampleWalker = (()=>{
        try {
            return nativeCreateTreeWalker.call(realDocument, realDocument, SHOW_ALL);
        } catch  {
            return;
        }
    })();
    const treeWalkerPrototype = sampleWalker ? Object.getPrototypeOf(sampleWalker) : globalThis.TreeWalker?.prototype;
    if (!treeWalkerPrototype) return false;
    if (isMarked(treeWalkerPrototype, PATCHED_CURRENT_NODE) || isMarked(treeWalkerPrototype, GARFISH_PATCHED_CURRENT_NODE)) return true;
    const descriptor = Object.getOwnPropertyDescriptor(treeWalkerPrototype, 'currentNode');
    if (!descriptor?.set) return false;
    const originalSetter = descriptor.set;
    Object.defineProperty(treeWalkerPrototype, 'currentNode', {
        ...descriptor,
        set (value) {
            originalSetter.call(this, unwrapGarfishProxy(value));
        }
    });
    mark(treeWalkerPrototype, PATCHED_CURRENT_NODE);
    mark(treeWalkerPrototype, GARFISH_PATCHED_CURRENT_NODE);
    return true;
}
function shouldFallbackForLitTemplateWalker(nativeCreateTreeWalker, documentLike) {
    try {
        const template = documentLike.createElement('template');
        template.innerHTML = '<div></div>';
        const fragment = documentLike.importNode(template.content, true);
        const walker = nativeCreateTreeWalker.call(unwrapGarfishProxy(documentLike), unwrapGarfishProxy(documentLike), LIT_TEMPLATE_WALKER_MASK);
        walker.currentNode = fragment;
        return false;
    } catch  {
        return true;
    }
}
function installDocumentCreateTreeWalkerPatch(documentLike) {
    const realDocument = unwrapGarfishProxy(documentLike);
    const nativeCreateTreeWalker = realDocument.createTreeWalker ?? documentLike.createTreeWalker;
    if ('function' != typeof nativeCreateTreeWalker) return false;
    if (isMarked(nativeCreateTreeWalker, PATCHED_CREATE_TREE_WALKER) || isMarked(nativeCreateTreeWalker, GARFISH_PATCHED_CREATE_TREE_WALKER)) return true;
    const hasProxyDocument = realDocument !== documentLike || hasGarfishProxyTarget(documentLike);
    const shouldPatchForGarfish = hasProxyDocument || hasGarfishRuntime();
    if (!shouldPatchForGarfish) return false;
    const currentNodePatchApplied = patchTreeWalkerCurrentNode(nativeCreateTreeWalker, realDocument);
    const needsSafeLitWalker = !currentNodePatchApplied && (hasProxyDocument || shouldFallbackForLitTemplateWalker(nativeCreateTreeWalker, documentLike));
    if (!currentNodePatchApplied && !needsSafeLitWalker) return false;
    const patchedCreateTreeWalker = function(...args) {
        const [root, whatToShow, filter] = args;
        const realRoot = unwrapGarfishProxy(root);
        const resolvedWhatToShow = whatToShow ?? SHOW_ALL;
        const resolvedFilter = filter ?? null;
        if (needsSafeLitWalker && isLitTemplateWalkerMask(resolvedWhatToShow)) return createSafeTreeWalker(realRoot, resolvedWhatToShow, resolvedFilter);
        try {
            return nativeCreateTreeWalker.call(unwrapGarfishProxy(this), realRoot, resolvedWhatToShow, resolvedFilter);
        } catch (error) {
            if (!isLitTemplateWalkerMask(resolvedWhatToShow)) throw error;
            return createSafeTreeWalker(realRoot, resolvedWhatToShow, resolvedFilter);
        }
    };
    mark(patchedCreateTreeWalker, PATCHED_CREATE_TREE_WALKER);
    mark(patchedCreateTreeWalker, GARFISH_PATCHED_CREATE_TREE_WALKER);
    setCreateTreeWalker(documentLike, patchedCreateTreeWalker);
    if (realDocument !== documentLike) setCreateTreeWalker(realDocument, patchedCreateTreeWalker);
    if (globalThis.Document?.prototype.createTreeWalker === nativeCreateTreeWalker) setCreateTreeWalker(globalThis.Document.prototype, patchedCreateTreeWalker);
    return true;
}
function setCreateTreeWalker(target, value) {
    try {
        Object.defineProperty(target, 'createTreeWalker', {
            configurable: true,
            writable: true,
            value
        });
        return true;
    } catch  {
        try {
            target.createTreeWalker = value;
            return target.createTreeWalker === value;
        } catch  {
            return false;
        }
    }
}
function installGarfishAdapter() {
    if ("u" < typeof document) return false;
    return installDocumentCreateTreeWalkerPatch(document);
}
const garfishAdapterInstalled = installGarfishAdapter();
export { garfishAdapterInstalled, installGarfishAdapter };
